{
 "metadata": {
  "name": "",
  "signature": "sha256:1e99c4c7405635e87f7443fec6d847141e1e5e1ed4a2845f6c649c220719072c"
 },
 "nbformat": 3,
 "nbformat_minor": 0,
 "worksheets": [
  {
   "cells": [
    {
     "cell_type": "code",
     "collapsed": false,
     "input": [
      "import os\n",
      "import pandas as pd\n",
      "import numpy as np\n",
      "import scipy\n",
      "import pickle\n",
      "from numpy import *\n",
      "import matplotlib as mpl\n",
      "from matplotlib import pyplot as plt\n",
      "eu = os.path.expanduser\n",
      "jn = os.path.join\n",
      "meta_dir = eu(\"~/vervet_project/metadata\")\n",
      "admix_ana_dir = eu(\"~/vervet_project/analyses/20140902_admixture/_data\")\n",
      "snapp_ana_dir = eu(\"~/vervet_project/analyses/20140909_snapp/_data\")\n",
      "meta_df = pd.read_csv(jn(meta_dir,\"163_population_ucla_id_taxon.csv\"),index_col=0)\n",
      "meta_df[\"index\"] = meta_df.index\n",
      "if float(\".\".join(pd.__version__.split(\".\")[0:2])) >= 0.14:\n",
      "    meta_df.drop_duplicates(subset='index', take_last=True, inplace=True)\n",
      "else:\n",
      "    meta_df.drop_duplicates(cols='index', take_last=True, inplace=True)\n",
      "del meta_df[\"index\"]\n",
      "ucla_ids = meta_df.index.unique()"
     ],
     "language": "python",
     "metadata": {},
     "outputs": [],
     "prompt_number": 9
    },
    {
     "cell_type": "code",
     "collapsed": false,
     "input": [
      "ped_fn = jn(admix_ana_dir,\"163_f2_maf0p05_CAE28_t_ld_prune_0p1_012.ped\")\n",
      "nexus_fn = jn(snapp_ana_dir,\"vervet_CAE28_0p01.nexus\")"
     ],
     "language": "python",
     "metadata": {},
     "outputs": [],
     "prompt_number": 21
    },
    {
     "cell_type": "code",
     "collapsed": false,
     "input": [
      "\" \"*5"
     ],
     "language": "python",
     "metadata": {},
     "outputs": [
      {
       "metadata": {},
       "output_type": "pyout",
       "prompt_number": 26,
       "text": [
        "'     '"
       ]
      }
     ],
     "prompt_number": 26
    },
    {
     "cell_type": "code",
     "collapsed": false,
     "input": [
      "def ped_to_nexus(ped_fn, nexus_fn, individual_names=None, n_individuals=None,\n",
      "                                 max_name_len = 12, comments=None,dominant=False):\n",
      "    \"\"\"\n",
      "    Write snps from ped file to nexus alignment file.\n",
      "    Attention: The bed file has to be encoded with:\n",
      "                0.. missing\n",
      "                1.. first allele \n",
      "                2.. second allele\n",
      "            Such a ped file can be obtained from a regular ped file\n",
      "            with \"plink --file <in_filename> --recode12 --out <out_filename>\"\n",
      "    This function assumes that individuals are diploid. \n",
      "    The two haploid columns in the ped become a single genotype (0/1/2) \n",
      "        in the nexus file.\n",
      "    -------Parameters---------\n",
      "    individual_names... list of individual (or taxa) names\n",
      "                        if not supplied, the names from the bed file are used\n",
      "    n_individuals... number of individuals (ntax in the nexus format)\n",
      "                     (if not supplied, the length of indiviudal_names is used)\n",
      "    max_name_len... maximal number of characters of individual names, \n",
      "                    used to determine intent of alignments\n",
      "    comments... list of strings that are added to the nexus header as comments\n",
      "                alternatively, it can be a single string\n",
      "    dominant... if True, the 0 allele is assumed to be dominant and the resulting\n",
      "                nexus file is encoded 0/1 instead of 0/1/2\n",
      "    \"\"\"\n",
      "     #\n",
      "    assert individual_names is not None or n_individuals is not None,\\\n",
      "            \"specify either individual names or the total number of individuals\" \n",
      "    if individual_names is not None and n_individuals is not None:\n",
      "        assert len(individual_names) == n_individuals,\\\n",
      "                \"number of individual names must equal argument n_individuals\"\n",
      "    if n_individuals is None:\n",
      "        n_individuals = len(individual_names)\n",
      "    if individual_names is None:\n",
      "        individual_names = [None]*n_individuals\n",
      "        \n",
      "    with open(ped_fn,\"r\") as bf, open(nexus_fn,\"w\") as nf:\n",
      "        nf.write(\"#NEXUS\\n\")\n",
      "        if comments is not None:\n",
      "            if not isinstance(comments, basestring):\n",
      "                for comment in comments:\n",
      "                    nf.write(\"[{}]\\n\".format(comment))\n",
      "            else:\n",
      "                nf.write(\"[{}]\\n\".format(comments))\n",
      "        \n",
      "        for i,(line,name) in enumerate(zip(bf,individual_names)):\n",
      "            if i >= n_individuals:\n",
      "                break\n",
      "            fields = line.split()\n",
      "            if name is None:\n",
      "                name = fields[0]\n",
      "            n_whitespace = max_name_len + 1 - len(name)\n",
      "            assert n_whitespace >= 1, \"max taxon name length is \" \\\n",
      "                                    + str(max_name_len) + \" but name is \" + name\n",
      "            data = fields[6:]\n",
      "            \n",
      "            if i == 0:\n",
      "                if dominant:\n",
      "                    symbols = \"01\"\n",
      "                else:\n",
      "                    symbols = \"012\"\n",
      "                nf.write(\"\\nBegin data;\\n\"\n",
      "                 \"        Dimensions ntax={} nchar={};\\n\"\n",
      "                 '        Format datatype=integerdata symbols=\"{}\" missing=? gap=-;\\n'\n",
      "                 \"        Matrix\\n\".format(n_individuals,len(data)/2,symbols))\n",
      "            nf.write(name)\n",
      "            nf.write(\" \" * n_whitespace)\n",
      "            for i in xrange(0,len(data),2):\n",
      "                h0 = int(data[i])\n",
      "                h1 = int(data[i+1])\n",
      "                if h0!=0 and h1!=0:\n",
      "                    gt = h0 + h1 - 2\n",
      "                    if dominant:\n",
      "                        #assume 0 allele is dominant\n",
      "                        gt = int(gt/2)\n",
      "                    gt = str(gt)\n",
      "                else:\n",
      "                    gt = \"?\"\n",
      "                nf.write(gt)\n",
      "            nf.write(\"\\n\")\n",
      "        nf.write(\"        ;\\nEnd;\")\n",
      "                \n",
      "    "
     ],
     "language": "python",
     "metadata": {},
     "outputs": [],
     "prompt_number": 51
    },
    {
     "cell_type": "code",
     "collapsed": false,
     "input": [
      "with open(ped_fn,\"r\") as pf:\n",
      "    names = []\n",
      "    for line in pf:\n",
      "        name = line.split()[0]\n",
      "        #print name\n",
      "        names.append(meta_df.ix[name][\"population\"]+ \"_\" +name)"
     ],
     "language": "python",
     "metadata": {},
     "outputs": [],
     "prompt_number": 49
    },
    {
     "cell_type": "code",
     "collapsed": false,
     "input": [
      "ped_to_nexus(ped_fn, nexus_fn, individual_names=names[:10], comments=\"SNPs from CAE28 filter1 LD pruning to r2=0.1\")"
     ],
     "language": "python",
     "metadata": {},
     "outputs": [],
     "prompt_number": 52
    },
    {
     "cell_type": "code",
     "collapsed": false,
     "input": [
      "#NEXUS\n",
      "[SNP data from Leache AD, Fujita MK, Minin VN, Bouckaert R. 2014. Species delimitation using genome-wide SNP data. Systematic Biology doi:10.1\n",
      "[Also available on Dryad: http://datadryad.org/resource/doi:10.5061/dryad.r55fb]\n",
      "\n",
      "Begin data;\n",
      "        Dimensions ntax=46 nchar=129;\n",
      "        Format datatype=integerdata symbols=\"012\"  gap=-;\n",
      "        Matrix\n",
      "eng_NG_1        2-0202-22120211220222-10212-01222022202202202000220122020220220022-20220202020202121022200021100-022200220202222-0202222200220\n",
      "coal_CA1_2      12020212202020222022222012211022-02221210210201022002202021020002122222010202020202202220020220020122012202022212020222020-200\n",
      "coal_CA1_3      220202022020102210222220222110222022202202202000220022020220220022221220202020202122002200202210100220122020222220202210200220\n",
      "coal_CA1_4      12020202202010221022222022-11022202-202202212010-2002202021022002122-22020-020202122022200202210-01220122020222120202210201220\n",
      "coal_CA1_5      120202022020102220222220222110222022202202202000220022120220220020221100212020202122012200202200002220122020222120202210200220\n",
      "coal_CG_6       220202022020202220222222222121222022212202202000220022120220220020220210201020202122022200202220202220122010222220202220200220\n",
      "kya_GH3_7       222222220220202220222020220000221022212222202220020222022222022202200220202222000202020022200002222222002020202202220222210220\n",
      "kya_GH3_8       222222220220202220222020220000222022202222202220020222022222022202200220202222000102020022200002222222002020202202220222210220\n",
      "kya_GH3_9       22222222022020222022202022000022202-202222202220020222022222021202200220202222000202020022200002222222002020202202221222210220\n",
      "fas_GH2_10      222222220220202-20222020220000212022202222202220020222022222022202200220202222000022020022200002222222001020202202220222210220\n",
      "fas_GH2_11      2222222202212022202220202200002-2022202222202220020222022222022202200220202222000122020022200002222222000020202202220222210220\n",
      "fas_GH1_12      222222220220202220222020220000212022202222202220020222022222022202200220202222000022020022201002222222000020202202220222200220\n",
      "fas_GH1_13      222222220220202-202220202200002-2022202222202220020222022222022202200220202222000122020022200002222222000020202202220222210220\n",
      "fas_GH1_14      222222220220202120222010220000212022202222202220020222022222022202200220202222000122020022200002222222000020202202220222200220\n",
      "eng_NG_15       2102021221202122202222102120002222212222022020002200220222202200221202002020200022220222001122002-1200022020222220202222200120\n",
      "fas_GH2_16      222222220220202220122020220000212022201222202220020222022222022202200220202222000022020022200002222222000020202202220222210220\n",
      "fas_GH2_17      2222222202212022202220202200002-2022202221212220020222022222022202200220202222000022020022200002222222000020202202220222210220\n",
      "eng_NG_18       210202212010212220222200022-012222221222021020022200220212-0220022020200202020102121022200-12200001200012020222220202222200220\n",
      "eng_CA2_20      22222222022020222022202022000022202220-222202220020212022222022202200220202222000202020022200002222222002020202202220222210220\n",
      "eng_EG_21       220102222020202220202-00222-00222222122202202001221022020220220022220200202020201122022200102200000200022-20222220202222200220\n",
      "eng_EG_22       220002222020222220212200222000222222222202202000211021021220220022220222202020200212022200002200001100022020221220202222100221\n",
      "eng_EG_23       220202022020222220211200222000222222222202202000221021022220220022220221202020202201221200002200001200002020221220202222200221\n",
      "eng_EG_24       220002122020212221222210222000222222122202202002220022020220220022220220202020202012022210112200002100022021122220202222200220\n",
      "eng_EG_25       220000122120222220222-10222-0022222212220220201022102102222022002222022120202020201202220010-200000100022120221220202122200220\n",
      "coal_GA_26      22020222202020222022200022202022200221-2022020202200220121202200102202202020202221220222002022002022-0022020222220102220200220\n",
      "kya_GH3_27      222222220220202220222020220000222022202222202220020222022222022202200220202222000202020022200002222222002020202202220222210220\n",
      "kya_GH3_28      22222222022020222022202022000022202220-222202220020222022222022202200220202222000202020022200002222222002020202202220222210220\n",
      "kya_GH3_29      222222220220202220222020220000222022202222202220020222022222022202200220202222000202020022200002222222002020202202220222210220\n",
      "kya_GH3_30      222222220220202220222020220000222022202222202220020222022222022202200220202222000202020022200002222222002020202202220222210220\n",
      "kya_GH3_31      2222222202202022202220202200001220222022222012200202220222220222022002202022220002020200222000022222220020202022022202222-0220\n",
      "fas_GH4_32      2222222202202022202220202200002-2022202222202220020222022222022202200220202222000012020022200002222222000020202202220222210220\n",
      "fas_GH4_33      2222222202212021202220202200002-2022202222202220020222022222022202200220202222000122020022200002222222000020202202220222210220\n",
      "fas_GH4_34      2222222202212021202220202200002-2022202222202220020222022222022202200220202222000022020022200002222222000020202202220222210220\n",
      "fas_GH4_35      2222222202202022202220202200002-20222022222022200202220222220222022002202022220000220200222000022222220000202022022202222-0220\n",
      "fas_GH4_36      2222222202212022202220202200002-2022202222202220020222022222022202200220202222000022020022200002222222000020202202220222210220\n",
      "fas_GH4_37      222222220220202220222020220000212022202222202220020222022222022202200220202222000022020022200002222222000020202202220222210220\n",
      "fas_GH4_38      2222222202212022202220202200002-2022202222202220020222022222022202200220202222000022020022200002222222000020202202220222210220\n",
      "fas_GH4_39      222222220221202-202220202200002-2022202222202220020222022222022202200220202212000022020022200002222222000020202202220222210220\n",
      "fas_GH4_40      222222220220202220222020220000212022202222202220020222022222022202200220202222000022020022200002222222000020202202220222210220\n",
      "kya_GH3_41      222222220220202220222020220000222022202222202220020222022222022202200220202222000202020022201002222222002020202202220222210220\n",
      "kya_GH3_42      222222220220202220222020220000222022202222202220020222022222022202200220202222000202020022200002222222002020202202220222210220\n",
      "kya_GH3_43      2222222202202022202220202200002220222022222022200202220222220222022002202022220002020200222000022222220020202022022202222-0220\n",
      "kya_GH3_44      222222220220202220222020220000222022202222202220020222022222022202200220202222000202020022201002222222002020202202220222210220\n",
      "kya_GH3_45      222222220220202220222020220000222022202222202220020222022222022202200220202222000202020022200002222222002020202202220222210220\n",
      "kya_GH3_46      222222220220202220222020220000222022202222202220020222022222022202200220202222000202020022200002222222002020202202220222210220\n",
      "kya_GH3_47      222222220220202220222020220000222022202222202220020222022222022202200220202222000202020022200002222222002020202202220222200220\n",
      "        ;\n",
      "End;"
     ],
     "language": "python",
     "metadata": {},
     "outputs": []
    },
    {
     "cell_type": "code",
     "collapsed": false,
     "input": [],
     "language": "python",
     "metadata": {},
     "outputs": []
    },
    {
     "cell_type": "code",
     "collapsed": false,
     "input": [],
     "language": "python",
     "metadata": {},
     "outputs": []
    }
   ],
   "metadata": {}
  }
 ]
}