{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 316,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "%matplotlib inline"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 318,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "#parallel\n",
    "from ipyparallel import Client\n",
    "rc = Client(profile=\"default\")\n",
    "dview = rc[:] # use all engines\n",
    "lv = rc.load_balanced_view()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 319,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "%%px --local\n",
    "import os,  ast\n",
    "import numpy as np\n",
    "import pandas as pd\n",
    "import vcf, pyfasta\n",
    "\n",
    "chromosomes = [\"CAE\" + str(i) for i in range(1,30) + [\"X\"] + ['Y'] ]\n",
    "eu = os.path.expanduser\n",
    "jn = os.path.join\n",
    "var_ana_dir = eu(\"~/vervet_project/analyses/20150504_163_UnifiedGenotyper/_data\")\n",
    "meta_dir = eu(\"~/vervet_project/metadata\")\n",
    "\n",
    "def str_to_dic(s):\n",
    "    try:\n",
    "        return ast.literal_eval(s)\n",
    "    except ValueError:\n",
    "        print s\n",
    "        raise\n",
    "        \n",
    "mt = pd.read_csv(jn(meta_dir,\"163_master_table.csv\"),\n",
    "                 index_col=0,\n",
    "                 converters={\"library\":str_to_dic, \"phenotypes\":str_to_dic})\n",
    "\n",
    "\n",
    "#gene_df_ensembl = pd.read_csv(jn(meta_dir,\"annot_ensemble_1.1.77_genes.tsv\"),sep=\"\\t\",index_col=[0,1])\n",
    "#gene_df_ensembl_CDS = pd.read_csv(jn(meta_dir,\"annot_ensemble_1.1.77_CDS.tsv\"),sep=\"\\t\",index_col=[0,1])\n",
    "mac_seq = pyfasta.Fasta(eu(\"~/vervet_project/data/other_species/Macaca_mulatta/CR_1.0_in_vervet_coord_ref3500.fa\"))\n",
    "#gene_df_r100 = pd.read_csv(jn(meta_dir,\"annot_release100_genes.tsv\"),sep=\"\\t\",index_col=[0,1])\n",
    "#gene_df_r100_exons = pd.read_csv(jn(meta_dir,\"annot_release100_exons.tsv\"),sep=\"\\t\",index_col=[0,1])\n",
    "#attention, upgrade to ensembl 1.1.78!!!\n",
    "#gene_df_ensembl = pd.read_csv(jn(meta_dir,\"annot_ensemble_1.1.77_genes.tsv\"),sep=\"\\t\",index_col=[0,1])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 320,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "%%px --local\n",
    "\n",
    "import re\n",
    "import subprocess, gzip\n",
    "\n",
    "def add_info_line(info_dic, line):\n",
    "    #split all commas except if in quotes\n",
    "    r = re.compile('[^,]+\".+\"[^,]*'\n",
    "                '|'\n",
    "                '[^,]+')\n",
    "    try:\n",
    "        category, value = line.strip().split('=<')\n",
    "    except ValueError:\n",
    "        try: \n",
    "            info_dic[\"uncategorised\"].update({line.strip().split('=')[0][2:]:line.strip().split('=')[1]})\n",
    "        except KeyError:\n",
    "            try:\n",
    "                info_dic.update({\"uncategorised\":{line.strip().split('=')[0][2:]:line.strip().split('=')[1]}})\n",
    "            except ValueError, e:\n",
    "                print line.strip().split('=')\n",
    "                raise e\n",
    "        return\n",
    "    category = category[2:]\n",
    "    tags = r.findall(value[:-1])\n",
    "    subdic = {k:v for k,v in [t.split('=',1) for t  in tags]}\n",
    "    try:\n",
    "        info_dic[category].update({subdic[\"ID\"]:subdic})\n",
    "    except KeyError:\n",
    "        info_dic.update({category:{subdic[\"ID\"]:subdic}})\n",
    "        \n",
    "def parse_vcf_header(fn):\n",
    "    info_dic = {}\n",
    "    fh = gzip.open(fn) if fn[-3:] == \".gz\" else open(fn)\n",
    "    for line in fh:\n",
    "        if line[:6] == \"#CHROM\":\n",
    "            header = line.strip().split('\\t')\n",
    "            header[0] = header[0][1:]\n",
    "        else:\n",
    "            add_info_line(info_dic, line)\n",
    "        if line[0] != '#':\n",
    "            break\n",
    "    return header, info_dic"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 321,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "%%px --local\n",
    "def get_vcf_df(fn, chrom, start, end, header=None, **kwa):\n",
    "    if header is None:\n",
    "        header, _ = parse_vcf_header(fn)\n",
    "    tabix_stream = subprocess.Popen(['tabix',fn,\"{}:{}-{}\".format(chrom,start,end)], stdout=subprocess.PIPE,stderr=subprocess.PIPE)\n",
    "    vcf_df = pd.read_csv(tabix_stream.stdout, sep=\"\\t\", index_col=[0,1], header=False, names = header, **kwa)\n",
    "    \n",
    "    return vcf_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Get SNPeff annotations for peak regions, including filtered sites and including macacaque differences"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "peaks = [(\"CAE1\",2467224),(\"CAE1\",2757224),(\"CAE1\",7005224),\n",
    "         (\"CAE2\",50381341),(\"CAE4\",5189544),(\"CAE5\",13975063),\n",
    "        (\"CAE5\",60269063),(\"CAE5\",74650063), (\"CAE6\",3823905),\n",
    "        (\"CAE6\",5564905),(\"CAE6\",11705905),(\"CAE6\",29586905),\n",
    "        (\"CAE8\",117925766),(\"CAE10\",104389305),\n",
    "        (\"CAE14\",17153077),(\"CAE14\",98746077),(\"CAE15\",62257003),\n",
    "        (\"CAE16\",1907271),(\"CAE16\",6963271),(\"CAE16\",7241271),(\"CAE16\",57481271),\n",
    "        (\"CAE16\",59802271),(\"CAE16\",68026271),(\"CAE17\",44458229),(\"CAE17\",54486229),\n",
    "        (\"CAE18\",62028538),(\"CAE20\",18933658),(\"CAE20\",106492658),(\"CAE20\",111605658),\n",
    "        (\"CAE22\",10431400),(\"CAE23\",250753),(\"CAE23\",42363534),(\"CAE24\",3453536),\n",
    "        (\"CAE24\",8944536),(\"CAE25\",61650817),(\"CAE26\",18910982),(\"CAE29\",18370875),\n",
    "        (\"CAE29\",21874875)]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 322,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "%%px --local\n",
    "def get_genotype_df(chrom, start,end):\n",
    "    vcf_fn = jn(var_ana_dir,\"163_UG_ref3500_all_sites_{}_filter_anc_mac_diff_ensembl_1.1.78.vcf.gz\".format(chrom))\n",
    "\n",
    "\n",
    "    header, info_dic = parse_vcf_header(vcf_fn)\n",
    "    samples = header[9:]\n",
    "\n",
    "    converters = {\"INFO\":get_info_dic}\n",
    "    converters.update({n:get_genotype for n in samples})#{i:get_genotype for t in range(9,len(header))}\n",
    "\n",
    "    vcf_df = get_vcf_df(vcf_fn, chrom, start, end, header=header, converters=converters)\n",
    "    \n",
    "    return vcf_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 323,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "%%px --local\n",
    "def get_genotype(s):\n",
    "    gt_str = s.split(':',1)[0]\n",
    "    if gt_str[:3] in [\"0/0\",\"0|0\"]:\n",
    "        return 0\n",
    "    elif gt_str[:3] in [\"1/1\",\"1|1\"]:\n",
    "        return 2\n",
    "    elif gt_str[:3] in [\"0/1\",\"0|1\",\"1|0\"]:\n",
    "        return 1\n",
    "    else:\n",
    "        return np.nan\n",
    "    \n",
    "def get_info_dic(s):\n",
    "    info_tuples = [t.split('=') for t in s.split(';')]\n",
    "    info_tuples = [t for t in info_tuples if len(t)==2]\n",
    "    tags = [t for t in info_tuples if len(t)!=2]\n",
    "#    try:\n",
    "    d = {k:v for (k,v) in info_tuples}\n",
    "    d.update({'_tags':tags})\n",
    "#    except ValueError:\n",
    "#        d = {}\n",
    "#        logging.warning(\"Can not parse info column at {}:{}\".format(line[0],line[1]))\n",
    "    return d\n",
    "\n",
    "def group_by_pop(col,poptype=\"pop3\"):\n",
    "    return mt.ix[col][poptype]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 324,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "%%px --local\n",
    "def try_dic(dic,key):\n",
    "    try:\n",
    "        return dic[key]\n",
    "    except KeyError:\n",
    "        return np.nan\n",
    "\n",
    "def get_ith_annotation(el,i):\n",
    "    try:\n",
    "        return el.split(',')[i]\n",
    "    except IndexError:\n",
    "        return np.nan\n",
    "def split_annotation(s):\n",
    "    return pd.Series(s.split('|')[:4]+s.split('|')[-1:],index=['allele','effect','impact','gene_symbol','errors'])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 341,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "%%px --local\n",
    "def get_variant_annot_df(vcf_df, sample_ids=None):\n",
    "    #remove non-seggregating\n",
    "    \n",
    "    if sample_ids is None:\n",
    "        sample_ids = vcf_df.columns.values[7:]\n",
    "    vcf_df = vcf_df[vcf_df['ALT'] != '.']\n",
    "\n",
    "    subpop_af_df = vcf_df[sample_ids].groupby(group_by_pop,axis=1).apply(lambda df: df.mean(axis=1)/2.)\n",
    "    annot_df = vcf_df[[\"REF\",\"ALT\",\"FILTER\"]]\n",
    "    annot_df.loc[:,'AA'] = vcf_df['INFO'].apply(lambda d:try_dic(d,\"AA\"))\n",
    "    annot_df.loc[:,'AF'] = vcf_df[sample_ids].mean(axis=1)/2.\n",
    "\n",
    "    annot_s = vcf_df['INFO'].apply(lambda d:try_dic(d,\"ANN\"))\n",
    "\n",
    "\n",
    "\n",
    "    #annotation lowest level\n",
    "    level0_df = annot_s.apply(lambda el:get_ith_annotation(el,0)).apply(split_annotation)\n",
    "    level0_df = pd.concat([annot_df,level0_df,subpop_af_df],axis=1)\n",
    "    level0_df.set_index(\"gene_symbol\",append=True,inplace=True)\n",
    "    tot_level_df = level0_df\n",
    "    i =1\n",
    "    for i in range(1,100):\n",
    "        level_info = annot_s.apply(lambda el:get_ith_annotation(el,i)).dropna()   \n",
    "        if not len(level_info):\n",
    "            break\n",
    "        level_df = level_info.apply(split_annotation)\n",
    "        level_df = annot_df.ix[level_df.index].join(level_df,how='left').join(subpop_af_df, how='left')\n",
    "        level_df.set_index(\"gene_symbol\",append=True,inplace=True)\n",
    "        tot_level_df = pd.concat([tot_level_df, level_df])\n",
    "    return tot_level_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 348,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "%%px --local\n",
    "def get_annot(chrom, pos, distance):\n",
    "    start = pos-distance\n",
    "    end = pos+distance\n",
    "    try:\n",
    "        vcf_df = get_genotype_df(chrom, start,end)\n",
    "        annot_df = get_variant_annot_df(vcf_df)\n",
    "    except Exception, e:\n",
    "        return str(e)\n",
    "    return annot_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 349,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "peak_map = lv.map_async(lambda p:get_annot(p[0],p[1],50000),peaks)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 351,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "peak_map.wait(10)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 353,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "['Too many columns specified: expected 172 and found -1']"
      ]
     },
     "execution_count": 353,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "[err for err in peak_map.result if type(err)==str]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 355,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "annot_f = pd.concat([r for r in peak_map.result if type(r)!=str])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 361,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div style=\"max-height:1000px;max-width:1500px;overflow:auto;\">\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th>REF</th>\n",
       "      <th>ALT</th>\n",
       "      <th>FILTER</th>\n",
       "      <th>AA</th>\n",
       "      <th>AF</th>\n",
       "      <th>allele</th>\n",
       "      <th>effect</th>\n",
       "      <th>impact</th>\n",
       "      <th>errors</th>\n",
       "      <th>aet</th>\n",
       "      <th>cyn</th>\n",
       "      <th>pyn</th>\n",
       "      <th>pys</th>\n",
       "      <th>sab</th>\n",
       "      <th>sac</th>\n",
       "      <th>sar</th>\n",
       "      <th>tan</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>CHROM</th>\n",
       "      <th>POS</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th rowspan=\"5\" valign=\"top\">CAE6</th>\n",
       "      <th>5576627</th>\n",
       "      <td> G</td>\n",
       "      <td> A</td>\n",
       "      <td>  PASS</td>\n",
       "      <td> NaN</td>\n",
       "      <td> 0.000000</td>\n",
       "      <td> A</td>\n",
       "      <td> missense_variant</td>\n",
       "      <td> MODERATE</td>\n",
       "      <td> </td>\n",
       "      <td> 0.000000</td>\n",
       "      <td> 0.00000</td>\n",
       "      <td> 0</td>\n",
       "      <td> 0</td>\n",
       "      <td> 0</td>\n",
       "      <td> 0</td>\n",
       "      <td> 0</td>\n",
       "      <td> 0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5576646</th>\n",
       "      <td> C</td>\n",
       "      <td> T</td>\n",
       "      <td> ExCov</td>\n",
       "      <td> NaN</td>\n",
       "      <td> 0.081250</td>\n",
       "      <td> T</td>\n",
       "      <td> missense_variant</td>\n",
       "      <td> MODERATE</td>\n",
       "      <td> </td>\n",
       "      <td> 0.766667</td>\n",
       "      <td> 0.09375</td>\n",
       "      <td> 0</td>\n",
       "      <td> 0</td>\n",
       "      <td> 0</td>\n",
       "      <td> 0</td>\n",
       "      <td> 0</td>\n",
       "      <td> 0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5576672</th>\n",
       "      <td> G</td>\n",
       "      <td> A</td>\n",
       "      <td> ExCov</td>\n",
       "      <td> NaN</td>\n",
       "      <td> 0.000000</td>\n",
       "      <td> A</td>\n",
       "      <td> missense_variant</td>\n",
       "      <td> MODERATE</td>\n",
       "      <td> </td>\n",
       "      <td> 0.000000</td>\n",
       "      <td> 0.00000</td>\n",
       "      <td> 0</td>\n",
       "      <td> 0</td>\n",
       "      <td> 0</td>\n",
       "      <td> 0</td>\n",
       "      <td> 0</td>\n",
       "      <td> 0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5576717</th>\n",
       "      <td> G</td>\n",
       "      <td> A</td>\n",
       "      <td> ExCov</td>\n",
       "      <td> NaN</td>\n",
       "      <td> 0.000000</td>\n",
       "      <td> A</td>\n",
       "      <td> missense_variant</td>\n",
       "      <td> MODERATE</td>\n",
       "      <td> </td>\n",
       "      <td> 0.000000</td>\n",
       "      <td> 0.00000</td>\n",
       "      <td> 0</td>\n",
       "      <td> 0</td>\n",
       "      <td> 0</td>\n",
       "      <td> 0</td>\n",
       "      <td> 0</td>\n",
       "      <td> 0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5586838</th>\n",
       "      <td> A</td>\n",
       "      <td> G</td>\n",
       "      <td>  PASS</td>\n",
       "      <td> NaN</td>\n",
       "      <td> 0.015924</td>\n",
       "      <td> G</td>\n",
       "      <td> missense_variant</td>\n",
       "      <td> MODERATE</td>\n",
       "      <td> </td>\n",
       "      <td> 0.000000</td>\n",
       "      <td> 0.15625</td>\n",
       "      <td> 0</td>\n",
       "      <td> 0</td>\n",
       "      <td> 0</td>\n",
       "      <td> 0</td>\n",
       "      <td> 0</td>\n",
       "      <td> 0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "              REF ALT FILTER   AA        AF allele            effect  \\\n",
       "CHROM POS                                                              \n",
       "CAE6  5576627   G   A   PASS  NaN  0.000000      A  missense_variant   \n",
       "      5576646   C   T  ExCov  NaN  0.081250      T  missense_variant   \n",
       "      5576672   G   A  ExCov  NaN  0.000000      A  missense_variant   \n",
       "      5576717   G   A  ExCov  NaN  0.000000      A  missense_variant   \n",
       "      5586838   A   G   PASS  NaN  0.015924      G  missense_variant   \n",
       "\n",
       "                 impact errors       aet      cyn  pyn  pys  sab  sac  sar  \\\n",
       "CHROM POS                                                                    \n",
       "CAE6  5576627  MODERATE         0.000000  0.00000    0    0    0    0    0   \n",
       "      5576646  MODERATE         0.766667  0.09375    0    0    0    0    0   \n",
       "      5576672  MODERATE         0.000000  0.00000    0    0    0    0    0   \n",
       "      5576717  MODERATE         0.000000  0.00000    0    0    0    0    0   \n",
       "      5586838  MODERATE         0.000000  0.15625    0    0    0    0    0   \n",
       "\n",
       "               tan  \n",
       "CHROM POS           \n",
       "CAE6  5576627    0  \n",
       "      5576646    0  \n",
       "      5576672    0  \n",
       "      5576717    0  \n",
       "      5586838    0  "
      ]
     },
     "execution_count": 361,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "annot_f[annot_f['impact']=='MODERATE'].xs('RANBP3',level=2)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 362,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div style=\"max-height:1000px;max-width:1500px;overflow:auto;\">\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th>REF</th>\n",
       "      <th>ALT</th>\n",
       "      <th>FILTER</th>\n",
       "      <th>AA</th>\n",
       "      <th>AF</th>\n",
       "      <th>allele</th>\n",
       "      <th>effect</th>\n",
       "      <th>impact</th>\n",
       "      <th>errors</th>\n",
       "      <th>aet</th>\n",
       "      <th>cyn</th>\n",
       "      <th>pyn</th>\n",
       "      <th>pys</th>\n",
       "      <th>sab</th>\n",
       "      <th>sac</th>\n",
       "      <th>sar</th>\n",
       "      <th>tan</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>CHROM</th>\n",
       "      <th>POS</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>CAE6</th>\n",
       "      <th>11707284</th>\n",
       "      <td> T</td>\n",
       "      <td> G</td>\n",
       "      <td> PASS</td>\n",
       "      <td> NaN</td>\n",
       "      <td> 0</td>\n",
       "      <td> G</td>\n",
       "      <td> missense_variant</td>\n",
       "      <td> MODERATE</td>\n",
       "      <td> </td>\n",
       "      <td> 0</td>\n",
       "      <td> 0</td>\n",
       "      <td> 0</td>\n",
       "      <td> 0</td>\n",
       "      <td> 0</td>\n",
       "      <td> 0</td>\n",
       "      <td> 0</td>\n",
       "      <td> 0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "               REF ALT FILTER   AA  AF allele            effect    impact  \\\n",
       "CHROM POS                                                                   \n",
       "CAE6  11707284   T   G   PASS  NaN   0      G  missense_variant  MODERATE   \n",
       "\n",
       "               errors  aet  cyn  pyn  pys  sab  sac  sar  tan  \n",
       "CHROM POS                                                      \n",
       "CAE6  11707284           0    0    0    0    0    0    0    0  "
      ]
     },
     "execution_count": 362,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "annot_f[annot_f['impact']=='MODERATE'].xs('NFIX',level=2)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "annot_f[annot_f['impact']=='MODERATE'].xs('NFIX',level=2)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 309,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div style=\"max-height:1000px;max-width:1500px;overflow:auto;\">\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th>REF</th>\n",
       "      <th>ALT</th>\n",
       "      <th>FILTER</th>\n",
       "      <th>AA</th>\n",
       "      <th>AF</th>\n",
       "      <th>allele</th>\n",
       "      <th>effect</th>\n",
       "      <th>impact</th>\n",
       "      <th>errors</th>\n",
       "      <th>aet</th>\n",
       "      <th>cyn</th>\n",
       "      <th>pyn</th>\n",
       "      <th>pys</th>\n",
       "      <th>sab</th>\n",
       "      <th>sac</th>\n",
       "      <th>sar</th>\n",
       "      <th>tan</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>CHROM</th>\n",
       "      <th>POS</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th rowspan=\"61\" valign=\"top\">CAE16</th>\n",
       "      <th>6955661</th>\n",
       "      <td>      T</td>\n",
       "      <td>    C</td>\n",
       "      <td>                         ExCov</td>\n",
       "      <td> T</td>\n",
       "      <td> 0.215190</td>\n",
       "      <td>   C</td>\n",
       "      <td> upstream_gene_variant</td>\n",
       "      <td> MODIFIER</td>\n",
       "      <td> </td>\n",
       "      <td> 0.000000</td>\n",
       "      <td> 0.718750</td>\n",
       "      <td> 0.333333</td>\n",
       "      <td> 0.427083</td>\n",
       "      <td> 0.000000</td>\n",
       "      <td> 0.000000</td>\n",
       "      <td> 0</td>\n",
       "      <td> 0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6955663</th>\n",
       "      <td>      G</td>\n",
       "      <td>    A</td>\n",
       "      <td>                5bpIndel;ExCov</td>\n",
       "      <td> G</td>\n",
       "      <td> 0.035948</td>\n",
       "      <td>   A</td>\n",
       "      <td> upstream_gene_variant</td>\n",
       "      <td> MODIFIER</td>\n",
       "      <td> </td>\n",
       "      <td> 0.000000</td>\n",
       "      <td> 0.031250</td>\n",
       "      <td> 0.000000</td>\n",
       "      <td> 0.108696</td>\n",
       "      <td> 0.000000</td>\n",
       "      <td> 0.000000</td>\n",
       "      <td> 0</td>\n",
       "      <td> 0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6955667</th>\n",
       "      <td>      T</td>\n",
       "      <td>  TAG</td>\n",
       "      <td>                5bpIndel;ExCov</td>\n",
       "      <td> T</td>\n",
       "      <td> 0.613924</td>\n",
       "      <td>  AG</td>\n",
       "      <td> upstream_gene_variant</td>\n",
       "      <td> MODIFIER</td>\n",
       "      <td> </td>\n",
       "      <td> 1.000000</td>\n",
       "      <td> 1.000000</td>\n",
       "      <td> 1.000000</td>\n",
       "      <td> 1.000000</td>\n",
       "      <td> 0.000000</td>\n",
       "      <td> 0.000000</td>\n",
       "      <td> 0</td>\n",
       "      <td> 1.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6955676</th>\n",
       "      <td>      G</td>\n",
       "      <td>    C</td>\n",
       "      <td>                         ExCov</td>\n",
       "      <td> C</td>\n",
       "      <td> 0.000000</td>\n",
       "      <td>   C</td>\n",
       "      <td> upstream_gene_variant</td>\n",
       "      <td> MODIFIER</td>\n",
       "      <td> </td>\n",
       "      <td> 0.000000</td>\n",
       "      <td> 0.000000</td>\n",
       "      <td> 0.000000</td>\n",
       "      <td> 0.000000</td>\n",
       "      <td> 0.000000</td>\n",
       "      <td> 0.000000</td>\n",
       "      <td> 0</td>\n",
       "      <td> 0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6955704</th>\n",
       "      <td>      T</td>\n",
       "      <td>    C</td>\n",
       "      <td>                         ExCov</td>\n",
       "      <td> C</td>\n",
       "      <td> 0.706250</td>\n",
       "      <td>   C</td>\n",
       "      <td> upstream_gene_variant</td>\n",
       "      <td> MODIFIER</td>\n",
       "      <td> </td>\n",
       "      <td> 1.000000</td>\n",
       "      <td> 1.000000</td>\n",
       "      <td> 1.000000</td>\n",
       "      <td> 1.000000</td>\n",
       "      <td> 0.586957</td>\n",
       "      <td> 0.015152</td>\n",
       "      <td> 0</td>\n",
       "      <td> 1.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6955828</th>\n",
       "      <td>      C</td>\n",
       "      <td>    G</td>\n",
       "      <td>                      5bpIndel</td>\n",
       "      <td> G</td>\n",
       "      <td> 0.000000</td>\n",
       "      <td>   G</td>\n",
       "      <td> upstream_gene_variant</td>\n",
       "      <td> MODIFIER</td>\n",
       "      <td> </td>\n",
       "      <td> 0.000000</td>\n",
       "      <td> 0.000000</td>\n",
       "      <td> 0.000000</td>\n",
       "      <td> 0.000000</td>\n",
       "      <td> 0.000000</td>\n",
       "      <td> 0.000000</td>\n",
       "      <td> 0</td>\n",
       "      <td> 0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6955832</th>\n",
       "      <td>      T</td>\n",
       "      <td>  TCC</td>\n",
       "      <td>                      5bpIndel</td>\n",
       "      <td> T</td>\n",
       "      <td> 0.009375</td>\n",
       "      <td>  CC</td>\n",
       "      <td> upstream_gene_variant</td>\n",
       "      <td> MODIFIER</td>\n",
       "      <td> </td>\n",
       "      <td> 0.000000</td>\n",
       "      <td> 0.000000</td>\n",
       "      <td> 0.000000</td>\n",
       "      <td> 0.000000</td>\n",
       "      <td> 0.000000</td>\n",
       "      <td> 0.000000</td>\n",
       "      <td> 0</td>\n",
       "      <td> 0.136364</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6955865</th>\n",
       "      <td>      C</td>\n",
       "      <td>    T</td>\n",
       "      <td>                          PASS</td>\n",
       "      <td> C</td>\n",
       "      <td> 0.003106</td>\n",
       "      <td>   T</td>\n",
       "      <td> upstream_gene_variant</td>\n",
       "      <td> MODIFIER</td>\n",
       "      <td> </td>\n",
       "      <td> 0.000000</td>\n",
       "      <td> 0.000000</td>\n",
       "      <td> 0.000000</td>\n",
       "      <td> 0.000000</td>\n",
       "      <td> 0.000000</td>\n",
       "      <td> 0.000000</td>\n",
       "      <td> 0</td>\n",
       "      <td> 0.045455</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6956049</th>\n",
       "      <td>      T</td>\n",
       "      <td>    C</td>\n",
       "      <td>                         ExCov</td>\n",
       "      <td> C</td>\n",
       "      <td> 0.000000</td>\n",
       "      <td>   C</td>\n",
       "      <td> upstream_gene_variant</td>\n",
       "      <td> MODIFIER</td>\n",
       "      <td> </td>\n",
       "      <td> 0.000000</td>\n",
       "      <td> 0.000000</td>\n",
       "      <td> 0.000000</td>\n",
       "      <td> 0.000000</td>\n",
       "      <td> 0.000000</td>\n",
       "      <td> 0.000000</td>\n",
       "      <td> 0</td>\n",
       "      <td> 0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6956068</th>\n",
       "      <td>      G</td>\n",
       "      <td>    T</td>\n",
       "      <td>                         ExCov</td>\n",
       "      <td> G</td>\n",
       "      <td> 0.560127</td>\n",
       "      <td>   T</td>\n",
       "      <td> upstream_gene_variant</td>\n",
       "      <td> MODIFIER</td>\n",
       "      <td> </td>\n",
       "      <td> 0.968750</td>\n",
       "      <td> 0.937500</td>\n",
       "      <td> 0.750000</td>\n",
       "      <td> 0.852941</td>\n",
       "      <td> 0.000000</td>\n",
       "      <td> 0.000000</td>\n",
       "      <td> 0</td>\n",
       "      <td> 0.909091</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6956089</th>\n",
       "      <td>     AG</td>\n",
       "      <td>    A</td>\n",
       "      <td> 5bpIndel;ExCov;QualityByDepth</td>\n",
       "      <td> A</td>\n",
       "      <td> 0.018868</td>\n",
       "      <td>    </td>\n",
       "      <td> upstream_gene_variant</td>\n",
       "      <td> MODIFIER</td>\n",
       "      <td> </td>\n",
       "      <td> 0.031250</td>\n",
       "      <td> 0.000000</td>\n",
       "      <td> 0.083333</td>\n",
       "      <td> 0.030000</td>\n",
       "      <td> 0.020833</td>\n",
       "      <td> 0.000000</td>\n",
       "      <td> 0</td>\n",
       "      <td> 0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6956108</th>\n",
       "      <td>      C</td>\n",
       "      <td>    T</td>\n",
       "      <td>  ExCov;LowQual;QualityByDepth</td>\n",
       "      <td> C</td>\n",
       "      <td> 0.016026</td>\n",
       "      <td>   T</td>\n",
       "      <td> upstream_gene_variant</td>\n",
       "      <td> MODIFIER</td>\n",
       "      <td> </td>\n",
       "      <td> 0.033333</td>\n",
       "      <td> 0.000000</td>\n",
       "      <td> 0.083333</td>\n",
       "      <td> 0.030000</td>\n",
       "      <td> 0.000000</td>\n",
       "      <td> 0.000000</td>\n",
       "      <td> 0</td>\n",
       "      <td> 0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6956109</th>\n",
       "      <td>      A</td>\n",
       "      <td>    G</td>\n",
       "      <td>  ExCov;LowQual;QualityByDepth</td>\n",
       "      <td> A</td>\n",
       "      <td> 0.016129</td>\n",
       "      <td>   G</td>\n",
       "      <td> upstream_gene_variant</td>\n",
       "      <td> MODIFIER</td>\n",
       "      <td> </td>\n",
       "      <td> 0.033333</td>\n",
       "      <td> 0.000000</td>\n",
       "      <td> 0.083333</td>\n",
       "      <td> 0.030000</td>\n",
       "      <td> 0.000000</td>\n",
       "      <td> 0.000000</td>\n",
       "      <td> 0</td>\n",
       "      <td> 0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6956128</th>\n",
       "      <td>      A</td>\n",
       "      <td>   AT</td>\n",
       "      <td> 5bpIndel;ExCov;QualityByDepth</td>\n",
       "      <td> A</td>\n",
       "      <td> 0.036184</td>\n",
       "      <td>   T</td>\n",
       "      <td> upstream_gene_variant</td>\n",
       "      <td> MODIFIER</td>\n",
       "      <td> </td>\n",
       "      <td> 0.035714</td>\n",
       "      <td> 0.033333</td>\n",
       "      <td> 0.200000</td>\n",
       "      <td> 0.070000</td>\n",
       "      <td> 0.000000</td>\n",
       "      <td> 0.000000</td>\n",
       "      <td> 0</td>\n",
       "      <td> 0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6956129</th>\n",
       "      <td>      T</td>\n",
       "      <td>    G</td>\n",
       "      <td> 5bpIndel;ExCov;QualityByDepth</td>\n",
       "      <td> T</td>\n",
       "      <td> 0.019481</td>\n",
       "      <td>   G</td>\n",
       "      <td> upstream_gene_variant</td>\n",
       "      <td> MODIFIER</td>\n",
       "      <td> </td>\n",
       "      <td> 0.035714</td>\n",
       "      <td> 0.000000</td>\n",
       "      <td> 0.166667</td>\n",
       "      <td> 0.030000</td>\n",
       "      <td> 0.000000</td>\n",
       "      <td> 0.000000</td>\n",
       "      <td> 0</td>\n",
       "      <td> 0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6956242</th>\n",
       "      <td>      C</td>\n",
       "      <td>   CG</td>\n",
       "      <td>                5bpIndel;ExCov</td>\n",
       "      <td> C</td>\n",
       "      <td> 0.020000</td>\n",
       "      <td>   G</td>\n",
       "      <td> upstream_gene_variant</td>\n",
       "      <td> MODIFIER</td>\n",
       "      <td> </td>\n",
       "      <td> 0.133333</td>\n",
       "      <td> 0.000000</td>\n",
       "      <td> 0.100000</td>\n",
       "      <td> 0.000000</td>\n",
       "      <td> 0.000000</td>\n",
       "      <td> 0.000000</td>\n",
       "      <td> 0</td>\n",
       "      <td> 0.050000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6956243</th>\n",
       "      <td>      C</td>\n",
       "      <td>  A,G</td>\n",
       "      <td>                5bpIndel;ExCov</td>\n",
       "      <td> C</td>\n",
       "      <td> 0.011719</td>\n",
       "      <td>   A</td>\n",
       "      <td> upstream_gene_variant</td>\n",
       "      <td> MODIFIER</td>\n",
       "      <td> </td>\n",
       "      <td> 1.000000</td>\n",
       "      <td> 0.000000</td>\n",
       "      <td> 0.000000</td>\n",
       "      <td> 0.000000</td>\n",
       "      <td> 0.000000</td>\n",
       "      <td> 0.000000</td>\n",
       "      <td> 0</td>\n",
       "      <td> 0.250000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6956294</th>\n",
       "      <td>      G</td>\n",
       "      <td>    A</td>\n",
       "      <td>                         ExCov</td>\n",
       "      <td> G</td>\n",
       "      <td> 0.006329</td>\n",
       "      <td>   A</td>\n",
       "      <td> upstream_gene_variant</td>\n",
       "      <td> MODIFIER</td>\n",
       "      <td> </td>\n",
       "      <td> 0.000000</td>\n",
       "      <td> 0.000000</td>\n",
       "      <td> 0.000000</td>\n",
       "      <td> 0.020000</td>\n",
       "      <td> 0.000000</td>\n",
       "      <td> 0.000000</td>\n",
       "      <td> 0</td>\n",
       "      <td> 0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6956384</th>\n",
       "      <td>      C</td>\n",
       "      <td>    T</td>\n",
       "      <td>                 ExCov;LowQual</td>\n",
       "      <td> C</td>\n",
       "      <td> 0.003205</td>\n",
       "      <td>   T</td>\n",
       "      <td> upstream_gene_variant</td>\n",
       "      <td> MODIFIER</td>\n",
       "      <td> </td>\n",
       "      <td> 0.000000</td>\n",
       "      <td> 0.000000</td>\n",
       "      <td> 0.000000</td>\n",
       "      <td> 0.010000</td>\n",
       "      <td> 0.000000</td>\n",
       "      <td> 0.000000</td>\n",
       "      <td> 0</td>\n",
       "      <td> 0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6956404</th>\n",
       "      <td>      G</td>\n",
       "      <td>    A</td>\n",
       "      <td>                         ExCov</td>\n",
       "      <td> G</td>\n",
       "      <td> 0.215686</td>\n",
       "      <td>   A</td>\n",
       "      <td> upstream_gene_variant</td>\n",
       "      <td> MODIFIER</td>\n",
       "      <td> </td>\n",
       "      <td> 0.000000</td>\n",
       "      <td> 0.269231</td>\n",
       "      <td> 0.000000</td>\n",
       "      <td> 0.590000</td>\n",
       "      <td> 0.000000</td>\n",
       "      <td> 0.000000</td>\n",
       "      <td> 0</td>\n",
       "      <td> 0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6956431</th>\n",
       "      <td>      G</td>\n",
       "      <td>    C</td>\n",
       "      <td>                         ExCov</td>\n",
       "      <td> G</td>\n",
       "      <td> 0.013245</td>\n",
       "      <td>   C</td>\n",
       "      <td> upstream_gene_variant</td>\n",
       "      <td> MODIFIER</td>\n",
       "      <td> </td>\n",
       "      <td> 0.000000</td>\n",
       "      <td> 0.000000</td>\n",
       "      <td> 0.000000</td>\n",
       "      <td> 0.000000</td>\n",
       "      <td> 0.083333</td>\n",
       "      <td> 0.000000</td>\n",
       "      <td> 0</td>\n",
       "      <td> 0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6956546</th>\n",
       "      <td>      G</td>\n",
       "      <td>    C</td>\n",
       "      <td>                      5bpIndel</td>\n",
       "      <td> G</td>\n",
       "      <td> 0.036424</td>\n",
       "      <td>   C</td>\n",
       "      <td> upstream_gene_variant</td>\n",
       "      <td> MODIFIER</td>\n",
       "      <td> </td>\n",
       "      <td> 0.000000</td>\n",
       "      <td> 0.000000</td>\n",
       "      <td> 0.000000</td>\n",
       "      <td> 0.117021</td>\n",
       "      <td> 0.000000</td>\n",
       "      <td> 0.000000</td>\n",
       "      <td> 0</td>\n",
       "      <td> 0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6956548</th>\n",
       "      <td>      G</td>\n",
       "      <td> GCTT</td>\n",
       "      <td>                      5bpIndel</td>\n",
       "      <td> G</td>\n",
       "      <td> 0.029801</td>\n",
       "      <td> CTT</td>\n",
       "      <td> upstream_gene_variant</td>\n",
       "      <td> MODIFIER</td>\n",
       "      <td> </td>\n",
       "      <td> 0.000000</td>\n",
       "      <td> 0.000000</td>\n",
       "      <td> 0.000000</td>\n",
       "      <td> 0.093750</td>\n",
       "      <td> 0.000000</td>\n",
       "      <td> 0.000000</td>\n",
       "      <td> 0</td>\n",
       "      <td> 0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6956569</th>\n",
       "      <td>     CT</td>\n",
       "      <td>    C</td>\n",
       "      <td>                      5bpIndel</td>\n",
       "      <td> C</td>\n",
       "      <td> 0.019481</td>\n",
       "      <td>    </td>\n",
       "      <td> upstream_gene_variant</td>\n",
       "      <td> MODIFIER</td>\n",
       "      <td> </td>\n",
       "      <td> 0.000000</td>\n",
       "      <td> 0.200000</td>\n",
       "      <td> 0.000000</td>\n",
       "      <td> 0.000000</td>\n",
       "      <td> 0.000000</td>\n",
       "      <td> 0.000000</td>\n",
       "      <td> 0</td>\n",
       "      <td> 0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6956573</th>\n",
       "      <td>      C</td>\n",
       "      <td>    G</td>\n",
       "      <td>                      5bpIndel</td>\n",
       "      <td> C</td>\n",
       "      <td> 0.417219</td>\n",
       "      <td>   G</td>\n",
       "      <td> upstream_gene_variant</td>\n",
       "      <td> MODIFIER</td>\n",
       "      <td> </td>\n",
       "      <td> 0.000000</td>\n",
       "      <td> 1.000000</td>\n",
       "      <td> 0.916667</td>\n",
       "      <td> 0.988636</td>\n",
       "      <td> 0.000000</td>\n",
       "      <td> 0.000000</td>\n",
       "      <td> 0</td>\n",
       "      <td> 0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6956606</th>\n",
       "      <td>      C</td>\n",
       "      <td>    T</td>\n",
       "      <td>                          PASS</td>\n",
       "      <td> C</td>\n",
       "      <td> 0.019231</td>\n",
       "      <td>   T</td>\n",
       "      <td> upstream_gene_variant</td>\n",
       "      <td> MODIFIER</td>\n",
       "      <td> </td>\n",
       "      <td> 0.000000</td>\n",
       "      <td> 0.000000</td>\n",
       "      <td> 0.000000</td>\n",
       "      <td> 0.000000</td>\n",
       "      <td> 0.000000</td>\n",
       "      <td> 0.000000</td>\n",
       "      <td> 0</td>\n",
       "      <td> 0.272727</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6956615</th>\n",
       "      <td>      A</td>\n",
       "      <td>    G</td>\n",
       "      <td>                          PASS</td>\n",
       "      <td> G</td>\n",
       "      <td> 0.593333</td>\n",
       "      <td>   G</td>\n",
       "      <td> upstream_gene_variant</td>\n",
       "      <td> MODIFIER</td>\n",
       "      <td> </td>\n",
       "      <td> 1.000000</td>\n",
       "      <td> 1.000000</td>\n",
       "      <td> 1.000000</td>\n",
       "      <td> 1.000000</td>\n",
       "      <td> 0.000000</td>\n",
       "      <td> 0.000000</td>\n",
       "      <td> 0</td>\n",
       "      <td> 1.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6956663</th>\n",
       "      <td>      A</td>\n",
       "      <td>    G</td>\n",
       "      <td>                          PASS</td>\n",
       "      <td> A</td>\n",
       "      <td> 0.024845</td>\n",
       "      <td>   G</td>\n",
       "      <td> upstream_gene_variant</td>\n",
       "      <td> MODIFIER</td>\n",
       "      <td> </td>\n",
       "      <td> 0.000000</td>\n",
       "      <td> 0.000000</td>\n",
       "      <td> 0.000000</td>\n",
       "      <td> 0.000000</td>\n",
       "      <td> 0.020833</td>\n",
       "      <td> 0.106061</td>\n",
       "      <td> 0</td>\n",
       "      <td> 0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6956676</th>\n",
       "      <td>      G</td>\n",
       "      <td>    C</td>\n",
       "      <td>                          PASS</td>\n",
       "      <td> C</td>\n",
       "      <td> 0.000000</td>\n",
       "      <td>   C</td>\n",
       "      <td> upstream_gene_variant</td>\n",
       "      <td> MODIFIER</td>\n",
       "      <td> </td>\n",
       "      <td> 0.000000</td>\n",
       "      <td> 0.000000</td>\n",
       "      <td> 0.000000</td>\n",
       "      <td> 0.000000</td>\n",
       "      <td> 0.000000</td>\n",
       "      <td> 0.000000</td>\n",
       "      <td> 0</td>\n",
       "      <td> 0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6956677</th>\n",
       "      <td>      G</td>\n",
       "      <td>    A</td>\n",
       "      <td>                          PASS</td>\n",
       "      <td> A</td>\n",
       "      <td> 0.000000</td>\n",
       "      <td>   A</td>\n",
       "      <td> upstream_gene_variant</td>\n",
       "      <td> MODIFIER</td>\n",
       "      <td> </td>\n",
       "      <td> 0.000000</td>\n",
       "      <td> 0.000000</td>\n",
       "      <td> 0.000000</td>\n",
       "      <td> 0.000000</td>\n",
       "      <td> 0.000000</td>\n",
       "      <td> 0.000000</td>\n",
       "      <td> 0</td>\n",
       "      <td> 0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6958637</th>\n",
       "      <td>      G</td>\n",
       "      <td>    A</td>\n",
       "      <td>                          PASS</td>\n",
       "      <td> A</td>\n",
       "      <td> 0.000000</td>\n",
       "      <td>   A</td>\n",
       "      <td>        intron_variant</td>\n",
       "      <td> MODIFIER</td>\n",
       "      <td> </td>\n",
       "      <td> 0.000000</td>\n",
       "      <td> 0.000000</td>\n",
       "      <td> 0.000000</td>\n",
       "      <td> 0.000000</td>\n",
       "      <td> 0.000000</td>\n",
       "      <td> 0.000000</td>\n",
       "      <td> 0</td>\n",
       "      <td> 0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6958672</th>\n",
       "      <td>      G</td>\n",
       "      <td>    T</td>\n",
       "      <td>              5bpIndel;LowQual</td>\n",
       "      <td> T</td>\n",
       "      <td> 0.000000</td>\n",
       "      <td>   T</td>\n",
       "      <td>        intron_variant</td>\n",
       "      <td> MODIFIER</td>\n",
       "      <td> </td>\n",
       "      <td> 0.000000</td>\n",
       "      <td> 0.000000</td>\n",
       "      <td> 0.000000</td>\n",
       "      <td> 0.000000</td>\n",
       "      <td> 0.000000</td>\n",
       "      <td> 0.000000</td>\n",
       "      <td> 0</td>\n",
       "      <td> 0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6958672</th>\n",
       "      <td>      G</td>\n",
       "      <td>    T</td>\n",
       "      <td>              5bpIndel;LowQual</td>\n",
       "      <td> T</td>\n",
       "      <td> 0.000000</td>\n",
       "      <td>   T</td>\n",
       "      <td>        intron_variant</td>\n",
       "      <td> MODIFIER</td>\n",
       "      <td> </td>\n",
       "      <td> 0.062500</td>\n",
       "      <td> 0.000000</td>\n",
       "      <td> 0.000000</td>\n",
       "      <td> 0.000000</td>\n",
       "      <td> 0.000000</td>\n",
       "      <td> 0.000000</td>\n",
       "      <td> 0</td>\n",
       "      <td> 0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6958672</th>\n",
       "      <td>      G</td>\n",
       "      <td>    T</td>\n",
       "      <td>              5bpIndel;LowQual</td>\n",
       "      <td> T</td>\n",
       "      <td> 0.000000</td>\n",
       "      <td>   C</td>\n",
       "      <td>        intron_variant</td>\n",
       "      <td> MODIFIER</td>\n",
       "      <td> </td>\n",
       "      <td> 0.000000</td>\n",
       "      <td> 0.000000</td>\n",
       "      <td> 0.000000</td>\n",
       "      <td> 0.000000</td>\n",
       "      <td> 0.000000</td>\n",
       "      <td> 0.000000</td>\n",
       "      <td> 0</td>\n",
       "      <td> 0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6958672</th>\n",
       "      <td>      G</td>\n",
       "      <td>    T</td>\n",
       "      <td>              5bpIndel;LowQual</td>\n",
       "      <td> T</td>\n",
       "      <td> 0.000000</td>\n",
       "      <td>   C</td>\n",
       "      <td>        intron_variant</td>\n",
       "      <td> MODIFIER</td>\n",
       "      <td> </td>\n",
       "      <td> 0.062500</td>\n",
       "      <td> 0.000000</td>\n",
       "      <td> 0.000000</td>\n",
       "      <td> 0.000000</td>\n",
       "      <td> 0.000000</td>\n",
       "      <td> 0.000000</td>\n",
       "      <td> 0</td>\n",
       "      <td> 0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6958672</th>\n",
       "      <td>      G</td>\n",
       "      <td>   GC</td>\n",
       "      <td>                      5bpIndel</td>\n",
       "      <td> T</td>\n",
       "      <td> 0.006536</td>\n",
       "      <td>   T</td>\n",
       "      <td>        intron_variant</td>\n",
       "      <td> MODIFIER</td>\n",
       "      <td> </td>\n",
       "      <td> 0.000000</td>\n",
       "      <td> 0.000000</td>\n",
       "      <td> 0.000000</td>\n",
       "      <td> 0.000000</td>\n",
       "      <td> 0.000000</td>\n",
       "      <td> 0.000000</td>\n",
       "      <td> 0</td>\n",
       "      <td> 0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6958672</th>\n",
       "      <td>      G</td>\n",
       "      <td>   GC</td>\n",
       "      <td>                      5bpIndel</td>\n",
       "      <td> T</td>\n",
       "      <td> 0.006536</td>\n",
       "      <td>   T</td>\n",
       "      <td>        intron_variant</td>\n",
       "      <td> MODIFIER</td>\n",
       "      <td> </td>\n",
       "      <td> 0.062500</td>\n",
       "      <td> 0.000000</td>\n",
       "      <td> 0.000000</td>\n",
       "      <td> 0.000000</td>\n",
       "      <td> 0.000000</td>\n",
       "      <td> 0.000000</td>\n",
       "      <td> 0</td>\n",
       "      <td> 0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6958672</th>\n",
       "      <td>      G</td>\n",
       "      <td>   GC</td>\n",
       "      <td>                      5bpIndel</td>\n",
       "      <td> T</td>\n",
       "      <td> 0.006536</td>\n",
       "      <td>   C</td>\n",
       "      <td>        intron_variant</td>\n",
       "      <td> MODIFIER</td>\n",
       "      <td> </td>\n",
       "      <td> 0.000000</td>\n",
       "      <td> 0.000000</td>\n",
       "      <td> 0.000000</td>\n",
       "      <td> 0.000000</td>\n",
       "      <td> 0.000000</td>\n",
       "      <td> 0.000000</td>\n",
       "      <td> 0</td>\n",
       "      <td> 0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6958672</th>\n",
       "      <td>      G</td>\n",
       "      <td>   GC</td>\n",
       "      <td>                      5bpIndel</td>\n",
       "      <td> T</td>\n",
       "      <td> 0.006536</td>\n",
       "      <td>   C</td>\n",
       "      <td>        intron_variant</td>\n",
       "      <td> MODIFIER</td>\n",
       "      <td> </td>\n",
       "      <td> 0.062500</td>\n",
       "      <td> 0.000000</td>\n",
       "      <td> 0.000000</td>\n",
       "      <td> 0.000000</td>\n",
       "      <td> 0.000000</td>\n",
       "      <td> 0.000000</td>\n",
       "      <td> 0</td>\n",
       "      <td> 0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6958672</th>\n",
       "      <td>      G</td>\n",
       "      <td>    T</td>\n",
       "      <td>              5bpIndel;LowQual</td>\n",
       "      <td> T</td>\n",
       "      <td> 0.000000</td>\n",
       "      <td>   T</td>\n",
       "      <td>        intron_variant</td>\n",
       "      <td> MODIFIER</td>\n",
       "      <td> </td>\n",
       "      <td> 0.000000</td>\n",
       "      <td> 0.000000</td>\n",
       "      <td> 0.000000</td>\n",
       "      <td> 0.000000</td>\n",
       "      <td> 0.000000</td>\n",
       "      <td> 0.000000</td>\n",
       "      <td> 0</td>\n",
       "      <td> 0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6958672</th>\n",
       "      <td>      G</td>\n",
       "      <td>    T</td>\n",
       "      <td>              5bpIndel;LowQual</td>\n",
       "      <td> T</td>\n",
       "      <td> 0.000000</td>\n",
       "      <td>   T</td>\n",
       "      <td>        intron_variant</td>\n",
       "      <td> MODIFIER</td>\n",
       "      <td> </td>\n",
       "      <td> 0.062500</td>\n",
       "      <td> 0.000000</td>\n",
       "      <td> 0.000000</td>\n",
       "      <td> 0.000000</td>\n",
       "      <td> 0.000000</td>\n",
       "      <td> 0.000000</td>\n",
       "      <td> 0</td>\n",
       "      <td> 0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6958672</th>\n",
       "      <td>      G</td>\n",
       "      <td>    T</td>\n",
       "      <td>              5bpIndel;LowQual</td>\n",
       "      <td> T</td>\n",
       "      <td> 0.000000</td>\n",
       "      <td>   C</td>\n",
       "      <td>        intron_variant</td>\n",
       "      <td> MODIFIER</td>\n",
       "      <td> </td>\n",
       "      <td> 0.000000</td>\n",
       "      <td> 0.000000</td>\n",
       "      <td> 0.000000</td>\n",
       "      <td> 0.000000</td>\n",
       "      <td> 0.000000</td>\n",
       "      <td> 0.000000</td>\n",
       "      <td> 0</td>\n",
       "      <td> 0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6958672</th>\n",
       "      <td>      G</td>\n",
       "      <td>    T</td>\n",
       "      <td>              5bpIndel;LowQual</td>\n",
       "      <td> T</td>\n",
       "      <td> 0.000000</td>\n",
       "      <td>   C</td>\n",
       "      <td>        intron_variant</td>\n",
       "      <td> MODIFIER</td>\n",
       "      <td> </td>\n",
       "      <td> 0.062500</td>\n",
       "      <td> 0.000000</td>\n",
       "      <td> 0.000000</td>\n",
       "      <td> 0.000000</td>\n",
       "      <td> 0.000000</td>\n",
       "      <td> 0.000000</td>\n",
       "      <td> 0</td>\n",
       "      <td> 0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6958672</th>\n",
       "      <td>      G</td>\n",
       "      <td>   GC</td>\n",
       "      <td>                      5bpIndel</td>\n",
       "      <td> T</td>\n",
       "      <td> 0.006536</td>\n",
       "      <td>   T</td>\n",
       "      <td>        intron_variant</td>\n",
       "      <td> MODIFIER</td>\n",
       "      <td> </td>\n",
       "      <td> 0.000000</td>\n",
       "      <td> 0.000000</td>\n",
       "      <td> 0.000000</td>\n",
       "      <td> 0.000000</td>\n",
       "      <td> 0.000000</td>\n",
       "      <td> 0.000000</td>\n",
       "      <td> 0</td>\n",
       "      <td> 0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6958672</th>\n",
       "      <td>      G</td>\n",
       "      <td>   GC</td>\n",
       "      <td>                      5bpIndel</td>\n",
       "      <td> T</td>\n",
       "      <td> 0.006536</td>\n",
       "      <td>   T</td>\n",
       "      <td>        intron_variant</td>\n",
       "      <td> MODIFIER</td>\n",
       "      <td> </td>\n",
       "      <td> 0.062500</td>\n",
       "      <td> 0.000000</td>\n",
       "      <td> 0.000000</td>\n",
       "      <td> 0.000000</td>\n",
       "      <td> 0.000000</td>\n",
       "      <td> 0.000000</td>\n",
       "      <td> 0</td>\n",
       "      <td> 0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6958672</th>\n",
       "      <td>      G</td>\n",
       "      <td>   GC</td>\n",
       "      <td>                      5bpIndel</td>\n",
       "      <td> T</td>\n",
       "      <td> 0.006536</td>\n",
       "      <td>   C</td>\n",
       "      <td>        intron_variant</td>\n",
       "      <td> MODIFIER</td>\n",
       "      <td> </td>\n",
       "      <td> 0.000000</td>\n",
       "      <td> 0.000000</td>\n",
       "      <td> 0.000000</td>\n",
       "      <td> 0.000000</td>\n",
       "      <td> 0.000000</td>\n",
       "      <td> 0.000000</td>\n",
       "      <td> 0</td>\n",
       "      <td> 0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6958672</th>\n",
       "      <td>      G</td>\n",
       "      <td>   GC</td>\n",
       "      <td>                      5bpIndel</td>\n",
       "      <td> T</td>\n",
       "      <td> 0.006536</td>\n",
       "      <td>   C</td>\n",
       "      <td>        intron_variant</td>\n",
       "      <td> MODIFIER</td>\n",
       "      <td> </td>\n",
       "      <td> 0.062500</td>\n",
       "      <td> 0.000000</td>\n",
       "      <td> 0.000000</td>\n",
       "      <td> 0.000000</td>\n",
       "      <td> 0.000000</td>\n",
       "      <td> 0.000000</td>\n",
       "      <td> 0</td>\n",
       "      <td> 0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6958686</th>\n",
       "      <td> AACCCC</td>\n",
       "      <td>    A</td>\n",
       "      <td>                      5bpIndel</td>\n",
       "      <td> A</td>\n",
       "      <td> 0.003226</td>\n",
       "      <td>    </td>\n",
       "      <td>        intron_variant</td>\n",
       "      <td> MODIFIER</td>\n",
       "      <td> </td>\n",
       "      <td> 0.000000</td>\n",
       "      <td> 0.000000</td>\n",
       "      <td> 0.000000</td>\n",
       "      <td> 0.000000</td>\n",
       "      <td> 0.000000</td>\n",
       "      <td> 0.000000</td>\n",
       "      <td> 0</td>\n",
       "      <td> 0.045455</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6958703</th>\n",
       "      <td>      C</td>\n",
       "      <td>    T</td>\n",
       "      <td>                          PASS</td>\n",
       "      <td> C</td>\n",
       "      <td> 0.519481</td>\n",
       "      <td>   T</td>\n",
       "      <td>        intron_variant</td>\n",
       "      <td> MODIFIER</td>\n",
       "      <td> </td>\n",
       "      <td> 0.000000</td>\n",
       "      <td> 1.000000</td>\n",
       "      <td> 1.000000</td>\n",
       "      <td> 1.000000</td>\n",
       "      <td> 0.000000</td>\n",
       "      <td> 0.000000</td>\n",
       "      <td> 0</td>\n",
       "      <td> 1.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6958758</th>\n",
       "      <td>      C</td>\n",
       "      <td>    T</td>\n",
       "      <td>                          PASS</td>\n",
       "      <td> T</td>\n",
       "      <td> 0.000000</td>\n",
       "      <td>   T</td>\n",
       "      <td>        intron_variant</td>\n",
       "      <td> MODIFIER</td>\n",
       "      <td> </td>\n",
       "      <td> 0.000000</td>\n",
       "      <td> 0.000000</td>\n",
       "      <td> 0.000000</td>\n",
       "      <td> 0.000000</td>\n",
       "      <td> 0.000000</td>\n",
       "      <td> 0.000000</td>\n",
       "      <td> 0</td>\n",
       "      <td> 0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6958770</th>\n",
       "      <td>      C</td>\n",
       "      <td>    T</td>\n",
       "      <td>                       LowQual</td>\n",
       "      <td> C</td>\n",
       "      <td> 0.003145</td>\n",
       "      <td>   T</td>\n",
       "      <td>        intron_variant</td>\n",
       "      <td> MODIFIER</td>\n",
       "      <td> </td>\n",
       "      <td> 0.000000</td>\n",
       "      <td> 0.000000</td>\n",
       "      <td> 0.083333</td>\n",
       "      <td> 0.000000</td>\n",
       "      <td> 0.000000</td>\n",
       "      <td> 0.000000</td>\n",
       "      <td> 0</td>\n",
       "      <td> 0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6958778</th>\n",
       "      <td>      T</td>\n",
       "      <td>    G</td>\n",
       "      <td>                          PASS</td>\n",
       "      <td> G</td>\n",
       "      <td> 0.000000</td>\n",
       "      <td>   G</td>\n",
       "      <td>        intron_variant</td>\n",
       "      <td> MODIFIER</td>\n",
       "      <td> </td>\n",
       "      <td> 0.000000</td>\n",
       "      <td> 0.000000</td>\n",
       "      <td> 0.000000</td>\n",
       "      <td> 0.000000</td>\n",
       "      <td> 0.000000</td>\n",
       "      <td> 0.000000</td>\n",
       "      <td> 0</td>\n",
       "      <td> 0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6958797</th>\n",
       "      <td>      T</td>\n",
       "      <td>    C</td>\n",
       "      <td>                          PASS</td>\n",
       "      <td> C</td>\n",
       "      <td> 0.000000</td>\n",
       "      <td>   C</td>\n",
       "      <td>        intron_variant</td>\n",
       "      <td> MODIFIER</td>\n",
       "      <td> </td>\n",
       "      <td> 0.000000</td>\n",
       "      <td> 0.000000</td>\n",
       "      <td> 0.000000</td>\n",
       "      <td> 0.000000</td>\n",
       "      <td> 0.000000</td>\n",
       "      <td> 0.000000</td>\n",
       "      <td> 0</td>\n",
       "      <td> 0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6958798</th>\n",
       "      <td>      G</td>\n",
       "      <td>    A</td>\n",
       "      <td>                          PASS</td>\n",
       "      <td> A</td>\n",
       "      <td> 0.000000</td>\n",
       "      <td>   A</td>\n",
       "      <td>        intron_variant</td>\n",
       "      <td> MODIFIER</td>\n",
       "      <td> </td>\n",
       "      <td> 0.000000</td>\n",
       "      <td> 0.000000</td>\n",
       "      <td> 0.000000</td>\n",
       "      <td> 0.000000</td>\n",
       "      <td> 0.000000</td>\n",
       "      <td> 0.000000</td>\n",
       "      <td> 0</td>\n",
       "      <td> 0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6958857</th>\n",
       "      <td>      T</td>\n",
       "      <td>    C</td>\n",
       "      <td>                          PASS</td>\n",
       "      <td> C</td>\n",
       "      <td> 0.000000</td>\n",
       "      <td>   C</td>\n",
       "      <td>        intron_variant</td>\n",
       "      <td> MODIFIER</td>\n",
       "      <td> </td>\n",
       "      <td> 0.000000</td>\n",
       "      <td> 0.000000</td>\n",
       "      <td> 0.000000</td>\n",
       "      <td> 0.000000</td>\n",
       "      <td> 0.000000</td>\n",
       "      <td> 0.000000</td>\n",
       "      <td> 0</td>\n",
       "      <td> 0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6958868</th>\n",
       "      <td>      T</td>\n",
       "      <td>    C</td>\n",
       "      <td>                          PASS</td>\n",
       "      <td> C</td>\n",
       "      <td> 0.000000</td>\n",
       "      <td>   C</td>\n",
       "      <td>        intron_variant</td>\n",
       "      <td> MODIFIER</td>\n",
       "      <td> </td>\n",
       "      <td> 0.000000</td>\n",
       "      <td> 0.000000</td>\n",
       "      <td> 0.000000</td>\n",
       "      <td> 0.000000</td>\n",
       "      <td> 0.000000</td>\n",
       "      <td> 0.000000</td>\n",
       "      <td> 0</td>\n",
       "      <td> 0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6958877</th>\n",
       "      <td>      T</td>\n",
       "      <td>    A</td>\n",
       "      <td>        LowQual;QualityByDepth</td>\n",
       "      <td> T</td>\n",
       "      <td> 0.006369</td>\n",
       "      <td>   A</td>\n",
       "      <td>        intron_variant</td>\n",
       "      <td> MODIFIER</td>\n",
       "      <td> </td>\n",
       "      <td> 0.000000</td>\n",
       "      <td> 0.000000</td>\n",
       "      <td> 0.083333</td>\n",
       "      <td> 0.010000</td>\n",
       "      <td> 0.000000</td>\n",
       "      <td> 0.000000</td>\n",
       "      <td> 0</td>\n",
       "      <td> 0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6958878</th>\n",
       "      <td>      G</td>\n",
       "      <td>    A</td>\n",
       "      <td>                          PASS</td>\n",
       "      <td> G</td>\n",
       "      <td> 0.009615</td>\n",
       "      <td>   A</td>\n",
       "      <td>        intron_variant</td>\n",
       "      <td> MODIFIER</td>\n",
       "      <td> </td>\n",
       "      <td> 0.000000</td>\n",
       "      <td> 0.000000</td>\n",
       "      <td> 0.166667</td>\n",
       "      <td> 0.010204</td>\n",
       "      <td> 0.000000</td>\n",
       "      <td> 0.000000</td>\n",
       "      <td> 0</td>\n",
       "      <td> 0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6958884</th>\n",
       "      <td>      G</td>\n",
       "      <td>    C</td>\n",
       "      <td>                          PASS</td>\n",
       "      <td> C</td>\n",
       "      <td> 0.000000</td>\n",
       "      <td>   C</td>\n",
       "      <td>        intron_variant</td>\n",
       "      <td> MODIFIER</td>\n",
       "      <td> </td>\n",
       "      <td> 0.000000</td>\n",
       "      <td> 0.000000</td>\n",
       "      <td> 0.000000</td>\n",
       "      <td> 0.000000</td>\n",
       "      <td> 0.000000</td>\n",
       "      <td> 0.000000</td>\n",
       "      <td> 0</td>\n",
       "      <td> 0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6958896</th>\n",
       "      <td>      G</td>\n",
       "      <td>    A</td>\n",
       "      <td>                          PASS</td>\n",
       "      <td> A</td>\n",
       "      <td> 0.000000</td>\n",
       "      <td>   A</td>\n",
       "      <td>        intron_variant</td>\n",
       "      <td> MODIFIER</td>\n",
       "      <td> </td>\n",
       "      <td> 0.000000</td>\n",
       "      <td> 0.000000</td>\n",
       "      <td> 0.000000</td>\n",
       "      <td> 0.000000</td>\n",
       "      <td> 0.000000</td>\n",
       "      <td> 0.000000</td>\n",
       "      <td> 0</td>\n",
       "      <td> 0.000000</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>614 rows  17 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                  REF   ALT                         FILTER AA        AF  \\\n",
       "CHROM POS                                                                 \n",
       "CAE16 6955661       T     C                          ExCov  T  0.215190   \n",
       "      6955663       G     A                 5bpIndel;ExCov  G  0.035948   \n",
       "      6955667       T   TAG                 5bpIndel;ExCov  T  0.613924   \n",
       "      6955676       G     C                          ExCov  C  0.000000   \n",
       "      6955704       T     C                          ExCov  C  0.706250   \n",
       "      6955828       C     G                       5bpIndel  G  0.000000   \n",
       "      6955832       T   TCC                       5bpIndel  T  0.009375   \n",
       "      6955865       C     T                           PASS  C  0.003106   \n",
       "      6956049       T     C                          ExCov  C  0.000000   \n",
       "      6956068       G     T                          ExCov  G  0.560127   \n",
       "      6956089      AG     A  5bpIndel;ExCov;QualityByDepth  A  0.018868   \n",
       "      6956108       C     T   ExCov;LowQual;QualityByDepth  C  0.016026   \n",
       "      6956109       A     G   ExCov;LowQual;QualityByDepth  A  0.016129   \n",
       "      6956128       A    AT  5bpIndel;ExCov;QualityByDepth  A  0.036184   \n",
       "      6956129       T     G  5bpIndel;ExCov;QualityByDepth  T  0.019481   \n",
       "      6956242       C    CG                 5bpIndel;ExCov  C  0.020000   \n",
       "      6956243       C   A,G                 5bpIndel;ExCov  C  0.011719   \n",
       "      6956294       G     A                          ExCov  G  0.006329   \n",
       "      6956384       C     T                  ExCov;LowQual  C  0.003205   \n",
       "      6956404       G     A                          ExCov  G  0.215686   \n",
       "      6956431       G     C                          ExCov  G  0.013245   \n",
       "      6956546       G     C                       5bpIndel  G  0.036424   \n",
       "      6956548       G  GCTT                       5bpIndel  G  0.029801   \n",
       "      6956569      CT     C                       5bpIndel  C  0.019481   \n",
       "      6956573       C     G                       5bpIndel  C  0.417219   \n",
       "      6956606       C     T                           PASS  C  0.019231   \n",
       "      6956615       A     G                           PASS  G  0.593333   \n",
       "      6956663       A     G                           PASS  A  0.024845   \n",
       "      6956676       G     C                           PASS  C  0.000000   \n",
       "      6956677       G     A                           PASS  A  0.000000   \n",
       "...               ...   ...                            ... ..       ...   \n",
       "      6958637       G     A                           PASS  A  0.000000   \n",
       "      6958672       G     T               5bpIndel;LowQual  T  0.000000   \n",
       "      6958672       G     T               5bpIndel;LowQual  T  0.000000   \n",
       "      6958672       G     T               5bpIndel;LowQual  T  0.000000   \n",
       "      6958672       G     T               5bpIndel;LowQual  T  0.000000   \n",
       "      6958672       G    GC                       5bpIndel  T  0.006536   \n",
       "      6958672       G    GC                       5bpIndel  T  0.006536   \n",
       "      6958672       G    GC                       5bpIndel  T  0.006536   \n",
       "      6958672       G    GC                       5bpIndel  T  0.006536   \n",
       "      6958672       G     T               5bpIndel;LowQual  T  0.000000   \n",
       "      6958672       G     T               5bpIndel;LowQual  T  0.000000   \n",
       "      6958672       G     T               5bpIndel;LowQual  T  0.000000   \n",
       "      6958672       G     T               5bpIndel;LowQual  T  0.000000   \n",
       "      6958672       G    GC                       5bpIndel  T  0.006536   \n",
       "      6958672       G    GC                       5bpIndel  T  0.006536   \n",
       "      6958672       G    GC                       5bpIndel  T  0.006536   \n",
       "      6958672       G    GC                       5bpIndel  T  0.006536   \n",
       "      6958686  AACCCC     A                       5bpIndel  A  0.003226   \n",
       "      6958703       C     T                           PASS  C  0.519481   \n",
       "      6958758       C     T                           PASS  T  0.000000   \n",
       "      6958770       C     T                        LowQual  C  0.003145   \n",
       "      6958778       T     G                           PASS  G  0.000000   \n",
       "      6958797       T     C                           PASS  C  0.000000   \n",
       "      6958798       G     A                           PASS  A  0.000000   \n",
       "      6958857       T     C                           PASS  C  0.000000   \n",
       "      6958868       T     C                           PASS  C  0.000000   \n",
       "      6958877       T     A         LowQual;QualityByDepth  T  0.006369   \n",
       "      6958878       G     A                           PASS  G  0.009615   \n",
       "      6958884       G     C                           PASS  C  0.000000   \n",
       "      6958896       G     A                           PASS  A  0.000000   \n",
       "\n",
       "              allele                 effect    impact errors       aet  \\\n",
       "CHROM POS                                                                \n",
       "CAE16 6955661      C  upstream_gene_variant  MODIFIER         0.000000   \n",
       "      6955663      A  upstream_gene_variant  MODIFIER         0.000000   \n",
       "      6955667     AG  upstream_gene_variant  MODIFIER         1.000000   \n",
       "      6955676      C  upstream_gene_variant  MODIFIER         0.000000   \n",
       "      6955704      C  upstream_gene_variant  MODIFIER         1.000000   \n",
       "      6955828      G  upstream_gene_variant  MODIFIER         0.000000   \n",
       "      6955832     CC  upstream_gene_variant  MODIFIER         0.000000   \n",
       "      6955865      T  upstream_gene_variant  MODIFIER         0.000000   \n",
       "      6956049      C  upstream_gene_variant  MODIFIER         0.000000   \n",
       "      6956068      T  upstream_gene_variant  MODIFIER         0.968750   \n",
       "      6956089         upstream_gene_variant  MODIFIER         0.031250   \n",
       "      6956108      T  upstream_gene_variant  MODIFIER         0.033333   \n",
       "      6956109      G  upstream_gene_variant  MODIFIER         0.033333   \n",
       "      6956128      T  upstream_gene_variant  MODIFIER         0.035714   \n",
       "      6956129      G  upstream_gene_variant  MODIFIER         0.035714   \n",
       "      6956242      G  upstream_gene_variant  MODIFIER         0.133333   \n",
       "      6956243      A  upstream_gene_variant  MODIFIER         1.000000   \n",
       "      6956294      A  upstream_gene_variant  MODIFIER         0.000000   \n",
       "      6956384      T  upstream_gene_variant  MODIFIER         0.000000   \n",
       "      6956404      A  upstream_gene_variant  MODIFIER         0.000000   \n",
       "      6956431      C  upstream_gene_variant  MODIFIER         0.000000   \n",
       "      6956546      C  upstream_gene_variant  MODIFIER         0.000000   \n",
       "      6956548    CTT  upstream_gene_variant  MODIFIER         0.000000   \n",
       "      6956569         upstream_gene_variant  MODIFIER         0.000000   \n",
       "      6956573      G  upstream_gene_variant  MODIFIER         0.000000   \n",
       "      6956606      T  upstream_gene_variant  MODIFIER         0.000000   \n",
       "      6956615      G  upstream_gene_variant  MODIFIER         1.000000   \n",
       "      6956663      G  upstream_gene_variant  MODIFIER         0.000000   \n",
       "      6956676      C  upstream_gene_variant  MODIFIER         0.000000   \n",
       "      6956677      A  upstream_gene_variant  MODIFIER         0.000000   \n",
       "...              ...                    ...       ...    ...       ...   \n",
       "      6958637      A         intron_variant  MODIFIER         0.000000   \n",
       "      6958672      T         intron_variant  MODIFIER         0.000000   \n",
       "      6958672      T         intron_variant  MODIFIER         0.062500   \n",
       "      6958672      C         intron_variant  MODIFIER         0.000000   \n",
       "      6958672      C         intron_variant  MODIFIER         0.062500   \n",
       "      6958672      T         intron_variant  MODIFIER         0.000000   \n",
       "      6958672      T         intron_variant  MODIFIER         0.062500   \n",
       "      6958672      C         intron_variant  MODIFIER         0.000000   \n",
       "      6958672      C         intron_variant  MODIFIER         0.062500   \n",
       "      6958672      T         intron_variant  MODIFIER         0.000000   \n",
       "      6958672      T         intron_variant  MODIFIER         0.062500   \n",
       "      6958672      C         intron_variant  MODIFIER         0.000000   \n",
       "      6958672      C         intron_variant  MODIFIER         0.062500   \n",
       "      6958672      T         intron_variant  MODIFIER         0.000000   \n",
       "      6958672      T         intron_variant  MODIFIER         0.062500   \n",
       "      6958672      C         intron_variant  MODIFIER         0.000000   \n",
       "      6958672      C         intron_variant  MODIFIER         0.062500   \n",
       "      6958686                intron_variant  MODIFIER         0.000000   \n",
       "      6958703      T         intron_variant  MODIFIER         0.000000   \n",
       "      6958758      T         intron_variant  MODIFIER         0.000000   \n",
       "      6958770      T         intron_variant  MODIFIER         0.000000   \n",
       "      6958778      G         intron_variant  MODIFIER         0.000000   \n",
       "      6958797      C         intron_variant  MODIFIER         0.000000   \n",
       "      6958798      A         intron_variant  MODIFIER         0.000000   \n",
       "      6958857      C         intron_variant  MODIFIER         0.000000   \n",
       "      6958868      C         intron_variant  MODIFIER         0.000000   \n",
       "      6958877      A         intron_variant  MODIFIER         0.000000   \n",
       "      6958878      A         intron_variant  MODIFIER         0.000000   \n",
       "      6958884      C         intron_variant  MODIFIER         0.000000   \n",
       "      6958896      A         intron_variant  MODIFIER         0.000000   \n",
       "\n",
       "                    cyn       pyn       pys       sab       sac  sar       tan  \n",
       "CHROM POS                                                                       \n",
       "CAE16 6955661  0.718750  0.333333  0.427083  0.000000  0.000000    0  0.000000  \n",
       "      6955663  0.031250  0.000000  0.108696  0.000000  0.000000    0  0.000000  \n",
       "      6955667  1.000000  1.000000  1.000000  0.000000  0.000000    0  1.000000  \n",
       "      6955676  0.000000  0.000000  0.000000  0.000000  0.000000    0  0.000000  \n",
       "      6955704  1.000000  1.000000  1.000000  0.586957  0.015152    0  1.000000  \n",
       "      6955828  0.000000  0.000000  0.000000  0.000000  0.000000    0  0.000000  \n",
       "      6955832  0.000000  0.000000  0.000000  0.000000  0.000000    0  0.136364  \n",
       "      6955865  0.000000  0.000000  0.000000  0.000000  0.000000    0  0.045455  \n",
       "      6956049  0.000000  0.000000  0.000000  0.000000  0.000000    0  0.000000  \n",
       "      6956068  0.937500  0.750000  0.852941  0.000000  0.000000    0  0.909091  \n",
       "      6956089  0.000000  0.083333  0.030000  0.020833  0.000000    0  0.000000  \n",
       "      6956108  0.000000  0.083333  0.030000  0.000000  0.000000    0  0.000000  \n",
       "      6956109  0.000000  0.083333  0.030000  0.000000  0.000000    0  0.000000  \n",
       "      6956128  0.033333  0.200000  0.070000  0.000000  0.000000    0  0.000000  \n",
       "      6956129  0.000000  0.166667  0.030000  0.000000  0.000000    0  0.000000  \n",
       "      6956242  0.000000  0.100000  0.000000  0.000000  0.000000    0  0.050000  \n",
       "      6956243  0.000000  0.000000  0.000000  0.000000  0.000000    0  0.250000  \n",
       "      6956294  0.000000  0.000000  0.020000  0.000000  0.000000    0  0.000000  \n",
       "      6956384  0.000000  0.000000  0.010000  0.000000  0.000000    0  0.000000  \n",
       "      6956404  0.269231  0.000000  0.590000  0.000000  0.000000    0  0.000000  \n",
       "      6956431  0.000000  0.000000  0.000000  0.083333  0.000000    0  0.000000  \n",
       "      6956546  0.000000  0.000000  0.117021  0.000000  0.000000    0  0.000000  \n",
       "      6956548  0.000000  0.000000  0.093750  0.000000  0.000000    0  0.000000  \n",
       "      6956569  0.200000  0.000000  0.000000  0.000000  0.000000    0  0.000000  \n",
       "      6956573  1.000000  0.916667  0.988636  0.000000  0.000000    0  0.000000  \n",
       "      6956606  0.000000  0.000000  0.000000  0.000000  0.000000    0  0.272727  \n",
       "      6956615  1.000000  1.000000  1.000000  0.000000  0.000000    0  1.000000  \n",
       "      6956663  0.000000  0.000000  0.000000  0.020833  0.106061    0  0.000000  \n",
       "      6956676  0.000000  0.000000  0.000000  0.000000  0.000000    0  0.000000  \n",
       "      6956677  0.000000  0.000000  0.000000  0.000000  0.000000    0  0.000000  \n",
       "...                 ...       ...       ...       ...       ...  ...       ...  \n",
       "      6958637  0.000000  0.000000  0.000000  0.000000  0.000000    0  0.000000  \n",
       "      6958672  0.000000  0.000000  0.000000  0.000000  0.000000    0  0.000000  \n",
       "      6958672  0.000000  0.000000  0.000000  0.000000  0.000000    0  0.000000  \n",
       "      6958672  0.000000  0.000000  0.000000  0.000000  0.000000    0  0.000000  \n",
       "      6958672  0.000000  0.000000  0.000000  0.000000  0.000000    0  0.000000  \n",
       "      6958672  0.000000  0.000000  0.000000  0.000000  0.000000    0  0.000000  \n",
       "      6958672  0.000000  0.000000  0.000000  0.000000  0.000000    0  0.000000  \n",
       "      6958672  0.000000  0.000000  0.000000  0.000000  0.000000    0  0.000000  \n",
       "      6958672  0.000000  0.000000  0.000000  0.000000  0.000000    0  0.000000  \n",
       "      6958672  0.000000  0.000000  0.000000  0.000000  0.000000    0  0.000000  \n",
       "      6958672  0.000000  0.000000  0.000000  0.000000  0.000000    0  0.000000  \n",
       "      6958672  0.000000  0.000000  0.000000  0.000000  0.000000    0  0.000000  \n",
       "      6958672  0.000000  0.000000  0.000000  0.000000  0.000000    0  0.000000  \n",
       "      6958672  0.000000  0.000000  0.000000  0.000000  0.000000    0  0.000000  \n",
       "      6958672  0.000000  0.000000  0.000000  0.000000  0.000000    0  0.000000  \n",
       "      6958672  0.000000  0.000000  0.000000  0.000000  0.000000    0  0.000000  \n",
       "      6958672  0.000000  0.000000  0.000000  0.000000  0.000000    0  0.000000  \n",
       "      6958686  0.000000  0.000000  0.000000  0.000000  0.000000    0  0.045455  \n",
       "      6958703  1.000000  1.000000  1.000000  0.000000  0.000000    0  1.000000  \n",
       "      6958758  0.000000  0.000000  0.000000  0.000000  0.000000    0  0.000000  \n",
       "      6958770  0.000000  0.083333  0.000000  0.000000  0.000000    0  0.000000  \n",
       "      6958778  0.000000  0.000000  0.000000  0.000000  0.000000    0  0.000000  \n",
       "      6958797  0.000000  0.000000  0.000000  0.000000  0.000000    0  0.000000  \n",
       "      6958798  0.000000  0.000000  0.000000  0.000000  0.000000    0  0.000000  \n",
       "      6958857  0.000000  0.000000  0.000000  0.000000  0.000000    0  0.000000  \n",
       "      6958868  0.000000  0.000000  0.000000  0.000000  0.000000    0  0.000000  \n",
       "      6958877  0.000000  0.083333  0.010000  0.000000  0.000000    0  0.000000  \n",
       "      6958878  0.000000  0.166667  0.010204  0.000000  0.000000    0  0.000000  \n",
       "      6958884  0.000000  0.000000  0.000000  0.000000  0.000000    0  0.000000  \n",
       "      6958896  0.000000  0.000000  0.000000  0.000000  0.000000    0  0.000000  \n",
       "\n",
       "[614 rows x 17 columns]"
      ]
     },
     "execution_count": 309,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "define a metric for af differentiation...."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "from pypopgen import vcfparser"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 310,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div style=\"max-height:1000px;max-width:1500px;overflow:auto;\">\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th>REF</th>\n",
       "      <th>ALT</th>\n",
       "      <th>FILTER</th>\n",
       "      <th>AA</th>\n",
       "      <th>AF</th>\n",
       "      <th>allele</th>\n",
       "      <th>effect</th>\n",
       "      <th>impact</th>\n",
       "      <th>errors</th>\n",
       "      <th>aet</th>\n",
       "      <th>cyn</th>\n",
       "      <th>pyn</th>\n",
       "      <th>pys</th>\n",
       "      <th>sab</th>\n",
       "      <th>sac</th>\n",
       "      <th>sar</th>\n",
       "      <th>tan</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>CHROM</th>\n",
       "      <th>POS</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th rowspan=\"13\" valign=\"top\">CAE16</th>\n",
       "      <th>6957216</th>\n",
       "      <td> G</td>\n",
       "      <td> A</td>\n",
       "      <td>         ExCov</td>\n",
       "      <td> G</td>\n",
       "      <td> 0.003205</td>\n",
       "      <td> A</td>\n",
       "      <td> missense_variant</td>\n",
       "      <td> MODERATE</td>\n",
       "      <td> </td>\n",
       "      <td> 0.03125</td>\n",
       "      <td> 0.000000</td>\n",
       "      <td> 0</td>\n",
       "      <td> 0.000000</td>\n",
       "      <td> 0.000000</td>\n",
       "      <td> 0.000000</td>\n",
       "      <td> 0.0</td>\n",
       "      <td> 0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6957221</th>\n",
       "      <td> G</td>\n",
       "      <td> A</td>\n",
       "      <td>         ExCov</td>\n",
       "      <td> G</td>\n",
       "      <td> 0.009615</td>\n",
       "      <td> A</td>\n",
       "      <td> missense_variant</td>\n",
       "      <td> MODERATE</td>\n",
       "      <td> </td>\n",
       "      <td> 0.00000</td>\n",
       "      <td> 0.000000</td>\n",
       "      <td> 0</td>\n",
       "      <td> 0.030000</td>\n",
       "      <td> 0.000000</td>\n",
       "      <td> 0.000000</td>\n",
       "      <td> 0.0</td>\n",
       "      <td> 0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6957354</th>\n",
       "      <td> G</td>\n",
       "      <td> A</td>\n",
       "      <td>         ExCov</td>\n",
       "      <td> G</td>\n",
       "      <td> 0.056962</td>\n",
       "      <td> A</td>\n",
       "      <td> missense_variant</td>\n",
       "      <td> MODERATE</td>\n",
       "      <td> </td>\n",
       "      <td> 0.56250</td>\n",
       "      <td> 0.000000</td>\n",
       "      <td> 0</td>\n",
       "      <td> 0.000000</td>\n",
       "      <td> 0.000000</td>\n",
       "      <td> 0.000000</td>\n",
       "      <td> 0.0</td>\n",
       "      <td> 0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6957485</th>\n",
       "      <td> A</td>\n",
       "      <td> G</td>\n",
       "      <td>         ExCov</td>\n",
       "      <td> A</td>\n",
       "      <td> 0.022727</td>\n",
       "      <td> G</td>\n",
       "      <td> missense_variant</td>\n",
       "      <td> MODERATE</td>\n",
       "      <td> </td>\n",
       "      <td> 0.21875</td>\n",
       "      <td> 0.000000</td>\n",
       "      <td> 0</td>\n",
       "      <td> 0.000000</td>\n",
       "      <td> 0.000000</td>\n",
       "      <td> 0.000000</td>\n",
       "      <td> 0.0</td>\n",
       "      <td> 0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6957509</th>\n",
       "      <td> G</td>\n",
       "      <td> A</td>\n",
       "      <td>         ExCov</td>\n",
       "      <td> A</td>\n",
       "      <td> 0.000000</td>\n",
       "      <td> A</td>\n",
       "      <td> missense_variant</td>\n",
       "      <td> MODERATE</td>\n",
       "      <td> </td>\n",
       "      <td> 0.00000</td>\n",
       "      <td> 0.000000</td>\n",
       "      <td> 0</td>\n",
       "      <td> 0.000000</td>\n",
       "      <td> 0.000000</td>\n",
       "      <td> 0.000000</td>\n",
       "      <td> 0.0</td>\n",
       "      <td> 0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6957557</th>\n",
       "      <td> A</td>\n",
       "      <td> G</td>\n",
       "      <td>         ExCov</td>\n",
       "      <td> G</td>\n",
       "      <td> 0.000000</td>\n",
       "      <td> G</td>\n",
       "      <td> missense_variant</td>\n",
       "      <td> MODERATE</td>\n",
       "      <td> </td>\n",
       "      <td> 0.00000</td>\n",
       "      <td> 0.000000</td>\n",
       "      <td> 0</td>\n",
       "      <td> 0.000000</td>\n",
       "      <td> 0.000000</td>\n",
       "      <td> 0.000000</td>\n",
       "      <td> 0.0</td>\n",
       "      <td> 0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6957623</th>\n",
       "      <td> C</td>\n",
       "      <td> T</td>\n",
       "      <td>         ExCov</td>\n",
       "      <td> C</td>\n",
       "      <td> 0.022876</td>\n",
       "      <td> T</td>\n",
       "      <td> missense_variant</td>\n",
       "      <td> MODERATE</td>\n",
       "      <td> </td>\n",
       "      <td> 0.00000</td>\n",
       "      <td> 0.033333</td>\n",
       "      <td> 0</td>\n",
       "      <td> 0.058824</td>\n",
       "      <td> 0.000000</td>\n",
       "      <td> 0.000000</td>\n",
       "      <td> 0.0</td>\n",
       "      <td> 0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6957677</th>\n",
       "      <td> A</td>\n",
       "      <td> G</td>\n",
       "      <td> ExCov;LowQual</td>\n",
       "      <td> G</td>\n",
       "      <td> 0.000000</td>\n",
       "      <td> G</td>\n",
       "      <td> missense_variant</td>\n",
       "      <td> MODERATE</td>\n",
       "      <td> </td>\n",
       "      <td> 0.00000</td>\n",
       "      <td> 0.000000</td>\n",
       "      <td> 0</td>\n",
       "      <td> 0.000000</td>\n",
       "      <td> 0.000000</td>\n",
       "      <td> 0.000000</td>\n",
       "      <td> 0.0</td>\n",
       "      <td> 0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6957762</th>\n",
       "      <td> T</td>\n",
       "      <td> C</td>\n",
       "      <td>         ExCov</td>\n",
       "      <td> T</td>\n",
       "      <td> 0.097484</td>\n",
       "      <td> C</td>\n",
       "      <td> missense_variant</td>\n",
       "      <td> MODERATE</td>\n",
       "      <td> </td>\n",
       "      <td> 0.00000</td>\n",
       "      <td> 0.000000</td>\n",
       "      <td> 0</td>\n",
       "      <td> 0.303922</td>\n",
       "      <td> 0.000000</td>\n",
       "      <td> 0.000000</td>\n",
       "      <td> 0.0</td>\n",
       "      <td> 0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6957774</th>\n",
       "      <td> C</td>\n",
       "      <td> T</td>\n",
       "      <td>         ExCov</td>\n",
       "      <td> T</td>\n",
       "      <td> 0.000000</td>\n",
       "      <td> T</td>\n",
       "      <td> missense_variant</td>\n",
       "      <td> MODERATE</td>\n",
       "      <td> </td>\n",
       "      <td> 0.00000</td>\n",
       "      <td> 0.000000</td>\n",
       "      <td> 0</td>\n",
       "      <td> 0.000000</td>\n",
       "      <td> 0.000000</td>\n",
       "      <td> 0.000000</td>\n",
       "      <td> 0.0</td>\n",
       "      <td> 0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6957835</th>\n",
       "      <td> G</td>\n",
       "      <td> A</td>\n",
       "      <td> ExCov;LowQual</td>\n",
       "      <td> G</td>\n",
       "      <td> 0.003185</td>\n",
       "      <td> A</td>\n",
       "      <td> missense_variant</td>\n",
       "      <td> MODERATE</td>\n",
       "      <td> </td>\n",
       "      <td> 0.00000</td>\n",
       "      <td> 0.000000</td>\n",
       "      <td> 0</td>\n",
       "      <td> 0.000000</td>\n",
       "      <td> 0.000000</td>\n",
       "      <td> 0.000000</td>\n",
       "      <td> 0.0</td>\n",
       "      <td> 0.045455</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6958256</th>\n",
       "      <td> T</td>\n",
       "      <td> C</td>\n",
       "      <td>         ExCov</td>\n",
       "      <td> C</td>\n",
       "      <td> 0.000000</td>\n",
       "      <td> C</td>\n",
       "      <td> missense_variant</td>\n",
       "      <td> MODERATE</td>\n",
       "      <td> </td>\n",
       "      <td> 0.00000</td>\n",
       "      <td> 0.000000</td>\n",
       "      <td> 0</td>\n",
       "      <td> 0.000000</td>\n",
       "      <td> 0.000000</td>\n",
       "      <td> 0.000000</td>\n",
       "      <td> 0.0</td>\n",
       "      <td> 0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6958279</th>\n",
       "      <td> G</td>\n",
       "      <td> C</td>\n",
       "      <td>         ExCov</td>\n",
       "      <td> G</td>\n",
       "      <td> 0.276730</td>\n",
       "      <td> C</td>\n",
       "      <td> missense_variant</td>\n",
       "      <td> MODERATE</td>\n",
       "      <td> </td>\n",
       "      <td> 0.00000</td>\n",
       "      <td> 0.000000</td>\n",
       "      <td> 0</td>\n",
       "      <td> 0.000000</td>\n",
       "      <td> 0.708333</td>\n",
       "      <td> 0.758065</td>\n",
       "      <td> 0.7</td>\n",
       "      <td> 0.000000</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "              REF ALT         FILTER AA        AF allele            effect  \\\n",
       "CHROM POS                                                                    \n",
       "CAE16 6957216   G   A          ExCov  G  0.003205      A  missense_variant   \n",
       "      6957221   G   A          ExCov  G  0.009615      A  missense_variant   \n",
       "      6957354   G   A          ExCov  G  0.056962      A  missense_variant   \n",
       "      6957485   A   G          ExCov  A  0.022727      G  missense_variant   \n",
       "      6957509   G   A          ExCov  A  0.000000      A  missense_variant   \n",
       "      6957557   A   G          ExCov  G  0.000000      G  missense_variant   \n",
       "      6957623   C   T          ExCov  C  0.022876      T  missense_variant   \n",
       "      6957677   A   G  ExCov;LowQual  G  0.000000      G  missense_variant   \n",
       "      6957762   T   C          ExCov  T  0.097484      C  missense_variant   \n",
       "      6957774   C   T          ExCov  T  0.000000      T  missense_variant   \n",
       "      6957835   G   A  ExCov;LowQual  G  0.003185      A  missense_variant   \n",
       "      6958256   T   C          ExCov  C  0.000000      C  missense_variant   \n",
       "      6958279   G   C          ExCov  G  0.276730      C  missense_variant   \n",
       "\n",
       "                 impact errors      aet       cyn  pyn       pys       sab  \\\n",
       "CHROM POS                                                                    \n",
       "CAE16 6957216  MODERATE         0.03125  0.000000    0  0.000000  0.000000   \n",
       "      6957221  MODERATE         0.00000  0.000000    0  0.030000  0.000000   \n",
       "      6957354  MODERATE         0.56250  0.000000    0  0.000000  0.000000   \n",
       "      6957485  MODERATE         0.21875  0.000000    0  0.000000  0.000000   \n",
       "      6957509  MODERATE         0.00000  0.000000    0  0.000000  0.000000   \n",
       "      6957557  MODERATE         0.00000  0.000000    0  0.000000  0.000000   \n",
       "      6957623  MODERATE         0.00000  0.033333    0  0.058824  0.000000   \n",
       "      6957677  MODERATE         0.00000  0.000000    0  0.000000  0.000000   \n",
       "      6957762  MODERATE         0.00000  0.000000    0  0.303922  0.000000   \n",
       "      6957774  MODERATE         0.00000  0.000000    0  0.000000  0.000000   \n",
       "      6957835  MODERATE         0.00000  0.000000    0  0.000000  0.000000   \n",
       "      6958256  MODERATE         0.00000  0.000000    0  0.000000  0.000000   \n",
       "      6958279  MODERATE         0.00000  0.000000    0  0.000000  0.708333   \n",
       "\n",
       "                    sac  sar       tan  \n",
       "CHROM POS                               \n",
       "CAE16 6957216  0.000000  0.0  0.000000  \n",
       "      6957221  0.000000  0.0  0.000000  \n",
       "      6957354  0.000000  0.0  0.000000  \n",
       "      6957485  0.000000  0.0  0.000000  \n",
       "      6957509  0.000000  0.0  0.000000  \n",
       "      6957557  0.000000  0.0  0.000000  \n",
       "      6957623  0.000000  0.0  0.000000  \n",
       "      6957677  0.000000  0.0  0.000000  \n",
       "      6957762  0.000000  0.0  0.000000  \n",
       "      6957774  0.000000  0.0  0.000000  \n",
       "      6957835  0.000000  0.0  0.045455  \n",
       "      6958256  0.000000  0.0  0.000000  \n",
       "      6958279  0.758065  0.7  0.000000  "
      ]
     },
     "execution_count": 310,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "tot_level_df[tot_level_df[\"impact\"]=='MODERATE'].xs(\"CD68\",level=2)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 29,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "class SNPEffAnnotationAlleleFreq(vcfparser.Parser):\n",
    "    def __init__(self):\n",
    "        self.annot_df = pd.DataFrame() \n",
    "    \n",
    "    def parse_fun(self,sline):\n",
    "        output = self._filter_(sline)\n",
    "        if output is not None:\n",
    "            self.out_tsv.write(\"\\t\".join(output)+\"\\n\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 37,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "chrom = \"CAE1\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 38,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "variant = open(jn(var_ana_dir,\"163_UG_ref3500_all_sites_{}_filter_anc_mac_diff_ensembl_1.1.78.vcf.gz\".format(chrom)))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 40,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "t = tabix.open(variant.name)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 45,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "ename": "TypeError",
     "evalue": "unsupported operand type(s) for -: 'tuple' and 'int'",
     "output_type": "error",
     "traceback": [
      "\u001b[1;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[1;31mTypeError\u001b[0m                                 Traceback (most recent call last)",
      "\u001b[1;32m<ipython-input-45-29c46fc0938f>\u001b[0m in \u001b[0;36m<module>\u001b[1;34m()\u001b[0m\n\u001b[1;32m----> 1\u001b[1;33m \u001b[0mpeaks\u001b[0m\u001b[1;33m[\u001b[0m\u001b[1;36m1\u001b[0m\u001b[1;33m]\u001b[0m\u001b[1;33m-\u001b[0m\u001b[0mdistance\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n\u001b[0m",
      "\u001b[1;31mTypeError\u001b[0m: unsupported operand type(s) for -: 'tuple' and 'int'"
     ]
    }
   ],
   "source": [
    "peaks[1]-distance"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 46,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "region = t.query(peaks[0][0],peaks[0][1]-distance,peaks[0][1]+distance)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 48,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "['__class__',\n",
       " '__delattr__',\n",
       " '__doc__',\n",
       " '__format__',\n",
       " '__getattribute__',\n",
       " '__hash__',\n",
       " '__init__',\n",
       " '__iter__',\n",
       " '__new__',\n",
       " '__reduce__',\n",
       " '__reduce_ex__',\n",
       " '__repr__',\n",
       " '__setattr__',\n",
       " '__sizeof__',\n",
       " '__str__',\n",
       " '__subclasshook__',\n",
       " 'next']"
      ]
     },
     "execution_count": 48,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "dir(region)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 44,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "distance = 50000"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 32,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "<__main__.SNPEffAnnotationAlleleFreq at 0x58c08d0>"
      ]
     },
     "execution_count": 32,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "distance = 50000 #bp to left and right\n",
    "#chrom, pos = peaks[0]\n",
    "\n",
    "intervals = [(chrom,pos-distance,pos+distance) for chrom,pos in peaks][:1]\n",
    "\n",
    "snp_eff_parser = SNPEffAnnotationAlleleFreq(variant=)\n",
    "\n",
    "variant = open(jn(var_ana_dir,\"163_UG_ref3500_all_sites_{}_filter_anc_mac_diff_ensembl_1.1.78.vcf.gz\".format(chrom)))\n",
    "walker = vcfparser.get_walker(avariant, snp_eff_parser, intervals=intervals, chunk=False, skip_multiple_entries=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "parser_class = parser_classes[args.parser]\n",
    "#    logging.warning(\"INTERVAL INPUT: {}\".format(args.intervals))\n",
    "    parser = parser_class(**{arg:getattr(sub_args,arg) for arg in vars(sub_args)})\n",
    "\n",
    "\n",
    "    walker = get_walker(args.variant, parser, intervals=args.intervals,\n",
    "                                            auto_tabix=args.auto_tabix,\n",
    "                            chunk=not args.no_chunk, chunk_factor=args.chunk_factor,\n",
    "                                 tmp_dir=args.temp_dir, ncpus=args.ncpus,\n",
    "                           skip_multiple_entries=args.skip_multiple_entries,\n",
    "                            progress_report_interval=args.progress_report_interval)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "\n",
    "gene_df_ensembl."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div style=\"max-height:1000px;max-width:1500px;overflow:auto;\">\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th>end</th>\n",
       "      <th>gene_id</th>\n",
       "      <th>gene_name</th>\n",
       "      <th>biotype</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>chrom</th>\n",
       "      <th>start</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th rowspan=\"30\" valign=\"top\">CAE1</th>\n",
       "      <th>3829     </th>\n",
       "      <td>      5154</td>\n",
       "      <td> ENSCSAG00000011340</td>\n",
       "      <td> ENSCSAG00000011340</td>\n",
       "      <td>       protein_coding</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7313     </th>\n",
       "      <td>     10166</td>\n",
       "      <td> ENSCSAG00000011331</td>\n",
       "      <td>               ODF3</td>\n",
       "      <td>       protein_coding</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11325    </th>\n",
       "      <td>     11434</td>\n",
       "      <td> ENSCSAG00000022912</td>\n",
       "      <td> ENSCSAG00000022912</td>\n",
       "      <td>                miRNA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15562    </th>\n",
       "      <td>     17569</td>\n",
       "      <td> ENSCSAG00000011317</td>\n",
       "      <td>              BET1L</td>\n",
       "      <td>       protein_coding</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>19013    </th>\n",
       "      <td>     25473</td>\n",
       "      <td> ENSCSAG00000011307</td>\n",
       "      <td>              RIC8A</td>\n",
       "      <td>       protein_coding</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>19573    </th>\n",
       "      <td>     19643</td>\n",
       "      <td> ENSCSAG00000020592</td>\n",
       "      <td> ENSCSAG00000020592</td>\n",
       "      <td>                miRNA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>26927    </th>\n",
       "      <td>     48944</td>\n",
       "      <td> ENSCSAG00000011291</td>\n",
       "      <td>              SIRT3</td>\n",
       "      <td>       protein_coding</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>49558    </th>\n",
       "      <td>     66974</td>\n",
       "      <td> ENSCSAG00000011249</td>\n",
       "      <td>             PSMD13</td>\n",
       "      <td>       protein_coding</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>93334    </th>\n",
       "      <td>     98972</td>\n",
       "      <td> ENSCSAG00000011172</td>\n",
       "      <td>              NLRP6</td>\n",
       "      <td>       protein_coding</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>102136   </th>\n",
       "      <td>    108370</td>\n",
       "      <td> ENSCSAG00000011141</td>\n",
       "      <td>              ATHL1</td>\n",
       "      <td>       protein_coding</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>112469   </th>\n",
       "      <td>    113448</td>\n",
       "      <td> ENSCSAG00000011109</td>\n",
       "      <td>             IFITM5</td>\n",
       "      <td>       protein_coding</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>128138   </th>\n",
       "      <td>    129734</td>\n",
       "      <td> ENSCSAG00000011104</td>\n",
       "      <td> ENSCSAG00000011104</td>\n",
       "      <td>       protein_coding</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>147184   </th>\n",
       "      <td>    148057</td>\n",
       "      <td> ENSCSAG00000011099</td>\n",
       "      <td> ENSCSAG00000011099</td>\n",
       "      <td>       protein_coding</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>194757   </th>\n",
       "      <td>    206519</td>\n",
       "      <td> ENSCSAG00000011092</td>\n",
       "      <td>           B4GALNT4</td>\n",
       "      <td>       protein_coding</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>215874   </th>\n",
       "      <td>    224015</td>\n",
       "      <td> ENSCSAG00000011081</td>\n",
       "      <td>               PKP3</td>\n",
       "      <td>       protein_coding</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>224726   </th>\n",
       "      <td>    228976</td>\n",
       "      <td> ENSCSAG00000011075</td>\n",
       "      <td>             SIGIRR</td>\n",
       "      <td>       protein_coding</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>235702   </th>\n",
       "      <td>    249990</td>\n",
       "      <td> ENSCSAG00000011063</td>\n",
       "      <td>               ANO9</td>\n",
       "      <td>       protein_coding</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>255089   </th>\n",
       "      <td>    255396</td>\n",
       "      <td> ENSCSAG00000021823</td>\n",
       "      <td>        Metazoa_SRP</td>\n",
       "      <td>             misc_RNA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>265246   </th>\n",
       "      <td>    303824</td>\n",
       "      <td> ENSCSAG00000011045</td>\n",
       "      <td>             PTDSS2</td>\n",
       "      <td>       protein_coding</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>305282   </th>\n",
       "      <td>    321270</td>\n",
       "      <td> ENSCSAG00000011038</td>\n",
       "      <td>               RNH1</td>\n",
       "      <td>       protein_coding</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>343506   </th>\n",
       "      <td>    346637</td>\n",
       "      <td> ENSCSAG00000011033</td>\n",
       "      <td>               HRAS</td>\n",
       "      <td>       protein_coding</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>355982   </th>\n",
       "      <td>    369633</td>\n",
       "      <td> ENSCSAG00000011029</td>\n",
       "      <td>             LRRC56</td>\n",
       "      <td>       protein_coding</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>370338   </th>\n",
       "      <td>    374345</td>\n",
       "      <td> ENSCSAG00000011007</td>\n",
       "      <td>             LMNTD2</td>\n",
       "      <td>       protein_coding</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>376405   </th>\n",
       "      <td>    379640</td>\n",
       "      <td> ENSCSAG00000010983</td>\n",
       "      <td>             RASSF7</td>\n",
       "      <td>       protein_coding</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>383379   </th>\n",
       "      <td>    383489</td>\n",
       "      <td> ENSCSAG00000027014</td>\n",
       "      <td> ENSCSAG00000027014</td>\n",
       "      <td>                miRNA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>397520   </th>\n",
       "      <td>    429668</td>\n",
       "      <td> ENSCSAG00000010968</td>\n",
       "      <td>              PHRF1</td>\n",
       "      <td>       protein_coding</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>429310   </th>\n",
       "      <td>    432925</td>\n",
       "      <td> ENSCSAG00000010883</td>\n",
       "      <td>               IRF7</td>\n",
       "      <td>       protein_coding</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>435622   </th>\n",
       "      <td>    443156</td>\n",
       "      <td> ENSCSAG00000010823</td>\n",
       "      <td>              CDHR5</td>\n",
       "      <td>       protein_coding</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>444601   </th>\n",
       "      <td>    445339</td>\n",
       "      <td> ENSCSAG00000010809</td>\n",
       "      <td>                SCT</td>\n",
       "      <td>       protein_coding</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>455700   </th>\n",
       "      <td>    459482</td>\n",
       "      <td> ENSCSAG00000010793</td>\n",
       "      <td>               DRD4</td>\n",
       "      <td>       protein_coding</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th rowspan=\"12\" valign=\"top\">CAEX</th>\n",
       "      <th>129315465</th>\n",
       "      <td> 129342193</td>\n",
       "      <td> ENSCSAG00000006563</td>\n",
       "      <td>             FUNDC2</td>\n",
       "      <td>       protein_coding</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>129349054</th>\n",
       "      <td> 129351326</td>\n",
       "      <td> ENSCSAG00000006553</td>\n",
       "      <td>               CMC4</td>\n",
       "      <td>       protein_coding</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>129352805</th>\n",
       "      <td> 129353322</td>\n",
       "      <td> ENSCSAG00000006547</td>\n",
       "      <td>              MTCP1</td>\n",
       "      <td>       protein_coding</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>129358862</th>\n",
       "      <td> 129403352</td>\n",
       "      <td> ENSCSAG00000006538</td>\n",
       "      <td> ENSCSAG00000006538</td>\n",
       "      <td>       protein_coding</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>129487616</th>\n",
       "      <td> 129509313</td>\n",
       "      <td> ENSCSAG00000006524</td>\n",
       "      <td> ENSCSAG00000006524</td>\n",
       "      <td>       protein_coding</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>129536347</th>\n",
       "      <td> 129542075</td>\n",
       "      <td> ENSCSAG00000006507</td>\n",
       "      <td>             RAB39B</td>\n",
       "      <td>       protein_coding</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>129560658</th>\n",
       "      <td> 129606438</td>\n",
       "      <td> ENSCSAG00000006505</td>\n",
       "      <td>              CLIC2</td>\n",
       "      <td>       protein_coding</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>129626286</th>\n",
       "      <td> 129729592</td>\n",
       "      <td> ENSCSAG00000006499</td>\n",
       "      <td>              TMLHE</td>\n",
       "      <td>       protein_coding</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>129808874</th>\n",
       "      <td> 129809416</td>\n",
       "      <td> ENSCSAG00000000481</td>\n",
       "      <td> ENSCSAG00000000481</td>\n",
       "      <td> processed_pseudogene</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>129838996</th>\n",
       "      <td> 129839862</td>\n",
       "      <td> ENSCSAG00000000518</td>\n",
       "      <td>              SPRY3</td>\n",
       "      <td>       protein_coding</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>129937046</th>\n",
       "      <td> 129984119</td>\n",
       "      <td> ENSCSAG00000006492</td>\n",
       "      <td>              VAMP7</td>\n",
       "      <td>       protein_coding</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>130006899</th>\n",
       "      <td> 130020894</td>\n",
       "      <td> ENSCSAG00000006485</td>\n",
       "      <td>               IL9R</td>\n",
       "      <td>       protein_coding</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th rowspan=\"18\" valign=\"top\">CAEY</th>\n",
       "      <th>3899     </th>\n",
       "      <td>     32376</td>\n",
       "      <td> ENSCSAG00000018322</td>\n",
       "      <td> ENSCSAG00000018322</td>\n",
       "      <td>       protein_coding</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>99000    </th>\n",
       "      <td>    128638</td>\n",
       "      <td> ENSCSAG00000018321</td>\n",
       "      <td> ENSCSAG00000018321</td>\n",
       "      <td>       protein_coding</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>373229   </th>\n",
       "      <td>    400632</td>\n",
       "      <td> ENSCSAG00000018320</td>\n",
       "      <td> ENSCSAG00000018320</td>\n",
       "      <td>       protein_coding</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>489840   </th>\n",
       "      <td>    626098</td>\n",
       "      <td> ENSCSAG00000018318</td>\n",
       "      <td> ENSCSAG00000018318</td>\n",
       "      <td>       protein_coding</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>689380   </th>\n",
       "      <td>    706092</td>\n",
       "      <td> ENSCSAG00000018317</td>\n",
       "      <td> ENSCSAG00000018317</td>\n",
       "      <td>       protein_coding</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>968810   </th>\n",
       "      <td>   1188253</td>\n",
       "      <td> ENSCSAG00000018316</td>\n",
       "      <td> ENSCSAG00000018316</td>\n",
       "      <td>       protein_coding</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2512762  </th>\n",
       "      <td>   2531915</td>\n",
       "      <td> ENSCSAG00000018315</td>\n",
       "      <td> ENSCSAG00000018315</td>\n",
       "      <td>       protein_coding</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2626940  </th>\n",
       "      <td>   2668114</td>\n",
       "      <td> ENSCSAG00000018314</td>\n",
       "      <td> ENSCSAG00000018314</td>\n",
       "      <td>       protein_coding</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2638256  </th>\n",
       "      <td>   2638312</td>\n",
       "      <td> ENSCSAG00000020832</td>\n",
       "      <td> ENSCSAG00000020832</td>\n",
       "      <td>                miRNA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2885467  </th>\n",
       "      <td>   2902561</td>\n",
       "      <td> ENSCSAG00000018313</td>\n",
       "      <td> ENSCSAG00000018313</td>\n",
       "      <td>       protein_coding</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3016251  </th>\n",
       "      <td>   3048094</td>\n",
       "      <td> ENSCSAG00000018312</td>\n",
       "      <td> ENSCSAG00000018312</td>\n",
       "      <td>       protein_coding</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3938440  </th>\n",
       "      <td>   3938501</td>\n",
       "      <td> ENSCSAG00000021637</td>\n",
       "      <td>                 U7</td>\n",
       "      <td>                snRNA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4017611  </th>\n",
       "      <td>   4021871</td>\n",
       "      <td> ENSCSAG00000018311</td>\n",
       "      <td> ENSCSAG00000018311</td>\n",
       "      <td>       protein_coding</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4036571  </th>\n",
       "      <td>   4047431</td>\n",
       "      <td> ENSCSAG00000018310</td>\n",
       "      <td> ENSCSAG00000018310</td>\n",
       "      <td>       protein_coding</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4114243  </th>\n",
       "      <td>   4142355</td>\n",
       "      <td> ENSCSAG00000018309</td>\n",
       "      <td>               XKR3</td>\n",
       "      <td>       protein_coding</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4487008  </th>\n",
       "      <td>   4488693</td>\n",
       "      <td> ENSCSAG00000018308</td>\n",
       "      <td> ENSCSAG00000018308</td>\n",
       "      <td>       protein_coding</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5350554  </th>\n",
       "      <td>   5368631</td>\n",
       "      <td> ENSCSAG00000018306</td>\n",
       "      <td> ENSCSAG00000018306</td>\n",
       "      <td>       protein_coding</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5352628  </th>\n",
       "      <td>   5364577</td>\n",
       "      <td> ENSCSAG00000018304</td>\n",
       "      <td> ENSCSAG00000018304</td>\n",
       "      <td>       protein_coding</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>27516 rows  4 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                       end             gene_id           gene_name  \\\n",
       "chrom start                                                          \n",
       "CAE1  3829            5154  ENSCSAG00000011340  ENSCSAG00000011340   \n",
       "      7313           10166  ENSCSAG00000011331                ODF3   \n",
       "      11325          11434  ENSCSAG00000022912  ENSCSAG00000022912   \n",
       "      15562          17569  ENSCSAG00000011317               BET1L   \n",
       "      19013          25473  ENSCSAG00000011307               RIC8A   \n",
       "      19573          19643  ENSCSAG00000020592  ENSCSAG00000020592   \n",
       "      26927          48944  ENSCSAG00000011291               SIRT3   \n",
       "      49558          66974  ENSCSAG00000011249              PSMD13   \n",
       "      93334          98972  ENSCSAG00000011172               NLRP6   \n",
       "      102136        108370  ENSCSAG00000011141               ATHL1   \n",
       "      112469        113448  ENSCSAG00000011109              IFITM5   \n",
       "      128138        129734  ENSCSAG00000011104  ENSCSAG00000011104   \n",
       "      147184        148057  ENSCSAG00000011099  ENSCSAG00000011099   \n",
       "      194757        206519  ENSCSAG00000011092            B4GALNT4   \n",
       "      215874        224015  ENSCSAG00000011081                PKP3   \n",
       "      224726        228976  ENSCSAG00000011075              SIGIRR   \n",
       "      235702        249990  ENSCSAG00000011063                ANO9   \n",
       "      255089        255396  ENSCSAG00000021823         Metazoa_SRP   \n",
       "      265246        303824  ENSCSAG00000011045              PTDSS2   \n",
       "      305282        321270  ENSCSAG00000011038                RNH1   \n",
       "      343506        346637  ENSCSAG00000011033                HRAS   \n",
       "      355982        369633  ENSCSAG00000011029              LRRC56   \n",
       "      370338        374345  ENSCSAG00000011007              LMNTD2   \n",
       "      376405        379640  ENSCSAG00000010983              RASSF7   \n",
       "      383379        383489  ENSCSAG00000027014  ENSCSAG00000027014   \n",
       "      397520        429668  ENSCSAG00000010968               PHRF1   \n",
       "      429310        432925  ENSCSAG00000010883                IRF7   \n",
       "      435622        443156  ENSCSAG00000010823               CDHR5   \n",
       "      444601        445339  ENSCSAG00000010809                 SCT   \n",
       "      455700        459482  ENSCSAG00000010793                DRD4   \n",
       "...                    ...                 ...                 ...   \n",
       "CAEX  129315465  129342193  ENSCSAG00000006563              FUNDC2   \n",
       "      129349054  129351326  ENSCSAG00000006553                CMC4   \n",
       "      129352805  129353322  ENSCSAG00000006547               MTCP1   \n",
       "      129358862  129403352  ENSCSAG00000006538  ENSCSAG00000006538   \n",
       "      129487616  129509313  ENSCSAG00000006524  ENSCSAG00000006524   \n",
       "      129536347  129542075  ENSCSAG00000006507              RAB39B   \n",
       "      129560658  129606438  ENSCSAG00000006505               CLIC2   \n",
       "      129626286  129729592  ENSCSAG00000006499               TMLHE   \n",
       "      129808874  129809416  ENSCSAG00000000481  ENSCSAG00000000481   \n",
       "      129838996  129839862  ENSCSAG00000000518               SPRY3   \n",
       "      129937046  129984119  ENSCSAG00000006492               VAMP7   \n",
       "      130006899  130020894  ENSCSAG00000006485                IL9R   \n",
       "CAEY  3899           32376  ENSCSAG00000018322  ENSCSAG00000018322   \n",
       "      99000         128638  ENSCSAG00000018321  ENSCSAG00000018321   \n",
       "      373229        400632  ENSCSAG00000018320  ENSCSAG00000018320   \n",
       "      489840        626098  ENSCSAG00000018318  ENSCSAG00000018318   \n",
       "      689380        706092  ENSCSAG00000018317  ENSCSAG00000018317   \n",
       "      968810       1188253  ENSCSAG00000018316  ENSCSAG00000018316   \n",
       "      2512762      2531915  ENSCSAG00000018315  ENSCSAG00000018315   \n",
       "      2626940      2668114  ENSCSAG00000018314  ENSCSAG00000018314   \n",
       "      2638256      2638312  ENSCSAG00000020832  ENSCSAG00000020832   \n",
       "      2885467      2902561  ENSCSAG00000018313  ENSCSAG00000018313   \n",
       "      3016251      3048094  ENSCSAG00000018312  ENSCSAG00000018312   \n",
       "      3938440      3938501  ENSCSAG00000021637                  U7   \n",
       "      4017611      4021871  ENSCSAG00000018311  ENSCSAG00000018311   \n",
       "      4036571      4047431  ENSCSAG00000018310  ENSCSAG00000018310   \n",
       "      4114243      4142355  ENSCSAG00000018309                XKR3   \n",
       "      4487008      4488693  ENSCSAG00000018308  ENSCSAG00000018308   \n",
       "      5350554      5368631  ENSCSAG00000018306  ENSCSAG00000018306   \n",
       "      5352628      5364577  ENSCSAG00000018304  ENSCSAG00000018304   \n",
       "\n",
       "                              biotype  \n",
       "chrom start                            \n",
       "CAE1  3829             protein_coding  \n",
       "      7313             protein_coding  \n",
       "      11325                     miRNA  \n",
       "      15562            protein_coding  \n",
       "      19013            protein_coding  \n",
       "      19573                     miRNA  \n",
       "      26927            protein_coding  \n",
       "      49558            protein_coding  \n",
       "      93334            protein_coding  \n",
       "      102136           protein_coding  \n",
       "      112469           protein_coding  \n",
       "      128138           protein_coding  \n",
       "      147184           protein_coding  \n",
       "      194757           protein_coding  \n",
       "      215874           protein_coding  \n",
       "      224726           protein_coding  \n",
       "      235702           protein_coding  \n",
       "      255089                 misc_RNA  \n",
       "      265246           protein_coding  \n",
       "      305282           protein_coding  \n",
       "      343506           protein_coding  \n",
       "      355982           protein_coding  \n",
       "      370338           protein_coding  \n",
       "      376405           protein_coding  \n",
       "      383379                    miRNA  \n",
       "      397520           protein_coding  \n",
       "      429310           protein_coding  \n",
       "      435622           protein_coding  \n",
       "      444601           protein_coding  \n",
       "      455700           protein_coding  \n",
       "...                               ...  \n",
       "CAEX  129315465        protein_coding  \n",
       "      129349054        protein_coding  \n",
       "      129352805        protein_coding  \n",
       "      129358862        protein_coding  \n",
       "      129487616        protein_coding  \n",
       "      129536347        protein_coding  \n",
       "      129560658        protein_coding  \n",
       "      129626286        protein_coding  \n",
       "      129808874  processed_pseudogene  \n",
       "      129838996        protein_coding  \n",
       "      129937046        protein_coding  \n",
       "      130006899        protein_coding  \n",
       "CAEY  3899             protein_coding  \n",
       "      99000            protein_coding  \n",
       "      373229           protein_coding  \n",
       "      489840           protein_coding  \n",
       "      689380           protein_coding  \n",
       "      968810           protein_coding  \n",
       "      2512762          protein_coding  \n",
       "      2626940          protein_coding  \n",
       "      2638256                   miRNA  \n",
       "      2885467          protein_coding  \n",
       "      3016251          protein_coding  \n",
       "      3938440                   snRNA  \n",
       "      4017611          protein_coding  \n",
       "      4036571          protein_coding  \n",
       "      4114243          protein_coding  \n",
       "      4487008          protein_coding  \n",
       "      5350554          protein_coding  \n",
       "      5352628          protein_coding  \n",
       "\n",
       "[27516 rows x 4 columns]"
      ]
     },
     "execution_count": 24,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "gene_df_ensembl"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "collapsed": true
   },
   "source": [
    "## For all genes: get for \"pass\" variants number of syn, nonsyn, high impact and allele frequencies in each category in all pops"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 2",
   "language": "python",
   "name": "python2"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 2
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython2",
   "version": "2.7.3"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 0
}
