{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 137,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "%matplotlib inline"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 138,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "#parallel\n",
    "from ipyparallel import Client\n",
    "rc = Client(profile=\"default\")\n",
    "dview = rc[:] # use all engines\n",
    "lv = rc.load_balanced_view()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 201,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "%%px --local\n",
    "import os,  ast\n",
    "import numpy as np\n",
    "import pandas as pd\n",
    "import vcf, pyfasta\n",
    "import matplotlib as mpl\n",
    "from matplotlib import pyplot as plt\n",
    "\n",
    "africa_pops3 = [\"aet\",\"cyn\",\"pyn\",\"pys\",\"sab\",\"tan\"]\n",
    "pops3 = [\"aet\",\"cyn\",\"pyn\",\"pys\",\"sab\",'sac','sar',\"tan\"]\n",
    "\n",
    "chromosomes = [\"CAE\" + str(i) for i in range(1,30) + [\"X\"] + ['Y'] ]\n",
    "eu = os.path.expanduser\n",
    "jn = os.path.join\n",
    "var_ana_dir = eu(\"~/vervet_project/analyses/20150504_163_UnifiedGenotyper/_data\")\n",
    "variant_impact_dir = eu(\"~/vervet_project/analyses/20160516_variant_impact/_data\")\n",
    "meta_dir = eu(\"~/vervet_project/metadata\")\n",
    "\n",
    "\n",
    "\n",
    "def str_to_dic(s):\n",
    "    try:\n",
    "        return ast.literal_eval(s)\n",
    "    except ValueError:\n",
    "        print s\n",
    "        raise\n",
    "        \n",
    "mt = pd.read_csv(jn(meta_dir,\"163_master_table.csv\"),\n",
    "                 index_col=0,\n",
    "                 converters={\"library\":str_to_dic, \"phenotypes\":str_to_dic})\n",
    "\n",
    "\n",
    "#gene_df_ensembl = pd.read_csv(jn(meta_dir,\"annot_ensemble_1.1.77_genes.tsv\"),sep=\"\\t\",index_col=[0,1])\n",
    "#gene_df_ensembl_CDS = pd.read_csv(jn(meta_dir,\"annot_ensemble_1.1.77_CDS.tsv\"),sep=\"\\t\",index_col=[0,1])\n",
    "mac_seq = pyfasta.Fasta(eu(\"~/vervet_project/data/other_species/Macaca_mulatta/CR_1.0_in_vervet_coord_ref3500.fa\"))\n",
    "#gene_df_r100 = pd.read_csv(jn(meta_dir,\"annot_release100_genes.tsv\"),sep=\"\\t\",index_col=[0,1])\n",
    "#gene_df_r100_exons = pd.read_csv(jn(meta_dir,\"annot_release100_exons.tsv\"),sep=\"\\t\",index_col=[0,1])\n",
    "#attention, upgrade to ensembl 1.1.78!!!\n",
    "#gene_df_ensembl = pd.read_csv(jn(meta_dir,\"annot_ensemble_1.1.77_genes.tsv\"),sep=\"\\t\",index_col=[0,1])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 150,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "%%px --local\n",
    "\n",
    "import re\n",
    "import subprocess, gzip\n",
    "\n",
    "def add_info_line(info_dic, line):\n",
    "    #split all commas except if in quotes\n",
    "    r = re.compile('[^,]+\".+\"[^,]*'\n",
    "                '|'\n",
    "                '[^,]+')\n",
    "    try:\n",
    "        category, value = line.strip().split('=<')\n",
    "    except ValueError:\n",
    "        try: \n",
    "            info_dic[\"uncategorised\"].update({line.strip().split('=')[0][2:]:line.strip().split('=')[1]})\n",
    "        except KeyError:\n",
    "            try:\n",
    "                info_dic.update({\"uncategorised\":{line.strip().split('=')[0][2:]:line.strip().split('=')[1]}})\n",
    "            except ValueError, e:\n",
    "                print line.strip().split('=')\n",
    "                raise e\n",
    "        return\n",
    "    category = category[2:]\n",
    "    tags = r.findall(value[:-1])\n",
    "    subdic = {k:v for k,v in [t.split('=',1) for t  in tags]}\n",
    "    try:\n",
    "        info_dic[category].update({subdic[\"ID\"]:subdic})\n",
    "    except KeyError:\n",
    "        info_dic.update({category:{subdic[\"ID\"]:subdic}})\n",
    "        \n",
    "def parse_vcf_header(fn):\n",
    "    info_dic = {}\n",
    "    fh = gzip.open(fn) if fn[-3:] == \".gz\" else open(fn)\n",
    "    for line in fh:\n",
    "        if line[:6] == \"#CHROM\":\n",
    "            header = line.strip().split('\\t')\n",
    "            header[0] = header[0][1:]\n",
    "        else:\n",
    "            add_info_line(info_dic, line)\n",
    "        if line[0] != '#':\n",
    "            break\n",
    "    return header, info_dic"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 140,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "%%px --local\n",
    "def get_vcf_df(fn, chrom, start, end, header=None, **kwa):\n",
    "    if header is None:\n",
    "        header, _ = parse_vcf_header(fn)\n",
    "    tabix_stream = subprocess.Popen(['tabix',fn,\"{}:{}-{}\".format(chrom,start,end)], stdout=subprocess.PIPE,stderr=subprocess.PIPE)\n",
    "    vcf_df = pd.read_csv(tabix_stream.stdout, sep=\"\\t\", index_col=[0,1], header=False, names = header, **kwa)\n",
    "    \n",
    "    return vcf_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Get SNPeff annotations for peak regions, including filtered sites and including macacaque differences"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 141,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "peaks = [(\"CAE1\",2467224),(\"CAE1\",2757224),(\"CAE1\",7005224),\n",
    "         (\"CAE2\",50381341),(\"CAE4\",5189544),(\"CAE5\",13975063),\n",
    "        (\"CAE5\",60269063),(\"CAE5\",74650063), (\"CAE6\",3823905),\n",
    "        (\"CAE6\",5564905),(\"CAE6\",11705905),(\"CAE6\",29586905),\n",
    "        (\"CAE8\",117925766),(\"CAE10\",104389305),\n",
    "        (\"CAE14\",17153077),(\"CAE14\",98746077),(\"CAE15\",62257003),\n",
    "        (\"CAE16\",1907271),(\"CAE16\",6963271),(\"CAE16\",7241271),(\"CAE16\",57481271),\n",
    "        (\"CAE16\",59802271),(\"CAE16\",68026271),(\"CAE17\",44458229),(\"CAE17\",54486229),\n",
    "        (\"CAE18\",62028538),(\"CAE20\",18933658),(\"CAE20\",106492658),(\"CAE20\",111605658),\n",
    "        (\"CAE22\",10431400),(\"CAE23\",250753),(\"CAE23\",42363534),(\"CAE24\",3453536),\n",
    "        (\"CAE24\",8944536),(\"CAE25\",61650817),(\"CAE26\",18910982),(\"CAE29\",18370875),\n",
    "        (\"CAE29\",21874875)]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 142,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "%%px --local\n",
    "def get_genotype_df(chrom, start,end):\n",
    "    vcf_fn = jn(var_ana_dir,\"163_UG_ref3500_all_sites_{}_filter_anc_mac_diff_ensembl_1.1.78.vcf.gz\".format(chrom))\n",
    "\n",
    "\n",
    "    header, info_dic = parse_vcf_header(vcf_fn)\n",
    "    samples = header[9:]\n",
    "\n",
    "    converters = {\"INFO\":get_info_dic}\n",
    "    converters.update({n:get_genotype for n in samples})#{i:get_genotype for t in range(9,len(header))}\n",
    "\n",
    "    vcf_df = get_vcf_df(vcf_fn, chrom, start, end, header=header, converters=converters)\n",
    "    \n",
    "    return vcf_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 143,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "%%px --local\n",
    "def get_genotype(s):\n",
    "    gt_str = s.split(':',1)[0]\n",
    "    if gt_str[:3] in [\"0/0\",\"0|0\"]:\n",
    "        return 0\n",
    "    elif gt_str[:3] in [\"1/1\",\"1|1\"]:\n",
    "        return 2\n",
    "    elif gt_str[:3] in [\"0/1\",\"0|1\",\"1|0\"]:\n",
    "        return 1\n",
    "    else:\n",
    "        return np.nan\n",
    "    \n",
    "def get_info_dic(s):\n",
    "    info_tuples = [t.split('=') for t in s.split(';')]\n",
    "    info_tuples = [t for t in info_tuples if len(t)==2]\n",
    "    tags = [t for t in info_tuples if len(t)!=2]\n",
    "#    try:\n",
    "    d = {k:v for (k,v) in info_tuples}\n",
    "    d.update({'_tags':tags})\n",
    "#    except ValueError:\n",
    "#        d = {}\n",
    "#        logging.warning(\"Can not parse info column at {}:{}\".format(line[0],line[1]))\n",
    "    return d\n",
    "\n",
    "def group_by_pop(col,poptype=\"pop3\"):\n",
    "    return mt.ix[col][poptype]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 144,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "%%px --local\n",
    "def try_dic(dic,key):\n",
    "    try:\n",
    "        return dic[key]\n",
    "    except KeyError:\n",
    "        return np.nan\n",
    "\n",
    "def get_ith_annotation(el,i):\n",
    "    try:\n",
    "        return el.split(',')[i]\n",
    "    except IndexError:\n",
    "        return np.nan\n",
    "def split_annotation(s):\n",
    "    return pd.Series(s.split('|')[:4]+s.split('|')[-1:],index=['allele','effect','impact','gene_symbol','errors'])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 146,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "%%px --local\n",
    "def get_variant_annot_df(vcf_df, sample_ids=None):\n",
    "    #remove non-seggregating\n",
    "    \n",
    "    if sample_ids is None:\n",
    "        sample_ids = vcf_df.columns.values[7:]\n",
    "    vcf_df = vcf_df[vcf_df['ALT'] != '.']\n",
    "\n",
    "    subpop_af_df = vcf_df[sample_ids].groupby(group_by_pop,axis=1).apply(lambda df: df.mean(axis=1)/2.)\n",
    "    annot_df = vcf_df[[\"REF\",\"ALT\",\"FILTER\"]]\n",
    "    annot_df.loc[:,'AA'] = vcf_df['INFO'].apply(lambda d:try_dic(d,\"AA\"))\n",
    "    annot_df.loc[:,'AF'] = vcf_df[sample_ids].mean(axis=1)/2.\n",
    "\n",
    "    annot_s = vcf_df['INFO'].apply(lambda d:try_dic(d,\"ANN\"))\n",
    "\n",
    "\n",
    "\n",
    "    #annotation lowest level\n",
    "    level0_df = annot_s.apply(lambda el:get_ith_annotation(el,0)).apply(split_annotation)\n",
    "    level0_df = pd.concat([annot_df,level0_df,subpop_af_df],axis=1)\n",
    "    level0_df.set_index(\"gene_symbol\",append=True,inplace=True)\n",
    "    tot_level_df = level0_df\n",
    "    i =1\n",
    "    for i in range(1,100):\n",
    "        level_info = annot_s.apply(lambda el:get_ith_annotation(el,i)).dropna()   \n",
    "        if not len(level_info):\n",
    "            break\n",
    "        level_df = level_info.apply(split_annotation)\n",
    "        level_df = annot_df.ix[level_df.index].join(level_df,how='left').join(subpop_af_df, how='left')\n",
    "        level_df.set_index(\"gene_symbol\",append=True,inplace=True)\n",
    "        tot_level_df = pd.concat([tot_level_df, level_df])\n",
    "    return tot_level_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 145,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "%%px --local\n",
    "def get_annot(chrom, pos, distance):\n",
    "    start = pos-distance\n",
    "    end = pos+distance\n",
    "    try:\n",
    "        vcf_df = get_genotype_df(chrom, start,end)\n",
    "        annot_df = get_variant_annot_df(vcf_df)\n",
    "    except Exception, e:\n",
    "        return str(e)\n",
    "    return annot_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 151,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "#look at peak regions\n",
    "peak_map = lv.map_async(lambda p:get_annot(p[0],p[1],50000),peaks)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 152,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "peak_map.wait(10)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 149,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "\"global name 'parse_vcf_header' is not defined\""
      ]
     },
     "execution_count": 149,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "peak_map.result[0]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 353,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "['Too many columns specified: expected 172 and found -1']"
      ]
     },
     "execution_count": 353,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "[err for err in peak_map.result if type(err)==str]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 355,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "annot_f = pd.concat([r for r in peak_map.result if type(r)!=str])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "ename": "NameError",
     "evalue": "name 'annot_f' is not defined",
     "output_type": "error",
     "traceback": [
      "\u001b[1;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[1;31mNameError\u001b[0m                                 Traceback (most recent call last)",
      "\u001b[1;32m<ipython-input-21-3872b95c989a>\u001b[0m in \u001b[0;36m<module>\u001b[1;34m()\u001b[0m\n\u001b[1;32m----> 1\u001b[1;33m \u001b[0mannot_f\u001b[0m\u001b[1;33m.\u001b[0m\u001b[0mshape\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n\u001b[0m",
      "\u001b[1;31mNameError\u001b[0m: name 'annot_f' is not defined"
     ]
    }
   ],
   "source": [
    "annot_f.shape"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 361,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div style=\"max-height:1000px;max-width:1500px;overflow:auto;\">\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th>REF</th>\n",
       "      <th>ALT</th>\n",
       "      <th>FILTER</th>\n",
       "      <th>AA</th>\n",
       "      <th>AF</th>\n",
       "      <th>allele</th>\n",
       "      <th>effect</th>\n",
       "      <th>impact</th>\n",
       "      <th>errors</th>\n",
       "      <th>aet</th>\n",
       "      <th>cyn</th>\n",
       "      <th>pyn</th>\n",
       "      <th>pys</th>\n",
       "      <th>sab</th>\n",
       "      <th>sac</th>\n",
       "      <th>sar</th>\n",
       "      <th>tan</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>CHROM</th>\n",
       "      <th>POS</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th rowspan=\"5\" valign=\"top\">CAE6</th>\n",
       "      <th>5576627</th>\n",
       "      <td> G</td>\n",
       "      <td> A</td>\n",
       "      <td>  PASS</td>\n",
       "      <td> NaN</td>\n",
       "      <td> 0.000000</td>\n",
       "      <td> A</td>\n",
       "      <td> missense_variant</td>\n",
       "      <td> MODERATE</td>\n",
       "      <td> </td>\n",
       "      <td> 0.000000</td>\n",
       "      <td> 0.00000</td>\n",
       "      <td> 0</td>\n",
       "      <td> 0</td>\n",
       "      <td> 0</td>\n",
       "      <td> 0</td>\n",
       "      <td> 0</td>\n",
       "      <td> 0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5576646</th>\n",
       "      <td> C</td>\n",
       "      <td> T</td>\n",
       "      <td> ExCov</td>\n",
       "      <td> NaN</td>\n",
       "      <td> 0.081250</td>\n",
       "      <td> T</td>\n",
       "      <td> missense_variant</td>\n",
       "      <td> MODERATE</td>\n",
       "      <td> </td>\n",
       "      <td> 0.766667</td>\n",
       "      <td> 0.09375</td>\n",
       "      <td> 0</td>\n",
       "      <td> 0</td>\n",
       "      <td> 0</td>\n",
       "      <td> 0</td>\n",
       "      <td> 0</td>\n",
       "      <td> 0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5576672</th>\n",
       "      <td> G</td>\n",
       "      <td> A</td>\n",
       "      <td> ExCov</td>\n",
       "      <td> NaN</td>\n",
       "      <td> 0.000000</td>\n",
       "      <td> A</td>\n",
       "      <td> missense_variant</td>\n",
       "      <td> MODERATE</td>\n",
       "      <td> </td>\n",
       "      <td> 0.000000</td>\n",
       "      <td> 0.00000</td>\n",
       "      <td> 0</td>\n",
       "      <td> 0</td>\n",
       "      <td> 0</td>\n",
       "      <td> 0</td>\n",
       "      <td> 0</td>\n",
       "      <td> 0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5576717</th>\n",
       "      <td> G</td>\n",
       "      <td> A</td>\n",
       "      <td> ExCov</td>\n",
       "      <td> NaN</td>\n",
       "      <td> 0.000000</td>\n",
       "      <td> A</td>\n",
       "      <td> missense_variant</td>\n",
       "      <td> MODERATE</td>\n",
       "      <td> </td>\n",
       "      <td> 0.000000</td>\n",
       "      <td> 0.00000</td>\n",
       "      <td> 0</td>\n",
       "      <td> 0</td>\n",
       "      <td> 0</td>\n",
       "      <td> 0</td>\n",
       "      <td> 0</td>\n",
       "      <td> 0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5586838</th>\n",
       "      <td> A</td>\n",
       "      <td> G</td>\n",
       "      <td>  PASS</td>\n",
       "      <td> NaN</td>\n",
       "      <td> 0.015924</td>\n",
       "      <td> G</td>\n",
       "      <td> missense_variant</td>\n",
       "      <td> MODERATE</td>\n",
       "      <td> </td>\n",
       "      <td> 0.000000</td>\n",
       "      <td> 0.15625</td>\n",
       "      <td> 0</td>\n",
       "      <td> 0</td>\n",
       "      <td> 0</td>\n",
       "      <td> 0</td>\n",
       "      <td> 0</td>\n",
       "      <td> 0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "              REF ALT FILTER   AA        AF allele            effect  \\\n",
       "CHROM POS                                                              \n",
       "CAE6  5576627   G   A   PASS  NaN  0.000000      A  missense_variant   \n",
       "      5576646   C   T  ExCov  NaN  0.081250      T  missense_variant   \n",
       "      5576672   G   A  ExCov  NaN  0.000000      A  missense_variant   \n",
       "      5576717   G   A  ExCov  NaN  0.000000      A  missense_variant   \n",
       "      5586838   A   G   PASS  NaN  0.015924      G  missense_variant   \n",
       "\n",
       "                 impact errors       aet      cyn  pyn  pys  sab  sac  sar  \\\n",
       "CHROM POS                                                                    \n",
       "CAE6  5576627  MODERATE         0.000000  0.00000    0    0    0    0    0   \n",
       "      5576646  MODERATE         0.766667  0.09375    0    0    0    0    0   \n",
       "      5576672  MODERATE         0.000000  0.00000    0    0    0    0    0   \n",
       "      5576717  MODERATE         0.000000  0.00000    0    0    0    0    0   \n",
       "      5586838  MODERATE         0.000000  0.15625    0    0    0    0    0   \n",
       "\n",
       "               tan  \n",
       "CHROM POS           \n",
       "CAE6  5576627    0  \n",
       "      5576646    0  \n",
       "      5576672    0  \n",
       "      5576717    0  \n",
       "      5586838    0  "
      ]
     },
     "execution_count": 361,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "annot_f[annot_f['impact']=='MODERATE'].xs('RANBP3',level=2)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 362,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div style=\"max-height:1000px;max-width:1500px;overflow:auto;\">\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th>REF</th>\n",
       "      <th>ALT</th>\n",
       "      <th>FILTER</th>\n",
       "      <th>AA</th>\n",
       "      <th>AF</th>\n",
       "      <th>allele</th>\n",
       "      <th>effect</th>\n",
       "      <th>impact</th>\n",
       "      <th>errors</th>\n",
       "      <th>aet</th>\n",
       "      <th>cyn</th>\n",
       "      <th>pyn</th>\n",
       "      <th>pys</th>\n",
       "      <th>sab</th>\n",
       "      <th>sac</th>\n",
       "      <th>sar</th>\n",
       "      <th>tan</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>CHROM</th>\n",
       "      <th>POS</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>CAE6</th>\n",
       "      <th>11707284</th>\n",
       "      <td> T</td>\n",
       "      <td> G</td>\n",
       "      <td> PASS</td>\n",
       "      <td> NaN</td>\n",
       "      <td> 0</td>\n",
       "      <td> G</td>\n",
       "      <td> missense_variant</td>\n",
       "      <td> MODERATE</td>\n",
       "      <td> </td>\n",
       "      <td> 0</td>\n",
       "      <td> 0</td>\n",
       "      <td> 0</td>\n",
       "      <td> 0</td>\n",
       "      <td> 0</td>\n",
       "      <td> 0</td>\n",
       "      <td> 0</td>\n",
       "      <td> 0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "               REF ALT FILTER   AA  AF allele            effect    impact  \\\n",
       "CHROM POS                                                                   \n",
       "CAE6  11707284   T   G   PASS  NaN   0      G  missense_variant  MODERATE   \n",
       "\n",
       "               errors  aet  cyn  pyn  pys  sab  sac  sar  tan  \n",
       "CHROM POS                                                      \n",
       "CAE6  11707284           0    0    0    0    0    0    0    0  "
      ]
     },
     "execution_count": 362,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "annot_f[annot_f['impact']=='MODERATE'].xs('NFIX',level=2)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "annot_f[annot_f['impact']=='MODERATE'].xs('NFIX',level=2)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 309,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div style=\"max-height:1000px;max-width:1500px;overflow:auto;\">\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th>REF</th>\n",
       "      <th>ALT</th>\n",
       "      <th>FILTER</th>\n",
       "      <th>AA</th>\n",
       "      <th>AF</th>\n",
       "      <th>allele</th>\n",
       "      <th>effect</th>\n",
       "      <th>impact</th>\n",
       "      <th>errors</th>\n",
       "      <th>aet</th>\n",
       "      <th>cyn</th>\n",
       "      <th>pyn</th>\n",
       "      <th>pys</th>\n",
       "      <th>sab</th>\n",
       "      <th>sac</th>\n",
       "      <th>sar</th>\n",
       "      <th>tan</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>CHROM</th>\n",
       "      <th>POS</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th rowspan=\"61\" valign=\"top\">CAE16</th>\n",
       "      <th>6955661</th>\n",
       "      <td>      T</td>\n",
       "      <td>    C</td>\n",
       "      <td>                         ExCov</td>\n",
       "      <td> T</td>\n",
       "      <td> 0.215190</td>\n",
       "      <td>   C</td>\n",
       "      <td> upstream_gene_variant</td>\n",
       "      <td> MODIFIER</td>\n",
       "      <td> </td>\n",
       "      <td> 0.000000</td>\n",
       "      <td> 0.718750</td>\n",
       "      <td> 0.333333</td>\n",
       "      <td> 0.427083</td>\n",
       "      <td> 0.000000</td>\n",
       "      <td> 0.000000</td>\n",
       "      <td> 0</td>\n",
       "      <td> 0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6955663</th>\n",
       "      <td>      G</td>\n",
       "      <td>    A</td>\n",
       "      <td>                5bpIndel;ExCov</td>\n",
       "      <td> G</td>\n",
       "      <td> 0.035948</td>\n",
       "      <td>   A</td>\n",
       "      <td> upstream_gene_variant</td>\n",
       "      <td> MODIFIER</td>\n",
       "      <td> </td>\n",
       "      <td> 0.000000</td>\n",
       "      <td> 0.031250</td>\n",
       "      <td> 0.000000</td>\n",
       "      <td> 0.108696</td>\n",
       "      <td> 0.000000</td>\n",
       "      <td> 0.000000</td>\n",
       "      <td> 0</td>\n",
       "      <td> 0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6955667</th>\n",
       "      <td>      T</td>\n",
       "      <td>  TAG</td>\n",
       "      <td>                5bpIndel;ExCov</td>\n",
       "      <td> T</td>\n",
       "      <td> 0.613924</td>\n",
       "      <td>  AG</td>\n",
       "      <td> upstream_gene_variant</td>\n",
       "      <td> MODIFIER</td>\n",
       "      <td> </td>\n",
       "      <td> 1.000000</td>\n",
       "      <td> 1.000000</td>\n",
       "      <td> 1.000000</td>\n",
       "      <td> 1.000000</td>\n",
       "      <td> 0.000000</td>\n",
       "      <td> 0.000000</td>\n",
       "      <td> 0</td>\n",
       "      <td> 1.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6955676</th>\n",
       "      <td>      G</td>\n",
       "      <td>    C</td>\n",
       "      <td>                         ExCov</td>\n",
       "      <td> C</td>\n",
       "      <td> 0.000000</td>\n",
       "      <td>   C</td>\n",
       "      <td> upstream_gene_variant</td>\n",
       "      <td> MODIFIER</td>\n",
       "      <td> </td>\n",
       "      <td> 0.000000</td>\n",
       "      <td> 0.000000</td>\n",
       "      <td> 0.000000</td>\n",
       "      <td> 0.000000</td>\n",
       "      <td> 0.000000</td>\n",
       "      <td> 0.000000</td>\n",
       "      <td> 0</td>\n",
       "      <td> 0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6955704</th>\n",
       "      <td>      T</td>\n",
       "      <td>    C</td>\n",
       "      <td>                         ExCov</td>\n",
       "      <td> C</td>\n",
       "      <td> 0.706250</td>\n",
       "      <td>   C</td>\n",
       "      <td> upstream_gene_variant</td>\n",
       "      <td> MODIFIER</td>\n",
       "      <td> </td>\n",
       "      <td> 1.000000</td>\n",
       "      <td> 1.000000</td>\n",
       "      <td> 1.000000</td>\n",
       "      <td> 1.000000</td>\n",
       "      <td> 0.586957</td>\n",
       "      <td> 0.015152</td>\n",
       "      <td> 0</td>\n",
       "      <td> 1.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6955828</th>\n",
       "      <td>      C</td>\n",
       "      <td>    G</td>\n",
       "      <td>                      5bpIndel</td>\n",
       "      <td> G</td>\n",
       "      <td> 0.000000</td>\n",
       "      <td>   G</td>\n",
       "      <td> upstream_gene_variant</td>\n",
       "      <td> MODIFIER</td>\n",
       "      <td> </td>\n",
       "      <td> 0.000000</td>\n",
       "      <td> 0.000000</td>\n",
       "      <td> 0.000000</td>\n",
       "      <td> 0.000000</td>\n",
       "      <td> 0.000000</td>\n",
       "      <td> 0.000000</td>\n",
       "      <td> 0</td>\n",
       "      <td> 0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6955832</th>\n",
       "      <td>      T</td>\n",
       "      <td>  TCC</td>\n",
       "      <td>                      5bpIndel</td>\n",
       "      <td> T</td>\n",
       "      <td> 0.009375</td>\n",
       "      <td>  CC</td>\n",
       "      <td> upstream_gene_variant</td>\n",
       "      <td> MODIFIER</td>\n",
       "      <td> </td>\n",
       "      <td> 0.000000</td>\n",
       "      <td> 0.000000</td>\n",
       "      <td> 0.000000</td>\n",
       "      <td> 0.000000</td>\n",
       "      <td> 0.000000</td>\n",
       "      <td> 0.000000</td>\n",
       "      <td> 0</td>\n",
       "      <td> 0.136364</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6955865</th>\n",
       "      <td>      C</td>\n",
       "      <td>    T</td>\n",
       "      <td>                          PASS</td>\n",
       "      <td> C</td>\n",
       "      <td> 0.003106</td>\n",
       "      <td>   T</td>\n",
       "      <td> upstream_gene_variant</td>\n",
       "      <td> MODIFIER</td>\n",
       "      <td> </td>\n",
       "      <td> 0.000000</td>\n",
       "      <td> 0.000000</td>\n",
       "      <td> 0.000000</td>\n",
       "      <td> 0.000000</td>\n",
       "      <td> 0.000000</td>\n",
       "      <td> 0.000000</td>\n",
       "      <td> 0</td>\n",
       "      <td> 0.045455</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6956049</th>\n",
       "      <td>      T</td>\n",
       "      <td>    C</td>\n",
       "      <td>                         ExCov</td>\n",
       "      <td> C</td>\n",
       "      <td> 0.000000</td>\n",
       "      <td>   C</td>\n",
       "      <td> upstream_gene_variant</td>\n",
       "      <td> MODIFIER</td>\n",
       "      <td> </td>\n",
       "      <td> 0.000000</td>\n",
       "      <td> 0.000000</td>\n",
       "      <td> 0.000000</td>\n",
       "      <td> 0.000000</td>\n",
       "      <td> 0.000000</td>\n",
       "      <td> 0.000000</td>\n",
       "      <td> 0</td>\n",
       "      <td> 0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6956068</th>\n",
       "      <td>      G</td>\n",
       "      <td>    T</td>\n",
       "      <td>                         ExCov</td>\n",
       "      <td> G</td>\n",
       "      <td> 0.560127</td>\n",
       "      <td>   T</td>\n",
       "      <td> upstream_gene_variant</td>\n",
       "      <td> MODIFIER</td>\n",
       "      <td> </td>\n",
       "      <td> 0.968750</td>\n",
       "      <td> 0.937500</td>\n",
       "      <td> 0.750000</td>\n",
       "      <td> 0.852941</td>\n",
       "      <td> 0.000000</td>\n",
       "      <td> 0.000000</td>\n",
       "      <td> 0</td>\n",
       "      <td> 0.909091</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6956089</th>\n",
       "      <td>     AG</td>\n",
       "      <td>    A</td>\n",
       "      <td> 5bpIndel;ExCov;QualityByDepth</td>\n",
       "      <td> A</td>\n",
       "      <td> 0.018868</td>\n",
       "      <td>    </td>\n",
       "      <td> upstream_gene_variant</td>\n",
       "      <td> MODIFIER</td>\n",
       "      <td> </td>\n",
       "      <td> 0.031250</td>\n",
       "      <td> 0.000000</td>\n",
       "      <td> 0.083333</td>\n",
       "      <td> 0.030000</td>\n",
       "      <td> 0.020833</td>\n",
       "      <td> 0.000000</td>\n",
       "      <td> 0</td>\n",
       "      <td> 0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6956108</th>\n",
       "      <td>      C</td>\n",
       "      <td>    T</td>\n",
       "      <td>  ExCov;LowQual;QualityByDepth</td>\n",
       "      <td> C</td>\n",
       "      <td> 0.016026</td>\n",
       "      <td>   T</td>\n",
       "      <td> upstream_gene_variant</td>\n",
       "      <td> MODIFIER</td>\n",
       "      <td> </td>\n",
       "      <td> 0.033333</td>\n",
       "      <td> 0.000000</td>\n",
       "      <td> 0.083333</td>\n",
       "      <td> 0.030000</td>\n",
       "      <td> 0.000000</td>\n",
       "      <td> 0.000000</td>\n",
       "      <td> 0</td>\n",
       "      <td> 0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6956109</th>\n",
       "      <td>      A</td>\n",
       "      <td>    G</td>\n",
       "      <td>  ExCov;LowQual;QualityByDepth</td>\n",
       "      <td> A</td>\n",
       "      <td> 0.016129</td>\n",
       "      <td>   G</td>\n",
       "      <td> upstream_gene_variant</td>\n",
       "      <td> MODIFIER</td>\n",
       "      <td> </td>\n",
       "      <td> 0.033333</td>\n",
       "      <td> 0.000000</td>\n",
       "      <td> 0.083333</td>\n",
       "      <td> 0.030000</td>\n",
       "      <td> 0.000000</td>\n",
       "      <td> 0.000000</td>\n",
       "      <td> 0</td>\n",
       "      <td> 0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6956128</th>\n",
       "      <td>      A</td>\n",
       "      <td>   AT</td>\n",
       "      <td> 5bpIndel;ExCov;QualityByDepth</td>\n",
       "      <td> A</td>\n",
       "      <td> 0.036184</td>\n",
       "      <td>   T</td>\n",
       "      <td> upstream_gene_variant</td>\n",
       "      <td> MODIFIER</td>\n",
       "      <td> </td>\n",
       "      <td> 0.035714</td>\n",
       "      <td> 0.033333</td>\n",
       "      <td> 0.200000</td>\n",
       "      <td> 0.070000</td>\n",
       "      <td> 0.000000</td>\n",
       "      <td> 0.000000</td>\n",
       "      <td> 0</td>\n",
       "      <td> 0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6956129</th>\n",
       "      <td>      T</td>\n",
       "      <td>    G</td>\n",
       "      <td> 5bpIndel;ExCov;QualityByDepth</td>\n",
       "      <td> T</td>\n",
       "      <td> 0.019481</td>\n",
       "      <td>   G</td>\n",
       "      <td> upstream_gene_variant</td>\n",
       "      <td> MODIFIER</td>\n",
       "      <td> </td>\n",
       "      <td> 0.035714</td>\n",
       "      <td> 0.000000</td>\n",
       "      <td> 0.166667</td>\n",
       "      <td> 0.030000</td>\n",
       "      <td> 0.000000</td>\n",
       "      <td> 0.000000</td>\n",
       "      <td> 0</td>\n",
       "      <td> 0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6956242</th>\n",
       "      <td>      C</td>\n",
       "      <td>   CG</td>\n",
       "      <td>                5bpIndel;ExCov</td>\n",
       "      <td> C</td>\n",
       "      <td> 0.020000</td>\n",
       "      <td>   G</td>\n",
       "      <td> upstream_gene_variant</td>\n",
       "      <td> MODIFIER</td>\n",
       "      <td> </td>\n",
       "      <td> 0.133333</td>\n",
       "      <td> 0.000000</td>\n",
       "      <td> 0.100000</td>\n",
       "      <td> 0.000000</td>\n",
       "      <td> 0.000000</td>\n",
       "      <td> 0.000000</td>\n",
       "      <td> 0</td>\n",
       "      <td> 0.050000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6956243</th>\n",
       "      <td>      C</td>\n",
       "      <td>  A,G</td>\n",
       "      <td>                5bpIndel;ExCov</td>\n",
       "      <td> C</td>\n",
       "      <td> 0.011719</td>\n",
       "      <td>   A</td>\n",
       "      <td> upstream_gene_variant</td>\n",
       "      <td> MODIFIER</td>\n",
       "      <td> </td>\n",
       "      <td> 1.000000</td>\n",
       "      <td> 0.000000</td>\n",
       "      <td> 0.000000</td>\n",
       "      <td> 0.000000</td>\n",
       "      <td> 0.000000</td>\n",
       "      <td> 0.000000</td>\n",
       "      <td> 0</td>\n",
       "      <td> 0.250000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6956294</th>\n",
       "      <td>      G</td>\n",
       "      <td>    A</td>\n",
       "      <td>                         ExCov</td>\n",
       "      <td> G</td>\n",
       "      <td> 0.006329</td>\n",
       "      <td>   A</td>\n",
       "      <td> upstream_gene_variant</td>\n",
       "      <td> MODIFIER</td>\n",
       "      <td> </td>\n",
       "      <td> 0.000000</td>\n",
       "      <td> 0.000000</td>\n",
       "      <td> 0.000000</td>\n",
       "      <td> 0.020000</td>\n",
       "      <td> 0.000000</td>\n",
       "      <td> 0.000000</td>\n",
       "      <td> 0</td>\n",
       "      <td> 0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6956384</th>\n",
       "      <td>      C</td>\n",
       "      <td>    T</td>\n",
       "      <td>                 ExCov;LowQual</td>\n",
       "      <td> C</td>\n",
       "      <td> 0.003205</td>\n",
       "      <td>   T</td>\n",
       "      <td> upstream_gene_variant</td>\n",
       "      <td> MODIFIER</td>\n",
       "      <td> </td>\n",
       "      <td> 0.000000</td>\n",
       "      <td> 0.000000</td>\n",
       "      <td> 0.000000</td>\n",
       "      <td> 0.010000</td>\n",
       "      <td> 0.000000</td>\n",
       "      <td> 0.000000</td>\n",
       "      <td> 0</td>\n",
       "      <td> 0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6956404</th>\n",
       "      <td>      G</td>\n",
       "      <td>    A</td>\n",
       "      <td>                         ExCov</td>\n",
       "      <td> G</td>\n",
       "      <td> 0.215686</td>\n",
       "      <td>   A</td>\n",
       "      <td> upstream_gene_variant</td>\n",
       "      <td> MODIFIER</td>\n",
       "      <td> </td>\n",
       "      <td> 0.000000</td>\n",
       "      <td> 0.269231</td>\n",
       "      <td> 0.000000</td>\n",
       "      <td> 0.590000</td>\n",
       "      <td> 0.000000</td>\n",
       "      <td> 0.000000</td>\n",
       "      <td> 0</td>\n",
       "      <td> 0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6956431</th>\n",
       "      <td>      G</td>\n",
       "      <td>    C</td>\n",
       "      <td>                         ExCov</td>\n",
       "      <td> G</td>\n",
       "      <td> 0.013245</td>\n",
       "      <td>   C</td>\n",
       "      <td> upstream_gene_variant</td>\n",
       "      <td> MODIFIER</td>\n",
       "      <td> </td>\n",
       "      <td> 0.000000</td>\n",
       "      <td> 0.000000</td>\n",
       "      <td> 0.000000</td>\n",
       "      <td> 0.000000</td>\n",
       "      <td> 0.083333</td>\n",
       "      <td> 0.000000</td>\n",
       "      <td> 0</td>\n",
       "      <td> 0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6956546</th>\n",
       "      <td>      G</td>\n",
       "      <td>    C</td>\n",
       "      <td>                      5bpIndel</td>\n",
       "      <td> G</td>\n",
       "      <td> 0.036424</td>\n",
       "      <td>   C</td>\n",
       "      <td> upstream_gene_variant</td>\n",
       "      <td> MODIFIER</td>\n",
       "      <td> </td>\n",
       "      <td> 0.000000</td>\n",
       "      <td> 0.000000</td>\n",
       "      <td> 0.000000</td>\n",
       "      <td> 0.117021</td>\n",
       "      <td> 0.000000</td>\n",
       "      <td> 0.000000</td>\n",
       "      <td> 0</td>\n",
       "      <td> 0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6956548</th>\n",
       "      <td>      G</td>\n",
       "      <td> GCTT</td>\n",
       "      <td>                      5bpIndel</td>\n",
       "      <td> G</td>\n",
       "      <td> 0.029801</td>\n",
       "      <td> CTT</td>\n",
       "      <td> upstream_gene_variant</td>\n",
       "      <td> MODIFIER</td>\n",
       "      <td> </td>\n",
       "      <td> 0.000000</td>\n",
       "      <td> 0.000000</td>\n",
       "      <td> 0.000000</td>\n",
       "      <td> 0.093750</td>\n",
       "      <td> 0.000000</td>\n",
       "      <td> 0.000000</td>\n",
       "      <td> 0</td>\n",
       "      <td> 0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6956569</th>\n",
       "      <td>     CT</td>\n",
       "      <td>    C</td>\n",
       "      <td>                      5bpIndel</td>\n",
       "      <td> C</td>\n",
       "      <td> 0.019481</td>\n",
       "      <td>    </td>\n",
       "      <td> upstream_gene_variant</td>\n",
       "      <td> MODIFIER</td>\n",
       "      <td> </td>\n",
       "      <td> 0.000000</td>\n",
       "      <td> 0.200000</td>\n",
       "      <td> 0.000000</td>\n",
       "      <td> 0.000000</td>\n",
       "      <td> 0.000000</td>\n",
       "      <td> 0.000000</td>\n",
       "      <td> 0</td>\n",
       "      <td> 0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6956573</th>\n",
       "      <td>      C</td>\n",
       "      <td>    G</td>\n",
       "      <td>                      5bpIndel</td>\n",
       "      <td> C</td>\n",
       "      <td> 0.417219</td>\n",
       "      <td>   G</td>\n",
       "      <td> upstream_gene_variant</td>\n",
       "      <td> MODIFIER</td>\n",
       "      <td> </td>\n",
       "      <td> 0.000000</td>\n",
       "      <td> 1.000000</td>\n",
       "      <td> 0.916667</td>\n",
       "      <td> 0.988636</td>\n",
       "      <td> 0.000000</td>\n",
       "      <td> 0.000000</td>\n",
       "      <td> 0</td>\n",
       "      <td> 0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6956606</th>\n",
       "      <td>      C</td>\n",
       "      <td>    T</td>\n",
       "      <td>                          PASS</td>\n",
       "      <td> C</td>\n",
       "      <td> 0.019231</td>\n",
       "      <td>   T</td>\n",
       "      <td> upstream_gene_variant</td>\n",
       "      <td> MODIFIER</td>\n",
       "      <td> </td>\n",
       "      <td> 0.000000</td>\n",
       "      <td> 0.000000</td>\n",
       "      <td> 0.000000</td>\n",
       "      <td> 0.000000</td>\n",
       "      <td> 0.000000</td>\n",
       "      <td> 0.000000</td>\n",
       "      <td> 0</td>\n",
       "      <td> 0.272727</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6956615</th>\n",
       "      <td>      A</td>\n",
       "      <td>    G</td>\n",
       "      <td>                          PASS</td>\n",
       "      <td> G</td>\n",
       "      <td> 0.593333</td>\n",
       "      <td>   G</td>\n",
       "      <td> upstream_gene_variant</td>\n",
       "      <td> MODIFIER</td>\n",
       "      <td> </td>\n",
       "      <td> 1.000000</td>\n",
       "      <td> 1.000000</td>\n",
       "      <td> 1.000000</td>\n",
       "      <td> 1.000000</td>\n",
       "      <td> 0.000000</td>\n",
       "      <td> 0.000000</td>\n",
       "      <td> 0</td>\n",
       "      <td> 1.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6956663</th>\n",
       "      <td>      A</td>\n",
       "      <td>    G</td>\n",
       "      <td>                          PASS</td>\n",
       "      <td> A</td>\n",
       "      <td> 0.024845</td>\n",
       "      <td>   G</td>\n",
       "      <td> upstream_gene_variant</td>\n",
       "      <td> MODIFIER</td>\n",
       "      <td> </td>\n",
       "      <td> 0.000000</td>\n",
       "      <td> 0.000000</td>\n",
       "      <td> 0.000000</td>\n",
       "      <td> 0.000000</td>\n",
       "      <td> 0.020833</td>\n",
       "      <td> 0.106061</td>\n",
       "      <td> 0</td>\n",
       "      <td> 0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6956676</th>\n",
       "      <td>      G</td>\n",
       "      <td>    C</td>\n",
       "      <td>                          PASS</td>\n",
       "      <td> C</td>\n",
       "      <td> 0.000000</td>\n",
       "      <td>   C</td>\n",
       "      <td> upstream_gene_variant</td>\n",
       "      <td> MODIFIER</td>\n",
       "      <td> </td>\n",
       "      <td> 0.000000</td>\n",
       "      <td> 0.000000</td>\n",
       "      <td> 0.000000</td>\n",
       "      <td> 0.000000</td>\n",
       "      <td> 0.000000</td>\n",
       "      <td> 0.000000</td>\n",
       "      <td> 0</td>\n",
       "      <td> 0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6956677</th>\n",
       "      <td>      G</td>\n",
       "      <td>    A</td>\n",
       "      <td>                          PASS</td>\n",
       "      <td> A</td>\n",
       "      <td> 0.000000</td>\n",
       "      <td>   A</td>\n",
       "      <td> upstream_gene_variant</td>\n",
       "      <td> MODIFIER</td>\n",
       "      <td> </td>\n",
       "      <td> 0.000000</td>\n",
       "      <td> 0.000000</td>\n",
       "      <td> 0.000000</td>\n",
       "      <td> 0.000000</td>\n",
       "      <td> 0.000000</td>\n",
       "      <td> 0.000000</td>\n",
       "      <td> 0</td>\n",
       "      <td> 0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6958637</th>\n",
       "      <td>      G</td>\n",
       "      <td>    A</td>\n",
       "      <td>                          PASS</td>\n",
       "      <td> A</td>\n",
       "      <td> 0.000000</td>\n",
       "      <td>   A</td>\n",
       "      <td>        intron_variant</td>\n",
       "      <td> MODIFIER</td>\n",
       "      <td> </td>\n",
       "      <td> 0.000000</td>\n",
       "      <td> 0.000000</td>\n",
       "      <td> 0.000000</td>\n",
       "      <td> 0.000000</td>\n",
       "      <td> 0.000000</td>\n",
       "      <td> 0.000000</td>\n",
       "      <td> 0</td>\n",
       "      <td> 0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6958672</th>\n",
       "      <td>      G</td>\n",
       "      <td>    T</td>\n",
       "      <td>              5bpIndel;LowQual</td>\n",
       "      <td> T</td>\n",
       "      <td> 0.000000</td>\n",
       "      <td>   T</td>\n",
       "      <td>        intron_variant</td>\n",
       "      <td> MODIFIER</td>\n",
       "      <td> </td>\n",
       "      <td> 0.000000</td>\n",
       "      <td> 0.000000</td>\n",
       "      <td> 0.000000</td>\n",
       "      <td> 0.000000</td>\n",
       "      <td> 0.000000</td>\n",
       "      <td> 0.000000</td>\n",
       "      <td> 0</td>\n",
       "      <td> 0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6958672</th>\n",
       "      <td>      G</td>\n",
       "      <td>    T</td>\n",
       "      <td>              5bpIndel;LowQual</td>\n",
       "      <td> T</td>\n",
       "      <td> 0.000000</td>\n",
       "      <td>   T</td>\n",
       "      <td>        intron_variant</td>\n",
       "      <td> MODIFIER</td>\n",
       "      <td> </td>\n",
       "      <td> 0.062500</td>\n",
       "      <td> 0.000000</td>\n",
       "      <td> 0.000000</td>\n",
       "      <td> 0.000000</td>\n",
       "      <td> 0.000000</td>\n",
       "      <td> 0.000000</td>\n",
       "      <td> 0</td>\n",
       "      <td> 0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6958672</th>\n",
       "      <td>      G</td>\n",
       "      <td>    T</td>\n",
       "      <td>              5bpIndel;LowQual</td>\n",
       "      <td> T</td>\n",
       "      <td> 0.000000</td>\n",
       "      <td>   C</td>\n",
       "      <td>        intron_variant</td>\n",
       "      <td> MODIFIER</td>\n",
       "      <td> </td>\n",
       "      <td> 0.000000</td>\n",
       "      <td> 0.000000</td>\n",
       "      <td> 0.000000</td>\n",
       "      <td> 0.000000</td>\n",
       "      <td> 0.000000</td>\n",
       "      <td> 0.000000</td>\n",
       "      <td> 0</td>\n",
       "      <td> 0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6958672</th>\n",
       "      <td>      G</td>\n",
       "      <td>    T</td>\n",
       "      <td>              5bpIndel;LowQual</td>\n",
       "      <td> T</td>\n",
       "      <td> 0.000000</td>\n",
       "      <td>   C</td>\n",
       "      <td>        intron_variant</td>\n",
       "      <td> MODIFIER</td>\n",
       "      <td> </td>\n",
       "      <td> 0.062500</td>\n",
       "      <td> 0.000000</td>\n",
       "      <td> 0.000000</td>\n",
       "      <td> 0.000000</td>\n",
       "      <td> 0.000000</td>\n",
       "      <td> 0.000000</td>\n",
       "      <td> 0</td>\n",
       "      <td> 0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6958672</th>\n",
       "      <td>      G</td>\n",
       "      <td>   GC</td>\n",
       "      <td>                      5bpIndel</td>\n",
       "      <td> T</td>\n",
       "      <td> 0.006536</td>\n",
       "      <td>   T</td>\n",
       "      <td>        intron_variant</td>\n",
       "      <td> MODIFIER</td>\n",
       "      <td> </td>\n",
       "      <td> 0.000000</td>\n",
       "      <td> 0.000000</td>\n",
       "      <td> 0.000000</td>\n",
       "      <td> 0.000000</td>\n",
       "      <td> 0.000000</td>\n",
       "      <td> 0.000000</td>\n",
       "      <td> 0</td>\n",
       "      <td> 0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6958672</th>\n",
       "      <td>      G</td>\n",
       "      <td>   GC</td>\n",
       "      <td>                      5bpIndel</td>\n",
       "      <td> T</td>\n",
       "      <td> 0.006536</td>\n",
       "      <td>   T</td>\n",
       "      <td>        intron_variant</td>\n",
       "      <td> MODIFIER</td>\n",
       "      <td> </td>\n",
       "      <td> 0.062500</td>\n",
       "      <td> 0.000000</td>\n",
       "      <td> 0.000000</td>\n",
       "      <td> 0.000000</td>\n",
       "      <td> 0.000000</td>\n",
       "      <td> 0.000000</td>\n",
       "      <td> 0</td>\n",
       "      <td> 0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6958672</th>\n",
       "      <td>      G</td>\n",
       "      <td>   GC</td>\n",
       "      <td>                      5bpIndel</td>\n",
       "      <td> T</td>\n",
       "      <td> 0.006536</td>\n",
       "      <td>   C</td>\n",
       "      <td>        intron_variant</td>\n",
       "      <td> MODIFIER</td>\n",
       "      <td> </td>\n",
       "      <td> 0.000000</td>\n",
       "      <td> 0.000000</td>\n",
       "      <td> 0.000000</td>\n",
       "      <td> 0.000000</td>\n",
       "      <td> 0.000000</td>\n",
       "      <td> 0.000000</td>\n",
       "      <td> 0</td>\n",
       "      <td> 0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6958672</th>\n",
       "      <td>      G</td>\n",
       "      <td>   GC</td>\n",
       "      <td>                      5bpIndel</td>\n",
       "      <td> T</td>\n",
       "      <td> 0.006536</td>\n",
       "      <td>   C</td>\n",
       "      <td>        intron_variant</td>\n",
       "      <td> MODIFIER</td>\n",
       "      <td> </td>\n",
       "      <td> 0.062500</td>\n",
       "      <td> 0.000000</td>\n",
       "      <td> 0.000000</td>\n",
       "      <td> 0.000000</td>\n",
       "      <td> 0.000000</td>\n",
       "      <td> 0.000000</td>\n",
       "      <td> 0</td>\n",
       "      <td> 0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6958672</th>\n",
       "      <td>      G</td>\n",
       "      <td>    T</td>\n",
       "      <td>              5bpIndel;LowQual</td>\n",
       "      <td> T</td>\n",
       "      <td> 0.000000</td>\n",
       "      <td>   T</td>\n",
       "      <td>        intron_variant</td>\n",
       "      <td> MODIFIER</td>\n",
       "      <td> </td>\n",
       "      <td> 0.000000</td>\n",
       "      <td> 0.000000</td>\n",
       "      <td> 0.000000</td>\n",
       "      <td> 0.000000</td>\n",
       "      <td> 0.000000</td>\n",
       "      <td> 0.000000</td>\n",
       "      <td> 0</td>\n",
       "      <td> 0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6958672</th>\n",
       "      <td>      G</td>\n",
       "      <td>    T</td>\n",
       "      <td>              5bpIndel;LowQual</td>\n",
       "      <td> T</td>\n",
       "      <td> 0.000000</td>\n",
       "      <td>   T</td>\n",
       "      <td>        intron_variant</td>\n",
       "      <td> MODIFIER</td>\n",
       "      <td> </td>\n",
       "      <td> 0.062500</td>\n",
       "      <td> 0.000000</td>\n",
       "      <td> 0.000000</td>\n",
       "      <td> 0.000000</td>\n",
       "      <td> 0.000000</td>\n",
       "      <td> 0.000000</td>\n",
       "      <td> 0</td>\n",
       "      <td> 0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6958672</th>\n",
       "      <td>      G</td>\n",
       "      <td>    T</td>\n",
       "      <td>              5bpIndel;LowQual</td>\n",
       "      <td> T</td>\n",
       "      <td> 0.000000</td>\n",
       "      <td>   C</td>\n",
       "      <td>        intron_variant</td>\n",
       "      <td> MODIFIER</td>\n",
       "      <td> </td>\n",
       "      <td> 0.000000</td>\n",
       "      <td> 0.000000</td>\n",
       "      <td> 0.000000</td>\n",
       "      <td> 0.000000</td>\n",
       "      <td> 0.000000</td>\n",
       "      <td> 0.000000</td>\n",
       "      <td> 0</td>\n",
       "      <td> 0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6958672</th>\n",
       "      <td>      G</td>\n",
       "      <td>    T</td>\n",
       "      <td>              5bpIndel;LowQual</td>\n",
       "      <td> T</td>\n",
       "      <td> 0.000000</td>\n",
       "      <td>   C</td>\n",
       "      <td>        intron_variant</td>\n",
       "      <td> MODIFIER</td>\n",
       "      <td> </td>\n",
       "      <td> 0.062500</td>\n",
       "      <td> 0.000000</td>\n",
       "      <td> 0.000000</td>\n",
       "      <td> 0.000000</td>\n",
       "      <td> 0.000000</td>\n",
       "      <td> 0.000000</td>\n",
       "      <td> 0</td>\n",
       "      <td> 0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6958672</th>\n",
       "      <td>      G</td>\n",
       "      <td>   GC</td>\n",
       "      <td>                      5bpIndel</td>\n",
       "      <td> T</td>\n",
       "      <td> 0.006536</td>\n",
       "      <td>   T</td>\n",
       "      <td>        intron_variant</td>\n",
       "      <td> MODIFIER</td>\n",
       "      <td> </td>\n",
       "      <td> 0.000000</td>\n",
       "      <td> 0.000000</td>\n",
       "      <td> 0.000000</td>\n",
       "      <td> 0.000000</td>\n",
       "      <td> 0.000000</td>\n",
       "      <td> 0.000000</td>\n",
       "      <td> 0</td>\n",
       "      <td> 0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6958672</th>\n",
       "      <td>      G</td>\n",
       "      <td>   GC</td>\n",
       "      <td>                      5bpIndel</td>\n",
       "      <td> T</td>\n",
       "      <td> 0.006536</td>\n",
       "      <td>   T</td>\n",
       "      <td>        intron_variant</td>\n",
       "      <td> MODIFIER</td>\n",
       "      <td> </td>\n",
       "      <td> 0.062500</td>\n",
       "      <td> 0.000000</td>\n",
       "      <td> 0.000000</td>\n",
       "      <td> 0.000000</td>\n",
       "      <td> 0.000000</td>\n",
       "      <td> 0.000000</td>\n",
       "      <td> 0</td>\n",
       "      <td> 0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6958672</th>\n",
       "      <td>      G</td>\n",
       "      <td>   GC</td>\n",
       "      <td>                      5bpIndel</td>\n",
       "      <td> T</td>\n",
       "      <td> 0.006536</td>\n",
       "      <td>   C</td>\n",
       "      <td>        intron_variant</td>\n",
       "      <td> MODIFIER</td>\n",
       "      <td> </td>\n",
       "      <td> 0.000000</td>\n",
       "      <td> 0.000000</td>\n",
       "      <td> 0.000000</td>\n",
       "      <td> 0.000000</td>\n",
       "      <td> 0.000000</td>\n",
       "      <td> 0.000000</td>\n",
       "      <td> 0</td>\n",
       "      <td> 0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6958672</th>\n",
       "      <td>      G</td>\n",
       "      <td>   GC</td>\n",
       "      <td>                      5bpIndel</td>\n",
       "      <td> T</td>\n",
       "      <td> 0.006536</td>\n",
       "      <td>   C</td>\n",
       "      <td>        intron_variant</td>\n",
       "      <td> MODIFIER</td>\n",
       "      <td> </td>\n",
       "      <td> 0.062500</td>\n",
       "      <td> 0.000000</td>\n",
       "      <td> 0.000000</td>\n",
       "      <td> 0.000000</td>\n",
       "      <td> 0.000000</td>\n",
       "      <td> 0.000000</td>\n",
       "      <td> 0</td>\n",
       "      <td> 0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6958686</th>\n",
       "      <td> AACCCC</td>\n",
       "      <td>    A</td>\n",
       "      <td>                      5bpIndel</td>\n",
       "      <td> A</td>\n",
       "      <td> 0.003226</td>\n",
       "      <td>    </td>\n",
       "      <td>        intron_variant</td>\n",
       "      <td> MODIFIER</td>\n",
       "      <td> </td>\n",
       "      <td> 0.000000</td>\n",
       "      <td> 0.000000</td>\n",
       "      <td> 0.000000</td>\n",
       "      <td> 0.000000</td>\n",
       "      <td> 0.000000</td>\n",
       "      <td> 0.000000</td>\n",
       "      <td> 0</td>\n",
       "      <td> 0.045455</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6958703</th>\n",
       "      <td>      C</td>\n",
       "      <td>    T</td>\n",
       "      <td>                          PASS</td>\n",
       "      <td> C</td>\n",
       "      <td> 0.519481</td>\n",
       "      <td>   T</td>\n",
       "      <td>        intron_variant</td>\n",
       "      <td> MODIFIER</td>\n",
       "      <td> </td>\n",
       "      <td> 0.000000</td>\n",
       "      <td> 1.000000</td>\n",
       "      <td> 1.000000</td>\n",
       "      <td> 1.000000</td>\n",
       "      <td> 0.000000</td>\n",
       "      <td> 0.000000</td>\n",
       "      <td> 0</td>\n",
       "      <td> 1.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6958758</th>\n",
       "      <td>      C</td>\n",
       "      <td>    T</td>\n",
       "      <td>                          PASS</td>\n",
       "      <td> T</td>\n",
       "      <td> 0.000000</td>\n",
       "      <td>   T</td>\n",
       "      <td>        intron_variant</td>\n",
       "      <td> MODIFIER</td>\n",
       "      <td> </td>\n",
       "      <td> 0.000000</td>\n",
       "      <td> 0.000000</td>\n",
       "      <td> 0.000000</td>\n",
       "      <td> 0.000000</td>\n",
       "      <td> 0.000000</td>\n",
       "      <td> 0.000000</td>\n",
       "      <td> 0</td>\n",
       "      <td> 0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6958770</th>\n",
       "      <td>      C</td>\n",
       "      <td>    T</td>\n",
       "      <td>                       LowQual</td>\n",
       "      <td> C</td>\n",
       "      <td> 0.003145</td>\n",
       "      <td>   T</td>\n",
       "      <td>        intron_variant</td>\n",
       "      <td> MODIFIER</td>\n",
       "      <td> </td>\n",
       "      <td> 0.000000</td>\n",
       "      <td> 0.000000</td>\n",
       "      <td> 0.083333</td>\n",
       "      <td> 0.000000</td>\n",
       "      <td> 0.000000</td>\n",
       "      <td> 0.000000</td>\n",
       "      <td> 0</td>\n",
       "      <td> 0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6958778</th>\n",
       "      <td>      T</td>\n",
       "      <td>    G</td>\n",
       "      <td>                          PASS</td>\n",
       "      <td> G</td>\n",
       "      <td> 0.000000</td>\n",
       "      <td>   G</td>\n",
       "      <td>        intron_variant</td>\n",
       "      <td> MODIFIER</td>\n",
       "      <td> </td>\n",
       "      <td> 0.000000</td>\n",
       "      <td> 0.000000</td>\n",
       "      <td> 0.000000</td>\n",
       "      <td> 0.000000</td>\n",
       "      <td> 0.000000</td>\n",
       "      <td> 0.000000</td>\n",
       "      <td> 0</td>\n",
       "      <td> 0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6958797</th>\n",
       "      <td>      T</td>\n",
       "      <td>    C</td>\n",
       "      <td>                          PASS</td>\n",
       "      <td> C</td>\n",
       "      <td> 0.000000</td>\n",
       "      <td>   C</td>\n",
       "      <td>        intron_variant</td>\n",
       "      <td> MODIFIER</td>\n",
       "      <td> </td>\n",
       "      <td> 0.000000</td>\n",
       "      <td> 0.000000</td>\n",
       "      <td> 0.000000</td>\n",
       "      <td> 0.000000</td>\n",
       "      <td> 0.000000</td>\n",
       "      <td> 0.000000</td>\n",
       "      <td> 0</td>\n",
       "      <td> 0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6958798</th>\n",
       "      <td>      G</td>\n",
       "      <td>    A</td>\n",
       "      <td>                          PASS</td>\n",
       "      <td> A</td>\n",
       "      <td> 0.000000</td>\n",
       "      <td>   A</td>\n",
       "      <td>        intron_variant</td>\n",
       "      <td> MODIFIER</td>\n",
       "      <td> </td>\n",
       "      <td> 0.000000</td>\n",
       "      <td> 0.000000</td>\n",
       "      <td> 0.000000</td>\n",
       "      <td> 0.000000</td>\n",
       "      <td> 0.000000</td>\n",
       "      <td> 0.000000</td>\n",
       "      <td> 0</td>\n",
       "      <td> 0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6958857</th>\n",
       "      <td>      T</td>\n",
       "      <td>    C</td>\n",
       "      <td>                          PASS</td>\n",
       "      <td> C</td>\n",
       "      <td> 0.000000</td>\n",
       "      <td>   C</td>\n",
       "      <td>        intron_variant</td>\n",
       "      <td> MODIFIER</td>\n",
       "      <td> </td>\n",
       "      <td> 0.000000</td>\n",
       "      <td> 0.000000</td>\n",
       "      <td> 0.000000</td>\n",
       "      <td> 0.000000</td>\n",
       "      <td> 0.000000</td>\n",
       "      <td> 0.000000</td>\n",
       "      <td> 0</td>\n",
       "      <td> 0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6958868</th>\n",
       "      <td>      T</td>\n",
       "      <td>    C</td>\n",
       "      <td>                          PASS</td>\n",
       "      <td> C</td>\n",
       "      <td> 0.000000</td>\n",
       "      <td>   C</td>\n",
       "      <td>        intron_variant</td>\n",
       "      <td> MODIFIER</td>\n",
       "      <td> </td>\n",
       "      <td> 0.000000</td>\n",
       "      <td> 0.000000</td>\n",
       "      <td> 0.000000</td>\n",
       "      <td> 0.000000</td>\n",
       "      <td> 0.000000</td>\n",
       "      <td> 0.000000</td>\n",
       "      <td> 0</td>\n",
       "      <td> 0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6958877</th>\n",
       "      <td>      T</td>\n",
       "      <td>    A</td>\n",
       "      <td>        LowQual;QualityByDepth</td>\n",
       "      <td> T</td>\n",
       "      <td> 0.006369</td>\n",
       "      <td>   A</td>\n",
       "      <td>        intron_variant</td>\n",
       "      <td> MODIFIER</td>\n",
       "      <td> </td>\n",
       "      <td> 0.000000</td>\n",
       "      <td> 0.000000</td>\n",
       "      <td> 0.083333</td>\n",
       "      <td> 0.010000</td>\n",
       "      <td> 0.000000</td>\n",
       "      <td> 0.000000</td>\n",
       "      <td> 0</td>\n",
       "      <td> 0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6958878</th>\n",
       "      <td>      G</td>\n",
       "      <td>    A</td>\n",
       "      <td>                          PASS</td>\n",
       "      <td> G</td>\n",
       "      <td> 0.009615</td>\n",
       "      <td>   A</td>\n",
       "      <td>        intron_variant</td>\n",
       "      <td> MODIFIER</td>\n",
       "      <td> </td>\n",
       "      <td> 0.000000</td>\n",
       "      <td> 0.000000</td>\n",
       "      <td> 0.166667</td>\n",
       "      <td> 0.010204</td>\n",
       "      <td> 0.000000</td>\n",
       "      <td> 0.000000</td>\n",
       "      <td> 0</td>\n",
       "      <td> 0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6958884</th>\n",
       "      <td>      G</td>\n",
       "      <td>    C</td>\n",
       "      <td>                          PASS</td>\n",
       "      <td> C</td>\n",
       "      <td> 0.000000</td>\n",
       "      <td>   C</td>\n",
       "      <td>        intron_variant</td>\n",
       "      <td> MODIFIER</td>\n",
       "      <td> </td>\n",
       "      <td> 0.000000</td>\n",
       "      <td> 0.000000</td>\n",
       "      <td> 0.000000</td>\n",
       "      <td> 0.000000</td>\n",
       "      <td> 0.000000</td>\n",
       "      <td> 0.000000</td>\n",
       "      <td> 0</td>\n",
       "      <td> 0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6958896</th>\n",
       "      <td>      G</td>\n",
       "      <td>    A</td>\n",
       "      <td>                          PASS</td>\n",
       "      <td> A</td>\n",
       "      <td> 0.000000</td>\n",
       "      <td>   A</td>\n",
       "      <td>        intron_variant</td>\n",
       "      <td> MODIFIER</td>\n",
       "      <td> </td>\n",
       "      <td> 0.000000</td>\n",
       "      <td> 0.000000</td>\n",
       "      <td> 0.000000</td>\n",
       "      <td> 0.000000</td>\n",
       "      <td> 0.000000</td>\n",
       "      <td> 0.000000</td>\n",
       "      <td> 0</td>\n",
       "      <td> 0.000000</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>614 rows  17 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                  REF   ALT                         FILTER AA        AF  \\\n",
       "CHROM POS                                                                 \n",
       "CAE16 6955661       T     C                          ExCov  T  0.215190   \n",
       "      6955663       G     A                 5bpIndel;ExCov  G  0.035948   \n",
       "      6955667       T   TAG                 5bpIndel;ExCov  T  0.613924   \n",
       "      6955676       G     C                          ExCov  C  0.000000   \n",
       "      6955704       T     C                          ExCov  C  0.706250   \n",
       "      6955828       C     G                       5bpIndel  G  0.000000   \n",
       "      6955832       T   TCC                       5bpIndel  T  0.009375   \n",
       "      6955865       C     T                           PASS  C  0.003106   \n",
       "      6956049       T     C                          ExCov  C  0.000000   \n",
       "      6956068       G     T                          ExCov  G  0.560127   \n",
       "      6956089      AG     A  5bpIndel;ExCov;QualityByDepth  A  0.018868   \n",
       "      6956108       C     T   ExCov;LowQual;QualityByDepth  C  0.016026   \n",
       "      6956109       A     G   ExCov;LowQual;QualityByDepth  A  0.016129   \n",
       "      6956128       A    AT  5bpIndel;ExCov;QualityByDepth  A  0.036184   \n",
       "      6956129       T     G  5bpIndel;ExCov;QualityByDepth  T  0.019481   \n",
       "      6956242       C    CG                 5bpIndel;ExCov  C  0.020000   \n",
       "      6956243       C   A,G                 5bpIndel;ExCov  C  0.011719   \n",
       "      6956294       G     A                          ExCov  G  0.006329   \n",
       "      6956384       C     T                  ExCov;LowQual  C  0.003205   \n",
       "      6956404       G     A                          ExCov  G  0.215686   \n",
       "      6956431       G     C                          ExCov  G  0.013245   \n",
       "      6956546       G     C                       5bpIndel  G  0.036424   \n",
       "      6956548       G  GCTT                       5bpIndel  G  0.029801   \n",
       "      6956569      CT     C                       5bpIndel  C  0.019481   \n",
       "      6956573       C     G                       5bpIndel  C  0.417219   \n",
       "      6956606       C     T                           PASS  C  0.019231   \n",
       "      6956615       A     G                           PASS  G  0.593333   \n",
       "      6956663       A     G                           PASS  A  0.024845   \n",
       "      6956676       G     C                           PASS  C  0.000000   \n",
       "      6956677       G     A                           PASS  A  0.000000   \n",
       "...               ...   ...                            ... ..       ...   \n",
       "      6958637       G     A                           PASS  A  0.000000   \n",
       "      6958672       G     T               5bpIndel;LowQual  T  0.000000   \n",
       "      6958672       G     T               5bpIndel;LowQual  T  0.000000   \n",
       "      6958672       G     T               5bpIndel;LowQual  T  0.000000   \n",
       "      6958672       G     T               5bpIndel;LowQual  T  0.000000   \n",
       "      6958672       G    GC                       5bpIndel  T  0.006536   \n",
       "      6958672       G    GC                       5bpIndel  T  0.006536   \n",
       "      6958672       G    GC                       5bpIndel  T  0.006536   \n",
       "      6958672       G    GC                       5bpIndel  T  0.006536   \n",
       "      6958672       G     T               5bpIndel;LowQual  T  0.000000   \n",
       "      6958672       G     T               5bpIndel;LowQual  T  0.000000   \n",
       "      6958672       G     T               5bpIndel;LowQual  T  0.000000   \n",
       "      6958672       G     T               5bpIndel;LowQual  T  0.000000   \n",
       "      6958672       G    GC                       5bpIndel  T  0.006536   \n",
       "      6958672       G    GC                       5bpIndel  T  0.006536   \n",
       "      6958672       G    GC                       5bpIndel  T  0.006536   \n",
       "      6958672       G    GC                       5bpIndel  T  0.006536   \n",
       "      6958686  AACCCC     A                       5bpIndel  A  0.003226   \n",
       "      6958703       C     T                           PASS  C  0.519481   \n",
       "      6958758       C     T                           PASS  T  0.000000   \n",
       "      6958770       C     T                        LowQual  C  0.003145   \n",
       "      6958778       T     G                           PASS  G  0.000000   \n",
       "      6958797       T     C                           PASS  C  0.000000   \n",
       "      6958798       G     A                           PASS  A  0.000000   \n",
       "      6958857       T     C                           PASS  C  0.000000   \n",
       "      6958868       T     C                           PASS  C  0.000000   \n",
       "      6958877       T     A         LowQual;QualityByDepth  T  0.006369   \n",
       "      6958878       G     A                           PASS  G  0.009615   \n",
       "      6958884       G     C                           PASS  C  0.000000   \n",
       "      6958896       G     A                           PASS  A  0.000000   \n",
       "\n",
       "              allele                 effect    impact errors       aet  \\\n",
       "CHROM POS                                                                \n",
       "CAE16 6955661      C  upstream_gene_variant  MODIFIER         0.000000   \n",
       "      6955663      A  upstream_gene_variant  MODIFIER         0.000000   \n",
       "      6955667     AG  upstream_gene_variant  MODIFIER         1.000000   \n",
       "      6955676      C  upstream_gene_variant  MODIFIER         0.000000   \n",
       "      6955704      C  upstream_gene_variant  MODIFIER         1.000000   \n",
       "      6955828      G  upstream_gene_variant  MODIFIER         0.000000   \n",
       "      6955832     CC  upstream_gene_variant  MODIFIER         0.000000   \n",
       "      6955865      T  upstream_gene_variant  MODIFIER         0.000000   \n",
       "      6956049      C  upstream_gene_variant  MODIFIER         0.000000   \n",
       "      6956068      T  upstream_gene_variant  MODIFIER         0.968750   \n",
       "      6956089         upstream_gene_variant  MODIFIER         0.031250   \n",
       "      6956108      T  upstream_gene_variant  MODIFIER         0.033333   \n",
       "      6956109      G  upstream_gene_variant  MODIFIER         0.033333   \n",
       "      6956128      T  upstream_gene_variant  MODIFIER         0.035714   \n",
       "      6956129      G  upstream_gene_variant  MODIFIER         0.035714   \n",
       "      6956242      G  upstream_gene_variant  MODIFIER         0.133333   \n",
       "      6956243      A  upstream_gene_variant  MODIFIER         1.000000   \n",
       "      6956294      A  upstream_gene_variant  MODIFIER         0.000000   \n",
       "      6956384      T  upstream_gene_variant  MODIFIER         0.000000   \n",
       "      6956404      A  upstream_gene_variant  MODIFIER         0.000000   \n",
       "      6956431      C  upstream_gene_variant  MODIFIER         0.000000   \n",
       "      6956546      C  upstream_gene_variant  MODIFIER         0.000000   \n",
       "      6956548    CTT  upstream_gene_variant  MODIFIER         0.000000   \n",
       "      6956569         upstream_gene_variant  MODIFIER         0.000000   \n",
       "      6956573      G  upstream_gene_variant  MODIFIER         0.000000   \n",
       "      6956606      T  upstream_gene_variant  MODIFIER         0.000000   \n",
       "      6956615      G  upstream_gene_variant  MODIFIER         1.000000   \n",
       "      6956663      G  upstream_gene_variant  MODIFIER         0.000000   \n",
       "      6956676      C  upstream_gene_variant  MODIFIER         0.000000   \n",
       "      6956677      A  upstream_gene_variant  MODIFIER         0.000000   \n",
       "...              ...                    ...       ...    ...       ...   \n",
       "      6958637      A         intron_variant  MODIFIER         0.000000   \n",
       "      6958672      T         intron_variant  MODIFIER         0.000000   \n",
       "      6958672      T         intron_variant  MODIFIER         0.062500   \n",
       "      6958672      C         intron_variant  MODIFIER         0.000000   \n",
       "      6958672      C         intron_variant  MODIFIER         0.062500   \n",
       "      6958672      T         intron_variant  MODIFIER         0.000000   \n",
       "      6958672      T         intron_variant  MODIFIER         0.062500   \n",
       "      6958672      C         intron_variant  MODIFIER         0.000000   \n",
       "      6958672      C         intron_variant  MODIFIER         0.062500   \n",
       "      6958672      T         intron_variant  MODIFIER         0.000000   \n",
       "      6958672      T         intron_variant  MODIFIER         0.062500   \n",
       "      6958672      C         intron_variant  MODIFIER         0.000000   \n",
       "      6958672      C         intron_variant  MODIFIER         0.062500   \n",
       "      6958672      T         intron_variant  MODIFIER         0.000000   \n",
       "      6958672      T         intron_variant  MODIFIER         0.062500   \n",
       "      6958672      C         intron_variant  MODIFIER         0.000000   \n",
       "      6958672      C         intron_variant  MODIFIER         0.062500   \n",
       "      6958686                intron_variant  MODIFIER         0.000000   \n",
       "      6958703      T         intron_variant  MODIFIER         0.000000   \n",
       "      6958758      T         intron_variant  MODIFIER         0.000000   \n",
       "      6958770      T         intron_variant  MODIFIER         0.000000   \n",
       "      6958778      G         intron_variant  MODIFIER         0.000000   \n",
       "      6958797      C         intron_variant  MODIFIER         0.000000   \n",
       "      6958798      A         intron_variant  MODIFIER         0.000000   \n",
       "      6958857      C         intron_variant  MODIFIER         0.000000   \n",
       "      6958868      C         intron_variant  MODIFIER         0.000000   \n",
       "      6958877      A         intron_variant  MODIFIER         0.000000   \n",
       "      6958878      A         intron_variant  MODIFIER         0.000000   \n",
       "      6958884      C         intron_variant  MODIFIER         0.000000   \n",
       "      6958896      A         intron_variant  MODIFIER         0.000000   \n",
       "\n",
       "                    cyn       pyn       pys       sab       sac  sar       tan  \n",
       "CHROM POS                                                                       \n",
       "CAE16 6955661  0.718750  0.333333  0.427083  0.000000  0.000000    0  0.000000  \n",
       "      6955663  0.031250  0.000000  0.108696  0.000000  0.000000    0  0.000000  \n",
       "      6955667  1.000000  1.000000  1.000000  0.000000  0.000000    0  1.000000  \n",
       "      6955676  0.000000  0.000000  0.000000  0.000000  0.000000    0  0.000000  \n",
       "      6955704  1.000000  1.000000  1.000000  0.586957  0.015152    0  1.000000  \n",
       "      6955828  0.000000  0.000000  0.000000  0.000000  0.000000    0  0.000000  \n",
       "      6955832  0.000000  0.000000  0.000000  0.000000  0.000000    0  0.136364  \n",
       "      6955865  0.000000  0.000000  0.000000  0.000000  0.000000    0  0.045455  \n",
       "      6956049  0.000000  0.000000  0.000000  0.000000  0.000000    0  0.000000  \n",
       "      6956068  0.937500  0.750000  0.852941  0.000000  0.000000    0  0.909091  \n",
       "      6956089  0.000000  0.083333  0.030000  0.020833  0.000000    0  0.000000  \n",
       "      6956108  0.000000  0.083333  0.030000  0.000000  0.000000    0  0.000000  \n",
       "      6956109  0.000000  0.083333  0.030000  0.000000  0.000000    0  0.000000  \n",
       "      6956128  0.033333  0.200000  0.070000  0.000000  0.000000    0  0.000000  \n",
       "      6956129  0.000000  0.166667  0.030000  0.000000  0.000000    0  0.000000  \n",
       "      6956242  0.000000  0.100000  0.000000  0.000000  0.000000    0  0.050000  \n",
       "      6956243  0.000000  0.000000  0.000000  0.000000  0.000000    0  0.250000  \n",
       "      6956294  0.000000  0.000000  0.020000  0.000000  0.000000    0  0.000000  \n",
       "      6956384  0.000000  0.000000  0.010000  0.000000  0.000000    0  0.000000  \n",
       "      6956404  0.269231  0.000000  0.590000  0.000000  0.000000    0  0.000000  \n",
       "      6956431  0.000000  0.000000  0.000000  0.083333  0.000000    0  0.000000  \n",
       "      6956546  0.000000  0.000000  0.117021  0.000000  0.000000    0  0.000000  \n",
       "      6956548  0.000000  0.000000  0.093750  0.000000  0.000000    0  0.000000  \n",
       "      6956569  0.200000  0.000000  0.000000  0.000000  0.000000    0  0.000000  \n",
       "      6956573  1.000000  0.916667  0.988636  0.000000  0.000000    0  0.000000  \n",
       "      6956606  0.000000  0.000000  0.000000  0.000000  0.000000    0  0.272727  \n",
       "      6956615  1.000000  1.000000  1.000000  0.000000  0.000000    0  1.000000  \n",
       "      6956663  0.000000  0.000000  0.000000  0.020833  0.106061    0  0.000000  \n",
       "      6956676  0.000000  0.000000  0.000000  0.000000  0.000000    0  0.000000  \n",
       "      6956677  0.000000  0.000000  0.000000  0.000000  0.000000    0  0.000000  \n",
       "...                 ...       ...       ...       ...       ...  ...       ...  \n",
       "      6958637  0.000000  0.000000  0.000000  0.000000  0.000000    0  0.000000  \n",
       "      6958672  0.000000  0.000000  0.000000  0.000000  0.000000    0  0.000000  \n",
       "      6958672  0.000000  0.000000  0.000000  0.000000  0.000000    0  0.000000  \n",
       "      6958672  0.000000  0.000000  0.000000  0.000000  0.000000    0  0.000000  \n",
       "      6958672  0.000000  0.000000  0.000000  0.000000  0.000000    0  0.000000  \n",
       "      6958672  0.000000  0.000000  0.000000  0.000000  0.000000    0  0.000000  \n",
       "      6958672  0.000000  0.000000  0.000000  0.000000  0.000000    0  0.000000  \n",
       "      6958672  0.000000  0.000000  0.000000  0.000000  0.000000    0  0.000000  \n",
       "      6958672  0.000000  0.000000  0.000000  0.000000  0.000000    0  0.000000  \n",
       "      6958672  0.000000  0.000000  0.000000  0.000000  0.000000    0  0.000000  \n",
       "      6958672  0.000000  0.000000  0.000000  0.000000  0.000000    0  0.000000  \n",
       "      6958672  0.000000  0.000000  0.000000  0.000000  0.000000    0  0.000000  \n",
       "      6958672  0.000000  0.000000  0.000000  0.000000  0.000000    0  0.000000  \n",
       "      6958672  0.000000  0.000000  0.000000  0.000000  0.000000    0  0.000000  \n",
       "      6958672  0.000000  0.000000  0.000000  0.000000  0.000000    0  0.000000  \n",
       "      6958672  0.000000  0.000000  0.000000  0.000000  0.000000    0  0.000000  \n",
       "      6958672  0.000000  0.000000  0.000000  0.000000  0.000000    0  0.000000  \n",
       "      6958686  0.000000  0.000000  0.000000  0.000000  0.000000    0  0.045455  \n",
       "      6958703  1.000000  1.000000  1.000000  0.000000  0.000000    0  1.000000  \n",
       "      6958758  0.000000  0.000000  0.000000  0.000000  0.000000    0  0.000000  \n",
       "      6958770  0.000000  0.083333  0.000000  0.000000  0.000000    0  0.000000  \n",
       "      6958778  0.000000  0.000000  0.000000  0.000000  0.000000    0  0.000000  \n",
       "      6958797  0.000000  0.000000  0.000000  0.000000  0.000000    0  0.000000  \n",
       "      6958798  0.000000  0.000000  0.000000  0.000000  0.000000    0  0.000000  \n",
       "      6958857  0.000000  0.000000  0.000000  0.000000  0.000000    0  0.000000  \n",
       "      6958868  0.000000  0.000000  0.000000  0.000000  0.000000    0  0.000000  \n",
       "      6958877  0.000000  0.083333  0.010000  0.000000  0.000000    0  0.000000  \n",
       "      6958878  0.000000  0.166667  0.010204  0.000000  0.000000    0  0.000000  \n",
       "      6958884  0.000000  0.000000  0.000000  0.000000  0.000000    0  0.000000  \n",
       "      6958896  0.000000  0.000000  0.000000  0.000000  0.000000    0  0.000000  \n",
       "\n",
       "[614 rows x 17 columns]"
      ]
     },
     "execution_count": 309,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "define a metric for af differentiation...."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 310,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div style=\"max-height:1000px;max-width:1500px;overflow:auto;\">\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th>REF</th>\n",
       "      <th>ALT</th>\n",
       "      <th>FILTER</th>\n",
       "      <th>AA</th>\n",
       "      <th>AF</th>\n",
       "      <th>allele</th>\n",
       "      <th>effect</th>\n",
       "      <th>impact</th>\n",
       "      <th>errors</th>\n",
       "      <th>aet</th>\n",
       "      <th>cyn</th>\n",
       "      <th>pyn</th>\n",
       "      <th>pys</th>\n",
       "      <th>sab</th>\n",
       "      <th>sac</th>\n",
       "      <th>sar</th>\n",
       "      <th>tan</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>CHROM</th>\n",
       "      <th>POS</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th rowspan=\"13\" valign=\"top\">CAE16</th>\n",
       "      <th>6957216</th>\n",
       "      <td> G</td>\n",
       "      <td> A</td>\n",
       "      <td>         ExCov</td>\n",
       "      <td> G</td>\n",
       "      <td> 0.003205</td>\n",
       "      <td> A</td>\n",
       "      <td> missense_variant</td>\n",
       "      <td> MODERATE</td>\n",
       "      <td> </td>\n",
       "      <td> 0.03125</td>\n",
       "      <td> 0.000000</td>\n",
       "      <td> 0</td>\n",
       "      <td> 0.000000</td>\n",
       "      <td> 0.000000</td>\n",
       "      <td> 0.000000</td>\n",
       "      <td> 0.0</td>\n",
       "      <td> 0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6957221</th>\n",
       "      <td> G</td>\n",
       "      <td> A</td>\n",
       "      <td>         ExCov</td>\n",
       "      <td> G</td>\n",
       "      <td> 0.009615</td>\n",
       "      <td> A</td>\n",
       "      <td> missense_variant</td>\n",
       "      <td> MODERATE</td>\n",
       "      <td> </td>\n",
       "      <td> 0.00000</td>\n",
       "      <td> 0.000000</td>\n",
       "      <td> 0</td>\n",
       "      <td> 0.030000</td>\n",
       "      <td> 0.000000</td>\n",
       "      <td> 0.000000</td>\n",
       "      <td> 0.0</td>\n",
       "      <td> 0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6957354</th>\n",
       "      <td> G</td>\n",
       "      <td> A</td>\n",
       "      <td>         ExCov</td>\n",
       "      <td> G</td>\n",
       "      <td> 0.056962</td>\n",
       "      <td> A</td>\n",
       "      <td> missense_variant</td>\n",
       "      <td> MODERATE</td>\n",
       "      <td> </td>\n",
       "      <td> 0.56250</td>\n",
       "      <td> 0.000000</td>\n",
       "      <td> 0</td>\n",
       "      <td> 0.000000</td>\n",
       "      <td> 0.000000</td>\n",
       "      <td> 0.000000</td>\n",
       "      <td> 0.0</td>\n",
       "      <td> 0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6957485</th>\n",
       "      <td> A</td>\n",
       "      <td> G</td>\n",
       "      <td>         ExCov</td>\n",
       "      <td> A</td>\n",
       "      <td> 0.022727</td>\n",
       "      <td> G</td>\n",
       "      <td> missense_variant</td>\n",
       "      <td> MODERATE</td>\n",
       "      <td> </td>\n",
       "      <td> 0.21875</td>\n",
       "      <td> 0.000000</td>\n",
       "      <td> 0</td>\n",
       "      <td> 0.000000</td>\n",
       "      <td> 0.000000</td>\n",
       "      <td> 0.000000</td>\n",
       "      <td> 0.0</td>\n",
       "      <td> 0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6957509</th>\n",
       "      <td> G</td>\n",
       "      <td> A</td>\n",
       "      <td>         ExCov</td>\n",
       "      <td> A</td>\n",
       "      <td> 0.000000</td>\n",
       "      <td> A</td>\n",
       "      <td> missense_variant</td>\n",
       "      <td> MODERATE</td>\n",
       "      <td> </td>\n",
       "      <td> 0.00000</td>\n",
       "      <td> 0.000000</td>\n",
       "      <td> 0</td>\n",
       "      <td> 0.000000</td>\n",
       "      <td> 0.000000</td>\n",
       "      <td> 0.000000</td>\n",
       "      <td> 0.0</td>\n",
       "      <td> 0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6957557</th>\n",
       "      <td> A</td>\n",
       "      <td> G</td>\n",
       "      <td>         ExCov</td>\n",
       "      <td> G</td>\n",
       "      <td> 0.000000</td>\n",
       "      <td> G</td>\n",
       "      <td> missense_variant</td>\n",
       "      <td> MODERATE</td>\n",
       "      <td> </td>\n",
       "      <td> 0.00000</td>\n",
       "      <td> 0.000000</td>\n",
       "      <td> 0</td>\n",
       "      <td> 0.000000</td>\n",
       "      <td> 0.000000</td>\n",
       "      <td> 0.000000</td>\n",
       "      <td> 0.0</td>\n",
       "      <td> 0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6957623</th>\n",
       "      <td> C</td>\n",
       "      <td> T</td>\n",
       "      <td>         ExCov</td>\n",
       "      <td> C</td>\n",
       "      <td> 0.022876</td>\n",
       "      <td> T</td>\n",
       "      <td> missense_variant</td>\n",
       "      <td> MODERATE</td>\n",
       "      <td> </td>\n",
       "      <td> 0.00000</td>\n",
       "      <td> 0.033333</td>\n",
       "      <td> 0</td>\n",
       "      <td> 0.058824</td>\n",
       "      <td> 0.000000</td>\n",
       "      <td> 0.000000</td>\n",
       "      <td> 0.0</td>\n",
       "      <td> 0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6957677</th>\n",
       "      <td> A</td>\n",
       "      <td> G</td>\n",
       "      <td> ExCov;LowQual</td>\n",
       "      <td> G</td>\n",
       "      <td> 0.000000</td>\n",
       "      <td> G</td>\n",
       "      <td> missense_variant</td>\n",
       "      <td> MODERATE</td>\n",
       "      <td> </td>\n",
       "      <td> 0.00000</td>\n",
       "      <td> 0.000000</td>\n",
       "      <td> 0</td>\n",
       "      <td> 0.000000</td>\n",
       "      <td> 0.000000</td>\n",
       "      <td> 0.000000</td>\n",
       "      <td> 0.0</td>\n",
       "      <td> 0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6957762</th>\n",
       "      <td> T</td>\n",
       "      <td> C</td>\n",
       "      <td>         ExCov</td>\n",
       "      <td> T</td>\n",
       "      <td> 0.097484</td>\n",
       "      <td> C</td>\n",
       "      <td> missense_variant</td>\n",
       "      <td> MODERATE</td>\n",
       "      <td> </td>\n",
       "      <td> 0.00000</td>\n",
       "      <td> 0.000000</td>\n",
       "      <td> 0</td>\n",
       "      <td> 0.303922</td>\n",
       "      <td> 0.000000</td>\n",
       "      <td> 0.000000</td>\n",
       "      <td> 0.0</td>\n",
       "      <td> 0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6957774</th>\n",
       "      <td> C</td>\n",
       "      <td> T</td>\n",
       "      <td>         ExCov</td>\n",
       "      <td> T</td>\n",
       "      <td> 0.000000</td>\n",
       "      <td> T</td>\n",
       "      <td> missense_variant</td>\n",
       "      <td> MODERATE</td>\n",
       "      <td> </td>\n",
       "      <td> 0.00000</td>\n",
       "      <td> 0.000000</td>\n",
       "      <td> 0</td>\n",
       "      <td> 0.000000</td>\n",
       "      <td> 0.000000</td>\n",
       "      <td> 0.000000</td>\n",
       "      <td> 0.0</td>\n",
       "      <td> 0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6957835</th>\n",
       "      <td> G</td>\n",
       "      <td> A</td>\n",
       "      <td> ExCov;LowQual</td>\n",
       "      <td> G</td>\n",
       "      <td> 0.003185</td>\n",
       "      <td> A</td>\n",
       "      <td> missense_variant</td>\n",
       "      <td> MODERATE</td>\n",
       "      <td> </td>\n",
       "      <td> 0.00000</td>\n",
       "      <td> 0.000000</td>\n",
       "      <td> 0</td>\n",
       "      <td> 0.000000</td>\n",
       "      <td> 0.000000</td>\n",
       "      <td> 0.000000</td>\n",
       "      <td> 0.0</td>\n",
       "      <td> 0.045455</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6958256</th>\n",
       "      <td> T</td>\n",
       "      <td> C</td>\n",
       "      <td>         ExCov</td>\n",
       "      <td> C</td>\n",
       "      <td> 0.000000</td>\n",
       "      <td> C</td>\n",
       "      <td> missense_variant</td>\n",
       "      <td> MODERATE</td>\n",
       "      <td> </td>\n",
       "      <td> 0.00000</td>\n",
       "      <td> 0.000000</td>\n",
       "      <td> 0</td>\n",
       "      <td> 0.000000</td>\n",
       "      <td> 0.000000</td>\n",
       "      <td> 0.000000</td>\n",
       "      <td> 0.0</td>\n",
       "      <td> 0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6958279</th>\n",
       "      <td> G</td>\n",
       "      <td> C</td>\n",
       "      <td>         ExCov</td>\n",
       "      <td> G</td>\n",
       "      <td> 0.276730</td>\n",
       "      <td> C</td>\n",
       "      <td> missense_variant</td>\n",
       "      <td> MODERATE</td>\n",
       "      <td> </td>\n",
       "      <td> 0.00000</td>\n",
       "      <td> 0.000000</td>\n",
       "      <td> 0</td>\n",
       "      <td> 0.000000</td>\n",
       "      <td> 0.708333</td>\n",
       "      <td> 0.758065</td>\n",
       "      <td> 0.7</td>\n",
       "      <td> 0.000000</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "              REF ALT         FILTER AA        AF allele            effect  \\\n",
       "CHROM POS                                                                    \n",
       "CAE16 6957216   G   A          ExCov  G  0.003205      A  missense_variant   \n",
       "      6957221   G   A          ExCov  G  0.009615      A  missense_variant   \n",
       "      6957354   G   A          ExCov  G  0.056962      A  missense_variant   \n",
       "      6957485   A   G          ExCov  A  0.022727      G  missense_variant   \n",
       "      6957509   G   A          ExCov  A  0.000000      A  missense_variant   \n",
       "      6957557   A   G          ExCov  G  0.000000      G  missense_variant   \n",
       "      6957623   C   T          ExCov  C  0.022876      T  missense_variant   \n",
       "      6957677   A   G  ExCov;LowQual  G  0.000000      G  missense_variant   \n",
       "      6957762   T   C          ExCov  T  0.097484      C  missense_variant   \n",
       "      6957774   C   T          ExCov  T  0.000000      T  missense_variant   \n",
       "      6957835   G   A  ExCov;LowQual  G  0.003185      A  missense_variant   \n",
       "      6958256   T   C          ExCov  C  0.000000      C  missense_variant   \n",
       "      6958279   G   C          ExCov  G  0.276730      C  missense_variant   \n",
       "\n",
       "                 impact errors      aet       cyn  pyn       pys       sab  \\\n",
       "CHROM POS                                                                    \n",
       "CAE16 6957216  MODERATE         0.03125  0.000000    0  0.000000  0.000000   \n",
       "      6957221  MODERATE         0.00000  0.000000    0  0.030000  0.000000   \n",
       "      6957354  MODERATE         0.56250  0.000000    0  0.000000  0.000000   \n",
       "      6957485  MODERATE         0.21875  0.000000    0  0.000000  0.000000   \n",
       "      6957509  MODERATE         0.00000  0.000000    0  0.000000  0.000000   \n",
       "      6957557  MODERATE         0.00000  0.000000    0  0.000000  0.000000   \n",
       "      6957623  MODERATE         0.00000  0.033333    0  0.058824  0.000000   \n",
       "      6957677  MODERATE         0.00000  0.000000    0  0.000000  0.000000   \n",
       "      6957762  MODERATE         0.00000  0.000000    0  0.303922  0.000000   \n",
       "      6957774  MODERATE         0.00000  0.000000    0  0.000000  0.000000   \n",
       "      6957835  MODERATE         0.00000  0.000000    0  0.000000  0.000000   \n",
       "      6958256  MODERATE         0.00000  0.000000    0  0.000000  0.000000   \n",
       "      6958279  MODERATE         0.00000  0.000000    0  0.000000  0.708333   \n",
       "\n",
       "                    sac  sar       tan  \n",
       "CHROM POS                               \n",
       "CAE16 6957216  0.000000  0.0  0.000000  \n",
       "      6957221  0.000000  0.0  0.000000  \n",
       "      6957354  0.000000  0.0  0.000000  \n",
       "      6957485  0.000000  0.0  0.000000  \n",
       "      6957509  0.000000  0.0  0.000000  \n",
       "      6957557  0.000000  0.0  0.000000  \n",
       "      6957623  0.000000  0.0  0.000000  \n",
       "      6957677  0.000000  0.0  0.000000  \n",
       "      6957762  0.000000  0.0  0.000000  \n",
       "      6957774  0.000000  0.0  0.000000  \n",
       "      6957835  0.000000  0.0  0.045455  \n",
       "      6958256  0.000000  0.0  0.000000  \n",
       "      6958279  0.758065  0.7  0.000000  "
      ]
     },
     "execution_count": 310,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "tot_level_df[tot_level_df[\"impact\"]=='MODERATE'].xs(\"CD68\",level=2)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 37,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "chrom = \"CAE1\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 38,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "variant = open(jn(var_ana_dir,\"163_UG_ref3500_all_sites_{}_filter_anc_mac_diff_ensembl_1.1.78.vcf.gz\".format(chrom)))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 40,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "t = tabix.open(variant.name)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "collapsed": true
   },
   "source": [
    "## For all genes: get for \"pass\" variants number of syn, nonsyn, high impact and allele frequencies in each category in all pops"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 118,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "%%px --local\n",
    "from pypopgen.modules import vcfpandas"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 119,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "%%px --local\n",
    "import gc\n",
    "def get_annot_pass_all_genes(chrom):\n",
    "    try:\n",
    "        fn = jn(var_ana_dir,\n",
    "                            \"GATK_UG_163NonVRC_ref3500_{}_snp_pass\"\n",
    "                            \"_beagle4_shapeit_notrb_ensembl_1.1.78.vcf.gz\".format(chrom))\n",
    "\n",
    "        header, _ = vcfpandas.parse_vcf_header(fn)\n",
    "        samples = header[9:]\n",
    "\n",
    "\n",
    "        converters = {\"INFO\":vcfpandas.get_info_dic}\n",
    "        converters.update(vcfpandas.converters.genotype_converter(samples))\n",
    "\n",
    "        vcf_df = vcfpandas.get_vcf_df(fn, chrom, converters=converters,\n",
    "                                                                   header=header)\n",
    "\n",
    "        annot_df = get_variant_annot_df(vcf_df)\n",
    "        del vcf_df\n",
    "        annot_df.to_csv(jn(variant_impact_dir,\"163_snp_pas_ensembl_1.1.78_variant_impact_{}.tsv\".format(chrom)),sep=\"\\t\")\n",
    "        del annot_df\n",
    "        gc.collect()\n",
    "        return chrom\n",
    "    except Exception, e:\n",
    "        return str(e)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 73,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "array(['CAE18', 'CAE1', 'CAE4', 'CAE8', 'CAE7', 'CAE26', 'CAE12', 'CAE9',\n",
       "       'CAEX', 'CAE16', 'CAE2', 'CAE25', 'CAE20', 'CAE24', 'CAE3', 'CAE29',\n",
       "       'CAE15', 'CAE17', 'CAE27', 'CAE23', 'CAE22', 'CAE11', 'CAE14',\n",
       "       'CAE6', 'CAE19', 'CAE5', 'CAE13', 'CAE21', 'CAE10'], \n",
       "      dtype='|S5')"
      ]
     },
     "execution_count": 73,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 124,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "annot_all_genome = lv.map_async(get_annot_pass_all_genes,['CAE25', 'CAE24', 'CAE22'])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 50,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "chrom = 'CAE28'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 132,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "impact_var_df = pd.DataFrame()\n",
    "for chrom in chromosomes[:-1]:\n",
    "    annot_df = pd.read_csv(jn(variant_impact_dir,\"163_snp_pas_ensembl_1.1.78_variant_impact_{}.tsv\".format(chrom)),sep=\"\\t\",index_col=[0,1])\n",
    "    impact_var_df = pd.concat([impact_var_df,annot_df[annot_df[\"impact\"] != \"MODIFIER\"]])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 135,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "get dn/ds"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 171,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "def get_s_n(df):\n",
    "    missense = df[df['effect']=='missense_variant']\n",
    "    af_std = missense[\"af_std\"].mean()\n",
    "    return pd.Series({\"s\":(df['effect']=='synonymous_variant').sum(),\"n\":len(missense),\"n_af_std\":af_std})"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 172,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "sn_df = impact_var_df.groupby('gene_symbol').apply(get_s_n)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 182,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "sn_df[\"diff\"] = (sn_df[\"n\"]-sn_df[\"s\"])/(sn_df[\"s\"]+sn_df[\"n\"])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 199,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "#sn_df[\"diff\"] = (sn_df[\"diff\"]-sn_df[\"diff\"].mean())/sn_df[\"diff\"].std()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 203,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "<matplotlib.collections.PathCollection at 0x1d434b90>"
      ]
     },
     "execution_count": 203,
     "metadata": {},
     "output_type": "execute_result"
    },
    {
     "data": {
      "image/png": "iVBORw0KGgoAAAANSUhEUgAAAX4AAAEACAYAAAC08h1NAAAABHNCSVQICAgIfAhkiAAAAAlwSFlz\nAAALEgAACxIB0t1+/AAAIABJREFUeJztvX94XVd55/tdkixbtmRLR5JtJbIdckpQjA1RYKgYdR6p\nJY4CTydM7GFoqecKWkgyQJ0fx4njC7QeotyUNkkphTaEBtsXOoWZyQ3XPBQpBsZ+mvRSSnCMKeQn\nIRMnMRDSkIQIFFvv/eNdr9baa699dHSOfh2d9/M8enR+7LP32nuf811rvb+WISIoiqIotUPdQjdA\nURRFmV9U+BVFUWoMFX5FUZQaQ4VfURSlxlDhVxRFqTFU+BVFUWqMioXfGHOJMeZBY8wjxpg9kfcH\njTE/N8Ycs38frvSYiqIoSvk0VPJhY0w9gE8CuAjAUwD+2RhziIh+EGx6lIgureRYiqIoyuxQ6Yj/\nTQAeJaIfEdErAL4A4O2R7UyFx1EURVFmiUqF/2wAT3rPT9rXfAjAvzXGHDfG/L0xZnOFx1QURVEq\noCJTD1jUp+M7ADYQ0cvGmLcC+BKA8yo8rqIoilImlQr/UwA2eM83gEf9UxDRi97jrxpj/soYkyOi\n5/ztjDFaNEhRFGWGENGMTemVmnq+DeDVxphzjDGNAN4J4JC/gTFmnTHG2MdvAmBC0ReIaEn+/fEf\n//GCt0HPT89Pz2/p/ZVLRSN+IjptjPkggDEA9QDuJKIfGGOusO9/GsB/BPBfjDGnAbwM4HcqOaai\nKIpSGZWaekBEXwXw1eC1T3uPPwXgU5UeR1EURZkdNHN3HhgcHFzoJswpen7VjZ5f7WEqsRPNJsYY\nWixtURRFqQaMMaAFcO4qiqIoVYYKv6IoSo2hwq8oilJjqPAriqLUGCr8iqIoNYYKv6IoSo2hwq8o\nilJjqPAriqLUGCr8iqIoNYYKv6IoSo2hwq8oilJjqPAriqLUGCr8iqIoNYYKv6IoSo2hwq8oilJj\nqPAriqLUGCr8iqIoNYYKv6IoSo2hwq8oilJjqPAriqLUGCr8iqIoNYYK/yJlbGwMF1+8AxdfvANj\nY2ML3RxFUZYQhogWug0AAGMMLZa2LDRjY2O47LJhjI9/DADQ1LQHd999EENDQwvcMkVRFhPGGBCR\nmfHnFovYqvA7Lr54Bw4fvhTAsH3lILZtO4R77rlrIZulKMoio1zhV1OPoihKjdGw0A1Q0hQKl+Pe\ne4cxPs7Pm5r2oFA4uLCNUhRlyVDxiN8Yc4kx5kFjzCPGmD1Ftvs3xpjTxpjtlR5zqTM0NIS772bz\nzrZth9S+ryjKrFKRjd8YUw/gIQAXAXgKwD8D+F0i+kFku8MAXgawn4hSxmq18SuKosyMhbLxvwnA\no0T0IyJ6BcAXALw9st0fAvifAH5a4fEURVGUCqlU+M8G8KT3/KR9bQpjzNngzuCv7Us6rFcURVlA\nKhX+UkT84wBusHYcY/8URVGUBaLSqJ6nAGzwnm8Aj/p93gDgC8YYAOgA8FZjzCtEdCjc2b59+6Ye\nDw4OYnBwsMLmKYqiLB2OHDmCI0eOVLyfSp27DWDn7lsAPA3gW4g4d73t9wP4MhH9P5H31LmrKIoy\nA8p17lY04iei08aYDwIYA1AP4E4i+oEx5gr7/qcr2b+iKIoy+2jJhkXC2NgYbr31DgCcwKVx+4qi\nTIfW6qlitCiboijloMJfxWhRNkVRykGLtCmKoigloUXaFgFalE1RlPlETT2LBHXuKooyUxYknFOp\nHBV8RVHmGx3xLyAazaMoSiVoVE8VotE8iqJUgkb1KIqiKCWhNv4FRKN5FEVZCNTUs8Coc1dRlHJR\nG7+iKEqNoTZ+RVEUpSRU+BVFUWoMFX5FUZQaQ4VfURSlxlDhVxRFqTFU+BVFUWoMFX5FUZQaQ4Vf\nUcpgbGwMF1+8AxdfvANjY2ML3RxFmRGawKUoM0SrqiqLBc3cVZR5QquqKosFzdxVFEVRSkKrcyrK\nDNGqqkq1o6YeRSkDraqqLAbUxq8oilJjqI1fURRFKQkVfkVRlBpDhV9RqghNHFNmg4qF3xhziTHm\nQWPMI8aYPZH3326MOW6MOWaMud8Y81uVHlNRahFJHDt8+FIcPnwpLrtsWMVfKYuKnLvGmHoADwG4\nCMBTAP4ZwO8S0Q+8bVYR0S/s460A7iaiX4vsS527iwiNWll8aOKYElKuc7fSOP43AXiUiH5kG/EF\nAG8HMCX8IvqWZgDPVnhMZY4JSxLce++wliRQlCVEpcJ/NoAnvecnAfx6uJEx5j8AuBlAF4CLKzym\nMsfceusdVvR5ZDk+zq+p8C8smjimzBaVCn9Jthki+hKALxlj/h2AzwF4TWy7ffv2TT0eHBzE4OBg\nhc1TlKXD0NAQ7r77oGeC01lYrXHkyBEcOXKk4v1UauPvA7CPiC6xz/cCmCSijxX5zGMA3kREPwte\nVxt/wELZ2WPVJz/0oT/E0aPfmfe2KIqSzYJk7hpjGsDO3bcAeBrAt5B27uYB/JCIyBhzIYD/QUT5\nyL5U+D0WuvSv3+kMDFyIm276Sy1DrCiLjHKFH0RU0R+At4LF/1EAe+1rVwC4wj6+HsD3ABwD8A8A\n/k3GfqhWGB0dpW3bttO2bdtpdHQ0us22bdsJOEAA2b8DtG3b9lk9RqlU2pb5ZDbPW1EWO1Y3Z67b\n5XxoLv5qRfhHR0epqWmdFdID1NS0LipQlYhtqcfwGRkZoVwuT7lcnkZGRmatLfNJOeetKNWMCn+V\nUKqIViJisWP09g5kbj8yMkLA6qljAasT4l8tglotHZSizBblCr+WbFikSATHtm2HsG3boYpt6seP\nfy8zy/O22/YD+AQ4fHMYwCfsa8m29PbuRy53I3p6Uvl3NUE55RK0xIKyKCmnt5iLP9TIiH8+Rs+j\no6NUV9fmjeDXEVDIHP3mcvnUSDmXy897u2dCzJY/l20sZ9+L7ZopSw+oqad6KNcBOZPP9fb2E9BH\nwHYCRouaPaYz9RAtLjNKMUGdK+duOee/mK6ZsjRR4V/izHT0ONPtizl3iRaXiM1GW2baQSxF4dcI\nqOpHhX+JU46IzOYPezGZLWYj1HU+zDazec1mW6QX0/1UykeFf4mTJXbzOWpbLCPESkWr3I6j1PP3\ntxsZGan4ms2FSC/22YhSGir8S5zYj39kZKRmR22VdEJzKXrVItIq/EsDFf4aIBQ7/fGWx1yaOapF\npNXUszQoV/grrc6pzCNDQ0OJWH6ppTNfLJXFWaqtyuVclGOutmugzDLl9BZz8Qcd8c+Y+Ry16Qix\nNObqOi0W/4qyuECZI/6KqnPOJlqdszzmaxS+0Mv+VdNso5raqlQ3C7X0orLAhOafxcJsil/WUpAA\nFqXAzsY90c5DmVPKmSbMxR/U1LOoKdWEMTo6Sr29A1RX105AYVbMHfGic/2LIkZ+Lkwwi9Gspqam\nxQk0qkcplXLi0UdHR6f9XChYXCOoeLmIUogJf6y+UDnHqERk50qgF1u01ujoKDU2thKXAOmjxsZW\nFf9Fggq/UhKhWNXVtUVLNJQjajHB4lpBlQlXrC29vQPTiuNcL3gzVwKdVVZ7oUbcXPepw+vQO6i3\nt39e26DEUeFfIsz1lDomKnV17aljzVZtGqBvxiPhrMqb4eyjWMc0HwvezJXwh21vbOy0I2553jrV\nEcxHJ1BK9VZlYVDhXwLMh203Ls7dlMvlE8cqtzaQ335jmqmlZQP19g7MifmkWCc51wvezNSXMdMO\n3d+eR9xyLqOJ0fd82P9js6tiC/so84cK/xJgPmy7o6Nhrf7VUeFKCmKB6uraSxJwEaze3n5qbOyc\nsUDN1jXI8gtk+SZmKsqhuay3t7+o6M9ebaH5t/+zjd/dy8bGTrXxLxJU+JcA8+XUGxkZsSPVbiv6\n8ePxqLY/0VGUWvu+3HOZjcqb27Ztp3x+MxnjzCM8Si4sSO2cSiurJmsy9c278IftUdFfPKjwLwFK\nGRnO1g9wdHS0pMiYYlVBi7W1kgqYpY6OwzUE3GcLU0LPQpkjYGTWhDLL+Ro7l23bttvrnN3BlnIN\npMpnuTMpZWmiwr9EKFbSd7Z9AKXsL0vApxP2SsMkp+vckquGFQhopaam9fZxVnTR7Ah/aPoAOlIh\njunQVmdSm845W8q11dG3QqTCv+SIhzD2FxWEco8zk9h8EfBSRvRzKVButjJKnC/gm3QGIsLfl2h/\npW2bbmnLmEM0l8sHI/a478Rd21G7/z4Nn1SiqPAvMWYjaWk2zUKx8MqFzC511yI2uu8hP/LFmFbK\n5zfP6sypWMfHDvT2aWZKyQ4r9J1w+KY7h5hDVUf+igr/EiPLjjyTUMe5FuZShGeuxMmZetLOzlwu\nb527Oft+0qk7k1DPrLY7B3nW/gsJYa+rawtmSsWTtPL5C6Y19yx0WQcJaZWZjHY+848K/xIjy8FX\n6g9tps7V2Rbo0dFRyue3ErCm6Ki1EkZGRqi5uStxjFJMUaWaqbKENRbO6Wc/FzPVuM+mOyw3SzgQ\nnTH44ajTzTjmeiZQip9DmXtU+Jcg2SF904/wZiL8c+E0ZlHYkmpDPn9B2YlMM5lVFItGKiXxqpKO\no1hUjtzLMEyWH/uRP4Ug32INAR3U0LCqaMeWdS+zrmO5nUTs+EDfvISWKg4V/iXOTM0TMwn7m0kU\nSSmLhzvHZtonwaGVpXUwlUYGTbdGcbHEq6xs1VLDYEvptP1tYo57rkC6noDWqU4KWE2NjbnM+zsT\nE2El11eFf3GwYMIP4BIADwJ4BMCeyPu/B+A4gO8CuA/A6zL2M4eXp/rJcvZmhxAWCGii+vpOamnZ\nGC3EVmzfvtnAxcb3kAuhzM4zcGaKzVa0/IibnszjlDpqL5VwnzMxj8QKk+Xzm4M8gbipZybXN37v\n3LVtaFgbFVjgADU0rEmEhY6OjlJLy8bU9rHXSgnJne7aqqln4VkQ4QdQD+BRAOcAWAbgAQDnB9u8\nGcAacp3ENzP2NZfXp+pJx4VzglJjYyf19vYHo8ZRK7jFo0Ky9p3tqJw+a9Rtv4Z4dO8nUS2nMJGp\nt7c/0/QyU3PVdDOR2P7y+Qui2cl8LSUngM8pOdIftee1joBNBOSoq+vcqXvh+2PYDzH9jKq3d2Dq\n83IOMdH2cxIkcSzZQSc7rKzjlyP84UxGnbsLy0IJ/5sBjHrPbwBwQ5Ht2wCczHhvrq5NVVFMwJyZ\noc8KT3IE6OzE20sS6VKOy+Ig+5peKJyYDERFNll1sjUYNbr6/SwmA4mqlFmmiGIRNuF2roxDwXZG\na4hnIWnzSFiDPm3+4cSxZGe8moAd5M+MgDYCVkXbN525JZmoJsdw916qZCZFXDqljQT0UHNzV9Qs\nNFNTz2KIJFKSLJTw/0cAn/Ge7wTwl0W23w3gjoz35uraVA2l/LDittXt3uMccQ2e2anpkjTdZMee\np88hfvykXTvdObjOpY9kpsKzgv5otiu3z3eC8qg3l8sntnXt6rF//mdCxyrPRBoaXJRNQ0N7ylbv\n/BV++/so7ttwnXM8Aih9n+RadXVtJKDZHs+397u6+PHvhfOnxLKFxaQlnex0Il6p6U2ZfcoV/krX\n3KVSNzTG/CaA3wfQn7XNvn37ph4PDg5icHCwgqZVH7feeoddV3YYADA+7taUlf8DAxfi3nv3YHxc\nPrUbwOe9vZwHvsR/bd9jGhuvQ6HwudQxp1vbdWhoCB/96DX4yEc+CKItANpgzDW44ILX4eabD0a3\nv/vug9i790YcP34NJif59aamPSgUDibWo7344h2Rq/C0bffVAA5hYiKPF154AU8//RN7bU7ga197\nF5qbV6G5eRl+8pOfY3Lyz+EWgQeA2/Hcc8Dhw0/j61//PXz0o9fg6NHv2M8fsse4IfjM1QC+CKAV\njY0n8cIL5+D06Xfb7YHTp9+Nu+46jLvvPohbb70Dzz77MzzwAIFK/gWcBWAYk5PAbbfdiDe+8Y1F\n19FNrjP8KgCfAXCb19b1MOaXU9sWCpfj3nuHve/FVQDeO3WOExNAR8ch3HPPXZH9A+Pje0o9kTlD\n1xmeniNHjuDIkSOV76ic3kL+APQhaerZi7iD93VgX8CvFdnXnPWK1UI8IiO5tmxdXRsNDw9PjZh9\nU0jSDFCg5uauoqO5Uqfu7GRMjn5nIxQzbcZYTU1NncSmEn9kvca+1h+M1DvsSDg20m0m519os6Pm\nPmITVFvGZ9z5NTa2UWgrb2rqSLXfmDVBe2KmnqR5xl+cJnYNJHLKtTFrNJ/0ifhF6ziHInt0Xq59\nf65MPWpGKg8skKmnAcBjYOduI+LO3Y1W9Pum2dccXp7qIPblj5tDclPmAhHXmVaAJIqHLMY+U85C\nHNP5KrZt226djpuITSMDBBSsozWdvMRinTYf8WfDePdu+yfOzi2UNJGsCsR2TeraxTqHurpc4pw4\nsqXV7r+bgGbK5zdTb28/dXWda/chx/bNSq62DwvwDnseeQJ2RByv8dBJ357f0rIxtUpXaNf3Q3HL\nrfs0V8lhakYqjwURfj4u3grgISvue+1rVwC4wj7+GwA/A3DM/n0rYz9zeoGqhVJCEIG+qRIA/ueK\nJQ0VD7uM/9hGRyX7VkofpJ2KWeeQNXpjR6zY1P0RNYt0tu1fCqLFRr6brcDKfg8QsNIKabv3Witx\nB9Nnjy0O4bRzt76+M+NYfqecFs98fqt3ft32fZfBy7MWd63TYaOrp1Yta2hY5bXPdVTsoC5Q0uci\nzuqtU+co9YliuQTJ5RxLX2hnrlDhL48FE/7Z+lPhjxN3XqarQRIl69MPDw+nfuh+qGAyTNONfCUs\nj807ayg0d4hgF6sWWSyr1IVtph2gsvZv2vwhI/JRSo7uYyP67Vb4VgfbjQTn4o/ok8sZNjSsoVzu\nbApNMDwyd+2NJXI5E0wstDK92lmykxvN3N6P2WezlV+nSMJ3e8ifXfiDg6zErqyFduZqZJ+FmnrK\nQ4V/CcNC6I+6eXrvJ3CFP5x0CQAXWZIcrY5aoUzajPn9dMkFoJsaG1uLxnDHRuxNTWfZfUrlzLTZ\nRsxHbrsuK2Q7PEHsobAcMre9i9KdlG9XD0U6PP4OAtqpqanLdngyim6z+18RbW+65v6WYN98rxoa\n1nq+GRehlJw1lLaOgDPryTnJrCE7kitL+Itn+rKPpK6uvWiC2mwx353NUkCFv4rJ+sL7rw8PD0dN\nJMWSnZwdOC0iSdFKCwbb32Phiu2Uz29NOHuBDqqrW04tLRspl8tTV9c5EREuWBNFvI6+P9J0Me/+\n0pASm94cHFtmA+nOJhnm2h2878fgF2g6Z2y4jYRlsoNXOuUeu50f/rp96r1Y2QhjWr2OJubDSCdr\nuY5VrmG3t21oerpg6rsUy7TNWjcgq7qoivPiQoW/Ssma4mbZ7GPmBYnciZsd4mLmShMMRMsCcMZo\naAoS00MoUKFwilkkTzwC7ifJfk2bWPqmSkps27adGhtzVHzk3koXXXRRILZx0XPtbCOgO4jAEfOP\nZBUX6zRkdiF+hGYCuiLCKTOzbmIfg38eHK0UL8HQSvX17dTU1BF0aqvtPiRaqI3y+a00MjJiRbxA\nSd9GelGa5uauqe9abPEYrvkTJqrJdulOpJREOWX+UOGvUrLs4bHXpe5OzISTzoLtsEK0nYAtZEzz\n1Ht+58I/+i5yESiFIJrIH7Vuso+nM5uI2CYXROHHaykcXTc0rPLaHo7MfQGXDiVmnw/t6e3EM4du\nAnqmzFPpzOewMwqPKdcxTNziJLAVK6TTCB2tcj39Ti9H8WvVR+JwbmhYRfn8BbZjWxk5Lw5VzeXW\nk1+O2nXKMuMYsPfyrOC7lixDERZ7a2zs9DKhY/fU7zgLCxoJpKjwVy0zEf6kABa8H+AoSdx+S8tG\nqq9vt8LkRsri2PV/fDyya6bQoTg8PJyacbCZRgR7B7GZREwgIgh+J7HKCldoj99Azh691XsundlA\nIE47iEfL3QScY/eTNtvkcnlqaurw2rXDHt+FWhqz0rZ1pXe+4mz2z3WNV+agOXI8/16ssaYauRax\nXAMR702U7aCW+73FXuts8w1/JhZu2k5hJU9gRSKiKpyZZS344qKvnKknNuBYyNh/RYW/ailm6nF1\nZWSk55s8uu1rI5SOBvHFJntK7ka/adORmIHEeesSgkJzwmpqbl5jxVQiTfrsc19gnbPQCXPMrh4u\noh5G97QQi3m6SihHIskovzU4fo64M5J9dtjXxIHtd1rNXvvE15BtCuIaRJ2U5WB14t1uj3MRifnJ\nJauJ2UnuXbGlJTso5nhvadlALodho/3rppaWDRmznezKnfLdLBb3L1FYxdAwzblFhb+KiU2FWfib\nPaFcTUnh92vaZItNS8vGzB8nm3Pijt3shd7TP+SGhrXU3LyW0p1PT9AmSV4apew6N357YqKXJ9fp\ndSQim5JF2PzORD67xdtnp30vGSPvhFdq48Q6oGSGNJCjpqazKDlyD9vssoi5Smmfbc8q4llP6CM5\nl9zMLtapN1PS1NNBdXUtZEwLOce1fC5svwsJ5v2kM4dj39GsDPKZVkRV4Z89VPiXGMVNPf4PV0au\nSdutCFCxURnH1bdQaOqJTf9bWjbY6X4snLKb4nby7uD5Ou95zIbcHiROxa6BL+yrKZ/f7J1LbJ9h\nZI/sM0diSqqrW0N1dZLxKwlQfkSRCHy3fa9YNFDoD+jwthvxRDk2O/M7xQLxjEUykFcRdxR+RyZm\nqLzdt5jkwvs0kPFd8tdZcN+dUkpfl7oinJp65hYV/iXE6OionbaHP9awOuNqAoYpbTYRJ2i2A05+\nxFJeoL6+k+rr26mrayM1NaXrtwPriTNkQ6FbaUUnFvPfGrRJTC9ifxfR6yMXr++LaczUk7Qz19d3\netm+YackYi2mJymvIE5T/1rJdS0eUeTa1W6vSUxQfUduOEsrZgryRTq9EIzz22z3zsMfxUv7lwfX\nPtYh5sl13EnhL6UcR9JslM4riX3X1Lk7+6jwLxGS5YOT0Rw8Ku0h4AL7w+0hHqHGxMT9sMNyyMPD\nw2SM2Jh7vKiaHSQmlHQoqIijf5wCAWeRc8KGNeiX2c+0kdS9b2nZ4IU0jlDa0SmCnff2KwXTYgXZ\n1pKzy/ttlqgYv+MpeO/FkrtiMwy/41hOzi+QFQ0ks4KsSKHwGvrH9f0JsazgdOG45HlItFFoRouF\n2/K14Blf8j1Z1zfLBJlMWmslNpvxc3/BHxX8uUeFf4ngTDy+yeYCK2wxp+jqDDHpsz/stqB8Q8w8\nsZJ4hBmGB0pFS9/xKccJHcqriYVenIsiksnOw5icNedk2cPlvPPe4z7iUWt95NyX2fezRt+++Pn7\nHLDn3G2FK6s9LnonWRZhBbmOze9spHPqj7S1J3L//E51jd1v3Hmb7kwKxP4BGQQ0k0t0Cz+7iSQL\n16/uygOA9DmHVWGzEwXTx2pp2ZiqHDvd8pRKeajwVxlZoyGXPNNPkvSTFOR0WF1TU2cQermGXJhm\nT/DjHsgQt6zRq7wW1vaJiYsv9GI68d8Pk5Nio9p8RDDFNt5mBWydPdY6cp1PVnx8rJMS4Q47uph5\nxZky+PkKcgu5rLD3Z71tzzLbvhZ7jE2UTrASM42cf4f3vtzf5fYYfvvknoq5x3eQS1tbyM2Yktew\nq+ucVIluFvFYB9MXNTXGC+hlXff0jKSUKCBlZqjwVxHFQjiTSVhily42kj1gqzn2U0vLRivyBe+H\nJ/bmA/ZHGLP39lE8ysYvXyxiUiiyfd80z7cEz8McAhGvcL8D3nZiX98SnEtsBiKj75g4xUI0W8lV\n+hQHazp81X1W7o/Y9cVfEGbsSimHcI2BrEgbCQvtJ57tuTpKLos3LthilsrlNtloq65UwT5XIE5m\nJWGnuIpiAwFOLGsL2rOK0h1Q6Ndw7dOIntmlXOGvdAUupQxiK23t3XszOjraMTHxZ3ArQ91u/58A\nIKtVvQrAgwAO2ue7MD5ucOzY++zz3QC2AZDViw4BeA94RabzAQwCeL+37+/a/43wV+wCdgFoBvAL\n+/iM/fuiff8XAK6d2tqYq0D03uBMH/bauRtAR/B+K4BfB3CjfT4A4BtI8xiAWwCsB7AHwK329au9\nbYbA1+16ey7jAFYB+DCAX0b2OQG+BuvhrlUdgH8F8C4AnwawH8B9ANy9Yq63n1tp2wXbrvcB+JJ9\nzd/+FgB3Avg4wpXCks/vAHCpfXylfW+HPc9hAGMAfg3ASQBPRc4J4OroDXjuuf8KAHj55Wtw1133\nYHx8J/i78DOMj7fhHe+4HGvXrgHwCXsud4Dv1wSAvwLwZfB9Z4y5Gj/84WkQfdK+cg2Adcjnz8WT\nTz6JiYnbbbuG4a4nwKudHbTXZyeAxzParcwnKvyLhOPHv4fXv74n8s7Z4GX3PmGf70Jvbx5PPMFi\nuXx5B5555o+QFhD/x7cVuVwzXnjhhzh9+nkAK8DCArB4T9jHw2Bx+DH4q8FLHbJo1du/EbvtbrCg\nXgtjCES/BAvl1ql2cgd0yNv3nXAdwdUAcgAOJ86NP+8vA3g1AOOdlwgYALwFSfG/0/7/z+BlID5u\nn18HX8T48fvssXbatn0e3EF+EXy9V9jr8gjSNNi2hAJ/e2RbgMXvTMZ74XbXgq/rNWDxPWpfPwng\nL8H3BAA+EJzTbgAvg5felI4CmJwEXnppD/i63zK17YsvDuOll+4E318R6/PBi+XtB3AcfI1uB/Aw\nli2bxMTEJ+E6idcAeB7nnvtqvOc978Rtt+3HxMQZ/PKXB3D6NH8H6uuvwZkzr4A7399AU9PnUSjI\n/VcWlHKmCXPxhxoz9cQWCA/L/HIWanrKLdNlTsEXp68fVy+mDXFGSvkEidoJTQRSEmF18HnfzBEz\njchrvilpOzn7tG8SEBt6u93nGsq2D8t+JLpJSi7Hav9sJud4lmsVc9Jusefp+0jEdr+WJA6enaW+\nKStd5sCVgoiZirop7evoIbbbt0X2I+arNtu+5bYNoSM7dt/E0Sxx+2IGC7dLZ+fKNXKmwVgiWxgx\nFJqrVkfWfWi1df7VuTsfQE091cPQ0BBe//rNOHbsdvAi3AcBnEJHx+PeYt4/xokTDQAo9fn77z+O\nCy8cxLF9Oe7CAAAgAElEQVRj30Jy9P4uAL9CV9c6PPPM1QD+ADyqldH3veARYcg68Ijx/WhqugHj\n4zIDuAPOzLE/8rm8fU9GujLL2AngL+zja+z+fwXgB3ALhu8GsNzb15jdz0PgEWsrgOfBJqUVAJrA\n5o2/RHKUvR/A5QD22ecnwAu+hZwEcBo8Wt9qjzcMN4KWEfN/se9fY1//kP3/YfAM4HoA3eDlpa+1\nx7sPbH47bbf9Bdi0Vg93D3bDmcfO2NeO2ddPw81OdgP4T+BZx/vgRtgr7bF8WgE8i/TMw58J7AGb\nh2KcAI+5Pgu+R59AeuYI8H1ptNslF7b/8pdv9MxIwMTEH6Cj43H72Jm2JidP4Lbb9uPo0e/oQuqL\ngXJ6i7n4Qw2N+Immz2jkiIu+6CgrOYpORs4Yk7Nx8mFEjYyUQ2dlMqa9t3eAhoeHKbnCk4yOc0E7\nhslFqUgZgqyCZrHX+ym+UpWfueofU7JX/dnNOkq3K/ycOF3FIRuemz8rkOcFclE14iAf8I7bQxzF\nE0uck0id2AxJMnflHP3IJpmBSImHnsi9l5G9OHljyXZbKRlNlBVCGuZdxL4zYW5E8lhcRTQ5C5OC\ngMnzSq8hoVQOdMRfXQwNDU2N7gGgUDiYMQraCp4R3AHgW+BR4C3e+0m7MtFKnD49grSTd4Uc2e5v\nH9iZ9x77Go+4jx9/FDwC3Qa2v1/l7f1XAAoAttj3/1+4keousD/iZOQcHkJydC+cAc8O9oNnAsPe\ne/vB9mZxco6BR/3igN4JHoGfBeCG4LN/DODncNfmDPi6ATxyHoZzUvs02+PcAbarv2yP9wrcDOad\ndn+/AtASafeN4Pvj+ymEDrAvwR/d5+zj2AzklwA+ibQf4W/ssSfAtn7/WLsANKKx8TTa21fjmWe+\naLfbCp5tnGXb8Q2kR/gFJH00KwGcA+cAP4GkT2W3HbSFM479KBQux733DmN8XNrsthkf5wAHHfUv\nHCr8C8jQ0FDml79QuBxHj/4OJiZ2g380l8KYIyDaGmwZRs68Bu5HuA/AKQAfBLAWyR+tCLx8lp1/\nk5PAsWNXgZ17VwD4JlgwCMAFYAHdA6ATySiVE2BTwO8jGR10FVgow9d3gUXocXD0TchJewwh5kz9\nOIBzI5/9GbhzOgtsBjoFZ7aQfWwDdx7CNeBIoJ12mxMAvg82k0jnI9xu32uLHFtoA5+7mIIeBguo\nHyF0AsBfg6/FZu89id75nm3LIXsesOd0JVikfwU2O+2E65DfB+BOnD49gVOnXoDrZK4CdyTN9vmp\nSJu74ExzywH8iX19GPz9kO/eHwFYBo7CegBh1FlHR3tiYHP//T/Fc88VuVTK/FPONGEu/lBjpp5S\nkFhrSbxhE4yf1NNJruZ9rIKnOP38KpNt1jywkVymaVbWqx/HL6YGcYKG0/7QLNRnj9EVeX29bbfv\nWIzVpfFNNjHTTDulF1GXDNik45zNLKFDVl7PE5uq1lPSuZzlKJbXOsgvV+Ac2HJdzw7aIvWIpCaP\n5AT4Jh8xxcXMXy3k1isQB7RcqwFyJS78wn1+u9vs9ZJzT95bXs+gi7LNdb5DWs73otR+QiduWNtf\nTT2zBzSBa2njfjz+j76feNm+Tmpq6rRRQC6KongZAt9XELNHZxUUE/tzJyULxmUlmoX7kJLCIsRS\nA2c5uUzclcSdhtih5bUw4ckv8LbFtm0g0g7fZxAmoY1QevGUdd5+Qp9IrKSxS5xKrpoVu35hZc4L\nguPIZ4qVj/CzmTsp3TlLsbpwICClL6T9srbxFnsuxfwf3eTWMwhfT27rJ2k5XxZf8/lauL1WKFf4\n1dSziBkbG5uK8HnggR+AyI+o2A229f4FzpwBxsd3o6GhAb29n0FHxzo8++xmHDu2FfGEGTEXHIIz\na/hmIIn73hf57EsI8wo44uNVSMfKT4BNC/K6mDZWwuUDXAuOdlkGNgmJ/+JqsH15Ndic8QCASbCZ\nxUVCsXmpGxxv/2pwslcY/bLV7nerPW++bu5cJxBGq7CZY5c9z522PQTgvfa4u8AJX76dXGLlH4cz\n0YSsRzK34ZvgPIM/A/AbYNv77eCIoJDl3rH22/M8D3FT1JXg78jVYB+IJFABfF8fsttsA/A7YDPR\nSfB9WR/sT74PpxD6lBoaJnD6NDIJkxUnJw/i6NFD+NCHsj+jzD0q/IuUsbExXHbZsP3R3I50GOZ9\nYPFyP9DTp29HR8c63HPPXbjppptw7NgHwIIsYYeSHCU/YoDtyfeBvwofB/AE2B5/Chx+GYr5GQCf\nQlpoHgGLiGThbgPwT2AHqSRyHUfaYXseWIRWA/ivwX6vATulT4JFl8C+iStt+3aD7dFDAP43nA9B\n2iwOzfPBNuhXgR2bI8FxrkWan4Bt6GLPfguArwL4W3BH0YC4OD/vPb48OM4HwOJ9g31+td3PerAf\nRjoSaVOYoPUa7/lDYMfyIaQzu5fDJaE9A+eo/SyA3wLwj+AO+T57Tg3ga+qH5UrH/yOwv+EOu++H\nIH6h+vpr0NnZgmeeuQbcUTagru4ZDAxIKKyyWFHhn2dkFA9gKp7Zf21g4EIcPfod3H//cW+kdAj8\no/OjNx7KPMZNN92Ej3zkY+DbuxvO8UpgAZMRqzg4ZZR9LXhE/1mwyLwXHFVyA9gBez04jj7kLLDg\nHYWfHepEdhfYCfsCODIHSEex7EJ6pP4asBPVgIXrNQD6kRwxP27/wgiVa8HO69MA+sAi9w2wiIec\nRny2shLAR+1rV4FnHBJJNAQWSrm+4sB9GXyvxCm/E9yB1YMF/iNBO6+2+7oP6ZH7HvDIvt2+fh9Y\ndP1IqwsB/CmSM7BxAGvg7sUuAL8H7gjfD+4Y/tw7r/fa/Z8MrsP3wZ3f+7z9jEM6iDNnXsYzzwA8\nUHDBATfdtAdvfOMbMTQ0FET3AE1NezR7dzFQjn1oLv5QAzb+WOx+uJJR0gYdxkHvIMkM7e3tpTA2\nu6FhjecLkByAAXILih8gYA319vbT8PCwVxfft9lKeV+xWzd7dt1Ra+f1nY5i845lsg5YW7Rv1xaH\nbVZ1T2n3GuKsXN+5G+Y0SHZpVqZuW+QzKynpcBVHbHitpqs+KpVTw0VPVtv9r6Xkwu+SYRvayCXr\nOnYOYQ6DLP6ymXjh+WL2+Ni9oIztpZieXwrcL4kd29bfV/HlFbUu/9wBtfEvfmLF2W677cbEa8wh\n8DTbhRs2NIzjzJmvgYht0w88cDV4xP5xcMw5sGnT2Th69DuYnPxz8FTfr88iJp6/wKlT/xeOH38Q\nk5NNSI+yX4KLHR+z7fisfe8r4FHjcnChtKfAeQCnptqQ5s+RLk62Gzw6DhF79dXgOPXV4FnGMHi0\n/C57vNvBmbIT9tihf+FacPjqKqSLrJ2w12E3XP2cbnte3wT7CR5H3IxzHtImrlaki7h9GHwdJ+1+\nvo5kvgPgzG5SuOxye36CZPreAOBFuGsO8Cj9V+Bw14eRvochJ+zfIJKmKHnvYdue0+AicfKdOQi+\nfpVRLGxZWRhU+BctXG0yl7sRb3jD6/HDH74ajz3mim/xJOkWAD8FT/WBxx+/BuxoBfjWhnHvXP3x\nmWd+grgQicMSSJtixAYt9uk9AH4TbPNuAIvtHiRLGDSDhTRECqD5Jovd4OQmAOgBdyS++A6BRf+/\ngxOqToOTn/aDhUscv7DvPQ4Wf98kdhO4kNtysDB/yr5+HThnQcxJrwLwD0h2JlICw+dBxEpqcEfS\nb4/1sP2cfx+uBXciO+FKM5yCqxh6lr0Wp8CmIXFK+53XnXD+FP8eXgfuMK72tg2d8R+0xzkb6SJ5\n7/fauQt8XcU0cw24w5Hn34UzFbkcjbq6a1Ao/F3kuiiLhnKmCf4fgEvAv4BHAOyJvN8D4P8DZ48U\niuxn7uZDi4SZmXqS8c68xmk4RU+HYTY2Smx31pQ+vW6tq8G+mbgMgSzx6Bd/E1OLbwaQpf667TbD\nlF4jdwdlLz4u8eYSyx6GToahiv7SjK3kchBCk5W83028GIqEWPomHr8IWfjeOm8f68iVpAjLF0ie\nQykLrvthlVvILem4wrvOAxn3ODSjxe5tm70W/bbdLeRCXLPCQmPfBVmlTO6xhHz2UUPDKurqOpca\nGtZSU1MHNTevtceVa72FjMlNhWuqiWfuwULE8YOHZI+C4+6WgWPuzg+26QTwRrCXr6aFn4jj8XO5\nPOVy+cQPpLd3gHK5POXzW6dqnciPhe32ayhM1mpoWBP5UTfbH/x6crZ6ETapdBkK+Dpyi4q0UnoZ\nxnX2hx0mWa2ipBBKMlNop5ZEqW7i+Pw8OX9FLIbd//xKYtt2jtxi7BJvLjbubnKJUWFMvqzVG6tu\nKYuEZHWSG+wx2sktnNJHLND+4uxiC58ud0LaI/H1o+Q6gFHiDiPWAa2i5ELvWev5JgcO2Qu2bA8+\nF3YgI97zCwjoobq6NWSMnPMme03D5LQe6u3tn/pOxwY52hHMLgsl/G8GMOo9vwHADRnb/nGtC3+x\nlbeyCrZx6WX5MYeFx5ZTUoxXUHr0WfC2PUA8qg8TflaRW60qPpOIvxauqBUT1w1WPMIELMkCldHu\nCKWdkunsUjebCbN9Y05xX9yyVgzLaneO4mv8SkcgHY10XFKsbiXF27DRnp90wLK0poj5GvsnHWQb\n8QxKOpR+e4/8DidrRuELexelk8bCcsv+DGsHuVlGs22DXHOZBWXPKMWpmy7S5jvbk99xnRmUT7nC\nX6mN/2wAT3rPT4KXVFIixJy7EsaZ9fq+fZ8A22APIV0Q7Haw8/UOcOhdE9KJSIfs528E22M3IB36\neDt44ibFuLIWFAl5GmwP3wS2/cpqXcJue5zPgO39f4q0rftKb9uXkYyp/2ykrWLXfh5pH8YhsH08\nxq+QdgBP2H38e6TDOSfBDubYqllX2m02gW3oe5AMS30AbGMXf8dD4Fo63fa4rwXnOOTBP5n/G5yT\ncQLsszgLwHPg1bz88tbdcAXptoHv/Q+RXvXK583ge/RHYGf5y2DfwUHblpfBzug8+PvxcbhQz2sB\n3AV29h+y59IDl/yXpK7uERQK+4JXx8AL4/yZfX4dgM9hfPxjU99xl68C3HvvMO6+O6tgoTJbVCr8\nNCutsOzbt2/q8eDgIAYHB2dz91XFs8/+DJddNozTp2NFyGK8BO6Hs+gGFy8LozoAl8krK3f1I+l4\nvQrAReAfrbAbnBj1L+AOox/c+XTDJSZ1gQVqa/BZIYyS+QiAjWAxPmn35XMC7Nj9fpHzjEXHvGzb\nIytKPQSO+DnfvnbI+y+1/Fsjxwf4Wkmbr0OsRj13TsvBWcqvAzuMbwZHEbUh6VC9BhwHL8tKynoF\n1yLd0e+27ZeibVeCr5VfQE1WF5PIpWGwI/i19vUPgu9PM9j5fRLsfpNoqbAz/TB4JbBvg79jkkgY\nJqddhY9+9LopwXbx+51g0fe3vQHACtx//0+xd++N0UGPCn+cI0eO4MiRI5XvqJxpgvyBM2N8U89e\nRBy8pKYeIkqbeurq2qZi6sMiVrwQtm//Tk7vGxulNo8/VW+mZNEwf7F12V+4opRfe0am72Hxt5XE\n5omt5GzSI5SubyMmC98M0W4/szw4bhjTLn4AWVWq1ZoZ5HqFZp9VlLQx+7btVmLzkqwq1hOYJeSa\nbLHv+w5QMVFITaJ1wTEHyPlGYo5XOf4mivtEYtv32OOJrZ8obUoJz1/qF4nZaICSzuIB+7qYxfrI\n+TsuIDY9NRObdsQXFIv/b/eOK/4V+V712GuwYcq2H37f47ki7jvrakq593VB9tLBAtn4G8DFUc4B\nxxGmnLvetvtqXfiJnCO3rs5PqpIFPFwRq7SNlJOpurrOm7KF5vMXRH5Um6y4+JE5K8kJ9gaKJ3at\nJlc8rTty7G5PIMSOHLOZZy3RWCDuOHLkqlSKMIqgiOiLuK0hZ+fuiOxXbM3iI/AdrL5zM5YM53cY\nKyiZKOZ3kOI3iFUqlSUb/TZ3UtIf47e3jdK+klDQ/cgl//X2yP5iUTktlO4gVlG62ql8XjrGNbZt\nMZ+GH91VoHQnvpouuuii6PedBzCxyquuzfxbcIOeLDu/+gLSLIjw83HxVvDc+VEAe+1rVwC4wj5e\nD/YD/BzAv4KLqjRH9jO3V2gRkRR1+cE6h5x8sWPr8vqjoXiIp4RHxkTnAMVFvY2So2cRlDC8Mkcu\nrDGr8mQ4uhYHYyi2EiXURtyh+A5EP+JIMl5j68a2BvsV5+R6u69ltr0rve2yMl2lk/SdruLobKPs\nsEjfKSsdV4HiQr2O0uGfWVVN5T72EY/QY/vzndZyXZfb44sztYtcNnasM2r1rk+3vWY5clFUsSij\n2H7aU2IsZcWNcZng/DjZWfX2Dkwr6NOtWFerLJjwz9afCn9S+IlktNRDPErfSH64nHvfDyfMEY9e\ns8SEKB7Vsz6yvZhCYkLRRemRpZRY8E0B0qlsiuynLdjmALHoZoWMhiNgGZ02W0HaaD8vHccouSgg\n3ywhcen+CHY9JevXSxSKiLh0QOE5DATtbSO3oP0mipuiws4tdo0lL2KUkqWk/ZmFiPAAuZmWmAP9\n/IQ8uTUCikU8SUlr2daPoPKjgPzONrmfMFLHCXWB6uraqbd3IJW3UqqAx34zahJS4a8qwtFLVtKW\nC+X0R4fNifh/tvP7giDiH35ORG2A0qPqmPA3U3YHIuYBEdMtVjDICl/s+OF+/PVkRfwGItut857L\nCHw9uVhzvyOTmPeziDvLmG3fb1eO3MIifsKWhFP6oreD0iaZfkq3VxK7zqH4IjmhgPlx+f53QTog\nf6Q/Qsl6Ru22DTI48Gvt+CYtEfxwBicmRnltlXf8WD2mA/aYsbUR+P14OGdSqEOTTSkmHBX+OCr8\nVYD/BR8eHqZcLk8tLRuoq+ucqVW20guup4W3rq5tarve3phYikNPBExssv7IMBT5MKtVOgtfKGRU\nv8W+Lqt79ROPGDcTi3LMcRvOMvyR7ig5oQ3PpYOcDd/3CeSITSCSfzBKacdyaAPPGvHGBFH8BL7o\nrSA3O9hRZH/S5tisLvQxdBD7MfIkBfiS4hvLbZCR+EpK+kpkfz3B9fXbIvkSa8nNTrba98LVwMLz\nkusrGdPt5Bz9vHBPPr8187sbE+pSTThq6olTrvBrrZ55IllfH3BlkY/ixReHAdyH55//Hr797W9P\nE8p2FiYnr5yKgX7iidji5gCHI/4EHFJ52j7uh6vJA+/xDgDfQbJOzLuRXOj9aW+ffunl94Dj9LeC\nwy39ui/fBpcUfpV9bT84Pr0OXODNLy52Bhxq6i8Is8vu92kAR+yxToLDE/21ZCX+/Xwkyza/124r\nhcYejFyn58F18mOF8uQ83wuuibQSbh3aq8FhkH7YqxRdewjxonVPg2PoXwaHfP4MnEvwFe9YUg9H\nwkTDUsmfAReu+xD43sji7n7bdyFZ5+hy2y4pUOffP7LtGbPncxBcfz/kISQXnd8NLvr2VbB7j/f5\n2GO7cNNNN5VcjjkrtyX8Dfhr+AJAoaCx/hVRTm8xF39Y4iN+NwIanRodufIJyRBP307a2OiHZ7rR\nZ2/vgLekXaxEQTO5ui3yXie5ED8pH7yZ3IjbH6GFpo0Oips2JAzRN0nEImeWRfaZVYLaz1YNR+JZ\nWcXiXA1Hv/4asWu8dspMJ6u0cVjWIGauklpE4g8R05efZRu79hLV4s80wtG13yYZpctMI7z+MXNT\nOFuTNY7DbbeQm0FJe2MzoJXee/L99dccdvvM5fJT31814cwt0BF/NRBmeV4FHu35S9MlRzyvfe15\nePDBvRgf/yVkZayGhg/gu99diTNnlgN4Fpxlegt49PoSOLK2E5xclZV5KolDfw1epOMV8Cj2FnAS\n12HwTOHD4FHpAHh0GK721AlXSlm4A+lR6C0AvoZ0Ju4N4BGuMAQeFV8HngXsBI++fwzOTD6DNMvB\no9WXIue7J3i+F25m8wbwLCJcatBPgiq2DCXAM5f14JG5Pwt5BRzcJslofwCevfhVPofgZkM+Z8GN\n0gFOJHvZHudeuOqYMmv0s513g68FkJytrQHPvkLawTO82xBfgeskOOnsBcS+v3V1/J31eeGFF3Hx\nxTtQKFyOe+65K3JMh5sZcJZzXd0juoLXfFBObzEXf1jiI34Oz4yF5PnRJG7EE9o0Gxtbqbd3gPJ5\ncWbKSFmKqsVG2lmLqPvHloVExEkrSViycLifxNVP6RG7OFk3UfGQSQkvjM0YwoiVDuLZkDgbpW7M\nFnL1cPxtW8hVFY35CGK1aYiStm+ZaXQQz4I2UtK5G/NTiH8hazYnETP+QvLdlEyyC2Prw6gcCemU\nGZyfV7HcXr8w1l5mM75juUBxX44/iwu/P/6MUhzryevb1XVe5LrsoJnY4d3iQWq/nylQ5+7iJ9sR\nm47qiU2Bu7rOIWdWCBOdYhEjMXONhAlKpxEWepMEpGZKm02KdSR+mKKEBPqflSSicJ8iPKFQtdjj\nSTim/zm/9LTEhecoHqq6lpyJRETJdxLHTE8xEXRJds5h7jtaQ9NKGzkR3mLb4UfXbCcX999i319P\n6U7N77Q6yZmTpCPeTq7zlPsSdqSdU22pq5PrKt+hsJOWaynntIbq6iRENx16KlVmc7m8zdJNmqKK\nmW3ke875KJq9Ww4q/FVAOozT2ezr69sTpZrTnUSWiPsdSEx8/ZmB1KX3Bc0P9fT3FRP5Vm//Ejoo\nYhCG+K0kngWspWSUkIiN3+GJqIaloqWdsQgZiT0X8ZMM3bB9ImxyDBH1PnJ2d/ER+KGNcg5SOz92\njWRffsZx2CnKdRN/i9yTWF6CzKjWFzmmX3E11mkdoHiyF7fFZYzLaL4n2M86cktFcu4IdxYx239H\nIq9kJvb69G8h+X1W4S8NFf4qwdUvCRNjOGxP6pa7OjwiZtOZbdrJOSuLmVuyQvVCgY8dbxk54Rbx\nWWF/tMuouOhKGKS8voZ4FpDz9uGLgL+WbUz4Oyg9Ip/uGoXn2RE5Rui8lCzWUIR7yC2eImGtsesq\nJa/DZDe/dr9/vcSRKglU4ahdwi3DjkG+I1kJc3I+MkMaJVe2Qxy7cl99R7C0158J8b4aGtYkTDIz\nCbnMClWe7nNyHC3dwKjwVxHx+iX9Uz+AFSv8CAv54cXs176whtExsR9WO8U7hLB+i8T9+z/+ULz8\nwnHiAwjFYS1xjHiH3a/Y6WX7dZQ0lYgI7vDEJucdxxd5yej1hS9cSD2cFYXRT63e9iPe9ZF9ZY2o\nw2vRQekZmaxKlRU15OccyHlKHSUxcfmrjA1QMpfBP7+YWco3GUkpDZ51GSOrc8m1le8Ab5PLxTqO\ncEH5bmpsbI2WaShFlGPCn8vlp/2cxvMnUeGvIjhMMzR5yA84y2wQio2fOCWCGzorQ5POCKVDPFus\nCOUIOJeSzkGpJpm3ohTreETU/NFklkiL+ErWr3Q4o5Q0Afkjd98f4Y9Kw218EW2NHLeDXOXKPnIO\n11WUrOUTirJ/vuJobaV0+2RVMbGf+8lXWaaquPmER+wyWwjvebF7EAq1LIu4iVw5hh4yRjo7yf4d\nINfxHojWgMrnL7Az1XXEnTnPUkSoZzoCL1fANfwziQp/lSFVOo2RmjO+Uy32Q+6jXO5sctPyZnI/\n/Cyxl4gYcUbK67LMYg8lI0zWkaty2U5Jx2/WTMEfUYs5JyaYMX+EiPcosdjK682RfYRC7NvIs8xa\nYVTSBeTMJKE/IoyKidnJ19rrJn6NmGh3ePciS9z9zOVYpyDXJTyv2HmKfyLWIbTZqBu/QJvvYA7b\nzo723t7+qCjHZqr5/OayR+DldBgq/ElU+KuQZFKXjKh952Zyql5f32nFv5ncKDUUbxGVHAFN5EIz\nxRSUI7bHZ0XpFKsbE5qfwmSigQxx6qNk6QOpkSPLGIbRPlLvxm9HrEORWUNM+OQ6is07FnIoTvBY\nBcpQ5PzOVkQzdq6d5ExWYR0hGYUXvH3EirTJkpaxuj7heaykbDOc3K+YCTDW4WxJLAcainIsKq2l\nJV01dS6FWE09ScoVfk3gWnBOgBOUJIlJVjbaCU6oeRCcSHUDzpwBnntuN7h0wsvg1PorwYk9HwAn\nU/13uAScGwB0gJN61gN4HJzk9Dg4sUhS933ORjrhS2iAS366Dpz0I8lE1wF4GzghzS8xIAlQp8AJ\nQd+35/rb4GUc/gZcBiJcbWovuAzDCXAyFSL73Wm3+WnkvT8EJybVg5Op/MSxE3BlH24HlyHwj10A\n8LfgJLYbbDvHwat2yXKXB8FlLUImAFxqH38NnGC1H8D3wIlb94ETueQe7QTfH2E3+J4BySQugK/V\nNriSFO+zr/0FeJWsq8FLI+4El3botuflL5Mo+0yXZcjlxvHf/psrhRCWROjoSCeALVu2LPXaXKKl\nG2aJcnqLufhDDY74k9U3/VGvjFRXUzqlv2BHlX3Exb3EPBSrgy+jepkV+KN+sfcXXyQj6VwNy0NI\nQtBqSiZ3ia07jFzaSG7Uv5myTRpiNorlJch+ZUQroYcFclEq4mjNMlPFRsB+NM8m73rI7CrmzA0T\novwZkj/bCYuk+TOnWHSPf43FzyKzk9hsyv9u5L39dFO2CXBr4jiljJxjo+1yyywrswPU1FN9JKfO\no+SyOH1hywXb5CgtNGKK8EVIzCkHvP2F5iO/fn0/OX9Awdt/B7FtW8TRT7QSwZWkIF9Mw+NJkpkI\nVOicXU1JG78s1hIKXTMlxX8dJTvHUXJOWP+10JQUHjt07kodpaxQTWm/+GYkGircTu6nbyKSDlh8\nL2Hn202uI/A79JjDPuxowsxs3raxsdWu2OaHprJjuqVlY9l2efFVcXXZfhX9eUaFv0qQHw6Lfhi9\nE7P3in1bIkfC9/0RnqyJK3HwvhDHRtWS7h9G4kgykQiUfDYr9rotEEcpRuavBdtMScEKhdnv0CS0\nMGbTjoW1bgq2i4Wtygg4lqgV26eEmg5E9hWOtEWww8xhcZZLklRsPzLLGpi6j8b41yIWVrqG3BoI\nWWLIii8AAA9bSURBVCGnq2l4eDhlpx8ZGSFj3IyjsbFzaqnPmcbFq7194VHhrwLS2Yqy1mveE0/f\nYRrL7owJUCiKodnBNy/4n5VQv9iC2L4TUIQ5K6qkQEA9uUihMF5+pX2/0wqvxOjLtYjtd4DcSlid\nxLOKMFfBb0MnuQiXfkqHaIpz9lzv+GL2igm/mJFis5eYSUeuzwi58gy+matYkpdcp0JKiIeHh4md\n9DmSJL3h4WEbYSMOdqncWU+uc1s+lQUe+x7K/isx1WiEzcKjwl8FpBdQDwuT+eULDlBarAuUNueI\nuGTFcvdZ8RT7uD/K9U1BMUHyE4QkqSscYS4n4CLv9ayZhdTr8csV+CaPmPDHUvpj9upYjZ6N9njd\n5EI7/UgeP3wz7GD9tWV9/4Z0FMMU7wikneFIXPYXK5LmOoeGhrWZi5CEI3L+LsXMd25/Uh659O/k\nzMRbhX/hKVf4NapnwbgD6RLCh8DRJ7sBfAMclRHyClykzelpjtEMV4J5NzgKRMr0ngHwV+Bon3ok\nyypfDS71PAmOxvmw3WYzOCroA+AIn+sB/Ct40RCJmvEjSITnwREwXwRHFN0F4Ef2c6+x+/WjcnaB\nI4fCxVHusJ8dBi9A8np7Tl9CutzzjQA+5b12EMkIpSu9906BS1lLxJJEIU2Co3A6wFFABlzK+hbw\nwiX7ADxs2+AipBoarsfq1cvx3HPX2PPbibq6z2Jy8gpwVM4dAP7RHt9FpGzden40QmVoaCgjcuW+\njGs0P1EupS62oiw+VPjnkeQPJbZCk7ACXFN9CByaKOwHi7UvZvvAIvV9JFeDuhZcp/4QgN8Ci/nH\nwaGGu8Cdil9f/QS4g3gdgLeAQxmvAIvh8+DVp660+94NDsn8EIA3g8VNCEMQ94BF/z5wGOqPwfX8\nnwaHSALcCb1kjw9wuOMTkevytD32frvN5fbcvxTZNsbD4FW6gGQN+2sB/AK8ktSV4DUPCpicHMfk\nJK9w1tDwCjZt2oTHHjtoj837q6s7g8nJrfD5zd/8Ddxzz10YGxuzYYeP49lnN+PYMQnjHAKwG8bc\nCSL+bGPjdbj55s+VeB78Xfr6138vVQufr9FBALtw7bXXRz/r2gUMDFyIe+/dU5Z4a2hlFVPONGEu\n/lADph4i37nbH6yu5Zt6JMFKTA0SpTIQMYnIeyvJlSSI1ayXdValxMMoxbNTxZzhvx7bTl4TW3iY\nCSshiH7S1nJKO3Fj5htxioYmneURU0kYSnrAey27dIQxzZTPXzAVjTIyMkK9vf1Tax+PjIwk7o/Y\n3l2pDbnGScf4TNaMLdepKoR17Bsa2qm5uStR5XU+2qEsHFAbf/UxOjpq66LkicMHNxA7W9eTKxQm\nZYEPUDz1X5b420Fs6/YX+vBj01fa/fuRQbGOJE/phdBjzl8pseDH0Yv9PO+9JtuuoIaGmHNTwhZj\nr0unJ07vWHjnOnKVKv1YeAlnzXuP3aLgfjlhuRe++GXZr5M15P1r1Ee5XH7eq0rOdJ9ql19alCv8\nauqZR/wpdqFwOYaGhrBpUzeee85f/m8nOPt1P3ih9B+BTT+7wHbsnWCzjQFnkr4LnL35FbudZKOG\ny+TtAvCfvMcA8EakbeuNAP432Pwhi3MPI5ldehV4AXKxa2+1f9vstv7SkisBdKG3tx1PPHEKzz0X\nXpUsk9fTYPPQTvCi7f8ewD+ltsrlmvHKK6/gxRe3eu05CPZJrAT7AcQkxMtONjXtwc03O3PG2NgY\nLrts2C76Ddx77zB6enqirRJ7+8UX78Dhw1OvAjiFN7zhUFFTR7atvnzmYp9KDVBObzEXf1jiI/5w\n4XRjWmlkZMRbStFP2gpNFE3kEqf6KVmbp9P7nD8CDUfXflanmIe221GwrMbkJ271E0fGJEe1yZo1\n/r5XZozI+4oU+VpNrlJnaKqROH63Aldd3cpEDLpfV8Y3eSSTpPy2tFE+f0FqZBwbBWcVKvPvZzXG\nsFdru5U4UFPP4iZW4MqYNlvkqscT41g4pMRmS5hj+L6UAfbNO2IuEtEOM1d9H0AsnLJY3L6UguDO\nqrGxlerqYsv4HUiYP9LlqJeT1Oevq1tJLS0bvOX7tttz3UQNDWunskJlmb/Qjs327navTZ3U0OD7\nA9bQ8PBw9N5MZ9bJMqNU64Ig1dpuJY0K/yInVuOcF7PwR+xZYiu13v2FuUOR7qb4bCEr6WkNsc1f\narbHBN5fG0BG0sPBa2uoqWm91+kUH02GyUOhABUbkU43Wo2VEyhF4HQUrFQrKvyLnPiqW7KeqR/R\nE2acdnpiKxE5YbkBcfDGF8OOR+XIDGOUYglQzc1t1NKygZqbu6ir6zxbRiCrE/FNPKU5OouRJdhz\n6ZjUUbBSjZQr/BU7d40xl4A9ivUA/oaIPhbZ5hMA3gquJfxuIjpW6XGrjZtv/gh++7ffidOnJS7+\nFQA/xuQkARgAJzZtBXAcnDB0AzjO/SKw4/YPwAlQW8HOS0nEOg3gArBz+BvRY190US++9rXQibsN\nwBGw0/N6AO8H8EcAlgHYhje/mXDPPXdNfUIc0/ff/9OIg3YTfOdvU9PjifK+M2UhHJbqJFVqinJ6\nC/kDi/2j4ILlywA8AOD8YJu3Afh7+/jXAXwzY19z1y0uEkZHRymf30rpWHSJBxfzzGpyIZrhaDo0\nvRSopWWjHdnHUvgL1NS0jnp7e8nVstlBUlq4rq6dWlpkrdfpR9LpekOywEqhpDVTK71+apJRFAcW\naMT/JgCPEtGPAMAY8wUAbwfwA2+bS2FX6yCifzLGtBpj1hHRjys8dtUxNDSEc8+9A489VkByUZAv\ngssX1AOoA/BaNDf/EC+99JR9PrUH8Mg+WVqgr+9xAMDhw+Fs4LUAbsH4+FZ0dBzC6OjNuPXWO/Ds\ns88CeA06OgiFwt8CgA1n5CzSYtmbkq25d+/NOH78e5icfDeAU2hq+nxFo/xS0ExRRZkdKhX+swE8\n6T0/CR7VT7dNNzh3v8YZAwu1xJ/vBpt0DmLdui5MTDyIiYlJ+GaUhoZnUFf3M0xMcGkBX6S5HMTH\nwH3tbvAKVI5i5oyZCKrsxy9JMF8irCYZRamcSoWfStzOlPm5JUehcDmOHv3PmJgAeOR+C9KF2m7B\n6tX78drX5nDsmCR38Sh+69YtuPnmj0RFWsT72Wd/hn/5l9OYmDgF4GBJ9VfKEVQVYUWpTioV/qcA\nbPCebwCP6Itt021fS7Fv376px4ODgxgcHKyweYuPoaEhHDr0ObzjHe/Diy+OZ26XXN9UCnsdREfH\noUzB9V9PZgmrSURRlgJHjhzBkSNHKt6PYf9AmR82pgHAQ+Byjk8D+BaA3yWiH3jbvA3AB4nobcaY\nPgAfJ6K+yL6okrZUG2NjY7j00t/BxEQDkqaeYTQ1fR53380jdL+UQFPTHtx9t4q4oiiMMQZEFFpU\npv9cpWJrjHkrXDjnnUR0szHmCgAgok/bbT4J4BJw7dv3ENF3IvupKeEHWPz37r0RTzxxCm1tLVi9\nehU6OtZN1fGRbcL6PoqiKMACCv9sUYvCryiKUgnlCn/d9JsoiqIoSwkVfkVRlBpDhV9RFKXGUOFX\nFEWpMVT4FUVRagwVfkVRlBpDhV9RFKXGUOFXFEWpMVT4FUVRagwVfkVRlBpDhV9RFKXGUOFXFEWp\nMVT4FUVRagwVfkVRlBpDhV9RFKXGUOFXFEWpMVT4FUVRagwVfkVRlBpDhV9RFKXGUOFXFEWpMVT4\nFUVRagwVfkVRlBpDhV9RFKXGUOFXFEWpMVT4FUVRagwVfkVRlBpDhV9RFKXGUOFXFEWpMcoWfmNM\nzhhz2BjzsDHmHmNMa8Z2nzXG/NgYc6L8ZiqKoiizRSUj/hsAHCai8wB83T6PsR/AJRUcp+o5cuTI\nQjdhTtHzq270/GqPSoT/UgAH7eODAP5DbCMi+gcA/1rBcaqepf7F0/OrbvT8ao9KhH8dEf3YPv4x\ngHWz0B5FURRljmko9qYx5jCA9ZG3PuQ/ISIyxtBsNkxRFEWZGwxReXptjHkQwCARnTLGdAH4X0TU\nk7HtOQC+TERbi+xPOw5FUZQZQkRmpp8pOuKfhkMAhgF8zP7/UgX7KqvxiqIoysypxMb/JwC2GWMe\nBvBb9jmMMWcZY74iGxlj/g7APwI4zxjzpDHmPZU0WFEURamMsk09iqIoSnWyIJm7xph3GGP+xRhz\nxhhzYZHtfmSM+a4x5pgx5lvz2cZKmMH5XWKMedAY84gxZs98trESZpC8V1X3r5T7YYz5hH3/uDGm\nd77bWC7TnZsxZtAY83N7r44ZYz68EO0sh1KSRKv1vgHTn19Z946I5v0PQA+A8wD8LwAXFtnucQC5\nhWjjXJ8fgHoAjwI4B8AyAA8AOH+h217i+f0pgOvt4z0A/qTa718p9wPA2wD8vX386wC+udDtnsVz\nGwRwaKHbWub5/TsAvQBOZLxflfdtBuc343u3ICN+InqQiB4ucfOqc/qWeH5vAvAoEf2IiF4B8AUA\nb5/71s0KJSXvWarl/pVyP6bOm4j+CUCrMaYa8ldK/a5Vy71KQNMniVbrfQNQchLsjO7dYi/SRgC+\nZoz5tjHmfQvdmFnmbABPes9P2teqgVKT96rp/pVyP2LbdM9xu2aDUs6NAPxbawr5e2PM5nlr3dxT\nrfetVGZ87yoJ5yxKkeSv/5OIvlzibvqJ6BljTCeAw8aYB23vt+DMwvktaq/6LCXvLdr7F6HU+xGO\nrBb1fbSU0sbvANhARC8bY94KDs8+b26bNa9U430rlRnfuzkTfiLaNgv7eMb+/6kx5m7wlHVRCMcs\nnN9TADZ4zzeARyKLgmLnZx1N68kl7/0kYx+L9v5FKOV+hNt029cWO9OeGxG96D3+qjHmr4wxOSJ6\nbp7aOJdU630riXLu3WIw9URtU8aYlcaYFvt4FYCLAVRjaecs29u3AbzaGHOOMaYRwDvBSXHVgCTv\nARnJe1V4/0q5H4cA/B8AYIzpA/C8Z/JazEx7bsaYdcYYYx+/CRzqvRREH6je+1YSZd27BfJSXwa2\nuY0DOAXgq/b1swB8xT4+Fxx98ACA7wHYu9De9dk8P/v8rQAeAkdcVNP55QB8DcDDAO4B0LoU7l/s\nfgC4AsAV3jaftO8fR5GItMX2N925AfiAvU8PgBMu+xa6zTM4t78D8DSACfu7+/2lct9KOb9y7p0m\ncCmKotQYi8HUoyiKoswjKvyKoig1hgq/oihKjaHCryiKUmOo8CuKotQYKvyKoig1hgq/oihKjaHC\nryiKUmP8/+3YcGm5VOP/AAAAAElFTkSuQmCC\n",
      "text/plain": [
       "<matplotlib.figure.Figure at 0x7f9ca50>"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "plt.scatter(sn_df_gt_20[\"diff\"],sn_df_gt_20['n_af_std'])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 183,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "sn_df_gt_20 = sn_df[(sn_df['n']+sn_df[\"s\"])>20]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 189,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "n            0\n",
       "n_af_std   NaN\n",
       "s            5\n",
       "diff        -1\n",
       "Name: NFIX, dtype: float64"
      ]
     },
     "execution_count": 189,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "sn_df.ix['NFIX']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 196,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/home/GMI/hannes.svardal/lib/python/virtualenvs/ver/lib/python2.7/site-packages/ipykernel/__main__.py:1: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame.\n",
      "Try using .loc[row_indexer,col_indexer] = value instead\n",
      "\n",
      "See the the caveats in the documentation: http://pandas.pydata.org/pandas-docs/stable/indexing.html#indexing-view-versus-copy\n",
      "  if __name__ == '__main__':\n"
     ]
    }
   ],
   "source": [
    "sn_df_gt_20['compound'] = np.sign(sn_df_gt_20['diff'])*sn_df_gt_20['diff']**2+sn_df_gt_20['n_af_std']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 197,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div style=\"max-height:1000px;max-width:1500px;overflow:auto;\">\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>n</th>\n",
       "      <th>n_af_std</th>\n",
       "      <th>s</th>\n",
       "      <th>diff</th>\n",
       "      <th>compound</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>gene_symbol</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>ENSCSAG00000008486</th>\n",
       "      <td> 37</td>\n",
       "      <td> 0.087828</td>\n",
       "      <td>  2</td>\n",
       "      <td> 0.897436</td>\n",
       "      <td> 0.893219</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>OR2C3</th>\n",
       "      <td> 22</td>\n",
       "      <td> 0.059000</td>\n",
       "      <td>  1</td>\n",
       "      <td> 0.913043</td>\n",
       "      <td> 0.892649</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSCSAG00000001298</th>\n",
       "      <td> 21</td>\n",
       "      <td> 0.191565</td>\n",
       "      <td>  2</td>\n",
       "      <td> 0.826087</td>\n",
       "      <td> 0.873985</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSCSAG00000006038</th>\n",
       "      <td> 19</td>\n",
       "      <td> 0.097611</td>\n",
       "      <td>  2</td>\n",
       "      <td> 0.809524</td>\n",
       "      <td> 0.752940</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSCSAG00000008252</th>\n",
       "      <td> 20</td>\n",
       "      <td> 0.068894</td>\n",
       "      <td>  2</td>\n",
       "      <td> 0.818182</td>\n",
       "      <td> 0.738316</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>C1orf110</th>\n",
       "      <td> 19</td>\n",
       "      <td> 0.078408</td>\n",
       "      <td>  2</td>\n",
       "      <td> 0.809524</td>\n",
       "      <td> 0.733736</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>CD4</th>\n",
       "      <td> 20</td>\n",
       "      <td> 0.172251</td>\n",
       "      <td>  3</td>\n",
       "      <td> 0.739130</td>\n",
       "      <td> 0.718565</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSCSAG00000009660</th>\n",
       "      <td> 34</td>\n",
       "      <td> 0.135081</td>\n",
       "      <td>  5</td>\n",
       "      <td> 0.743590</td>\n",
       "      <td> 0.688007</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSCSAG00000011762</th>\n",
       "      <td> 38</td>\n",
       "      <td> 0.092997</td>\n",
       "      <td>  5</td>\n",
       "      <td> 0.767442</td>\n",
       "      <td> 0.681964</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSCSAG00000001676</th>\n",
       "      <td> 19</td>\n",
       "      <td> 0.152873</td>\n",
       "      <td>  3</td>\n",
       "      <td> 0.727273</td>\n",
       "      <td> 0.681798</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ZNF37A</th>\n",
       "      <td> 25</td>\n",
       "      <td> 0.064266</td>\n",
       "      <td>  3</td>\n",
       "      <td> 0.785714</td>\n",
       "      <td> 0.681613</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSCSAG00000007682</th>\n",
       "      <td> 29</td>\n",
       "      <td> 0.101541</td>\n",
       "      <td>  4</td>\n",
       "      <td> 0.757576</td>\n",
       "      <td> 0.675462</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSCSAG00000000429</th>\n",
       "      <td> 20</td>\n",
       "      <td> 0.125685</td>\n",
       "      <td>  3</td>\n",
       "      <td> 0.739130</td>\n",
       "      <td> 0.671999</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>OR13G1</th>\n",
       "      <td> 24</td>\n",
       "      <td> 0.155915</td>\n",
       "      <td>  4</td>\n",
       "      <td> 0.714286</td>\n",
       "      <td> 0.666119</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSCSAG00000003968</th>\n",
       "      <td> 18</td>\n",
       "      <td> 0.142385</td>\n",
       "      <td>  3</td>\n",
       "      <td> 0.714286</td>\n",
       "      <td> 0.652589</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSCSAG00000014373</th>\n",
       "      <td> 24</td>\n",
       "      <td> 0.212765</td>\n",
       "      <td>  5</td>\n",
       "      <td> 0.655172</td>\n",
       "      <td> 0.642016</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ZNF534</th>\n",
       "      <td> 26</td>\n",
       "      <td> 0.099229</td>\n",
       "      <td>  4</td>\n",
       "      <td> 0.733333</td>\n",
       "      <td> 0.637007</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>N4BP2L2</th>\n",
       "      <td> 26</td>\n",
       "      <td> 0.096366</td>\n",
       "      <td>  4</td>\n",
       "      <td> 0.733333</td>\n",
       "      <td> 0.634144</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>TSPAN8</th>\n",
       "      <td> 23</td>\n",
       "      <td> 0.212648</td>\n",
       "      <td>  5</td>\n",
       "      <td> 0.642857</td>\n",
       "      <td> 0.625914</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>U8</th>\n",
       "      <td> 43</td>\n",
       "      <td> 0.152555</td>\n",
       "      <td>  8</td>\n",
       "      <td> 0.686275</td>\n",
       "      <td> 0.623528</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSCSAG00000001778</th>\n",
       "      <td> 24</td>\n",
       "      <td> 0.105093</td>\n",
       "      <td>  4</td>\n",
       "      <td> 0.714286</td>\n",
       "      <td> 0.615297</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>SLFN13</th>\n",
       "      <td> 81</td>\n",
       "      <td> 0.240721</td>\n",
       "      <td> 20</td>\n",
       "      <td> 0.603960</td>\n",
       "      <td> 0.605489</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>CCNB3</th>\n",
       "      <td> 24</td>\n",
       "      <td> 0.243602</td>\n",
       "      <td>  6</td>\n",
       "      <td> 0.600000</td>\n",
       "      <td> 0.603602</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>UPP2</th>\n",
       "      <td> 25</td>\n",
       "      <td> 0.158459</td>\n",
       "      <td>  5</td>\n",
       "      <td> 0.666667</td>\n",
       "      <td> 0.602903</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSCSAG00000007185</th>\n",
       "      <td> 20</td>\n",
       "      <td> 0.153511</td>\n",
       "      <td>  4</td>\n",
       "      <td> 0.666667</td>\n",
       "      <td> 0.597955</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSCSAG00000014713</th>\n",
       "      <td> 18</td>\n",
       "      <td> 0.085261</td>\n",
       "      <td>  3</td>\n",
       "      <td> 0.714286</td>\n",
       "      <td> 0.595465</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSCSAG00000000158</th>\n",
       "      <td> 25</td>\n",
       "      <td> 0.141414</td>\n",
       "      <td>  5</td>\n",
       "      <td> 0.666667</td>\n",
       "      <td> 0.585858</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSCSAG00000016189</th>\n",
       "      <td> 34</td>\n",
       "      <td> 0.201536</td>\n",
       "      <td>  8</td>\n",
       "      <td> 0.619048</td>\n",
       "      <td> 0.584756</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSCSAG00000009726</th>\n",
       "      <td> 20</td>\n",
       "      <td> 0.135916</td>\n",
       "      <td>  4</td>\n",
       "      <td> 0.666667</td>\n",
       "      <td> 0.580361</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSCSAG00000016129</th>\n",
       "      <td> 23</td>\n",
       "      <td> 0.162240</td>\n",
       "      <td>  5</td>\n",
       "      <td> 0.642857</td>\n",
       "      <td> 0.575505</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>TRIO</th>\n",
       "      <td>  2</td>\n",
       "      <td> 0.070168</td>\n",
       "      <td> 93</td>\n",
       "      <td>-0.957895</td>\n",
       "      <td>-0.847395</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>TRIM71</th>\n",
       "      <td>  1</td>\n",
       "      <td> 0.000000</td>\n",
       "      <td> 25</td>\n",
       "      <td>-0.923077</td>\n",
       "      <td>-0.852071</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>MYH7</th>\n",
       "      <td>  1</td>\n",
       "      <td> 0.076547</td>\n",
       "      <td> 56</td>\n",
       "      <td>-0.964912</td>\n",
       "      <td>-0.854509</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ERCC2</th>\n",
       "      <td>  1</td>\n",
       "      <td> 0.004002</td>\n",
       "      <td> 29</td>\n",
       "      <td>-0.933333</td>\n",
       "      <td>-0.867109</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>KIRREL</th>\n",
       "      <td>  1</td>\n",
       "      <td> 0.000000</td>\n",
       "      <td> 29</td>\n",
       "      <td>-0.933333</td>\n",
       "      <td>-0.871111</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GRIK3</th>\n",
       "      <td>  1</td>\n",
       "      <td> 0.037113</td>\n",
       "      <td> 44</td>\n",
       "      <td>-0.955556</td>\n",
       "      <td>-0.875973</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>LRRC8A</th>\n",
       "      <td>  1</td>\n",
       "      <td> 0.004002</td>\n",
       "      <td> 34</td>\n",
       "      <td>-0.942857</td>\n",
       "      <td>-0.884977</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ATP2B2</th>\n",
       "      <td>  1</td>\n",
       "      <td> 0.013845</td>\n",
       "      <td> 43</td>\n",
       "      <td>-0.954545</td>\n",
       "      <td>-0.897312</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>MTOR</th>\n",
       "      <td>  1</td>\n",
       "      <td> 0.004002</td>\n",
       "      <td> 39</td>\n",
       "      <td>-0.950000</td>\n",
       "      <td>-0.898498</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GRIN2B</th>\n",
       "      <td>  1</td>\n",
       "      <td> 0.015674</td>\n",
       "      <td> 53</td>\n",
       "      <td>-0.962963</td>\n",
       "      <td>-0.911624</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ITPR1</th>\n",
       "      <td>  1</td>\n",
       "      <td> 0.018557</td>\n",
       "      <td> 57</td>\n",
       "      <td>-0.965517</td>\n",
       "      <td>-0.913667</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ADCY5</th>\n",
       "      <td>  0</td>\n",
       "      <td>      NaN</td>\n",
       "      <td> 27</td>\n",
       "      <td>-1.000000</td>\n",
       "      <td>      NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ANO3</th>\n",
       "      <td>  0</td>\n",
       "      <td>      NaN</td>\n",
       "      <td> 22</td>\n",
       "      <td>-1.000000</td>\n",
       "      <td>      NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ATP1A2</th>\n",
       "      <td>  0</td>\n",
       "      <td>      NaN</td>\n",
       "      <td> 42</td>\n",
       "      <td>-1.000000</td>\n",
       "      <td>      NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ATP1A3</th>\n",
       "      <td>  0</td>\n",
       "      <td>      NaN</td>\n",
       "      <td> 32</td>\n",
       "      <td>-1.000000</td>\n",
       "      <td>      NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ATP2A1</th>\n",
       "      <td>  0</td>\n",
       "      <td>      NaN</td>\n",
       "      <td> 27</td>\n",
       "      <td>-1.000000</td>\n",
       "      <td>      NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>CYFIP1</th>\n",
       "      <td>  0</td>\n",
       "      <td>      NaN</td>\n",
       "      <td> 24</td>\n",
       "      <td>-1.000000</td>\n",
       "      <td>      NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSCSAG00000000561</th>\n",
       "      <td>  0</td>\n",
       "      <td>      NaN</td>\n",
       "      <td> 26</td>\n",
       "      <td>-1.000000</td>\n",
       "      <td>      NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GRIN2D</th>\n",
       "      <td>  0</td>\n",
       "      <td>      NaN</td>\n",
       "      <td> 28</td>\n",
       "      <td>-1.000000</td>\n",
       "      <td>      NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>HS6ST1</th>\n",
       "      <td>  0</td>\n",
       "      <td>      NaN</td>\n",
       "      <td> 21</td>\n",
       "      <td>-1.000000</td>\n",
       "      <td>      NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>LARGE</th>\n",
       "      <td>  0</td>\n",
       "      <td>      NaN</td>\n",
       "      <td> 26</td>\n",
       "      <td>-1.000000</td>\n",
       "      <td>      NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>MYH10</th>\n",
       "      <td>  0</td>\n",
       "      <td>      NaN</td>\n",
       "      <td> 57</td>\n",
       "      <td>-1.000000</td>\n",
       "      <td>      NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>PCSK2</th>\n",
       "      <td>  0</td>\n",
       "      <td>      NaN</td>\n",
       "      <td> 29</td>\n",
       "      <td>-1.000000</td>\n",
       "      <td>      NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>PRKCE</th>\n",
       "      <td>  0</td>\n",
       "      <td>      NaN</td>\n",
       "      <td> 21</td>\n",
       "      <td>-1.000000</td>\n",
       "      <td>      NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>SCN2A</th>\n",
       "      <td>  0</td>\n",
       "      <td>      NaN</td>\n",
       "      <td> 35</td>\n",
       "      <td>-1.000000</td>\n",
       "      <td>      NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>SCN8A</th>\n",
       "      <td>  0</td>\n",
       "      <td>      NaN</td>\n",
       "      <td> 26</td>\n",
       "      <td>-1.000000</td>\n",
       "      <td>      NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>SLC5A1</th>\n",
       "      <td>  0</td>\n",
       "      <td>      NaN</td>\n",
       "      <td> 22</td>\n",
       "      <td>-1.000000</td>\n",
       "      <td>      NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>SLC6A3</th>\n",
       "      <td>  0</td>\n",
       "      <td>      NaN</td>\n",
       "      <td> 29</td>\n",
       "      <td>-1.000000</td>\n",
       "      <td>      NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>SLC8A2</th>\n",
       "      <td>  0</td>\n",
       "      <td>      NaN</td>\n",
       "      <td> 25</td>\n",
       "      <td>-1.000000</td>\n",
       "      <td>      NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>TNPO2</th>\n",
       "      <td>  0</td>\n",
       "      <td>      NaN</td>\n",
       "      <td> 35</td>\n",
       "      <td>-1.000000</td>\n",
       "      <td>      NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5329 rows  5 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                     n  n_af_std   s      diff  compound\n",
       "gene_symbol                                             \n",
       "ENSCSAG00000008486  37  0.087828   2  0.897436  0.893219\n",
       "OR2C3               22  0.059000   1  0.913043  0.892649\n",
       "ENSCSAG00000001298  21  0.191565   2  0.826087  0.873985\n",
       "ENSCSAG00000006038  19  0.097611   2  0.809524  0.752940\n",
       "ENSCSAG00000008252  20  0.068894   2  0.818182  0.738316\n",
       "C1orf110            19  0.078408   2  0.809524  0.733736\n",
       "CD4                 20  0.172251   3  0.739130  0.718565\n",
       "ENSCSAG00000009660  34  0.135081   5  0.743590  0.688007\n",
       "ENSCSAG00000011762  38  0.092997   5  0.767442  0.681964\n",
       "ENSCSAG00000001676  19  0.152873   3  0.727273  0.681798\n",
       "ZNF37A              25  0.064266   3  0.785714  0.681613\n",
       "ENSCSAG00000007682  29  0.101541   4  0.757576  0.675462\n",
       "ENSCSAG00000000429  20  0.125685   3  0.739130  0.671999\n",
       "OR13G1              24  0.155915   4  0.714286  0.666119\n",
       "ENSCSAG00000003968  18  0.142385   3  0.714286  0.652589\n",
       "ENSCSAG00000014373  24  0.212765   5  0.655172  0.642016\n",
       "ZNF534              26  0.099229   4  0.733333  0.637007\n",
       "N4BP2L2             26  0.096366   4  0.733333  0.634144\n",
       "TSPAN8              23  0.212648   5  0.642857  0.625914\n",
       "U8                  43  0.152555   8  0.686275  0.623528\n",
       "ENSCSAG00000001778  24  0.105093   4  0.714286  0.615297\n",
       "SLFN13              81  0.240721  20  0.603960  0.605489\n",
       "CCNB3               24  0.243602   6  0.600000  0.603602\n",
       "UPP2                25  0.158459   5  0.666667  0.602903\n",
       "ENSCSAG00000007185  20  0.153511   4  0.666667  0.597955\n",
       "ENSCSAG00000014713  18  0.085261   3  0.714286  0.595465\n",
       "ENSCSAG00000000158  25  0.141414   5  0.666667  0.585858\n",
       "ENSCSAG00000016189  34  0.201536   8  0.619048  0.584756\n",
       "ENSCSAG00000009726  20  0.135916   4  0.666667  0.580361\n",
       "ENSCSAG00000016129  23  0.162240   5  0.642857  0.575505\n",
       "...                 ..       ...  ..       ...       ...\n",
       "TRIO                 2  0.070168  93 -0.957895 -0.847395\n",
       "TRIM71               1  0.000000  25 -0.923077 -0.852071\n",
       "MYH7                 1  0.076547  56 -0.964912 -0.854509\n",
       "ERCC2                1  0.004002  29 -0.933333 -0.867109\n",
       "KIRREL               1  0.000000  29 -0.933333 -0.871111\n",
       "GRIK3                1  0.037113  44 -0.955556 -0.875973\n",
       "LRRC8A               1  0.004002  34 -0.942857 -0.884977\n",
       "ATP2B2               1  0.013845  43 -0.954545 -0.897312\n",
       "MTOR                 1  0.004002  39 -0.950000 -0.898498\n",
       "GRIN2B               1  0.015674  53 -0.962963 -0.911624\n",
       "ITPR1                1  0.018557  57 -0.965517 -0.913667\n",
       "ADCY5                0       NaN  27 -1.000000       NaN\n",
       "ANO3                 0       NaN  22 -1.000000       NaN\n",
       "ATP1A2               0       NaN  42 -1.000000       NaN\n",
       "ATP1A3               0       NaN  32 -1.000000       NaN\n",
       "ATP2A1               0       NaN  27 -1.000000       NaN\n",
       "CYFIP1               0       NaN  24 -1.000000       NaN\n",
       "ENSCSAG00000000561   0       NaN  26 -1.000000       NaN\n",
       "GRIN2D               0       NaN  28 -1.000000       NaN\n",
       "HS6ST1               0       NaN  21 -1.000000       NaN\n",
       "LARGE                0       NaN  26 -1.000000       NaN\n",
       "MYH10                0       NaN  57 -1.000000       NaN\n",
       "PCSK2                0       NaN  29 -1.000000       NaN\n",
       "PRKCE                0       NaN  21 -1.000000       NaN\n",
       "SCN2A                0       NaN  35 -1.000000       NaN\n",
       "SCN8A                0       NaN  26 -1.000000       NaN\n",
       "SLC5A1               0       NaN  22 -1.000000       NaN\n",
       "SLC6A3               0       NaN  29 -1.000000       NaN\n",
       "SLC8A2               0       NaN  25 -1.000000       NaN\n",
       "TNPO2                0       NaN  35 -1.000000       NaN\n",
       "\n",
       "[5329 rows x 5 columns]"
      ]
     },
     "execution_count": 197,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "sn_df_gt_20.sort('compound',ascending=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 184,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div style=\"max-height:1000px;max-width:1500px;overflow:auto;\">\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>n</th>\n",
       "      <th>n_af_std</th>\n",
       "      <th>s</th>\n",
       "      <th>diff</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>gene_symbol</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>OR2C3</th>\n",
       "      <td> 22</td>\n",
       "      <td> 0.059000</td>\n",
       "      <td>  1</td>\n",
       "      <td> 0.913043</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSCSAG00000008486</th>\n",
       "      <td> 37</td>\n",
       "      <td> 0.087828</td>\n",
       "      <td>  2</td>\n",
       "      <td> 0.897436</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSCSAG00000001298</th>\n",
       "      <td> 21</td>\n",
       "      <td> 0.191565</td>\n",
       "      <td>  2</td>\n",
       "      <td> 0.826087</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSCSAG00000008252</th>\n",
       "      <td> 20</td>\n",
       "      <td> 0.068894</td>\n",
       "      <td>  2</td>\n",
       "      <td> 0.818182</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>C1orf110</th>\n",
       "      <td> 19</td>\n",
       "      <td> 0.078408</td>\n",
       "      <td>  2</td>\n",
       "      <td> 0.809524</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSCSAG00000006038</th>\n",
       "      <td> 19</td>\n",
       "      <td> 0.097611</td>\n",
       "      <td>  2</td>\n",
       "      <td> 0.809524</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ZNF37A</th>\n",
       "      <td> 25</td>\n",
       "      <td> 0.064266</td>\n",
       "      <td>  3</td>\n",
       "      <td> 0.785714</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSCSAG00000011762</th>\n",
       "      <td> 38</td>\n",
       "      <td> 0.092997</td>\n",
       "      <td>  5</td>\n",
       "      <td> 0.767442</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSCSAG00000007682</th>\n",
       "      <td> 29</td>\n",
       "      <td> 0.101541</td>\n",
       "      <td>  4</td>\n",
       "      <td> 0.757576</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSCSAG00000009660</th>\n",
       "      <td> 34</td>\n",
       "      <td> 0.135081</td>\n",
       "      <td>  5</td>\n",
       "      <td> 0.743590</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSCSAG00000000429</th>\n",
       "      <td> 20</td>\n",
       "      <td> 0.125685</td>\n",
       "      <td>  3</td>\n",
       "      <td> 0.739130</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>CD4</th>\n",
       "      <td> 20</td>\n",
       "      <td> 0.172251</td>\n",
       "      <td>  3</td>\n",
       "      <td> 0.739130</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>N4BP2L2</th>\n",
       "      <td> 26</td>\n",
       "      <td> 0.096366</td>\n",
       "      <td>  4</td>\n",
       "      <td> 0.733333</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ZNF534</th>\n",
       "      <td> 26</td>\n",
       "      <td> 0.099229</td>\n",
       "      <td>  4</td>\n",
       "      <td> 0.733333</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSCSAG00000001676</th>\n",
       "      <td> 19</td>\n",
       "      <td> 0.152873</td>\n",
       "      <td>  3</td>\n",
       "      <td> 0.727273</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSCSAG00000003968</th>\n",
       "      <td> 18</td>\n",
       "      <td> 0.142385</td>\n",
       "      <td>  3</td>\n",
       "      <td> 0.714286</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSCSAG00000001778</th>\n",
       "      <td> 24</td>\n",
       "      <td> 0.105093</td>\n",
       "      <td>  4</td>\n",
       "      <td> 0.714286</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>OR13G1</th>\n",
       "      <td> 24</td>\n",
       "      <td> 0.155915</td>\n",
       "      <td>  4</td>\n",
       "      <td> 0.714286</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSCSAG00000014713</th>\n",
       "      <td> 18</td>\n",
       "      <td> 0.085261</td>\n",
       "      <td>  3</td>\n",
       "      <td> 0.714286</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSCSAG00000004330</th>\n",
       "      <td> 22</td>\n",
       "      <td> 0.069908</td>\n",
       "      <td>  4</td>\n",
       "      <td> 0.692308</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>U8</th>\n",
       "      <td> 43</td>\n",
       "      <td> 0.152555</td>\n",
       "      <td>  8</td>\n",
       "      <td> 0.686275</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>SNORA73</th>\n",
       "      <td> 31</td>\n",
       "      <td> 0.117827</td>\n",
       "      <td>  6</td>\n",
       "      <td> 0.675676</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSCSAG00000019331</th>\n",
       "      <td> 25</td>\n",
       "      <td> 0.128451</td>\n",
       "      <td>  5</td>\n",
       "      <td> 0.666667</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSCSAG00000009726</th>\n",
       "      <td> 20</td>\n",
       "      <td> 0.135916</td>\n",
       "      <td>  4</td>\n",
       "      <td> 0.666667</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>SERPINB10</th>\n",
       "      <td> 20</td>\n",
       "      <td> 0.114867</td>\n",
       "      <td>  4</td>\n",
       "      <td> 0.666667</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSCSAG00000007185</th>\n",
       "      <td> 20</td>\n",
       "      <td> 0.153511</td>\n",
       "      <td>  4</td>\n",
       "      <td> 0.666667</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>OR52J3</th>\n",
       "      <td> 35</td>\n",
       "      <td> 0.051908</td>\n",
       "      <td>  7</td>\n",
       "      <td> 0.666667</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>PARP15</th>\n",
       "      <td> 25</td>\n",
       "      <td> 0.075902</td>\n",
       "      <td>  5</td>\n",
       "      <td> 0.666667</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSCSAG00000000158</th>\n",
       "      <td> 25</td>\n",
       "      <td> 0.141414</td>\n",
       "      <td>  5</td>\n",
       "      <td> 0.666667</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSCSAG00000006010</th>\n",
       "      <td> 25</td>\n",
       "      <td> 0.076820</td>\n",
       "      <td>  5</td>\n",
       "      <td> 0.666667</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>PANK4</th>\n",
       "      <td>  1</td>\n",
       "      <td> 0.063110</td>\n",
       "      <td> 35</td>\n",
       "      <td>-0.944444</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>KSR2</th>\n",
       "      <td>  1</td>\n",
       "      <td> 0.283227</td>\n",
       "      <td> 37</td>\n",
       "      <td>-0.947368</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>MTOR</th>\n",
       "      <td>  1</td>\n",
       "      <td> 0.004002</td>\n",
       "      <td> 39</td>\n",
       "      <td>-0.950000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>KLHL25</th>\n",
       "      <td>  1</td>\n",
       "      <td> 0.114820</td>\n",
       "      <td> 43</td>\n",
       "      <td>-0.954545</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ATP2B2</th>\n",
       "      <td>  1</td>\n",
       "      <td> 0.013845</td>\n",
       "      <td> 43</td>\n",
       "      <td>-0.954545</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GRIK3</th>\n",
       "      <td>  1</td>\n",
       "      <td> 0.037113</td>\n",
       "      <td> 44</td>\n",
       "      <td>-0.955556</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>TRIO</th>\n",
       "      <td>  2</td>\n",
       "      <td> 0.070168</td>\n",
       "      <td> 93</td>\n",
       "      <td>-0.957895</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>MYH2</th>\n",
       "      <td>  1</td>\n",
       "      <td> 0.102062</td>\n",
       "      <td> 50</td>\n",
       "      <td>-0.960784</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GRIN2B</th>\n",
       "      <td>  1</td>\n",
       "      <td> 0.015674</td>\n",
       "      <td> 53</td>\n",
       "      <td>-0.962963</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>MYH7</th>\n",
       "      <td>  1</td>\n",
       "      <td> 0.076547</td>\n",
       "      <td> 56</td>\n",
       "      <td>-0.964912</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ITPR1</th>\n",
       "      <td>  1</td>\n",
       "      <td> 0.018557</td>\n",
       "      <td> 57</td>\n",
       "      <td>-0.965517</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>PCSK2</th>\n",
       "      <td>  0</td>\n",
       "      <td>      NaN</td>\n",
       "      <td> 29</td>\n",
       "      <td>-1.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ADCY5</th>\n",
       "      <td>  0</td>\n",
       "      <td>      NaN</td>\n",
       "      <td> 27</td>\n",
       "      <td>-1.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSCSAG00000000561</th>\n",
       "      <td>  0</td>\n",
       "      <td>      NaN</td>\n",
       "      <td> 26</td>\n",
       "      <td>-1.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>SCN2A</th>\n",
       "      <td>  0</td>\n",
       "      <td>      NaN</td>\n",
       "      <td> 35</td>\n",
       "      <td>-1.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ATP2A1</th>\n",
       "      <td>  0</td>\n",
       "      <td>      NaN</td>\n",
       "      <td> 27</td>\n",
       "      <td>-1.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>MYH10</th>\n",
       "      <td>  0</td>\n",
       "      <td>      NaN</td>\n",
       "      <td> 57</td>\n",
       "      <td>-1.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ANO3</th>\n",
       "      <td>  0</td>\n",
       "      <td>      NaN</td>\n",
       "      <td> 22</td>\n",
       "      <td>-1.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>LARGE</th>\n",
       "      <td>  0</td>\n",
       "      <td>      NaN</td>\n",
       "      <td> 26</td>\n",
       "      <td>-1.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ATP1A2</th>\n",
       "      <td>  0</td>\n",
       "      <td>      NaN</td>\n",
       "      <td> 42</td>\n",
       "      <td>-1.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>SLC8A2</th>\n",
       "      <td>  0</td>\n",
       "      <td>      NaN</td>\n",
       "      <td> 25</td>\n",
       "      <td>-1.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>PRKCE</th>\n",
       "      <td>  0</td>\n",
       "      <td>      NaN</td>\n",
       "      <td> 21</td>\n",
       "      <td>-1.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>TNPO2</th>\n",
       "      <td>  0</td>\n",
       "      <td>      NaN</td>\n",
       "      <td> 35</td>\n",
       "      <td>-1.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ATP1A3</th>\n",
       "      <td>  0</td>\n",
       "      <td>      NaN</td>\n",
       "      <td> 32</td>\n",
       "      <td>-1.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>SLC5A1</th>\n",
       "      <td>  0</td>\n",
       "      <td>      NaN</td>\n",
       "      <td> 22</td>\n",
       "      <td>-1.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>HS6ST1</th>\n",
       "      <td>  0</td>\n",
       "      <td>      NaN</td>\n",
       "      <td> 21</td>\n",
       "      <td>-1.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>CYFIP1</th>\n",
       "      <td>  0</td>\n",
       "      <td>      NaN</td>\n",
       "      <td> 24</td>\n",
       "      <td>-1.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>SLC6A3</th>\n",
       "      <td>  0</td>\n",
       "      <td>      NaN</td>\n",
       "      <td> 29</td>\n",
       "      <td>-1.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>SCN8A</th>\n",
       "      <td>  0</td>\n",
       "      <td>      NaN</td>\n",
       "      <td> 26</td>\n",
       "      <td>-1.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GRIN2D</th>\n",
       "      <td>  0</td>\n",
       "      <td>      NaN</td>\n",
       "      <td> 28</td>\n",
       "      <td>-1.000000</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5329 rows  4 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                     n  n_af_std   s      diff\n",
       "gene_symbol                                   \n",
       "OR2C3               22  0.059000   1  0.913043\n",
       "ENSCSAG00000008486  37  0.087828   2  0.897436\n",
       "ENSCSAG00000001298  21  0.191565   2  0.826087\n",
       "ENSCSAG00000008252  20  0.068894   2  0.818182\n",
       "C1orf110            19  0.078408   2  0.809524\n",
       "ENSCSAG00000006038  19  0.097611   2  0.809524\n",
       "ZNF37A              25  0.064266   3  0.785714\n",
       "ENSCSAG00000011762  38  0.092997   5  0.767442\n",
       "ENSCSAG00000007682  29  0.101541   4  0.757576\n",
       "ENSCSAG00000009660  34  0.135081   5  0.743590\n",
       "ENSCSAG00000000429  20  0.125685   3  0.739130\n",
       "CD4                 20  0.172251   3  0.739130\n",
       "N4BP2L2             26  0.096366   4  0.733333\n",
       "ZNF534              26  0.099229   4  0.733333\n",
       "ENSCSAG00000001676  19  0.152873   3  0.727273\n",
       "ENSCSAG00000003968  18  0.142385   3  0.714286\n",
       "ENSCSAG00000001778  24  0.105093   4  0.714286\n",
       "OR13G1              24  0.155915   4  0.714286\n",
       "ENSCSAG00000014713  18  0.085261   3  0.714286\n",
       "ENSCSAG00000004330  22  0.069908   4  0.692308\n",
       "U8                  43  0.152555   8  0.686275\n",
       "SNORA73             31  0.117827   6  0.675676\n",
       "ENSCSAG00000019331  25  0.128451   5  0.666667\n",
       "ENSCSAG00000009726  20  0.135916   4  0.666667\n",
       "SERPINB10           20  0.114867   4  0.666667\n",
       "ENSCSAG00000007185  20  0.153511   4  0.666667\n",
       "OR52J3              35  0.051908   7  0.666667\n",
       "PARP15              25  0.075902   5  0.666667\n",
       "ENSCSAG00000000158  25  0.141414   5  0.666667\n",
       "ENSCSAG00000006010  25  0.076820   5  0.666667\n",
       "...                 ..       ...  ..       ...\n",
       "PANK4                1  0.063110  35 -0.944444\n",
       "KSR2                 1  0.283227  37 -0.947368\n",
       "MTOR                 1  0.004002  39 -0.950000\n",
       "KLHL25               1  0.114820  43 -0.954545\n",
       "ATP2B2               1  0.013845  43 -0.954545\n",
       "GRIK3                1  0.037113  44 -0.955556\n",
       "TRIO                 2  0.070168  93 -0.957895\n",
       "MYH2                 1  0.102062  50 -0.960784\n",
       "GRIN2B               1  0.015674  53 -0.962963\n",
       "MYH7                 1  0.076547  56 -0.964912\n",
       "ITPR1                1  0.018557  57 -0.965517\n",
       "PCSK2                0       NaN  29 -1.000000\n",
       "ADCY5                0       NaN  27 -1.000000\n",
       "ENSCSAG00000000561   0       NaN  26 -1.000000\n",
       "SCN2A                0       NaN  35 -1.000000\n",
       "ATP2A1               0       NaN  27 -1.000000\n",
       "MYH10                0       NaN  57 -1.000000\n",
       "ANO3                 0       NaN  22 -1.000000\n",
       "LARGE                0       NaN  26 -1.000000\n",
       "ATP1A2               0       NaN  42 -1.000000\n",
       "SLC8A2               0       NaN  25 -1.000000\n",
       "PRKCE                0       NaN  21 -1.000000\n",
       "TNPO2                0       NaN  35 -1.000000\n",
       "ATP1A3               0       NaN  32 -1.000000\n",
       "SLC5A1               0       NaN  22 -1.000000\n",
       "HS6ST1               0       NaN  21 -1.000000\n",
       "CYFIP1               0       NaN  24 -1.000000\n",
       "SLC6A3               0       NaN  29 -1.000000\n",
       "SCN8A                0       NaN  26 -1.000000\n",
       "GRIN2D               0       NaN  28 -1.000000\n",
       "\n",
       "[5329 rows x 4 columns]"
      ]
     },
     "execution_count": 184,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "sn_df_gt_20.sort(\"diff\",ascending=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 162,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div style=\"max-height:1000px;max-width:1500px;overflow:auto;\">\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th>gene_symbol</th>\n",
       "      <th>REF</th>\n",
       "      <th>ALT</th>\n",
       "      <th>FILTER</th>\n",
       "      <th>AA</th>\n",
       "      <th>AF</th>\n",
       "      <th>allele</th>\n",
       "      <th>effect</th>\n",
       "      <th>impact</th>\n",
       "      <th>errors</th>\n",
       "      <th>aet</th>\n",
       "      <th>cyn</th>\n",
       "      <th>pyn</th>\n",
       "      <th>pys</th>\n",
       "      <th>sab</th>\n",
       "      <th>sac</th>\n",
       "      <th>sar</th>\n",
       "      <th>tan</th>\n",
       "      <th>af_std</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>CHROM</th>\n",
       "      <th>POS</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>CAEX</th>\n",
       "      <th>47049265 </th>\n",
       "      <td>              AKAP4</td>\n",
       "      <td> C</td>\n",
       "      <td> A</td>\n",
       "      <td> PASS</td>\n",
       "      <td>NaN</td>\n",
       "      <td> 0.466258</td>\n",
       "      <td> A</td>\n",
       "      <td>                                       stop_gained</td>\n",
       "      <td> HIGH</td>\n",
       "      <td>                                     NaN</td>\n",
       "      <td> 0.18750</td>\n",
       "      <td> 1.00000</td>\n",
       "      <td> 1.000000</td>\n",
       "      <td> 1.000000</td>\n",
       "      <td> 0.000000</td>\n",
       "      <td> 0.000000</td>\n",
       "      <td> 0.0</td>\n",
       "      <td> 0.000000</td>\n",
       "      <td> 0.518034</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>CAE16</th>\n",
       "      <th>46180030 </th>\n",
       "      <td> ENSCSAG00000023715</td>\n",
       "      <td> A</td>\n",
       "      <td> G</td>\n",
       "      <td> PASS</td>\n",
       "      <td>NaN</td>\n",
       "      <td> 0.546012</td>\n",
       "      <td> G</td>\n",
       "      <td>                   stop_lost&amp;splice_region_variant</td>\n",
       "      <td> HIGH</td>\n",
       "      <td>       WARNING_TRANSCRIPT_NO_START_CODON</td>\n",
       "      <td> 1.00000</td>\n",
       "      <td> 1.00000</td>\n",
       "      <td> 1.000000</td>\n",
       "      <td> 1.000000</td>\n",
       "      <td> 0.000000</td>\n",
       "      <td> 0.000000</td>\n",
       "      <td> 0.0</td>\n",
       "      <td> 0.000000</td>\n",
       "      <td> 0.516398</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>CAE11</th>\n",
       "      <th>75076181 </th>\n",
       "      <td>        Metazoa_SRP</td>\n",
       "      <td> A</td>\n",
       "      <td> T</td>\n",
       "      <td> PASS</td>\n",
       "      <td>NaN</td>\n",
       "      <td> 0.447853</td>\n",
       "      <td> T</td>\n",
       "      <td>                                         stop_lost</td>\n",
       "      <td> HIGH</td>\n",
       "      <td> WARNING_TRANSCRIPT_MULTIPLE_STOP_CODONS</td>\n",
       "      <td> 0.06250</td>\n",
       "      <td> 1.00000</td>\n",
       "      <td> 0.833333</td>\n",
       "      <td> 1.000000</td>\n",
       "      <td> 0.000000</td>\n",
       "      <td> 0.000000</td>\n",
       "      <td> 0.0</td>\n",
       "      <td> 0.000000</td>\n",
       "      <td> 0.510041</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>CAE7</th>\n",
       "      <th>97439521 </th>\n",
       "      <td> ENSCSAG00000003348</td>\n",
       "      <td> G</td>\n",
       "      <td> A</td>\n",
       "      <td> PASS</td>\n",
       "      <td>NaN</td>\n",
       "      <td> 0.524540</td>\n",
       "      <td> A</td>\n",
       "      <td> splice_acceptor_variant&amp;splice_donor_variant&amp;i...</td>\n",
       "      <td> HIGH</td>\n",
       "      <td>       WARNING_TRANSCRIPT_NO_START_CODON</td>\n",
       "      <td> 0.00000</td>\n",
       "      <td> 1.00000</td>\n",
       "      <td> 1.000000</td>\n",
       "      <td> 1.000000</td>\n",
       "      <td> 0.041667</td>\n",
       "      <td> 0.014706</td>\n",
       "      <td> 0.0</td>\n",
       "      <td> 1.000000</td>\n",
       "      <td> 0.505811</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>CAE22</th>\n",
       "      <th>95483322 </th>\n",
       "      <td>                 U2</td>\n",
       "      <td> A</td>\n",
       "      <td> G</td>\n",
       "      <td> PASS</td>\n",
       "      <td>NaN</td>\n",
       "      <td> 0.438650</td>\n",
       "      <td> G</td>\n",
       "      <td>                                         stop_lost</td>\n",
       "      <td> HIGH</td>\n",
       "      <td> WARNING_TRANSCRIPT_MULTIPLE_STOP_CODONS</td>\n",
       "      <td> 0.00000</td>\n",
       "      <td> 0.93750</td>\n",
       "      <td> 0.833333</td>\n",
       "      <td> 1.000000</td>\n",
       "      <td> 0.000000</td>\n",
       "      <td> 0.000000</td>\n",
       "      <td> 0.0</td>\n",
       "      <td> 0.045455</td>\n",
       "      <td> 0.500700</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>CAE11</th>\n",
       "      <th>101983690</th>\n",
       "      <td> ENSCSAG00000018940</td>\n",
       "      <td> G</td>\n",
       "      <td> C</td>\n",
       "      <td> PASS</td>\n",
       "      <td>NaN</td>\n",
       "      <td> 0.503067</td>\n",
       "      <td> C</td>\n",
       "      <td>                                       stop_gained</td>\n",
       "      <td> HIGH</td>\n",
       "      <td>        WARNING_TRANSCRIPT_NO_STOP_CODON</td>\n",
       "      <td> 0.00000</td>\n",
       "      <td> 1.00000</td>\n",
       "      <td> 0.916667</td>\n",
       "      <td> 0.980392</td>\n",
       "      <td> 0.000000</td>\n",
       "      <td> 0.000000</td>\n",
       "      <td> 0.0</td>\n",
       "      <td> 0.954545</td>\n",
       "      <td> 0.498022</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>CAE1</th>\n",
       "      <th>9938424  </th>\n",
       "      <td>              Y_RNA</td>\n",
       "      <td> G</td>\n",
       "      <td> A</td>\n",
       "      <td> PASS</td>\n",
       "      <td>NaN</td>\n",
       "      <td> 0.500000</td>\n",
       "      <td> A</td>\n",
       "      <td>                                       stop_gained</td>\n",
       "      <td> HIGH</td>\n",
       "      <td>       WARNING_TRANSCRIPT_NO_START_CODON</td>\n",
       "      <td> 0.00000</td>\n",
       "      <td> 0.93750</td>\n",
       "      <td> 0.916667</td>\n",
       "      <td> 0.980392</td>\n",
       "      <td> 0.000000</td>\n",
       "      <td> 0.000000</td>\n",
       "      <td> 0.0</td>\n",
       "      <td> 1.000000</td>\n",
       "      <td> 0.495926</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>CAE18</th>\n",
       "      <th>17036302 </th>\n",
       "      <td>                 U6</td>\n",
       "      <td> A</td>\n",
       "      <td> T</td>\n",
       "      <td> PASS</td>\n",
       "      <td>NaN</td>\n",
       "      <td> 0.533742</td>\n",
       "      <td> T</td>\n",
       "      <td>                                       stop_gained</td>\n",
       "      <td> HIGH</td>\n",
       "      <td> WARNING_TRANSCRIPT_MULTIPLE_STOP_CODONS</td>\n",
       "      <td> 0.09375</td>\n",
       "      <td> 1.00000</td>\n",
       "      <td> 1.000000</td>\n",
       "      <td> 1.000000</td>\n",
       "      <td> 0.000000</td>\n",
       "      <td> 0.044118</td>\n",
       "      <td> 0.0</td>\n",
       "      <td> 1.000000</td>\n",
       "      <td> 0.493084</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>CAE11</th>\n",
       "      <th>48208899 </th>\n",
       "      <td>              NR4A1</td>\n",
       "      <td> T</td>\n",
       "      <td> C</td>\n",
       "      <td> PASS</td>\n",
       "      <td>NaN</td>\n",
       "      <td> 0.484663</td>\n",
       "      <td> C</td>\n",
       "      <td>               splice_donor_variant&amp;intron_variant</td>\n",
       "      <td> HIGH</td>\n",
       "      <td>                                     NaN</td>\n",
       "      <td> 0.00000</td>\n",
       "      <td> 1.00000</td>\n",
       "      <td> 1.000000</td>\n",
       "      <td> 1.000000</td>\n",
       "      <td> 0.000000</td>\n",
       "      <td> 0.000000</td>\n",
       "      <td> 0.0</td>\n",
       "      <td> 0.545455</td>\n",
       "      <td> 0.490404</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>CAE1</th>\n",
       "      <th>255393   </th>\n",
       "      <td>        Metazoa_SRP</td>\n",
       "      <td> G</td>\n",
       "      <td> A</td>\n",
       "      <td> PASS</td>\n",
       "      <td>NaN</td>\n",
       "      <td> 0.444785</td>\n",
       "      <td> A</td>\n",
       "      <td>                                       stop_gained</td>\n",
       "      <td> HIGH</td>\n",
       "      <td> WARNING_TRANSCRIPT_MULTIPLE_STOP_CODONS</td>\n",
       "      <td> 0.06250</td>\n",
       "      <td> 1.00000</td>\n",
       "      <td> 1.000000</td>\n",
       "      <td> 0.872549</td>\n",
       "      <td> 0.000000</td>\n",
       "      <td> 0.000000</td>\n",
       "      <td> 0.6</td>\n",
       "      <td> 0.181818</td>\n",
       "      <td> 0.485626</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>CAE10</th>\n",
       "      <th>76025378 </th>\n",
       "      <td>           TMEM194B</td>\n",
       "      <td> C</td>\n",
       "      <td> T</td>\n",
       "      <td> PASS</td>\n",
       "      <td>NaN</td>\n",
       "      <td> 0.463190</td>\n",
       "      <td> T</td>\n",
       "      <td>                                        start_lost</td>\n",
       "      <td> HIGH</td>\n",
       "      <td>                                     NaN</td>\n",
       "      <td> 0.00000</td>\n",
       "      <td> 0.96875</td>\n",
       "      <td> 0.916667</td>\n",
       "      <td> 1.000000</td>\n",
       "      <td> 0.000000</td>\n",
       "      <td> 0.000000</td>\n",
       "      <td> 0.0</td>\n",
       "      <td> 0.318182</td>\n",
       "      <td> 0.483629</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>CAE14</th>\n",
       "      <th>12377693 </th>\n",
       "      <td>            CCDC138</td>\n",
       "      <td> G</td>\n",
       "      <td> A</td>\n",
       "      <td> PASS</td>\n",
       "      <td>NaN</td>\n",
       "      <td> 0.447853</td>\n",
       "      <td> A</td>\n",
       "      <td>            splice_acceptor_variant&amp;intron_variant</td>\n",
       "      <td> HIGH</td>\n",
       "      <td>                                     NaN</td>\n",
       "      <td> 0.00000</td>\n",
       "      <td> 0.96875</td>\n",
       "      <td> 1.000000</td>\n",
       "      <td> 0.931373</td>\n",
       "      <td> 0.000000</td>\n",
       "      <td> 0.000000</td>\n",
       "      <td> 0.0</td>\n",
       "      <td> 0.363636</td>\n",
       "      <td> 0.482247</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>CAE2</th>\n",
       "      <th>25313286 </th>\n",
       "      <td>            SNORA71</td>\n",
       "      <td> A</td>\n",
       "      <td> G</td>\n",
       "      <td> PASS</td>\n",
       "      <td>NaN</td>\n",
       "      <td> 0.546012</td>\n",
       "      <td> G</td>\n",
       "      <td>                                         stop_lost</td>\n",
       "      <td> HIGH</td>\n",
       "      <td> WARNING_TRANSCRIPT_MULTIPLE_STOP_CODONS</td>\n",
       "      <td> 0.00000</td>\n",
       "      <td> 1.00000</td>\n",
       "      <td> 0.916667</td>\n",
       "      <td> 1.000000</td>\n",
       "      <td> 0.104167</td>\n",
       "      <td> 0.088235</td>\n",
       "      <td> 0.0</td>\n",
       "      <td> 1.000000</td>\n",
       "      <td> 0.480960</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>CAE20</th>\n",
       "      <th>42794243 </th>\n",
       "      <td> ENSCSAG00000025756</td>\n",
       "      <td> A</td>\n",
       "      <td> C</td>\n",
       "      <td> PASS</td>\n",
       "      <td>NaN</td>\n",
       "      <td> 0.515337</td>\n",
       "      <td> C</td>\n",
       "      <td>                                         stop_lost</td>\n",
       "      <td> HIGH</td>\n",
       "      <td>           WARNING_TRANSCRIPT_INCOMPLETE</td>\n",
       "      <td> 0.00000</td>\n",
       "      <td> 1.00000</td>\n",
       "      <td> 1.000000</td>\n",
       "      <td> 1.000000</td>\n",
       "      <td> 0.062500</td>\n",
       "      <td> 0.000000</td>\n",
       "      <td> 0.2</td>\n",
       "      <td> 0.772727</td>\n",
       "      <td> 0.479483</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>CAE2</th>\n",
       "      <th>33907866 </th>\n",
       "      <td>            SLC4A11</td>\n",
       "      <td> A</td>\n",
       "      <td> G</td>\n",
       "      <td> PASS</td>\n",
       "      <td>NaN</td>\n",
       "      <td> 0.500000</td>\n",
       "      <td> G</td>\n",
       "      <td>               splice_donor_variant&amp;intron_variant</td>\n",
       "      <td> HIGH</td>\n",
       "      <td>                                     NaN</td>\n",
       "      <td> 0.06250</td>\n",
       "      <td> 0.71875</td>\n",
       "      <td> 1.000000</td>\n",
       "      <td> 1.000000</td>\n",
       "      <td> 0.000000</td>\n",
       "      <td> 0.029412</td>\n",
       "      <td> 0.0</td>\n",
       "      <td> 1.000000</td>\n",
       "      <td> 0.476976</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>CAE20</th>\n",
       "      <th>14359047 </th>\n",
       "      <td>             HSD3B2</td>\n",
       "      <td> G</td>\n",
       "      <td> A</td>\n",
       "      <td> PASS</td>\n",
       "      <td>NaN</td>\n",
       "      <td> 0.128834</td>\n",
       "      <td> A</td>\n",
       "      <td>                                       stop_gained</td>\n",
       "      <td> HIGH</td>\n",
       "      <td>                                     NaN</td>\n",
       "      <td> 1.00000</td>\n",
       "      <td> 0.00000</td>\n",
       "      <td> 0.833333</td>\n",
       "      <td> 0.000000</td>\n",
       "      <td> 0.000000</td>\n",
       "      <td> 0.000000</td>\n",
       "      <td> 0.0</td>\n",
       "      <td> 0.000000</td>\n",
       "      <td> 0.476290</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>CAE2</th>\n",
       "      <th>7344987  </th>\n",
       "      <td>                 U6</td>\n",
       "      <td> T</td>\n",
       "      <td> G</td>\n",
       "      <td> PASS</td>\n",
       "      <td>NaN</td>\n",
       "      <td> 0.527607</td>\n",
       "      <td> G</td>\n",
       "      <td>                                         stop_lost</td>\n",
       "      <td> HIGH</td>\n",
       "      <td> WARNING_TRANSCRIPT_MULTIPLE_STOP_CODONS</td>\n",
       "      <td> 0.15625</td>\n",
       "      <td> 1.00000</td>\n",
       "      <td> 1.000000</td>\n",
       "      <td> 1.000000</td>\n",
       "      <td> 0.000000</td>\n",
       "      <td> 0.000000</td>\n",
       "      <td> 0.0</td>\n",
       "      <td> 0.954545</td>\n",
       "      <td> 0.473103</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>CAE4</th>\n",
       "      <th>76207128 </th>\n",
       "      <td>        Metazoa_SRP</td>\n",
       "      <td> A</td>\n",
       "      <td> G</td>\n",
       "      <td> PASS</td>\n",
       "      <td>NaN</td>\n",
       "      <td> 0.398773</td>\n",
       "      <td> G</td>\n",
       "      <td>                                         stop_lost</td>\n",
       "      <td> HIGH</td>\n",
       "      <td> WARNING_TRANSCRIPT_MULTIPLE_STOP_CODONS</td>\n",
       "      <td> 0.00000</td>\n",
       "      <td> 0.78125</td>\n",
       "      <td> 0.916667</td>\n",
       "      <td> 0.911765</td>\n",
       "      <td> 0.000000</td>\n",
       "      <td> 0.000000</td>\n",
       "      <td> 0.0</td>\n",
       "      <td> 0.045455</td>\n",
       "      <td> 0.470968</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>CAE13</th>\n",
       "      <th>76998487 </th>\n",
       "      <td>            NDUFAF4</td>\n",
       "      <td> G</td>\n",
       "      <td> T</td>\n",
       "      <td> PASS</td>\n",
       "      <td>NaN</td>\n",
       "      <td> 0.404908</td>\n",
       "      <td> T</td>\n",
       "      <td>            splice_acceptor_variant&amp;intron_variant</td>\n",
       "      <td> HIGH</td>\n",
       "      <td>                                     NaN</td>\n",
       "      <td> 0.00000</td>\n",
       "      <td> 0.96875</td>\n",
       "      <td> 0.500000</td>\n",
       "      <td> 0.931373</td>\n",
       "      <td> 0.000000</td>\n",
       "      <td> 0.000000</td>\n",
       "      <td> 0.0</td>\n",
       "      <td> 0.000000</td>\n",
       "      <td> 0.468152</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>CAE6</th>\n",
       "      <th>48190642 </th>\n",
       "      <td>             ZNF865</td>\n",
       "      <td> C</td>\n",
       "      <td> T</td>\n",
       "      <td> PASS</td>\n",
       "      <td>NaN</td>\n",
       "      <td> 0.162577</td>\n",
       "      <td> T</td>\n",
       "      <td>               splice_donor_variant&amp;intron_variant</td>\n",
       "      <td> HIGH</td>\n",
       "      <td>       WARNING_TRANSCRIPT_NO_START_CODON</td>\n",
       "      <td> 1.00000</td>\n",
       "      <td> 0.00000</td>\n",
       "      <td> 0.000000</td>\n",
       "      <td> 0.000000</td>\n",
       "      <td> 0.041667</td>\n",
       "      <td> 0.014706</td>\n",
       "      <td> 0.0</td>\n",
       "      <td> 0.818182</td>\n",
       "      <td> 0.467900</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>CAE9</th>\n",
       "      <th>45370011 </th>\n",
       "      <td>                 U6</td>\n",
       "      <td> G</td>\n",
       "      <td> A</td>\n",
       "      <td> PASS</td>\n",
       "      <td>NaN</td>\n",
       "      <td> 0.180982</td>\n",
       "      <td> A</td>\n",
       "      <td>                                       stop_gained</td>\n",
       "      <td> HIGH</td>\n",
       "      <td> WARNING_TRANSCRIPT_MULTIPLE_STOP_CODONS</td>\n",
       "      <td> 0.06250</td>\n",
       "      <td> 0.15625</td>\n",
       "      <td> 1.000000</td>\n",
       "      <td> 0.166667</td>\n",
       "      <td> 0.020833</td>\n",
       "      <td> 0.000000</td>\n",
       "      <td> 0.0</td>\n",
       "      <td> 1.000000</td>\n",
       "      <td> 0.467231</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>CAE13</th>\n",
       "      <th>369349   </th>\n",
       "      <td>                7SK</td>\n",
       "      <td> A</td>\n",
       "      <td> G</td>\n",
       "      <td> PASS</td>\n",
       "      <td>NaN</td>\n",
       "      <td> 0.407975</td>\n",
       "      <td> G</td>\n",
       "      <td>                                         stop_lost</td>\n",
       "      <td> HIGH</td>\n",
       "      <td> WARNING_TRANSCRIPT_MULTIPLE_STOP_CODONS</td>\n",
       "      <td> 0.00000</td>\n",
       "      <td> 0.96875</td>\n",
       "      <td> 0.083333</td>\n",
       "      <td> 0.950980</td>\n",
       "      <td> 0.000000</td>\n",
       "      <td> 0.000000</td>\n",
       "      <td> 0.0</td>\n",
       "      <td> 0.181818</td>\n",
       "      <td> 0.466310</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>CAE28</th>\n",
       "      <th>12426794 </th>\n",
       "      <td>                ZAN</td>\n",
       "      <td> C</td>\n",
       "      <td> T</td>\n",
       "      <td> PASS</td>\n",
       "      <td>NaN</td>\n",
       "      <td> 0.401840</td>\n",
       "      <td> T</td>\n",
       "      <td>               splice_donor_variant&amp;intron_variant</td>\n",
       "      <td> HIGH</td>\n",
       "      <td>       WARNING_TRANSCRIPT_NO_START_CODON</td>\n",
       "      <td> 0.03125</td>\n",
       "      <td> 0.93750</td>\n",
       "      <td> 0.833333</td>\n",
       "      <td> 0.862745</td>\n",
       "      <td> 0.000000</td>\n",
       "      <td> 0.000000</td>\n",
       "      <td> 0.0</td>\n",
       "      <td> 0.090909</td>\n",
       "      <td> 0.460704</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>CAE1</th>\n",
       "      <th>65928019 </th>\n",
       "      <td>                7SK</td>\n",
       "      <td> G</td>\n",
       "      <td> A</td>\n",
       "      <td> PASS</td>\n",
       "      <td>NaN</td>\n",
       "      <td> 0.546012</td>\n",
       "      <td> A</td>\n",
       "      <td>                                       stop_gained</td>\n",
       "      <td> HIGH</td>\n",
       "      <td> WARNING_TRANSCRIPT_MULTIPLE_STOP_CODONS</td>\n",
       "      <td> 0.18750</td>\n",
       "      <td> 1.00000</td>\n",
       "      <td> 1.000000</td>\n",
       "      <td> 1.000000</td>\n",
       "      <td> 0.041667</td>\n",
       "      <td> 0.029412</td>\n",
       "      <td> 0.0</td>\n",
       "      <td> 1.000000</td>\n",
       "      <td> 0.459547</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>CAE6</th>\n",
       "      <th>38843413 </th>\n",
       "      <td>           PPP1R13L</td>\n",
       "      <td> C</td>\n",
       "      <td> G</td>\n",
       "      <td> PASS</td>\n",
       "      <td>NaN</td>\n",
       "      <td> 0.481595</td>\n",
       "      <td> G</td>\n",
       "      <td>                                         stop_lost</td>\n",
       "      <td> HIGH</td>\n",
       "      <td>                                     NaN</td>\n",
       "      <td> 1.00000</td>\n",
       "      <td> 0.78125</td>\n",
       "      <td> 0.333333</td>\n",
       "      <td> 0.941176</td>\n",
       "      <td> 0.000000</td>\n",
       "      <td> 0.000000</td>\n",
       "      <td> 0.0</td>\n",
       "      <td> 0.000000</td>\n",
       "      <td> 0.458475</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>CAE7</th>\n",
       "      <th>129628649</th>\n",
       "      <td>                 U6</td>\n",
       "      <td> G</td>\n",
       "      <td> T</td>\n",
       "      <td> PASS</td>\n",
       "      <td>NaN</td>\n",
       "      <td> 0.404908</td>\n",
       "      <td> T</td>\n",
       "      <td>                                       stop_gained</td>\n",
       "      <td> HIGH</td>\n",
       "      <td>       WARNING_TRANSCRIPT_NO_START_CODON</td>\n",
       "      <td> 0.00000</td>\n",
       "      <td> 0.87500</td>\n",
       "      <td> 0.583333</td>\n",
       "      <td> 0.950980</td>\n",
       "      <td> 0.000000</td>\n",
       "      <td> 0.000000</td>\n",
       "      <td> 0.0</td>\n",
       "      <td> 0.000000</td>\n",
       "      <td> 0.456684</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>CAE12</th>\n",
       "      <th>60988799 </th>\n",
       "      <td> ENSCSAG00000018483</td>\n",
       "      <td> C</td>\n",
       "      <td> T</td>\n",
       "      <td> PASS</td>\n",
       "      <td>NaN</td>\n",
       "      <td> 0.481595</td>\n",
       "      <td> T</td>\n",
       "      <td>                                       stop_gained</td>\n",
       "      <td> HIGH</td>\n",
       "      <td>                                     NaN</td>\n",
       "      <td> 0.03125</td>\n",
       "      <td> 0.96875</td>\n",
       "      <td> 1.000000</td>\n",
       "      <td> 0.882353</td>\n",
       "      <td> 0.062500</td>\n",
       "      <td> 0.029412</td>\n",
       "      <td> 0.0</td>\n",
       "      <td> 0.818182</td>\n",
       "      <td> 0.454149</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>CAE2</th>\n",
       "      <th>562475   </th>\n",
       "      <td>              RTEL1</td>\n",
       "      <td> C</td>\n",
       "      <td> T</td>\n",
       "      <td> PASS</td>\n",
       "      <td>NaN</td>\n",
       "      <td> 0.174847</td>\n",
       "      <td> T</td>\n",
       "      <td>                                       stop_gained</td>\n",
       "      <td> HIGH</td>\n",
       "      <td>        WARNING_TRANSCRIPT_NO_STOP_CODON</td>\n",
       "      <td> 1.00000</td>\n",
       "      <td> 0.06250</td>\n",
       "      <td> 0.250000</td>\n",
       "      <td> 0.009804</td>\n",
       "      <td> 0.000000</td>\n",
       "      <td> 0.000000</td>\n",
       "      <td> 0.0</td>\n",
       "      <td> 0.863636</td>\n",
       "      <td> 0.450773</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>CAE20</th>\n",
       "      <th>78966415 </th>\n",
       "      <td>             LRRC42</td>\n",
       "      <td> C</td>\n",
       "      <td> T</td>\n",
       "      <td> PASS</td>\n",
       "      <td>NaN</td>\n",
       "      <td> 0.500000</td>\n",
       "      <td> T</td>\n",
       "      <td>                                       stop_gained</td>\n",
       "      <td> HIGH</td>\n",
       "      <td>                                     NaN</td>\n",
       "      <td> 0.12500</td>\n",
       "      <td> 1.00000</td>\n",
       "      <td> 0.833333</td>\n",
       "      <td> 1.000000</td>\n",
       "      <td> 0.000000</td>\n",
       "      <td> 0.000000</td>\n",
       "      <td> 0.0</td>\n",
       "      <td> 0.681818</td>\n",
       "      <td> 0.439658</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>CAE25</th>\n",
       "      <th>77331379 </th>\n",
       "      <td>               FMN2</td>\n",
       "      <td> G</td>\n",
       "      <td> A</td>\n",
       "      <td> PASS</td>\n",
       "      <td>NaN</td>\n",
       "      <td> 0.411043</td>\n",
       "      <td> A</td>\n",
       "      <td>            splice_acceptor_variant&amp;intron_variant</td>\n",
       "      <td> HIGH</td>\n",
       "      <td>        WARNING_TRANSCRIPT_NO_STOP_CODON</td>\n",
       "      <td> 0.03125</td>\n",
       "      <td> 0.90625</td>\n",
       "      <td> 1.000000</td>\n",
       "      <td> 0.754902</td>\n",
       "      <td> 0.000000</td>\n",
       "      <td> 0.000000</td>\n",
       "      <td> 0.0</td>\n",
       "      <td> 0.681818</td>\n",
       "      <td> 0.438054</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>CAE12</th>\n",
       "      <th>96863638 </th>\n",
       "      <td>           C9orf153</td>\n",
       "      <td> T</td>\n",
       "      <td> A</td>\n",
       "      <td> PASS</td>\n",
       "      <td>NaN</td>\n",
       "      <td> 0.015337</td>\n",
       "      <td> A</td>\n",
       "      <td>                                       stop_gained</td>\n",
       "      <td> HIGH</td>\n",
       "      <td>                                     NaN</td>\n",
       "      <td> 0.00000</td>\n",
       "      <td> 0.00000</td>\n",
       "      <td> 0.000000</td>\n",
       "      <td> 0.000000</td>\n",
       "      <td> 0.000000</td>\n",
       "      <td> 0.000000</td>\n",
       "      <td> 0.5</td>\n",
       "      <td> 0.000000</td>\n",
       "      <td> 0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>CAE9</th>\n",
       "      <th>107123094</th>\n",
       "      <td>               VWA2</td>\n",
       "      <td> G</td>\n",
       "      <td> A</td>\n",
       "      <td> PASS</td>\n",
       "      <td>NaN</td>\n",
       "      <td> 0.006135</td>\n",
       "      <td> A</td>\n",
       "      <td>               splice_donor_variant&amp;intron_variant</td>\n",
       "      <td> HIGH</td>\n",
       "      <td>        WARNING_TRANSCRIPT_NO_STOP_CODON</td>\n",
       "      <td> 0.00000</td>\n",
       "      <td> 0.00000</td>\n",
       "      <td> 0.000000</td>\n",
       "      <td> 0.000000</td>\n",
       "      <td> 0.000000</td>\n",
       "      <td> 0.029412</td>\n",
       "      <td> 0.0</td>\n",
       "      <td> 0.000000</td>\n",
       "      <td> 0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>CAE10</th>\n",
       "      <th>51550334 </th>\n",
       "      <td> ENSCSAG00000015438</td>\n",
       "      <td> G</td>\n",
       "      <td> T</td>\n",
       "      <td> PASS</td>\n",
       "      <td>NaN</td>\n",
       "      <td> 0.012270</td>\n",
       "      <td> T</td>\n",
       "      <td>                                       stop_gained</td>\n",
       "      <td> HIGH</td>\n",
       "      <td>                                     NaN</td>\n",
       "      <td> 0.00000</td>\n",
       "      <td> 0.00000</td>\n",
       "      <td> 0.000000</td>\n",
       "      <td> 0.000000</td>\n",
       "      <td> 0.000000</td>\n",
       "      <td> 0.058824</td>\n",
       "      <td> 0.0</td>\n",
       "      <td> 0.000000</td>\n",
       "      <td> 0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>CAE6</th>\n",
       "      <th>35435749 </th>\n",
       "      <td> ENSCSAG00000003601</td>\n",
       "      <td> G</td>\n",
       "      <td> A</td>\n",
       "      <td> PASS</td>\n",
       "      <td>NaN</td>\n",
       "      <td> 0.006135</td>\n",
       "      <td> A</td>\n",
       "      <td>                                       stop_gained</td>\n",
       "      <td> HIGH</td>\n",
       "      <td>                                     NaN</td>\n",
       "      <td> 0.00000</td>\n",
       "      <td> 0.00000</td>\n",
       "      <td> 0.000000</td>\n",
       "      <td> 0.000000</td>\n",
       "      <td> 0.000000</td>\n",
       "      <td> 0.000000</td>\n",
       "      <td> 0.2</td>\n",
       "      <td> 0.000000</td>\n",
       "      <td> 0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>CAE22</th>\n",
       "      <th>36444200 </th>\n",
       "      <td>               CHL1</td>\n",
       "      <td> G</td>\n",
       "      <td> T</td>\n",
       "      <td> PASS</td>\n",
       "      <td>NaN</td>\n",
       "      <td> 0.003067</td>\n",
       "      <td> T</td>\n",
       "      <td>                 stop_gained&amp;splice_region_variant</td>\n",
       "      <td> HIGH</td>\n",
       "      <td>                                     NaN</td>\n",
       "      <td> 0.00000</td>\n",
       "      <td> 0.00000</td>\n",
       "      <td> 0.000000</td>\n",
       "      <td> 0.000000</td>\n",
       "      <td> 0.000000</td>\n",
       "      <td> 0.000000</td>\n",
       "      <td> 0.1</td>\n",
       "      <td> 0.000000</td>\n",
       "      <td> 0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>CAE28</th>\n",
       "      <th>3443945  </th>\n",
       "      <td>            5S_rRNA</td>\n",
       "      <td> T</td>\n",
       "      <td> G</td>\n",
       "      <td> PASS</td>\n",
       "      <td>NaN</td>\n",
       "      <td> 0.024540</td>\n",
       "      <td> G</td>\n",
       "      <td>                                         stop_lost</td>\n",
       "      <td> HIGH</td>\n",
       "      <td>           WARNING_TRANSCRIPT_INCOMPLETE</td>\n",
       "      <td> 0.00000</td>\n",
       "      <td> 0.00000</td>\n",
       "      <td> 0.000000</td>\n",
       "      <td> 0.000000</td>\n",
       "      <td> 0.000000</td>\n",
       "      <td> 0.117647</td>\n",
       "      <td> 0.0</td>\n",
       "      <td> 0.000000</td>\n",
       "      <td> 0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>CAE5</th>\n",
       "      <th>58398005 </th>\n",
       "      <td>                 U6</td>\n",
       "      <td> C</td>\n",
       "      <td> T</td>\n",
       "      <td> PASS</td>\n",
       "      <td>NaN</td>\n",
       "      <td> 0.021472</td>\n",
       "      <td> T</td>\n",
       "      <td>                                       stop_gained</td>\n",
       "      <td> HIGH</td>\n",
       "      <td> WARNING_TRANSCRIPT_MULTIPLE_STOP_CODONS</td>\n",
       "      <td> 0.00000</td>\n",
       "      <td> 0.00000</td>\n",
       "      <td> 0.000000</td>\n",
       "      <td> 0.000000</td>\n",
       "      <td> 0.000000</td>\n",
       "      <td> 0.088235</td>\n",
       "      <td> 0.1</td>\n",
       "      <td> 0.000000</td>\n",
       "      <td> 0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>CAE26</th>\n",
       "      <th>21280477 </th>\n",
       "      <td> ENSCSAG00000012210</td>\n",
       "      <td> G</td>\n",
       "      <td> A</td>\n",
       "      <td> PASS</td>\n",
       "      <td>NaN</td>\n",
       "      <td> 0.049080</td>\n",
       "      <td> A</td>\n",
       "      <td>                                       stop_gained</td>\n",
       "      <td> HIGH</td>\n",
       "      <td>                                     NaN</td>\n",
       "      <td> 0.00000</td>\n",
       "      <td> 0.00000</td>\n",
       "      <td> 0.000000</td>\n",
       "      <td> 0.000000</td>\n",
       "      <td> 0.000000</td>\n",
       "      <td> 0.235294</td>\n",
       "      <td> 0.0</td>\n",
       "      <td> 0.000000</td>\n",
       "      <td> 0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>CAE3</th>\n",
       "      <th>27890028 </th>\n",
       "      <td>             SPRYD7</td>\n",
       "      <td> T</td>\n",
       "      <td> C</td>\n",
       "      <td> PASS</td>\n",
       "      <td>NaN</td>\n",
       "      <td> 0.935583</td>\n",
       "      <td> C</td>\n",
       "      <td>                                        start_lost</td>\n",
       "      <td> HIGH</td>\n",
       "      <td>                                     NaN</td>\n",
       "      <td> 1.00000</td>\n",
       "      <td> 1.00000</td>\n",
       "      <td> 1.000000</td>\n",
       "      <td> 1.000000</td>\n",
       "      <td> 1.000000</td>\n",
       "      <td> 0.691176</td>\n",
       "      <td> 1.0</td>\n",
       "      <td> 1.000000</td>\n",
       "      <td> 0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>CAE6</th>\n",
       "      <th>34768192 </th>\n",
       "      <td>                 U6</td>\n",
       "      <td> T</td>\n",
       "      <td> C</td>\n",
       "      <td> PASS</td>\n",
       "      <td>NaN</td>\n",
       "      <td> 0.030675</td>\n",
       "      <td> C</td>\n",
       "      <td>                                         stop_lost</td>\n",
       "      <td> HIGH</td>\n",
       "      <td>           WARNING_TRANSCRIPT_INCOMPLETE</td>\n",
       "      <td> 0.00000</td>\n",
       "      <td> 0.00000</td>\n",
       "      <td> 0.000000</td>\n",
       "      <td> 0.000000</td>\n",
       "      <td> 0.000000</td>\n",
       "      <td> 0.014706</td>\n",
       "      <td> 0.9</td>\n",
       "      <td> 0.000000</td>\n",
       "      <td> 0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>CAE19</th>\n",
       "      <th>23607566 </th>\n",
       "      <td>              Y_RNA</td>\n",
       "      <td> G</td>\n",
       "      <td> A</td>\n",
       "      <td> PASS</td>\n",
       "      <td>NaN</td>\n",
       "      <td> 0.049080</td>\n",
       "      <td> A</td>\n",
       "      <td>                                       stop_gained</td>\n",
       "      <td> HIGH</td>\n",
       "      <td> WARNING_TRANSCRIPT_MULTIPLE_STOP_CODONS</td>\n",
       "      <td> 0.00000</td>\n",
       "      <td> 0.00000</td>\n",
       "      <td> 0.000000</td>\n",
       "      <td> 0.000000</td>\n",
       "      <td> 0.000000</td>\n",
       "      <td> 0.161765</td>\n",
       "      <td> 0.5</td>\n",
       "      <td> 0.000000</td>\n",
       "      <td> 0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>CAE14</th>\n",
       "      <th>89103174 </th>\n",
       "      <td> ENSCSAG00000007999</td>\n",
       "      <td> T</td>\n",
       "      <td> C</td>\n",
       "      <td> PASS</td>\n",
       "      <td>NaN</td>\n",
       "      <td> 0.003067</td>\n",
       "      <td> C</td>\n",
       "      <td>               splice_donor_variant&amp;intron_variant</td>\n",
       "      <td> HIGH</td>\n",
       "      <td>                                     NaN</td>\n",
       "      <td> 0.00000</td>\n",
       "      <td> 0.00000</td>\n",
       "      <td> 0.000000</td>\n",
       "      <td> 0.000000</td>\n",
       "      <td> 0.000000</td>\n",
       "      <td> 0.014706</td>\n",
       "      <td> 0.0</td>\n",
       "      <td> 0.000000</td>\n",
       "      <td> 0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>CAE3</th>\n",
       "      <th>24325369 </th>\n",
       "      <td>             LRRC63</td>\n",
       "      <td> G</td>\n",
       "      <td> A</td>\n",
       "      <td> PASS</td>\n",
       "      <td>NaN</td>\n",
       "      <td> 0.018405</td>\n",
       "      <td> A</td>\n",
       "      <td>            splice_acceptor_variant&amp;intron_variant</td>\n",
       "      <td> HIGH</td>\n",
       "      <td>        WARNING_TRANSCRIPT_NO_STOP_CODON</td>\n",
       "      <td> 0.00000</td>\n",
       "      <td> 0.00000</td>\n",
       "      <td> 0.000000</td>\n",
       "      <td> 0.000000</td>\n",
       "      <td> 0.000000</td>\n",
       "      <td> 0.073529</td>\n",
       "      <td> 0.1</td>\n",
       "      <td> 0.000000</td>\n",
       "      <td> 0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>CAE23</th>\n",
       "      <th>47818308 </th>\n",
       "      <td> ENSCSAG00000012331</td>\n",
       "      <td> C</td>\n",
       "      <td> T</td>\n",
       "      <td> PASS</td>\n",
       "      <td>NaN</td>\n",
       "      <td> 0.021472</td>\n",
       "      <td> T</td>\n",
       "      <td>                                       stop_gained</td>\n",
       "      <td> HIGH</td>\n",
       "      <td>       WARNING_TRANSCRIPT_NO_START_CODON</td>\n",
       "      <td> 0.00000</td>\n",
       "      <td> 0.00000</td>\n",
       "      <td> 0.000000</td>\n",
       "      <td> 0.000000</td>\n",
       "      <td> 0.000000</td>\n",
       "      <td> 0.102941</td>\n",
       "      <td> 0.0</td>\n",
       "      <td> 0.000000</td>\n",
       "      <td> 0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>CAE11</th>\n",
       "      <th>61962524 </th>\n",
       "      <td>                7SK</td>\n",
       "      <td> C</td>\n",
       "      <td> T</td>\n",
       "      <td> PASS</td>\n",
       "      <td>NaN</td>\n",
       "      <td> 0.006135</td>\n",
       "      <td> T</td>\n",
       "      <td>                                       stop_gained</td>\n",
       "      <td> HIGH</td>\n",
       "      <td> WARNING_TRANSCRIPT_MULTIPLE_STOP_CODONS</td>\n",
       "      <td> 0.00000</td>\n",
       "      <td> 0.00000</td>\n",
       "      <td> 0.000000</td>\n",
       "      <td> 0.000000</td>\n",
       "      <td> 0.000000</td>\n",
       "      <td> 0.014706</td>\n",
       "      <td> 0.1</td>\n",
       "      <td> 0.000000</td>\n",
       "      <td> 0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>CAE24</th>\n",
       "      <th>22322993 </th>\n",
       "      <td>              FANCM</td>\n",
       "      <td> G</td>\n",
       "      <td> A</td>\n",
       "      <td> PASS</td>\n",
       "      <td>NaN</td>\n",
       "      <td> 0.003067</td>\n",
       "      <td> A</td>\n",
       "      <td>            splice_acceptor_variant&amp;intron_variant</td>\n",
       "      <td> HIGH</td>\n",
       "      <td>                                     NaN</td>\n",
       "      <td> 0.00000</td>\n",
       "      <td> 0.00000</td>\n",
       "      <td> 0.000000</td>\n",
       "      <td> 0.000000</td>\n",
       "      <td> 0.000000</td>\n",
       "      <td> 0.000000</td>\n",
       "      <td> 0.1</td>\n",
       "      <td> 0.000000</td>\n",
       "      <td> 0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>CAE13</th>\n",
       "      <th>11995990 </th>\n",
       "      <td>              RARS2</td>\n",
       "      <td> A</td>\n",
       "      <td> C</td>\n",
       "      <td> PASS</td>\n",
       "      <td>NaN</td>\n",
       "      <td> 0.015337</td>\n",
       "      <td> C</td>\n",
       "      <td>               splice_donor_variant&amp;intron_variant</td>\n",
       "      <td> HIGH</td>\n",
       "      <td>                                     NaN</td>\n",
       "      <td> 0.00000</td>\n",
       "      <td> 0.00000</td>\n",
       "      <td> 0.000000</td>\n",
       "      <td> 0.000000</td>\n",
       "      <td> 0.000000</td>\n",
       "      <td> 0.073529</td>\n",
       "      <td> 0.0</td>\n",
       "      <td> 0.000000</td>\n",
       "      <td> 0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th rowspan=\"2\" valign=\"top\">CAE20</th>\n",
       "      <th>130570772</th>\n",
       "      <td> ENSCSAG00000000661</td>\n",
       "      <td> C</td>\n",
       "      <td> T</td>\n",
       "      <td> PASS</td>\n",
       "      <td>NaN</td>\n",
       "      <td> 0.015337</td>\n",
       "      <td> T</td>\n",
       "      <td>                                       stop_gained</td>\n",
       "      <td> HIGH</td>\n",
       "      <td>                                     NaN</td>\n",
       "      <td> 0.00000</td>\n",
       "      <td> 0.00000</td>\n",
       "      <td> 0.000000</td>\n",
       "      <td> 0.000000</td>\n",
       "      <td> 0.000000</td>\n",
       "      <td> 0.073529</td>\n",
       "      <td> 0.0</td>\n",
       "      <td> 0.000000</td>\n",
       "      <td> 0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>80034861 </th>\n",
       "      <td>             ECHDC2</td>\n",
       "      <td> C</td>\n",
       "      <td> T</td>\n",
       "      <td> PASS</td>\n",
       "      <td>NaN</td>\n",
       "      <td> 0.012270</td>\n",
       "      <td> T</td>\n",
       "      <td>                                       stop_gained</td>\n",
       "      <td> HIGH</td>\n",
       "      <td>                                     NaN</td>\n",
       "      <td> 0.00000</td>\n",
       "      <td> 0.00000</td>\n",
       "      <td> 0.000000</td>\n",
       "      <td> 0.000000</td>\n",
       "      <td> 0.000000</td>\n",
       "      <td> 0.058824</td>\n",
       "      <td> 0.0</td>\n",
       "      <td> 0.000000</td>\n",
       "      <td> 0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>CAE22</th>\n",
       "      <th>59394111 </th>\n",
       "      <td>               POLQ</td>\n",
       "      <td> C</td>\n",
       "      <td> T</td>\n",
       "      <td> PASS</td>\n",
       "      <td>NaN</td>\n",
       "      <td> 0.042945</td>\n",
       "      <td> T</td>\n",
       "      <td>                                       stop_gained</td>\n",
       "      <td> HIGH</td>\n",
       "      <td>                                     NaN</td>\n",
       "      <td> 0.00000</td>\n",
       "      <td> 0.00000</td>\n",
       "      <td> 0.000000</td>\n",
       "      <td> 0.000000</td>\n",
       "      <td> 0.000000</td>\n",
       "      <td> 0.205882</td>\n",
       "      <td> 0.0</td>\n",
       "      <td> 0.000000</td>\n",
       "      <td> 0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>CAE24</th>\n",
       "      <th>8515944  </th>\n",
       "      <td> ENSCSAG00000015689</td>\n",
       "      <td> G</td>\n",
       "      <td> A</td>\n",
       "      <td> PASS</td>\n",
       "      <td>NaN</td>\n",
       "      <td> 0.030675</td>\n",
       "      <td> A</td>\n",
       "      <td>                                       stop_gained</td>\n",
       "      <td> HIGH</td>\n",
       "      <td>        WARNING_TRANSCRIPT_NO_STOP_CODON</td>\n",
       "      <td> 0.00000</td>\n",
       "      <td> 0.00000</td>\n",
       "      <td> 0.000000</td>\n",
       "      <td> 0.000000</td>\n",
       "      <td> 0.000000</td>\n",
       "      <td> 0.147059</td>\n",
       "      <td> 0.0</td>\n",
       "      <td> 0.000000</td>\n",
       "      <td> 0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>CAE1</th>\n",
       "      <th>21503929 </th>\n",
       "      <td>             ALKBH3</td>\n",
       "      <td> C</td>\n",
       "      <td> T</td>\n",
       "      <td> PASS</td>\n",
       "      <td>NaN</td>\n",
       "      <td> 0.033742</td>\n",
       "      <td> T</td>\n",
       "      <td>                                        start_lost</td>\n",
       "      <td> HIGH</td>\n",
       "      <td>                                     NaN</td>\n",
       "      <td> 0.00000</td>\n",
       "      <td> 0.00000</td>\n",
       "      <td> 0.000000</td>\n",
       "      <td> 0.000000</td>\n",
       "      <td> 0.000000</td>\n",
       "      <td> 0.161765</td>\n",
       "      <td> 0.0</td>\n",
       "      <td> 0.000000</td>\n",
       "      <td> 0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>CAE6</th>\n",
       "      <th>28041036 </th>\n",
       "      <td>              RHPN2</td>\n",
       "      <td> A</td>\n",
       "      <td> G</td>\n",
       "      <td> PASS</td>\n",
       "      <td>NaN</td>\n",
       "      <td> 0.015337</td>\n",
       "      <td> G</td>\n",
       "      <td>               splice_donor_variant&amp;intron_variant</td>\n",
       "      <td> HIGH</td>\n",
       "      <td>                                     NaN</td>\n",
       "      <td> 0.00000</td>\n",
       "      <td> 0.00000</td>\n",
       "      <td> 0.000000</td>\n",
       "      <td> 0.000000</td>\n",
       "      <td> 0.000000</td>\n",
       "      <td> 0.000000</td>\n",
       "      <td> 0.5</td>\n",
       "      <td> 0.000000</td>\n",
       "      <td> 0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>CAE8</th>\n",
       "      <th>35798069 </th>\n",
       "      <td>              PROSC</td>\n",
       "      <td> G</td>\n",
       "      <td> A</td>\n",
       "      <td> PASS</td>\n",
       "      <td>NaN</td>\n",
       "      <td> 0.012270</td>\n",
       "      <td> A</td>\n",
       "      <td>                                        start_lost</td>\n",
       "      <td> HIGH</td>\n",
       "      <td>                                     NaN</td>\n",
       "      <td> 0.00000</td>\n",
       "      <td> 0.00000</td>\n",
       "      <td> 0.000000</td>\n",
       "      <td> 0.000000</td>\n",
       "      <td> 0.000000</td>\n",
       "      <td> 0.058824</td>\n",
       "      <td> 0.0</td>\n",
       "      <td> 0.000000</td>\n",
       "      <td> 0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>CAE1</th>\n",
       "      <th>116200400</th>\n",
       "      <td>             CCDC15</td>\n",
       "      <td> G</td>\n",
       "      <td> T</td>\n",
       "      <td> PASS</td>\n",
       "      <td>NaN</td>\n",
       "      <td> 0.006135</td>\n",
       "      <td> T</td>\n",
       "      <td>                                       stop_gained</td>\n",
       "      <td> HIGH</td>\n",
       "      <td>       WARNING_TRANSCRIPT_NO_START_CODON</td>\n",
       "      <td> 0.00000</td>\n",
       "      <td> 0.00000</td>\n",
       "      <td> 0.000000</td>\n",
       "      <td> 0.000000</td>\n",
       "      <td> 0.000000</td>\n",
       "      <td> 0.000000</td>\n",
       "      <td> 0.2</td>\n",
       "      <td> 0.000000</td>\n",
       "      <td> 0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>CAE8</th>\n",
       "      <th>31001692 </th>\n",
       "      <td>                 U6</td>\n",
       "      <td> T</td>\n",
       "      <td> A</td>\n",
       "      <td> PASS</td>\n",
       "      <td>NaN</td>\n",
       "      <td> 0.963190</td>\n",
       "      <td> A</td>\n",
       "      <td>                                         stop_lost</td>\n",
       "      <td> HIGH</td>\n",
       "      <td> WARNING_TRANSCRIPT_MULTIPLE_STOP_CODONS</td>\n",
       "      <td> 1.00000</td>\n",
       "      <td> 1.00000</td>\n",
       "      <td> 1.000000</td>\n",
       "      <td> 1.000000</td>\n",
       "      <td> 1.000000</td>\n",
       "      <td> 0.867647</td>\n",
       "      <td> 0.7</td>\n",
       "      <td> 1.000000</td>\n",
       "      <td> 0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>CAE10</th>\n",
       "      <th>57303137 </th>\n",
       "      <td> ENSCSAG00000022571</td>\n",
       "      <td> C</td>\n",
       "      <td> G</td>\n",
       "      <td> PASS</td>\n",
       "      <td>NaN</td>\n",
       "      <td> 1.000000</td>\n",
       "      <td> G</td>\n",
       "      <td>                                         stop_lost</td>\n",
       "      <td> HIGH</td>\n",
       "      <td>           WARNING_TRANSCRIPT_INCOMPLETE</td>\n",
       "      <td> 1.00000</td>\n",
       "      <td> 1.00000</td>\n",
       "      <td> 1.000000</td>\n",
       "      <td> 1.000000</td>\n",
       "      <td> 1.000000</td>\n",
       "      <td> 1.000000</td>\n",
       "      <td> 1.0</td>\n",
       "      <td> 1.000000</td>\n",
       "      <td> 0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>CAE16</th>\n",
       "      <th>59272885 </th>\n",
       "      <td>            EFCAB13</td>\n",
       "      <td> T</td>\n",
       "      <td> C</td>\n",
       "      <td> PASS</td>\n",
       "      <td>NaN</td>\n",
       "      <td> 0.009202</td>\n",
       "      <td> C</td>\n",
       "      <td>            splice_acceptor_variant&amp;intron_variant</td>\n",
       "      <td> HIGH</td>\n",
       "      <td>       WARNING_TRANSCRIPT_NO_START_CODON</td>\n",
       "      <td> 0.00000</td>\n",
       "      <td> 0.00000</td>\n",
       "      <td> 0.000000</td>\n",
       "      <td> 0.000000</td>\n",
       "      <td> 0.000000</td>\n",
       "      <td> 0.000000</td>\n",
       "      <td> 0.3</td>\n",
       "      <td> 0.000000</td>\n",
       "      <td> 0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>CAE12</th>\n",
       "      <th>22402461 </th>\n",
       "      <td>                7SK</td>\n",
       "      <td> T</td>\n",
       "      <td> A</td>\n",
       "      <td> PASS</td>\n",
       "      <td>NaN</td>\n",
       "      <td> 0.923313</td>\n",
       "      <td> A</td>\n",
       "      <td>                                         stop_lost</td>\n",
       "      <td> HIGH</td>\n",
       "      <td> WARNING_TRANSCRIPT_MULTIPLE_STOP_CODONS</td>\n",
       "      <td> 1.00000</td>\n",
       "      <td> 1.00000</td>\n",
       "      <td> 1.000000</td>\n",
       "      <td> 1.000000</td>\n",
       "      <td> 1.000000</td>\n",
       "      <td> 0.691176</td>\n",
       "      <td> 0.6</td>\n",
       "      <td> 1.000000</td>\n",
       "      <td> 0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>CAE25</th>\n",
       "      <th>24784772 </th>\n",
       "      <td>            5S_rRNA</td>\n",
       "      <td> G</td>\n",
       "      <td> T</td>\n",
       "      <td> PASS</td>\n",
       "      <td>NaN</td>\n",
       "      <td> 0.012270</td>\n",
       "      <td> T</td>\n",
       "      <td>                                       stop_gained</td>\n",
       "      <td> HIGH</td>\n",
       "      <td>           WARNING_TRANSCRIPT_INCOMPLETE</td>\n",
       "      <td> 0.00000</td>\n",
       "      <td> 0.00000</td>\n",
       "      <td> 0.000000</td>\n",
       "      <td> 0.000000</td>\n",
       "      <td> 0.000000</td>\n",
       "      <td> 0.000000</td>\n",
       "      <td> 0.4</td>\n",
       "      <td> 0.000000</td>\n",
       "      <td> 0.000000</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>6410 rows  19 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                        gene_symbol REF ALT FILTER  AA        AF allele  \\\n",
       "CHROM POS                                                                 \n",
       "CAEX  47049265                AKAP4   C   A   PASS NaN  0.466258      A   \n",
       "CAE16 46180030   ENSCSAG00000023715   A   G   PASS NaN  0.546012      G   \n",
       "CAE11 75076181          Metazoa_SRP   A   T   PASS NaN  0.447853      T   \n",
       "CAE7  97439521   ENSCSAG00000003348   G   A   PASS NaN  0.524540      A   \n",
       "CAE22 95483322                   U2   A   G   PASS NaN  0.438650      G   \n",
       "CAE11 101983690  ENSCSAG00000018940   G   C   PASS NaN  0.503067      C   \n",
       "CAE1  9938424                 Y_RNA   G   A   PASS NaN  0.500000      A   \n",
       "CAE18 17036302                   U6   A   T   PASS NaN  0.533742      T   \n",
       "CAE11 48208899                NR4A1   T   C   PASS NaN  0.484663      C   \n",
       "CAE1  255393            Metazoa_SRP   G   A   PASS NaN  0.444785      A   \n",
       "CAE10 76025378             TMEM194B   C   T   PASS NaN  0.463190      T   \n",
       "CAE14 12377693              CCDC138   G   A   PASS NaN  0.447853      A   \n",
       "CAE2  25313286              SNORA71   A   G   PASS NaN  0.546012      G   \n",
       "CAE20 42794243   ENSCSAG00000025756   A   C   PASS NaN  0.515337      C   \n",
       "CAE2  33907866              SLC4A11   A   G   PASS NaN  0.500000      G   \n",
       "CAE20 14359047               HSD3B2   G   A   PASS NaN  0.128834      A   \n",
       "CAE2  7344987                    U6   T   G   PASS NaN  0.527607      G   \n",
       "CAE4  76207128          Metazoa_SRP   A   G   PASS NaN  0.398773      G   \n",
       "CAE13 76998487              NDUFAF4   G   T   PASS NaN  0.404908      T   \n",
       "CAE6  48190642               ZNF865   C   T   PASS NaN  0.162577      T   \n",
       "CAE9  45370011                   U6   G   A   PASS NaN  0.180982      A   \n",
       "CAE13 369349                    7SK   A   G   PASS NaN  0.407975      G   \n",
       "CAE28 12426794                  ZAN   C   T   PASS NaN  0.401840      T   \n",
       "CAE1  65928019                  7SK   G   A   PASS NaN  0.546012      A   \n",
       "CAE6  38843413             PPP1R13L   C   G   PASS NaN  0.481595      G   \n",
       "CAE7  129628649                  U6   G   T   PASS NaN  0.404908      T   \n",
       "CAE12 60988799   ENSCSAG00000018483   C   T   PASS NaN  0.481595      T   \n",
       "CAE2  562475                  RTEL1   C   T   PASS NaN  0.174847      T   \n",
       "CAE20 78966415               LRRC42   C   T   PASS NaN  0.500000      T   \n",
       "CAE25 77331379                 FMN2   G   A   PASS NaN  0.411043      A   \n",
       "...                             ...  ..  ..    ...  ..       ...    ...   \n",
       "CAE12 96863638             C9orf153   T   A   PASS NaN  0.015337      A   \n",
       "CAE9  107123094                VWA2   G   A   PASS NaN  0.006135      A   \n",
       "CAE10 51550334   ENSCSAG00000015438   G   T   PASS NaN  0.012270      T   \n",
       "CAE6  35435749   ENSCSAG00000003601   G   A   PASS NaN  0.006135      A   \n",
       "CAE22 36444200                 CHL1   G   T   PASS NaN  0.003067      T   \n",
       "CAE28 3443945               5S_rRNA   T   G   PASS NaN  0.024540      G   \n",
       "CAE5  58398005                   U6   C   T   PASS NaN  0.021472      T   \n",
       "CAE26 21280477   ENSCSAG00000012210   G   A   PASS NaN  0.049080      A   \n",
       "CAE3  27890028               SPRYD7   T   C   PASS NaN  0.935583      C   \n",
       "CAE6  34768192                   U6   T   C   PASS NaN  0.030675      C   \n",
       "CAE19 23607566                Y_RNA   G   A   PASS NaN  0.049080      A   \n",
       "CAE14 89103174   ENSCSAG00000007999   T   C   PASS NaN  0.003067      C   \n",
       "CAE3  24325369               LRRC63   G   A   PASS NaN  0.018405      A   \n",
       "CAE23 47818308   ENSCSAG00000012331   C   T   PASS NaN  0.021472      T   \n",
       "CAE11 61962524                  7SK   C   T   PASS NaN  0.006135      T   \n",
       "CAE24 22322993                FANCM   G   A   PASS NaN  0.003067      A   \n",
       "CAE13 11995990                RARS2   A   C   PASS NaN  0.015337      C   \n",
       "CAE20 130570772  ENSCSAG00000000661   C   T   PASS NaN  0.015337      T   \n",
       "      80034861               ECHDC2   C   T   PASS NaN  0.012270      T   \n",
       "CAE22 59394111                 POLQ   C   T   PASS NaN  0.042945      T   \n",
       "CAE24 8515944    ENSCSAG00000015689   G   A   PASS NaN  0.030675      A   \n",
       "CAE1  21503929               ALKBH3   C   T   PASS NaN  0.033742      T   \n",
       "CAE6  28041036                RHPN2   A   G   PASS NaN  0.015337      G   \n",
       "CAE8  35798069                PROSC   G   A   PASS NaN  0.012270      A   \n",
       "CAE1  116200400              CCDC15   G   T   PASS NaN  0.006135      T   \n",
       "CAE8  31001692                   U6   T   A   PASS NaN  0.963190      A   \n",
       "CAE10 57303137   ENSCSAG00000022571   C   G   PASS NaN  1.000000      G   \n",
       "CAE16 59272885              EFCAB13   T   C   PASS NaN  0.009202      C   \n",
       "CAE12 22402461                  7SK   T   A   PASS NaN  0.923313      A   \n",
       "CAE25 24784772              5S_rRNA   G   T   PASS NaN  0.012270      T   \n",
       "\n",
       "                                                            effect impact  \\\n",
       "CHROM POS                                                                   \n",
       "CAEX  47049265                                         stop_gained   HIGH   \n",
       "CAE16 46180030                     stop_lost&splice_region_variant   HIGH   \n",
       "CAE11 75076181                                           stop_lost   HIGH   \n",
       "CAE7  97439521   splice_acceptor_variant&splice_donor_variant&i...   HIGH   \n",
       "CAE22 95483322                                           stop_lost   HIGH   \n",
       "CAE11 101983690                                        stop_gained   HIGH   \n",
       "CAE1  9938424                                          stop_gained   HIGH   \n",
       "CAE18 17036302                                         stop_gained   HIGH   \n",
       "CAE11 48208899                 splice_donor_variant&intron_variant   HIGH   \n",
       "CAE1  255393                                           stop_gained   HIGH   \n",
       "CAE10 76025378                                          start_lost   HIGH   \n",
       "CAE14 12377693              splice_acceptor_variant&intron_variant   HIGH   \n",
       "CAE2  25313286                                           stop_lost   HIGH   \n",
       "CAE20 42794243                                           stop_lost   HIGH   \n",
       "CAE2  33907866                 splice_donor_variant&intron_variant   HIGH   \n",
       "CAE20 14359047                                         stop_gained   HIGH   \n",
       "CAE2  7344987                                            stop_lost   HIGH   \n",
       "CAE4  76207128                                           stop_lost   HIGH   \n",
       "CAE13 76998487              splice_acceptor_variant&intron_variant   HIGH   \n",
       "CAE6  48190642                 splice_donor_variant&intron_variant   HIGH   \n",
       "CAE9  45370011                                         stop_gained   HIGH   \n",
       "CAE13 369349                                             stop_lost   HIGH   \n",
       "CAE28 12426794                 splice_donor_variant&intron_variant   HIGH   \n",
       "CAE1  65928019                                         stop_gained   HIGH   \n",
       "CAE6  38843413                                           stop_lost   HIGH   \n",
       "CAE7  129628649                                        stop_gained   HIGH   \n",
       "CAE12 60988799                                         stop_gained   HIGH   \n",
       "CAE2  562475                                           stop_gained   HIGH   \n",
       "CAE20 78966415                                         stop_gained   HIGH   \n",
       "CAE25 77331379              splice_acceptor_variant&intron_variant   HIGH   \n",
       "...                                                            ...    ...   \n",
       "CAE12 96863638                                         stop_gained   HIGH   \n",
       "CAE9  107123094                splice_donor_variant&intron_variant   HIGH   \n",
       "CAE10 51550334                                         stop_gained   HIGH   \n",
       "CAE6  35435749                                         stop_gained   HIGH   \n",
       "CAE22 36444200                   stop_gained&splice_region_variant   HIGH   \n",
       "CAE28 3443945                                            stop_lost   HIGH   \n",
       "CAE5  58398005                                         stop_gained   HIGH   \n",
       "CAE26 21280477                                         stop_gained   HIGH   \n",
       "CAE3  27890028                                          start_lost   HIGH   \n",
       "CAE6  34768192                                           stop_lost   HIGH   \n",
       "CAE19 23607566                                         stop_gained   HIGH   \n",
       "CAE14 89103174                 splice_donor_variant&intron_variant   HIGH   \n",
       "CAE3  24325369              splice_acceptor_variant&intron_variant   HIGH   \n",
       "CAE23 47818308                                         stop_gained   HIGH   \n",
       "CAE11 61962524                                         stop_gained   HIGH   \n",
       "CAE24 22322993              splice_acceptor_variant&intron_variant   HIGH   \n",
       "CAE13 11995990                 splice_donor_variant&intron_variant   HIGH   \n",
       "CAE20 130570772                                        stop_gained   HIGH   \n",
       "      80034861                                         stop_gained   HIGH   \n",
       "CAE22 59394111                                         stop_gained   HIGH   \n",
       "CAE24 8515944                                          stop_gained   HIGH   \n",
       "CAE1  21503929                                          start_lost   HIGH   \n",
       "CAE6  28041036                 splice_donor_variant&intron_variant   HIGH   \n",
       "CAE8  35798069                                          start_lost   HIGH   \n",
       "CAE1  116200400                                        stop_gained   HIGH   \n",
       "CAE8  31001692                                           stop_lost   HIGH   \n",
       "CAE10 57303137                                           stop_lost   HIGH   \n",
       "CAE16 59272885              splice_acceptor_variant&intron_variant   HIGH   \n",
       "CAE12 22402461                                           stop_lost   HIGH   \n",
       "CAE25 24784772                                         stop_gained   HIGH   \n",
       "\n",
       "                                                  errors      aet      cyn  \\\n",
       "CHROM POS                                                                    \n",
       "CAEX  47049265                                       NaN  0.18750  1.00000   \n",
       "CAE16 46180030         WARNING_TRANSCRIPT_NO_START_CODON  1.00000  1.00000   \n",
       "CAE11 75076181   WARNING_TRANSCRIPT_MULTIPLE_STOP_CODONS  0.06250  1.00000   \n",
       "CAE7  97439521         WARNING_TRANSCRIPT_NO_START_CODON  0.00000  1.00000   \n",
       "CAE22 95483322   WARNING_TRANSCRIPT_MULTIPLE_STOP_CODONS  0.00000  0.93750   \n",
       "CAE11 101983690         WARNING_TRANSCRIPT_NO_STOP_CODON  0.00000  1.00000   \n",
       "CAE1  9938424          WARNING_TRANSCRIPT_NO_START_CODON  0.00000  0.93750   \n",
       "CAE18 17036302   WARNING_TRANSCRIPT_MULTIPLE_STOP_CODONS  0.09375  1.00000   \n",
       "CAE11 48208899                                       NaN  0.00000  1.00000   \n",
       "CAE1  255393     WARNING_TRANSCRIPT_MULTIPLE_STOP_CODONS  0.06250  1.00000   \n",
       "CAE10 76025378                                       NaN  0.00000  0.96875   \n",
       "CAE14 12377693                                       NaN  0.00000  0.96875   \n",
       "CAE2  25313286   WARNING_TRANSCRIPT_MULTIPLE_STOP_CODONS  0.00000  1.00000   \n",
       "CAE20 42794243             WARNING_TRANSCRIPT_INCOMPLETE  0.00000  1.00000   \n",
       "CAE2  33907866                                       NaN  0.06250  0.71875   \n",
       "CAE20 14359047                                       NaN  1.00000  0.00000   \n",
       "CAE2  7344987    WARNING_TRANSCRIPT_MULTIPLE_STOP_CODONS  0.15625  1.00000   \n",
       "CAE4  76207128   WARNING_TRANSCRIPT_MULTIPLE_STOP_CODONS  0.00000  0.78125   \n",
       "CAE13 76998487                                       NaN  0.00000  0.96875   \n",
       "CAE6  48190642         WARNING_TRANSCRIPT_NO_START_CODON  1.00000  0.00000   \n",
       "CAE9  45370011   WARNING_TRANSCRIPT_MULTIPLE_STOP_CODONS  0.06250  0.15625   \n",
       "CAE13 369349     WARNING_TRANSCRIPT_MULTIPLE_STOP_CODONS  0.00000  0.96875   \n",
       "CAE28 12426794         WARNING_TRANSCRIPT_NO_START_CODON  0.03125  0.93750   \n",
       "CAE1  65928019   WARNING_TRANSCRIPT_MULTIPLE_STOP_CODONS  0.18750  1.00000   \n",
       "CAE6  38843413                                       NaN  1.00000  0.78125   \n",
       "CAE7  129628649        WARNING_TRANSCRIPT_NO_START_CODON  0.00000  0.87500   \n",
       "CAE12 60988799                                       NaN  0.03125  0.96875   \n",
       "CAE2  562475            WARNING_TRANSCRIPT_NO_STOP_CODON  1.00000  0.06250   \n",
       "CAE20 78966415                                       NaN  0.12500  1.00000   \n",
       "CAE25 77331379          WARNING_TRANSCRIPT_NO_STOP_CODON  0.03125  0.90625   \n",
       "...                                                  ...      ...      ...   \n",
       "CAE12 96863638                                       NaN  0.00000  0.00000   \n",
       "CAE9  107123094         WARNING_TRANSCRIPT_NO_STOP_CODON  0.00000  0.00000   \n",
       "CAE10 51550334                                       NaN  0.00000  0.00000   \n",
       "CAE6  35435749                                       NaN  0.00000  0.00000   \n",
       "CAE22 36444200                                       NaN  0.00000  0.00000   \n",
       "CAE28 3443945              WARNING_TRANSCRIPT_INCOMPLETE  0.00000  0.00000   \n",
       "CAE5  58398005   WARNING_TRANSCRIPT_MULTIPLE_STOP_CODONS  0.00000  0.00000   \n",
       "CAE26 21280477                                       NaN  0.00000  0.00000   \n",
       "CAE3  27890028                                       NaN  1.00000  1.00000   \n",
       "CAE6  34768192             WARNING_TRANSCRIPT_INCOMPLETE  0.00000  0.00000   \n",
       "CAE19 23607566   WARNING_TRANSCRIPT_MULTIPLE_STOP_CODONS  0.00000  0.00000   \n",
       "CAE14 89103174                                       NaN  0.00000  0.00000   \n",
       "CAE3  24325369          WARNING_TRANSCRIPT_NO_STOP_CODON  0.00000  0.00000   \n",
       "CAE23 47818308         WARNING_TRANSCRIPT_NO_START_CODON  0.00000  0.00000   \n",
       "CAE11 61962524   WARNING_TRANSCRIPT_MULTIPLE_STOP_CODONS  0.00000  0.00000   \n",
       "CAE24 22322993                                       NaN  0.00000  0.00000   \n",
       "CAE13 11995990                                       NaN  0.00000  0.00000   \n",
       "CAE20 130570772                                      NaN  0.00000  0.00000   \n",
       "      80034861                                       NaN  0.00000  0.00000   \n",
       "CAE22 59394111                                       NaN  0.00000  0.00000   \n",
       "CAE24 8515944           WARNING_TRANSCRIPT_NO_STOP_CODON  0.00000  0.00000   \n",
       "CAE1  21503929                                       NaN  0.00000  0.00000   \n",
       "CAE6  28041036                                       NaN  0.00000  0.00000   \n",
       "CAE8  35798069                                       NaN  0.00000  0.00000   \n",
       "CAE1  116200400        WARNING_TRANSCRIPT_NO_START_CODON  0.00000  0.00000   \n",
       "CAE8  31001692   WARNING_TRANSCRIPT_MULTIPLE_STOP_CODONS  1.00000  1.00000   \n",
       "CAE10 57303137             WARNING_TRANSCRIPT_INCOMPLETE  1.00000  1.00000   \n",
       "CAE16 59272885         WARNING_TRANSCRIPT_NO_START_CODON  0.00000  0.00000   \n",
       "CAE12 22402461   WARNING_TRANSCRIPT_MULTIPLE_STOP_CODONS  1.00000  1.00000   \n",
       "CAE25 24784772             WARNING_TRANSCRIPT_INCOMPLETE  0.00000  0.00000   \n",
       "\n",
       "                      pyn       pys       sab       sac  sar       tan  \\\n",
       "CHROM POS                                                                \n",
       "CAEX  47049265   1.000000  1.000000  0.000000  0.000000  0.0  0.000000   \n",
       "CAE16 46180030   1.000000  1.000000  0.000000  0.000000  0.0  0.000000   \n",
       "CAE11 75076181   0.833333  1.000000  0.000000  0.000000  0.0  0.000000   \n",
       "CAE7  97439521   1.000000  1.000000  0.041667  0.014706  0.0  1.000000   \n",
       "CAE22 95483322   0.833333  1.000000  0.000000  0.000000  0.0  0.045455   \n",
       "CAE11 101983690  0.916667  0.980392  0.000000  0.000000  0.0  0.954545   \n",
       "CAE1  9938424    0.916667  0.980392  0.000000  0.000000  0.0  1.000000   \n",
       "CAE18 17036302   1.000000  1.000000  0.000000  0.044118  0.0  1.000000   \n",
       "CAE11 48208899   1.000000  1.000000  0.000000  0.000000  0.0  0.545455   \n",
       "CAE1  255393     1.000000  0.872549  0.000000  0.000000  0.6  0.181818   \n",
       "CAE10 76025378   0.916667  1.000000  0.000000  0.000000  0.0  0.318182   \n",
       "CAE14 12377693   1.000000  0.931373  0.000000  0.000000  0.0  0.363636   \n",
       "CAE2  25313286   0.916667  1.000000  0.104167  0.088235  0.0  1.000000   \n",
       "CAE20 42794243   1.000000  1.000000  0.062500  0.000000  0.2  0.772727   \n",
       "CAE2  33907866   1.000000  1.000000  0.000000  0.029412  0.0  1.000000   \n",
       "CAE20 14359047   0.833333  0.000000  0.000000  0.000000  0.0  0.000000   \n",
       "CAE2  7344987    1.000000  1.000000  0.000000  0.000000  0.0  0.954545   \n",
       "CAE4  76207128   0.916667  0.911765  0.000000  0.000000  0.0  0.045455   \n",
       "CAE13 76998487   0.500000  0.931373  0.000000  0.000000  0.0  0.000000   \n",
       "CAE6  48190642   0.000000  0.000000  0.041667  0.014706  0.0  0.818182   \n",
       "CAE9  45370011   1.000000  0.166667  0.020833  0.000000  0.0  1.000000   \n",
       "CAE13 369349     0.083333  0.950980  0.000000  0.000000  0.0  0.181818   \n",
       "CAE28 12426794   0.833333  0.862745  0.000000  0.000000  0.0  0.090909   \n",
       "CAE1  65928019   1.000000  1.000000  0.041667  0.029412  0.0  1.000000   \n",
       "CAE6  38843413   0.333333  0.941176  0.000000  0.000000  0.0  0.000000   \n",
       "CAE7  129628649  0.583333  0.950980  0.000000  0.000000  0.0  0.000000   \n",
       "CAE12 60988799   1.000000  0.882353  0.062500  0.029412  0.0  0.818182   \n",
       "CAE2  562475     0.250000  0.009804  0.000000  0.000000  0.0  0.863636   \n",
       "CAE20 78966415   0.833333  1.000000  0.000000  0.000000  0.0  0.681818   \n",
       "CAE25 77331379   1.000000  0.754902  0.000000  0.000000  0.0  0.681818   \n",
       "...                   ...       ...       ...       ...  ...       ...   \n",
       "CAE12 96863638   0.000000  0.000000  0.000000  0.000000  0.5  0.000000   \n",
       "CAE9  107123094  0.000000  0.000000  0.000000  0.029412  0.0  0.000000   \n",
       "CAE10 51550334   0.000000  0.000000  0.000000  0.058824  0.0  0.000000   \n",
       "CAE6  35435749   0.000000  0.000000  0.000000  0.000000  0.2  0.000000   \n",
       "CAE22 36444200   0.000000  0.000000  0.000000  0.000000  0.1  0.000000   \n",
       "CAE28 3443945    0.000000  0.000000  0.000000  0.117647  0.0  0.000000   \n",
       "CAE5  58398005   0.000000  0.000000  0.000000  0.088235  0.1  0.000000   \n",
       "CAE26 21280477   0.000000  0.000000  0.000000  0.235294  0.0  0.000000   \n",
       "CAE3  27890028   1.000000  1.000000  1.000000  0.691176  1.0  1.000000   \n",
       "CAE6  34768192   0.000000  0.000000  0.000000  0.014706  0.9  0.000000   \n",
       "CAE19 23607566   0.000000  0.000000  0.000000  0.161765  0.5  0.000000   \n",
       "CAE14 89103174   0.000000  0.000000  0.000000  0.014706  0.0  0.000000   \n",
       "CAE3  24325369   0.000000  0.000000  0.000000  0.073529  0.1  0.000000   \n",
       "CAE23 47818308   0.000000  0.000000  0.000000  0.102941  0.0  0.000000   \n",
       "CAE11 61962524   0.000000  0.000000  0.000000  0.014706  0.1  0.000000   \n",
       "CAE24 22322993   0.000000  0.000000  0.000000  0.000000  0.1  0.000000   \n",
       "CAE13 11995990   0.000000  0.000000  0.000000  0.073529  0.0  0.000000   \n",
       "CAE20 130570772  0.000000  0.000000  0.000000  0.073529  0.0  0.000000   \n",
       "      80034861   0.000000  0.000000  0.000000  0.058824  0.0  0.000000   \n",
       "CAE22 59394111   0.000000  0.000000  0.000000  0.205882  0.0  0.000000   \n",
       "CAE24 8515944    0.000000  0.000000  0.000000  0.147059  0.0  0.000000   \n",
       "CAE1  21503929   0.000000  0.000000  0.000000  0.161765  0.0  0.000000   \n",
       "CAE6  28041036   0.000000  0.000000  0.000000  0.000000  0.5  0.000000   \n",
       "CAE8  35798069   0.000000  0.000000  0.000000  0.058824  0.0  0.000000   \n",
       "CAE1  116200400  0.000000  0.000000  0.000000  0.000000  0.2  0.000000   \n",
       "CAE8  31001692   1.000000  1.000000  1.000000  0.867647  0.7  1.000000   \n",
       "CAE10 57303137   1.000000  1.000000  1.000000  1.000000  1.0  1.000000   \n",
       "CAE16 59272885   0.000000  0.000000  0.000000  0.000000  0.3  0.000000   \n",
       "CAE12 22402461   1.000000  1.000000  1.000000  0.691176  0.6  1.000000   \n",
       "CAE25 24784772   0.000000  0.000000  0.000000  0.000000  0.4  0.000000   \n",
       "\n",
       "                   af_std  \n",
       "CHROM POS                  \n",
       "CAEX  47049265   0.518034  \n",
       "CAE16 46180030   0.516398  \n",
       "CAE11 75076181   0.510041  \n",
       "CAE7  97439521   0.505811  \n",
       "CAE22 95483322   0.500700  \n",
       "CAE11 101983690  0.498022  \n",
       "CAE1  9938424    0.495926  \n",
       "CAE18 17036302   0.493084  \n",
       "CAE11 48208899   0.490404  \n",
       "CAE1  255393     0.485626  \n",
       "CAE10 76025378   0.483629  \n",
       "CAE14 12377693   0.482247  \n",
       "CAE2  25313286   0.480960  \n",
       "CAE20 42794243   0.479483  \n",
       "CAE2  33907866   0.476976  \n",
       "CAE20 14359047   0.476290  \n",
       "CAE2  7344987    0.473103  \n",
       "CAE4  76207128   0.470968  \n",
       "CAE13 76998487   0.468152  \n",
       "CAE6  48190642   0.467900  \n",
       "CAE9  45370011   0.467231  \n",
       "CAE13 369349     0.466310  \n",
       "CAE28 12426794   0.460704  \n",
       "CAE1  65928019   0.459547  \n",
       "CAE6  38843413   0.458475  \n",
       "CAE7  129628649  0.456684  \n",
       "CAE12 60988799   0.454149  \n",
       "CAE2  562475     0.450773  \n",
       "CAE20 78966415   0.439658  \n",
       "CAE25 77331379   0.438054  \n",
       "...                   ...  \n",
       "CAE12 96863638   0.000000  \n",
       "CAE9  107123094  0.000000  \n",
       "CAE10 51550334   0.000000  \n",
       "CAE6  35435749   0.000000  \n",
       "CAE22 36444200   0.000000  \n",
       "CAE28 3443945    0.000000  \n",
       "CAE5  58398005   0.000000  \n",
       "CAE26 21280477   0.000000  \n",
       "CAE3  27890028   0.000000  \n",
       "CAE6  34768192   0.000000  \n",
       "CAE19 23607566   0.000000  \n",
       "CAE14 89103174   0.000000  \n",
       "CAE3  24325369   0.000000  \n",
       "CAE23 47818308   0.000000  \n",
       "CAE11 61962524   0.000000  \n",
       "CAE24 22322993   0.000000  \n",
       "CAE13 11995990   0.000000  \n",
       "CAE20 130570772  0.000000  \n",
       "      80034861   0.000000  \n",
       "CAE22 59394111   0.000000  \n",
       "CAE24 8515944    0.000000  \n",
       "CAE1  21503929   0.000000  \n",
       "CAE6  28041036   0.000000  \n",
       "CAE8  35798069   0.000000  \n",
       "CAE1  116200400  0.000000  \n",
       "CAE8  31001692   0.000000  \n",
       "CAE10 57303137   0.000000  \n",
       "CAE16 59272885   0.000000  \n",
       "CAE12 22402461   0.000000  \n",
       "CAE25 24784772   0.000000  \n",
       "\n",
       "[6410 rows x 19 columns]"
      ]
     },
     "execution_count": 162,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "impact_var_df[impact_var_df[\"impact\"]==\"HIGH\"].sort(\"af_std\",ascending=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 2",
   "language": "python",
   "name": "python2"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 2
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython2",
   "version": "2.7.3"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 0
}
