{
 "metadata": {
  "name": "",
  "signature": "sha256:8ca6f03683c47baee14013b1889b4f7f9a19c08f69d589af3ae28044ddfa3e79"
 },
 "nbformat": 3,
 "nbformat_minor": 0,
 "worksheets": [
  {
   "cells": [
    {
     "cell_type": "code",
     "collapsed": false,
     "input": [
      "import os\n",
      "import pandas as pd\n",
      "import numpy as np\n",
      "eu = os.path.expanduser\n",
      "jn = os.path.join"
     ],
     "language": "python",
     "metadata": {},
     "outputs": [],
     "prompt_number": 1
    },
    {
     "cell_type": "code",
     "collapsed": false,
     "input": [
      "assoc_dir = eu(\"~/vervet_project/data/gene_associations\")"
     ],
     "language": "python",
     "metadata": {},
     "outputs": [],
     "prompt_number": 2
    },
    {
     "cell_type": "code",
     "collapsed": false,
     "input": [
      "meta_dir = eu(\"~/vervet_project/metadata\")\n",
      "def load_gene_df():\n",
      "    gene_df = pd.read_csv(jn(meta_dir,\"annot_mike_genes_minimal.tsv\"),sep=\"\\t\",index_col = [0,1])\n",
      "    return gene_df"
     ],
     "language": "python",
     "metadata": {},
     "outputs": [],
     "prompt_number": 3
    },
    {
     "cell_type": "code",
     "collapsed": false,
     "input": [
      "gene_df = load_gene_df()"
     ],
     "language": "python",
     "metadata": {},
     "outputs": [],
     "prompt_number": 4
    },
    {
     "cell_type": "code",
     "collapsed": false,
     "input": [
      "pops = [\"aet\",\"cyn\",\"pyn\",\"pys\",\"sab\",\"tan\"]"
     ],
     "language": "python",
     "metadata": {},
     "outputs": [],
     "prompt_number": 5
    },
    {
     "cell_type": "code",
     "collapsed": false,
     "input": [
      "gene_df.to_csv(jn(meta_dir,\"annot_mike_genes_minimal.tsv\"),sep=\"\\t\")"
     ],
     "language": "python",
     "metadata": {},
     "outputs": [],
     "prompt_number": 32
    },
    {
     "cell_type": "code",
     "collapsed": false,
     "input": [
      "def index_rolling(s,window,func,overlap=0,*args,**kwargs):\n",
      "    \"\"\"\n",
      "    Apply function in rolling windows, where the window\n",
      "    size is defined with respect to the index values.\n",
      "    This means that different windows can comprise different\n",
      "    numbers of elements.\n",
      "    \n",
      "    s ... pandas Series\n",
      "    window ... window size in units of the index values\n",
      "    func ... function to apply to the series values within each\n",
      "                window\n",
      "    overlap ... oberlap size of windows\n",
      "    args, kwarg ... additional arguments for func\n",
      "    \n",
      "    Example: index_rolling(pd.Series(range(20),index=np.random.rand(20))),0.1,max)\n",
      "    \"\"\"\n",
      "    #note that basis must be sorted in order for this to work properly\n",
      "    windows_min = s.index.min()\n",
      "    windows_max = s.index.max()\n",
      "    window_starts = np.arange(windows_min, windows_max, window-overlap)\n",
      "    window_starts = pd.Series(window_starts, index = window_starts)\n",
      "    def applyToWindow(val):\n",
      "        # using slice_indexer rather that what.loc [val:val+window] allows\n",
      "        # window limits that are not specifically in the index\n",
      "        indexer = s.index.slice_indexer(val,val+window,1)\n",
      "        chunk = s.iloc[indexer]\n",
      "        try:\n",
      "            return func(chunk,*args,**kwargs)\n",
      "        except ValueError, e:\n",
      "            if \"empty sequence\" in str(e):\n",
      "                #print indexer, chunk\n",
      "                return None\n",
      "            else:\n",
      "                raise           \n",
      "    rolled = window_starts.apply(applyToWindow)\n",
      "    return rolled"
     ],
     "language": "python",
     "metadata": {},
     "outputs": [],
     "prompt_number": 6
    },
    {
     "cell_type": "heading",
     "level": 2,
     "metadata": {},
     "source": [
      "ROD (reference ordered data) permutation test"
     ]
    },
    {
     "cell_type": "markdown",
     "metadata": {},
     "source": [
      "tools that, given an analysis result, one can rotate (shift) the results randomly around the genome"
     ]
    },
    {
     "cell_type": "code",
     "collapsed": false,
     "input": [
      "def get_chrom_length():\n",
      "    return pd.read_csv(eu(\"~/vervet_project/metadata/ref3500.tsv\"),\n",
      "                       sep=\"\\t\",index_col=0,squeeze=True,header=False)"
     ],
     "language": "python",
     "metadata": {},
     "outputs": [],
     "prompt_number": 7
    },
    {
     "cell_type": "code",
     "collapsed": false,
     "input": [
      "def shift_rod(rod_df, rnd, mode = \"grid\"):\n",
      "    \"\"\"\n",
      "    shift reference ordered data across the whole genome\n",
      "\n",
      "    Input:\n",
      "    rod_df ... pandas dataframe or series with mulitiindex (chrom, pos)\n",
      "                \n",
      "    modes ...\n",
      "        'grid' ... just rotate the index of the rod data frame\n",
      "                    this means that the positions stay the same only the\n",
      "                    value for each position becomes different\n",
      "                    Faster, but means that you only hit the same grid-point\n",
      "                    this should make it conservative on large grids. Large\n",
      "                    grids are problematic if the fraction of top windows \n",
      "                    considered becomes large \n",
      "        'continuous' ... add the random shift to each index value.\n",
      "                         NOT IMPLEMENTED\n",
      "    \"\"\"\n",
      "    if mode == \"grid\":\n",
      "        new_start_i = int(len(rod_df)*rnd)\n",
      "        rotate_data = np.concatenate((rod_df.iloc[new_start_i:].values,rod_df.iloc[:new_start_i].values))\n",
      "        if  isinstance(rod_df,pd.core.series.Series):\n",
      "            r = pd.Series(rotate_data,index=rod_df.index)\n",
      "            return r\n",
      "        elif isinstance(rod_df,pd.core.frame.DataFrame):\n",
      "            r = pd.DataFrame(rotate_data,index=rod_df.index,columns=rod_df.columns)\n",
      "            return r\n",
      "        \n",
      "    else:\n",
      "        raise UserException(\"Only mode grid supported.\")"
     ],
     "language": "python",
     "metadata": {},
     "outputs": [],
     "prompt_number": 8
    },
    {
     "cell_type": "code",
     "collapsed": false,
     "input": [
      "%timeit shift_rod(tot_sweed_df[\"Likelihood\"], 0.32424) "
     ],
     "language": "python",
     "metadata": {},
     "outputs": [
      {
       "output_type": "stream",
       "stream": "stdout",
       "text": [
        "100 loops, best of 3: 14.5 ms per loop\n"
       ]
      }
     ],
     "prompt_number": 900
    },
    {
     "cell_type": "heading",
     "level": 2,
     "metadata": {},
     "source": [
      "get genes for peaks"
     ]
    },
    {
     "cell_type": "code",
     "collapsed": false,
     "input": [
      "def get_peaks(value_s,min_peak_dist):\n",
      "    \"\"\"\n",
      "    ranked df of peaks\n",
      "    assuming a minimum distance between peaks\n",
      "    \"\"\"\n",
      "    assert len(value_s.index.droplevel(1).unique()) == 1, \"Multiple chromsosomes not supported.\"\n",
      "    chrom = value_s.index[0][0]\n",
      "    x = index_rolling(value_s.ix[chrom],2*min_peak_dist,np.argmax,min_peak_dist)\n",
      "    x.name = \"pos\"\n",
      "    #only take peaks that are found in at least two windows \n",
      "    #this makes sure that we have a minimal distance between peaks\n",
      "    pos_true = (x.reset_index().groupby(\"pos\").apply(len)>1)\n",
      "    pos = pos_true[pos_true.values].index\n",
      "    s = value_s.ix[chrom].ix[pos]\n",
      "    return s"
     ],
     "language": "python",
     "metadata": {},
     "outputs": [],
     "prompt_number": 9
    },
    {
     "cell_type": "code",
     "collapsed": false,
     "input": [
      "def get_genes(peak_s, gene_df, max_dist):\n",
      "    \"\"\"\n",
      "    take the input series and gets \n",
      "    names of genes nearby\n",
      "    \n",
      "    Input:\n",
      "    peak_s ... pandas series with (chrom, pos) index and value of\n",
      "                the statistic ('peak height'). Series should be named.\n",
      "    gene_df ... data frame with gene info \n",
      "    \"\"\"    \n",
      "    all_genes = []\n",
      "    if not gene_df.index.is_monotonic:\n",
      "        gene_df = gene_df.sort_index()\n",
      "    tot_hit_df = pd.DataFrame()\n",
      "    for chrom in peak_s.index.droplevel(1).unique():\n",
      "        loc_gene_df = gene_df.ix[chrom]\n",
      "        #loc_gene_df = loc_gene_df.append(pd.DataFrame(np.nan,index=[np.inf],columns=loc_gene_df.columns))\n",
      "        pos_rel_to_start = np.searchsorted(loc_gene_df.index.values-max_dist,peak_s.ix[chrom].index.values)\n",
      "        pos_rel_to_end = np.searchsorted(loc_gene_df[\"end\"].values+max_dist,peak_s.ix[chrom].index.values)\n",
      "        genes = list(set(loc_gene_df[\"gene_id\"].iloc[np.hstack([range(a,b) for a,b in zip(pos_rel_to_end,pos_rel_to_start)])]))\n",
      "        #print chrom, genes\n",
      "        all_genes += genes\n",
      "    return all_genes\n",
      "\n",
      "def get_gene_info(gene_ls,gene_df):\n",
      "    \"\"\"\n",
      "    for a list of gene ids,\n",
      "    get a data frame with their \n",
      "    position\n",
      "    \"\"\"\n",
      "    gi = gene_df[gene_df[\"gene_id\"].apply(lambda x: x in gene_ls)]\n",
      "    return gi\n",
      "\n",
      "def get_peaks(gene_info,top_s,max_dist):\n",
      "    \"\"\"\n",
      "    For each gene in gene_info get the\n",
      "    peaks within max_dist in top_s. This \n",
      "    is basically reverse engineering to get\n",
      "    the peak info for each gene that was found \n",
      "    to be associated with a peak. \n",
      "    \n",
      "    Input:\n",
      "    gene_info ... data frame with index ('chrom','start')\n",
      "                and columns 'gene_id' and 'end'\n",
      "    top_s ... series of peak positions with index (chrom, pos)\n",
      "                and values peak height\n",
      "    max_dist ... maximum distance between gene and peak\n",
      "    \"\"\"\n",
      "    def get_dist(df,gene_pos):\n",
      "        s = pd.Series(df.index.droplevel(0).values - gene_pos.ix[df.index[0][0]],\n",
      "                                                  index=df.index.droplevel(0).values)\n",
      "        return s\n",
      "    tot_gene_peaks_df = pd.DataFrame()\n",
      "    if not top_s.index.is_monotonic:\n",
      "        top_s = top_s.sort_index()\n",
      "    for chrom in gene_info.index.droplevel(1).unique():\n",
      "        loc_top_s = top_s.ix[chrom]\n",
      "        start = np.searchsorted(loc_top_s.index.values+max_dist,gene_info.ix[chrom].index.values)\n",
      "        end = np.searchsorted(loc_top_s.index.values-max_dist,gene_info.ix[chrom][\"end\"].values)\n",
      "        \n",
      "        x = pd.concat([loc_top_s.iloc[st:ed] for st,ed in zip(start,end)],\n",
      "                      keys=gene_info.ix[chrom][\"gene_id\"].values)\n",
      "        x.name = \"peak_height\"\n",
      "        \n",
      "\n",
      "\n",
      "        dist_start = x.groupby(lambda i: i[0]).\\\n",
      "                    apply(lambda df: get_dist(df,\n",
      "                                              gene_info.ix[chrom].reset_index().set_index(\"gene_id\")[\"pos\"]))\n",
      "        dist_start.name = \"dist_start\"\n",
      "        dist_end = x.groupby(lambda i: i[0]).\\\n",
      "                    apply(lambda df: get_dist(df,\n",
      "                                              gene_info.ix[chrom].set_index(\"gene_id\")[\"end\"]))\n",
      "        dist_end.name = \"dist_end\"\n",
      "        gene_peaks_df = pd.concat([x,dist_start,dist_end],axis=1)\n",
      "        gene_peaks_df.index = pd.MultiIndex.from_arrays([gene_peaks_df.index.droplevel(1),\n",
      "                                         [chrom]*len(x),\n",
      "                                         gene_peaks_df.index.droplevel(0)])\n",
      "        tot_gene_peaks_df = pd.concat([tot_gene_peaks_df, gene_peaks_df])\n",
      "        tot_gene_peaks_df.index.names = [\"gene_id\",\"chrom\",\"peak_pos\"]\n",
      "        if chrom == \"CAE19\":\n",
      "            print gene_info.ix[chrom].reset_index().set_index(\"gene_id\")[\"pos\"]\n",
      "            print loc_top_s\n",
      "            print start\n",
      "            print end\n",
      "            print x\n",
      "            print dist_start\n",
      "    return tot_gene_peaks_df"
     ],
     "language": "python",
     "metadata": {},
     "outputs": [],
     "prompt_number": 10
    },
    {
     "cell_type": "code",
     "collapsed": false,
     "input": [
      "chromosomes = [\"CAE\" + str(i) for i in range(1,30)+[\"X\",\"Y\"]]\n",
      "def get_sweed_df(chrom,pop):\n",
      "    sweed_res = pd.read_csv(eu(\"~/vervet_project/analyses/20140611_163_subpop_sweepfinder/\"\n",
      "                             \"_data/SweeD_Report.163_ref3500_{}_{}\".format(chrom,pop)),\n",
      "                                                          sep=\"\\t\",skiprows=2,index_col=0)\n",
      "    sweed_res.index = pd.MultiIndex.from_tuples(zip([chrom]*len(sweed_res.index),sweed_res.index))\n",
      "    return sweed_res"
     ],
     "language": "python",
     "metadata": {},
     "outputs": [],
     "prompt_number": 11
    },
    {
     "cell_type": "code",
     "collapsed": false,
     "input": [],
     "language": "python",
     "metadata": {},
     "outputs": [
      {
       "metadata": {},
       "output_type": "pyout",
       "prompt_number": 1158,
       "text": [
        "'0.14.0'"
       ]
      }
     ],
     "prompt_number": 1158
    },
    {
     "cell_type": "code",
     "collapsed": false,
     "input": [
      "pop = \"aet\"\n",
      "eu(\"~/vervet_project/analyses/20140611_163_subpop_sweepfinder/\"\n",
      "                             \"_data/SweeD_Report.163_ref3500_{}_{}\".format(chrom,pop))"
     ],
     "language": "python",
     "metadata": {},
     "outputs": [
      {
       "metadata": {},
       "output_type": "pyout",
       "prompt_number": 1057,
       "text": [
        "'/home/GMI/hannes.svardal/vervet_project/analyses/20140611_163_subpop_sweepfinder/_data/SweeD_Report.163_ref3500_CAEY_aet'"
       ]
      }
     ],
     "prompt_number": 1057
    },
    {
     "cell_type": "code",
     "collapsed": false,
     "input": [
      "gene_df = load_gene_df()\n",
      "gene_df = gene_df.sortlevel(level=[0,1])"
     ],
     "language": "python",
     "metadata": {},
     "outputs": [],
     "prompt_number": 372
    },
    {
     "cell_type": "code",
     "collapsed": false,
     "input": [
      "pop = \"aet\"\n",
      "tot_sweed_df = pd.read_csv(eu(\"~/vervet_project/analyses/20140611_163_subpop_sweepfinder/\"\n",
      "                                 \"_data/tot_sweed_result_{}.tsv\".format(pop)),sep=\"\\t\",index_col=[0,1])"
     ],
     "language": "python",
     "metadata": {},
     "outputs": [],
     "prompt_number": 19
    },
    {
     "cell_type": "code",
     "collapsed": false,
     "input": [
      "top_s = tot_sweed_df.sort(\"Likelihood\",ascending=False)[\"Likelihood\"].iloc[:2000]\n",
      "#top_s = pd.concat([pd.Series([17],index=pd.MultiIndex.from_arrays([[\"CAE1\"],[4922]])),top_s])"
     ],
     "language": "python",
     "metadata": {},
     "outputs": [],
     "prompt_number": 1152
    },
    {
     "cell_type": "code",
     "collapsed": false,
     "input": [
      "max_dist = 10000\n",
      "cand_genes = get_genes(top_s, gene_df, max_dist=max_dist)\n",
      "gene_info = get_gene_info(cand_genes, gene_df)\n",
      "peaks = get_peaks(gene_info,top_s,max_dist=max_dist)"
     ],
     "language": "python",
     "metadata": {},
     "outputs": [
      {
       "output_type": "stream",
       "stream": "stdout",
       "text": [
        "gene_id\n",
        "SAMM50     26467319\n",
        "Name: pos, dtype: int64\n",
        "pos\n",
        "9483311.2373     21.93971\n",
        "9483796.1723     23.32173\n",
        "26477373.2917    20.04123\n",
        "26478828.0967    19.82511\n",
        "Name: Likelihood, dtype: float64\n",
        "[2]\n",
        "[4]\n",
        "        pos          \n",
        "SAMM50  26477373.2917    20.04123\n",
        "        26478828.0967    19.82511\n",
        "Name: peak_height, dtype: float64\n",
        "SAMM50  26477373.2917    10054.2917\n",
        "        26478828.0967    11509.0967\n",
        "Name: dist_start, dtype: float64\n"
       ]
      }
     ],
     "prompt_number": 1225
    },
    {
     "cell_type": "code",
     "collapsed": false,
     "input": [
      "def genes_per_peak_s(df):\n",
      "    s = df.iloc[0]#[\"peak_height\"]\n",
      "    s[\"genes\"] = 3#list(df[\"gene_id\"].unique())\n",
      "    return s\n",
      "p = peaks.reset_index([0])"
     ],
     "language": "python",
     "metadata": {},
     "outputs": [],
     "prompt_number": 1248
    },
    {
     "cell_type": "code",
     "collapsed": false,
     "input": [
      "gene_list_peak_pos = p[\"gene_id\"].groupby(lambda x: x).apply(list)\n",
      "gene_list_peak_pos.name = \"genes\"\n",
      "gene_list_peak_pos.index = pd.MultiIndex.from_tuples(gene_list_peak_pos.index)\n",
      "peak_info = pd.concat([top_s,gene_list_peak_pos],axis=1)\n",
      "peak_info.sort(\"Likelihood\",ascending=False,inplace=True)"
     ],
     "language": "python",
     "metadata": {},
     "outputs": [],
     "prompt_number": 1263
    },
    {
     "cell_type": "code",
     "collapsed": false,
     "input": [],
     "language": "python",
     "metadata": {},
     "outputs": [],
     "prompt_number": 1270
    },
    {
     "cell_type": "code",
     "collapsed": false,
     "input": [],
     "language": "python",
     "metadata": {},
     "outputs": [],
     "prompt_number": 1272
    },
    {
     "cell_type": "heading",
     "level": 2,
     "metadata": {},
     "source": [
      "get go associations for gene list"
     ]
    },
    {
     "cell_type": "code",
     "collapsed": false,
     "input": [
      "gene_to_go = pd.read_csv(jn(assoc_dir,\"gene_to_cat.csv\"),index_col=0)\n",
      "gene_to_name = pd.read_csv(jn(assoc_dir,\"gene_to_name.csv\"),index_col=0)\n",
      "go_to_name = pd.read_csv(jn(assoc_dir,\"cat_to_name.csv\"),index_col=0)"
     ],
     "language": "python",
     "metadata": {},
     "outputs": [],
     "prompt_number": 11
    },
    {
     "cell_type": "code",
     "collapsed": false,
     "input": [
      "rank_table.join(go_to_name.set_index(\"go_identifier\",drop=True))"
     ],
     "language": "python",
     "metadata": {},
     "outputs": [
      {
       "html": [
        "<div style=\"max-height:1000px;max-width:1500px;overflow:auto;\">\n",
        "<table border=\"1\" class=\"dataframe\">\n",
        "  <thead>\n",
        "    <tr style=\"text-align: right;\">\n",
        "      <th></th>\n",
        "      <th>n_genes</th>\n",
        "      <th>out_of</th>\n",
        "      <th>rank</th>\n",
        "      <th>go_name</th>\n",
        "    </tr>\n",
        "    <tr>\n",
        "      <th>go_identifier</th>\n",
        "      <th></th>\n",
        "      <th></th>\n",
        "      <th></th>\n",
        "      <th></th>\n",
        "    </tr>\n",
        "  </thead>\n",
        "  <tbody>\n",
        "    <tr>\n",
        "      <th>GO:0000014</th>\n",
        "      <td>  1</td>\n",
        "      <td> 0</td>\n",
        "      <td> 0</td>\n",
        "      <td> single-stranded DNA endodeoxyribonuclease acti...</td>\n",
        "    </tr>\n",
        "    <tr>\n",
        "      <th>GO:0000060</th>\n",
        "      <td>  1</td>\n",
        "      <td> 0</td>\n",
        "      <td> 0</td>\n",
        "      <td>        protein import into nucleus, translocation</td>\n",
        "    </tr>\n",
        "    <tr>\n",
        "      <th>GO:0000075</th>\n",
        "      <td>  2</td>\n",
        "      <td> 0</td>\n",
        "      <td> 0</td>\n",
        "      <td>                             cell cycle checkpoint</td>\n",
        "    </tr>\n",
        "    <tr>\n",
        "      <th>GO:0000082</th>\n",
        "      <td>  2</td>\n",
        "      <td> 0</td>\n",
        "      <td> 0</td>\n",
        "      <td>             G1/S transition of mitotic cell cycle</td>\n",
        "    </tr>\n",
        "    <tr>\n",
        "      <th>GO:0000087</th>\n",
        "      <td>  2</td>\n",
        "      <td> 0</td>\n",
        "      <td> 0</td>\n",
        "      <td>                                   mitotic M phase</td>\n",
        "    </tr>\n",
        "    <tr>\n",
        "      <th>GO:0000088</th>\n",
        "      <td>  1</td>\n",
        "      <td> 0</td>\n",
        "      <td> 0</td>\n",
        "      <td>                                  mitotic prophase</td>\n",
        "    </tr>\n",
        "    <tr>\n",
        "      <th>GO:0000090</th>\n",
        "      <td>  1</td>\n",
        "      <td> 0</td>\n",
        "      <td> 0</td>\n",
        "      <td>                                  mitotic anaphase</td>\n",
        "    </tr>\n",
        "    <tr>\n",
        "      <th>GO:0000096</th>\n",
        "      <td>  1</td>\n",
        "      <td> 0</td>\n",
        "      <td> 0</td>\n",
        "      <td>               sulfur amino acid metabolic process</td>\n",
        "    </tr>\n",
        "    <tr>\n",
        "      <th>GO:0000098</th>\n",
        "      <td>  1</td>\n",
        "      <td> 0</td>\n",
        "      <td> 0</td>\n",
        "      <td>               sulfur amino acid catabolic process</td>\n",
        "    </tr>\n",
        "    <tr>\n",
        "      <th>GO:0000122</th>\n",
        "      <td>  2</td>\n",
        "      <td> 0</td>\n",
        "      <td> 0</td>\n",
        "      <td> negative regulation of transcription from RNA ...</td>\n",
        "    </tr>\n",
        "    <tr>\n",
        "      <th>GO:0000151</th>\n",
        "      <td>  1</td>\n",
        "      <td> 0</td>\n",
        "      <td> 0</td>\n",
        "      <td>                          ubiquitin ligase complex</td>\n",
        "    </tr>\n",
        "    <tr>\n",
        "      <th>GO:0000166</th>\n",
        "      <td> 14</td>\n",
        "      <td> 0</td>\n",
        "      <td> 0</td>\n",
        "      <td>                                nucleotide binding</td>\n",
        "    </tr>\n",
        "    <tr>\n",
        "      <th>GO:0000185</th>\n",
        "      <td>  1</td>\n",
        "      <td> 0</td>\n",
        "      <td> 0</td>\n",
        "      <td>                     activation of MAPKKK activity</td>\n",
        "    </tr>\n",
        "    <tr>\n",
        "      <th>GO:0000255</th>\n",
        "      <td>  1</td>\n",
        "      <td> 0</td>\n",
        "      <td> 0</td>\n",
        "      <td>                       allantoin metabolic process</td>\n",
        "    </tr>\n",
        "    <tr>\n",
        "      <th>GO:0000278</th>\n",
        "      <td>  2</td>\n",
        "      <td> 0</td>\n",
        "      <td> 0</td>\n",
        "      <td>                                mitotic cell cycle</td>\n",
        "    </tr>\n",
        "    <tr>\n",
        "      <th>GO:0000390</th>\n",
        "      <td>  1</td>\n",
        "      <td> 0</td>\n",
        "      <td> 0</td>\n",
        "      <td>                  spliceosomal complex disassembly</td>\n",
        "    </tr>\n",
        "    <tr>\n",
        "      <th>GO:0000398</th>\n",
        "      <td>  2</td>\n",
        "      <td> 0</td>\n",
        "      <td> 0</td>\n",
        "      <td>                    mRNA splicing, via spliceosome</td>\n",
        "    </tr>\n",
        "    <tr>\n",
        "      <th>GO:0000422</th>\n",
        "      <td>  1</td>\n",
        "      <td> 0</td>\n",
        "      <td> 0</td>\n",
        "      <td>                         mitochondrion degradation</td>\n",
        "    </tr>\n",
        "    <tr>\n",
        "      <th>GO:0000724</th>\n",
        "      <td>  1</td>\n",
        "      <td> 0</td>\n",
        "      <td> 0</td>\n",
        "      <td> double-strand break repair via homologous reco...</td>\n",
        "    </tr>\n",
        "    <tr>\n",
        "      <th>GO:0000737</th>\n",
        "      <td>  1</td>\n",
        "      <td> 0</td>\n",
        "      <td> 0</td>\n",
        "      <td>            DNA catabolic process, endonucleolytic</td>\n",
        "    </tr>\n",
        "    <tr>\n",
        "      <th>GO:0000910</th>\n",
        "      <td>  2</td>\n",
        "      <td> 0</td>\n",
        "      <td> 0</td>\n",
        "      <td>                                       cytokinesis</td>\n",
        "    </tr>\n",
        "    <tr>\n",
        "      <th>GO:0000922</th>\n",
        "      <td>  1</td>\n",
        "      <td> 0</td>\n",
        "      <td> 0</td>\n",
        "      <td>                                      spindle pole</td>\n",
        "    </tr>\n",
        "    <tr>\n",
        "      <th>GO:0000979</th>\n",
        "      <td>  1</td>\n",
        "      <td> 0</td>\n",
        "      <td> 0</td>\n",
        "      <td> RNA polymerase II core promoter sequence-speci...</td>\n",
        "    </tr>\n",
        "    <tr>\n",
        "      <th>GO:0001103</th>\n",
        "      <td>  1</td>\n",
        "      <td> 0</td>\n",
        "      <td> 0</td>\n",
        "      <td> RNA polymerase II repressing transcription fac...</td>\n",
        "    </tr>\n",
        "    <tr>\n",
        "      <th>GO:0001104</th>\n",
        "      <td>  1</td>\n",
        "      <td> 0</td>\n",
        "      <td> 0</td>\n",
        "      <td> RNA polymerase II transcription cofactor activity</td>\n",
        "    </tr>\n",
        "    <tr>\n",
        "      <th>GO:0001106</th>\n",
        "      <td>  1</td>\n",
        "      <td> 0</td>\n",
        "      <td> 0</td>\n",
        "      <td> RNA polymerase II transcription corepressor ac...</td>\n",
        "    </tr>\n",
        "    <tr>\n",
        "      <th>GO:0001523</th>\n",
        "      <td>  1</td>\n",
        "      <td> 0</td>\n",
        "      <td> 0</td>\n",
        "      <td>                        retinoid metabolic process</td>\n",
        "    </tr>\n",
        "    <tr>\n",
        "      <th>GO:0001525</th>\n",
        "      <td>  2</td>\n",
        "      <td> 0</td>\n",
        "      <td> 0</td>\n",
        "      <td>                                      angiogenesis</td>\n",
        "    </tr>\n",
        "    <tr>\n",
        "      <th>GO:0001541</th>\n",
        "      <td>  2</td>\n",
        "      <td> 0</td>\n",
        "      <td> 0</td>\n",
        "      <td>                      ovarian follicle development</td>\n",
        "    </tr>\n",
        "    <tr>\n",
        "      <th>GO:0001553</th>\n",
        "      <td>  1</td>\n",
        "      <td> 0</td>\n",
        "      <td> 0</td>\n",
        "      <td>                                     luteinization</td>\n",
        "    </tr>\n",
        "    <tr>\n",
        "      <th>...</th>\n",
        "      <td>...</td>\n",
        "      <td>...</td>\n",
        "      <td>...</td>\n",
        "      <td>...</td>\n",
        "    </tr>\n",
        "    <tr>\n",
        "      <th>GO:0097105</th>\n",
        "      <td>  1</td>\n",
        "      <td> 0</td>\n",
        "      <td> 0</td>\n",
        "      <td>                     presynaptic membrane assembly</td>\n",
        "    </tr>\n",
        "    <tr>\n",
        "      <th>GO:0097109</th>\n",
        "      <td>  1</td>\n",
        "      <td> 0</td>\n",
        "      <td> 0</td>\n",
        "      <td>                 neuroligin family protein binding</td>\n",
        "    </tr>\n",
        "    <tr>\n",
        "      <th>GO:0097112</th>\n",
        "      <td>  1</td>\n",
        "      <td> 0</td>\n",
        "      <td> 0</td>\n",
        "      <td>       gamma-aminobutyric acid receptor clustering</td>\n",
        "    </tr>\n",
        "    <tr>\n",
        "      <th>GO:0097114</th>\n",
        "      <td>  1</td>\n",
        "      <td> 0</td>\n",
        "      <td> 0</td>\n",
        "      <td>          N-methyl-D-aspartate receptor clustering</td>\n",
        "    </tr>\n",
        "    <tr>\n",
        "      <th>GO:0097116</th>\n",
        "      <td>  1</td>\n",
        "      <td> 0</td>\n",
        "      <td> 0</td>\n",
        "      <td>                               gephyrin clustering</td>\n",
        "    </tr>\n",
        "    <tr>\n",
        "      <th>GO:0097117</th>\n",
        "      <td>  1</td>\n",
        "      <td> 0</td>\n",
        "      <td> 0</td>\n",
        "      <td>    guanylate kinase-associated protein clustering</td>\n",
        "    </tr>\n",
        "    <tr>\n",
        "      <th>GO:0097118</th>\n",
        "      <td>  1</td>\n",
        "      <td> 0</td>\n",
        "      <td> 0</td>\n",
        "      <td>                             neuroligin clustering</td>\n",
        "    </tr>\n",
        "    <tr>\n",
        "      <th>GO:0097119</th>\n",
        "      <td>  1</td>\n",
        "      <td> 0</td>\n",
        "      <td> 0</td>\n",
        "      <td>        postsynaptic density protein 95 clustering</td>\n",
        "    </tr>\n",
        "    <tr>\n",
        "      <th>GO:0097120</th>\n",
        "      <td>  1</td>\n",
        "      <td> 0</td>\n",
        "      <td> 0</td>\n",
        "      <td>                  receptor localization to synapse</td>\n",
        "    </tr>\n",
        "    <tr>\n",
        "      <th>GO:0097286</th>\n",
        "      <td>  1</td>\n",
        "      <td> 0</td>\n",
        "      <td> 0</td>\n",
        "      <td>                                   iron ion import</td>\n",
        "    </tr>\n",
        "    <tr>\n",
        "      <th>GO:0097449</th>\n",
        "      <td>  1</td>\n",
        "      <td> 0</td>\n",
        "      <td> 0</td>\n",
        "      <td>                              astrocyte projection</td>\n",
        "    </tr>\n",
        "    <tr>\n",
        "      <th>GO:0098655</th>\n",
        "      <td>  1</td>\n",
        "      <td> 0</td>\n",
        "      <td> 0</td>\n",
        "      <td>                    cation transmembrane transport</td>\n",
        "    </tr>\n",
        "    <tr>\n",
        "      <th>GO:1900025</th>\n",
        "      <td>  1</td>\n",
        "      <td> 0</td>\n",
        "      <td> 0</td>\n",
        "      <td> negative regulation of substrate adhesion-depe...</td>\n",
        "    </tr>\n",
        "    <tr>\n",
        "      <th>GO:1902186</th>\n",
        "      <td>  1</td>\n",
        "      <td> 0</td>\n",
        "      <td> 0</td>\n",
        "      <td>        regulation of viral release from host cell</td>\n",
        "    </tr>\n",
        "    <tr>\n",
        "      <th>GO:1902305</th>\n",
        "      <td>  1</td>\n",
        "      <td> 0</td>\n",
        "      <td> 0</td>\n",
        "      <td>  regulation of sodium ion transmembrane transport</td>\n",
        "    </tr>\n",
        "    <tr>\n",
        "      <th>GO:1902560</th>\n",
        "      <td>  1</td>\n",
        "      <td> 0</td>\n",
        "      <td> 0</td>\n",
        "      <td>                             GMP reductase complex</td>\n",
        "    </tr>\n",
        "    <tr>\n",
        "      <th>GO:1902749</th>\n",
        "      <td>  1</td>\n",
        "      <td> 0</td>\n",
        "      <td> 0</td>\n",
        "      <td>    regulation of cell cycle G2/M phase transition</td>\n",
        "    </tr>\n",
        "    <tr>\n",
        "      <th>GO:1902911</th>\n",
        "      <td>  1</td>\n",
        "      <td> 0</td>\n",
        "      <td> 0</td>\n",
        "      <td>                            protein kinase complex</td>\n",
        "    </tr>\n",
        "    <tr>\n",
        "      <th>GO:1990090</th>\n",
        "      <td>  1</td>\n",
        "      <td> 0</td>\n",
        "      <td> 0</td>\n",
        "      <td> cellular response to nerve growth factor stimulus</td>\n",
        "    </tr>\n",
        "    <tr>\n",
        "      <th>GO:2000114</th>\n",
        "      <td>  1</td>\n",
        "      <td> 0</td>\n",
        "      <td> 0</td>\n",
        "      <td>      regulation of establishment of cell polarity</td>\n",
        "    </tr>\n",
        "    <tr>\n",
        "      <th>GO:2000134</th>\n",
        "      <td>  1</td>\n",
        "      <td> 0</td>\n",
        "      <td> 0</td>\n",
        "      <td> negative regulation of G1/S transition of mito...</td>\n",
        "    </tr>\n",
        "    <tr>\n",
        "      <th>GO:2000145</th>\n",
        "      <td>  1</td>\n",
        "      <td> 0</td>\n",
        "      <td> 0</td>\n",
        "      <td>                       regulation of cell motility</td>\n",
        "    </tr>\n",
        "    <tr>\n",
        "      <th>GO:2000304</th>\n",
        "      <td>  1</td>\n",
        "      <td> 0</td>\n",
        "      <td> 0</td>\n",
        "      <td> positive regulation of ceramide biosynthetic p...</td>\n",
        "    </tr>\n",
        "    <tr>\n",
        "      <th>GO:2000310</th>\n",
        "      <td>  1</td>\n",
        "      <td> 0</td>\n",
        "      <td> 0</td>\n",
        "      <td> regulation of N-methyl-D-aspartate selective g...</td>\n",
        "    </tr>\n",
        "    <tr>\n",
        "      <th>GO:2000311</th>\n",
        "      <td>  1</td>\n",
        "      <td> 0</td>\n",
        "      <td> 0</td>\n",
        "      <td> regulation of alpha-amino-3-hydroxy-5-methyl-4...</td>\n",
        "    </tr>\n",
        "    <tr>\n",
        "      <th>GO:2000463</th>\n",
        "      <td>  1</td>\n",
        "      <td> 0</td>\n",
        "      <td> 0</td>\n",
        "      <td> positive regulation of excitatory postsynaptic...</td>\n",
        "    </tr>\n",
        "    <tr>\n",
        "      <th>GO:2000599</th>\n",
        "      <td>  1</td>\n",
        "      <td> 0</td>\n",
        "      <td> 0</td>\n",
        "      <td>   negative regulation of cyclin catabolic process</td>\n",
        "    </tr>\n",
        "    <tr>\n",
        "      <th>GO:2001014</th>\n",
        "      <td>  1</td>\n",
        "      <td> 0</td>\n",
        "      <td> 0</td>\n",
        "      <td> regulation of skeletal muscle cell differentia...</td>\n",
        "    </tr>\n",
        "    <tr>\n",
        "      <th>GO:2001234</th>\n",
        "      <td>  1</td>\n",
        "      <td> 0</td>\n",
        "      <td> 0</td>\n",
        "      <td> negative regulation of apoptotic signaling pat...</td>\n",
        "    </tr>\n",
        "    <tr>\n",
        "      <th>GO:2001237</th>\n",
        "      <td>  1</td>\n",
        "      <td> 0</td>\n",
        "      <td> 0</td>\n",
        "      <td> negative regulation of extrinsic apoptotic sig...</td>\n",
        "    </tr>\n",
        "  </tbody>\n",
        "</table>\n",
        "<p>888 rows \u00d7 4 columns</p>\n",
        "</div>"
       ],
       "metadata": {},
       "output_type": "pyout",
       "prompt_number": 1077,
       "text": [
        "               n_genes  out_of  rank  \\\n",
        "go_identifier                          \n",
        "GO:0000014           1       0     0   \n",
        "GO:0000060           1       0     0   \n",
        "GO:0000075           2       0     0   \n",
        "GO:0000082           2       0     0   \n",
        "GO:0000087           2       0     0   \n",
        "GO:0000088           1       0     0   \n",
        "GO:0000090           1       0     0   \n",
        "GO:0000096           1       0     0   \n",
        "GO:0000098           1       0     0   \n",
        "GO:0000122           2       0     0   \n",
        "GO:0000151           1       0     0   \n",
        "GO:0000166          14       0     0   \n",
        "GO:0000185           1       0     0   \n",
        "GO:0000255           1       0     0   \n",
        "GO:0000278           2       0     0   \n",
        "GO:0000390           1       0     0   \n",
        "GO:0000398           2       0     0   \n",
        "GO:0000422           1       0     0   \n",
        "GO:0000724           1       0     0   \n",
        "GO:0000737           1       0     0   \n",
        "GO:0000910           2       0     0   \n",
        "GO:0000922           1       0     0   \n",
        "GO:0000979           1       0     0   \n",
        "GO:0001103           1       0     0   \n",
        "GO:0001104           1       0     0   \n",
        "GO:0001106           1       0     0   \n",
        "GO:0001523           1       0     0   \n",
        "GO:0001525           2       0     0   \n",
        "GO:0001541           2       0     0   \n",
        "GO:0001553           1       0     0   \n",
        "...                ...     ...   ...   \n",
        "GO:0097105           1       0     0   \n",
        "GO:0097109           1       0     0   \n",
        "GO:0097112           1       0     0   \n",
        "GO:0097114           1       0     0   \n",
        "GO:0097116           1       0     0   \n",
        "GO:0097117           1       0     0   \n",
        "GO:0097118           1       0     0   \n",
        "GO:0097119           1       0     0   \n",
        "GO:0097120           1       0     0   \n",
        "GO:0097286           1       0     0   \n",
        "GO:0097449           1       0     0   \n",
        "GO:0098655           1       0     0   \n",
        "GO:1900025           1       0     0   \n",
        "GO:1902186           1       0     0   \n",
        "GO:1902305           1       0     0   \n",
        "GO:1902560           1       0     0   \n",
        "GO:1902749           1       0     0   \n",
        "GO:1902911           1       0     0   \n",
        "GO:1990090           1       0     0   \n",
        "GO:2000114           1       0     0   \n",
        "GO:2000134           1       0     0   \n",
        "GO:2000145           1       0     0   \n",
        "GO:2000304           1       0     0   \n",
        "GO:2000310           1       0     0   \n",
        "GO:2000311           1       0     0   \n",
        "GO:2000463           1       0     0   \n",
        "GO:2000599           1       0     0   \n",
        "GO:2001014           1       0     0   \n",
        "GO:2001234           1       0     0   \n",
        "GO:2001237           1       0     0   \n",
        "\n",
        "                                                         go_name  \n",
        "go_identifier                                                     \n",
        "GO:0000014     single-stranded DNA endodeoxyribonuclease acti...  \n",
        "GO:0000060            protein import into nucleus, translocation  \n",
        "GO:0000075                                 cell cycle checkpoint  \n",
        "GO:0000082                 G1/S transition of mitotic cell cycle  \n",
        "GO:0000087                                       mitotic M phase  \n",
        "GO:0000088                                      mitotic prophase  \n",
        "GO:0000090                                      mitotic anaphase  \n",
        "GO:0000096                   sulfur amino acid metabolic process  \n",
        "GO:0000098                   sulfur amino acid catabolic process  \n",
        "GO:0000122     negative regulation of transcription from RNA ...  \n",
        "GO:0000151                              ubiquitin ligase complex  \n",
        "GO:0000166                                    nucleotide binding  \n",
        "GO:0000185                         activation of MAPKKK activity  \n",
        "GO:0000255                           allantoin metabolic process  \n",
        "GO:0000278                                    mitotic cell cycle  \n",
        "GO:0000390                      spliceosomal complex disassembly  \n",
        "GO:0000398                        mRNA splicing, via spliceosome  \n",
        "GO:0000422                             mitochondrion degradation  \n",
        "GO:0000724     double-strand break repair via homologous reco...  \n",
        "GO:0000737                DNA catabolic process, endonucleolytic  \n",
        "GO:0000910                                           cytokinesis  \n",
        "GO:0000922                                          spindle pole  \n",
        "GO:0000979     RNA polymerase II core promoter sequence-speci...  \n",
        "GO:0001103     RNA polymerase II repressing transcription fac...  \n",
        "GO:0001104     RNA polymerase II transcription cofactor activity  \n",
        "GO:0001106     RNA polymerase II transcription corepressor ac...  \n",
        "GO:0001523                            retinoid metabolic process  \n",
        "GO:0001525                                          angiogenesis  \n",
        "GO:0001541                          ovarian follicle development  \n",
        "GO:0001553                                         luteinization  \n",
        "...                                                          ...  \n",
        "GO:0097105                         presynaptic membrane assembly  \n",
        "GO:0097109                     neuroligin family protein binding  \n",
        "GO:0097112           gamma-aminobutyric acid receptor clustering  \n",
        "GO:0097114              N-methyl-D-aspartate receptor clustering  \n",
        "GO:0097116                                   gephyrin clustering  \n",
        "GO:0097117        guanylate kinase-associated protein clustering  \n",
        "GO:0097118                                 neuroligin clustering  \n",
        "GO:0097119            postsynaptic density protein 95 clustering  \n",
        "GO:0097120                      receptor localization to synapse  \n",
        "GO:0097286                                       iron ion import  \n",
        "GO:0097449                                  astrocyte projection  \n",
        "GO:0098655                        cation transmembrane transport  \n",
        "GO:1900025     negative regulation of substrate adhesion-depe...  \n",
        "GO:1902186            regulation of viral release from host cell  \n",
        "GO:1902305      regulation of sodium ion transmembrane transport  \n",
        "GO:1902560                                 GMP reductase complex  \n",
        "GO:1902749        regulation of cell cycle G2/M phase transition  \n",
        "GO:1902911                                protein kinase complex  \n",
        "GO:1990090     cellular response to nerve growth factor stimulus  \n",
        "GO:2000114          regulation of establishment of cell polarity  \n",
        "GO:2000134     negative regulation of G1/S transition of mito...  \n",
        "GO:2000145                           regulation of cell motility  \n",
        "GO:2000304     positive regulation of ceramide biosynthetic p...  \n",
        "GO:2000310     regulation of N-methyl-D-aspartate selective g...  \n",
        "GO:2000311     regulation of alpha-amino-3-hydroxy-5-methyl-4...  \n",
        "GO:2000463     positive regulation of excitatory postsynaptic...  \n",
        "GO:2000599       negative regulation of cyclin catabolic process  \n",
        "GO:2001014     regulation of skeletal muscle cell differentia...  \n",
        "GO:2001234     negative regulation of apoptotic signaling pat...  \n",
        "GO:2001237     negative regulation of extrinsic apoptotic sig...  \n",
        "\n",
        "[888 rows x 4 columns]"
       ]
      }
     ],
     "prompt_number": 1077
    },
    {
     "cell_type": "code",
     "collapsed": false,
     "input": [
      "def permut_assoc(rod_s, rnd, gene_df, gene_to_go, top_n, max_dist):\n",
      "    \"\"\"\n",
      "    This is the main function.\n",
      "    Use with rnd = 0 to get the real assoc.\n",
      "    \"\"\"\n",
      "    s = shift_rod(rod_s, rnd)\n",
      "    s.sort(ascending=False, inplace=True)\n",
      "    top_s = s.iloc[:top_n]\n",
      "    cand_genes = get_genes(top_s, gene_df, max_dist=max_dist)\n",
      "    assoc = get_go_assoc(cand_genes, gene_to_go)\n",
      "    return assoc\n",
      "\n",
      "def get_go_assoc(gene_ls, gene_to_go):\n",
      "    \"\"\"\n",
      "    Get series with number of genes associated with each\n",
      "    category in gene_to_go\n",
      "    \"\"\"\n",
      "    s = gene_to_go.set_index(\"gene_symbol\").ix[gene_ls].groupby(\"go_identifier\").apply(len)\n",
      "    return s\n",
      "\n",
      "def multiple_permut_assoc(rod_s, gene_df, gene_to_go, top_n, max_dist, n_runs, rnd_seed=None):\n",
      "    if rnd_seed is not None:\n",
      "        np.random.seed(rnd_seed)\n",
      "    assoc_table = pd.concat([permut_assoc(rod_s, rnd, gene_df, gene_to_go, top_n, max_dist) for rnd in np.random.rand(n_runs)],axis=1)\n",
      "    assoc_table = assoc_table.fillna(0)\n",
      "    return assoc_table\n",
      "\n",
      "def save_permut_assoc_table(assoc_table,fn):\n",
      "    assoc_table.to_csv(fn)\n",
      "\n",
      "def get_initial_rank_table(real_assoc):\n",
      "    return pd.DataFrame({\"n_genes\":real_assoc.values,\"rank\":0,\"out_of\":0},index=real_assoc.index)\n",
      "\n",
      "def get_p_val(rank_table):\n",
      "    \"\"\"\n",
      "    Input:\n",
      "    \n",
      "    \"\"\"\n",
      "    r =  1-rank_table[\"rank\"]*1./(rank_table[\"out_of\"]+1)\n",
      "    r.sort()\n",
      "    return r\n",
      "\n",
      "def update_rank(rank_table,permut_assoc_fh):\n",
      "    assoc_table = pd.read_csv(permut_assoc_fh,index_col=0)\n",
      "    r = assoc_table.apply(lambda row: get_rank(row,rank_table),axis=1)\n",
      "    return r\n",
      "    \n",
      "def total_rank(rank_table, permut_fns):\n",
      "    rt = rank_table.copy()\n",
      "    for f in permut_fns:\n",
      "        rt = update_rank(rt,permut_assoc_fh)\n",
      "    return rt\n",
      "    \n",
      "#def save_p_value_total(rank_table, permut_fns):\n",
      "    \n",
      "    \n",
      "def empirical_rank(value,dist):\n",
      "    \"\"\"\n",
      "    get empirical p value of\n",
      "    value with respect to list of \n",
      "    values in dist\n",
      "    \"\"\"\n",
      "    array = np.append(value,dist)\n",
      "    temp = array.argsort()\n",
      "    ranks = np.empty(len(array), int)\n",
      "    ranks[temp] = np.arange(len(array))\n",
      "    return ranks[0]\n",
      "\n",
      "def get_rank(series,rank_df):\n",
      "    go = series.name\n",
      "    try:\n",
      "        go_s = rank_df.ix[go]\n",
      "    except KeyError:\n",
      "        go_s = pd.Series({\"n_genes\":0,\"rank\":0,\"out_of\":0})\n",
      "    real_val = go_s[\"n_genes\"]\n",
      "    old_rank = go_s[\"rank\"]\n",
      "    old_out_of = go_s[\"out_of\"]\n",
      "    rank = empirical_rank(real_val,series.values)\n",
      "    new_rank = old_rank + rank\n",
      "    new_out_of = old_out_of + len(series)\n",
      "    return pd.Series({\"n_genes\":real_val,\"rank\":new_rank,\"out_of\":new_out_of})\n",
      "\n",
      "def get_genes_per_go(gene_ls,gene_to_go):\n",
      "    s = gene_to_go.set_index(\"gene_symbol\").ix[gene_ls].groupby(\"go_identifier\").apply(lambda x: list(x.index))\n",
      "    s.name = \"genes\"\n",
      "    return sqst"
     ],
     "language": "python",
     "metadata": {},
     "outputs": [],
     "prompt_number": 1199
    },
    {
     "cell_type": "code",
     "collapsed": false,
     "input": [
      "r = get_go_assoc(cand_genes,gene_to_go)\n",
      "rank_table = get_initial_rank_table(r)"
     ],
     "language": "python",
     "metadata": {},
     "outputs": [],
     "prompt_number": 1195
    },
    {
     "cell_type": "code",
     "collapsed": false,
     "input": [
      "genes_per_go = pd.Series(r0.index.values,index=r0.index)"
     ],
     "language": "python",
     "metadata": {},
     "outputs": [],
     "prompt_number": 1178
    },
    {
     "cell_type": "code",
     "collapsed": false,
     "input": [],
     "language": "python",
     "metadata": {},
     "outputs": [],
     "prompt_number": 1190
    },
    {
     "cell_type": "code",
     "collapsed": false,
     "input": [
      "pd.concat([rank_table,gg], axis =1)"
     ],
     "language": "python",
     "metadata": {},
     "outputs": [
      {
       "html": [
        "<div style=\"max-height:1000px;max-width:1500px;overflow:auto;\">\n",
        "<table border=\"1\" class=\"dataframe\">\n",
        "  <thead>\n",
        "    <tr style=\"text-align: right;\">\n",
        "      <th></th>\n",
        "      <th>n_genes</th>\n",
        "      <th>out_of</th>\n",
        "      <th>rank</th>\n",
        "      <th>genes</th>\n",
        "    </tr>\n",
        "    <tr>\n",
        "      <th>go_identifier</th>\n",
        "      <th></th>\n",
        "      <th></th>\n",
        "      <th></th>\n",
        "      <th></th>\n",
        "    </tr>\n",
        "  </thead>\n",
        "  <tbody>\n",
        "    <tr>\n",
        "      <th>GO:0000075</th>\n",
        "      <td>  1</td>\n",
        "      <td> 0</td>\n",
        "      <td> 0</td>\n",
        "      <td>                                          [ANAPC1]</td>\n",
        "    </tr>\n",
        "    <tr>\n",
        "      <th>GO:0000079</th>\n",
        "      <td>  1</td>\n",
        "      <td> 0</td>\n",
        "      <td> 0</td>\n",
        "      <td>                                           [HERC5]</td>\n",
        "    </tr>\n",
        "    <tr>\n",
        "      <th>GO:0000087</th>\n",
        "      <td>  1</td>\n",
        "      <td> 0</td>\n",
        "      <td> 0</td>\n",
        "      <td>                                          [ANAPC1]</td>\n",
        "    </tr>\n",
        "    <tr>\n",
        "      <th>GO:0000090</th>\n",
        "      <td>  1</td>\n",
        "      <td> 0</td>\n",
        "      <td> 0</td>\n",
        "      <td>                                          [ANAPC1]</td>\n",
        "    </tr>\n",
        "    <tr>\n",
        "      <th>GO:0000122</th>\n",
        "      <td>  2</td>\n",
        "      <td> 0</td>\n",
        "      <td> 0</td>\n",
        "      <td>                                     [TP63, TENM2]</td>\n",
        "    </tr>\n",
        "    <tr>\n",
        "      <th>GO:0000139</th>\n",
        "      <td>  2</td>\n",
        "      <td> 0</td>\n",
        "      <td> 0</td>\n",
        "      <td>                                  [HS3ST4, SCARA3]</td>\n",
        "    </tr>\n",
        "    <tr>\n",
        "      <th>GO:0000166</th>\n",
        "      <td> 24</td>\n",
        "      <td> 0</td>\n",
        "      <td> 0</td>\n",
        "      <td> [SCN8A, RAPGEF3, SYK, RFK, EPHB1, RYK, ATP13A4...</td>\n",
        "    </tr>\n",
        "    <tr>\n",
        "      <th>GO:0000185</th>\n",
        "      <td>  1</td>\n",
        "      <td> 0</td>\n",
        "      <td> 0</td>\n",
        "      <td>                                            [GRM5]</td>\n",
        "    </tr>\n",
        "    <tr>\n",
        "      <th>GO:0000187</th>\n",
        "      <td>  2</td>\n",
        "      <td> 0</td>\n",
        "      <td> 0</td>\n",
        "      <td>                                      [SYK, NTRK3]</td>\n",
        "    </tr>\n",
        "    <tr>\n",
        "      <th>GO:0000226</th>\n",
        "      <td>  1</td>\n",
        "      <td> 0</td>\n",
        "      <td> 0</td>\n",
        "      <td>                                           [PARD3]</td>\n",
        "    </tr>\n",
        "    <tr>\n",
        "      <th>GO:0000278</th>\n",
        "      <td>  1</td>\n",
        "      <td> 0</td>\n",
        "      <td> 0</td>\n",
        "      <td>                                          [ANAPC1]</td>\n",
        "    </tr>\n",
        "    <tr>\n",
        "      <th>GO:0000287</th>\n",
        "      <td>  1</td>\n",
        "      <td> 0</td>\n",
        "      <td> 0</td>\n",
        "      <td>                                          [ATP8A2]</td>\n",
        "    </tr>\n",
        "    <tr>\n",
        "      <th>GO:0000506</th>\n",
        "      <td>  1</td>\n",
        "      <td> 0</td>\n",
        "      <td> 0</td>\n",
        "      <td>                                            [PIGY]</td>\n",
        "    </tr>\n",
        "    <tr>\n",
        "      <th>GO:0000723</th>\n",
        "      <td>  1</td>\n",
        "      <td> 0</td>\n",
        "      <td> 0</td>\n",
        "      <td>                                           [XRCC5]</td>\n",
        "    </tr>\n",
        "    <tr>\n",
        "      <th>GO:0000783</th>\n",
        "      <td>  1</td>\n",
        "      <td> 0</td>\n",
        "      <td> 0</td>\n",
        "      <td>                                           [XRCC5]</td>\n",
        "    </tr>\n",
        "    <tr>\n",
        "      <th>GO:0000784</th>\n",
        "      <td>  1</td>\n",
        "      <td> 0</td>\n",
        "      <td> 0</td>\n",
        "      <td>                                           [XRCC5]</td>\n",
        "    </tr>\n",
        "    <tr>\n",
        "      <th>GO:0000785</th>\n",
        "      <td>  1</td>\n",
        "      <td> 0</td>\n",
        "      <td> 0</td>\n",
        "      <td>                                            [TP63]</td>\n",
        "    </tr>\n",
        "    <tr>\n",
        "      <th>GO:0000790</th>\n",
        "      <td>  1</td>\n",
        "      <td> 0</td>\n",
        "      <td> 0</td>\n",
        "      <td>                                            [TP63]</td>\n",
        "    </tr>\n",
        "    <tr>\n",
        "      <th>GO:0000978</th>\n",
        "      <td>  1</td>\n",
        "      <td> 0</td>\n",
        "      <td> 0</td>\n",
        "      <td>                                           [NR3C1]</td>\n",
        "    </tr>\n",
        "    <tr>\n",
        "      <th>GO:0000989</th>\n",
        "      <td>  1</td>\n",
        "      <td> 0</td>\n",
        "      <td> 0</td>\n",
        "      <td>                                            [TP63]</td>\n",
        "    </tr>\n",
        "    <tr>\n",
        "      <th>GO:0001077</th>\n",
        "      <td>  3</td>\n",
        "      <td> 0</td>\n",
        "      <td> 0</td>\n",
        "      <td>                              [TP63, NR3C1, CREB5]</td>\n",
        "    </tr>\n",
        "    <tr>\n",
        "      <th>GO:0001401</th>\n",
        "      <td>  1</td>\n",
        "      <td> 0</td>\n",
        "      <td> 0</td>\n",
        "      <td>                                          [SAMM50]</td>\n",
        "    </tr>\n",
        "    <tr>\n",
        "      <th>GO:0001501</th>\n",
        "      <td>  3</td>\n",
        "      <td> 0</td>\n",
        "      <td> 0</td>\n",
        "      <td>                                [TP63, ANKH, ARSE]</td>\n",
        "    </tr>\n",
        "    <tr>\n",
        "      <th>GO:0001508</th>\n",
        "      <td>  1</td>\n",
        "      <td> 0</td>\n",
        "      <td> 0</td>\n",
        "      <td>                                           [AKAP6]</td>\n",
        "    </tr>\n",
        "    <tr>\n",
        "      <th>GO:0001518</th>\n",
        "      <td>  1</td>\n",
        "      <td> 0</td>\n",
        "      <td> 0</td>\n",
        "      <td>                                           [SCN8A]</td>\n",
        "    </tr>\n",
        "    <tr>\n",
        "      <th>GO:0001523</th>\n",
        "      <td>  1</td>\n",
        "      <td> 0</td>\n",
        "      <td> 0</td>\n",
        "      <td>                                            [GPC5]</td>\n",
        "    </tr>\n",
        "    <tr>\n",
        "      <th>GO:0001525</th>\n",
        "      <td>  4</td>\n",
        "      <td> 0</td>\n",
        "      <td> 0</td>\n",
        "      <td>                     [RAPGEF3, SYK, EPHB1, THSD7A]</td>\n",
        "    </tr>\n",
        "    <tr>\n",
        "      <th>GO:0001539</th>\n",
        "      <td>  1</td>\n",
        "      <td> 0</td>\n",
        "      <td> 0</td>\n",
        "      <td>                                          [DNAH17]</td>\n",
        "    </tr>\n",
        "    <tr>\n",
        "      <th>GO:0001541</th>\n",
        "      <td>  1</td>\n",
        "      <td> 0</td>\n",
        "      <td> 0</td>\n",
        "      <td>                                            [FSHB]</td>\n",
        "    </tr>\n",
        "    <tr>\n",
        "      <th>GO:0001558</th>\n",
        "      <td>  1</td>\n",
        "      <td> 0</td>\n",
        "      <td> 0</td>\n",
        "      <td>                                            [NRG3]</td>\n",
        "    </tr>\n",
        "    <tr>\n",
        "      <th>...</th>\n",
        "      <td>...</td>\n",
        "      <td>...</td>\n",
        "      <td>...</td>\n",
        "      <td>...</td>\n",
        "    </tr>\n",
        "    <tr>\n",
        "      <th>GO:1900087</th>\n",
        "      <td>  1</td>\n",
        "      <td> 0</td>\n",
        "      <td> 0</td>\n",
        "      <td>                                         [PPP1R1C]</td>\n",
        "    </tr>\n",
        "    <tr>\n",
        "      <th>GO:1900170</th>\n",
        "      <td>  1</td>\n",
        "      <td> 0</td>\n",
        "      <td> 0</td>\n",
        "      <td>                                           [NR3C1]</td>\n",
        "    </tr>\n",
        "    <tr>\n",
        "      <th>GO:1900180</th>\n",
        "      <td>  1</td>\n",
        "      <td> 0</td>\n",
        "      <td> 0</td>\n",
        "      <td>                                          [RASSF5]</td>\n",
        "    </tr>\n",
        "    <tr>\n",
        "      <th>GO:1901018</th>\n",
        "      <td>  1</td>\n",
        "      <td> 0</td>\n",
        "      <td> 0</td>\n",
        "      <td>                                            [ANK2]</td>\n",
        "    </tr>\n",
        "    <tr>\n",
        "      <th>GO:1901019</th>\n",
        "      <td>  1</td>\n",
        "      <td> 0</td>\n",
        "      <td> 0</td>\n",
        "      <td>                                            [ANK2]</td>\n",
        "    </tr>\n",
        "    <tr>\n",
        "      <th>GO:1901021</th>\n",
        "      <td>  1</td>\n",
        "      <td> 0</td>\n",
        "      <td> 0</td>\n",
        "      <td>                                            [ANK2]</td>\n",
        "    </tr>\n",
        "    <tr>\n",
        "      <th>GO:1901374</th>\n",
        "      <td>  1</td>\n",
        "      <td> 0</td>\n",
        "      <td> 0</td>\n",
        "      <td>                                         [SLC22A1]</td>\n",
        "    </tr>\n",
        "    <tr>\n",
        "      <th>GO:1901379</th>\n",
        "      <td>  1</td>\n",
        "      <td> 0</td>\n",
        "      <td> 0</td>\n",
        "      <td>                                            [DPP6]</td>\n",
        "    </tr>\n",
        "    <tr>\n",
        "      <th>GO:1901381</th>\n",
        "      <td>  2</td>\n",
        "      <td> 0</td>\n",
        "      <td> 0</td>\n",
        "      <td>                                     [DPP6, AKAP6]</td>\n",
        "    </tr>\n",
        "    <tr>\n",
        "      <th>GO:1902187</th>\n",
        "      <td>  1</td>\n",
        "      <td> 0</td>\n",
        "      <td> 0</td>\n",
        "      <td>                                          [TRIM35]</td>\n",
        "    </tr>\n",
        "    <tr>\n",
        "      <th>GO:1902261</th>\n",
        "      <td>  1</td>\n",
        "      <td> 0</td>\n",
        "      <td> 0</td>\n",
        "      <td>                                           [AKAP6]</td>\n",
        "    </tr>\n",
        "    <tr>\n",
        "      <th>GO:1902287</th>\n",
        "      <td>  1</td>\n",
        "      <td> 0</td>\n",
        "      <td> 0</td>\n",
        "      <td>                                          [PLXNA4]</td>\n",
        "    </tr>\n",
        "    <tr>\n",
        "      <th>GO:1902808</th>\n",
        "      <td>  1</td>\n",
        "      <td> 0</td>\n",
        "      <td> 0</td>\n",
        "      <td>                                            [TP63]</td>\n",
        "    </tr>\n",
        "    <tr>\n",
        "      <th>GO:1990254</th>\n",
        "      <td>  1</td>\n",
        "      <td> 0</td>\n",
        "      <td> 0</td>\n",
        "      <td>                                           [KRT74]</td>\n",
        "    </tr>\n",
        "    <tr>\n",
        "      <th>GO:2000010</th>\n",
        "      <td>  1</td>\n",
        "      <td> 0</td>\n",
        "      <td> 0</td>\n",
        "      <td>                                           [GPD1L]</td>\n",
        "    </tr>\n",
        "    <tr>\n",
        "      <th>GO:2000096</th>\n",
        "      <td>  1</td>\n",
        "      <td> 0</td>\n",
        "      <td> 0</td>\n",
        "      <td>                                           [MLLT3]</td>\n",
        "    </tr>\n",
        "    <tr>\n",
        "      <th>GO:2000249</th>\n",
        "      <td>  1</td>\n",
        "      <td> 0</td>\n",
        "      <td> 0</td>\n",
        "      <td>                                         [RAPGEF3]</td>\n",
        "    </tr>\n",
        "    <tr>\n",
        "      <th>GO:2000251</th>\n",
        "      <td>  1</td>\n",
        "      <td> 0</td>\n",
        "      <td> 0</td>\n",
        "      <td>                                           [NTRK3]</td>\n",
        "    </tr>\n",
        "    <tr>\n",
        "      <th>GO:2000271</th>\n",
        "      <td>  1</td>\n",
        "      <td> 0</td>\n",
        "      <td> 0</td>\n",
        "      <td>                                            [TP63]</td>\n",
        "    </tr>\n",
        "    <tr>\n",
        "      <th>GO:2000366</th>\n",
        "      <td>  1</td>\n",
        "      <td> 0</td>\n",
        "      <td> 0</td>\n",
        "      <td>                                           [ERBB4]</td>\n",
        "    </tr>\n",
        "    <tr>\n",
        "      <th>GO:2000381</th>\n",
        "      <td>  1</td>\n",
        "      <td> 0</td>\n",
        "      <td> 0</td>\n",
        "      <td>                                            [TP63]</td>\n",
        "    </tr>\n",
        "    <tr>\n",
        "      <th>GO:2000599</th>\n",
        "      <td>  1</td>\n",
        "      <td> 0</td>\n",
        "      <td> 0</td>\n",
        "      <td>                                           [NELL1]</td>\n",
        "    </tr>\n",
        "    <tr>\n",
        "      <th>GO:2000642</th>\n",
        "      <td>  1</td>\n",
        "      <td> 0</td>\n",
        "      <td> 0</td>\n",
        "      <td>                                            [SNX3]</td>\n",
        "    </tr>\n",
        "    <tr>\n",
        "      <th>GO:2000649</th>\n",
        "      <td>  1</td>\n",
        "      <td> 0</td>\n",
        "      <td> 0</td>\n",
        "      <td>                                           [GPD1L]</td>\n",
        "    </tr>\n",
        "    <tr>\n",
        "      <th>GO:2000773</th>\n",
        "      <td>  1</td>\n",
        "      <td> 0</td>\n",
        "      <td> 0</td>\n",
        "      <td>                                            [TP63]</td>\n",
        "    </tr>\n",
        "    <tr>\n",
        "      <th>GO:2001235</th>\n",
        "      <td>  1</td>\n",
        "      <td> 0</td>\n",
        "      <td> 0</td>\n",
        "      <td>                                            [TP63]</td>\n",
        "    </tr>\n",
        "    <tr>\n",
        "      <th>GO:2001237</th>\n",
        "      <td>  1</td>\n",
        "      <td> 0</td>\n",
        "      <td> 0</td>\n",
        "      <td>                                           [ITGA6]</td>\n",
        "    </tr>\n",
        "    <tr>\n",
        "      <th>GO:2001244</th>\n",
        "      <td>  1</td>\n",
        "      <td> 0</td>\n",
        "      <td> 0</td>\n",
        "      <td>                                          [IL20RA]</td>\n",
        "    </tr>\n",
        "    <tr>\n",
        "      <th>GO:2001257</th>\n",
        "      <td>  1</td>\n",
        "      <td> 0</td>\n",
        "      <td> 0</td>\n",
        "      <td>                                            [ANK2]</td>\n",
        "    </tr>\n",
        "    <tr>\n",
        "      <th>GO:2001259</th>\n",
        "      <td>  1</td>\n",
        "      <td> 0</td>\n",
        "      <td> 0</td>\n",
        "      <td>                                            [ANK2]</td>\n",
        "    </tr>\n",
        "  </tbody>\n",
        "</table>\n",
        "<p>1265 rows \u00d7 4 columns</p>\n",
        "</div>"
       ],
       "metadata": {},
       "output_type": "pyout",
       "prompt_number": 1203,
       "text": [
        "               n_genes  out_of  rank  \\\n",
        "go_identifier                          \n",
        "GO:0000075           1       0     0   \n",
        "GO:0000079           1       0     0   \n",
        "GO:0000087           1       0     0   \n",
        "GO:0000090           1       0     0   \n",
        "GO:0000122           2       0     0   \n",
        "GO:0000139           2       0     0   \n",
        "GO:0000166          24       0     0   \n",
        "GO:0000185           1       0     0   \n",
        "GO:0000187           2       0     0   \n",
        "GO:0000226           1       0     0   \n",
        "GO:0000278           1       0     0   \n",
        "GO:0000287           1       0     0   \n",
        "GO:0000506           1       0     0   \n",
        "GO:0000723           1       0     0   \n",
        "GO:0000783           1       0     0   \n",
        "GO:0000784           1       0     0   \n",
        "GO:0000785           1       0     0   \n",
        "GO:0000790           1       0     0   \n",
        "GO:0000978           1       0     0   \n",
        "GO:0000989           1       0     0   \n",
        "GO:0001077           3       0     0   \n",
        "GO:0001401           1       0     0   \n",
        "GO:0001501           3       0     0   \n",
        "GO:0001508           1       0     0   \n",
        "GO:0001518           1       0     0   \n",
        "GO:0001523           1       0     0   \n",
        "GO:0001525           4       0     0   \n",
        "GO:0001539           1       0     0   \n",
        "GO:0001541           1       0     0   \n",
        "GO:0001558           1       0     0   \n",
        "...                ...     ...   ...   \n",
        "GO:1900087           1       0     0   \n",
        "GO:1900170           1       0     0   \n",
        "GO:1900180           1       0     0   \n",
        "GO:1901018           1       0     0   \n",
        "GO:1901019           1       0     0   \n",
        "GO:1901021           1       0     0   \n",
        "GO:1901374           1       0     0   \n",
        "GO:1901379           1       0     0   \n",
        "GO:1901381           2       0     0   \n",
        "GO:1902187           1       0     0   \n",
        "GO:1902261           1       0     0   \n",
        "GO:1902287           1       0     0   \n",
        "GO:1902808           1       0     0   \n",
        "GO:1990254           1       0     0   \n",
        "GO:2000010           1       0     0   \n",
        "GO:2000096           1       0     0   \n",
        "GO:2000249           1       0     0   \n",
        "GO:2000251           1       0     0   \n",
        "GO:2000271           1       0     0   \n",
        "GO:2000366           1       0     0   \n",
        "GO:2000381           1       0     0   \n",
        "GO:2000599           1       0     0   \n",
        "GO:2000642           1       0     0   \n",
        "GO:2000649           1       0     0   \n",
        "GO:2000773           1       0     0   \n",
        "GO:2001235           1       0     0   \n",
        "GO:2001237           1       0     0   \n",
        "GO:2001244           1       0     0   \n",
        "GO:2001257           1       0     0   \n",
        "GO:2001259           1       0     0   \n",
        "\n",
        "                                                           genes  \n",
        "go_identifier                                                     \n",
        "GO:0000075                                              [ANAPC1]  \n",
        "GO:0000079                                               [HERC5]  \n",
        "GO:0000087                                              [ANAPC1]  \n",
        "GO:0000090                                              [ANAPC1]  \n",
        "GO:0000122                                         [TP63, TENM2]  \n",
        "GO:0000139                                      [HS3ST4, SCARA3]  \n",
        "GO:0000166     [SCN8A, RAPGEF3, SYK, RFK, EPHB1, RYK, ATP13A4...  \n",
        "GO:0000185                                                [GRM5]  \n",
        "GO:0000187                                          [SYK, NTRK3]  \n",
        "GO:0000226                                               [PARD3]  \n",
        "GO:0000278                                              [ANAPC1]  \n",
        "GO:0000287                                              [ATP8A2]  \n",
        "GO:0000506                                                [PIGY]  \n",
        "GO:0000723                                               [XRCC5]  \n",
        "GO:0000783                                               [XRCC5]  \n",
        "GO:0000784                                               [XRCC5]  \n",
        "GO:0000785                                                [TP63]  \n",
        "GO:0000790                                                [TP63]  \n",
        "GO:0000978                                               [NR3C1]  \n",
        "GO:0000989                                                [TP63]  \n",
        "GO:0001077                                  [TP63, NR3C1, CREB5]  \n",
        "GO:0001401                                              [SAMM50]  \n",
        "GO:0001501                                    [TP63, ANKH, ARSE]  \n",
        "GO:0001508                                               [AKAP6]  \n",
        "GO:0001518                                               [SCN8A]  \n",
        "GO:0001523                                                [GPC5]  \n",
        "GO:0001525                         [RAPGEF3, SYK, EPHB1, THSD7A]  \n",
        "GO:0001539                                              [DNAH17]  \n",
        "GO:0001541                                                [FSHB]  \n",
        "GO:0001558                                                [NRG3]  \n",
        "...                                                          ...  \n",
        "GO:1900087                                             [PPP1R1C]  \n",
        "GO:1900170                                               [NR3C1]  \n",
        "GO:1900180                                              [RASSF5]  \n",
        "GO:1901018                                                [ANK2]  \n",
        "GO:1901019                                                [ANK2]  \n",
        "GO:1901021                                                [ANK2]  \n",
        "GO:1901374                                             [SLC22A1]  \n",
        "GO:1901379                                                [DPP6]  \n",
        "GO:1901381                                         [DPP6, AKAP6]  \n",
        "GO:1902187                                              [TRIM35]  \n",
        "GO:1902261                                               [AKAP6]  \n",
        "GO:1902287                                              [PLXNA4]  \n",
        "GO:1902808                                                [TP63]  \n",
        "GO:1990254                                               [KRT74]  \n",
        "GO:2000010                                               [GPD1L]  \n",
        "GO:2000096                                               [MLLT3]  \n",
        "GO:2000249                                             [RAPGEF3]  \n",
        "GO:2000251                                               [NTRK3]  \n",
        "GO:2000271                                                [TP63]  \n",
        "GO:2000366                                               [ERBB4]  \n",
        "GO:2000381                                                [TP63]  \n",
        "GO:2000599                                               [NELL1]  \n",
        "GO:2000642                                                [SNX3]  \n",
        "GO:2000649                                               [GPD1L]  \n",
        "GO:2000773                                                [TP63]  \n",
        "GO:2001235                                                [TP63]  \n",
        "GO:2001237                                               [ITGA6]  \n",
        "GO:2001244                                              [IL20RA]  \n",
        "GO:2001257                                                [ANK2]  \n",
        "GO:2001259                                                [ANK2]  \n",
        "\n",
        "[1265 rows x 4 columns]"
       ]
      }
     ],
     "prompt_number": 1203
    },
    {
     "cell_type": "code",
     "collapsed": false,
     "input": [
      "gg = get_genes_per_go(cand_genes,gene_to_go)"
     ],
     "language": "python",
     "metadata": {},
     "outputs": [],
     "prompt_number": 1202
    },
    {
     "cell_type": "code",
     "collapsed": false,
     "input": [
      "r = permut_assoc(tot_sweed_df[\"Likelihood\"], 0.1, gene_df, gene_to_go, top_n = 500, max_dist=10000)"
     ],
     "language": "python",
     "metadata": {},
     "outputs": [],
     "prompt_number": 927
    },
    {
     "cell_type": "code",
     "collapsed": false,
     "input": [
      "real_assoc = permut_assoc(tot_sweed_df[\"Likelihood\"], 0, gene_df, gene_to_go, top_n=1000, max_dist=10000)"
     ],
     "language": "python",
     "metadata": {},
     "outputs": [],
     "prompt_number": 934
    },
    {
     "cell_type": "code",
     "collapsed": false,
     "input": [
      "assoc_table = multiple_permut_assoc(tot_sweed_df[\"Likelihood\"], gene_df, gene_to_go, top_n=1000, max_dist = 10000, n_runs=100, rnd_seed=None)"
     ],
     "language": "python",
     "metadata": {},
     "outputs": [],
     "prompt_number": 933
    },
    {
     "cell_type": "code",
     "collapsed": false,
     "input": [
      "def f(r):\n",
      "    print r\n",
      "    raise"
     ],
     "language": "python",
     "metadata": {},
     "outputs": [],
     "prompt_number": 952
    },
    {
     "cell_type": "code",
     "collapsed": false,
     "input": [
      "p_vals = assoc_table.apply(lambda r: get_p_val(r,real_assoc),axis=1)"
     ],
     "language": "python",
     "metadata": {},
     "outputs": [],
     "prompt_number": 959
    },
    {
     "cell_type": "code",
     "collapsed": false,
     "input": [
      "p_vals.sort()"
     ],
     "language": "python",
     "metadata": {},
     "outputs": [],
     "prompt_number": 960
    },
    {
     "cell_type": "code",
     "collapsed": false,
     "input": [
      "sab_enrich = pd.read_csv(eu(\"~/vervet_project/analyses/20140611_163_subpop_sweepfinder/_data/enrichment_n2000_d10000_permut50000_sab.tsv\"),\n",
      "                            index_col=[0],sep=\"\\t\")"
     ],
     "language": "python",
     "metadata": {},
     "outputs": [],
     "prompt_number": 1206
    },
    {
     "cell_type": "code",
     "collapsed": false,
     "input": [
      "len(gene_to_go.set_index(\"go_identifier\").index.diff(sab_enrich.index))"
     ],
     "language": "python",
     "metadata": {},
     "outputs": [
      {
       "metadata": {},
       "output_type": "pyout",
       "prompt_number": 1218,
       "text": [
        "432"
       ]
      }
     ],
     "prompt_number": 1218
    },
    {
     "cell_type": "code",
     "collapsed": false,
     "input": [
      "len(sab_enrich)"
     ],
     "language": "python",
     "metadata": {},
     "outputs": [
      {
       "metadata": {},
       "output_type": "pyout",
       "prompt_number": 1220,
       "text": [
        "14566"
       ]
      }
     ],
     "prompt_number": 1220
    },
    {
     "cell_type": "code",
     "collapsed": false,
     "input": [
      "len(gene_to_go.set_index(\"go_identifier\").index.unique())"
     ],
     "language": "python",
     "metadata": {},
     "outputs": [
      {
       "metadata": {},
       "output_type": "pyout",
       "prompt_number": 1222,
       "text": [
        "14998"
       ]
      }
     ],
     "prompt_number": 1222
    },
    {
     "cell_type": "code",
     "collapsed": false,
     "input": [],
     "language": "python",
     "metadata": {},
     "outputs": [],
     "prompt_number": 1223
    },
    {
     "cell_type": "heading",
     "level": 2,
     "metadata": {},
     "source": [
      "Development for test_enrichment.py\n"
     ]
    },
    {
     "cell_type": "code",
     "collapsed": false,
     "input": [
      "from hs_vervet.scripts import test_enrichment as te"
     ],
     "language": "python",
     "metadata": {},
     "outputs": [],
     "prompt_number": 4
    },
    {
     "cell_type": "code",
     "collapsed": false,
     "input": [
      "from scipy import stats"
     ],
     "language": "python",
     "metadata": {},
     "outputs": [],
     "prompt_number": 10
    },
    {
     "cell_type": "code",
     "collapsed": false,
     "input": [
      "pop = \"aet\"\n",
      "tot_sweed_df = pd.read_csv(eu(\"~/vervet_project/analyses/20140611_163_subpop_sweepfinder/\"\n",
      "                                 \"_data/tot_sweed_result_{}.tsv\".format(pop)),sep=\"\\t\",index_col=[0,1])"
     ],
     "language": "python",
     "metadata": {},
     "outputs": []
    },
    {
     "cell_type": "code",
     "collapsed": false,
     "input": [
      "def gene_likelihood(outlier_locations,mode=\"max\"):\n",
      "    \"\"\"\n",
      "    get genes that overlap with the multi-index of the provided data frame\n",
      "    \"\"\"\n",
      "    modes = [\"max\",\"sum\"]\n",
      "    assert mode in modes\n",
      "    gene_hit_df = pd.DataFrame()\n",
      "    for chrom in outlier_locations.index.droplevel(1).unique():\n",
      "        pos_rel_to_start = gene_df.ix[chrom].index.searchsorted(outlier_locations.ix[chrom].index).values\n",
      "        pos_rel_to_end = np.searchsorted(gene_df.ix[chrom][\"end\"].values,outlier_locations.ix[chrom].index.values)\n",
      "        in_gene = (pos_rel_to_start - pos_rel_to_end) == 1\n",
      "        gene_hits = gene_df.ix[chrom].iloc[pos_rel_to_end[in_gene]]\n",
      "        gene_hits['chrom'] = chrom\n",
      "        gene_hits.set_index('chrom', append=True, inplace=True)\n",
      "        gene_hits = gene_hits.reorder_levels(['chrom', 'pos'])\n",
      "        gene_hits[\"likelihood\"] = outlier_locations.ix[chrom][in_gene][\"Likelihood\"].values\n",
      "        gene_hit_df = gene_hit_df.append(gene_hits)\n",
      "    #if mode == \"max\":    \n",
      "    return gene_hit_df.groupby(lambda x:x).max()\n",
      "    #elif mode == \"sum\":\n",
      "    #    return gene_hit_df.groupby(lambda x:x).sum()"
     ],
     "language": "python",
     "metadata": {},
     "outputs": [],
     "prompt_number": 79
    },
    {
     "cell_type": "code",
     "collapsed": false,
     "input": [
      "def gene_stats(rod_s,stats=None,stat_names=None):\n",
      "    \"\"\"\n",
      "    cacluclate statistics on the ROD series across genes\n",
      "    stats default: [np.max,np.nanmean,len]\n",
      "    \"\"\"\n",
      "    if stats == None:\n",
      "        stats = [np.max,np.nanmean,len]\n",
      "    def stats_apply(df):\n",
      "        #name = df.index.values[0]\n",
      "        c_dic = {\"gene_id\":df.iloc[0][\"gene_id\"],\"end\":df.iloc[0][\"end\"]}\n",
      "        if stat_names is None:\n",
      "            c_dic.update({stat.__name__:stat(df[\"value\"]) for stat in stats})\n",
      "        else:\n",
      "            c_dic.update({n:stat(df[\"value\"]) for n,stat in zip(stat_name,stats)})\n",
      "        s = pd.Series(c_dic)\n",
      "        return s\n",
      "    gene_hit_df = pd.DataFrame()\n",
      "    for chrom in rod_s.index.droplevel(1).unique():\n",
      "        pos_rel_to_start = gene_df.ix[chrom].index.searchsorted(rod_s.ix[chrom].index).values\n",
      "        pos_rel_to_end = np.searchsorted(gene_df.ix[chrom][\"end\"].values,rod_s.ix[chrom].index.values)\n",
      "        in_gene = (pos_rel_to_start - pos_rel_to_end) == 1\n",
      "        gene_hits = gene_df.ix[chrom].iloc[pos_rel_to_end[in_gene]]\n",
      "        gene_hits['chrom'] = chrom\n",
      "        gene_hits.set_index('chrom', append=True, inplace=True)\n",
      "        gene_hits = gene_hits.reorder_levels(['chrom', 'pos'])\n",
      "        gene_hits[\"value\"] = rod_s.ix[chrom][in_gene].values\n",
      "        gene_hits = gene_hits.groupby(lambda x:x).apply(stats_apply)\n",
      "        gene_hit_df = gene_hit_df.append(gene_hits)   \n",
      "    gene_hit_df.index = pd.MultiIndex.from_tuples(gene_hit_df.index)\n",
      "    return gene_hit_df"
     ],
     "language": "python",
     "metadata": {},
     "outputs": [],
     "prompt_number": 197
    },
    {
     "cell_type": "code",
     "collapsed": false,
     "input": [
      "def gene_stats_gb(rod_s,stats=None,stat_names=None):\n",
      "    \"\"\"\n",
      "    #implement avoiding loop\n",
      "    cacluclate statistics on the ROD series across genes\n",
      "    stats default: [np.max,np.nanmean,len]\n",
      "    \"\"\"\n",
      "    if stats == None:\n",
      "        stats = [np.max,np.nanmean,len]\n",
      "    def stats_apply(df):\n",
      "        #name = df.index.values[0]\n",
      "        c_dic = {\"gene_id\":df.iloc[0][\"gene_id\"],\"end\":df.iloc[0][\"end\"]}\n",
      "        if stat_names is None:\n",
      "            c_dic.update({stat.__name__:stat(df[\"value\"]) for stat in stats})\n",
      "        else:\n",
      "            c_dic.update({n:stat(df[\"value\"]) for n,stat in zip(stat_name,stats)})\n",
      "        s = pd.Series(c_dic)\n",
      "        return s\n",
      "    def get_gene_hits(rod_s):\n",
      "        #print rod_s.shape\n",
      "        chrom = rod_s.index.values[0][0]\n",
      "        #print chrom\n",
      "        pos_rel_to_start = gene_df.ix[chrom].index.searchsorted(rod_s.ix[chrom].index).values\n",
      "        pos_rel_to_end = np.searchsorted(gene_df.ix[chrom][\"end\"].values,rod_s.ix[chrom].index.values)\n",
      "        in_gene = (pos_rel_to_start - pos_rel_to_end) == 1\n",
      "        gene_hits = gene_df.ix[chrom].iloc[pos_rel_to_end[in_gene]]\n",
      "        gene_hits[\"value\"] = rod_s.ix[chrom][in_gene].values\n",
      "        gene_hits = gene_hits.groupby(lambda x:x).apply(stats_apply)\n",
      "        return gene_hits   \n",
      "    gene_hit_df = rod_s.groupby(lambda x: x[0]).apply(get_gene_hits)\n",
      "    return gene_hit_df"
     ],
     "language": "python",
     "metadata": {},
     "outputs": [],
     "prompt_number": 205
    },
    {
     "cell_type": "code",
     "collapsed": false,
     "input": [
      "tot_sweed_s = tot_sweed_df[\"Likelihood\"]"
     ],
     "language": "python",
     "metadata": {},
     "outputs": [],
     "prompt_number": 210
    },
    {
     "cell_type": "code",
     "collapsed": false,
     "input": [
      "tot_sweed_s"
     ],
     "language": "python",
     "metadata": {},
     "outputs": [
      {
       "metadata": {},
       "output_type": "pyout",
       "prompt_number": 211,
       "text": [
        "chrom  pos      \n",
        "CAE1   1161.0000    0.000000\n",
        "       1660.9969    0.000000\n",
        "       2160.9937    0.000000\n",
        "       2660.9906    0.000000\n",
        "       3160.9874    0.000000\n",
        "       3660.9843    0.038328\n",
        "       4160.9811    0.854687\n",
        "       4660.9780    0.109911\n",
        "       5160.9749    0.047829\n",
        "       5660.9717    0.000000\n",
        "       6160.9686    0.077996\n",
        "       6660.9654    0.000003\n",
        "       7160.9623    0.000000\n",
        "       7660.9591    0.004962\n",
        "       8160.9560    0.000000\n",
        "...\n",
        "CAEY   6174000.5120    0\n",
        "       6174499.0588    0\n",
        "       6174997.6057    0\n",
        "       6175496.1525    0\n",
        "       6175994.6993    0\n",
        "       6176493.2461    0\n",
        "       6176991.7929    0\n",
        "       6177490.3397    0\n",
        "       6177988.8865    0\n",
        "       6178487.4333    0\n",
        "       6178985.9802    0\n",
        "       6179484.5270    0\n",
        "       6179983.0738    0\n",
        "       6180481.6206    0\n",
        "       6180980.0000    0\n",
        "Name: Likelihood, Length: 5488230, dtype: float64"
       ]
      }
     ],
     "prompt_number": 211
    },
    {
     "cell_type": "code",
     "collapsed": false,
     "input": [
      "tot_sweed_s = tot_sweed_s.sortlevel(level=[0,1])"
     ],
     "language": "python",
     "metadata": {},
     "outputs": [],
     "prompt_number": 215
    },
    {
     "cell_type": "code",
     "collapsed": false,
     "input": [
      "tot_sweed_s.index.is_monotonic"
     ],
     "language": "python",
     "metadata": {},
     "outputs": [
      {
       "metadata": {},
       "output_type": "pyout",
       "prompt_number": 216,
       "text": [
        "True"
       ]
      }
     ],
     "prompt_number": 216
    },
    {
     "cell_type": "code",
     "collapsed": false,
     "input": [
      "tot_sweed_s"
     ],
     "language": "python",
     "metadata": {},
     "outputs": [
      {
       "metadata": {},
       "output_type": "pyout",
       "prompt_number": 220,
       "text": [
        "chrom  pos      \n",
        "CAE1   1161.0000    0.000000\n",
        "       1660.9969    0.000000\n",
        "       2160.9937    0.000000\n",
        "       2660.9906    0.000000\n",
        "       3160.9874    0.000000\n",
        "       3660.9843    0.038328\n",
        "       4160.9811    0.854687\n",
        "       4660.9780    0.109911\n",
        "       5160.9749    0.047829\n",
        "       5660.9717    0.000000\n",
        "       6160.9686    0.077996\n",
        "       6660.9654    0.000003\n",
        "       7160.9623    0.000000\n",
        "       7660.9591    0.004962\n",
        "       8160.9560    0.000000\n",
        "...\n",
        "CAEY   6174000.5120    0\n",
        "       6174499.0588    0\n",
        "       6174997.6057    0\n",
        "       6175496.1525    0\n",
        "       6175994.6993    0\n",
        "       6176493.2461    0\n",
        "       6176991.7929    0\n",
        "       6177490.3397    0\n",
        "       6177988.8865    0\n",
        "       6178487.4333    0\n",
        "       6178985.9802    0\n",
        "       6179484.5270    0\n",
        "       6179983.0738    0\n",
        "       6180481.6206    0\n",
        "       6180980.0000    0\n",
        "Name: Likelihood, Length: 5488230, dtype: float64"
       ]
      }
     ],
     "prompt_number": 220
    },
    {
     "cell_type": "code",
     "collapsed": false,
     "input": [
      "t_s = tot_sweed_s.iloc[::10]"
     ],
     "language": "python",
     "metadata": {},
     "outputs": [],
     "prompt_number": 217
    },
    {
     "cell_type": "code",
     "collapsed": false,
     "input": [
      "t_s.ix[\"CAE2\"]"
     ],
     "language": "python",
     "metadata": {},
     "outputs": [
      {
       "metadata": {},
       "output_type": "pyout",
       "prompt_number": 219,
       "text": [
        "pos\n",
        "4098.9189     0.000000\n",
        "9098.5137     0.000000\n",
        "14098.1084    0.000000\n",
        "19097.7031    0.063362\n",
        "24097.2979    0.098169\n",
        "29096.8926    0.000959\n",
        "34096.4873    0.632552\n",
        "39096.0820    0.000197\n",
        "44095.6768    0.000000\n",
        "49095.2715    0.000000\n",
        "54094.8662    0.000000\n",
        "59094.4609    0.000000\n",
        "64094.0557    0.000058\n",
        "69093.6504    2.168776\n",
        "74093.2451    0.000000\n",
        "...\n",
        "90296779.6807    0.000379\n",
        "90301779.2754    0.000000\n",
        "90306778.8701    0.000000\n",
        "90311778.4648    0.000000\n",
        "90316778.0596    0.000000\n",
        "90321777.6543    0.000000\n",
        "90326777.2490    0.000000\n",
        "90331776.8438    0.000000\n",
        "90336776.4385    0.000000\n",
        "90341776.0332    0.000000\n",
        "90346775.6279    0.000000\n",
        "90351775.2227    0.000000\n",
        "90356774.8174    0.000000\n",
        "90361774.4121    0.000000\n",
        "90366774.0068    0.000000\n",
        "Name: Likelihood, Length: 18075, dtype: float64"
       ]
      }
     ],
     "prompt_number": 219
    },
    {
     "cell_type": "code",
     "collapsed": false,
     "input": [
      "t_df.index.droplevel(1).unique()"
     ],
     "language": "python",
     "metadata": {},
     "outputs": [
      {
       "metadata": {},
       "output_type": "pyout",
       "prompt_number": 206,
       "text": [
        "array(['CAE1', 'CAE2', 'CAE3', 'CAE4', 'CAE5', 'CAE6', 'CAE7', 'CAE8',\n",
        "       'CAE9', 'CAE10', 'CAE11', 'CAE12', 'CAE13', 'CAE14', 'CAE15',\n",
        "       'CAE16', 'CAE17', 'CAE18', 'CAE19', 'CAE20', 'CAE21', 'CAE22',\n",
        "       'CAE23', 'CAE24', 'CAE25', 'CAE26', 'CAE27', 'CAE28', 'CAE29',\n",
        "       'CAEX', 'CAEY'], dtype=object)"
       ]
      }
     ],
     "prompt_number": 206
    },
    {
     "cell_type": "code",
     "collapsed": false,
     "input": [
      "t_df.ix[\"CAE2\"]"
     ],
     "language": "python",
     "metadata": {},
     "outputs": [
      {
       "html": [
        "<div style=\"max-height:1000px;max-width:1500px;overflow:auto;\">\n",
        "<table border=\"1\" class=\"dataframe\">\n",
        "  <thead>\n",
        "    <tr style=\"text-align: right;\">\n",
        "      <th></th>\n",
        "      <th>Likelihood</th>\n",
        "      <th>Alpha</th>\n",
        "    </tr>\n",
        "    <tr>\n",
        "      <th>pos</th>\n",
        "      <th></th>\n",
        "      <th></th>\n",
        "    </tr>\n",
        "  </thead>\n",
        "  <tbody>\n",
        "  </tbody>\n",
        "</table>\n",
        "</div>"
       ],
       "metadata": {},
       "output_type": "pyout",
       "prompt_number": 207,
       "text": [
        "Empty DataFrame\n",
        "Columns: [Likelihood, Alpha]\n",
        "Index: []"
       ]
      }
     ],
     "prompt_number": 207
    },
    {
     "cell_type": "code",
     "collapsed": false,
     "input": [
      "len(t_df.ix[\"CAE2\"])"
     ],
     "language": "python",
     "metadata": {},
     "outputs": [
      {
       "metadata": {},
       "output_type": "pyout",
       "prompt_number": 193,
       "text": [
        "0"
       ]
      }
     ],
     "prompt_number": 193
    },
    {
     "cell_type": "code",
     "collapsed": false,
     "input": [
      "t_s = t_df[\"Likelihood\"].iloc[::10]"
     ],
     "language": "python",
     "metadata": {},
     "outputs": [],
     "prompt_number": 208
    },
    {
     "cell_type": "code",
     "collapsed": false,
     "input": [
      "t_s.ix[\"CAE2\"]"
     ],
     "language": "python",
     "metadata": {},
     "outputs": [
      {
       "metadata": {},
       "output_type": "pyout",
       "prompt_number": 209,
       "text": [
        "Series([], name: Likelihood, dtype: float64)"
       ]
      }
     ],
     "prompt_number": 209
    },
    {
     "cell_type": "code",
     "collapsed": false,
     "input": [
      "%timeit gene_likely = gene_stats(t_s,stats=[np.max])"
     ],
     "language": "python",
     "metadata": {},
     "outputs": [
      {
       "output_type": "stream",
       "stream": "stdout",
       "text": [
        "1 loops, best of 3: 14.6 s per loop\n"
       ]
      }
     ],
     "prompt_number": 199
    },
    {
     "cell_type": "code",
     "collapsed": false,
     "input": [
      "%timeit gene_likely_gb = gene_stats_gb(t_s,stats=[np.max])"
     ],
     "language": "python",
     "metadata": {},
     "outputs": [
      {
       "output_type": "stream",
       "stream": "stdout",
       "text": [
        "(548823,)\n",
        "(548823,)"
       ]
      },
      {
       "output_type": "stream",
       "stream": "stdout",
       "text": [
        "\n",
        "(548823,)"
       ]
      },
      {
       "output_type": "stream",
       "stream": "stdout",
       "text": [
        "\n",
        "(548823,)"
       ]
      },
      {
       "output_type": "stream",
       "stream": "stdout",
       "text": [
        "\n",
        "1 loops, best of 3: 13.5 s per loop"
       ]
      },
      {
       "output_type": "stream",
       "stream": "stdout",
       "text": [
        "\n"
       ]
      }
     ],
     "prompt_number": 200
    },
    {
     "cell_type": "code",
     "collapsed": false,
     "input": [
      "%timeit gene_likely_gb = gene_stats_gb(t_s,stats=[np.max])"
     ],
     "language": "python",
     "metadata": {},
     "outputs": [
      {
       "output_type": "stream",
       "stream": "stdout",
       "text": [
        "1 loops, best of 3: 16.2 s per loop\n"
       ]
      }
     ],
     "prompt_number": 204
    },
    {
     "cell_type": "code",
     "collapsed": false,
     "input": [
      "len(gene_likely_gb[\"gene_id\"].unique())"
     ],
     "language": "python",
     "metadata": {},
     "outputs": [
      {
       "metadata": {},
       "output_type": "pyout",
       "prompt_number": 136,
       "text": [
        "9200"
       ]
      }
     ],
     "prompt_number": 136
    },
    {
     "cell_type": "code",
     "collapsed": false,
     "input": [
      "%timeit gene_likely = gene_stats(t_s,stats=[np.max])"
     ],
     "language": "python",
     "metadata": {},
     "outputs": [
      {
       "output_type": "stream",
       "stream": "stdout",
       "text": [
        "1 loops, best of 3: 31.8 s per loop\n"
       ]
      }
     ],
     "prompt_number": 98
    },
    {
     "cell_type": "code",
     "collapsed": false,
     "input": [
      "gene_likely = gene_stats(t_s,stats=[np.max])"
     ],
     "language": "python",
     "metadata": {},
     "outputs": [],
     "prompt_number": 92
    },
    {
     "cell_type": "code",
     "collapsed": false,
     "input": [
      "%timeit gene_likelihood(t_df,mode=\"max\")"
     ],
     "language": "python",
     "metadata": {},
     "outputs": [
      {
       "output_type": "stream",
       "stream": "stdout",
       "text": [
        "1 loops, best of 3: 35.6 s per loop\n"
       ]
      }
     ],
     "prompt_number": 99
    },
    {
     "cell_type": "code",
     "collapsed": false,
     "input": [
      "gene_likely.sort(\"amax\",ascending=False)"
     ],
     "language": "python",
     "metadata": {},
     "outputs": [
      {
       "html": [
        "<div style=\"max-height:1000px;max-width:1500px;overflow:auto;\">\n",
        "<table border=\"1\" class=\"dataframe\">\n",
        "  <thead>\n",
        "    <tr style=\"text-align: right;\">\n",
        "      <th></th>\n",
        "      <th>amax</th>\n",
        "      <th>end</th>\n",
        "      <th>gene_id</th>\n",
        "    </tr>\n",
        "  </thead>\n",
        "  <tbody>\n",
        "    <tr>\n",
        "      <th>(CAE3, 4747744)</th>\n",
        "      <td> 48.61760</td>\n",
        "      <td>   5378598</td>\n",
        "      <td>       ATP8A2</td>\n",
        "    </tr>\n",
        "    <tr>\n",
        "      <th>(CAE1, 47494862)</th>\n",
        "      <td> 36.05660</td>\n",
        "      <td>  47578818</td>\n",
        "      <td>        ABCC8</td>\n",
        "    </tr>\n",
        "    <tr>\n",
        "      <th>(CAE1, 117553388)</th>\n",
        "      <td> 32.38441</td>\n",
        "      <td> 118122895</td>\n",
        "      <td>      KIRREL3</td>\n",
        "    </tr>\n",
        "    <tr>\n",
        "      <th>(CAE1, 42201546)</th>\n",
        "      <td> 30.17795</td>\n",
        "      <td>  42335632</td>\n",
        "      <td>         GAS2</td>\n",
        "    </tr>\n",
        "    <tr>\n",
        "      <th>(CAE1, 79636636)</th>\n",
        "      <td> 29.88969</td>\n",
        "      <td>  80188349</td>\n",
        "      <td>         GRM5</td>\n",
        "    </tr>\n",
        "    <tr>\n",
        "      <th>(CAE1, 43416052)</th>\n",
        "      <td> 28.28739</td>\n",
        "      <td>  44335894</td>\n",
        "      <td>        NELL1</td>\n",
        "    </tr>\n",
        "    <tr>\n",
        "      <th>(CAE3, 8099783)</th>\n",
        "      <td> 25.72230</td>\n",
        "      <td>   8519645</td>\n",
        "      <td>        MTUS2</td>\n",
        "    </tr>\n",
        "    <tr>\n",
        "      <th>(CAE3, 3770901)</th>\n",
        "      <td> 24.93997</td>\n",
        "      <td>   4085379</td>\n",
        "      <td>      SPATA13</td>\n",
        "    </tr>\n",
        "    <tr>\n",
        "      <th>(CAE1, 96365646)</th>\n",
        "      <td> 22.69252</td>\n",
        "      <td>  96388966</td>\n",
        "      <td> LOC103248366</td>\n",
        "    </tr>\n",
        "    <tr>\n",
        "      <th>(CAE2, 20613352)</th>\n",
        "      <td> 21.20587</td>\n",
        "      <td>  21720417</td>\n",
        "      <td>        PTPRT</td>\n",
        "    </tr>\n",
        "    <tr>\n",
        "      <th>(CAE1, 66666049)</th>\n",
        "      <td> 20.69260</td>\n",
        "      <td>  66707160</td>\n",
        "      <td>        GDPD5</td>\n",
        "    </tr>\n",
        "    <tr>\n",
        "      <th>(CAE2, 66037224)</th>\n",
        "      <td> 20.34533</td>\n",
        "      <td>  66320860</td>\n",
        "      <td>          APP</td>\n",
        "    </tr>\n",
        "    <tr>\n",
        "      <th>(CAE3, 20233323)</th>\n",
        "      <td> 18.98894</td>\n",
        "      <td>  20408490</td>\n",
        "      <td>         DGKH</td>\n",
        "    </tr>\n",
        "    <tr>\n",
        "      <th>(CAE1, 17137896)</th>\n",
        "      <td> 18.35761</td>\n",
        "      <td>  17313787</td>\n",
        "      <td>        PTPRJ</td>\n",
        "    </tr>\n",
        "    <tr>\n",
        "      <th>(CAE1, 115768337)</th>\n",
        "      <td> 16.86788</td>\n",
        "      <td> 115778084</td>\n",
        "      <td>        TBRG1</td>\n",
        "    </tr>\n",
        "    <tr>\n",
        "      <th>(CAE1, 46956665)</th>\n",
        "      <td> 16.70152</td>\n",
        "      <td>  47099546</td>\n",
        "      <td>       SERGEF</td>\n",
        "    </tr>\n",
        "    <tr>\n",
        "      <th>(CAE3, 19767190)</th>\n",
        "      <td> 16.06920</td>\n",
        "      <td>  20110650</td>\n",
        "      <td>         VWA8</td>\n",
        "    </tr>\n",
        "    <tr>\n",
        "      <th>(CAE3, 23840947)</th>\n",
        "      <td> 15.84178</td>\n",
        "      <td>  23909954</td>\n",
        "      <td>        SIAH3</td>\n",
        "    </tr>\n",
        "    <tr>\n",
        "      <th>(CAE1, 67894124)</th>\n",
        "      <td> 15.70431</td>\n",
        "      <td>  67926732</td>\n",
        "      <td> LOC103248083</td>\n",
        "    </tr>\n",
        "    <tr>\n",
        "      <th>(CAE1, 94484090)</th>\n",
        "      <td> 15.15998</td>\n",
        "      <td>  94849960</td>\n",
        "      <td>      DYNC2H1</td>\n",
        "    </tr>\n",
        "    <tr>\n",
        "      <th>(CAE1, 31627447)</th>\n",
        "      <td> 14.97001</td>\n",
        "      <td>  31925520</td>\n",
        "      <td>    KIAA1549L</td>\n",
        "    </tr>\n",
        "    <tr>\n",
        "      <th>(CAE3, 568068)</th>\n",
        "      <td> 14.61980</td>\n",
        "      <td>    690030</td>\n",
        "      <td>        CRYL1</td>\n",
        "    </tr>\n",
        "    <tr>\n",
        "      <th>(CAE1, 34943544)</th>\n",
        "      <td> 14.55943</td>\n",
        "      <td>  34946866</td>\n",
        "      <td>         FSHB</td>\n",
        "    </tr>\n",
        "    <tr>\n",
        "      <th>(CAE1, 96284010)</th>\n",
        "      <td> 14.19321</td>\n",
        "      <td>  96298525</td>\n",
        "      <td> LOC103248368</td>\n",
        "    </tr>\n",
        "    <tr>\n",
        "      <th>(CAE1, 65282311)</th>\n",
        "      <td> 14.15360</td>\n",
        "      <td>  65436007</td>\n",
        "      <td>        C2CD3</td>\n",
        "    </tr>\n",
        "    <tr>\n",
        "      <th>(CAE3, 24820029)</th>\n",
        "      <td> 13.90847</td>\n",
        "      <td>  24882444</td>\n",
        "      <td>        HTR2A</td>\n",
        "    </tr>\n",
        "    <tr>\n",
        "      <th>(CAE1, 117479328)</th>\n",
        "      <td> 13.89792</td>\n",
        "      <td> 117543262</td>\n",
        "      <td>      ST3GAL4</td>\n",
        "    </tr>\n",
        "    <tr>\n",
        "      <th>(CAE3, 20988638)</th>\n",
        "      <td> 13.15384</td>\n",
        "      <td>  21128966</td>\n",
        "      <td>       EPSTI1</td>\n",
        "    </tr>\n",
        "    <tr>\n",
        "      <th>(CAE2, 17233287)</th>\n",
        "      <td> 13.12974</td>\n",
        "      <td>  17324761</td>\n",
        "      <td>      SLC13A3</td>\n",
        "    </tr>\n",
        "    <tr>\n",
        "      <th>(CAE3, 28791491)</th>\n",
        "      <td> 13.04286</td>\n",
        "      <td>  28841963</td>\n",
        "      <td>     RNASEH2B</td>\n",
        "    </tr>\n",
        "    <tr>\n",
        "      <th>...</th>\n",
        "      <td>...</td>\n",
        "      <td>...</td>\n",
        "      <td>...</td>\n",
        "    </tr>\n",
        "    <tr>\n",
        "      <th>(CAE1, 81057390)</th>\n",
        "      <td>  0.00000</td>\n",
        "      <td>  81058134</td>\n",
        "      <td> LOC103248206</td>\n",
        "    </tr>\n",
        "    <tr>\n",
        "      <th>(CAE1, 15582781)</th>\n",
        "      <td>  0.00000</td>\n",
        "      <td>  15584988</td>\n",
        "      <td>        YPEL4</td>\n",
        "    </tr>\n",
        "    <tr>\n",
        "      <th>(CAE1, 62536922)</th>\n",
        "      <td>  0.00000</td>\n",
        "      <td>  62538821</td>\n",
        "      <td> LOC103247969</td>\n",
        "    </tr>\n",
        "    <tr>\n",
        "      <th>(CAE1, 77180429)</th>\n",
        "      <td>  0.00000</td>\n",
        "      <td>  77181144</td>\n",
        "      <td> LOC103248157</td>\n",
        "    </tr>\n",
        "    <tr>\n",
        "      <th>(CAE2, 88710182)</th>\n",
        "      <td>  0.00000</td>\n",
        "      <td>  88712024</td>\n",
        "      <td> LOC103220067</td>\n",
        "    </tr>\n",
        "    <tr>\n",
        "      <th>(CAE2, 88777194)</th>\n",
        "      <td>  0.00000</td>\n",
        "      <td>  88779026</td>\n",
        "      <td> LOC103220073</td>\n",
        "    </tr>\n",
        "    <tr>\n",
        "      <th>(CAE1, 80929496)</th>\n",
        "      <td>  0.00000</td>\n",
        "      <td>  80931691</td>\n",
        "      <td> LOC103248199</td>\n",
        "    </tr>\n",
        "    <tr>\n",
        "      <th>(CAE2, 56617784)</th>\n",
        "      <td>  0.00000</td>\n",
        "      <td>  56619136</td>\n",
        "      <td>        ROMO1</td>\n",
        "    </tr>\n",
        "    <tr>\n",
        "      <th>(CAE2, 47855831)</th>\n",
        "      <td>  0.00000</td>\n",
        "      <td>  47856625</td>\n",
        "      <td>       SCP2D1</td>\n",
        "    </tr>\n",
        "    <tr>\n",
        "      <th>(CAE1, 80981885)</th>\n",
        "      <td>  0.00000</td>\n",
        "      <td>  80982678</td>\n",
        "      <td> LOC103248201</td>\n",
        "    </tr>\n",
        "    <tr>\n",
        "      <th>(CAE1, 26211019)</th>\n",
        "      <td>  0.00000</td>\n",
        "      <td>  26213633</td>\n",
        "      <td> LOC103237427</td>\n",
        "    </tr>\n",
        "    <tr>\n",
        "      <th>(CAE2, 55307268)</th>\n",
        "      <td>  0.00000</td>\n",
        "      <td>  55335200</td>\n",
        "      <td>      RAD21L1</td>\n",
        "    </tr>\n",
        "    <tr>\n",
        "      <th>(CAE1, 81032860)</th>\n",
        "      <td>  0.00000</td>\n",
        "      <td>  81038456</td>\n",
        "      <td>       TRIM51</td>\n",
        "    </tr>\n",
        "    <tr>\n",
        "      <th>(CAE2, 17380060)</th>\n",
        "      <td>  0.00000</td>\n",
        "      <td>  17388798</td>\n",
        "      <td> LOC103214199</td>\n",
        "    </tr>\n",
        "    <tr>\n",
        "      <th>(CAE1, 81058982)</th>\n",
        "      <td>  0.00000</td>\n",
        "      <td>  81065150</td>\n",
        "      <td> LOC103248207</td>\n",
        "    </tr>\n",
        "    <tr>\n",
        "      <th>(CAE1, 62075960)</th>\n",
        "      <td>  0.00000</td>\n",
        "      <td>  62077427</td>\n",
        "      <td> LOC103247939</td>\n",
        "    </tr>\n",
        "    <tr>\n",
        "      <th>(CAE2, 61909708)</th>\n",
        "      <td>  0.00000</td>\n",
        "      <td>  61910985</td>\n",
        "      <td> LOC103217332</td>\n",
        "    </tr>\n",
        "    <tr>\n",
        "      <th>(CAE2, 55356081)</th>\n",
        "      <td>  0.00000</td>\n",
        "      <td>  55361265</td>\n",
        "      <td>      TMEM74B</td>\n",
        "    </tr>\n",
        "    <tr>\n",
        "      <th>(CAE1, 81080875)</th>\n",
        "      <td>  0.00000</td>\n",
        "      <td>  81087759</td>\n",
        "      <td> LOC103248208</td>\n",
        "    </tr>\n",
        "    <tr>\n",
        "      <th>(CAE2, 7594255)</th>\n",
        "      <td>  0.00000</td>\n",
        "      <td>   7596124</td>\n",
        "      <td> LOC103243851</td>\n",
        "    </tr>\n",
        "    <tr>\n",
        "      <th>(CAE1, 57105725)</th>\n",
        "      <td>  0.00000</td>\n",
        "      <td>  57107180</td>\n",
        "      <td> LOC103241029</td>\n",
        "    </tr>\n",
        "    <tr>\n",
        "      <th>(CAE1, 16358982)</th>\n",
        "      <td>  0.00000</td>\n",
        "      <td>  16361043</td>\n",
        "      <td> LOC103235573</td>\n",
        "    </tr>\n",
        "    <tr>\n",
        "      <th>(CAE2, 46744411)</th>\n",
        "      <td>  0.00000</td>\n",
        "      <td>  46746730</td>\n",
        "      <td> LOC103220801</td>\n",
        "    </tr>\n",
        "    <tr>\n",
        "      <th>(CAE1, 81100872)</th>\n",
        "      <td>  0.00000</td>\n",
        "      <td>  81107924</td>\n",
        "      <td> LOC103248209</td>\n",
        "    </tr>\n",
        "    <tr>\n",
        "      <th>(CAE1, 16361110)</th>\n",
        "      <td>  0.00000</td>\n",
        "      <td>  16363220</td>\n",
        "      <td> LOC103242014</td>\n",
        "    </tr>\n",
        "    <tr>\n",
        "      <th>(CAE1, 81132388)</th>\n",
        "      <td>  0.00000</td>\n",
        "      <td>  81138319</td>\n",
        "      <td> LOC103248212</td>\n",
        "    </tr>\n",
        "    <tr>\n",
        "      <th>(CAE2, 45902473)</th>\n",
        "      <td>  0.00000</td>\n",
        "      <td>  45903834</td>\n",
        "      <td> LOC103214642</td>\n",
        "    </tr>\n",
        "    <tr>\n",
        "      <th>(CAE1, 81185865)</th>\n",
        "      <td>  0.00000</td>\n",
        "      <td>  81193080</td>\n",
        "      <td> LOC103248214</td>\n",
        "    </tr>\n",
        "    <tr>\n",
        "      <th>(CAE1, 16374611)</th>\n",
        "      <td>  0.00000</td>\n",
        "      <td>  16376189</td>\n",
        "      <td> LOC103235580</td>\n",
        "    </tr>\n",
        "    <tr>\n",
        "      <th>(CAE1, 61744193)</th>\n",
        "      <td>  0.00000</td>\n",
        "      <td>  61747676</td>\n",
        "      <td> LOC103247918</td>\n",
        "    </tr>\n",
        "  </tbody>\n",
        "</table>\n",
        "<p>2273 rows \u00d7 3 columns</p>\n",
        "</div>"
       ],
       "metadata": {},
       "output_type": "pyout",
       "prompt_number": 94,
       "text": [
        "                       amax        end       gene_id\n",
        "(CAE3, 4747744)    48.61760    5378598        ATP8A2\n",
        "(CAE1, 47494862)   36.05660   47578818         ABCC8\n",
        "(CAE1, 117553388)  32.38441  118122895       KIRREL3\n",
        "(CAE1, 42201546)   30.17795   42335632          GAS2\n",
        "(CAE1, 79636636)   29.88969   80188349          GRM5\n",
        "(CAE1, 43416052)   28.28739   44335894         NELL1\n",
        "(CAE3, 8099783)    25.72230    8519645         MTUS2\n",
        "(CAE3, 3770901)    24.93997    4085379       SPATA13\n",
        "(CAE1, 96365646)   22.69252   96388966  LOC103248366\n",
        "(CAE2, 20613352)   21.20587   21720417         PTPRT\n",
        "(CAE1, 66666049)   20.69260   66707160         GDPD5\n",
        "(CAE2, 66037224)   20.34533   66320860           APP\n",
        "(CAE3, 20233323)   18.98894   20408490          DGKH\n",
        "(CAE1, 17137896)   18.35761   17313787         PTPRJ\n",
        "(CAE1, 115768337)  16.86788  115778084         TBRG1\n",
        "(CAE1, 46956665)   16.70152   47099546        SERGEF\n",
        "(CAE3, 19767190)   16.06920   20110650          VWA8\n",
        "(CAE3, 23840947)   15.84178   23909954         SIAH3\n",
        "(CAE1, 67894124)   15.70431   67926732  LOC103248083\n",
        "(CAE1, 94484090)   15.15998   94849960       DYNC2H1\n",
        "(CAE1, 31627447)   14.97001   31925520     KIAA1549L\n",
        "(CAE3, 568068)     14.61980     690030         CRYL1\n",
        "(CAE1, 34943544)   14.55943   34946866          FSHB\n",
        "(CAE1, 96284010)   14.19321   96298525  LOC103248368\n",
        "(CAE1, 65282311)   14.15360   65436007         C2CD3\n",
        "(CAE3, 24820029)   13.90847   24882444         HTR2A\n",
        "(CAE1, 117479328)  13.89792  117543262       ST3GAL4\n",
        "(CAE3, 20988638)   13.15384   21128966        EPSTI1\n",
        "(CAE2, 17233287)   13.12974   17324761       SLC13A3\n",
        "(CAE3, 28791491)   13.04286   28841963      RNASEH2B\n",
        "...                     ...        ...           ...\n",
        "(CAE1, 81057390)    0.00000   81058134  LOC103248206\n",
        "(CAE1, 15582781)    0.00000   15584988         YPEL4\n",
        "(CAE1, 62536922)    0.00000   62538821  LOC103247969\n",
        "(CAE1, 77180429)    0.00000   77181144  LOC103248157\n",
        "(CAE2, 88710182)    0.00000   88712024  LOC103220067\n",
        "(CAE2, 88777194)    0.00000   88779026  LOC103220073\n",
        "(CAE1, 80929496)    0.00000   80931691  LOC103248199\n",
        "(CAE2, 56617784)    0.00000   56619136         ROMO1\n",
        "(CAE2, 47855831)    0.00000   47856625        SCP2D1\n",
        "(CAE1, 80981885)    0.00000   80982678  LOC103248201\n",
        "(CAE1, 26211019)    0.00000   26213633  LOC103237427\n",
        "(CAE2, 55307268)    0.00000   55335200       RAD21L1\n",
        "(CAE1, 81032860)    0.00000   81038456        TRIM51\n",
        "(CAE2, 17380060)    0.00000   17388798  LOC103214199\n",
        "(CAE1, 81058982)    0.00000   81065150  LOC103248207\n",
        "(CAE1, 62075960)    0.00000   62077427  LOC103247939\n",
        "(CAE2, 61909708)    0.00000   61910985  LOC103217332\n",
        "(CAE2, 55356081)    0.00000   55361265       TMEM74B\n",
        "(CAE1, 81080875)    0.00000   81087759  LOC103248208\n",
        "(CAE2, 7594255)     0.00000    7596124  LOC103243851\n",
        "(CAE1, 57105725)    0.00000   57107180  LOC103241029\n",
        "(CAE1, 16358982)    0.00000   16361043  LOC103235573\n",
        "(CAE2, 46744411)    0.00000   46746730  LOC103220801\n",
        "(CAE1, 81100872)    0.00000   81107924  LOC103248209\n",
        "(CAE1, 16361110)    0.00000   16363220  LOC103242014\n",
        "(CAE1, 81132388)    0.00000   81138319  LOC103248212\n",
        "(CAE2, 45902473)    0.00000   45903834  LOC103214642\n",
        "(CAE1, 81185865)    0.00000   81193080  LOC103248214\n",
        "(CAE1, 16374611)    0.00000   16376189  LOC103235580\n",
        "(CAE1, 61744193)    0.00000   61747676  LOC103247918\n",
        "\n",
        "[2273 rows x 3 columns]"
       ]
      }
     ],
     "prompt_number": 94
    },
    {
     "cell_type": "code",
     "collapsed": false,
     "input": [
      "stats.kurtosis(range(10), axis=0, fisher=True, bias=True)"
     ],
     "language": "python",
     "metadata": {},
     "outputs": [
      {
       "metadata": {},
       "output_type": "pyout",
       "prompt_number": 11,
       "text": [
        "-1.2242424242424244"
       ]
      }
     ],
     "prompt_number": 11
    },
    {
     "cell_type": "code",
     "collapsed": false,
     "input": [
      "stats.skew(range(10)+[3], axis=0)"
     ],
     "language": "python",
     "metadata": {},
     "outputs": [
      {
       "metadata": {},
       "output_type": "pyout",
       "prompt_number": 14,
       "text": [
        "0.13328073436899723"
       ]
      }
     ],
     "prompt_number": 14
    },
    {
     "cell_type": "code",
     "collapsed": false,
     "input": [],
     "language": "python",
     "metadata": {},
     "outputs": []
    },
    {
     "cell_type": "code",
     "collapsed": false,
     "input": [
      "s1 = pd.Series([1,2,3,4,8],index=[1,2,3,4,8])\n",
      "s2 = pd.Series([1,2,7,14],index=[1,2,7,14])"
     ],
     "language": "python",
     "metadata": {},
     "outputs": []
    },
    {
     "cell_type": "code",
     "collapsed": false,
     "input": [
      "df1 = pd.DataFrame(17, columns=[\"rank\",\"out_of\",\"n_genes\"], index=[\"G0:1\",\"G0:2\",\"G0:4\",\"G0:5\",\"G0:6\",\"G0:7\"])\n",
      "df2 = pd.DataFrame(19, columns=[\"rank\",\"out_of\",\"n_genes\"], index=[\"G0:1\",\"G0:3\",\"GO:7\",\"G0:4\",\"G0:5\",\"G0:6\",\"G0:7\"])"
     ],
     "language": "python",
     "metadata": {},
     "outputs": [],
     "prompt_number": 59
    },
    {
     "cell_type": "code",
     "collapsed": false,
     "input": [
      "df1[\"rank\"]+df2[\"rank\"]"
     ],
     "language": "python",
     "metadata": {},
     "outputs": [
      {
       "metadata": {},
       "output_type": "pyout",
       "prompt_number": 54,
       "text": [
        "G0:1    32\n",
        "G0:2   NaN\n",
        "G0:3   NaN\n",
        "G0:4    32\n",
        "G0:5    32\n",
        "G0:6    32\n",
        "G0:7    32\n",
        "Name: rank, dtype: float64"
       ]
      }
     ],
     "prompt_number": 54
    },
    {
     "cell_type": "code",
     "collapsed": false,
     "input": [
      "df1[\"rank\"]+(df1[\"rank\"]>df2[\"rank\"])"
     ],
     "language": "python",
     "metadata": {},
     "outputs": [
      {
       "metadata": {},
       "output_type": "pyout",
       "prompt_number": 38,
       "text": [
        "G0:1    18\n",
        "G0:2    18\n",
        "G0:4    18\n",
        "G0:5    18\n",
        "G0:6    18\n",
        "G0:7    18\n",
        "Name: rank, dtype: int64"
       ]
      }
     ],
     "prompt_number": 38
    },
    {
     "cell_type": "code",
     "collapsed": false,
     "input": [
      "df2"
     ],
     "language": "python",
     "metadata": {},
     "outputs": [
      {
       "html": [
        "<div style=\"max-height:1000px;max-width:1500px;overflow:auto;\">\n",
        "<table border=\"1\" class=\"dataframe\">\n",
        "  <thead>\n",
        "    <tr style=\"text-align: right;\">\n",
        "      <th></th>\n",
        "      <th>rank</th>\n",
        "      <th>out_of</th>\n",
        "      <th>n_genes</th>\n",
        "    </tr>\n",
        "  </thead>\n",
        "  <tbody>\n",
        "    <tr>\n",
        "      <th>G0:1</th>\n",
        "      <td> 15</td>\n",
        "      <td> 15</td>\n",
        "      <td> 15</td>\n",
        "    </tr>\n",
        "    <tr>\n",
        "      <th>G0:3</th>\n",
        "      <td> 15</td>\n",
        "      <td> 15</td>\n",
        "      <td> 15</td>\n",
        "    </tr>\n",
        "    <tr>\n",
        "      <th>G0:19</th>\n",
        "      <td> 15</td>\n",
        "      <td> 15</td>\n",
        "      <td> 15</td>\n",
        "    </tr>\n",
        "    <tr>\n",
        "      <th>G0:4</th>\n",
        "      <td> 15</td>\n",
        "      <td> 15</td>\n",
        "      <td> 15</td>\n",
        "    </tr>\n",
        "    <tr>\n",
        "      <th>G0:5</th>\n",
        "      <td> 15</td>\n",
        "      <td> 15</td>\n",
        "      <td> 15</td>\n",
        "    </tr>\n",
        "    <tr>\n",
        "      <th>G0:6</th>\n",
        "      <td> 15</td>\n",
        "      <td> 15</td>\n",
        "      <td> 15</td>\n",
        "    </tr>\n",
        "    <tr>\n",
        "      <th>G0:7</th>\n",
        "      <td> 15</td>\n",
        "      <td> 15</td>\n",
        "      <td> 15</td>\n",
        "    </tr>\n",
        "  </tbody>\n",
        "</table>\n",
        "</div>"
       ],
       "metadata": {},
       "output_type": "pyout",
       "prompt_number": 51,
       "text": [
        "       rank  out_of  n_genes\n",
        "G0:1     15      15       15\n",
        "G0:3     15      15       15\n",
        "G0:19    15      15       15\n",
        "G0:4     15      15       15\n",
        "G0:5     15      15       15\n",
        "G0:6     15      15       15\n",
        "G0:7     15      15       15"
       ]
      }
     ],
     "prompt_number": 51
    },
    {
     "cell_type": "code",
     "collapsed": false,
     "input": [
      "df1[\"rank\"]"
     ],
     "language": "python",
     "metadata": {},
     "outputs": [
      {
       "metadata": {},
       "output_type": "pyout",
       "prompt_number": 60,
       "text": [
        "G0:1    17\n",
        "G0:2    17\n",
        "G0:4    17\n",
        "G0:5    17\n",
        "G0:6    17\n",
        "G0:7    17\n",
        "Name: rank, dtype: int64"
       ]
      }
     ],
     "prompt_number": 60
    },
    {
     "cell_type": "code",
     "collapsed": false,
     "input": [
      "df1[\"rank\"]+(df1[\"n_genes\"]>df2[\"rank\"].reindex(df1.index).fillna(0))"
     ],
     "language": "python",
     "metadata": {},
     "outputs": [
      {
       "metadata": {},
       "output_type": "pyout",
       "prompt_number": 61,
       "text": [
        "G0:1    17\n",
        "G0:2    18\n",
        "G0:4    17\n",
        "G0:5    17\n",
        "G0:6    17\n",
        "G0:7    17\n",
        "dtype: int64"
       ]
      }
     ],
     "prompt_number": 61
    },
    {
     "cell_type": "code",
     "collapsed": false,
     "input": [
      "fn0 = open(eu(\"~/vervet_project/analyses/20140611_163_subpop_sweepfinder/_data/sweed_go_enrichment_q0.001_d10000_0_pys.temp.tsv\"))\n",
      "fn1 = open(eu(\"~/vervet_project/analyses/20140611_163_subpop_sweepfinder/_data/sweed_go_enrichment_q0.001_d10000_1_pys.temp.tsv\"))\n",
      "tot_rank = te.read_table(fn0,index_col=0)\n",
      "rank_table = te.read_table(fn1,index_col=0)"
     ],
     "language": "python",
     "metadata": {},
     "outputs": [],
     "prompt_number": 17
    },
    {
     "cell_type": "code",
     "collapsed": false,
     "input": [
      "tot_rank.index.diff(rank_table.index)"
     ],
     "language": "python",
     "metadata": {},
     "outputs": [
      {
       "metadata": {},
       "output_type": "pyout",
       "prompt_number": 20,
       "text": [
        "Index([u'000'], dtype='object')"
       ]
      }
     ],
     "prompt_number": 20
    },
    {
     "cell_type": "code",
     "collapsed": false,
     "input": [
      "top_n = 1000\n",
      "value_s = tot_sweed_df[\"Likelihood\"].sort(ascending=False,inplace=False)\n",
      "top_s = value_s.iloc[:top_n]"
     ],
     "language": "python",
     "metadata": {},
     "outputs": [],
     "prompt_number": 14
    },
    {
     "cell_type": "code",
     "collapsed": false,
     "input": [
      "max_dist = 10000\n",
      "cand_genes = te.get_genes(top_s, gene_df, max_dist=10000)"
     ],
     "language": "python",
     "metadata": {},
     "outputs": [],
     "prompt_number": 16
    },
    {
     "cell_type": "code",
     "collapsed": false,
     "input": [
      "real_assoc = te.get_go_assoc(cand_genes, gene_to_go)"
     ],
     "language": "python",
     "metadata": {},
     "outputs": [],
     "prompt_number": 19
    },
    {
     "cell_type": "code",
     "collapsed": false,
     "input": [
      "init_rank_table = te.get_initial_rank_table(real_assoc)"
     ],
     "language": "python",
     "metadata": {},
     "outputs": [],
     "prompt_number": 21
    },
    {
     "cell_type": "code",
     "collapsed": false,
     "input": [
      "assoc_table = te.multiple_permut_assoc_low_mem(value_s, init_rank_table,\n",
      "                                                gene_df,\n",
      "                                                gene_to_go,\n",
      "                                                top_n,\n",
      "                                                max_dist,\n",
      "                                                n_runs=1000,\n",
      "                                                ascending=False,\n",
      "                                                rnd_seed=None)"
     ],
     "language": "python",
     "metadata": {},
     "outputs": [
      {
       "ename": "KeyboardInterrupt",
       "evalue": "",
       "output_type": "pyerr",
       "traceback": [
        "\u001b[1;31m---------------------------------------------------------------------------\u001b[0m\n\u001b[1;31mKeyboardInterrupt\u001b[0m                         Traceback (most recent call last)",
        "\u001b[1;32m<ipython-input-67-01008f4c3fd3>\u001b[0m in \u001b[0;36m<module>\u001b[1;34m()\u001b[0m\n\u001b[0;32m      6\u001b[0m                                                 \u001b[0mn_runs\u001b[0m\u001b[1;33m=\u001b[0m\u001b[1;36m1000\u001b[0m\u001b[1;33m,\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n\u001b[0;32m      7\u001b[0m                                                 \u001b[0mascending\u001b[0m\u001b[1;33m=\u001b[0m\u001b[0mFalse\u001b[0m\u001b[1;33m,\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n\u001b[1;32m----> 8\u001b[1;33m                                                 rnd_seed=None)\n\u001b[0m",
        "\u001b[1;32m/home/GMI/hannes.svardal/script/hs_vervet/scripts/test_enrichment.py\u001b[0m in \u001b[0;36mmultiple_permut_assoc_low_mem\u001b[1;34m(rod_s, init_rank_table, gene_df, gene_to_go, top_n, max_dist, n_runs, ascending, rnd_seed)\u001b[0m\n\u001b[0;32m    142\u001b[0m     \u001b[1;32mfor\u001b[0m \u001b[0mrnd\u001b[0m \u001b[1;32min\u001b[0m \u001b[0mnp\u001b[0m\u001b[1;33m.\u001b[0m\u001b[0mrandom\u001b[0m\u001b[1;33m.\u001b[0m\u001b[0mrand\u001b[0m\u001b[1;33m(\u001b[0m\u001b[0mn_runs\u001b[0m\u001b[1;33m)\u001b[0m\u001b[1;33m:\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n\u001b[0;32m    143\u001b[0m         assoc = permut_assoc(rod_s, rnd, gene_df, gene_to_go, \n\u001b[1;32m--> 144\u001b[1;33m                                        top_n, max_dist, ascending)\n\u001b[0m\u001b[0;32m    145\u001b[0m         init_rank_table[\"rank\"] += (init_rank_table[\"n_genes\"] > \\\n\u001b[0;32m    146\u001b[0m                                     assoc.reindex(init_rank_table.index).fillna(0))\n",
        "\u001b[1;32m/home/GMI/hannes.svardal/script/hs_vervet/scripts/test_enrichment.py\u001b[0m in \u001b[0;36mpermut_assoc\u001b[1;34m(rod_s, rnd, gene_df, gene_to_go, top_n, max_dist, ascending)\u001b[0m\n\u001b[0;32m    103\u001b[0m     \"\"\"\n\u001b[0;32m    104\u001b[0m     \u001b[0ms\u001b[0m \u001b[1;33m=\u001b[0m \u001b[0mshift_rod\u001b[0m\u001b[1;33m(\u001b[0m\u001b[0mrod_s\u001b[0m\u001b[1;33m,\u001b[0m \u001b[0mrnd\u001b[0m\u001b[1;33m)\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n\u001b[1;32m--> 105\u001b[1;33m     \u001b[0ms\u001b[0m\u001b[1;33m.\u001b[0m\u001b[0msort\u001b[0m\u001b[1;33m(\u001b[0m\u001b[0mascending\u001b[0m\u001b[1;33m=\u001b[0m\u001b[0mascending\u001b[0m\u001b[1;33m,\u001b[0m \u001b[0minplace\u001b[0m\u001b[1;33m=\u001b[0m\u001b[0mTrue\u001b[0m\u001b[1;33m)\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n\u001b[0m\u001b[0;32m    106\u001b[0m     \u001b[0mtop_s\u001b[0m \u001b[1;33m=\u001b[0m \u001b[0ms\u001b[0m\u001b[1;33m.\u001b[0m\u001b[0miloc\u001b[0m\u001b[1;33m[\u001b[0m\u001b[1;33m:\u001b[0m\u001b[0mtop_n\u001b[0m\u001b[1;33m]\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n\u001b[0;32m    107\u001b[0m     \u001b[0mcand_genes\u001b[0m \u001b[1;33m=\u001b[0m \u001b[0mget_genes\u001b[0m\u001b[1;33m(\u001b[0m\u001b[0mtop_s\u001b[0m\u001b[1;33m,\u001b[0m \u001b[0mgene_df\u001b[0m\u001b[1;33m,\u001b[0m \u001b[0mmax_dist\u001b[0m\u001b[1;33m=\u001b[0m\u001b[0mmax_dist\u001b[0m\u001b[1;33m)\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n",
        "\u001b[1;32m/net/gmi.oeaw.ac.at/nordborg/user/hannes.svardal/Documents/Dropbox/Akademisches/lib/python/virtualenvs/ver/lib/python2.7/site-packages/pandas/core/series.pyc\u001b[0m in \u001b[0;36msort\u001b[1;34m(self, axis, ascending, kind, na_position, inplace)\u001b[0m\n\u001b[0;32m   1646\u001b[0m                           \u001b[0mkind\u001b[0m\u001b[1;33m=\u001b[0m\u001b[0mkind\u001b[0m\u001b[1;33m,\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n\u001b[0;32m   1647\u001b[0m                           \u001b[0mna_position\u001b[0m\u001b[1;33m=\u001b[0m\u001b[0mna_position\u001b[0m\u001b[1;33m,\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n\u001b[1;32m-> 1648\u001b[1;33m                           inplace=inplace)\n\u001b[0m\u001b[0;32m   1649\u001b[0m \u001b[1;33m\u001b[0m\u001b[0m\n\u001b[0;32m   1650\u001b[0m     \u001b[1;32mdef\u001b[0m \u001b[0morder\u001b[0m\u001b[1;33m(\u001b[0m\u001b[0mself\u001b[0m\u001b[1;33m,\u001b[0m \u001b[0mna_last\u001b[0m\u001b[1;33m=\u001b[0m\u001b[0mNone\u001b[0m\u001b[1;33m,\u001b[0m \u001b[0mascending\u001b[0m\u001b[1;33m=\u001b[0m\u001b[0mTrue\u001b[0m\u001b[1;33m,\u001b[0m \u001b[0mkind\u001b[0m\u001b[1;33m=\u001b[0m\u001b[1;34m'quicksort'\u001b[0m\u001b[1;33m,\u001b[0m \u001b[0mna_position\u001b[0m\u001b[1;33m=\u001b[0m\u001b[1;34m'last'\u001b[0m\u001b[1;33m,\u001b[0m \u001b[0minplace\u001b[0m\u001b[1;33m=\u001b[0m\u001b[0mFalse\u001b[0m\u001b[1;33m)\u001b[0m\u001b[1;33m:\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n",
        "\u001b[1;32m/net/gmi.oeaw.ac.at/nordborg/user/hannes.svardal/Documents/Dropbox/Akademisches/lib/python/virtualenvs/ver/lib/python2.7/site-packages/pandas/core/series.pyc\u001b[0m in \u001b[0;36morder\u001b[1;34m(self, na_last, ascending, kind, na_position, inplace)\u001b[0m\n\u001b[0;32m   1705\u001b[0m         \u001b[0midx\u001b[0m \u001b[1;33m=\u001b[0m \u001b[0mpa\u001b[0m\u001b[1;33m.\u001b[0m\u001b[0marange\u001b[0m\u001b[1;33m(\u001b[0m\u001b[0mlen\u001b[0m\u001b[1;33m(\u001b[0m\u001b[0mself\u001b[0m\u001b[1;33m)\u001b[0m\u001b[1;33m)\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n\u001b[0;32m   1706\u001b[0m \u001b[1;33m\u001b[0m\u001b[0m\n\u001b[1;32m-> 1707\u001b[1;33m         \u001b[0margsorted\u001b[0m \u001b[1;33m=\u001b[0m \u001b[0m_try_kind_sort\u001b[0m\u001b[1;33m(\u001b[0m\u001b[0marr\u001b[0m\u001b[1;33m[\u001b[0m\u001b[0mgood\u001b[0m\u001b[1;33m]\u001b[0m\u001b[1;33m)\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n\u001b[0m\u001b[0;32m   1708\u001b[0m \u001b[1;33m\u001b[0m\u001b[0m\n\u001b[0;32m   1709\u001b[0m         \u001b[1;32mif\u001b[0m \u001b[1;32mnot\u001b[0m \u001b[0mascending\u001b[0m\u001b[1;33m:\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n",
        "\u001b[1;32m/net/gmi.oeaw.ac.at/nordborg/user/hannes.svardal/Documents/Dropbox/Akademisches/lib/python/virtualenvs/ver/lib/python2.7/site-packages/pandas/core/series.pyc\u001b[0m in \u001b[0;36m_try_kind_sort\u001b[1;34m(arr)\u001b[0m\n\u001b[0;32m   1691\u001b[0m             \u001b[1;32mtry\u001b[0m\u001b[1;33m:\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n\u001b[0;32m   1692\u001b[0m                 \u001b[1;31m# if kind==mergesort, it can fail for object dtype\u001b[0m\u001b[1;33m\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n\u001b[1;32m-> 1693\u001b[1;33m                 \u001b[1;32mreturn\u001b[0m \u001b[0marr\u001b[0m\u001b[1;33m.\u001b[0m\u001b[0margsort\u001b[0m\u001b[1;33m(\u001b[0m\u001b[0mkind\u001b[0m\u001b[1;33m=\u001b[0m\u001b[0mkind\u001b[0m\u001b[1;33m)\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n\u001b[0m\u001b[0;32m   1694\u001b[0m             \u001b[1;32mexcept\u001b[0m \u001b[0mTypeError\u001b[0m\u001b[1;33m:\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n\u001b[0;32m   1695\u001b[0m                 \u001b[1;31m# stable sort not available for object dtype\u001b[0m\u001b[1;33m\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n",
        "\u001b[1;31mKeyboardInterrupt\u001b[0m: "
       ]
      }
     ],
     "prompt_number": 67
    },
    {
     "cell_type": "code",
     "collapsed": false,
     "input": [
      "import time, string, gc\n",
      "from memory_profiler import LineProfiler, show_results\n",
      "import memory_profiler as mprof\n",
      "\n",
      "prof = LineProfiler()\n",
      "\n",
      "@prof\n",
      "def test(nrow=1000000, ncol = 4, timetest = 5):\n",
      "    from_ = nrow // 10\n",
      "    to_ = 9 * nrow // 10\n",
      "    df = pd.DataFrame(np.random.randn(nrow, ncol),\n",
      "                          index = np.random.randn(nrow),\n",
      "                          columns = list(string.letters[0:ncol]))\n",
      "    #df_new = df[from_:to_].copy()\n",
      "    del df\n",
      "    #del df_new\n",
      "    gc.collect()\n",
      "\n",
      "test()\n",
      "# for _ in xrange(10):\n",
      "#     print mprof.memory_usage()\n",
      "\n",
      "show_results(prof)"
     ],
     "language": "python",
     "metadata": {},
     "outputs": [
      {
       "output_type": "stream",
       "stream": "stdout",
       "text": [
        "Filename: <ipython-input-66-2d400069d413>\n",
        "\n",
        "ERROR: Could not find file <ipython-input-66-2d400069d413>\n",
        "NOTE: %mprun can only be used on functions defined in physical files, and not in the IPython environment.\n"
       ]
      }
     ],
     "prompt_number": 66
    },
    {
     "cell_type": "code",
     "collapsed": false,
     "input": [],
     "language": "python",
     "metadata": {},
     "outputs": []
    }
   ],
   "metadata": {}
  }
 ]
}