{
 "metadata": {
  "name": "",
  "signature": "sha256:fcb4cb1f2ac732b5918762c636d89dfcc20dc74dc3cbb0982c7c5bb4cffe6604"
 },
 "nbformat": 3,
 "nbformat_minor": 0,
 "worksheets": [
  {
   "cells": [
    {
     "cell_type": "code",
     "collapsed": false,
     "input": [
      "import os\n",
      "import pandas as pd\n",
      "import numpy as np\n",
      "eu = os.path.expanduser\n",
      "jn = os.path.join"
     ],
     "language": "python",
     "metadata": {},
     "outputs": [],
     "prompt_number": 2
    },
    {
     "cell_type": "code",
     "collapsed": false,
     "input": [
      "assoc_dir = eu(\"~/vervet_project/data/gene_associations\")"
     ],
     "language": "python",
     "metadata": {},
     "outputs": [],
     "prompt_number": 3
    },
    {
     "cell_type": "code",
     "collapsed": false,
     "input": [
      "meta_dir = eu(\"~/vervet_project/metadata\")\n",
      "def load_gene_df():\n",
      "    gene_df = pd.read_csv(jn(meta_dir,\"annot_mike_genes_minimal.tsv\"),sep=\"\\t\",index_col = [0,1])\n",
      "    gene_df.index.names = [gene_df.index.names[0],'start']\n",
      "    return gene_df"
     ],
     "language": "python",
     "metadata": {},
     "outputs": [],
     "prompt_number": 825
    },
    {
     "cell_type": "code",
     "collapsed": false,
     "input": [
      "gene_df = load_gene_df()"
     ],
     "language": "python",
     "metadata": {},
     "outputs": [],
     "prompt_number": 902
    },
    {
     "cell_type": "code",
     "collapsed": false,
     "input": [
      "pops = [\"aet\",\"cyn\",\"pyn\",\"pys\",\"sab\",\"tan\"]"
     ],
     "language": "python",
     "metadata": {},
     "outputs": [],
     "prompt_number": 6
    },
    {
     "cell_type": "code",
     "collapsed": false,
     "input": [
      "gene_df.to_csv(jn(meta_dir,\"annot_mike_genes_minimal.tsv\"),sep=\"\\t\")"
     ],
     "language": "python",
     "metadata": {},
     "outputs": [],
     "prompt_number": 7
    },
    {
     "cell_type": "code",
     "collapsed": false,
     "input": [
      "def index_rolling(s,window,func,overlap=0,*args,**kwargs):\n",
      "    \"\"\"\n",
      "    Apply function in rolling windows, where the window\n",
      "    size is defined with respect to the index values.\n",
      "    This means that different windows can comprise different\n",
      "    numbers of elements.\n",
      "    \n",
      "    s ... pandas Series\n",
      "    window ... window size in units of the index values\n",
      "    func ... function to apply to the series values within each\n",
      "                window\n",
      "    overlap ... oberlap size of windows\n",
      "    args, kwarg ... additional arguments for func\n",
      "    \n",
      "    Example: index_rolling(pd.Series(range(20),index=np.random.rand(20))),0.1,max)\n",
      "    \"\"\"\n",
      "    #note that basis must be sorted in order for this to work properly\n",
      "    windows_min = s.index.min()\n",
      "    windows_max = s.index.max()\n",
      "    window_starts = np.arange(windows_min, windows_max, window-overlap)\n",
      "    window_starts = pd.Series(window_starts, index = window_starts)\n",
      "    def applyToWindow(val):\n",
      "        # using slice_indexer rather that what.loc [val:val+window] allows\n",
      "        # window limits that are not specifically in the index\n",
      "        indexer = s.index.slice_indexer(val,val+window,1)\n",
      "        chunk = s.iloc[indexer]\n",
      "        try:\n",
      "            return func(chunk,*args,**kwargs)\n",
      "        except ValueError, e:\n",
      "            if \"empty sequence\" in str(e):\n",
      "                #print indexer, chunk\n",
      "                return None\n",
      "            else:\n",
      "                raise           \n",
      "    rolled = window_starts.apply(applyToWindow)\n",
      "    return rolled"
     ],
     "language": "python",
     "metadata": {},
     "outputs": [],
     "prompt_number": 8
    },
    {
     "cell_type": "heading",
     "level": 2,
     "metadata": {},
     "source": [
      "ROD (reference ordered data) permutation test"
     ]
    },
    {
     "cell_type": "markdown",
     "metadata": {},
     "source": [
      "tools that, given an analysis result, one can rotate (shift) the results randomly around the genome"
     ]
    },
    {
     "cell_type": "code",
     "collapsed": false,
     "input": [
      "def get_chrom_length():\n",
      "    return pd.read_csv(eu(\"~/vervet_project/metadata/ref3500.tsv\"),\n",
      "                       sep=\"\\t\",index_col=0,squeeze=True,header=False)"
     ],
     "language": "python",
     "metadata": {},
     "outputs": [],
     "prompt_number": 7
    },
    {
     "cell_type": "code",
     "collapsed": false,
     "input": [
      "def shift_rod(rod_df, rnd, mode = \"grid\"):\n",
      "    \"\"\"\n",
      "    shift reference ordered data across the whole genome\n",
      "\n",
      "    Input:\n",
      "    rod_df ... pandas dataframe or series with mulitiindex (chrom, pos)\n",
      "                \n",
      "    modes ...\n",
      "        'grid' ... just rotate the index of the rod data frame\n",
      "                    this means that the positions stay the same only the\n",
      "                    value for each position becomes different\n",
      "                    Faster, but means that you only hit the same grid-point\n",
      "                    this should make it conservative on large grids. Large\n",
      "                    grids are problematic if the fraction of top windows \n",
      "                    considered becomes large \n",
      "        'continuous' ... add the random shift to each index value.\n",
      "                         NOT IMPLEMENTED\n",
      "    \"\"\"\n",
      "    if mode == \"grid\":\n",
      "        new_start_i = int(len(rod_df)*rnd)\n",
      "        rotate_data = np.concatenate((rod_df.iloc[new_start_i:].values,rod_df.iloc[:new_start_i].values))\n",
      "        if  isinstance(rod_df,pd.core.series.Series):\n",
      "            r = pd.Series(rotate_data,index=rod_df.index)\n",
      "            return r\n",
      "        elif isinstance(rod_df,pd.core.frame.DataFrame):\n",
      "            r = pd.DataFrame(rotate_data,index=rod_df.index,columns=rod_df.columns)\n",
      "            return r\n",
      "        \n",
      "    else:\n",
      "        raise UserException(\"Only mode grid supported.\")"
     ],
     "language": "python",
     "metadata": {},
     "outputs": [],
     "prompt_number": 8
    },
    {
     "cell_type": "code",
     "collapsed": false,
     "input": [
      "%timeit shift_rod(tot_sweed_df[\"Likelihood\"], 0.32424) "
     ],
     "language": "python",
     "metadata": {},
     "outputs": [
      {
       "output_type": "stream",
       "stream": "stdout",
       "text": [
        "100 loops, best of 3: 14.5 ms per loop\n"
       ]
      }
     ],
     "prompt_number": 900
    },
    {
     "cell_type": "heading",
     "level": 2,
     "metadata": {},
     "source": [
      "get genes for peaks"
     ]
    },
    {
     "cell_type": "code",
     "collapsed": false,
     "input": [
      "def get_peaks(value_s,min_peak_dist):\n",
      "    \"\"\"\n",
      "    ranked df of peaks\n",
      "    assuming a minimum distance between peaks\n",
      "    \"\"\"\n",
      "    assert len(value_s.index.droplevel(1).unique()) == 1, \"Multiple chromsosomes not supported.\"\n",
      "    chrom = value_s.index[0][0]\n",
      "    x = index_rolling(value_s.ix[chrom],2*min_peak_dist,np.argmax,min_peak_dist)\n",
      "    x.name = \"pos\"\n",
      "    #only take peaks that are found in at least two windows \n",
      "    #this makes sure that we have a minimal distance between peaks\n",
      "    pos_true = (x.reset_index().groupby(\"pos\").apply(len)>1)\n",
      "    pos = pos_true[pos_true.values].index\n",
      "    s = value_s.ix[chrom].ix[pos]\n",
      "    return s"
     ],
     "language": "python",
     "metadata": {},
     "outputs": [],
     "prompt_number": 9
    },
    {
     "cell_type": "code",
     "collapsed": false,
     "input": [
      "def get_genes(peak_s, gene_df, max_dist):\n",
      "    \"\"\"\n",
      "    take the input series and gets \n",
      "    names of genes nearby\n",
      "    \n",
      "    Input:\n",
      "    peak_s ... pandas series with (chrom, pos) index and value of\n",
      "                the statistic ('peak height'). Series should be named.\n",
      "    gene_df ... data frame with gene info \n",
      "    \"\"\"    \n",
      "    all_genes = []\n",
      "    if not gene_df.index.is_monotonic:\n",
      "        gene_df = gene_df.sort_index()\n",
      "    tot_hit_df = pd.DataFrame()\n",
      "    for chrom in peak_s.index.droplevel(1).unique():\n",
      "        loc_gene_df = gene_df.ix[chrom]\n",
      "        #loc_gene_df = loc_gene_df.append(pd.DataFrame(np.nan,index=[np.inf],columns=loc_gene_df.columns))\n",
      "        pos_rel_to_start = np.searchsorted(loc_gene_df.index.values-max_dist,peak_s.ix[chrom].index.values)\n",
      "        pos_rel_to_end = np.searchsorted(loc_gene_df[\"end\"].values+max_dist,peak_s.ix[chrom].index.values)\n",
      "        genes = list(set(loc_gene_df[\"gene_id\"].iloc[np.hstack([range(a,b) for a,b in zip(pos_rel_to_end,pos_rel_to_start)])]))\n",
      "        #print chrom, genes\n",
      "        all_genes += genes\n",
      "    return all_genes\n",
      "\n",
      "def get_gene_info(gene_ls,gene_df):\n",
      "    \"\"\"\n",
      "    for a list of gene ids,\n",
      "    get a data frame with their \n",
      "    position\n",
      "    \"\"\"\n",
      "    gi = gene_df[gene_df[\"gene_id\"].apply(lambda x: x in gene_ls)]\n",
      "    return gi\n",
      "\n",
      "def get_peaks(gene_info,top_s,max_dist):\n",
      "    \"\"\"\n",
      "    For each gene in gene_info get the\n",
      "    peaks within max_dist in top_s. This \n",
      "    is basically reverse engineering to get\n",
      "    the peak info for each gene that was found \n",
      "    to be associated with a peak. \n",
      "    \n",
      "    Input:\n",
      "    gene_info ... data frame with index ('chrom','start')\n",
      "                and columns 'gene_id' and 'end'\n",
      "    top_s ... series of peak positions with index (chrom, pos)\n",
      "                and values peak height\n",
      "    max_dist ... maximum distance between gene and peak\n",
      "    \"\"\"\n",
      "    def get_dist(df,gene_pos):\n",
      "        s = pd.Series(df.index.droplevel(0).values - gene_pos.ix[df.index[0][0]],\n",
      "                                                  index=df.index.droplevel(0).values)\n",
      "        return s\n",
      "    tot_gene_peaks_df = pd.DataFrame()\n",
      "    if not top_s.index.is_monotonic:\n",
      "        top_s = top_s.sort_index()\n",
      "    for chrom in gene_info.index.droplevel(1).unique():\n",
      "        loc_top_s = top_s.ix[chrom]\n",
      "        start = np.searchsorted(loc_top_s.index.values+max_dist,gene_info.ix[chrom].index.values)\n",
      "        end = np.searchsorted(loc_top_s.index.values-max_dist,gene_info.ix[chrom][\"end\"].values)\n",
      "        \n",
      "        x = pd.concat([loc_top_s.iloc[st:ed] for st,ed in zip(start,end)],\n",
      "                      keys=gene_info.ix[chrom][\"gene_id\"].values)\n",
      "        x.name = \"peak_height\"\n",
      "        \n",
      "\n",
      "\n",
      "        dist_start = x.groupby(lambda i: i[0]).\\\n",
      "                    apply(lambda df: get_dist(df,\n",
      "                                              gene_info.ix[chrom].reset_index().set_index(\"gene_id\")[\"pos\"]))\n",
      "        dist_start.name = \"dist_start\"\n",
      "        dist_end = x.groupby(lambda i: i[0]).\\\n",
      "                    apply(lambda df: get_dist(df,\n",
      "                                              gene_info.ix[chrom].set_index(\"gene_id\")[\"end\"]))\n",
      "        dist_end.name = \"dist_end\"\n",
      "        gene_peaks_df = pd.concat([x,dist_start,dist_end],axis=1)\n",
      "        gene_peaks_df.index = pd.MultiIndex.from_arrays([gene_peaks_df.index.droplevel(1),\n",
      "                                         [chrom]*len(x),\n",
      "                                         gene_peaks_df.index.droplevel(0)])\n",
      "        tot_gene_peaks_df = pd.concat([tot_gene_peaks_df, gene_peaks_df])\n",
      "        tot_gene_peaks_df.index.names = [\"gene_id\",\"chrom\",\"peak_pos\"]\n",
      "        if chrom == \"CAE19\":\n",
      "            print gene_info.ix[chrom].reset_index().set_index(\"gene_id\")[\"pos\"]\n",
      "            print loc_top_s\n",
      "            print start\n",
      "            print end\n",
      "            print x\n",
      "            print dist_start\n",
      "    return tot_gene_peaks_df"
     ],
     "language": "python",
     "metadata": {},
     "outputs": [],
     "prompt_number": 10
    },
    {
     "cell_type": "code",
     "collapsed": false,
     "input": [
      "chromosomes = [\"CAE\" + str(i) for i in range(1,30)+[\"X\",\"Y\"]]\n",
      "def get_sweed_df(chrom,pop):\n",
      "    sweed_res = pd.read_csv(eu(\"~/vervet_project/analyses/20140611_163_subpop_sweepfinder/\"\n",
      "                             \"_data/SweeD_Report.163_ref3500_{}_{}\".format(chrom,pop)),\n",
      "                                                          sep=\"\\t\",skiprows=2,index_col=0)\n",
      "    sweed_res.index = pd.MultiIndex.from_tuples(zip([chrom]*len(sweed_res.index),sweed_res.index))\n",
      "    return sweed_res"
     ],
     "language": "python",
     "metadata": {},
     "outputs": [],
     "prompt_number": 11
    },
    {
     "cell_type": "code",
     "collapsed": false,
     "input": [],
     "language": "python",
     "metadata": {},
     "outputs": [
      {
       "metadata": {},
       "output_type": "pyout",
       "prompt_number": 1158,
       "text": [
        "'0.14.0'"
       ]
      }
     ],
     "prompt_number": 1158
    },
    {
     "cell_type": "code",
     "collapsed": false,
     "input": [
      "pop = \"aet\"\n",
      "eu(\"~/vervet_project/analyses/20140611_163_subpop_sweepfinder/\"\n",
      "                             \"_data/SweeD_Report.163_ref3500_{}_{}\".format(chrom,pop))"
     ],
     "language": "python",
     "metadata": {},
     "outputs": [
      {
       "metadata": {},
       "output_type": "pyout",
       "prompt_number": 1057,
       "text": [
        "'/home/GMI/hannes.svardal/vervet_project/analyses/20140611_163_subpop_sweepfinder/_data/SweeD_Report.163_ref3500_CAEY_aet'"
       ]
      }
     ],
     "prompt_number": 1057
    },
    {
     "cell_type": "code",
     "collapsed": false,
     "input": [
      "gene_df = load_gene_df()\n",
      "gene_df = gene_df.sortlevel(level=[0,1])"
     ],
     "language": "python",
     "metadata": {},
     "outputs": [],
     "prompt_number": 372
    },
    {
     "cell_type": "code",
     "collapsed": false,
     "input": [
      "pop = \"aet\"\n",
      "tot_sweed_df = pd.read_csv(eu(\"~/vervet_project/analyses/20140611_163_subpop_sweepfinder/\"\n",
      "                                 \"_data/tot_sweed_result_{}.tsv\".format(pop)),sep=\"\\t\",index_col=[0,1])"
     ],
     "language": "python",
     "metadata": {},
     "outputs": [],
     "prompt_number": 19
    },
    {
     "cell_type": "code",
     "collapsed": false,
     "input": [
      "top_s = tot_sweed_df.sort(\"Likelihood\",ascending=False)[\"Likelihood\"].iloc[:2000]\n",
      "#top_s = pd.concat([pd.Series([17],index=pd.MultiIndex.from_arrays([[\"CAE1\"],[4922]])),top_s])"
     ],
     "language": "python",
     "metadata": {},
     "outputs": [],
     "prompt_number": 1152
    },
    {
     "cell_type": "code",
     "collapsed": false,
     "input": [
      "max_dist = 10000\n",
      "cand_genes = get_genes(top_s, gene_df, max_dist=max_dist)\n",
      "gene_info = get_gene_info(cand_genes, gene_df)\n",
      "peaks = get_peaks(gene_info,top_s,max_dist=max_dist)"
     ],
     "language": "python",
     "metadata": {},
     "outputs": [
      {
       "output_type": "stream",
       "stream": "stdout",
       "text": [
        "gene_id\n",
        "SAMM50     26467319\n",
        "Name: pos, dtype: int64\n",
        "pos\n",
        "9483311.2373     21.93971\n",
        "9483796.1723     23.32173\n",
        "26477373.2917    20.04123\n",
        "26478828.0967    19.82511\n",
        "Name: Likelihood, dtype: float64\n",
        "[2]\n",
        "[4]\n",
        "        pos          \n",
        "SAMM50  26477373.2917    20.04123\n",
        "        26478828.0967    19.82511\n",
        "Name: peak_height, dtype: float64\n",
        "SAMM50  26477373.2917    10054.2917\n",
        "        26478828.0967    11509.0967\n",
        "Name: dist_start, dtype: float64\n"
       ]
      }
     ],
     "prompt_number": 1225
    },
    {
     "cell_type": "code",
     "collapsed": false,
     "input": [
      "def genes_per_peak_s(df):\n",
      "    s = df.iloc[0]#[\"peak_height\"]\n",
      "    s[\"genes\"] = 3#list(df[\"gene_id\"].unique())\n",
      "    return s\n",
      "p = peaks.reset_index([0])"
     ],
     "language": "python",
     "metadata": {},
     "outputs": [],
     "prompt_number": 1248
    },
    {
     "cell_type": "code",
     "collapsed": false,
     "input": [
      "gene_list_peak_pos = p[\"gene_id\"].groupby(lambda x: x).apply(list)\n",
      "gene_list_peak_pos.name = \"genes\"\n",
      "gene_list_peak_pos.index = pd.MultiIndex.from_tuples(gene_list_peak_pos.index)\n",
      "peak_info = pd.concat([top_s,gene_list_peak_pos],axis=1)\n",
      "peak_info.sort(\"Likelihood\",ascending=False,inplace=True)"
     ],
     "language": "python",
     "metadata": {},
     "outputs": [],
     "prompt_number": 1263
    },
    {
     "cell_type": "code",
     "collapsed": false,
     "input": [],
     "language": "python",
     "metadata": {},
     "outputs": [],
     "prompt_number": 1270
    },
    {
     "cell_type": "code",
     "collapsed": false,
     "input": [],
     "language": "python",
     "metadata": {},
     "outputs": [],
     "prompt_number": 1272
    },
    {
     "cell_type": "heading",
     "level": 2,
     "metadata": {},
     "source": [
      "get go associations for gene list"
     ]
    },
    {
     "cell_type": "code",
     "collapsed": false,
     "input": [
      "gene_to_go = pd.read_csv(jn(assoc_dir,\"gene_to_cat.csv\"),index_col=0)\n",
      "gene_to_name = pd.read_csv(jn(assoc_dir,\"gene_to_name.csv\"),index_col=0)\n",
      "go_to_name = pd.read_csv(jn(assoc_dir,\"cat_to_name.csv\"),index_col=0)"
     ],
     "language": "python",
     "metadata": {},
     "outputs": [],
     "prompt_number": 11
    },
    {
     "cell_type": "code",
     "collapsed": false,
     "input": [
      "rank_table.join(go_to_name.set_index(\"go_identifier\",drop=True))"
     ],
     "language": "python",
     "metadata": {},
     "outputs": [
      {
       "html": [
        "<div style=\"max-height:1000px;max-width:1500px;overflow:auto;\">\n",
        "<table border=\"1\" class=\"dataframe\">\n",
        "  <thead>\n",
        "    <tr style=\"text-align: right;\">\n",
        "      <th></th>\n",
        "      <th>n_genes</th>\n",
        "      <th>out_of</th>\n",
        "      <th>rank</th>\n",
        "      <th>go_name</th>\n",
        "    </tr>\n",
        "    <tr>\n",
        "      <th>go_identifier</th>\n",
        "      <th></th>\n",
        "      <th></th>\n",
        "      <th></th>\n",
        "      <th></th>\n",
        "    </tr>\n",
        "  </thead>\n",
        "  <tbody>\n",
        "    <tr>\n",
        "      <th>GO:0000014</th>\n",
        "      <td>  1</td>\n",
        "      <td> 0</td>\n",
        "      <td> 0</td>\n",
        "      <td> single-stranded DNA endodeoxyribonuclease acti...</td>\n",
        "    </tr>\n",
        "    <tr>\n",
        "      <th>GO:0000060</th>\n",
        "      <td>  1</td>\n",
        "      <td> 0</td>\n",
        "      <td> 0</td>\n",
        "      <td>        protein import into nucleus, translocation</td>\n",
        "    </tr>\n",
        "    <tr>\n",
        "      <th>GO:0000075</th>\n",
        "      <td>  2</td>\n",
        "      <td> 0</td>\n",
        "      <td> 0</td>\n",
        "      <td>                             cell cycle checkpoint</td>\n",
        "    </tr>\n",
        "    <tr>\n",
        "      <th>GO:0000082</th>\n",
        "      <td>  2</td>\n",
        "      <td> 0</td>\n",
        "      <td> 0</td>\n",
        "      <td>             G1/S transition of mitotic cell cycle</td>\n",
        "    </tr>\n",
        "    <tr>\n",
        "      <th>GO:0000087</th>\n",
        "      <td>  2</td>\n",
        "      <td> 0</td>\n",
        "      <td> 0</td>\n",
        "      <td>                                   mitotic M phase</td>\n",
        "    </tr>\n",
        "    <tr>\n",
        "      <th>GO:0000088</th>\n",
        "      <td>  1</td>\n",
        "      <td> 0</td>\n",
        "      <td> 0</td>\n",
        "      <td>                                  mitotic prophase</td>\n",
        "    </tr>\n",
        "    <tr>\n",
        "      <th>GO:0000090</th>\n",
        "      <td>  1</td>\n",
        "      <td> 0</td>\n",
        "      <td> 0</td>\n",
        "      <td>                                  mitotic anaphase</td>\n",
        "    </tr>\n",
        "    <tr>\n",
        "      <th>GO:0000096</th>\n",
        "      <td>  1</td>\n",
        "      <td> 0</td>\n",
        "      <td> 0</td>\n",
        "      <td>               sulfur amino acid metabolic process</td>\n",
        "    </tr>\n",
        "    <tr>\n",
        "      <th>GO:0000098</th>\n",
        "      <td>  1</td>\n",
        "      <td> 0</td>\n",
        "      <td> 0</td>\n",
        "      <td>               sulfur amino acid catabolic process</td>\n",
        "    </tr>\n",
        "    <tr>\n",
        "      <th>GO:0000122</th>\n",
        "      <td>  2</td>\n",
        "      <td> 0</td>\n",
        "      <td> 0</td>\n",
        "      <td> negative regulation of transcription from RNA ...</td>\n",
        "    </tr>\n",
        "    <tr>\n",
        "      <th>GO:0000151</th>\n",
        "      <td>  1</td>\n",
        "      <td> 0</td>\n",
        "      <td> 0</td>\n",
        "      <td>                          ubiquitin ligase complex</td>\n",
        "    </tr>\n",
        "    <tr>\n",
        "      <th>GO:0000166</th>\n",
        "      <td> 14</td>\n",
        "      <td> 0</td>\n",
        "      <td> 0</td>\n",
        "      <td>                                nucleotide binding</td>\n",
        "    </tr>\n",
        "    <tr>\n",
        "      <th>GO:0000185</th>\n",
        "      <td>  1</td>\n",
        "      <td> 0</td>\n",
        "      <td> 0</td>\n",
        "      <td>                     activation of MAPKKK activity</td>\n",
        "    </tr>\n",
        "    <tr>\n",
        "      <th>GO:0000255</th>\n",
        "      <td>  1</td>\n",
        "      <td> 0</td>\n",
        "      <td> 0</td>\n",
        "      <td>                       allantoin metabolic process</td>\n",
        "    </tr>\n",
        "    <tr>\n",
        "      <th>GO:0000278</th>\n",
        "      <td>  2</td>\n",
        "      <td> 0</td>\n",
        "      <td> 0</td>\n",
        "      <td>                                mitotic cell cycle</td>\n",
        "    </tr>\n",
        "    <tr>\n",
        "      <th>GO:0000390</th>\n",
        "      <td>  1</td>\n",
        "      <td> 0</td>\n",
        "      <td> 0</td>\n",
        "      <td>                  spliceosomal complex disassembly</td>\n",
        "    </tr>\n",
        "    <tr>\n",
        "      <th>GO:0000398</th>\n",
        "      <td>  2</td>\n",
        "      <td> 0</td>\n",
        "      <td> 0</td>\n",
        "      <td>                    mRNA splicing, via spliceosome</td>\n",
        "    </tr>\n",
        "    <tr>\n",
        "      <th>GO:0000422</th>\n",
        "      <td>  1</td>\n",
        "      <td> 0</td>\n",
        "      <td> 0</td>\n",
        "      <td>                         mitochondrion degradation</td>\n",
        "    </tr>\n",
        "    <tr>\n",
        "      <th>GO:0000724</th>\n",
        "      <td>  1</td>\n",
        "      <td> 0</td>\n",
        "      <td> 0</td>\n",
        "      <td> double-strand break repair via homologous reco...</td>\n",
        "    </tr>\n",
        "    <tr>\n",
        "      <th>GO:0000737</th>\n",
        "      <td>  1</td>\n",
        "      <td> 0</td>\n",
        "      <td> 0</td>\n",
        "      <td>            DNA catabolic process, endonucleolytic</td>\n",
        "    </tr>\n",
        "    <tr>\n",
        "      <th>GO:0000910</th>\n",
        "      <td>  2</td>\n",
        "      <td> 0</td>\n",
        "      <td> 0</td>\n",
        "      <td>                                       cytokinesis</td>\n",
        "    </tr>\n",
        "    <tr>\n",
        "      <th>GO:0000922</th>\n",
        "      <td>  1</td>\n",
        "      <td> 0</td>\n",
        "      <td> 0</td>\n",
        "      <td>                                      spindle pole</td>\n",
        "    </tr>\n",
        "    <tr>\n",
        "      <th>GO:0000979</th>\n",
        "      <td>  1</td>\n",
        "      <td> 0</td>\n",
        "      <td> 0</td>\n",
        "      <td> RNA polymerase II core promoter sequence-speci...</td>\n",
        "    </tr>\n",
        "    <tr>\n",
        "      <th>GO:0001103</th>\n",
        "      <td>  1</td>\n",
        "      <td> 0</td>\n",
        "      <td> 0</td>\n",
        "      <td> RNA polymerase II repressing transcription fac...</td>\n",
        "    </tr>\n",
        "    <tr>\n",
        "      <th>GO:0001104</th>\n",
        "      <td>  1</td>\n",
        "      <td> 0</td>\n",
        "      <td> 0</td>\n",
        "      <td> RNA polymerase II transcription cofactor activity</td>\n",
        "    </tr>\n",
        "    <tr>\n",
        "      <th>GO:0001106</th>\n",
        "      <td>  1</td>\n",
        "      <td> 0</td>\n",
        "      <td> 0</td>\n",
        "      <td> RNA polymerase II transcription corepressor ac...</td>\n",
        "    </tr>\n",
        "    <tr>\n",
        "      <th>GO:0001523</th>\n",
        "      <td>  1</td>\n",
        "      <td> 0</td>\n",
        "      <td> 0</td>\n",
        "      <td>                        retinoid metabolic process</td>\n",
        "    </tr>\n",
        "    <tr>\n",
        "      <th>GO:0001525</th>\n",
        "      <td>  2</td>\n",
        "      <td> 0</td>\n",
        "      <td> 0</td>\n",
        "      <td>                                      angiogenesis</td>\n",
        "    </tr>\n",
        "    <tr>\n",
        "      <th>GO:0001541</th>\n",
        "      <td>  2</td>\n",
        "      <td> 0</td>\n",
        "      <td> 0</td>\n",
        "      <td>                      ovarian follicle development</td>\n",
        "    </tr>\n",
        "    <tr>\n",
        "      <th>GO:0001553</th>\n",
        "      <td>  1</td>\n",
        "      <td> 0</td>\n",
        "      <td> 0</td>\n",
        "      <td>                                     luteinization</td>\n",
        "    </tr>\n",
        "    <tr>\n",
        "      <th>...</th>\n",
        "      <td>...</td>\n",
        "      <td>...</td>\n",
        "      <td>...</td>\n",
        "      <td>...</td>\n",
        "    </tr>\n",
        "    <tr>\n",
        "      <th>GO:0097105</th>\n",
        "      <td>  1</td>\n",
        "      <td> 0</td>\n",
        "      <td> 0</td>\n",
        "      <td>                     presynaptic membrane assembly</td>\n",
        "    </tr>\n",
        "    <tr>\n",
        "      <th>GO:0097109</th>\n",
        "      <td>  1</td>\n",
        "      <td> 0</td>\n",
        "      <td> 0</td>\n",
        "      <td>                 neuroligin family protein binding</td>\n",
        "    </tr>\n",
        "    <tr>\n",
        "      <th>GO:0097112</th>\n",
        "      <td>  1</td>\n",
        "      <td> 0</td>\n",
        "      <td> 0</td>\n",
        "      <td>       gamma-aminobutyric acid receptor clustering</td>\n",
        "    </tr>\n",
        "    <tr>\n",
        "      <th>GO:0097114</th>\n",
        "      <td>  1</td>\n",
        "      <td> 0</td>\n",
        "      <td> 0</td>\n",
        "      <td>          N-methyl-D-aspartate receptor clustering</td>\n",
        "    </tr>\n",
        "    <tr>\n",
        "      <th>GO:0097116</th>\n",
        "      <td>  1</td>\n",
        "      <td> 0</td>\n",
        "      <td> 0</td>\n",
        "      <td>                               gephyrin clustering</td>\n",
        "    </tr>\n",
        "    <tr>\n",
        "      <th>GO:0097117</th>\n",
        "      <td>  1</td>\n",
        "      <td> 0</td>\n",
        "      <td> 0</td>\n",
        "      <td>    guanylate kinase-associated protein clustering</td>\n",
        "    </tr>\n",
        "    <tr>\n",
        "      <th>GO:0097118</th>\n",
        "      <td>  1</td>\n",
        "      <td> 0</td>\n",
        "      <td> 0</td>\n",
        "      <td>                             neuroligin clustering</td>\n",
        "    </tr>\n",
        "    <tr>\n",
        "      <th>GO:0097119</th>\n",
        "      <td>  1</td>\n",
        "      <td> 0</td>\n",
        "      <td> 0</td>\n",
        "      <td>        postsynaptic density protein 95 clustering</td>\n",
        "    </tr>\n",
        "    <tr>\n",
        "      <th>GO:0097120</th>\n",
        "      <td>  1</td>\n",
        "      <td> 0</td>\n",
        "      <td> 0</td>\n",
        "      <td>                  receptor localization to synapse</td>\n",
        "    </tr>\n",
        "    <tr>\n",
        "      <th>GO:0097286</th>\n",
        "      <td>  1</td>\n",
        "      <td> 0</td>\n",
        "      <td> 0</td>\n",
        "      <td>                                   iron ion import</td>\n",
        "    </tr>\n",
        "    <tr>\n",
        "      <th>GO:0097449</th>\n",
        "      <td>  1</td>\n",
        "      <td> 0</td>\n",
        "      <td> 0</td>\n",
        "      <td>                              astrocyte projection</td>\n",
        "    </tr>\n",
        "    <tr>\n",
        "      <th>GO:0098655</th>\n",
        "      <td>  1</td>\n",
        "      <td> 0</td>\n",
        "      <td> 0</td>\n",
        "      <td>                    cation transmembrane transport</td>\n",
        "    </tr>\n",
        "    <tr>\n",
        "      <th>GO:1900025</th>\n",
        "      <td>  1</td>\n",
        "      <td> 0</td>\n",
        "      <td> 0</td>\n",
        "      <td> negative regulation of substrate adhesion-depe...</td>\n",
        "    </tr>\n",
        "    <tr>\n",
        "      <th>GO:1902186</th>\n",
        "      <td>  1</td>\n",
        "      <td> 0</td>\n",
        "      <td> 0</td>\n",
        "      <td>        regulation of viral release from host cell</td>\n",
        "    </tr>\n",
        "    <tr>\n",
        "      <th>GO:1902305</th>\n",
        "      <td>  1</td>\n",
        "      <td> 0</td>\n",
        "      <td> 0</td>\n",
        "      <td>  regulation of sodium ion transmembrane transport</td>\n",
        "    </tr>\n",
        "    <tr>\n",
        "      <th>GO:1902560</th>\n",
        "      <td>  1</td>\n",
        "      <td> 0</td>\n",
        "      <td> 0</td>\n",
        "      <td>                             GMP reductase complex</td>\n",
        "    </tr>\n",
        "    <tr>\n",
        "      <th>GO:1902749</th>\n",
        "      <td>  1</td>\n",
        "      <td> 0</td>\n",
        "      <td> 0</td>\n",
        "      <td>    regulation of cell cycle G2/M phase transition</td>\n",
        "    </tr>\n",
        "    <tr>\n",
        "      <th>GO:1902911</th>\n",
        "      <td>  1</td>\n",
        "      <td> 0</td>\n",
        "      <td> 0</td>\n",
        "      <td>                            protein kinase complex</td>\n",
        "    </tr>\n",
        "    <tr>\n",
        "      <th>GO:1990090</th>\n",
        "      <td>  1</td>\n",
        "      <td> 0</td>\n",
        "      <td> 0</td>\n",
        "      <td> cellular response to nerve growth factor stimulus</td>\n",
        "    </tr>\n",
        "    <tr>\n",
        "      <th>GO:2000114</th>\n",
        "      <td>  1</td>\n",
        "      <td> 0</td>\n",
        "      <td> 0</td>\n",
        "      <td>      regulation of establishment of cell polarity</td>\n",
        "    </tr>\n",
        "    <tr>\n",
        "      <th>GO:2000134</th>\n",
        "      <td>  1</td>\n",
        "      <td> 0</td>\n",
        "      <td> 0</td>\n",
        "      <td> negative regulation of G1/S transition of mito...</td>\n",
        "    </tr>\n",
        "    <tr>\n",
        "      <th>GO:2000145</th>\n",
        "      <td>  1</td>\n",
        "      <td> 0</td>\n",
        "      <td> 0</td>\n",
        "      <td>                       regulation of cell motility</td>\n",
        "    </tr>\n",
        "    <tr>\n",
        "      <th>GO:2000304</th>\n",
        "      <td>  1</td>\n",
        "      <td> 0</td>\n",
        "      <td> 0</td>\n",
        "      <td> positive regulation of ceramide biosynthetic p...</td>\n",
        "    </tr>\n",
        "    <tr>\n",
        "      <th>GO:2000310</th>\n",
        "      <td>  1</td>\n",
        "      <td> 0</td>\n",
        "      <td> 0</td>\n",
        "      <td> regulation of N-methyl-D-aspartate selective g...</td>\n",
        "    </tr>\n",
        "    <tr>\n",
        "      <th>GO:2000311</th>\n",
        "      <td>  1</td>\n",
        "      <td> 0</td>\n",
        "      <td> 0</td>\n",
        "      <td> regulation of alpha-amino-3-hydroxy-5-methyl-4...</td>\n",
        "    </tr>\n",
        "    <tr>\n",
        "      <th>GO:2000463</th>\n",
        "      <td>  1</td>\n",
        "      <td> 0</td>\n",
        "      <td> 0</td>\n",
        "      <td> positive regulation of excitatory postsynaptic...</td>\n",
        "    </tr>\n",
        "    <tr>\n",
        "      <th>GO:2000599</th>\n",
        "      <td>  1</td>\n",
        "      <td> 0</td>\n",
        "      <td> 0</td>\n",
        "      <td>   negative regulation of cyclin catabolic process</td>\n",
        "    </tr>\n",
        "    <tr>\n",
        "      <th>GO:2001014</th>\n",
        "      <td>  1</td>\n",
        "      <td> 0</td>\n",
        "      <td> 0</td>\n",
        "      <td> regulation of skeletal muscle cell differentia...</td>\n",
        "    </tr>\n",
        "    <tr>\n",
        "      <th>GO:2001234</th>\n",
        "      <td>  1</td>\n",
        "      <td> 0</td>\n",
        "      <td> 0</td>\n",
        "      <td> negative regulation of apoptotic signaling pat...</td>\n",
        "    </tr>\n",
        "    <tr>\n",
        "      <th>GO:2001237</th>\n",
        "      <td>  1</td>\n",
        "      <td> 0</td>\n",
        "      <td> 0</td>\n",
        "      <td> negative regulation of extrinsic apoptotic sig...</td>\n",
        "    </tr>\n",
        "  </tbody>\n",
        "</table>\n",
        "<p>888 rows \u00d7 4 columns</p>\n",
        "</div>"
       ],
       "metadata": {},
       "output_type": "pyout",
       "prompt_number": 1077,
       "text": [
        "               n_genes  out_of  rank  \\\n",
        "go_identifier                          \n",
        "GO:0000014           1       0     0   \n",
        "GO:0000060           1       0     0   \n",
        "GO:0000075           2       0     0   \n",
        "GO:0000082           2       0     0   \n",
        "GO:0000087           2       0     0   \n",
        "GO:0000088           1       0     0   \n",
        "GO:0000090           1       0     0   \n",
        "GO:0000096           1       0     0   \n",
        "GO:0000098           1       0     0   \n",
        "GO:0000122           2       0     0   \n",
        "GO:0000151           1       0     0   \n",
        "GO:0000166          14       0     0   \n",
        "GO:0000185           1       0     0   \n",
        "GO:0000255           1       0     0   \n",
        "GO:0000278           2       0     0   \n",
        "GO:0000390           1       0     0   \n",
        "GO:0000398           2       0     0   \n",
        "GO:0000422           1       0     0   \n",
        "GO:0000724           1       0     0   \n",
        "GO:0000737           1       0     0   \n",
        "GO:0000910           2       0     0   \n",
        "GO:0000922           1       0     0   \n",
        "GO:0000979           1       0     0   \n",
        "GO:0001103           1       0     0   \n",
        "GO:0001104           1       0     0   \n",
        "GO:0001106           1       0     0   \n",
        "GO:0001523           1       0     0   \n",
        "GO:0001525           2       0     0   \n",
        "GO:0001541           2       0     0   \n",
        "GO:0001553           1       0     0   \n",
        "...                ...     ...   ...   \n",
        "GO:0097105           1       0     0   \n",
        "GO:0097109           1       0     0   \n",
        "GO:0097112           1       0     0   \n",
        "GO:0097114           1       0     0   \n",
        "GO:0097116           1       0     0   \n",
        "GO:0097117           1       0     0   \n",
        "GO:0097118           1       0     0   \n",
        "GO:0097119           1       0     0   \n",
        "GO:0097120           1       0     0   \n",
        "GO:0097286           1       0     0   \n",
        "GO:0097449           1       0     0   \n",
        "GO:0098655           1       0     0   \n",
        "GO:1900025           1       0     0   \n",
        "GO:1902186           1       0     0   \n",
        "GO:1902305           1       0     0   \n",
        "GO:1902560           1       0     0   \n",
        "GO:1902749           1       0     0   \n",
        "GO:1902911           1       0     0   \n",
        "GO:1990090           1       0     0   \n",
        "GO:2000114           1       0     0   \n",
        "GO:2000134           1       0     0   \n",
        "GO:2000145           1       0     0   \n",
        "GO:2000304           1       0     0   \n",
        "GO:2000310           1       0     0   \n",
        "GO:2000311           1       0     0   \n",
        "GO:2000463           1       0     0   \n",
        "GO:2000599           1       0     0   \n",
        "GO:2001014           1       0     0   \n",
        "GO:2001234           1       0     0   \n",
        "GO:2001237           1       0     0   \n",
        "\n",
        "                                                         go_name  \n",
        "go_identifier                                                     \n",
        "GO:0000014     single-stranded DNA endodeoxyribonuclease acti...  \n",
        "GO:0000060            protein import into nucleus, translocation  \n",
        "GO:0000075                                 cell cycle checkpoint  \n",
        "GO:0000082                 G1/S transition of mitotic cell cycle  \n",
        "GO:0000087                                       mitotic M phase  \n",
        "GO:0000088                                      mitotic prophase  \n",
        "GO:0000090                                      mitotic anaphase  \n",
        "GO:0000096                   sulfur amino acid metabolic process  \n",
        "GO:0000098                   sulfur amino acid catabolic process  \n",
        "GO:0000122     negative regulation of transcription from RNA ...  \n",
        "GO:0000151                              ubiquitin ligase complex  \n",
        "GO:0000166                                    nucleotide binding  \n",
        "GO:0000185                         activation of MAPKKK activity  \n",
        "GO:0000255                           allantoin metabolic process  \n",
        "GO:0000278                                    mitotic cell cycle  \n",
        "GO:0000390                      spliceosomal complex disassembly  \n",
        "GO:0000398                        mRNA splicing, via spliceosome  \n",
        "GO:0000422                             mitochondrion degradation  \n",
        "GO:0000724     double-strand break repair via homologous reco...  \n",
        "GO:0000737                DNA catabolic process, endonucleolytic  \n",
        "GO:0000910                                           cytokinesis  \n",
        "GO:0000922                                          spindle pole  \n",
        "GO:0000979     RNA polymerase II core promoter sequence-speci...  \n",
        "GO:0001103     RNA polymerase II repressing transcription fac...  \n",
        "GO:0001104     RNA polymerase II transcription cofactor activity  \n",
        "GO:0001106     RNA polymerase II transcription corepressor ac...  \n",
        "GO:0001523                            retinoid metabolic process  \n",
        "GO:0001525                                          angiogenesis  \n",
        "GO:0001541                          ovarian follicle development  \n",
        "GO:0001553                                         luteinization  \n",
        "...                                                          ...  \n",
        "GO:0097105                         presynaptic membrane assembly  \n",
        "GO:0097109                     neuroligin family protein binding  \n",
        "GO:0097112           gamma-aminobutyric acid receptor clustering  \n",
        "GO:0097114              N-methyl-D-aspartate receptor clustering  \n",
        "GO:0097116                                   gephyrin clustering  \n",
        "GO:0097117        guanylate kinase-associated protein clustering  \n",
        "GO:0097118                                 neuroligin clustering  \n",
        "GO:0097119            postsynaptic density protein 95 clustering  \n",
        "GO:0097120                      receptor localization to synapse  \n",
        "GO:0097286                                       iron ion import  \n",
        "GO:0097449                                  astrocyte projection  \n",
        "GO:0098655                        cation transmembrane transport  \n",
        "GO:1900025     negative regulation of substrate adhesion-depe...  \n",
        "GO:1902186            regulation of viral release from host cell  \n",
        "GO:1902305      regulation of sodium ion transmembrane transport  \n",
        "GO:1902560                                 GMP reductase complex  \n",
        "GO:1902749        regulation of cell cycle G2/M phase transition  \n",
        "GO:1902911                                protein kinase complex  \n",
        "GO:1990090     cellular response to nerve growth factor stimulus  \n",
        "GO:2000114          regulation of establishment of cell polarity  \n",
        "GO:2000134     negative regulation of G1/S transition of mito...  \n",
        "GO:2000145                           regulation of cell motility  \n",
        "GO:2000304     positive regulation of ceramide biosynthetic p...  \n",
        "GO:2000310     regulation of N-methyl-D-aspartate selective g...  \n",
        "GO:2000311     regulation of alpha-amino-3-hydroxy-5-methyl-4...  \n",
        "GO:2000463     positive regulation of excitatory postsynaptic...  \n",
        "GO:2000599       negative regulation of cyclin catabolic process  \n",
        "GO:2001014     regulation of skeletal muscle cell differentia...  \n",
        "GO:2001234     negative regulation of apoptotic signaling pat...  \n",
        "GO:2001237     negative regulation of extrinsic apoptotic sig...  \n",
        "\n",
        "[888 rows x 4 columns]"
       ]
      }
     ],
     "prompt_number": 1077
    },
    {
     "cell_type": "code",
     "collapsed": false,
     "input": [
      "def permut_assoc(rod_s, rnd, gene_df, gene_to_go, top_n, max_dist):\n",
      "    \"\"\"\n",
      "    This is the main function.\n",
      "    Use with rnd = 0 to get the real assoc.\n",
      "    \"\"\"\n",
      "    s = shift_rod(rod_s, rnd)\n",
      "    s.sort(ascending=False, inplace=True)\n",
      "    top_s = s.iloc[:top_n]\n",
      "    cand_genes = get_genes(top_s, gene_df, max_dist=max_dist)\n",
      "    assoc = get_go_assoc(cand_genes, gene_to_go)\n",
      "    return assoc\n",
      "\n",
      "def get_go_assoc(gene_ls, gene_to_go):\n",
      "    \"\"\"\n",
      "    Get series with number of genes associated with each\n",
      "    category in gene_to_go\n",
      "    \"\"\"\n",
      "    s = gene_to_go.set_index(\"gene_symbol\").ix[gene_ls].groupby(\"go_identifier\").apply(len)\n",
      "    return s\n",
      "\n",
      "def multiple_permut_assoc(rod_s, gene_df, gene_to_go, top_n, max_dist, n_runs, rnd_seed=None):\n",
      "    if rnd_seed is not None:\n",
      "        np.random.seed(rnd_seed)\n",
      "    assoc_table = pd.concat([permut_assoc(rod_s, rnd, gene_df, gene_to_go, top_n, max_dist) for rnd in np.random.rand(n_runs)],axis=1)\n",
      "    assoc_table = assoc_table.fillna(0)\n",
      "    return assoc_table\n",
      "\n",
      "def save_permut_assoc_table(assoc_table,fn):\n",
      "    assoc_table.to_csv(fn)\n",
      "\n",
      "def get_initial_rank_table(real_assoc):\n",
      "    return pd.DataFrame({\"n_genes\":real_assoc.values,\"rank\":0,\"out_of\":0},index=real_assoc.index)\n",
      "\n",
      "def get_p_val(rank_table):\n",
      "    \"\"\"\n",
      "    Input:\n",
      "    \n",
      "    \"\"\"\n",
      "    r =  1-rank_table[\"rank\"]*1./(rank_table[\"out_of\"]+1)\n",
      "    r.sort()\n",
      "    return r\n",
      "\n",
      "def update_rank(rank_table,permut_assoc_fh):\n",
      "    assoc_table = pd.read_csv(permut_assoc_fh,index_col=0)\n",
      "    r = assoc_table.apply(lambda row: get_rank(row,rank_table),axis=1)\n",
      "    return r\n",
      "    \n",
      "def total_rank(rank_table, permut_fns):\n",
      "    rt = rank_table.copy()\n",
      "    for f in permut_fns:\n",
      "        rt = update_rank(rt,permut_assoc_fh)\n",
      "    return rt\n",
      "    \n",
      "#def save_p_value_total(rank_table, permut_fns):\n",
      "    \n",
      "    \n",
      "def empirical_rank(value,dist):\n",
      "    \"\"\"\n",
      "    get empirical p value of\n",
      "    value with respect to list of \n",
      "    values in dist\n",
      "    \"\"\"\n",
      "    array = np.append(value,dist)\n",
      "    temp = array.argsort()\n",
      "    ranks = np.empty(len(array), int)\n",
      "    ranks[temp] = np.arange(len(array))\n",
      "    return ranks[0]\n",
      "\n",
      "def get_rank(series,rank_df):\n",
      "    go = series.name\n",
      "    try:\n",
      "        go_s = rank_df.ix[go]\n",
      "    except KeyError:\n",
      "        go_s = pd.Series({\"n_genes\":0,\"rank\":0,\"out_of\":0})\n",
      "    real_val = go_s[\"n_genes\"]\n",
      "    old_rank = go_s[\"rank\"]\n",
      "    old_out_of = go_s[\"out_of\"]\n",
      "    rank = empirical_rank(real_val,series.values)\n",
      "    new_rank = old_rank + rank\n",
      "    new_out_of = old_out_of + len(series)\n",
      "    return pd.Series({\"n_genes\":real_val,\"rank\":new_rank,\"out_of\":new_out_of})\n",
      "\n",
      "def get_genes_per_go(gene_ls,gene_to_go):\n",
      "    s = gene_to_go.set_index(\"gene_symbol\").ix[gene_ls].groupby(\"go_identifier\").apply(lambda x: list(x.index))\n",
      "    s.name = \"genes\"\n",
      "    return sqst"
     ],
     "language": "python",
     "metadata": {},
     "outputs": [],
     "prompt_number": 1199
    },
    {
     "cell_type": "code",
     "collapsed": false,
     "input": [
      "r = get_go_assoc(cand_genes,gene_to_go)\n",
      "rank_table = get_initial_rank_table(r)"
     ],
     "language": "python",
     "metadata": {},
     "outputs": [],
     "prompt_number": 1195
    },
    {
     "cell_type": "code",
     "collapsed": false,
     "input": [
      "genes_per_go = pd.Series(r0.index.values,index=r0.index)"
     ],
     "language": "python",
     "metadata": {},
     "outputs": [],
     "prompt_number": 1178
    },
    {
     "cell_type": "code",
     "collapsed": false,
     "input": [],
     "language": "python",
     "metadata": {},
     "outputs": [],
     "prompt_number": 1190
    },
    {
     "cell_type": "code",
     "collapsed": false,
     "input": [
      "pd.concat([rank_table,gg], axis =1)"
     ],
     "language": "python",
     "metadata": {},
     "outputs": [
      {
       "html": [
        "<div style=\"max-height:1000px;max-width:1500px;overflow:auto;\">\n",
        "<table border=\"1\" class=\"dataframe\">\n",
        "  <thead>\n",
        "    <tr style=\"text-align: right;\">\n",
        "      <th></th>\n",
        "      <th>n_genes</th>\n",
        "      <th>out_of</th>\n",
        "      <th>rank</th>\n",
        "      <th>genes</th>\n",
        "    </tr>\n",
        "    <tr>\n",
        "      <th>go_identifier</th>\n",
        "      <th></th>\n",
        "      <th></th>\n",
        "      <th></th>\n",
        "      <th></th>\n",
        "    </tr>\n",
        "  </thead>\n",
        "  <tbody>\n",
        "    <tr>\n",
        "      <th>GO:0000075</th>\n",
        "      <td>  1</td>\n",
        "      <td> 0</td>\n",
        "      <td> 0</td>\n",
        "      <td>                                          [ANAPC1]</td>\n",
        "    </tr>\n",
        "    <tr>\n",
        "      <th>GO:0000079</th>\n",
        "      <td>  1</td>\n",
        "      <td> 0</td>\n",
        "      <td> 0</td>\n",
        "      <td>                                           [HERC5]</td>\n",
        "    </tr>\n",
        "    <tr>\n",
        "      <th>GO:0000087</th>\n",
        "      <td>  1</td>\n",
        "      <td> 0</td>\n",
        "      <td> 0</td>\n",
        "      <td>                                          [ANAPC1]</td>\n",
        "    </tr>\n",
        "    <tr>\n",
        "      <th>GO:0000090</th>\n",
        "      <td>  1</td>\n",
        "      <td> 0</td>\n",
        "      <td> 0</td>\n",
        "      <td>                                          [ANAPC1]</td>\n",
        "    </tr>\n",
        "    <tr>\n",
        "      <th>GO:0000122</th>\n",
        "      <td>  2</td>\n",
        "      <td> 0</td>\n",
        "      <td> 0</td>\n",
        "      <td>                                     [TP63, TENM2]</td>\n",
        "    </tr>\n",
        "    <tr>\n",
        "      <th>GO:0000139</th>\n",
        "      <td>  2</td>\n",
        "      <td> 0</td>\n",
        "      <td> 0</td>\n",
        "      <td>                                  [HS3ST4, SCARA3]</td>\n",
        "    </tr>\n",
        "    <tr>\n",
        "      <th>GO:0000166</th>\n",
        "      <td> 24</td>\n",
        "      <td> 0</td>\n",
        "      <td> 0</td>\n",
        "      <td> [SCN8A, RAPGEF3, SYK, RFK, EPHB1, RYK, ATP13A4...</td>\n",
        "    </tr>\n",
        "    <tr>\n",
        "      <th>GO:0000185</th>\n",
        "      <td>  1</td>\n",
        "      <td> 0</td>\n",
        "      <td> 0</td>\n",
        "      <td>                                            [GRM5]</td>\n",
        "    </tr>\n",
        "    <tr>\n",
        "      <th>GO:0000187</th>\n",
        "      <td>  2</td>\n",
        "      <td> 0</td>\n",
        "      <td> 0</td>\n",
        "      <td>                                      [SYK, NTRK3]</td>\n",
        "    </tr>\n",
        "    <tr>\n",
        "      <th>GO:0000226</th>\n",
        "      <td>  1</td>\n",
        "      <td> 0</td>\n",
        "      <td> 0</td>\n",
        "      <td>                                           [PARD3]</td>\n",
        "    </tr>\n",
        "    <tr>\n",
        "      <th>GO:0000278</th>\n",
        "      <td>  1</td>\n",
        "      <td> 0</td>\n",
        "      <td> 0</td>\n",
        "      <td>                                          [ANAPC1]</td>\n",
        "    </tr>\n",
        "    <tr>\n",
        "      <th>GO:0000287</th>\n",
        "      <td>  1</td>\n",
        "      <td> 0</td>\n",
        "      <td> 0</td>\n",
        "      <td>                                          [ATP8A2]</td>\n",
        "    </tr>\n",
        "    <tr>\n",
        "      <th>GO:0000506</th>\n",
        "      <td>  1</td>\n",
        "      <td> 0</td>\n",
        "      <td> 0</td>\n",
        "      <td>                                            [PIGY]</td>\n",
        "    </tr>\n",
        "    <tr>\n",
        "      <th>GO:0000723</th>\n",
        "      <td>  1</td>\n",
        "      <td> 0</td>\n",
        "      <td> 0</td>\n",
        "      <td>                                           [XRCC5]</td>\n",
        "    </tr>\n",
        "    <tr>\n",
        "      <th>GO:0000783</th>\n",
        "      <td>  1</td>\n",
        "      <td> 0</td>\n",
        "      <td> 0</td>\n",
        "      <td>                                           [XRCC5]</td>\n",
        "    </tr>\n",
        "    <tr>\n",
        "      <th>GO:0000784</th>\n",
        "      <td>  1</td>\n",
        "      <td> 0</td>\n",
        "      <td> 0</td>\n",
        "      <td>                                           [XRCC5]</td>\n",
        "    </tr>\n",
        "    <tr>\n",
        "      <th>GO:0000785</th>\n",
        "      <td>  1</td>\n",
        "      <td> 0</td>\n",
        "      <td> 0</td>\n",
        "      <td>                                            [TP63]</td>\n",
        "    </tr>\n",
        "    <tr>\n",
        "      <th>GO:0000790</th>\n",
        "      <td>  1</td>\n",
        "      <td> 0</td>\n",
        "      <td> 0</td>\n",
        "      <td>                                            [TP63]</td>\n",
        "    </tr>\n",
        "    <tr>\n",
        "      <th>GO:0000978</th>\n",
        "      <td>  1</td>\n",
        "      <td> 0</td>\n",
        "      <td> 0</td>\n",
        "      <td>                                           [NR3C1]</td>\n",
        "    </tr>\n",
        "    <tr>\n",
        "      <th>GO:0000989</th>\n",
        "      <td>  1</td>\n",
        "      <td> 0</td>\n",
        "      <td> 0</td>\n",
        "      <td>                                            [TP63]</td>\n",
        "    </tr>\n",
        "    <tr>\n",
        "      <th>GO:0001077</th>\n",
        "      <td>  3</td>\n",
        "      <td> 0</td>\n",
        "      <td> 0</td>\n",
        "      <td>                              [TP63, NR3C1, CREB5]</td>\n",
        "    </tr>\n",
        "    <tr>\n",
        "      <th>GO:0001401</th>\n",
        "      <td>  1</td>\n",
        "      <td> 0</td>\n",
        "      <td> 0</td>\n",
        "      <td>                                          [SAMM50]</td>\n",
        "    </tr>\n",
        "    <tr>\n",
        "      <th>GO:0001501</th>\n",
        "      <td>  3</td>\n",
        "      <td> 0</td>\n",
        "      <td> 0</td>\n",
        "      <td>                                [TP63, ANKH, ARSE]</td>\n",
        "    </tr>\n",
        "    <tr>\n",
        "      <th>GO:0001508</th>\n",
        "      <td>  1</td>\n",
        "      <td> 0</td>\n",
        "      <td> 0</td>\n",
        "      <td>                                           [AKAP6]</td>\n",
        "    </tr>\n",
        "    <tr>\n",
        "      <th>GO:0001518</th>\n",
        "      <td>  1</td>\n",
        "      <td> 0</td>\n",
        "      <td> 0</td>\n",
        "      <td>                                           [SCN8A]</td>\n",
        "    </tr>\n",
        "    <tr>\n",
        "      <th>GO:0001523</th>\n",
        "      <td>  1</td>\n",
        "      <td> 0</td>\n",
        "      <td> 0</td>\n",
        "      <td>                                            [GPC5]</td>\n",
        "    </tr>\n",
        "    <tr>\n",
        "      <th>GO:0001525</th>\n",
        "      <td>  4</td>\n",
        "      <td> 0</td>\n",
        "      <td> 0</td>\n",
        "      <td>                     [RAPGEF3, SYK, EPHB1, THSD7A]</td>\n",
        "    </tr>\n",
        "    <tr>\n",
        "      <th>GO:0001539</th>\n",
        "      <td>  1</td>\n",
        "      <td> 0</td>\n",
        "      <td> 0</td>\n",
        "      <td>                                          [DNAH17]</td>\n",
        "    </tr>\n",
        "    <tr>\n",
        "      <th>GO:0001541</th>\n",
        "      <td>  1</td>\n",
        "      <td> 0</td>\n",
        "      <td> 0</td>\n",
        "      <td>                                            [FSHB]</td>\n",
        "    </tr>\n",
        "    <tr>\n",
        "      <th>GO:0001558</th>\n",
        "      <td>  1</td>\n",
        "      <td> 0</td>\n",
        "      <td> 0</td>\n",
        "      <td>                                            [NRG3]</td>\n",
        "    </tr>\n",
        "    <tr>\n",
        "      <th>...</th>\n",
        "      <td>...</td>\n",
        "      <td>...</td>\n",
        "      <td>...</td>\n",
        "      <td>...</td>\n",
        "    </tr>\n",
        "    <tr>\n",
        "      <th>GO:1900087</th>\n",
        "      <td>  1</td>\n",
        "      <td> 0</td>\n",
        "      <td> 0</td>\n",
        "      <td>                                         [PPP1R1C]</td>\n",
        "    </tr>\n",
        "    <tr>\n",
        "      <th>GO:1900170</th>\n",
        "      <td>  1</td>\n",
        "      <td> 0</td>\n",
        "      <td> 0</td>\n",
        "      <td>                                           [NR3C1]</td>\n",
        "    </tr>\n",
        "    <tr>\n",
        "      <th>GO:1900180</th>\n",
        "      <td>  1</td>\n",
        "      <td> 0</td>\n",
        "      <td> 0</td>\n",
        "      <td>                                          [RASSF5]</td>\n",
        "    </tr>\n",
        "    <tr>\n",
        "      <th>GO:1901018</th>\n",
        "      <td>  1</td>\n",
        "      <td> 0</td>\n",
        "      <td> 0</td>\n",
        "      <td>                                            [ANK2]</td>\n",
        "    </tr>\n",
        "    <tr>\n",
        "      <th>GO:1901019</th>\n",
        "      <td>  1</td>\n",
        "      <td> 0</td>\n",
        "      <td> 0</td>\n",
        "      <td>                                            [ANK2]</td>\n",
        "    </tr>\n",
        "    <tr>\n",
        "      <th>GO:1901021</th>\n",
        "      <td>  1</td>\n",
        "      <td> 0</td>\n",
        "      <td> 0</td>\n",
        "      <td>                                            [ANK2]</td>\n",
        "    </tr>\n",
        "    <tr>\n",
        "      <th>GO:1901374</th>\n",
        "      <td>  1</td>\n",
        "      <td> 0</td>\n",
        "      <td> 0</td>\n",
        "      <td>                                         [SLC22A1]</td>\n",
        "    </tr>\n",
        "    <tr>\n",
        "      <th>GO:1901379</th>\n",
        "      <td>  1</td>\n",
        "      <td> 0</td>\n",
        "      <td> 0</td>\n",
        "      <td>                                            [DPP6]</td>\n",
        "    </tr>\n",
        "    <tr>\n",
        "      <th>GO:1901381</th>\n",
        "      <td>  2</td>\n",
        "      <td> 0</td>\n",
        "      <td> 0</td>\n",
        "      <td>                                     [DPP6, AKAP6]</td>\n",
        "    </tr>\n",
        "    <tr>\n",
        "      <th>GO:1902187</th>\n",
        "      <td>  1</td>\n",
        "      <td> 0</td>\n",
        "      <td> 0</td>\n",
        "      <td>                                          [TRIM35]</td>\n",
        "    </tr>\n",
        "    <tr>\n",
        "      <th>GO:1902261</th>\n",
        "      <td>  1</td>\n",
        "      <td> 0</td>\n",
        "      <td> 0</td>\n",
        "      <td>                                           [AKAP6]</td>\n",
        "    </tr>\n",
        "    <tr>\n",
        "      <th>GO:1902287</th>\n",
        "      <td>  1</td>\n",
        "      <td> 0</td>\n",
        "      <td> 0</td>\n",
        "      <td>                                          [PLXNA4]</td>\n",
        "    </tr>\n",
        "    <tr>\n",
        "      <th>GO:1902808</th>\n",
        "      <td>  1</td>\n",
        "      <td> 0</td>\n",
        "      <td> 0</td>\n",
        "      <td>                                            [TP63]</td>\n",
        "    </tr>\n",
        "    <tr>\n",
        "      <th>GO:1990254</th>\n",
        "      <td>  1</td>\n",
        "      <td> 0</td>\n",
        "      <td> 0</td>\n",
        "      <td>                                           [KRT74]</td>\n",
        "    </tr>\n",
        "    <tr>\n",
        "      <th>GO:2000010</th>\n",
        "      <td>  1</td>\n",
        "      <td> 0</td>\n",
        "      <td> 0</td>\n",
        "      <td>                                           [GPD1L]</td>\n",
        "    </tr>\n",
        "    <tr>\n",
        "      <th>GO:2000096</th>\n",
        "      <td>  1</td>\n",
        "      <td> 0</td>\n",
        "      <td> 0</td>\n",
        "      <td>                                           [MLLT3]</td>\n",
        "    </tr>\n",
        "    <tr>\n",
        "      <th>GO:2000249</th>\n",
        "      <td>  1</td>\n",
        "      <td> 0</td>\n",
        "      <td> 0</td>\n",
        "      <td>                                         [RAPGEF3]</td>\n",
        "    </tr>\n",
        "    <tr>\n",
        "      <th>GO:2000251</th>\n",
        "      <td>  1</td>\n",
        "      <td> 0</td>\n",
        "      <td> 0</td>\n",
        "      <td>                                           [NTRK3]</td>\n",
        "    </tr>\n",
        "    <tr>\n",
        "      <th>GO:2000271</th>\n",
        "      <td>  1</td>\n",
        "      <td> 0</td>\n",
        "      <td> 0</td>\n",
        "      <td>                                            [TP63]</td>\n",
        "    </tr>\n",
        "    <tr>\n",
        "      <th>GO:2000366</th>\n",
        "      <td>  1</td>\n",
        "      <td> 0</td>\n",
        "      <td> 0</td>\n",
        "      <td>                                           [ERBB4]</td>\n",
        "    </tr>\n",
        "    <tr>\n",
        "      <th>GO:2000381</th>\n",
        "      <td>  1</td>\n",
        "      <td> 0</td>\n",
        "      <td> 0</td>\n",
        "      <td>                                            [TP63]</td>\n",
        "    </tr>\n",
        "    <tr>\n",
        "      <th>GO:2000599</th>\n",
        "      <td>  1</td>\n",
        "      <td> 0</td>\n",
        "      <td> 0</td>\n",
        "      <td>                                           [NELL1]</td>\n",
        "    </tr>\n",
        "    <tr>\n",
        "      <th>GO:2000642</th>\n",
        "      <td>  1</td>\n",
        "      <td> 0</td>\n",
        "      <td> 0</td>\n",
        "      <td>                                            [SNX3]</td>\n",
        "    </tr>\n",
        "    <tr>\n",
        "      <th>GO:2000649</th>\n",
        "      <td>  1</td>\n",
        "      <td> 0</td>\n",
        "      <td> 0</td>\n",
        "      <td>                                           [GPD1L]</td>\n",
        "    </tr>\n",
        "    <tr>\n",
        "      <th>GO:2000773</th>\n",
        "      <td>  1</td>\n",
        "      <td> 0</td>\n",
        "      <td> 0</td>\n",
        "      <td>                                            [TP63]</td>\n",
        "    </tr>\n",
        "    <tr>\n",
        "      <th>GO:2001235</th>\n",
        "      <td>  1</td>\n",
        "      <td> 0</td>\n",
        "      <td> 0</td>\n",
        "      <td>                                            [TP63]</td>\n",
        "    </tr>\n",
        "    <tr>\n",
        "      <th>GO:2001237</th>\n",
        "      <td>  1</td>\n",
        "      <td> 0</td>\n",
        "      <td> 0</td>\n",
        "      <td>                                           [ITGA6]</td>\n",
        "    </tr>\n",
        "    <tr>\n",
        "      <th>GO:2001244</th>\n",
        "      <td>  1</td>\n",
        "      <td> 0</td>\n",
        "      <td> 0</td>\n",
        "      <td>                                          [IL20RA]</td>\n",
        "    </tr>\n",
        "    <tr>\n",
        "      <th>GO:2001257</th>\n",
        "      <td>  1</td>\n",
        "      <td> 0</td>\n",
        "      <td> 0</td>\n",
        "      <td>                                            [ANK2]</td>\n",
        "    </tr>\n",
        "    <tr>\n",
        "      <th>GO:2001259</th>\n",
        "      <td>  1</td>\n",
        "      <td> 0</td>\n",
        "      <td> 0</td>\n",
        "      <td>                                            [ANK2]</td>\n",
        "    </tr>\n",
        "  </tbody>\n",
        "</table>\n",
        "<p>1265 rows \u00d7 4 columns</p>\n",
        "</div>"
       ],
       "metadata": {},
       "output_type": "pyout",
       "prompt_number": 1203,
       "text": [
        "               n_genes  out_of  rank  \\\n",
        "go_identifier                          \n",
        "GO:0000075           1       0     0   \n",
        "GO:0000079           1       0     0   \n",
        "GO:0000087           1       0     0   \n",
        "GO:0000090           1       0     0   \n",
        "GO:0000122           2       0     0   \n",
        "GO:0000139           2       0     0   \n",
        "GO:0000166          24       0     0   \n",
        "GO:0000185           1       0     0   \n",
        "GO:0000187           2       0     0   \n",
        "GO:0000226           1       0     0   \n",
        "GO:0000278           1       0     0   \n",
        "GO:0000287           1       0     0   \n",
        "GO:0000506           1       0     0   \n",
        "GO:0000723           1       0     0   \n",
        "GO:0000783           1       0     0   \n",
        "GO:0000784           1       0     0   \n",
        "GO:0000785           1       0     0   \n",
        "GO:0000790           1       0     0   \n",
        "GO:0000978           1       0     0   \n",
        "GO:0000989           1       0     0   \n",
        "GO:0001077           3       0     0   \n",
        "GO:0001401           1       0     0   \n",
        "GO:0001501           3       0     0   \n",
        "GO:0001508           1       0     0   \n",
        "GO:0001518           1       0     0   \n",
        "GO:0001523           1       0     0   \n",
        "GO:0001525           4       0     0   \n",
        "GO:0001539           1       0     0   \n",
        "GO:0001541           1       0     0   \n",
        "GO:0001558           1       0     0   \n",
        "...                ...     ...   ...   \n",
        "GO:1900087           1       0     0   \n",
        "GO:1900170           1       0     0   \n",
        "GO:1900180           1       0     0   \n",
        "GO:1901018           1       0     0   \n",
        "GO:1901019           1       0     0   \n",
        "GO:1901021           1       0     0   \n",
        "GO:1901374           1       0     0   \n",
        "GO:1901379           1       0     0   \n",
        "GO:1901381           2       0     0   \n",
        "GO:1902187           1       0     0   \n",
        "GO:1902261           1       0     0   \n",
        "GO:1902287           1       0     0   \n",
        "GO:1902808           1       0     0   \n",
        "GO:1990254           1       0     0   \n",
        "GO:2000010           1       0     0   \n",
        "GO:2000096           1       0     0   \n",
        "GO:2000249           1       0     0   \n",
        "GO:2000251           1       0     0   \n",
        "GO:2000271           1       0     0   \n",
        "GO:2000366           1       0     0   \n",
        "GO:2000381           1       0     0   \n",
        "GO:2000599           1       0     0   \n",
        "GO:2000642           1       0     0   \n",
        "GO:2000649           1       0     0   \n",
        "GO:2000773           1       0     0   \n",
        "GO:2001235           1       0     0   \n",
        "GO:2001237           1       0     0   \n",
        "GO:2001244           1       0     0   \n",
        "GO:2001257           1       0     0   \n",
        "GO:2001259           1       0     0   \n",
        "\n",
        "                                                           genes  \n",
        "go_identifier                                                     \n",
        "GO:0000075                                              [ANAPC1]  \n",
        "GO:0000079                                               [HERC5]  \n",
        "GO:0000087                                              [ANAPC1]  \n",
        "GO:0000090                                              [ANAPC1]  \n",
        "GO:0000122                                         [TP63, TENM2]  \n",
        "GO:0000139                                      [HS3ST4, SCARA3]  \n",
        "GO:0000166     [SCN8A, RAPGEF3, SYK, RFK, EPHB1, RYK, ATP13A4...  \n",
        "GO:0000185                                                [GRM5]  \n",
        "GO:0000187                                          [SYK, NTRK3]  \n",
        "GO:0000226                                               [PARD3]  \n",
        "GO:0000278                                              [ANAPC1]  \n",
        "GO:0000287                                              [ATP8A2]  \n",
        "GO:0000506                                                [PIGY]  \n",
        "GO:0000723                                               [XRCC5]  \n",
        "GO:0000783                                               [XRCC5]  \n",
        "GO:0000784                                               [XRCC5]  \n",
        "GO:0000785                                                [TP63]  \n",
        "GO:0000790                                                [TP63]  \n",
        "GO:0000978                                               [NR3C1]  \n",
        "GO:0000989                                                [TP63]  \n",
        "GO:0001077                                  [TP63, NR3C1, CREB5]  \n",
        "GO:0001401                                              [SAMM50]  \n",
        "GO:0001501                                    [TP63, ANKH, ARSE]  \n",
        "GO:0001508                                               [AKAP6]  \n",
        "GO:0001518                                               [SCN8A]  \n",
        "GO:0001523                                                [GPC5]  \n",
        "GO:0001525                         [RAPGEF3, SYK, EPHB1, THSD7A]  \n",
        "GO:0001539                                              [DNAH17]  \n",
        "GO:0001541                                                [FSHB]  \n",
        "GO:0001558                                                [NRG3]  \n",
        "...                                                          ...  \n",
        "GO:1900087                                             [PPP1R1C]  \n",
        "GO:1900170                                               [NR3C1]  \n",
        "GO:1900180                                              [RASSF5]  \n",
        "GO:1901018                                                [ANK2]  \n",
        "GO:1901019                                                [ANK2]  \n",
        "GO:1901021                                                [ANK2]  \n",
        "GO:1901374                                             [SLC22A1]  \n",
        "GO:1901379                                                [DPP6]  \n",
        "GO:1901381                                         [DPP6, AKAP6]  \n",
        "GO:1902187                                              [TRIM35]  \n",
        "GO:1902261                                               [AKAP6]  \n",
        "GO:1902287                                              [PLXNA4]  \n",
        "GO:1902808                                                [TP63]  \n",
        "GO:1990254                                               [KRT74]  \n",
        "GO:2000010                                               [GPD1L]  \n",
        "GO:2000096                                               [MLLT3]  \n",
        "GO:2000249                                             [RAPGEF3]  \n",
        "GO:2000251                                               [NTRK3]  \n",
        "GO:2000271                                                [TP63]  \n",
        "GO:2000366                                               [ERBB4]  \n",
        "GO:2000381                                                [TP63]  \n",
        "GO:2000599                                               [NELL1]  \n",
        "GO:2000642                                                [SNX3]  \n",
        "GO:2000649                                               [GPD1L]  \n",
        "GO:2000773                                                [TP63]  \n",
        "GO:2001235                                                [TP63]  \n",
        "GO:2001237                                               [ITGA6]  \n",
        "GO:2001244                                              [IL20RA]  \n",
        "GO:2001257                                                [ANK2]  \n",
        "GO:2001259                                                [ANK2]  \n",
        "\n",
        "[1265 rows x 4 columns]"
       ]
      }
     ],
     "prompt_number": 1203
    },
    {
     "cell_type": "code",
     "collapsed": false,
     "input": [
      "gg = get_genes_per_go(cand_genes,gene_to_go)"
     ],
     "language": "python",
     "metadata": {},
     "outputs": [],
     "prompt_number": 1202
    },
    {
     "cell_type": "code",
     "collapsed": false,
     "input": [
      "r = permut_assoc(tot_sweed_df[\"Likelihood\"], 0.1, gene_df, gene_to_go, top_n = 500, max_dist=10000)"
     ],
     "language": "python",
     "metadata": {},
     "outputs": [],
     "prompt_number": 927
    },
    {
     "cell_type": "code",
     "collapsed": false,
     "input": [
      "real_assoc = permut_assoc(tot_sweed_df[\"Likelihood\"], 0, gene_df, gene_to_go, top_n=1000, max_dist=10000)"
     ],
     "language": "python",
     "metadata": {},
     "outputs": [],
     "prompt_number": 934
    },
    {
     "cell_type": "code",
     "collapsed": false,
     "input": [
      "assoc_table = multiple_permut_assoc(tot_sweed_df[\"Likelihood\"], gene_df, gene_to_go, top_n=1000, max_dist = 10000, n_runs=100, rnd_seed=None)"
     ],
     "language": "python",
     "metadata": {},
     "outputs": [],
     "prompt_number": 933
    },
    {
     "cell_type": "code",
     "collapsed": false,
     "input": [
      "def f(r):\n",
      "    print r\n",
      "    raise"
     ],
     "language": "python",
     "metadata": {},
     "outputs": [],
     "prompt_number": 952
    },
    {
     "cell_type": "code",
     "collapsed": false,
     "input": [
      "p_vals = assoc_table.apply(lambda r: get_p_val(r,real_assoc),axis=1)"
     ],
     "language": "python",
     "metadata": {},
     "outputs": [],
     "prompt_number": 959
    },
    {
     "cell_type": "code",
     "collapsed": false,
     "input": [
      "p_vals.sort()"
     ],
     "language": "python",
     "metadata": {},
     "outputs": [],
     "prompt_number": 960
    },
    {
     "cell_type": "code",
     "collapsed": false,
     "input": [
      "sab_enrich = pd.read_csv(eu(\"~/vervet_project/analyses/20140611_163_subpop_sweepfinder/_data/enrichment_n2000_d10000_permut50000_sab.tsv\"),\n",
      "                            index_col=[0],sep=\"\\t\")"
     ],
     "language": "python",
     "metadata": {},
     "outputs": [],
     "prompt_number": 1206
    },
    {
     "cell_type": "code",
     "collapsed": false,
     "input": [
      "len(gene_to_go.set_index(\"go_identifier\").index.diff(sab_enrich.index))"
     ],
     "language": "python",
     "metadata": {},
     "outputs": [
      {
       "metadata": {},
       "output_type": "pyout",
       "prompt_number": 1218,
       "text": [
        "432"
       ]
      }
     ],
     "prompt_number": 1218
    },
    {
     "cell_type": "code",
     "collapsed": false,
     "input": [
      "len(sab_enrich)"
     ],
     "language": "python",
     "metadata": {},
     "outputs": [
      {
       "metadata": {},
       "output_type": "pyout",
       "prompt_number": 1220,
       "text": [
        "14566"
       ]
      }
     ],
     "prompt_number": 1220
    },
    {
     "cell_type": "code",
     "collapsed": false,
     "input": [
      "len(gene_to_go.set_index(\"go_identifier\").index.unique())"
     ],
     "language": "python",
     "metadata": {},
     "outputs": [
      {
       "metadata": {},
       "output_type": "pyout",
       "prompt_number": 1222,
       "text": [
        "14998"
       ]
      }
     ],
     "prompt_number": 1222
    },
    {
     "cell_type": "code",
     "collapsed": false,
     "input": [],
     "language": "python",
     "metadata": {},
     "outputs": [],
     "prompt_number": 1223
    },
    {
     "cell_type": "heading",
     "level": 2,
     "metadata": {},
     "source": [
      "Development for test_enrichment.py\n"
     ]
    },
    {
     "cell_type": "code",
     "collapsed": false,
     "input": [
      "from hs_vervet.scripts import test_enrichment as te"
     ],
     "language": "python",
     "metadata": {},
     "outputs": [],
     "prompt_number": 4
    },
    {
     "cell_type": "code",
     "collapsed": false,
     "input": [
      "from scipy import stats"
     ],
     "language": "python",
     "metadata": {},
     "outputs": [],
     "prompt_number": 10
    },
    {
     "cell_type": "code",
     "collapsed": false,
     "input": [
      "pop = \"aet\"\n",
      "tot_sweed_df = pd.read_csv(eu(\"~/vervet_project/analyses/20140611_163_subpop_sweepfinder/\"\n",
      "                                 \"_data/tot_sweed_result_{}.tsv\".format(pop)),sep=\"\\t\",index_col=[0,1])"
     ],
     "language": "python",
     "metadata": {},
     "outputs": []
    },
    {
     "cell_type": "code",
     "collapsed": false,
     "input": [
      "def gene_likelihood(outlier_locations,mode=\"max\"):\n",
      "    \"\"\"\n",
      "    get genes that overlap with the multi-index of the provided data frame\n",
      "    \"\"\"\n",
      "    modes = [\"max\",\"sum\"]\n",
      "    assert mode in modes\n",
      "    gene_hit_df = pd.DataFrame()\n",
      "    for chrom in outlier_locations.index.droplevel(1).unique():\n",
      "        pos_rel_to_start = gene_df.ix[chrom].index.searchsorted(outlier_locations.ix[chrom].index).values\n",
      "        pos_rel_to_end = np.searchsorted(gene_df.ix[chrom][\"end\"].values,outlier_locations.ix[chrom].index.values)\n",
      "        in_gene = (pos_rel_to_start - pos_rel_to_end) == 1\n",
      "        gene_hits = gene_df.ix[chrom].iloc[pos_rel_to_end[in_gene]]\n",
      "        gene_hits['chrom'] = chrom\n",
      "        gene_hits.set_index('chrom', append=True, inplace=True)\n",
      "        gene_hits = gene_hits.reorder_levels(['chrom', 'pos'])\n",
      "        gene_hits[\"likelihood\"] = outlier_locations.ix[chrom][in_gene][\"Likelihood\"].values\n",
      "        gene_hit_df = gene_hit_df.append(gene_hits)\n",
      "    #if mode == \"max\":    \n",
      "    return gene_hit_df.groupby(lambda x:x).max()\n",
      "    #elif mode == \"sum\":\n",
      "    #    return gene_hit_df.groupby(lambda x:x).sum()"
     ],
     "language": "python",
     "metadata": {},
     "outputs": [],
     "prompt_number": 79
    },
    {
     "cell_type": "code",
     "collapsed": false,
     "input": [
      "def gene_stats(rod_s,stats=None,stat_names=None):\n",
      "    \"\"\"\n",
      "    cacluclate statistics on the ROD series across genes\n",
      "    stats default: [np.max,np.nanmean,len]\n",
      "    \"\"\"\n",
      "    if stats == None:\n",
      "        stats = [np.max,np.nanmean,len]\n",
      "    def stats_apply(df):\n",
      "        #name = df.index.values[0]\n",
      "        c_dic = {\"gene_id\":df.iloc[0][\"gene_id\"],\"end\":df.iloc[0][\"end\"]}\n",
      "        if stat_names is None:\n",
      "            c_dic.update({stat.__name__:stat(df[\"value\"]) for stat in stats})\n",
      "        else:\n",
      "            c_dic.update({n:stat(df[\"value\"]) for n,stat in zip(stat_name,stats)})\n",
      "        s = pd.Series(c_dic)\n",
      "        return s\n",
      "    gene_hit_df = pd.DataFrame()\n",
      "    for chrom in rod_s.index.droplevel(1).unique():\n",
      "        pos_rel_to_start = gene_df.ix[chrom].index.searchsorted(rod_s.ix[chrom].index).values\n",
      "        pos_rel_to_end = np.searchsorted(gene_df.ix[chrom][\"end\"].values,rod_s.ix[chrom].index.values)\n",
      "        in_gene = (pos_rel_to_start - pos_rel_to_end) == 1\n",
      "        gene_hits = gene_df.ix[chrom].iloc[pos_rel_to_end[in_gene]]\n",
      "        gene_hits['chrom'] = chrom\n",
      "        gene_hits.set_index('chrom', append=True, inplace=True)\n",
      "        gene_hits = gene_hits.reorder_levels(['chrom', 'pos'])\n",
      "        gene_hits[\"value\"] = rod_s.ix[chrom][in_gene].values\n",
      "        gene_hits = gene_hits.groupby(lambda x:x).apply(stats_apply)\n",
      "        gene_hit_df = gene_hit_df.append(gene_hits)   \n",
      "    gene_hit_df.index = pd.MultiIndex.from_tuples(gene_hit_df.index)\n",
      "    return gene_hit_df"
     ],
     "language": "python",
     "metadata": {},
     "outputs": [],
     "prompt_number": 197
    },
    {
     "cell_type": "code",
     "collapsed": false,
     "input": [
      "def gene_stats_gb(rod_s,stats=None,stat_names=None):\n",
      "    \"\"\"\n",
      "    #implement avoiding loop\n",
      "    cacluclate statistics on the ROD series across genes\n",
      "    stats default: [np.max,np.nanmean,len]\n",
      "    \"\"\"\n",
      "    if stats == None:\n",
      "        stats = [np.max,np.nanmean,len]\n",
      "    def stats_apply(df):\n",
      "        #name = df.index.values[0]\n",
      "        c_dic = {\"gene_id\":df.iloc[0][\"gene_id\"],\"end\":df.iloc[0][\"end\"]}\n",
      "        if stat_names is None:\n",
      "            c_dic.update({stat.__name__:stat(df[\"value\"]) for stat in stats})\n",
      "        else:\n",
      "            c_dic.update({n:stat(df[\"value\"]) for n,stat in zip(stat_name,stats)})\n",
      "        s = pd.Series(c_dic)\n",
      "        return s\n",
      "    def get_gene_hits(rod_s):\n",
      "        #print rod_s.shape\n",
      "        chrom = rod_s.index.values[0][0]\n",
      "        #print chrom\n",
      "        pos_rel_to_start = gene_df.ix[chrom].index.searchsorted(rod_s.ix[chrom].index).values\n",
      "        pos_rel_to_end = np.searchsorted(gene_df.ix[chrom][\"end\"].values,rod_s.ix[chrom].index.values)\n",
      "        in_gene = (pos_rel_to_start - pos_rel_to_end) == 1\n",
      "        gene_hits = gene_df.ix[chrom].iloc[pos_rel_to_end[in_gene]]\n",
      "        gene_hits[\"value\"] = rod_s.ix[chrom][in_gene].values\n",
      "        gene_hits = gene_hits.groupby(lambda x:x).apply(stats_apply)\n",
      "        return gene_hits   \n",
      "    gene_hit_df = rod_s.groupby(lambda x: x[0]).apply(get_gene_hits)\n",
      "    return gene_hit_df"
     ],
     "language": "python",
     "metadata": {},
     "outputs": [],
     "prompt_number": 205
    },
    {
     "cell_type": "code",
     "collapsed": false,
     "input": [
      "tot_sweed_s = tot_sweed_df[\"Likelihood\"]"
     ],
     "language": "python",
     "metadata": {},
     "outputs": [],
     "prompt_number": 210
    },
    {
     "cell_type": "code",
     "collapsed": false,
     "input": [
      "fn0 = open(eu(\"~/vervet_project/analyses/20140611_163_subpop_sweepfinder/_data/sweed_go_enrichment_q0.001_d10000_0_pys.temp.tsv\"))\n",
      "fn1 = open(eu(\"~/vervet_project/analyses/20140611_163_subpop_sweepfinder/_data/sweed_go_enrichment_q0.001_d10000_1_pys.temp.tsv\"))\n",
      "tot_rank = te.read_table(fn0,index_col=0)\n",
      "rank_table = te.read_table(fn1,index_col=0)"
     ],
     "language": "python",
     "metadata": {},
     "outputs": [],
     "prompt_number": 17
    },
    {
     "cell_type": "code",
     "collapsed": false,
     "input": [
      "tot_rank.index.diff(rank_table.index)"
     ],
     "language": "python",
     "metadata": {},
     "outputs": [
      {
       "metadata": {},
       "output_type": "pyout",
       "prompt_number": 20,
       "text": [
        "Index([u'000'], dtype='object')"
       ]
      }
     ],
     "prompt_number": 20
    },
    {
     "cell_type": "code",
     "collapsed": false,
     "input": [
      "top_n = 1000\n",
      "value_s = tot_sweed_df[\"Likelihood\"].sort(ascending=False,inplace=False)\n",
      "top_s = value_s.iloc[:top_n]"
     ],
     "language": "python",
     "metadata": {},
     "outputs": [],
     "prompt_number": 14
    },
    {
     "cell_type": "code",
     "collapsed": false,
     "input": [
      "max_dist = 10000\n",
      "cand_genes = te.get_genes(top_s, gene_df, max_dist=10000)"
     ],
     "language": "python",
     "metadata": {},
     "outputs": [],
     "prompt_number": 16
    },
    {
     "cell_type": "code",
     "collapsed": false,
     "input": [
      "real_assoc = te.get_go_assoc(cand_genes, gene_to_go)"
     ],
     "language": "python",
     "metadata": {},
     "outputs": [],
     "prompt_number": 19
    },
    {
     "cell_type": "code",
     "collapsed": false,
     "input": [
      "init_rank_table = te.get_initial_rank_table(real_assoc)"
     ],
     "language": "python",
     "metadata": {},
     "outputs": [],
     "prompt_number": 21
    },
    {
     "cell_type": "heading",
     "level": 2,
     "metadata": {},
     "source": [
      "Enrichment according to values across gene (mean, max)..."
     ]
    },
    {
     "cell_type": "code",
     "collapsed": false,
     "input": [
      "from hs_vervet.tools import hs_pandas as hp\n",
      "from hs_vervet.scripts import test_enrichment as te\n",
      "assoc_dir = eu(\"~/vervet_project/data/gene_associations\")"
     ],
     "language": "python",
     "metadata": {},
     "outputs": [],
     "prompt_number": 24
    },
    {
     "cell_type": "code",
     "collapsed": false,
     "input": [
      "gene_to_cat = pd.read_csv(jn(assoc_dir,\"gene_to_cat.csv\"),index_col=0)"
     ],
     "language": "python",
     "metadata": {},
     "outputs": [],
     "prompt_number": 27
    },
    {
     "cell_type": "code",
     "collapsed": false,
     "input": [
      "pop = \"aet\"\n",
      "tot_sweed_df = pd.read_csv(eu(\"~/vervet_project/analyses/20140611_163_subpop_sweepfinder/\"\n",
      "                                 \"_data/tot_sweed_result_{}.tsv\".format(pop)),sep=\"\\t\",index_col=[0,1])"
     ],
     "language": "python",
     "metadata": {},
     "outputs": [],
     "prompt_number": 714
    },
    {
     "cell_type": "code",
     "collapsed": false,
     "input": [
      "rod_s = tot_sweed_df['Likelihood'].copy()"
     ],
     "language": "python",
     "metadata": {},
     "outputs": [],
     "prompt_number": 759
    },
    {
     "cell_type": "code",
     "collapsed": false,
     "input": [],
     "language": "python",
     "metadata": {},
     "outputs": [],
     "prompt_number": 773
    },
    {
     "cell_type": "code",
     "collapsed": false,
     "input": [
      "def data_per_feature(rod,feature_df):\n",
      "    \"\"\"\n",
      "    Get the entires in rod which lie within a feature\n",
      "    (e.g. gene) in feature_df.\n",
      "    Input:\n",
      "    rod (reference ordered data)... pandas series or data frame with multiindex (chrom, pos)\n",
      "                                    such as SNP genotypes\n",
      "    feature_df (gene annotation data frame)... index must be (chrom,feature_name), must have columns 'start', 'end'\n",
      "    \"\"\"\n",
      "    rod = pd.DataFrame(rod)\n",
      "    #feature_hit_df = pd.DataFrame()\n",
      "    chrom_features = []\n",
      "    chrpos_names = rod.index.names\n",
      "    feature_name = feature_df.index.names[1]\n",
      "    for chrom in rod.index.droplevel(1).unique():\n",
      "        #print feature_df.ix[chrom].index.is_monotonic\n",
      "        #print rod.ix[chrom].index.is_monotonic\n",
      "        #pos_rel_to_start = feature_df.ix[chrom]['start'].searchsorted(rod.ix[chrom].index)\n",
      "        pos_rel_to_end = np.searchsorted(feature_df.ix[chrom][\"end\"].values,rod.ix[chrom].index.values)\n",
      "        #in_feature = (pos_rel_to_start - pos_rel_to_end) == 1\n",
      "        #feature_id = feature_df.ix[chrom].iloc[pos_rel_to_end[in_feature]].index\n",
      "        #snp_df = rod.ix[chrom][in_feature]\n",
      "        #snp_df[chrpos_names[0]] = chrom\n",
      "        #snp_df[feature_name] = feature_id\n",
      "        #snp_df.set_index([chrpos_names[0],feature_name],append=True,inplace=True)\n",
      "        #snp_df = snp_df.reorder_levels([chrpos_names[0], feature_name,chrpos_names[1]])\n",
      "        #chrom_features.append(snp_df)\n",
      "    #return pd.concat(chrom_features)"
     ],
     "language": "python",
     "metadata": {},
     "outputs": [],
     "prompt_number": 481
    },
    {
     "cell_type": "code",
     "collapsed": false,
     "input": [
      "def data_per_feature3(rod,feature_df):\n",
      "    \"\"\"\n",
      "    Get the entires in rod which lie within a feature\n",
      "    (e.g. gene) in feature_df.\n",
      "    Input:\n",
      "    rod (reference ordered data)... pandas series or data frame with multiindex (chrom, pos)\n",
      "                                    such as SNP genotypes\n",
      "    feature_df (gene annotation data frame)... index must be (chrom,feature_name), must have columns 'start', 'end'\n",
      "    \"\"\"\n",
      "    rod = pd.DataFrame(rod).copy()\n",
      "    feature_df = feature_df.copy()\n",
      "    index = rod.reset_index().groupby('chrom').apply(lambda df: df['pos']+chrom_len_df.loc[:df['chrom'].iloc[0]].iloc[:-1].sum()).values\n",
      "    rod['index'] = index\n",
      "    rod.set_index(index,inplace=True)\n",
      "\n",
      "    feature_df.loc[:,['start','end']] = feature_df.reset_index().groupby('chrom').apply(lambda df: df[['start','end']]+chrom_len_df.loc[:df['chrom'].iloc[0]].iloc[:-1].sum()).values\n",
      "    feature_df.index = feature_df.index.droplevel(0)\n",
      "    feature_name = feature_df.index.name\n",
      "\n",
      "    pos_rel_to_start = feature_df['start'].searchsorted(rod.index)\n",
      "    pos_rel_to_end = np.searchsorted(feature_df[\"end\"].values,rod.index.values)\n",
      "    in_feature = (pos_rel_to_start - pos_rel_to_end) == 1\n",
      "    feature_id = feature_df.iloc[pos_rel_to_end[in_feature]].index\n",
      "    snp_df = rod[in_feature]\n",
      "    snp_df[feature_name] = feature_id\n",
      "    snp_df.set_index(['gene_id','index'],inplace=True)\n",
      "    return snp_df"
     ],
     "language": "python",
     "metadata": {},
     "outputs": [],
     "prompt_number": 448
    },
    {
     "cell_type": "code",
     "collapsed": false,
     "input": [
      "def rod_chrompos_to_1d_index(rod, chrom_len_s, drop=True):\n",
      "    \"\"\"\n",
      "    Converts a (chrom, pos) multiindex to a \n",
      "    single numeric index that runs through chromosomes.\n",
      "    Note that the (chrom, pos) index is sorted lexographically,\n",
      "    i.e., if chrom names are strings, the entries in the resulting\n",
      "    index are Chr1, Chr10, Chr2, Chr20, Chr3,...,ChrX, ChrY.\n",
      "    Example: (Chr2, 1) is converted to len(Chr1)+len(Chr10)+1.\n",
      "    The inverse operation is given by rod_1d_to_chrompos_index.\n",
      "    \n",
      "    Input:\n",
      "    rod ... series or dataframe with reference ordered data \n",
      "            with multiindex (chrom,pos)\n",
      "    chrom_len_s ... series with chromosome names as keys\n",
      "                    and chromosome length as values\n",
      "    drop ... If False, keep chrom, pos as columns\n",
      "    \"\"\"\n",
      "    if not chrom_len_s.index.is_monotonic:\n",
      "        chrom_len_s = chrom_len_s.sort_index()\n",
      "    rod = pd.DataFrame(rod).copy()\n",
      "    if not rod.index.is_monotonic:\n",
      "        rod.sortlevel(inplace=True)\n",
      "    columns = columns = [c for c in rod.columns if c not in ['index']]\n",
      "    rod.reset_index(inplace=True)\n",
      "    #return rod.groupby('chrom')\n",
      "    index = rod.groupby('chrom').apply(lambda df: df['pos']+chrom_len_s.loc[:df['chrom'].iloc[0]].iloc[:-1].sum()).values\n",
      "    rod['index'] = index\n",
      "    rod.set_index('index', inplace=True, drop=True)\n",
      "    if not drop:\n",
      "        columns = list(columns) + ['chrom', 'pos']\n",
      "    rod.index = rod.index.values.astype(int)\n",
      "    if not r.index.is_monotonic:\n",
      "        rod.sort_index(inplace=True)\n",
      "    \n",
      "    return rod[columns] if len(columns)>1 else rod[columns[0]]"
     ],
     "language": "python",
     "metadata": {},
     "outputs": [],
     "prompt_number": 1083
    },
    {
     "cell_type": "code",
     "collapsed": false,
     "input": [
      "def rod_1d_to_chrompos_index(rod_1d, chrom_len_s, drop=True):\n",
      "    \"\"\"\n",
      "    Reverts the action of rod_chrompos_to_1d_index.\n",
      "    Converts a single numeric index that runs through \n",
      "    chromosomes to a (chrom, pos) multiindex.\n",
      "    Note that the single index is expected to correspond\n",
      "    to (chrom, pos) sorted lexographically,\n",
      "    i.e., if chrom names are strings, the entries should be in\n",
      "    in the order Chr1, Chr10, Chr2, Chr20, Chr3,...,ChrX, ChrY.\n",
      "    Example:  len(Chr1)+len(Chr10)+1 is converted to (Chr2, 1).\n",
      "    \n",
      "    Input:\n",
      "    rod ... series or dataframe of reference ordered data \n",
      "            with single index running through chromosomes.\n",
      "            (as produced by rod_chrompos_to_1d_index())\n",
      "    chrom_len_s ... series with chromosome names as keys\n",
      "                    and chromosome length as values\n",
      "    drop ... If False, keep numeric index as columns\n",
      "    \"\"\"\n",
      "    if not chrom_len_s.index.is_monotonic:\n",
      "        chrom_len_s = chrom_len_s.sort_index()\n",
      "    rod = pd.DataFrame(rod_1d).copy()\n",
      "    if not rod.index.is_monotonic:\n",
      "        rod.sort_index(inplace=True)\n",
      "    columns = [c for c in rod.columns if c not in ['chrom','pos']]\n",
      "    cs = chrom_len_s.cumsum()\n",
      "    rod['chrom'] = np.nan\n",
      "    for chrom, (start, end) in  zip(cs.index,zip([0] + list(cs.values[:-1]),cs.values)):\n",
      "        end = min(end,rod.index[-1])\n",
      "        rod.loc[slice(start,end),'chrom'] = chrom\n",
      "        #print chrom, rod.loc[slice(start,end),columns[0]]\n",
      "        rod.loc[slice(start,end),'pos'] = rod.ix[slice(start,end)].index - start\n",
      "    if drop:\n",
      "        rod.set_index(['chrom','pos'], inplace=True, drop=True)\n",
      "    else:\n",
      "        rod = rod.reset_index().set_index(['chrom','pos'], drop=True)\n",
      "        columns = list(columns) + ['index']\n",
      "    #if not rod.index.is_monotonic:\n",
      "    #    rod.sort_index(inplace=True)\n",
      "\n",
      "    return rod[columns] if len(columns)>1 else rod[columns[0]]"
     ],
     "language": "python",
     "metadata": {},
     "outputs": [],
     "prompt_number": 1115
    },
    {
     "cell_type": "code",
     "collapsed": false,
     "input": [
      "\"\"\"\n",
      "Specifications:\n",
      "\n",
      "rod ... Series or DataFrame of reference ordered data\n",
      "index: \n",
      "    (chrom, pos)\n",
      "\n",
      "feature_df ... Data Frame of intervals along the genome,\n",
      "                equivalent of a bed file, but 1-indexed\n",
      "index: \n",
      "    (chrom, start)\n",
      "columns \n",
      "    required: \n",
      "        end\n",
      "    optional:\n",
      "        name\n",
      "        ...\n",
      "\"\"\""
     ],
     "language": "python",
     "metadata": {},
     "outputs": [
      {
       "metadata": {},
       "output_type": "pyout",
       "prompt_number": 764,
       "text": [
        "'\\nSpecifications:\\n\\nrod ... Series or DataFrame of reference ordered data\\nindex: \\n    (chrom, pos)\\n\\nfeature_df ... Data Frame of intervals along the genome,\\n                equivalent of a bed file, but 1-indexed\\nindex: \\n    (chrom, start)\\ncolumns \\n    required: \\n        end\\n    optional:\\n        name\\n        ...\\n'"
       ]
      }
     ],
     "prompt_number": 764
    },
    {
     "cell_type": "code",
     "collapsed": false,
     "input": [
      "def feature_df_to_1d_index(feature_df, chrom_len_s):\n",
      "    \"\"\"\n",
      "    Converts mulitindex feature_df (chrom,start)\n",
      "    to single numeric index running trough all \n",
      "    chromosomes. The column 'end' is also converted.\n",
      "    See rod_chrompos_to_1d_index for details.\n",
      "    \"\"\"\n",
      "    feature_df = feature_df.copy()\n",
      "    feature_df.index.names = (feature_df.index.names[0], 'pos') \n",
      "    feature_df_1d = rod_chrompos_to_1d_index(feature_df, chrom_len_s,drop=False)\n",
      "    #print feature_df_1d\n",
      "    end_df = feature_df_1d.reset_index().set_index(['chrom','end'])\n",
      "    #print end_df\n",
      "    end_df.drop('pos',axis=1,inplace=True)\n",
      "    end_df.rename(columns={'index':'start'}, inplace=True)\n",
      "    end_df.index.names = (end_df.index.names[0], 'pos') \n",
      "    end_1d = rod_chrompos_to_1d_index(end_df, chrom_len_s)\n",
      "    end_1d.index.name = 'end'\n",
      "    end_1d = end_1d.reset_index().set_index('start')\n",
      "    end_1d.index.name = 'index'\n",
      "    if not end_1d.index.is_monotonic:\n",
      "        end_1d.sort_index(inplace=True)\n",
      "    return end_1d"
     ],
     "language": "python",
     "metadata": {},
     "outputs": [],
     "prompt_number": 1125
    },
    {
     "cell_type": "code",
     "collapsed": false,
     "input": [
      "feature_df_1d = feature_df_to_1d_index(gene_df, chrom_len_df)"
     ],
     "language": "python",
     "metadata": {},
     "outputs": [],
     "prompt_number": 1133
    },
    {
     "cell_type": "code",
     "collapsed": false,
     "input": [
      "def feature_df_to_chrompos_index(feature_df_1d, chrom_len_s):\n",
      "    \"\"\"\n",
      "    Converts feature_df with single numberic index\n",
      "    running through all chromosomes to \n",
      "    multiindex (chrom,start)\n",
      "    The column 'end' is also converted.\n",
      "    This is the inverse function of \n",
      "    feature_df_to_1d_index().\n",
      "    See rod_1d_to_chrompos_index for details.\n",
      "    \"\"\"\n",
      "    feature_df_1d = feature_df_1d.copy()\n",
      "    feature_df = rod_1d_to_chrompos_index(feature_df_1d, chrom_len_s)\n",
      "    feature_df.index.names = (feature_df.index.names[0], 'start') \n",
      "    end_df_1d = feature_df.reset_index().set_index('end')\n",
      "    \n",
      "    end_df = rod_1d_to_chrompos_index(end_df_1d, chrom_len_s)\n",
      "    end_df.index.names = (end_df.index.names[0],'end')\n",
      "    end_df = end_df.reset_index().set_index(['chrom','start'])\n",
      "    if not end_df.index.is_monotonic:\n",
      "        end_df.sortlevel(inplace=True)\n",
      "    return end_df"
     ],
     "language": "python",
     "metadata": {},
     "outputs": [],
     "prompt_number": 1131
    },
    {
     "cell_type": "code",
     "collapsed": false,
     "input": [
      "(gene_df.index.values == feature.index.values).a"
     ],
     "language": "python",
     "metadata": {},
     "outputs": [
      {
       "metadata": {},
       "output_type": "pyout",
       "prompt_number": 905,
       "text": [
        "array([ True,  True,  True, ...,  True,  True,  True], dtype=bool)"
       ]
      }
     ],
     "prompt_number": 905
    },
    {
     "cell_type": "code",
     "collapsed": false,
     "input": [
      "feature = feature_df_to_chrompos_index(feature_df_1d, chrom_len_df)"
     ],
     "language": "python",
     "metadata": {},
     "outputs": [],
     "prompt_number": 1134
    },
    {
     "cell_type": "code",
     "collapsed": false,
     "input": [
      "feature.index.is_monotonic"
     ],
     "language": "python",
     "metadata": {},
     "outputs": [
      {
       "metadata": {},
       "output_type": "pyout",
       "prompt_number": 1122,
       "text": [
        "False"
       ]
      }
     ],
     "prompt_number": 1122
    },
    {
     "cell_type": "code",
     "collapsed": false,
     "input": [
      "feature.index.values == gene_df.index.values"
     ],
     "language": "python",
     "metadata": {},
     "outputs": [
      {
       "metadata": {},
       "output_type": "pyout",
       "prompt_number": 1120,
       "text": [
        "array([ True,  True,  True, ...,  True,  True,  True], dtype=bool)"
       ]
      }
     ],
     "prompt_number": 1120
    },
    {
     "cell_type": "code",
     "collapsed": false,
     "input": [
      "print (feature == gene_df).all()\n",
      "print feature.index.equals(gene_df.index)"
     ],
     "language": "python",
     "metadata": {},
     "outputs": [
      {
       "output_type": "stream",
       "stream": "stdout",
       "text": [
        "end        True\n",
        "gene_id    True\n",
        "dtype: bool\n",
        "True\n"
       ]
      }
     ],
     "prompt_number": 1135
    },
    {
     "cell_type": "code",
     "collapsed": false,
     "input": [
      "rod_1d = rod_chrompos_to_1d_index(rod_s,chrom_len_df)"
     ],
     "language": "python",
     "metadata": {},
     "outputs": [],
     "prompt_number": 1136
    },
    {
     "cell_type": "code",
     "collapsed": false,
     "input": [
      "rod = rod_1d_to_chrompos_index(rod_1d,chrom_len_df)"
     ],
     "language": "python",
     "metadata": {},
     "outputs": [],
     "prompt_number": 1137
    },
    {
     "cell_type": "code",
     "collapsed": false,
     "input": [
      "rod_orig = rod_s\n",
      "rod_orig = rod_orig.sortlevel()\n",
      "rod_orig.index = pd.MultiIndex.from_arrays([rod_orig.index.droplevel(1),rod_orig.index.droplevel(0).values.astype(int)])"
     ],
     "language": "python",
     "metadata": {},
     "outputs": [],
     "prompt_number": 949
    },
    {
     "cell_type": "code",
     "collapsed": false,
     "input": [
      "print (rod == rod_orig).all()\n",
      "print rod.index.equals(rod_orig.index)"
     ],
     "language": "python",
     "metadata": {},
     "outputs": [
      {
       "output_type": "stream",
       "stream": "stdout",
       "text": [
        "True\n",
        "True"
       ]
      },
      {
       "output_type": "stream",
       "stream": "stdout",
       "text": [
        "\n"
       ]
      }
     ],
     "prompt_number": 1138
    },
    {
     "cell_type": "code",
     "collapsed": false,
     "input": [
      "def data_per_feature_SI(rod,feature_df):\n",
      "    \"\"\"\n",
      "    Get the entires in rod which lie within a feature\n",
      "    (e.g. gene) in feature_df.\n",
      "    Input:\n",
      "    rod (reference ordered data)... pandas series or data frame with multiindex (chrom, pos)\n",
      "                                    such as SNP genotypes\n",
      "    feature_df (gene annotation data frame)... index must be (chrom,feature_name), must have columns 'start', 'end'\n",
      "    \"\"\"\n",
      "    feature_name = feature_df.index.name\n",
      "\n",
      "    pos_rel_to_start = feature_df['start'].searchsorted(rod.index)\n",
      "    pos_rel_to_end = np.searchsorted(feature_df[\"end\"].values,rod.index.values)\n",
      "    in_feature = (pos_rel_to_start - pos_rel_to_end) == 1\n",
      "    feature_id = feature_df.iloc[pos_rel_to_end[in_feature]].index\n",
      "    snp_df = rod[in_feature]\n",
      "    snp_df[feature_name] = feature_id\n",
      "    snp_df.set_index(['gene_id','index'],inplace=True)\n",
      "    return snp_df"
     ],
     "language": "python",
     "metadata": {},
     "outputs": [],
     "prompt_number": 456
    },
    {
     "cell_type": "code",
     "collapsed": false,
     "input": [
      "def data_per_feature2(rod,feature_df):\n",
      "    rod = pd.DataFrame(rod)\n",
      "    def gene_values(gene,sl):\n",
      "        gene_values = data.ix[sl].copy()\n",
      "        gene_values['gene'] = gene\n",
      "        gene_values.set_index('gene',append=True,inplace=True)\n",
      "        return gene_values\n",
      "    for chrom in rod.index.droplevel(1).unique():\n",
      "        data = rod.ix[chrom]\n",
      "        genes = feature_df.ix[chrom]\n",
      "        slices = genes.apply(lambda gene: slice(gene['start'], gene['end']), axis=1)\n",
      "        return pd.concat([gene_values(*sl) for sl in slices.iteritems()])"
     ],
     "language": "python",
     "metadata": {},
     "outputs": [],
     "prompt_number": 345
    },
    {
     "cell_type": "code",
     "collapsed": false,
     "input": [
      "def get_features(peak_s, feature_df, max_dist):\n",
      "    \"\"\"\n",
      "    take the input series and gets.\n",
      "    names of features nearby\n",
      "\n",
      "    Input:\n",
      "    peak_s ... pandas series with (chrom, pos) index and value of\n",
      "                the statistic ('peak height'). Series should be named.\n",
      "    feature_df ... data frame with feature info.\n",
      "    \"\"\"\n",
      "    all_features = []\n",
      "    if not feature_df.index.is_monotonic:\n",
      "        feature_df = feature_df.sort_index()\n",
      "    tot_hit_df = pd.DataFrame()\n",
      "    for chrom in peak_s.index.droplevel(1).unique():\n",
      "        loc_feature_df = feature_df.ix[chrom]\n",
      "        #loc_feature_df = loc_feature_df.append(pd.DataFrame(np.nan,index=[np.inf],columns=loc_feature_df.columns))\n",
      "        pos_rel_to_start = np.searchsorted(loc_feature_df['start']-max_dist,peak_s.ix[chrom].index.values)\n",
      "        pos_rel_to_end = np.searchsorted(loc_feature_df[\"end\"].values+max_dist,peak_s.ix[chrom].index.values)\n",
      "        #return pos_rel_to_start,pos_rel_to_end\n",
      "        #print [range(a,b) for a,b in zip(pos_rel_to_end,pos_rel_to_start)]\n",
      "        #features = list(set(loc_feature_df.index[np.hstack([range(a,b) for a,b in zip(pos_rel_to_end,pos_rel_to_start)])]))\n",
      "        features = list(set(loc_feature_df[\"feature_id\"].iloc[np.hstack([range(a,b) for a,b in zip(pos_rel_to_end,pos_rel_to_start)])]))\n",
      "        #features = list(set(loc_feature_df[\"feature_id\"].iloc[np.vstack([pos_rel_to_start,pos_rel_to_end]).T]))\n",
      "        all_features += features\n",
      "    return all_features"
     ],
     "language": "python",
     "metadata": {},
     "outputs": [],
     "prompt_number": 253
    },
    {
     "cell_type": "code",
     "collapsed": false,
     "input": [
      "gene_df['feature_id'] = gene_df.index.droplevel(0)"
     ],
     "language": "python",
     "metadata": {},
     "outputs": [],
     "prompt_number": 235
    },
    {
     "cell_type": "code",
     "collapsed": false,
     "input": [
      "rod_s = tot_sweed_df['Likelihood'].copy().loc[['CAE1','CAE2']]"
     ],
     "language": "python",
     "metadata": {},
     "outputs": [],
     "prompt_number": 529
    },
    {
     "cell_type": "code",
     "collapsed": false,
     "input": [
      "tot_sweed_df['Likelihood'].ix['index']"
     ],
     "language": "python",
     "metadata": {},
     "outputs": [
      {
       "metadata": {},
       "output_type": "pyout",
       "prompt_number": 526,
       "text": [
        "pos\n",
        "       [1161.0, 1660.9969, 2160.9937, 2660.9906, 3160...\n",
        "Name: Likelihood, dtype: object"
       ]
      }
     ],
     "prompt_number": 526
    },
    {
     "cell_type": "code",
     "collapsed": false,
     "input": [
      "gene_df_single_index.sort_index(inplace=True)"
     ],
     "language": "python",
     "metadata": {},
     "outputs": [],
     "prompt_number": 463
    },
    {
     "cell_type": "code",
     "collapsed": false,
     "input": [
      "rod_single_index.sort_index(inplace=True)"
     ],
     "language": "python",
     "metadata": {},
     "outputs": [],
     "prompt_number": 465
    },
    {
     "cell_type": "code",
     "collapsed": false,
     "input": [
      "gene_df_single_index.index.is_monotonic"
     ],
     "language": "python",
     "metadata": {},
     "outputs": [
      {
       "metadata": {},
       "output_type": "pyout",
       "prompt_number": 464,
       "text": [
        "True"
       ]
      }
     ],
     "prompt_number": 464
    },
    {
     "cell_type": "code",
     "collapsed": false,
     "input": [
      "gene_df.sort('start',inplace=True)"
     ],
     "language": "python",
     "metadata": {},
     "outputs": [],
     "prompt_number": 477
    },
    {
     "cell_type": "code",
     "collapsed": false,
     "input": [
      "%timeit data_per_feature(rod_s,gene_df)"
     ],
     "language": "python",
     "metadata": {},
     "outputs": [
      {
       "output_type": "stream",
       "stream": "stdout",
       "text": [
        "1 loops, best of 3: 3.65 s per loop\n"
       ]
      }
     ],
     "prompt_number": 482
    },
    {
     "cell_type": "code",
     "collapsed": false,
     "input": [
      "%timeit data_per_feature(rod_s,gene_df)"
     ],
     "language": "python",
     "metadata": {},
     "outputs": [
      {
       "output_type": "stream",
       "stream": "stdout",
       "text": [
        "1 loops, best of 3: 3.7 s per loop\n"
       ]
      }
     ],
     "prompt_number": 480
    },
    {
     "cell_type": "code",
     "collapsed": false,
     "input": [
      "%timeit data_per_feature_SI(rod_single_index,gene_df_single_index)"
     ],
     "language": "python",
     "metadata": {},
     "outputs": [
      {
       "output_type": "stream",
       "stream": "stdout",
       "text": [
        "1 loops, best of 3: 256 ms per loop\n"
       ]
      }
     ],
     "prompt_number": 466
    },
    {
     "cell_type": "code",
     "collapsed": false,
     "input": [
      "%timeit data_per_feature_SI(rod_single_index,gene_df_single_index)"
     ],
     "language": "python",
     "metadata": {},
     "outputs": [
      {
       "output_type": "stream",
       "stream": "stdout",
       "text": [
        "1 loops, best of 3: 2.61 s per loop\n"
       ]
      }
     ],
     "prompt_number": 460
    },
    {
     "cell_type": "code",
     "collapsed": false,
     "input": [
      "%timeit data_per_feature3(tot_sweed_df['Likelihood'],gene_df)"
     ],
     "language": "python",
     "metadata": {},
     "outputs": [
      {
       "output_type": "stream",
       "stream": "stdout",
       "text": [
        "1 loops, best of 3: 5.69 s per loop\n"
       ]
      }
     ],
     "prompt_number": 451
    },
    {
     "cell_type": "code",
     "collapsed": false,
     "input": [
      "np.hstack([range(a,b) for a,b in zip(pos_rel_to_end,pos_rel_to_start)])"
     ],
     "language": "python",
     "metadata": {},
     "outputs": [
      {
       "metadata": {},
       "output_type": "pyout",
       "prompt_number": 255,
       "text": [
        "array([  0.00000000e+00,   1.00000000e+00,   2.00000000e+00, ...,\n",
        "         1.33300000e+03,   1.33200000e+03,   1.33300000e+03])"
       ]
      }
     ],
     "prompt_number": 255
    },
    {
     "cell_type": "code",
     "collapsed": false,
     "input": [
      "%timeit data_per_feature( tot_sweed_df['Likelihood'].loc[['CAE28']],gene_df)"
     ],
     "language": "python",
     "metadata": {},
     "outputs": [
      {
       "output_type": "stream",
       "stream": "stdout",
       "text": [
        "1 loops, best of 3: 437 ms per loop\n"
       ]
      }
     ],
     "prompt_number": 370
    },
    {
     "cell_type": "code",
     "collapsed": false,
     "input": [
      "gene_df.index.searchsorted?"
     ],
     "language": "python",
     "metadata": {},
     "outputs": [],
     "prompt_number": 378
    },
    {
     "cell_type": "code",
     "collapsed": false,
     "input": [
      "chrom_len_df = pd.read_csv(eu('~/vervet_project/metadata/ref3500.tsv'),sep='\\t',index_col=0,squeeze=True)"
     ],
     "language": "python",
     "metadata": {},
     "outputs": [],
     "prompt_number": 396
    },
    {
     "cell_type": "code",
     "collapsed": false,
     "input": [
      "rod_s.reset_index().groupby('chrom').apply(lambda df: df['pos']+chrom_len_df.loc[:df['chrom'].iloc[0]].iloc[:-1].sum()).values"
     ],
     "language": "python",
     "metadata": {},
     "outputs": [
      {
       "metadata": {},
       "output_type": "pyout",
       "prompt_number": 419,
       "text": [
        "array([  1.81064081e+09,   1.81064129e+09,   1.81064178e+09, ...,\n",
        "         2.60789421e+09,   2.60789469e+09,   2.60789517e+09])"
       ]
      }
     ],
     "prompt_number": 419
    },
    {
     "cell_type": "code",
     "collapsed": false,
     "input": [
      "%timeit r.apply(lambda row:row['pos']+chrom_len_df.loc[:row['chrom']].iloc[:-1].sum(),axis=1)"
     ],
     "language": "python",
     "metadata": {},
     "outputs": [
      {
       "output_type": "stream",
       "stream": "stdout",
       "text": [
        "1 loops, best of 3: 38.8 s per loop\n"
       ]
      }
     ],
     "prompt_number": 409
    },
    {
     "cell_type": "code",
     "collapsed": false,
     "input": [
      "r.groupby('chrom').apply(lambda df: df['pos']+chrom_len_df.loc[:df['chrom'].iloc[0]].iloc[:-1].sum())"
     ],
     "language": "python",
     "metadata": {},
     "outputs": [
      {
       "metadata": {},
       "output_type": "pyout",
       "prompt_number": 417,
       "text": [
        "chrom    \n",
        "CAE19  0     1.810641e+09\n",
        "       1     1.810641e+09\n",
        "       2     1.810642e+09\n",
        "       3     1.810642e+09\n",
        "       4     1.810643e+09\n",
        "       5     1.810643e+09\n",
        "       6     1.810644e+09\n",
        "       7     1.810644e+09\n",
        "       8     1.810645e+09\n",
        "       9     1.810645e+09\n",
        "       10    1.810646e+09\n",
        "       11    1.810646e+09\n",
        "       12    1.810647e+09\n",
        "       13    1.810647e+09\n",
        "       14    1.810648e+09\n",
        "...\n",
        "CAE29  157988    2.607888e+09\n",
        "       157989    2.607889e+09\n",
        "       157990    2.607889e+09\n",
        "       157991    2.607890e+09\n",
        "       157992    2.607890e+09\n",
        "       157993    2.607891e+09\n",
        "       157994    2.607891e+09\n",
        "       157995    2.607892e+09\n",
        "       157996    2.607892e+09\n",
        "       157997    2.607893e+09\n",
        "       157998    2.607893e+09\n",
        "       157999    2.607894e+09\n",
        "       158000    2.607894e+09\n",
        "       158001    2.607895e+09\n",
        "       158002    2.607895e+09\n",
        "Name: pos, Length: 158003, dtype: float64"
       ]
      }
     ],
     "prompt_number": 417
    },
    {
     "cell_type": "code",
     "collapsed": false,
     "input": [
      "def get_summary_per_feature(value_s,feature_df,summary=None,summary_func=None):\n",
      "    \"\"\"\n",
      "    summary ... name of the function of the groupby object\n",
      "                         to apply to the data (e.g. 'mean', 'max',...)\n",
      "    summary_func ... alternatively: function object to apply\n",
      "                     This is slower than applying a summary string.\n",
      "    \"\"\"\n",
      "    values_per_feature = hp.data_per_feature(value_s,feature_df)\n",
      "    if summary is not None:\n",
      "        summary_per_feature = hp.apply_to_feature(values_per_feature,summary)\n",
      "    elif summary_func is not None:\n",
      "        summary_per_feature = hp.apply_to_feature(values_per_feature,function=summary_func)\n",
      "    else:\n",
      "        raise ValueError(\"Either summary or summary_func must be given.\")\n",
      "    return summary_per_feature"
     ],
     "language": "python",
     "metadata": {},
     "outputs": [],
     "prompt_number": 57
    },
    {
     "cell_type": "code",
     "collapsed": false,
     "input": [
      "def get_summary_per_category(value_per_feature,feature_to_category,summary='mean'):\n",
      "    \"\"\"\n",
      "    Calculates summary (e.g. mean) of values for the features\n",
      "    in each of the given categories.\n",
      "    \"\"\"\n",
      "    feature_to_category = feature_to_category.copy()\n",
      "    values_per_feature_to_category = value_per_feature.ix[feature_to_category['gene_symbol'].values].values\n",
      "    feature_to_category['value'] =  values_per_feature_to_category\n",
      "    del feature_to_category['gene_symbol']\n",
      "    summary_per_category = getattr(feature_to_category.groupby('go_identifier'),summary)()\n",
      "    return summary_per_category['value'] "
     ],
     "language": "python",
     "metadata": {},
     "outputs": [],
     "prompt_number": 118
    },
    {
     "cell_type": "code",
     "collapsed": false,
     "input": [
      "tot_sweed_df.sortlevel([0,1],inplace=True)"
     ],
     "language": "python",
     "metadata": {},
     "outputs": [],
     "prompt_number": 181
    },
    {
     "cell_type": "code",
     "collapsed": false,
     "input": [
      "rod_s = tot_sweed_df['Likelihood'].loc[['CAE1']]"
     ],
     "language": "python",
     "metadata": {},
     "outputs": [],
     "prompt_number": 347
    },
    {
     "cell_type": "code",
     "collapsed": false,
     "input": [
      "rod_s"
     ],
     "language": "python",
     "metadata": {},
     "outputs": [
      {
       "metadata": {},
       "output_type": "pyout",
       "prompt_number": 348,
       "text": [
        "chrom  pos      \n",
        "CAE1   1161.0000    0.000000\n",
        "       1660.9969    0.000000\n",
        "       2160.9937    0.000000\n",
        "       2660.9906    0.000000\n",
        "       3160.9874    0.000000\n",
        "       3660.9843    0.038328\n",
        "       4160.9811    0.854687\n",
        "       4660.9780    0.109911\n",
        "       5160.9749    0.047829\n",
        "       5660.9717    0.000000\n",
        "       6160.9686    0.077996\n",
        "       6660.9654    0.000003\n",
        "       7160.9623    0.000000\n",
        "       7660.9591    0.004962\n",
        "       8160.9560    0.000000\n",
        "...\n",
        "CAE1   1.260289e+08    0.001553\n",
        "       1.260294e+08    0.101537\n",
        "       1.260299e+08    0.000000\n",
        "       1.260304e+08    0.008782\n",
        "       1.260309e+08    0.887271\n",
        "       1.260314e+08    0.375868\n",
        "       1.260319e+08    0.128634\n",
        "       1.260324e+08    0.016026\n",
        "       1.260329e+08    0.012902\n",
        "       1.260334e+08    0.147273\n",
        "       1.260339e+08    0.158781\n",
        "       1.260344e+08    0.000000\n",
        "       1.260349e+08    0.000000\n",
        "       1.260354e+08    0.000000\n",
        "       1.260359e+08    0.000000\n",
        "Name: Likelihood, Length: 252072, dtype: float64"
       ]
      }
     ],
     "prompt_number": 348
    },
    {
     "cell_type": "code",
     "collapsed": false,
     "input": [
      "spf = get_summary_per_feature(rod_s,gene_df,summary='max')"
     ],
     "language": "python",
     "metadata": {},
     "outputs": [],
     "prompt_number": 59
    },
    {
     "cell_type": "code",
     "collapsed": false,
     "input": [
      "spc = get_summary_per_category(spf,gene_to_cat,'mean')"
     ],
     "language": "python",
     "metadata": {},
     "outputs": [],
     "prompt_number": 122
    },
    {
     "cell_type": "code",
     "collapsed": false,
     "input": [
      "init_rank_table = te.get_initial_rank_table(spc)"
     ],
     "language": "python",
     "metadata": {},
     "outputs": [],
     "prompt_number": 123
    },
    {
     "cell_type": "code",
     "collapsed": false,
     "input": [
      "from hs_vervet.scripts import test_enrichment_dev as ted"
     ],
     "language": "python",
     "metadata": {},
     "outputs": [],
     "prompt_number": 153
    },
    {
     "cell_type": "code",
     "collapsed": false,
     "input": [
      "del gene_to_cat['value']"
     ],
     "language": "python",
     "metadata": {},
     "outputs": [],
     "prompt_number": 158
    },
    {
     "cell_type": "code",
     "collapsed": false,
     "input": [
      "se = ted.SummaryEnrichment(rod_s, gene_df, gene_to_cat, feature_summary='max',category_summary='sum')"
     ],
     "language": "python",
     "metadata": {},
     "outputs": [],
     "prompt_number": 203
    },
    {
     "cell_type": "code",
     "collapsed": false,
     "input": [
      "def top_genes(values_per_feature,quantile=None,top_n=None):\n",
      "    \"\"\"\n",
      "    reverse engineering to get top quantile per gene category\n",
      "    \"\"\"\n",
      "    if top_n is None:\n",
      "        top_n = int(len(means)*quantile)\n",
      "    groups = values_per_feature.groupby(lambda idx: idx[1])\n",
      "    means = groups.mean()\n",
      "    sort_means = means.sort(ascending=False)\n",
      "    sort_means.iloc[:top_n] = 1\n",
      "    sort_means.iloc[top_n:] = 0\n",
      "    return sort_means"
     ],
     "language": "python",
     "metadata": {},
     "outputs": [],
     "prompt_number": 218
    },
    {
     "cell_type": "code",
     "collapsed": false,
     "input": [
      "se1 = ted.SummaryEnrichment(rod_s, gene_df, gene_to_cat, feature_summary_fun=feature_summary_fun,category_summary='sum')"
     ],
     "language": "python",
     "metadata": {},
     "outputs": [],
     "prompt_number": 219
    },
    {
     "cell_type": "code",
     "collapsed": false,
     "input": [
      "se1.real_assoc()"
     ],
     "language": "python",
     "metadata": {},
     "outputs": [
      {
       "output_type": "stream",
       "stream": "stdout",
       "text": [
        "                value\n",
        "category             \n",
        "GO:0000002          1\n",
        "GO:0000003        NaN\n",
        "GO:0000009          0\n",
        "GO:0000010        NaN\n",
        "GO:0000012        NaN\n",
        "GO:0000014        NaN\n",
        "GO:0000015        NaN\n",
        "GO:0000016        NaN\n",
        "GO:0000018        NaN\n",
        "GO:0000019        NaN\n",
        "GO:0000022        NaN\n",
        "GO:0000023        NaN\n",
        "GO:0000026        NaN\n",
        "GO:0000027        NaN\n",
        "GO:0000028        NaN\n",
        "GO:0000030        NaN\n",
        "GO:0000033        NaN\n",
        "GO:0000035        NaN\n",
        "GO:0000036        NaN\n",
        "GO:0000038        NaN\n",
        "GO:0000041        NaN\n",
        "GO:0000042        NaN\n",
        "GO:0000045          1\n",
        "GO:0000046          0\n",
        "GO:0000049          1\n",
        "GO:0000050          0\n",
        "GO:0000052        NaN\n",
        "GO:0000053        NaN\n",
        "GO:0000054        NaN\n",
        "GO:0000055        NaN\n",
        "...               ...\n",
        "GO:2001268        NaN\n",
        "GO:2001269        NaN\n",
        "GO:2001271        NaN\n",
        "GO:2001273        NaN\n",
        "GO:2001274        NaN\n",
        "GO:2001275        NaN\n",
        "GO:2001286        NaN\n",
        "GO:2001287        NaN\n",
        "GO:2001288        NaN\n",
        "GO:2001294        NaN\n",
        "GO:2001295        NaN\n",
        "GO:2001300        NaN\n",
        "GO:2001303        NaN\n",
        "GO:2001306        NaN\n",
        "GO:2001311        NaN\n",
        "hiv155030.0         0\n",
        "hiv155348.0         3\n",
        "hiv155459.0         2\n",
        "hiv155807.0         2\n",
        "hiv155871.0         8\n",
        "hiv155908.0         1\n",
        "hiv155945.0         2\n",
        "hiv155971.0         2\n",
        "hiv156110.0         1\n",
        "hiv19424028.0     NaN\n",
        "hs_hivprot         14\n",
        "hs_hivprotphen      7\n",
        "hs_hivrep           5\n",
        "hs_hivrepphen       2\n",
        "hs_sivcand          0\n",
        "\n",
        "[15013 rows x 1 columns]\n"
       ]
      }
     ],
     "prompt_number": 220
    },
    {
     "cell_type": "code",
     "collapsed": false,
     "input": [
      "se1.rank_table"
     ],
     "language": "python",
     "metadata": {},
     "outputs": [
      {
       "html": [
        "<div style=\"max-height:1000px;max-width:1500px;overflow:auto;\">\n",
        "<table border=\"1\" class=\"dataframe\">\n",
        "  <thead>\n",
        "    <tr style=\"text-align: right;\">\n",
        "      <th></th>\n",
        "      <th>out_of</th>\n",
        "      <th>rank</th>\n",
        "      <th>value</th>\n",
        "    </tr>\n",
        "    <tr>\n",
        "      <th>category</th>\n",
        "      <th></th>\n",
        "      <th></th>\n",
        "      <th></th>\n",
        "    </tr>\n",
        "  </thead>\n",
        "  <tbody>\n",
        "    <tr>\n",
        "      <th>GO:0000002</th>\n",
        "      <td> 0</td>\n",
        "      <td> 0</td>\n",
        "      <td> 0.007469</td>\n",
        "    </tr>\n",
        "    <tr>\n",
        "      <th>GO:0000003</th>\n",
        "      <td> 0</td>\n",
        "      <td> 0</td>\n",
        "      <td>      NaN</td>\n",
        "    </tr>\n",
        "    <tr>\n",
        "      <th>GO:0000009</th>\n",
        "      <td> 0</td>\n",
        "      <td> 0</td>\n",
        "      <td>      NaN</td>\n",
        "    </tr>\n",
        "    <tr>\n",
        "      <th>GO:0000010</th>\n",
        "      <td> 0</td>\n",
        "      <td> 0</td>\n",
        "      <td>      NaN</td>\n",
        "    </tr>\n",
        "    <tr>\n",
        "      <th>GO:0000012</th>\n",
        "      <td> 0</td>\n",
        "      <td> 0</td>\n",
        "      <td>      NaN</td>\n",
        "    </tr>\n",
        "    <tr>\n",
        "      <th>GO:0000014</th>\n",
        "      <td> 0</td>\n",
        "      <td> 0</td>\n",
        "      <td>      NaN</td>\n",
        "    </tr>\n",
        "    <tr>\n",
        "      <th>GO:0000015</th>\n",
        "      <td> 0</td>\n",
        "      <td> 0</td>\n",
        "      <td>      NaN</td>\n",
        "    </tr>\n",
        "    <tr>\n",
        "      <th>GO:0000016</th>\n",
        "      <td> 0</td>\n",
        "      <td> 0</td>\n",
        "      <td>      NaN</td>\n",
        "    </tr>\n",
        "    <tr>\n",
        "      <th>GO:0000018</th>\n",
        "      <td> 0</td>\n",
        "      <td> 0</td>\n",
        "      <td>      NaN</td>\n",
        "    </tr>\n",
        "    <tr>\n",
        "      <th>GO:0000019</th>\n",
        "      <td> 0</td>\n",
        "      <td> 0</td>\n",
        "      <td>      NaN</td>\n",
        "    </tr>\n",
        "    <tr>\n",
        "      <th>GO:0000022</th>\n",
        "      <td> 0</td>\n",
        "      <td> 0</td>\n",
        "      <td>      NaN</td>\n",
        "    </tr>\n",
        "    <tr>\n",
        "      <th>GO:0000023</th>\n",
        "      <td> 0</td>\n",
        "      <td> 0</td>\n",
        "      <td>      NaN</td>\n",
        "    </tr>\n",
        "    <tr>\n",
        "      <th>GO:0000026</th>\n",
        "      <td> 0</td>\n",
        "      <td> 0</td>\n",
        "      <td>      NaN</td>\n",
        "    </tr>\n",
        "    <tr>\n",
        "      <th>GO:0000027</th>\n",
        "      <td> 0</td>\n",
        "      <td> 0</td>\n",
        "      <td>      NaN</td>\n",
        "    </tr>\n",
        "    <tr>\n",
        "      <th>GO:0000028</th>\n",
        "      <td> 0</td>\n",
        "      <td> 0</td>\n",
        "      <td>      NaN</td>\n",
        "    </tr>\n",
        "    <tr>\n",
        "      <th>GO:0000030</th>\n",
        "      <td> 0</td>\n",
        "      <td> 0</td>\n",
        "      <td>      NaN</td>\n",
        "    </tr>\n",
        "    <tr>\n",
        "      <th>GO:0000033</th>\n",
        "      <td> 0</td>\n",
        "      <td> 0</td>\n",
        "      <td>      NaN</td>\n",
        "    </tr>\n",
        "    <tr>\n",
        "      <th>GO:0000035</th>\n",
        "      <td> 0</td>\n",
        "      <td> 0</td>\n",
        "      <td>      NaN</td>\n",
        "    </tr>\n",
        "    <tr>\n",
        "      <th>GO:0000036</th>\n",
        "      <td> 0</td>\n",
        "      <td> 0</td>\n",
        "      <td>      NaN</td>\n",
        "    </tr>\n",
        "    <tr>\n",
        "      <th>GO:0000038</th>\n",
        "      <td> 0</td>\n",
        "      <td> 0</td>\n",
        "      <td>      NaN</td>\n",
        "    </tr>\n",
        "    <tr>\n",
        "      <th>GO:0000041</th>\n",
        "      <td> 0</td>\n",
        "      <td> 0</td>\n",
        "      <td>      NaN</td>\n",
        "    </tr>\n",
        "    <tr>\n",
        "      <th>GO:0000042</th>\n",
        "      <td> 0</td>\n",
        "      <td> 0</td>\n",
        "      <td>      NaN</td>\n",
        "    </tr>\n",
        "    <tr>\n",
        "      <th>GO:0000045</th>\n",
        "      <td> 0</td>\n",
        "      <td> 0</td>\n",
        "      <td> 0.051480</td>\n",
        "    </tr>\n",
        "    <tr>\n",
        "      <th>GO:0000046</th>\n",
        "      <td> 0</td>\n",
        "      <td> 0</td>\n",
        "      <td>      NaN</td>\n",
        "    </tr>\n",
        "    <tr>\n",
        "      <th>GO:0000049</th>\n",
        "      <td> 0</td>\n",
        "      <td> 0</td>\n",
        "      <td> 0.250136</td>\n",
        "    </tr>\n",
        "    <tr>\n",
        "      <th>GO:0000050</th>\n",
        "      <td> 0</td>\n",
        "      <td> 0</td>\n",
        "      <td>      NaN</td>\n",
        "    </tr>\n",
        "    <tr>\n",
        "      <th>GO:0000052</th>\n",
        "      <td> 0</td>\n",
        "      <td> 0</td>\n",
        "      <td>      NaN</td>\n",
        "    </tr>\n",
        "    <tr>\n",
        "      <th>GO:0000053</th>\n",
        "      <td> 0</td>\n",
        "      <td> 0</td>\n",
        "      <td>      NaN</td>\n",
        "    </tr>\n",
        "    <tr>\n",
        "      <th>GO:0000054</th>\n",
        "      <td> 0</td>\n",
        "      <td> 0</td>\n",
        "      <td>      NaN</td>\n",
        "    </tr>\n",
        "    <tr>\n",
        "      <th>GO:0000055</th>\n",
        "      <td> 0</td>\n",
        "      <td> 0</td>\n",
        "      <td>      NaN</td>\n",
        "    </tr>\n",
        "    <tr>\n",
        "      <th>...</th>\n",
        "      <td>...</td>\n",
        "      <td>...</td>\n",
        "      <td>...</td>\n",
        "    </tr>\n",
        "    <tr>\n",
        "      <th>GO:2001268</th>\n",
        "      <td> 0</td>\n",
        "      <td> 0</td>\n",
        "      <td>      NaN</td>\n",
        "    </tr>\n",
        "    <tr>\n",
        "      <th>GO:2001269</th>\n",
        "      <td> 0</td>\n",
        "      <td> 0</td>\n",
        "      <td>      NaN</td>\n",
        "    </tr>\n",
        "    <tr>\n",
        "      <th>GO:2001271</th>\n",
        "      <td> 0</td>\n",
        "      <td> 0</td>\n",
        "      <td>      NaN</td>\n",
        "    </tr>\n",
        "    <tr>\n",
        "      <th>GO:2001273</th>\n",
        "      <td> 0</td>\n",
        "      <td> 0</td>\n",
        "      <td>      NaN</td>\n",
        "    </tr>\n",
        "    <tr>\n",
        "      <th>GO:2001274</th>\n",
        "      <td> 0</td>\n",
        "      <td> 0</td>\n",
        "      <td>      NaN</td>\n",
        "    </tr>\n",
        "    <tr>\n",
        "      <th>GO:2001275</th>\n",
        "      <td> 0</td>\n",
        "      <td> 0</td>\n",
        "      <td>      NaN</td>\n",
        "    </tr>\n",
        "    <tr>\n",
        "      <th>GO:2001286</th>\n",
        "      <td> 0</td>\n",
        "      <td> 0</td>\n",
        "      <td>      NaN</td>\n",
        "    </tr>\n",
        "    <tr>\n",
        "      <th>GO:2001287</th>\n",
        "      <td> 0</td>\n",
        "      <td> 0</td>\n",
        "      <td>      NaN</td>\n",
        "    </tr>\n",
        "    <tr>\n",
        "      <th>GO:2001288</th>\n",
        "      <td> 0</td>\n",
        "      <td> 0</td>\n",
        "      <td>      NaN</td>\n",
        "    </tr>\n",
        "    <tr>\n",
        "      <th>GO:2001294</th>\n",
        "      <td> 0</td>\n",
        "      <td> 0</td>\n",
        "      <td>      NaN</td>\n",
        "    </tr>\n",
        "    <tr>\n",
        "      <th>GO:2001295</th>\n",
        "      <td> 0</td>\n",
        "      <td> 0</td>\n",
        "      <td>      NaN</td>\n",
        "    </tr>\n",
        "    <tr>\n",
        "      <th>GO:2001300</th>\n",
        "      <td> 0</td>\n",
        "      <td> 0</td>\n",
        "      <td>      NaN</td>\n",
        "    </tr>\n",
        "    <tr>\n",
        "      <th>GO:2001303</th>\n",
        "      <td> 0</td>\n",
        "      <td> 0</td>\n",
        "      <td>      NaN</td>\n",
        "    </tr>\n",
        "    <tr>\n",
        "      <th>GO:2001306</th>\n",
        "      <td> 0</td>\n",
        "      <td> 0</td>\n",
        "      <td>      NaN</td>\n",
        "    </tr>\n",
        "    <tr>\n",
        "      <th>GO:2001311</th>\n",
        "      <td> 0</td>\n",
        "      <td> 0</td>\n",
        "      <td>      NaN</td>\n",
        "    </tr>\n",
        "    <tr>\n",
        "      <th>hiv155030.0</th>\n",
        "      <td> 0</td>\n",
        "      <td> 0</td>\n",
        "      <td>      NaN</td>\n",
        "    </tr>\n",
        "    <tr>\n",
        "      <th>hiv155348.0</th>\n",
        "      <td> 0</td>\n",
        "      <td> 0</td>\n",
        "      <td> 0.864496</td>\n",
        "    </tr>\n",
        "    <tr>\n",
        "      <th>hiv155459.0</th>\n",
        "      <td> 0</td>\n",
        "      <td> 0</td>\n",
        "      <td> 0.569378</td>\n",
        "    </tr>\n",
        "    <tr>\n",
        "      <th>hiv155807.0</th>\n",
        "      <td> 0</td>\n",
        "      <td> 0</td>\n",
        "      <td> 0.022611</td>\n",
        "    </tr>\n",
        "    <tr>\n",
        "      <th>hiv155871.0</th>\n",
        "      <td> 0</td>\n",
        "      <td> 0</td>\n",
        "      <td> 2.061431</td>\n",
        "    </tr>\n",
        "    <tr>\n",
        "      <th>hiv155908.0</th>\n",
        "      <td> 0</td>\n",
        "      <td> 0</td>\n",
        "      <td> 0.703132</td>\n",
        "    </tr>\n",
        "    <tr>\n",
        "      <th>hiv155945.0</th>\n",
        "      <td> 0</td>\n",
        "      <td> 0</td>\n",
        "      <td> 0.004365</td>\n",
        "    </tr>\n",
        "    <tr>\n",
        "      <th>hiv155971.0</th>\n",
        "      <td> 0</td>\n",
        "      <td> 0</td>\n",
        "      <td> 0.968794</td>\n",
        "    </tr>\n",
        "    <tr>\n",
        "      <th>hiv156110.0</th>\n",
        "      <td> 0</td>\n",
        "      <td> 0</td>\n",
        "      <td> 0.007835</td>\n",
        "    </tr>\n",
        "    <tr>\n",
        "      <th>hiv19424028.0</th>\n",
        "      <td> 0</td>\n",
        "      <td> 0</td>\n",
        "      <td>      NaN</td>\n",
        "    </tr>\n",
        "    <tr>\n",
        "      <th>hs_hivprot</th>\n",
        "      <td> 0</td>\n",
        "      <td> 0</td>\n",
        "      <td> 2.769919</td>\n",
        "    </tr>\n",
        "    <tr>\n",
        "      <th>hs_hivprotphen</th>\n",
        "      <td> 0</td>\n",
        "      <td> 0</td>\n",
        "      <td> 2.249645</td>\n",
        "    </tr>\n",
        "    <tr>\n",
        "      <th>hs_hivrep</th>\n",
        "      <td> 0</td>\n",
        "      <td> 0</td>\n",
        "      <td> 1.163367</td>\n",
        "    </tr>\n",
        "    <tr>\n",
        "      <th>hs_hivrepphen</th>\n",
        "      <td> 0</td>\n",
        "      <td> 0</td>\n",
        "      <td> 0.710967</td>\n",
        "    </tr>\n",
        "    <tr>\n",
        "      <th>hs_sivcand</th>\n",
        "      <td> 0</td>\n",
        "      <td> 0</td>\n",
        "      <td>      NaN</td>\n",
        "    </tr>\n",
        "  </tbody>\n",
        "</table>\n",
        "<p>15013 rows \u00d7 3 columns</p>\n",
        "</div>"
       ],
       "metadata": {},
       "output_type": "pyout",
       "prompt_number": 213,
       "text": [
        "                out_of  rank     value\n",
        "category                              \n",
        "GO:0000002           0     0  0.007469\n",
        "GO:0000003           0     0       NaN\n",
        "GO:0000009           0     0       NaN\n",
        "GO:0000010           0     0       NaN\n",
        "GO:0000012           0     0       NaN\n",
        "GO:0000014           0     0       NaN\n",
        "GO:0000015           0     0       NaN\n",
        "GO:0000016           0     0       NaN\n",
        "GO:0000018           0     0       NaN\n",
        "GO:0000019           0     0       NaN\n",
        "GO:0000022           0     0       NaN\n",
        "GO:0000023           0     0       NaN\n",
        "GO:0000026           0     0       NaN\n",
        "GO:0000027           0     0       NaN\n",
        "GO:0000028           0     0       NaN\n",
        "GO:0000030           0     0       NaN\n",
        "GO:0000033           0     0       NaN\n",
        "GO:0000035           0     0       NaN\n",
        "GO:0000036           0     0       NaN\n",
        "GO:0000038           0     0       NaN\n",
        "GO:0000041           0     0       NaN\n",
        "GO:0000042           0     0       NaN\n",
        "GO:0000045           0     0  0.051480\n",
        "GO:0000046           0     0       NaN\n",
        "GO:0000049           0     0  0.250136\n",
        "GO:0000050           0     0       NaN\n",
        "GO:0000052           0     0       NaN\n",
        "GO:0000053           0     0       NaN\n",
        "GO:0000054           0     0       NaN\n",
        "GO:0000055           0     0       NaN\n",
        "...                ...   ...       ...\n",
        "GO:2001268           0     0       NaN\n",
        "GO:2001269           0     0       NaN\n",
        "GO:2001271           0     0       NaN\n",
        "GO:2001273           0     0       NaN\n",
        "GO:2001274           0     0       NaN\n",
        "GO:2001275           0     0       NaN\n",
        "GO:2001286           0     0       NaN\n",
        "GO:2001287           0     0       NaN\n",
        "GO:2001288           0     0       NaN\n",
        "GO:2001294           0     0       NaN\n",
        "GO:2001295           0     0       NaN\n",
        "GO:2001300           0     0       NaN\n",
        "GO:2001303           0     0       NaN\n",
        "GO:2001306           0     0       NaN\n",
        "GO:2001311           0     0       NaN\n",
        "hiv155030.0          0     0       NaN\n",
        "hiv155348.0          0     0  0.864496\n",
        "hiv155459.0          0     0  0.569378\n",
        "hiv155807.0          0     0  0.022611\n",
        "hiv155871.0          0     0  2.061431\n",
        "hiv155908.0          0     0  0.703132\n",
        "hiv155945.0          0     0  0.004365\n",
        "hiv155971.0          0     0  0.968794\n",
        "hiv156110.0          0     0  0.007835\n",
        "hiv19424028.0        0     0       NaN\n",
        "hs_hivprot           0     0  2.769919\n",
        "hs_hivprotphen       0     0  2.249645\n",
        "hs_hivrep            0     0  1.163367\n",
        "hs_hivrepphen        0     0  0.710967\n",
        "hs_sivcand           0     0       NaN\n",
        "\n",
        "[15013 rows x 3 columns]"
       ]
      }
     ],
     "prompt_number": 213
    },
    {
     "cell_type": "code",
     "collapsed": false,
     "input": [
      "se.permute(3)"
     ],
     "language": "python",
     "metadata": {},
     "outputs": [],
     "prompt_number": 168
    },
    {
     "cell_type": "code",
     "collapsed": false,
     "input": [
      "se.rank_table"
     ],
     "language": "python",
     "metadata": {},
     "outputs": [
      {
       "html": [
        "<div style=\"max-height:1000px;max-width:1500px;overflow:auto;\">\n",
        "<table border=\"1\" class=\"dataframe\">\n",
        "  <thead>\n",
        "    <tr style=\"text-align: right;\">\n",
        "      <th></th>\n",
        "      <th>out_of</th>\n",
        "      <th>rank</th>\n",
        "      <th>value</th>\n",
        "    </tr>\n",
        "    <tr>\n",
        "      <th>category</th>\n",
        "      <th></th>\n",
        "      <th></th>\n",
        "      <th></th>\n",
        "    </tr>\n",
        "  </thead>\n",
        "  <tbody>\n",
        "    <tr>\n",
        "      <th>GO:0036025</th>\n",
        "      <td> 3</td>\n",
        "      <td> 3</td>\n",
        "      <td> 3.962679</td>\n",
        "    </tr>\n",
        "    <tr>\n",
        "      <th>GO:0006592</th>\n",
        "      <td> 3</td>\n",
        "      <td> 3</td>\n",
        "      <td> 4.770869</td>\n",
        "    </tr>\n",
        "    <tr>\n",
        "      <th>GO:0046456</th>\n",
        "      <td> 3</td>\n",
        "      <td> 3</td>\n",
        "      <td> 3.038312</td>\n",
        "    </tr>\n",
        "    <tr>\n",
        "      <th>GO:0030247</th>\n",
        "      <td> 3</td>\n",
        "      <td> 3</td>\n",
        "      <td> 2.690239</td>\n",
        "    </tr>\n",
        "    <tr>\n",
        "      <th>GO:0030238</th>\n",
        "      <td> 3</td>\n",
        "      <td> 3</td>\n",
        "      <td> 4.034693</td>\n",
        "    </tr>\n",
        "    <tr>\n",
        "      <th>GO:0006568</th>\n",
        "      <td> 3</td>\n",
        "      <td> 3</td>\n",
        "      <td> 2.521666</td>\n",
        "    </tr>\n",
        "    <tr>\n",
        "      <th>GO:0046475</th>\n",
        "      <td> 3</td>\n",
        "      <td> 3</td>\n",
        "      <td> 5.769107</td>\n",
        "    </tr>\n",
        "    <tr>\n",
        "      <th>GO:0006573</th>\n",
        "      <td> 3</td>\n",
        "      <td> 3</td>\n",
        "      <td> 3.626865</td>\n",
        "    </tr>\n",
        "    <tr>\n",
        "      <th>GO:0030228</th>\n",
        "      <td> 3</td>\n",
        "      <td> 3</td>\n",
        "      <td> 1.826372</td>\n",
        "    </tr>\n",
        "    <tr>\n",
        "      <th>GO:0030225</th>\n",
        "      <td> 3</td>\n",
        "      <td> 3</td>\n",
        "      <td> 1.919870</td>\n",
        "    </tr>\n",
        "    <tr>\n",
        "      <th>GO:0006577</th>\n",
        "      <td> 3</td>\n",
        "      <td> 3</td>\n",
        "      <td> 2.563207</td>\n",
        "    </tr>\n",
        "    <tr>\n",
        "      <th>GO:0006579</th>\n",
        "      <td> 3</td>\n",
        "      <td> 3</td>\n",
        "      <td> 6.050205</td>\n",
        "    </tr>\n",
        "    <tr>\n",
        "      <th>GO:0030223</th>\n",
        "      <td> 3</td>\n",
        "      <td> 3</td>\n",
        "      <td> 4.336767</td>\n",
        "    </tr>\n",
        "    <tr>\n",
        "      <th>GO:0030222</th>\n",
        "      <td> 3</td>\n",
        "      <td> 3</td>\n",
        "      <td> 7.435727</td>\n",
        "    </tr>\n",
        "    <tr>\n",
        "      <th>GO:0006595</th>\n",
        "      <td> 3</td>\n",
        "      <td> 3</td>\n",
        "      <td> 2.467605</td>\n",
        "    </tr>\n",
        "    <tr>\n",
        "      <th>GO:0006556</th>\n",
        "      <td> 3</td>\n",
        "      <td> 3</td>\n",
        "      <td> 3.057228</td>\n",
        "    </tr>\n",
        "    <tr>\n",
        "      <th>GO:0030215</th>\n",
        "      <td> 3</td>\n",
        "      <td> 3</td>\n",
        "      <td> 6.233193</td>\n",
        "    </tr>\n",
        "    <tr>\n",
        "      <th>GO:0006597</th>\n",
        "      <td> 3</td>\n",
        "      <td> 3</td>\n",
        "      <td> 3.073089</td>\n",
        "    </tr>\n",
        "    <tr>\n",
        "      <th>GO:0030213</th>\n",
        "      <td> 3</td>\n",
        "      <td> 3</td>\n",
        "      <td> 1.675088</td>\n",
        "    </tr>\n",
        "    <tr>\n",
        "      <th>GO:0046523</th>\n",
        "      <td> 3</td>\n",
        "      <td> 3</td>\n",
        "      <td> 1.412485</td>\n",
        "    </tr>\n",
        "    <tr>\n",
        "      <th>GO:0030209</th>\n",
        "      <td> 3</td>\n",
        "      <td> 3</td>\n",
        "      <td> 1.675432</td>\n",
        "    </tr>\n",
        "    <tr>\n",
        "      <th>GO:0030208</th>\n",
        "      <td> 3</td>\n",
        "      <td> 3</td>\n",
        "      <td> 3.667450</td>\n",
        "    </tr>\n",
        "    <tr>\n",
        "      <th>GO:0070539</th>\n",
        "      <td> 3</td>\n",
        "      <td> 3</td>\n",
        "      <td> 2.010929</td>\n",
        "    </tr>\n",
        "    <tr>\n",
        "      <th>GO:0070537</th>\n",
        "      <td> 3</td>\n",
        "      <td> 3</td>\n",
        "      <td> 3.552483</td>\n",
        "    </tr>\n",
        "    <tr>\n",
        "      <th>GO:0070536</th>\n",
        "      <td> 3</td>\n",
        "      <td> 3</td>\n",
        "      <td> 3.571566</td>\n",
        "    </tr>\n",
        "    <tr>\n",
        "      <th>GO:0030206</th>\n",
        "      <td> 3</td>\n",
        "      <td> 3</td>\n",
        "      <td> 3.633931</td>\n",
        "    </tr>\n",
        "    <tr>\n",
        "      <th>GO:0030204</th>\n",
        "      <td> 3</td>\n",
        "      <td> 3</td>\n",
        "      <td> 3.244351</td>\n",
        "    </tr>\n",
        "    <tr>\n",
        "      <th>GO:0030203</th>\n",
        "      <td> 3</td>\n",
        "      <td> 3</td>\n",
        "      <td> 2.814680</td>\n",
        "    </tr>\n",
        "    <tr>\n",
        "      <th>GO:0006557</th>\n",
        "      <td> 3</td>\n",
        "      <td> 3</td>\n",
        "      <td> 1.328816</td>\n",
        "    </tr>\n",
        "    <tr>\n",
        "      <th>GO:0006555</th>\n",
        "      <td> 3</td>\n",
        "      <td> 3</td>\n",
        "      <td> 4.140163</td>\n",
        "    </tr>\n",
        "    <tr>\n",
        "      <th>...</th>\n",
        "      <td>...</td>\n",
        "      <td>...</td>\n",
        "      <td>...</td>\n",
        "    </tr>\n",
        "    <tr>\n",
        "      <th>GO:0016286</th>\n",
        "      <td> 3</td>\n",
        "      <td> 0</td>\n",
        "      <td> 1.774311</td>\n",
        "    </tr>\n",
        "    <tr>\n",
        "      <th>GO:0050856</th>\n",
        "      <td> 3</td>\n",
        "      <td> 0</td>\n",
        "      <td> 0.732568</td>\n",
        "    </tr>\n",
        "    <tr>\n",
        "      <th>GO:0016298</th>\n",
        "      <td> 3</td>\n",
        "      <td> 0</td>\n",
        "      <td> 2.856799</td>\n",
        "    </tr>\n",
        "    <tr>\n",
        "      <th>GO:0016301</th>\n",
        "      <td> 3</td>\n",
        "      <td> 0</td>\n",
        "      <td> 2.818602</td>\n",
        "    </tr>\n",
        "    <tr>\n",
        "      <th>GO:0016303</th>\n",
        "      <td> 3</td>\n",
        "      <td> 0</td>\n",
        "      <td> 2.077833</td>\n",
        "    </tr>\n",
        "    <tr>\n",
        "      <th>GO:0050849</th>\n",
        "      <td> 3</td>\n",
        "      <td> 0</td>\n",
        "      <td> 0.639202</td>\n",
        "    </tr>\n",
        "    <tr>\n",
        "      <th>GO:0016310</th>\n",
        "      <td> 3</td>\n",
        "      <td> 0</td>\n",
        "      <td> 2.810935</td>\n",
        "    </tr>\n",
        "    <tr>\n",
        "      <th>GO:0050832</th>\n",
        "      <td> 3</td>\n",
        "      <td> 0</td>\n",
        "      <td> 1.073443</td>\n",
        "    </tr>\n",
        "    <tr>\n",
        "      <th>GO:0016324</th>\n",
        "      <td> 3</td>\n",
        "      <td> 0</td>\n",
        "      <td> 2.672580</td>\n",
        "    </tr>\n",
        "    <tr>\n",
        "      <th>GO:0016248</th>\n",
        "      <td> 3</td>\n",
        "      <td> 0</td>\n",
        "      <td> 0.444729</td>\n",
        "    </tr>\n",
        "    <tr>\n",
        "      <th>GO:0050878</th>\n",
        "      <td> 3</td>\n",
        "      <td> 0</td>\n",
        "      <td> 1.263935</td>\n",
        "    </tr>\n",
        "    <tr>\n",
        "      <th>GO:0050880</th>\n",
        "      <td> 3</td>\n",
        "      <td> 0</td>\n",
        "      <td> 0.661373</td>\n",
        "    </tr>\n",
        "    <tr>\n",
        "      <th>GO:0016192</th>\n",
        "      <td> 3</td>\n",
        "      <td> 0</td>\n",
        "      <td> 2.153042</td>\n",
        "    </tr>\n",
        "    <tr>\n",
        "      <th>GO:0016095</th>\n",
        "      <td> 3</td>\n",
        "      <td> 0</td>\n",
        "      <td> 1.011537</td>\n",
        "    </tr>\n",
        "    <tr>\n",
        "      <th>GO:0016098</th>\n",
        "      <td> 3</td>\n",
        "      <td> 0</td>\n",
        "      <td>      NaN</td>\n",
        "    </tr>\n",
        "    <tr>\n",
        "      <th>GO:0016125</th>\n",
        "      <td> 3</td>\n",
        "      <td> 0</td>\n",
        "      <td> 0.722103</td>\n",
        "    </tr>\n",
        "    <tr>\n",
        "      <th>GO:0016150</th>\n",
        "      <td> 3</td>\n",
        "      <td> 0</td>\n",
        "      <td> 0.102355</td>\n",
        "    </tr>\n",
        "    <tr>\n",
        "      <th>GO:0050911</th>\n",
        "      <td> 3</td>\n",
        "      <td> 0</td>\n",
        "      <td>      NaN</td>\n",
        "    </tr>\n",
        "    <tr>\n",
        "      <th>GO:0016154</th>\n",
        "      <td> 3</td>\n",
        "      <td> 0</td>\n",
        "      <td> 0.046945</td>\n",
        "    </tr>\n",
        "    <tr>\n",
        "      <th>GO:0016155</th>\n",
        "      <td> 3</td>\n",
        "      <td> 0</td>\n",
        "      <td> 1.406975</td>\n",
        "    </tr>\n",
        "    <tr>\n",
        "      <th>GO:0050902</th>\n",
        "      <td> 3</td>\n",
        "      <td> 0</td>\n",
        "      <td> 0.413836</td>\n",
        "    </tr>\n",
        "    <tr>\n",
        "      <th>GO:0016246</th>\n",
        "      <td> 3</td>\n",
        "      <td> 0</td>\n",
        "      <td> 1.558582</td>\n",
        "    </tr>\n",
        "    <tr>\n",
        "      <th>GO:0050901</th>\n",
        "      <td> 3</td>\n",
        "      <td> 0</td>\n",
        "      <td> 0.938901</td>\n",
        "    </tr>\n",
        "    <tr>\n",
        "      <th>GO:0050900</th>\n",
        "      <td> 3</td>\n",
        "      <td> 0</td>\n",
        "      <td> 2.154302</td>\n",
        "    </tr>\n",
        "    <tr>\n",
        "      <th>GO:0016198</th>\n",
        "      <td> 3</td>\n",
        "      <td> 0</td>\n",
        "      <td> 0.295629</td>\n",
        "    </tr>\n",
        "    <tr>\n",
        "      <th>GO:0016199</th>\n",
        "      <td> 3</td>\n",
        "      <td> 0</td>\n",
        "      <td> 5.047951</td>\n",
        "    </tr>\n",
        "    <tr>\n",
        "      <th>GO:0016206</th>\n",
        "      <td> 3</td>\n",
        "      <td> 0</td>\n",
        "      <td> 0.084573</td>\n",
        "    </tr>\n",
        "    <tr>\n",
        "      <th>GO:0016209</th>\n",
        "      <td> 3</td>\n",
        "      <td> 0</td>\n",
        "      <td> 0.655857</td>\n",
        "    </tr>\n",
        "    <tr>\n",
        "      <th>GO:0050891</th>\n",
        "      <td> 3</td>\n",
        "      <td> 0</td>\n",
        "      <td> 1.507443</td>\n",
        "    </tr>\n",
        "    <tr>\n",
        "      <th>GO:0033580</th>\n",
        "      <td> 3</td>\n",
        "      <td> 0</td>\n",
        "      <td>      NaN</td>\n",
        "    </tr>\n",
        "  </tbody>\n",
        "</table>\n",
        "<p>15013 rows \u00d7 3 columns</p>\n",
        "</div>"
       ],
       "metadata": {},
       "output_type": "pyout",
       "prompt_number": 169,
       "text": [
        "            out_of  rank     value\n",
        "category                          \n",
        "GO:0036025       3     3  3.962679\n",
        "GO:0006592       3     3  4.770869\n",
        "GO:0046456       3     3  3.038312\n",
        "GO:0030247       3     3  2.690239\n",
        "GO:0030238       3     3  4.034693\n",
        "GO:0006568       3     3  2.521666\n",
        "GO:0046475       3     3  5.769107\n",
        "GO:0006573       3     3  3.626865\n",
        "GO:0030228       3     3  1.826372\n",
        "GO:0030225       3     3  1.919870\n",
        "GO:0006577       3     3  2.563207\n",
        "GO:0006579       3     3  6.050205\n",
        "GO:0030223       3     3  4.336767\n",
        "GO:0030222       3     3  7.435727\n",
        "GO:0006595       3     3  2.467605\n",
        "GO:0006556       3     3  3.057228\n",
        "GO:0030215       3     3  6.233193\n",
        "GO:0006597       3     3  3.073089\n",
        "GO:0030213       3     3  1.675088\n",
        "GO:0046523       3     3  1.412485\n",
        "GO:0030209       3     3  1.675432\n",
        "GO:0030208       3     3  3.667450\n",
        "GO:0070539       3     3  2.010929\n",
        "GO:0070537       3     3  3.552483\n",
        "GO:0070536       3     3  3.571566\n",
        "GO:0030206       3     3  3.633931\n",
        "GO:0030204       3     3  3.244351\n",
        "GO:0030203       3     3  2.814680\n",
        "GO:0006557       3     3  1.328816\n",
        "GO:0006555       3     3  4.140163\n",
        "...            ...   ...       ...\n",
        "GO:0016286       3     0  1.774311\n",
        "GO:0050856       3     0  0.732568\n",
        "GO:0016298       3     0  2.856799\n",
        "GO:0016301       3     0  2.818602\n",
        "GO:0016303       3     0  2.077833\n",
        "GO:0050849       3     0  0.639202\n",
        "GO:0016310       3     0  2.810935\n",
        "GO:0050832       3     0  1.073443\n",
        "GO:0016324       3     0  2.672580\n",
        "GO:0016248       3     0  0.444729\n",
        "GO:0050878       3     0  1.263935\n",
        "GO:0050880       3     0  0.661373\n",
        "GO:0016192       3     0  2.153042\n",
        "GO:0016095       3     0  1.011537\n",
        "GO:0016098       3     0       NaN\n",
        "GO:0016125       3     0  0.722103\n",
        "GO:0016150       3     0  0.102355\n",
        "GO:0050911       3     0       NaN\n",
        "GO:0016154       3     0  0.046945\n",
        "GO:0016155       3     0  1.406975\n",
        "GO:0050902       3     0  0.413836\n",
        "GO:0016246       3     0  1.558582\n",
        "GO:0050901       3     0  0.938901\n",
        "GO:0050900       3     0  2.154302\n",
        "GO:0016198       3     0  0.295629\n",
        "GO:0016199       3     0  5.047951\n",
        "GO:0016206       3     0  0.084573\n",
        "GO:0016209       3     0  0.655857\n",
        "GO:0050891       3     0  1.507443\n",
        "GO:0033580       3     0       NaN\n",
        "\n",
        "[15013 rows x 3 columns]"
       ]
      }
     ],
     "prompt_number": 169
    },
    {
     "cell_type": "code",
     "collapsed": false,
     "input": [
      "s = te.shift_rod(rod_s, rnd)\n",
      "spf0 = get_summary_per_feature(s,gene_df,summary='max')\n",
      "spc0 = get_summary_per_category(spf0,gene_to_cat,'median')"
     ],
     "language": "python",
     "metadata": {},
     "outputs": [],
     "prompt_number": 109
    },
    {
     "cell_type": "code",
     "collapsed": false,
     "input": [
      "spc0 = spc0['value']"
     ],
     "language": "python",
     "metadata": {},
     "outputs": [],
     "prompt_number": 110
    },
    {
     "cell_type": "code",
     "collapsed": false,
     "input": [
      "(init_rank_table[\"n_genes\"] > \\\n",
      "                spc0.reindex(init_rank_table.index).fillna(0))"
     ],
     "language": "python",
     "metadata": {},
     "outputs": [
      {
       "metadata": {},
       "output_type": "pyout",
       "prompt_number": 115,
       "text": [
        "category\n",
        "GO:0000002    False\n",
        "GO:0000003    False\n",
        "GO:0000009    False\n",
        "GO:0000010    False\n",
        "GO:0000012    False\n",
        "GO:0000014    False\n",
        "GO:0000015    False\n",
        "GO:0000016     True\n",
        "GO:0000018    False\n",
        "GO:0000019    False\n",
        "GO:0000022    False\n",
        "GO:0000023     True\n",
        "GO:0000026    False\n",
        "GO:0000027    False\n",
        "GO:0000028     True\n",
        "...\n",
        "hiv155030.0       False\n",
        "hiv155348.0       False\n",
        "hiv155459.0       False\n",
        "hiv155807.0       False\n",
        "hiv155871.0       False\n",
        "hiv155908.0       False\n",
        "hiv155945.0       False\n",
        "hiv155971.0       False\n",
        "hiv156110.0       False\n",
        "hiv19424028.0     False\n",
        "hs_hivprot        False\n",
        "hs_hivprotphen    False\n",
        "hs_hivrep         False\n",
        "hs_hivrepphen     False\n",
        "hs_sivcand        False\n",
        "Length: 15013, dtype: bool"
       ]
      }
     ],
     "prompt_number": 115
    },
    {
     "cell_type": "code",
     "collapsed": false,
     "input": [
      "n = 100000"
     ],
     "language": "python",
     "metadata": {},
     "outputs": [],
     "prompt_number": 148
    },
    {
     "cell_type": "code",
     "collapsed": false,
     "input": [
      "%%timeit \n",
      "for i in np.random.rand(n):\n",
      "    pass"
     ],
     "language": "python",
     "metadata": {},
     "outputs": [
      {
       "output_type": "stream",
       "stream": "stdout",
       "text": [
        "100 loops, best of 3: 9.37 ms per loop\n"
       ]
      }
     ],
     "prompt_number": 149
    },
    {
     "cell_type": "code",
     "collapsed": false,
     "input": [
      "%%timeit \n",
      "for _ in range(n):\n",
      "    pass\n",
      "    #rnd = np.random.rand()"
     ],
     "language": "python",
     "metadata": {},
     "outputs": [
      {
       "output_type": "stream",
       "stream": "stdout",
       "text": [
        "100 loops, best of 3: 3.36 ms per loop\n"
       ]
      }
     ],
     "prompt_number": 152
    },
    {
     "cell_type": "code",
     "collapsed": false,
     "input": [
      "def multiple_permut_summary(rod_s, init_rank_table, gene_df, gene_to_cat, n_runs):\n",
      "    for rnd in np.random.rand(n_runs):\n",
      "        s = te.shift_rod(rod_s, rnd)\n",
      "        spf0 = get_summary_per_feature(s,gene_df,summary='max')\n",
      "        spc0 = get_summary_per_category(spf0,gene_to_cat,'mean')\n",
      "        init_rank_table[\"rank\"] += (init_rank_table[\"n_genes\"] > \\\n",
      "                                        spc0.reindex(init_rank_table.index).fillna(0))\n",
      "        init_rank_table[\"out_of\"] += 1\n",
      "    return init_rank_table"
     ],
     "language": "python",
     "metadata": {},
     "outputs": [],
     "prompt_number": 125
    },
    {
     "cell_type": "code",
     "collapsed": false,
     "input": [
      "rt = multiple_permut_summary(rod_s, init_rank_table, gene_df, gene_to_cat, 500)"
     ],
     "language": "python",
     "metadata": {},
     "outputs": []
    },
    {
     "cell_type": "code",
     "collapsed": false,
     "input": [
      "init_rank_table.sort('rank',ascending=False)"
     ],
     "language": "python",
     "metadata": {},
     "outputs": [
      {
       "html": [
        "<div style=\"max-height:1000px;max-width:1500px;overflow:auto;\">\n",
        "<table border=\"1\" class=\"dataframe\">\n",
        "  <thead>\n",
        "    <tr style=\"text-align: right;\">\n",
        "      <th></th>\n",
        "      <th>n_genes</th>\n",
        "      <th>out_of</th>\n",
        "      <th>rank</th>\n",
        "    </tr>\n",
        "    <tr>\n",
        "      <th>category</th>\n",
        "      <th></th>\n",
        "      <th></th>\n",
        "      <th></th>\n",
        "    </tr>\n",
        "  </thead>\n",
        "  <tbody>\n",
        "    <tr>\n",
        "      <th>GO:0071412</th>\n",
        "      <td> 17.441990</td>\n",
        "      <td> 500</td>\n",
        "      <td> 500</td>\n",
        "    </tr>\n",
        "    <tr>\n",
        "      <th>GO:0008281</th>\n",
        "      <td> 23.258225</td>\n",
        "      <td> 500</td>\n",
        "      <td> 500</td>\n",
        "    </tr>\n",
        "    <tr>\n",
        "      <th>GO:0005333</th>\n",
        "      <td> 27.544410</td>\n",
        "      <td> 500</td>\n",
        "      <td> 500</td>\n",
        "    </tr>\n",
        "    <tr>\n",
        "      <th>GO:0090038</th>\n",
        "      <td> 15.423555</td>\n",
        "      <td> 500</td>\n",
        "      <td> 500</td>\n",
        "    </tr>\n",
        "    <tr>\n",
        "      <th>GO:0009331</th>\n",
        "      <td> 12.012225</td>\n",
        "      <td> 500</td>\n",
        "      <td> 500</td>\n",
        "    </tr>\n",
        "    <tr>\n",
        "      <th>GO:0005334</th>\n",
        "      <td> 27.230980</td>\n",
        "      <td> 500</td>\n",
        "      <td> 500</td>\n",
        "    </tr>\n",
        "    <tr>\n",
        "      <th>GO:0032009</th>\n",
        "      <td> 10.860521</td>\n",
        "      <td> 500</td>\n",
        "      <td> 500</td>\n",
        "    </tr>\n",
        "    <tr>\n",
        "      <th>GO:0008504</th>\n",
        "      <td> 13.227751</td>\n",
        "      <td> 500</td>\n",
        "      <td> 500</td>\n",
        "    </tr>\n",
        "    <tr>\n",
        "      <th>GO:0061092</th>\n",
        "      <td> 48.617600</td>\n",
        "      <td> 500</td>\n",
        "      <td> 500</td>\n",
        "    </tr>\n",
        "    <tr>\n",
        "      <th>GO:0015844</th>\n",
        "      <td> 10.112849</td>\n",
        "      <td> 500</td>\n",
        "      <td> 500</td>\n",
        "    </tr>\n",
        "    <tr>\n",
        "      <th>GO:0015874</th>\n",
        "      <td> 27.544410</td>\n",
        "      <td> 500</td>\n",
        "      <td> 500</td>\n",
        "    </tr>\n",
        "    <tr>\n",
        "      <th>GO:0030822</th>\n",
        "      <td> 23.503007</td>\n",
        "      <td> 500</td>\n",
        "      <td> 500</td>\n",
        "    </tr>\n",
        "    <tr>\n",
        "      <th>GO:0003011</th>\n",
        "      <td> 48.617600</td>\n",
        "      <td> 500</td>\n",
        "      <td> 500</td>\n",
        "    </tr>\n",
        "    <tr>\n",
        "      <th>GO:0004367</th>\n",
        "      <td> 16.006388</td>\n",
        "      <td> 500</td>\n",
        "      <td> 499</td>\n",
        "    </tr>\n",
        "    <tr>\n",
        "      <th>GO:0030380</th>\n",
        "      <td> 14.008100</td>\n",
        "      <td> 500</td>\n",
        "      <td> 499</td>\n",
        "    </tr>\n",
        "    <tr>\n",
        "      <th>GO:0010248</th>\n",
        "      <td>  9.383399</td>\n",
        "      <td> 500</td>\n",
        "      <td> 499</td>\n",
        "    </tr>\n",
        "    <tr>\n",
        "      <th>GO:0090237</th>\n",
        "      <td> 26.309300</td>\n",
        "      <td> 500</td>\n",
        "      <td> 499</td>\n",
        "    </tr>\n",
        "    <tr>\n",
        "      <th>GO:0032928</th>\n",
        "      <td> 26.309300</td>\n",
        "      <td> 500</td>\n",
        "      <td> 499</td>\n",
        "    </tr>\n",
        "    <tr>\n",
        "      <th>GO:0071226</th>\n",
        "      <td> 26.309300</td>\n",
        "      <td> 500</td>\n",
        "      <td> 499</td>\n",
        "    </tr>\n",
        "    <tr>\n",
        "      <th>GO:0008513</th>\n",
        "      <td> 14.836556</td>\n",
        "      <td> 500</td>\n",
        "      <td> 499</td>\n",
        "    </tr>\n",
        "    <tr>\n",
        "      <th>GO:0008603</th>\n",
        "      <td>  9.813971</td>\n",
        "      <td> 500</td>\n",
        "      <td> 499</td>\n",
        "    </tr>\n",
        "    <tr>\n",
        "      <th>GO:0045401</th>\n",
        "      <td> 13.164026</td>\n",
        "      <td> 500</td>\n",
        "      <td> 499</td>\n",
        "    </tr>\n",
        "    <tr>\n",
        "      <th>GO:0046168</th>\n",
        "      <td> 16.006388</td>\n",
        "      <td> 500</td>\n",
        "      <td> 499</td>\n",
        "    </tr>\n",
        "    <tr>\n",
        "      <th>GO:0071799</th>\n",
        "      <td> 20.026342</td>\n",
        "      <td> 500</td>\n",
        "      <td> 499</td>\n",
        "    </tr>\n",
        "    <tr>\n",
        "      <th>GO:0043313</th>\n",
        "      <td> 13.536557</td>\n",
        "      <td> 500</td>\n",
        "      <td> 499</td>\n",
        "    </tr>\n",
        "    <tr>\n",
        "      <th>GO:0048241</th>\n",
        "      <td> 27.857840</td>\n",
        "      <td> 500</td>\n",
        "      <td> 499</td>\n",
        "    </tr>\n",
        "    <tr>\n",
        "      <th>GO:0001820</th>\n",
        "      <td> 13.164026</td>\n",
        "      <td> 500</td>\n",
        "      <td> 499</td>\n",
        "    </tr>\n",
        "    <tr>\n",
        "      <th>GO:2000249</th>\n",
        "      <td>  7.184613</td>\n",
        "      <td> 500</td>\n",
        "      <td> 499</td>\n",
        "    </tr>\n",
        "    <tr>\n",
        "      <th>GO:0060739</th>\n",
        "      <td> 31.531200</td>\n",
        "      <td> 500</td>\n",
        "      <td> 499</td>\n",
        "    </tr>\n",
        "    <tr>\n",
        "      <th>GO:1901374</th>\n",
        "      <td> 27.857840</td>\n",
        "      <td> 500</td>\n",
        "      <td> 499</td>\n",
        "    </tr>\n",
        "    <tr>\n",
        "      <th>...</th>\n",
        "      <td>...</td>\n",
        "      <td>...</td>\n",
        "      <td>...</td>\n",
        "    </tr>\n",
        "    <tr>\n",
        "      <th>GO:0006412</th>\n",
        "      <td>  1.346259</td>\n",
        "      <td> 500</td>\n",
        "      <td>   0</td>\n",
        "    </tr>\n",
        "    <tr>\n",
        "      <th>GO:0031324</th>\n",
        "      <td>       NaN</td>\n",
        "      <td> 500</td>\n",
        "      <td>   0</td>\n",
        "    </tr>\n",
        "    <tr>\n",
        "      <th>GO:0031087</th>\n",
        "      <td>  0.000000</td>\n",
        "      <td> 500</td>\n",
        "      <td>   0</td>\n",
        "    </tr>\n",
        "    <tr>\n",
        "      <th>GO:0006784</th>\n",
        "      <td>       NaN</td>\n",
        "      <td> 500</td>\n",
        "      <td>   0</td>\n",
        "    </tr>\n",
        "    <tr>\n",
        "      <th>GO:1902003</th>\n",
        "      <td>       NaN</td>\n",
        "      <td> 500</td>\n",
        "      <td>   0</td>\n",
        "    </tr>\n",
        "    <tr>\n",
        "      <th>GO:0061604</th>\n",
        "      <td>  0.000000</td>\n",
        "      <td> 500</td>\n",
        "      <td>   0</td>\n",
        "    </tr>\n",
        "    <tr>\n",
        "      <th>GO:0006711</th>\n",
        "      <td>       NaN</td>\n",
        "      <td> 500</td>\n",
        "      <td>   0</td>\n",
        "    </tr>\n",
        "    <tr>\n",
        "      <th>GO:0006705</th>\n",
        "      <td>       NaN</td>\n",
        "      <td> 500</td>\n",
        "      <td>   0</td>\n",
        "    </tr>\n",
        "    <tr>\n",
        "      <th>GO:1902045</th>\n",
        "      <td>       NaN</td>\n",
        "      <td> 500</td>\n",
        "      <td>   0</td>\n",
        "    </tr>\n",
        "    <tr>\n",
        "      <th>GO:1902081</th>\n",
        "      <td>       NaN</td>\n",
        "      <td> 500</td>\n",
        "      <td>   0</td>\n",
        "    </tr>\n",
        "    <tr>\n",
        "      <th>GO:0048640</th>\n",
        "      <td>  0.000000</td>\n",
        "      <td> 500</td>\n",
        "      <td>   0</td>\n",
        "    </tr>\n",
        "    <tr>\n",
        "      <th>GO:0031073</th>\n",
        "      <td>       NaN</td>\n",
        "      <td> 500</td>\n",
        "      <td>   0</td>\n",
        "    </tr>\n",
        "    <tr>\n",
        "      <th>GO:0006660</th>\n",
        "      <td>       NaN</td>\n",
        "      <td> 500</td>\n",
        "      <td>   0</td>\n",
        "    </tr>\n",
        "    <tr>\n",
        "      <th>GO:0048668</th>\n",
        "      <td>  0.000000</td>\n",
        "      <td> 500</td>\n",
        "      <td>   0</td>\n",
        "    </tr>\n",
        "    <tr>\n",
        "      <th>GO:0048672</th>\n",
        "      <td>  0.132715</td>\n",
        "      <td> 500</td>\n",
        "      <td>   0</td>\n",
        "    </tr>\n",
        "    <tr>\n",
        "      <th>GO:0006581</th>\n",
        "      <td>  0.000000</td>\n",
        "      <td> 500</td>\n",
        "      <td>   0</td>\n",
        "    </tr>\n",
        "    <tr>\n",
        "      <th>GO:0043105</th>\n",
        "      <td>  0.000000</td>\n",
        "      <td> 500</td>\n",
        "      <td>   0</td>\n",
        "    </tr>\n",
        "    <tr>\n",
        "      <th>GO:0070733</th>\n",
        "      <td>  0.000000</td>\n",
        "      <td> 500</td>\n",
        "      <td>   0</td>\n",
        "    </tr>\n",
        "    <tr>\n",
        "      <th>GO:0048683</th>\n",
        "      <td>       NaN</td>\n",
        "      <td> 500</td>\n",
        "      <td>   0</td>\n",
        "    </tr>\n",
        "    <tr>\n",
        "      <th>GO:1902178</th>\n",
        "      <td>  0.000000</td>\n",
        "      <td> 500</td>\n",
        "      <td>   0</td>\n",
        "    </tr>\n",
        "    <tr>\n",
        "      <th>GO:0048685</th>\n",
        "      <td>  0.000000</td>\n",
        "      <td> 500</td>\n",
        "      <td>   0</td>\n",
        "    </tr>\n",
        "    <tr>\n",
        "      <th>GO:0031129</th>\n",
        "      <td>  0.000000</td>\n",
        "      <td> 500</td>\n",
        "      <td>   0</td>\n",
        "    </tr>\n",
        "    <tr>\n",
        "      <th>GO:0006558</th>\n",
        "      <td>  0.000000</td>\n",
        "      <td> 500</td>\n",
        "      <td>   0</td>\n",
        "    </tr>\n",
        "    <tr>\n",
        "      <th>GO:0006521</th>\n",
        "      <td>  0.910422</td>\n",
        "      <td> 500</td>\n",
        "      <td>   0</td>\n",
        "    </tr>\n",
        "    <tr>\n",
        "      <th>GO:0006507</th>\n",
        "      <td>  0.000000</td>\n",
        "      <td> 500</td>\n",
        "      <td>   0</td>\n",
        "    </tr>\n",
        "    <tr>\n",
        "      <th>GO:0006506</th>\n",
        "      <td>  0.603288</td>\n",
        "      <td> 500</td>\n",
        "      <td>   0</td>\n",
        "    </tr>\n",
        "    <tr>\n",
        "      <th>GO:1902251</th>\n",
        "      <td>  0.000000</td>\n",
        "      <td> 500</td>\n",
        "      <td>   0</td>\n",
        "    </tr>\n",
        "    <tr>\n",
        "      <th>GO:0043112</th>\n",
        "      <td>       NaN</td>\n",
        "      <td> 500</td>\n",
        "      <td>   0</td>\n",
        "    </tr>\n",
        "    <tr>\n",
        "      <th>GO:0072331</th>\n",
        "      <td>       NaN</td>\n",
        "      <td> 500</td>\n",
        "      <td>   0</td>\n",
        "    </tr>\n",
        "    <tr>\n",
        "      <th>GO:0033580</th>\n",
        "      <td>       NaN</td>\n",
        "      <td> 500</td>\n",
        "      <td>   0</td>\n",
        "    </tr>\n",
        "  </tbody>\n",
        "</table>\n",
        "<p>15013 rows \u00d7 3 columns</p>\n",
        "</div>"
       ],
       "metadata": {},
       "output_type": "pyout",
       "prompt_number": 131,
       "text": [
        "              n_genes  out_of  rank\n",
        "category                           \n",
        "GO:0071412  17.441990     500   500\n",
        "GO:0008281  23.258225     500   500\n",
        "GO:0005333  27.544410     500   500\n",
        "GO:0090038  15.423555     500   500\n",
        "GO:0009331  12.012225     500   500\n",
        "GO:0005334  27.230980     500   500\n",
        "GO:0032009  10.860521     500   500\n",
        "GO:0008504  13.227751     500   500\n",
        "GO:0061092  48.617600     500   500\n",
        "GO:0015844  10.112849     500   500\n",
        "GO:0015874  27.544410     500   500\n",
        "GO:0030822  23.503007     500   500\n",
        "GO:0003011  48.617600     500   500\n",
        "GO:0004367  16.006388     500   499\n",
        "GO:0030380  14.008100     500   499\n",
        "GO:0010248   9.383399     500   499\n",
        "GO:0090237  26.309300     500   499\n",
        "GO:0032928  26.309300     500   499\n",
        "GO:0071226  26.309300     500   499\n",
        "GO:0008513  14.836556     500   499\n",
        "GO:0008603   9.813971     500   499\n",
        "GO:0045401  13.164026     500   499\n",
        "GO:0046168  16.006388     500   499\n",
        "GO:0071799  20.026342     500   499\n",
        "GO:0043313  13.536557     500   499\n",
        "GO:0048241  27.857840     500   499\n",
        "GO:0001820  13.164026     500   499\n",
        "GO:2000249   7.184613     500   499\n",
        "GO:0060739  31.531200     500   499\n",
        "GO:1901374  27.857840     500   499\n",
        "...               ...     ...   ...\n",
        "GO:0006412   1.346259     500     0\n",
        "GO:0031324        NaN     500     0\n",
        "GO:0031087   0.000000     500     0\n",
        "GO:0006784        NaN     500     0\n",
        "GO:1902003        NaN     500     0\n",
        "GO:0061604   0.000000     500     0\n",
        "GO:0006711        NaN     500     0\n",
        "GO:0006705        NaN     500     0\n",
        "GO:1902045        NaN     500     0\n",
        "GO:1902081        NaN     500     0\n",
        "GO:0048640   0.000000     500     0\n",
        "GO:0031073        NaN     500     0\n",
        "GO:0006660        NaN     500     0\n",
        "GO:0048668   0.000000     500     0\n",
        "GO:0048672   0.132715     500     0\n",
        "GO:0006581   0.000000     500     0\n",
        "GO:0043105   0.000000     500     0\n",
        "GO:0070733   0.000000     500     0\n",
        "GO:0048683        NaN     500     0\n",
        "GO:1902178   0.000000     500     0\n",
        "GO:0048685   0.000000     500     0\n",
        "GO:0031129   0.000000     500     0\n",
        "GO:0006558   0.000000     500     0\n",
        "GO:0006521   0.910422     500     0\n",
        "GO:0006507   0.000000     500     0\n",
        "GO:0006506   0.603288     500     0\n",
        "GO:1902251   0.000000     500     0\n",
        "GO:0043112        NaN     500     0\n",
        "GO:0072331        NaN     500     0\n",
        "GO:0033580        NaN     500     0\n",
        "\n",
        "[15013 rows x 3 columns]"
       ]
      }
     ],
     "prompt_number": 131
    },
    {
     "cell_type": "code",
     "collapsed": false,
     "input": [
      "s = gene_to_cat.groupby('go_identifier').apply(len)\n",
      "s.name = 'n_genes'"
     ],
     "language": "python",
     "metadata": {},
     "outputs": [],
     "prompt_number": 68
    },
    {
     "cell_type": "code",
     "collapsed": false,
     "input": [
      "pd.DataFrame(gene_to_cat.iloc[:,:2].values.values,columns=[\"hallo\",'kuh'])"
     ],
     "language": "python",
     "metadata": {},
     "outputs": [
      {
       "ename": "AttributeError",
       "evalue": "'numpy.ndarray' object has no attribute 'values'",
       "output_type": "pyerr",
       "traceback": [
        "\u001b[1;31m---------------------------------------------------------------------------\u001b[0m\n\u001b[1;31mAttributeError\u001b[0m                            Traceback (most recent call last)",
        "\u001b[1;32m<ipython-input-138-a9802da52974>\u001b[0m in \u001b[0;36m<module>\u001b[1;34m()\u001b[0m\n\u001b[1;32m----> 1\u001b[1;33m \u001b[0mpd\u001b[0m\u001b[1;33m.\u001b[0m\u001b[0mDataFrame\u001b[0m\u001b[1;33m(\u001b[0m\u001b[0mgene_to_cat\u001b[0m\u001b[1;33m.\u001b[0m\u001b[0miloc\u001b[0m\u001b[1;33m[\u001b[0m\u001b[1;33m:\u001b[0m\u001b[1;33m,\u001b[0m\u001b[1;33m:\u001b[0m\u001b[1;36m2\u001b[0m\u001b[1;33m]\u001b[0m\u001b[1;33m.\u001b[0m\u001b[0mvalues\u001b[0m\u001b[1;33m.\u001b[0m\u001b[0mvalues\u001b[0m\u001b[1;33m,\u001b[0m\u001b[0mcolumns\u001b[0m\u001b[1;33m=\u001b[0m\u001b[1;33m[\u001b[0m\u001b[1;34m\"hallo\"\u001b[0m\u001b[1;33m,\u001b[0m\u001b[1;34m'kuh'\u001b[0m\u001b[1;33m]\u001b[0m\u001b[1;33m)\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n\u001b[0m",
        "\u001b[1;31mAttributeError\u001b[0m: 'numpy.ndarray' object has no attribute 'values'"
       ]
      }
     ],
     "prompt_number": 138
    },
    {
     "cell_type": "code",
     "collapsed": false,
     "input": [
      "data = np.array(range(50))"
     ],
     "language": "python",
     "metadata": {},
     "outputs": [],
     "prompt_number": 257
    },
    {
     "cell_type": "code",
     "collapsed": false,
     "input": [
      "data = pd.DataFrame(range(0,50),index=range(0,100,2))"
     ],
     "language": "python",
     "metadata": {},
     "outputs": [],
     "prompt_number": 281
    },
    {
     "cell_type": "code",
     "collapsed": false,
     "input": [
      "genes = pd.DataFrame({'start':[2,21],'end':[9,26]},index=['gene1','gene2'])"
     ],
     "language": "python",
     "metadata": {},
     "outputs": [],
     "prompt_number": 293
    },
    {
     "cell_type": "code",
     "collapsed": false,
     "input": [
      "data[slice(i,j)] for i,j in ngenes['start'].values,genes['end'].values"
     ],
     "language": "python",
     "metadata": {},
     "outputs": [
      {
       "ename": "SyntaxError",
       "evalue": "invalid syntax (<ipython-input-283-a81fcb2143c2>, line 1)",
       "output_type": "pyerr",
       "traceback": [
        "\u001b[1;36m  File \u001b[1;32m\"<ipython-input-283-a81fcb2143c2>\"\u001b[1;36m, line \u001b[1;32m1\u001b[0m\n\u001b[1;33m    data[slice(i,j)] for i,j in ngenes['start'].values,genes['end'].values\u001b[0m\n\u001b[1;37m                       ^\u001b[0m\n\u001b[1;31mSyntaxError\u001b[0m\u001b[1;31m:\u001b[0m invalid syntax\n"
       ]
      }
     ],
     "prompt_number": 283
    },
    {
     "cell_type": "code",
     "collapsed": false,
     "input": [
      "[data[data.index.slice_indexer(j['start'],j['end'])] for _,j in genes.iterrows()]"
     ],
     "language": "python",
     "metadata": {},
     "outputs": [
      {
       "metadata": {},
       "output_type": "pyout",
       "prompt_number": 296,
       "text": [
        "[   0\n",
        " 2  1\n",
        " 4  2\n",
        " 6  3\n",
        " 8  4,      0\n",
        " 22  11\n",
        " 24  12\n",
        " 26  13]"
       ]
      }
     ],
     "prompt_number": 296
    },
    {
     "cell_type": "code",
     "collapsed": false,
     "input": [
      "%timeit data[data.index.slice_indexer(2, 9)]"
     ],
     "language": "python",
     "metadata": {},
     "outputs": [
      {
       "output_type": "stream",
       "stream": "stdout",
       "text": [
        "10000 loops, best of 3: 129 \u00b5s per loop\n"
       ]
      }
     ],
     "prompt_number": 308
    },
    {
     "cell_type": "code",
     "collapsed": false,
     "input": [
      "%timeit data.ix[slice(2, 9, None)]"
     ],
     "language": "python",
     "metadata": {},
     "outputs": [
      {
       "output_type": "stream",
       "stream": "stdout",
       "text": [
        "10000 loops, best of 3: 134 \u00b5s per loop\n"
       ]
      }
     ],
     "prompt_number": 309
    },
    {
     "cell_type": "code",
     "collapsed": false,
     "input": [],
     "language": "python",
     "metadata": {},
     "outputs": [],
     "prompt_number": 325
    },
    {
     "cell_type": "code",
     "collapsed": false,
     "input": [
      "def gene_values(gene,sl):\n",
      "    gene_values = data.ix[sl].copy()\n",
      "    gene_values['gene'] = gene\n",
      "    gene_values.set_index('gene',append=True,inplace=True)\n",
      "    return gene_values\n",
      "slices = genes.apply(lambda gene:slice(gene['start'],gene['end']),axis=1)\n",
      "pd.concat([gene_values(*sl) for sl in slices.iteritems()])"
     ],
     "language": "python",
     "metadata": {},
     "outputs": [],
     "prompt_number": 343
    },
    {
     "cell_type": "code",
     "collapsed": false,
     "input": [],
     "language": "python",
     "metadata": {},
     "outputs": [
      {
       "html": [
        "<div style=\"max-height:1000px;max-width:1500px;overflow:auto;\">\n",
        "<table border=\"1\" class=\"dataframe\">\n",
        "  <thead>\n",
        "    <tr style=\"text-align: right;\">\n",
        "      <th></th>\n",
        "      <th>0</th>\n",
        "    </tr>\n",
        "  </thead>\n",
        "  <tbody>\n",
        "    <tr>\n",
        "      <th>2 </th>\n",
        "      <td>  1</td>\n",
        "    </tr>\n",
        "    <tr>\n",
        "      <th>4 </th>\n",
        "      <td>  2</td>\n",
        "    </tr>\n",
        "    <tr>\n",
        "      <th>6 </th>\n",
        "      <td>  3</td>\n",
        "    </tr>\n",
        "    <tr>\n",
        "      <th>8 </th>\n",
        "      <td>  4</td>\n",
        "    </tr>\n",
        "    <tr>\n",
        "      <th>22</th>\n",
        "      <td> 11</td>\n",
        "    </tr>\n",
        "    <tr>\n",
        "      <th>24</th>\n",
        "      <td> 12</td>\n",
        "    </tr>\n",
        "    <tr>\n",
        "      <th>26</th>\n",
        "      <td> 13</td>\n",
        "    </tr>\n",
        "  </tbody>\n",
        "</table>\n",
        "</div>"
       ],
       "metadata": {},
       "output_type": "pyout",
       "prompt_number": 331,
       "text": [
        "     0\n",
        "2    1\n",
        "4    2\n",
        "6    3\n",
        "8    4\n",
        "22  11\n",
        "24  12\n",
        "26  13"
       ]
      }
     ],
     "prompt_number": 331
    },
    {
     "cell_type": "code",
     "collapsed": false,
     "input": [],
     "language": "python",
     "metadata": {},
     "outputs": []
    }
   ],
   "metadata": {}
  }
 ]
}