{
 "metadata": {
  "name": "",
  "signature": "sha256:985adaaa8297322015e542af61e40fecd2ed96700a5802cf374129f36c661918"
 },
 "nbformat": 3,
 "nbformat_minor": 0,
 "worksheets": [
  {
   "cells": [
    {
     "cell_type": "code",
     "collapsed": false,
     "input": [
      "import os\n",
      "import pandas as pd\n",
      "import numpy as np\n",
      "eu = os.path.expanduser\n",
      "jn = os.path.join"
     ],
     "language": "python",
     "metadata": {},
     "outputs": [],
     "prompt_number": 5
    },
    {
     "cell_type": "code",
     "collapsed": false,
     "input": [
      "meta_dir = eu(\"~/vervet_project/metadata\")"
     ],
     "language": "python",
     "metadata": {},
     "outputs": [],
     "prompt_number": 25
    },
    {
     "cell_type": "code",
     "collapsed": false,
     "input": [
      "def load_gene_df():\n",
      "    gene_df = pd.read_csv(jn(meta_dir,\"annot_mike_genes_minimal.tsv\"),sep=\"\\t\",index_col = [0,1])\n",
      "    return gene_df"
     ],
     "language": "python",
     "metadata": {},
     "outputs": [],
     "prompt_number": 33
    },
    {
     "cell_type": "code",
     "collapsed": false,
     "input": [
      "pops = [\"aet\",\"cyn\",\"pyn\",\"pys\",\"sab\",\"tan\"]"
     ],
     "language": "python",
     "metadata": {},
     "outputs": [],
     "prompt_number": 1065
    },
    {
     "cell_type": "code",
     "collapsed": false,
     "input": [
      "gene_df.to_csv(jn(meta_dir,\"annot_mike_genes_minimal.tsv\"),sep=\"\\t\")"
     ],
     "language": "python",
     "metadata": {},
     "outputs": [],
     "prompt_number": 32
    },
    {
     "cell_type": "code",
     "collapsed": false,
     "input": [
      "def index_rolling(s,window,func,overlap=0,*args,**kwargs):\n",
      "    \"\"\"\n",
      "    Apply function in rolling windows, where the window\n",
      "    size is defined with respect to the index values.\n",
      "    This means that different windows can comprise different\n",
      "    numbers of elements.\n",
      "    \n",
      "    s ... pandas Series\n",
      "    window ... window size in units of the index values\n",
      "    func ... function to apply to the series values within each\n",
      "                window\n",
      "    overlap ... oberlap size of windows\n",
      "    args, kwarg ... additional arguments for func\n",
      "    \n",
      "    Example: index_rolling(pd.Series(range(20),index=np.random.rand(20))),0.1,max)\n",
      "    \"\"\"\n",
      "    #note that basis must be sorted in order for this to work properly\n",
      "    windows_min = s.index.min()\n",
      "    windows_max = s.index.max()\n",
      "    window_starts = np.arange(windows_min, windows_max, window-overlap)\n",
      "    window_starts = pd.Series(window_starts, index = window_starts)\n",
      "    def applyToWindow(val):\n",
      "        # using slice_indexer rather that what.loc [val:val+window] allows\n",
      "        # window limits that are not specifically in the index\n",
      "        indexer = s.index.slice_indexer(val,val+window,1)\n",
      "        chunk = s.iloc[indexer]\n",
      "        try:\n",
      "            return func(chunk,*args,**kwargs)\n",
      "        except ValueError, e:\n",
      "            if \"empty sequence\" in str(e):\n",
      "                #print indexer, chunk\n",
      "                return None\n",
      "            else:\n",
      "                raise           \n",
      "    rolled = window_starts.apply(applyToWindow)\n",
      "    return rolled"
     ],
     "language": "python",
     "metadata": {},
     "outputs": []
    },
    {
     "cell_type": "heading",
     "level": 2,
     "metadata": {},
     "source": [
      "ROD (reference ordered data) permutation test"
     ]
    },
    {
     "cell_type": "markdown",
     "metadata": {},
     "source": [
      "tools that, given an analysis result, one can rotate (shift) the results randomly around the genome"
     ]
    },
    {
     "cell_type": "code",
     "collapsed": false,
     "input": [
      "def get_chrom_length():\n",
      "    return pd.read_csv(eu(\"~/vervet_project/metadata/ref3500.tsv\"),\n",
      "                       sep=\"\\t\",index_col=0,squeeze=True,header=False)"
     ],
     "language": "python",
     "metadata": {},
     "outputs": [],
     "prompt_number": 1
    },
    {
     "cell_type": "code",
     "collapsed": false,
     "input": [
      "def shift_rod(rod_df, rnd, mode = \"grid\"):\n",
      "    \"\"\"\n",
      "    shift reference ordered data across the whole genome\n",
      "\n",
      "    Input:\n",
      "    rod_df ... pandas dataframe or series with mulitiindex (chrom, pos)\n",
      "                \n",
      "    modes ...\n",
      "        'grid' ... just rotate the index of the rod data frame\n",
      "                    this means that the positions stay the same only the\n",
      "                    value for each position becomes different\n",
      "                    Faster, but means that you only hit the same grid-point\n",
      "                    this should make it conservative on large grids. Large\n",
      "                    grids are problematic if the fraction of top windows \n",
      "                    considered becomes large \n",
      "        'continuous' ... add the random shift to each index value.\n",
      "                         NOT IMPLEMENTED\n",
      "    \"\"\"\n",
      "    if mode == \"grid\":\n",
      "        new_start_i = int(len(rod_df)*rnd)\n",
      "        rotate_data = np.concatenate((rod_df.iloc[new_start_i:].values,rod_df.iloc[:new_start_i].values))\n",
      "        if  isinstance(rod_df,pd.core.series.Series):\n",
      "            r = pd.Series(rotate_data,index=rod_df.index)\n",
      "            return r\n",
      "        elif isinstance(rod_df,pd.core.frame.DataFrame):\n",
      "            r = pd.DataFrame(rotate_data,index=rod_df.index,columns=rod_df.columns)\n",
      "            return r\n",
      "        \n",
      "    else:\n",
      "        raise UserException(\"Only mode grid supported.\")"
     ],
     "language": "python",
     "metadata": {},
     "outputs": [],
     "prompt_number": 885
    },
    {
     "cell_type": "code",
     "collapsed": false,
     "input": [
      "%timeit shift_rod(tot_sweed_df[\"Likelihood\"], 0.32424) "
     ],
     "language": "python",
     "metadata": {},
     "outputs": [
      {
       "output_type": "stream",
       "stream": "stdout",
       "text": [
        "100 loops, best of 3: 14.5 ms per loop\n"
       ]
      }
     ],
     "prompt_number": 900
    },
    {
     "cell_type": "heading",
     "level": 2,
     "metadata": {},
     "source": [
      "get genes for peaks"
     ]
    },
    {
     "cell_type": "code",
     "collapsed": false,
     "input": [
      "def get_peaks(value_s,min_peak_dist):\n",
      "    \"\"\"\n",
      "    ranked df of peaks\n",
      "    assuming a minimum distance between peaks\n",
      "    \"\"\"\n",
      "    assert len(value_s.index.droplevel(1).unique()) == 1, \"Multiple chromsosomes not supported.\"\n",
      "    chrom = value_s.index[0][0]\n",
      "    x = index_rolling(value_s.ix[chrom],2*min_peak_dist,np.argmax,min_peak_dist)\n",
      "    x.name = \"pos\"\n",
      "    #only take peaks that are found in at least two windows \n",
      "    #this makes sure that we have a minimal distance between peaks\n",
      "    pos_true = (x.reset_index().groupby(\"pos\").apply(len)>1)\n",
      "    pos = pos_true[pos_true.values].index\n",
      "    s = value_s.ix[chrom].ix[pos]\n",
      "    return s"
     ],
     "language": "python",
     "metadata": {},
     "outputs": []
    },
    {
     "cell_type": "code",
     "collapsed": false,
     "input": [
      "def get_genes(peak_s, gene_df, max_dist):\n",
      "    \"\"\"\n",
      "    take the input series and gets \n",
      "    names of genes nearby\n",
      "    \n",
      "    Input:\n",
      "    peak_s ... pandas series with (chrom, pos) index and value of\n",
      "                the statistic ('peak height'). Series should be named.\n",
      "    gene_df ... data frame with gene info \n",
      "    \"\"\"    \n",
      "    all_genes = []\n",
      "    if not gene_df.index.is_monotonic:\n",
      "        gene_df = gene_df.sort_index()\n",
      "    tot_hit_df = pd.DataFrame()\n",
      "    for chrom in peak_s.index.droplevel(1).unique():\n",
      "        loc_gene_df = gene_df.ix[chrom]\n",
      "        #loc_gene_df = loc_gene_df.append(pd.DataFrame(np.nan,index=[np.inf],columns=loc_gene_df.columns))\n",
      "        pos_rel_to_start = np.searchsorted(loc_gene_df.index.values-max_dist,peak_s.ix[chrom].index.values)\n",
      "        pos_rel_to_end = np.searchsorted(loc_gene_df[\"end\"].values+max_dist,peak_s.ix[chrom].index.values)\n",
      "        genes = list(set(loc_gene_df[\"gene_id\"].iloc[np.hstack([range(a,b) for a,b in zip(pos_rel_to_end,pos_rel_to_start)])]))\n",
      "        #print chrom, genes\n",
      "        all_genes += genes\n",
      "    return all_genes\n",
      "\n",
      "def get_gene_info(gene_ls,gene_df):\n",
      "    \"\"\"\n",
      "    for a list of gene ids,\n",
      "    get a data frame with their \n",
      "    position\n",
      "    \"\"\"\n",
      "    gi = gene_df[gene_df[\"gene_id\"].apply(lambda x: x in gene_ls)]\n",
      "    return gi\n",
      "\n",
      "def get_peaks(gene_info,top_s,max_dist):\n",
      "    \"\"\"\n",
      "    For each gene in gene_info get the\n",
      "    peaks within max_dist in top_s. This \n",
      "    is basically reverse engineering to get\n",
      "    the peak info for each gene that was found \n",
      "    to be associated with a peak. \n",
      "    \n",
      "    Input:\n",
      "    gene_info ... data frame with index ('chrom','start')\n",
      "                and columns 'gene_id' and 'end'\n",
      "    top_s ... series of peak positions with index (chrom, pos)\n",
      "                and values peak height\n",
      "    max_dist ... maximum distance between gene and peak\n",
      "    \"\"\"\n",
      "    def get_dist(df,gene_pos):\n",
      "        s = pd.Series(df.index.droplevel(0).values - gene_pos.ix[df.index[0][0]],\n",
      "                                                  index=df.index.droplevel(0).values)\n",
      "        return s\n",
      "    tot_gene_peaks_df = pd.DataFrame()\n",
      "    if not top_s.index.is_monotonic:\n",
      "        top_s = top_s.sort_index()\n",
      "    for chrom in gene_info.index.droplevel(1).unique():\n",
      "        loc_top_s = top_s.ix[chrom]\n",
      "        start = np.searchsorted(loc_top_s.index.values+max_dist,gene_info.ix[chrom].index.values)\n",
      "        end = np.searchsorted(loc_top_s.index.values-max_dist,gene_info.ix[chrom][\"end\"].values)\n",
      "        \n",
      "        x = pd.concat([loc_top_s.iloc[st:ed] for st,ed in zip(start,end)],\n",
      "                      keys=gene_info.ix[chrom][\"gene_id\"].values)\n",
      "        x.name = \"peak_height\"\n",
      "        \n",
      "\n",
      "\n",
      "        dist_start = x.groupby(lambda i: i[0]).\\\n",
      "                    apply(lambda df: get_dist(df,\n",
      "                                              gene_info.ix[chrom].reset_index().set_index(\"gene_id\")[\"pos\"]))\n",
      "        dist_start.name = \"dist_start\"\n",
      "        dist_end = x.groupby(lambda i: i[0]).\\\n",
      "                    apply(lambda df: get_dist(df,\n",
      "                                              gene_info.ix[chrom].set_index(\"gene_id\")[\"end\"]))\n",
      "        dist_end.name = \"dist_end\"\n",
      "        gene_peaks_df = pd.concat([x,dist_start,dist_end],axis=1)\n",
      "        gene_peaks_df.index = pd.MultiIndex.from_arrays([gene_peaks_df.index.droplevel(1),\n",
      "                                         [chrom]*len(x),\n",
      "                                         gene_peaks_df.index.droplevel(0)])\n",
      "        tot_gene_peaks_df = pd.concat([tot_gene_peaks_df, gene_peaks_df])\n",
      "        tot_gene_peaks_df.index.names = [\"gene_id\",\"chrom\",\"peak_pos\"]\n",
      "        if chrom == \"CAE19\":\n",
      "            print gene_info.ix[chrom].reset_index().set_index(\"gene_id\")[\"pos\"]\n",
      "            print loc_top_s\n",
      "            print start\n",
      "            print end\n",
      "            print x\n",
      "            print dist_start\n",
      "    return tot_gene_peaks_df"
     ],
     "language": "python",
     "metadata": {},
     "outputs": [],
     "prompt_number": 1163
    },
    {
     "cell_type": "code",
     "collapsed": false,
     "input": [
      "chromosomes = [\"CAE\" + str(i) for i in range(1,30)+[\"X\",\"Y\"]]\n",
      "def get_sweed_df(chrom,pop):\n",
      "    sweed_res = pd.read_csv(eu(\"~/vervet_project/analyses/20140611_163_subpop_sweepfinder/\"\n",
      "                             \"_data/SweeD_Report.163_ref3500_{}_{}\".format(chrom,pop)),\n",
      "                                                          sep=\"\\t\",skiprows=2,index_col=0)\n",
      "    sweed_res.index = pd.MultiIndex.from_tuples(zip([chrom]*len(sweed_res.index),sweed_res.index))\n",
      "    return sweed_res"
     ],
     "language": "python",
     "metadata": {},
     "outputs": [],
     "prompt_number": 642
    },
    {
     "cell_type": "code",
     "collapsed": false,
     "input": [],
     "language": "python",
     "metadata": {},
     "outputs": [
      {
       "metadata": {},
       "output_type": "pyout",
       "prompt_number": 1158,
       "text": [
        "'0.14.0'"
       ]
      }
     ],
     "prompt_number": 1158
    },
    {
     "cell_type": "code",
     "collapsed": false,
     "input": [
      "pop = \"aet\"\n",
      "eu(\"~/vervet_project/analyses/20140611_163_subpop_sweepfinder/\"\n",
      "                             \"_data/SweeD_Report.163_ref3500_{}_{}\".format(chrom,pop))"
     ],
     "language": "python",
     "metadata": {},
     "outputs": [
      {
       "metadata": {},
       "output_type": "pyout",
       "prompt_number": 1057,
       "text": [
        "'/home/GMI/hannes.svardal/vervet_project/analyses/20140611_163_subpop_sweepfinder/_data/SweeD_Report.163_ref3500_CAEY_aet'"
       ]
      }
     ],
     "prompt_number": 1057
    },
    {
     "cell_type": "code",
     "collapsed": false,
     "input": [
      "gene_df = load_gene_df()\n",
      "gene_df = gene_df.sortlevel(level=[0,1])"
     ],
     "language": "python",
     "metadata": {},
     "outputs": [],
     "prompt_number": 372
    },
    {
     "cell_type": "code",
     "collapsed": false,
     "input": [
      "pop = \"aet\"\n",
      "tot_sweed_df = pd.read_csv(eu(\"~/vervet_project/analyses/20140611_163_subpop_sweepfinder/\"\n",
      "                                 \"_data/tot_sweed_result_{}.tsv\".format(pop)),sep=\"\\t\",index_col=[0,1])"
     ],
     "language": "python",
     "metadata": {},
     "outputs": [],
     "prompt_number": 1150
    },
    {
     "cell_type": "code",
     "collapsed": false,
     "input": [
      "top_s = tot_sweed_df.sort(\"Likelihood\",ascending=False)[\"Likelihood\"].iloc[:2000]\n",
      "#top_s = pd.concat([pd.Series([17],index=pd.MultiIndex.from_arrays([[\"CAE1\"],[4922]])),top_s])"
     ],
     "language": "python",
     "metadata": {},
     "outputs": [],
     "prompt_number": 1152
    },
    {
     "cell_type": "code",
     "collapsed": false,
     "input": [
      "max_dist = 10000\n",
      "cand_genes = get_genes(top_s, gene_df, max_dist=max_dist)\n",
      "gene_info = get_gene_info(cand_genes, gene_df)\n",
      "peaks = get_peaks(gene_info,top_s,max_dist=max_dist)"
     ],
     "language": "python",
     "metadata": {},
     "outputs": [
      {
       "output_type": "stream",
       "stream": "stdout",
       "text": [
        "gene_id\n",
        "SAMM50     26467319\n",
        "Name: pos, dtype: int64\n",
        "pos\n",
        "9483311.2373     21.93971\n",
        "9483796.1723     23.32173\n",
        "26477373.2917    20.04123\n",
        "26478828.0967    19.82511\n",
        "Name: Likelihood, dtype: float64\n",
        "[2]\n",
        "[4]\n",
        "        pos          \n",
        "SAMM50  26477373.2917    20.04123\n",
        "        26478828.0967    19.82511\n",
        "Name: peak_height, dtype: float64\n",
        "SAMM50  26477373.2917    10054.2917\n",
        "        26478828.0967    11509.0967\n",
        "Name: dist_start, dtype: float64\n"
       ]
      }
     ],
     "prompt_number": 1225
    },
    {
     "cell_type": "code",
     "collapsed": false,
     "input": [
      "def genes_per_peak_s(df):\n",
      "    s = df.iloc[0]#[\"peak_height\"]\n",
      "    s[\"genes\"] = 3#list(df[\"gene_id\"].unique())\n",
      "    return s\n",
      "p = peaks.reset_index([0])"
     ],
     "language": "python",
     "metadata": {},
     "outputs": [],
     "prompt_number": 1248
    },
    {
     "cell_type": "code",
     "collapsed": false,
     "input": [
      "gene_list_peak_pos = p[\"gene_id\"].groupby(lambda x: x).apply(list)\n",
      "gene_list_peak_pos.name = \"genes\"\n",
      "gene_list_peak_pos.index = pd.MultiIndex.from_tuples(gene_list_peak_pos.index)\n",
      "peak_info = pd.concat([top_s,gene_list_peak_pos],axis=1)\n",
      "peak_info.sort(\"Likelihood\",ascending=False,inplace=True)"
     ],
     "language": "python",
     "metadata": {},
     "outputs": [],
     "prompt_number": 1263
    },
    {
     "cell_type": "code",
     "collapsed": false,
     "input": [],
     "language": "python",
     "metadata": {},
     "outputs": [],
     "prompt_number": 1270
    },
    {
     "cell_type": "code",
     "collapsed": false,
     "input": [],
     "language": "python",
     "metadata": {},
     "outputs": [],
     "prompt_number": 1272
    },
    {
     "cell_type": "heading",
     "level": 2,
     "metadata": {},
     "source": [
      "get go associations for gene list"
     ]
    },
    {
     "cell_type": "code",
     "collapsed": false,
     "input": [
      "gene_to_go = pd.read_csv(jn(meta_dir,\"gene_to_go.csv\"),index_col=0)\n",
      "gene_to_name = pd.read_csv(jn(meta_dir,\"gene_to_name.csv\"),index_col=0)\n",
      "go_to_name = pd.read_csv(jn(meta_dir,\"go_to_name.csv\"),index_col=0)"
     ],
     "language": "python",
     "metadata": {},
     "outputs": [],
     "prompt_number": 824
    },
    {
     "cell_type": "code",
     "collapsed": false,
     "input": [
      "rank_table.join(go_to_name.set_index(\"go_identifier\",drop=True))"
     ],
     "language": "python",
     "metadata": {},
     "outputs": [
      {
       "html": [
        "<div style=\"max-height:1000px;max-width:1500px;overflow:auto;\">\n",
        "<table border=\"1\" class=\"dataframe\">\n",
        "  <thead>\n",
        "    <tr style=\"text-align: right;\">\n",
        "      <th></th>\n",
        "      <th>n_genes</th>\n",
        "      <th>out_of</th>\n",
        "      <th>rank</th>\n",
        "      <th>go_name</th>\n",
        "    </tr>\n",
        "    <tr>\n",
        "      <th>go_identifier</th>\n",
        "      <th></th>\n",
        "      <th></th>\n",
        "      <th></th>\n",
        "      <th></th>\n",
        "    </tr>\n",
        "  </thead>\n",
        "  <tbody>\n",
        "    <tr>\n",
        "      <th>GO:0000014</th>\n",
        "      <td>  1</td>\n",
        "      <td> 0</td>\n",
        "      <td> 0</td>\n",
        "      <td> single-stranded DNA endodeoxyribonuclease acti...</td>\n",
        "    </tr>\n",
        "    <tr>\n",
        "      <th>GO:0000060</th>\n",
        "      <td>  1</td>\n",
        "      <td> 0</td>\n",
        "      <td> 0</td>\n",
        "      <td>        protein import into nucleus, translocation</td>\n",
        "    </tr>\n",
        "    <tr>\n",
        "      <th>GO:0000075</th>\n",
        "      <td>  2</td>\n",
        "      <td> 0</td>\n",
        "      <td> 0</td>\n",
        "      <td>                             cell cycle checkpoint</td>\n",
        "    </tr>\n",
        "    <tr>\n",
        "      <th>GO:0000082</th>\n",
        "      <td>  2</td>\n",
        "      <td> 0</td>\n",
        "      <td> 0</td>\n",
        "      <td>             G1/S transition of mitotic cell cycle</td>\n",
        "    </tr>\n",
        "    <tr>\n",
        "      <th>GO:0000087</th>\n",
        "      <td>  2</td>\n",
        "      <td> 0</td>\n",
        "      <td> 0</td>\n",
        "      <td>                                   mitotic M phase</td>\n",
        "    </tr>\n",
        "    <tr>\n",
        "      <th>GO:0000088</th>\n",
        "      <td>  1</td>\n",
        "      <td> 0</td>\n",
        "      <td> 0</td>\n",
        "      <td>                                  mitotic prophase</td>\n",
        "    </tr>\n",
        "    <tr>\n",
        "      <th>GO:0000090</th>\n",
        "      <td>  1</td>\n",
        "      <td> 0</td>\n",
        "      <td> 0</td>\n",
        "      <td>                                  mitotic anaphase</td>\n",
        "    </tr>\n",
        "    <tr>\n",
        "      <th>GO:0000096</th>\n",
        "      <td>  1</td>\n",
        "      <td> 0</td>\n",
        "      <td> 0</td>\n",
        "      <td>               sulfur amino acid metabolic process</td>\n",
        "    </tr>\n",
        "    <tr>\n",
        "      <th>GO:0000098</th>\n",
        "      <td>  1</td>\n",
        "      <td> 0</td>\n",
        "      <td> 0</td>\n",
        "      <td>               sulfur amino acid catabolic process</td>\n",
        "    </tr>\n",
        "    <tr>\n",
        "      <th>GO:0000122</th>\n",
        "      <td>  2</td>\n",
        "      <td> 0</td>\n",
        "      <td> 0</td>\n",
        "      <td> negative regulation of transcription from RNA ...</td>\n",
        "    </tr>\n",
        "    <tr>\n",
        "      <th>GO:0000151</th>\n",
        "      <td>  1</td>\n",
        "      <td> 0</td>\n",
        "      <td> 0</td>\n",
        "      <td>                          ubiquitin ligase complex</td>\n",
        "    </tr>\n",
        "    <tr>\n",
        "      <th>GO:0000166</th>\n",
        "      <td> 14</td>\n",
        "      <td> 0</td>\n",
        "      <td> 0</td>\n",
        "      <td>                                nucleotide binding</td>\n",
        "    </tr>\n",
        "    <tr>\n",
        "      <th>GO:0000185</th>\n",
        "      <td>  1</td>\n",
        "      <td> 0</td>\n",
        "      <td> 0</td>\n",
        "      <td>                     activation of MAPKKK activity</td>\n",
        "    </tr>\n",
        "    <tr>\n",
        "      <th>GO:0000255</th>\n",
        "      <td>  1</td>\n",
        "      <td> 0</td>\n",
        "      <td> 0</td>\n",
        "      <td>                       allantoin metabolic process</td>\n",
        "    </tr>\n",
        "    <tr>\n",
        "      <th>GO:0000278</th>\n",
        "      <td>  2</td>\n",
        "      <td> 0</td>\n",
        "      <td> 0</td>\n",
        "      <td>                                mitotic cell cycle</td>\n",
        "    </tr>\n",
        "    <tr>\n",
        "      <th>GO:0000390</th>\n",
        "      <td>  1</td>\n",
        "      <td> 0</td>\n",
        "      <td> 0</td>\n",
        "      <td>                  spliceosomal complex disassembly</td>\n",
        "    </tr>\n",
        "    <tr>\n",
        "      <th>GO:0000398</th>\n",
        "      <td>  2</td>\n",
        "      <td> 0</td>\n",
        "      <td> 0</td>\n",
        "      <td>                    mRNA splicing, via spliceosome</td>\n",
        "    </tr>\n",
        "    <tr>\n",
        "      <th>GO:0000422</th>\n",
        "      <td>  1</td>\n",
        "      <td> 0</td>\n",
        "      <td> 0</td>\n",
        "      <td>                         mitochondrion degradation</td>\n",
        "    </tr>\n",
        "    <tr>\n",
        "      <th>GO:0000724</th>\n",
        "      <td>  1</td>\n",
        "      <td> 0</td>\n",
        "      <td> 0</td>\n",
        "      <td> double-strand break repair via homologous reco...</td>\n",
        "    </tr>\n",
        "    <tr>\n",
        "      <th>GO:0000737</th>\n",
        "      <td>  1</td>\n",
        "      <td> 0</td>\n",
        "      <td> 0</td>\n",
        "      <td>            DNA catabolic process, endonucleolytic</td>\n",
        "    </tr>\n",
        "    <tr>\n",
        "      <th>GO:0000910</th>\n",
        "      <td>  2</td>\n",
        "      <td> 0</td>\n",
        "      <td> 0</td>\n",
        "      <td>                                       cytokinesis</td>\n",
        "    </tr>\n",
        "    <tr>\n",
        "      <th>GO:0000922</th>\n",
        "      <td>  1</td>\n",
        "      <td> 0</td>\n",
        "      <td> 0</td>\n",
        "      <td>                                      spindle pole</td>\n",
        "    </tr>\n",
        "    <tr>\n",
        "      <th>GO:0000979</th>\n",
        "      <td>  1</td>\n",
        "      <td> 0</td>\n",
        "      <td> 0</td>\n",
        "      <td> RNA polymerase II core promoter sequence-speci...</td>\n",
        "    </tr>\n",
        "    <tr>\n",
        "      <th>GO:0001103</th>\n",
        "      <td>  1</td>\n",
        "      <td> 0</td>\n",
        "      <td> 0</td>\n",
        "      <td> RNA polymerase II repressing transcription fac...</td>\n",
        "    </tr>\n",
        "    <tr>\n",
        "      <th>GO:0001104</th>\n",
        "      <td>  1</td>\n",
        "      <td> 0</td>\n",
        "      <td> 0</td>\n",
        "      <td> RNA polymerase II transcription cofactor activity</td>\n",
        "    </tr>\n",
        "    <tr>\n",
        "      <th>GO:0001106</th>\n",
        "      <td>  1</td>\n",
        "      <td> 0</td>\n",
        "      <td> 0</td>\n",
        "      <td> RNA polymerase II transcription corepressor ac...</td>\n",
        "    </tr>\n",
        "    <tr>\n",
        "      <th>GO:0001523</th>\n",
        "      <td>  1</td>\n",
        "      <td> 0</td>\n",
        "      <td> 0</td>\n",
        "      <td>                        retinoid metabolic process</td>\n",
        "    </tr>\n",
        "    <tr>\n",
        "      <th>GO:0001525</th>\n",
        "      <td>  2</td>\n",
        "      <td> 0</td>\n",
        "      <td> 0</td>\n",
        "      <td>                                      angiogenesis</td>\n",
        "    </tr>\n",
        "    <tr>\n",
        "      <th>GO:0001541</th>\n",
        "      <td>  2</td>\n",
        "      <td> 0</td>\n",
        "      <td> 0</td>\n",
        "      <td>                      ovarian follicle development</td>\n",
        "    </tr>\n",
        "    <tr>\n",
        "      <th>GO:0001553</th>\n",
        "      <td>  1</td>\n",
        "      <td> 0</td>\n",
        "      <td> 0</td>\n",
        "      <td>                                     luteinization</td>\n",
        "    </tr>\n",
        "    <tr>\n",
        "      <th>...</th>\n",
        "      <td>...</td>\n",
        "      <td>...</td>\n",
        "      <td>...</td>\n",
        "      <td>...</td>\n",
        "    </tr>\n",
        "    <tr>\n",
        "      <th>GO:0097105</th>\n",
        "      <td>  1</td>\n",
        "      <td> 0</td>\n",
        "      <td> 0</td>\n",
        "      <td>                     presynaptic membrane assembly</td>\n",
        "    </tr>\n",
        "    <tr>\n",
        "      <th>GO:0097109</th>\n",
        "      <td>  1</td>\n",
        "      <td> 0</td>\n",
        "      <td> 0</td>\n",
        "      <td>                 neuroligin family protein binding</td>\n",
        "    </tr>\n",
        "    <tr>\n",
        "      <th>GO:0097112</th>\n",
        "      <td>  1</td>\n",
        "      <td> 0</td>\n",
        "      <td> 0</td>\n",
        "      <td>       gamma-aminobutyric acid receptor clustering</td>\n",
        "    </tr>\n",
        "    <tr>\n",
        "      <th>GO:0097114</th>\n",
        "      <td>  1</td>\n",
        "      <td> 0</td>\n",
        "      <td> 0</td>\n",
        "      <td>          N-methyl-D-aspartate receptor clustering</td>\n",
        "    </tr>\n",
        "    <tr>\n",
        "      <th>GO:0097116</th>\n",
        "      <td>  1</td>\n",
        "      <td> 0</td>\n",
        "      <td> 0</td>\n",
        "      <td>                               gephyrin clustering</td>\n",
        "    </tr>\n",
        "    <tr>\n",
        "      <th>GO:0097117</th>\n",
        "      <td>  1</td>\n",
        "      <td> 0</td>\n",
        "      <td> 0</td>\n",
        "      <td>    guanylate kinase-associated protein clustering</td>\n",
        "    </tr>\n",
        "    <tr>\n",
        "      <th>GO:0097118</th>\n",
        "      <td>  1</td>\n",
        "      <td> 0</td>\n",
        "      <td> 0</td>\n",
        "      <td>                             neuroligin clustering</td>\n",
        "    </tr>\n",
        "    <tr>\n",
        "      <th>GO:0097119</th>\n",
        "      <td>  1</td>\n",
        "      <td> 0</td>\n",
        "      <td> 0</td>\n",
        "      <td>        postsynaptic density protein 95 clustering</td>\n",
        "    </tr>\n",
        "    <tr>\n",
        "      <th>GO:0097120</th>\n",
        "      <td>  1</td>\n",
        "      <td> 0</td>\n",
        "      <td> 0</td>\n",
        "      <td>                  receptor localization to synapse</td>\n",
        "    </tr>\n",
        "    <tr>\n",
        "      <th>GO:0097286</th>\n",
        "      <td>  1</td>\n",
        "      <td> 0</td>\n",
        "      <td> 0</td>\n",
        "      <td>                                   iron ion import</td>\n",
        "    </tr>\n",
        "    <tr>\n",
        "      <th>GO:0097449</th>\n",
        "      <td>  1</td>\n",
        "      <td> 0</td>\n",
        "      <td> 0</td>\n",
        "      <td>                              astrocyte projection</td>\n",
        "    </tr>\n",
        "    <tr>\n",
        "      <th>GO:0098655</th>\n",
        "      <td>  1</td>\n",
        "      <td> 0</td>\n",
        "      <td> 0</td>\n",
        "      <td>                    cation transmembrane transport</td>\n",
        "    </tr>\n",
        "    <tr>\n",
        "      <th>GO:1900025</th>\n",
        "      <td>  1</td>\n",
        "      <td> 0</td>\n",
        "      <td> 0</td>\n",
        "      <td> negative regulation of substrate adhesion-depe...</td>\n",
        "    </tr>\n",
        "    <tr>\n",
        "      <th>GO:1902186</th>\n",
        "      <td>  1</td>\n",
        "      <td> 0</td>\n",
        "      <td> 0</td>\n",
        "      <td>        regulation of viral release from host cell</td>\n",
        "    </tr>\n",
        "    <tr>\n",
        "      <th>GO:1902305</th>\n",
        "      <td>  1</td>\n",
        "      <td> 0</td>\n",
        "      <td> 0</td>\n",
        "      <td>  regulation of sodium ion transmembrane transport</td>\n",
        "    </tr>\n",
        "    <tr>\n",
        "      <th>GO:1902560</th>\n",
        "      <td>  1</td>\n",
        "      <td> 0</td>\n",
        "      <td> 0</td>\n",
        "      <td>                             GMP reductase complex</td>\n",
        "    </tr>\n",
        "    <tr>\n",
        "      <th>GO:1902749</th>\n",
        "      <td>  1</td>\n",
        "      <td> 0</td>\n",
        "      <td> 0</td>\n",
        "      <td>    regulation of cell cycle G2/M phase transition</td>\n",
        "    </tr>\n",
        "    <tr>\n",
        "      <th>GO:1902911</th>\n",
        "      <td>  1</td>\n",
        "      <td> 0</td>\n",
        "      <td> 0</td>\n",
        "      <td>                            protein kinase complex</td>\n",
        "    </tr>\n",
        "    <tr>\n",
        "      <th>GO:1990090</th>\n",
        "      <td>  1</td>\n",
        "      <td> 0</td>\n",
        "      <td> 0</td>\n",
        "      <td> cellular response to nerve growth factor stimulus</td>\n",
        "    </tr>\n",
        "    <tr>\n",
        "      <th>GO:2000114</th>\n",
        "      <td>  1</td>\n",
        "      <td> 0</td>\n",
        "      <td> 0</td>\n",
        "      <td>      regulation of establishment of cell polarity</td>\n",
        "    </tr>\n",
        "    <tr>\n",
        "      <th>GO:2000134</th>\n",
        "      <td>  1</td>\n",
        "      <td> 0</td>\n",
        "      <td> 0</td>\n",
        "      <td> negative regulation of G1/S transition of mito...</td>\n",
        "    </tr>\n",
        "    <tr>\n",
        "      <th>GO:2000145</th>\n",
        "      <td>  1</td>\n",
        "      <td> 0</td>\n",
        "      <td> 0</td>\n",
        "      <td>                       regulation of cell motility</td>\n",
        "    </tr>\n",
        "    <tr>\n",
        "      <th>GO:2000304</th>\n",
        "      <td>  1</td>\n",
        "      <td> 0</td>\n",
        "      <td> 0</td>\n",
        "      <td> positive regulation of ceramide biosynthetic p...</td>\n",
        "    </tr>\n",
        "    <tr>\n",
        "      <th>GO:2000310</th>\n",
        "      <td>  1</td>\n",
        "      <td> 0</td>\n",
        "      <td> 0</td>\n",
        "      <td> regulation of N-methyl-D-aspartate selective g...</td>\n",
        "    </tr>\n",
        "    <tr>\n",
        "      <th>GO:2000311</th>\n",
        "      <td>  1</td>\n",
        "      <td> 0</td>\n",
        "      <td> 0</td>\n",
        "      <td> regulation of alpha-amino-3-hydroxy-5-methyl-4...</td>\n",
        "    </tr>\n",
        "    <tr>\n",
        "      <th>GO:2000463</th>\n",
        "      <td>  1</td>\n",
        "      <td> 0</td>\n",
        "      <td> 0</td>\n",
        "      <td> positive regulation of excitatory postsynaptic...</td>\n",
        "    </tr>\n",
        "    <tr>\n",
        "      <th>GO:2000599</th>\n",
        "      <td>  1</td>\n",
        "      <td> 0</td>\n",
        "      <td> 0</td>\n",
        "      <td>   negative regulation of cyclin catabolic process</td>\n",
        "    </tr>\n",
        "    <tr>\n",
        "      <th>GO:2001014</th>\n",
        "      <td>  1</td>\n",
        "      <td> 0</td>\n",
        "      <td> 0</td>\n",
        "      <td> regulation of skeletal muscle cell differentia...</td>\n",
        "    </tr>\n",
        "    <tr>\n",
        "      <th>GO:2001234</th>\n",
        "      <td>  1</td>\n",
        "      <td> 0</td>\n",
        "      <td> 0</td>\n",
        "      <td> negative regulation of apoptotic signaling pat...</td>\n",
        "    </tr>\n",
        "    <tr>\n",
        "      <th>GO:2001237</th>\n",
        "      <td>  1</td>\n",
        "      <td> 0</td>\n",
        "      <td> 0</td>\n",
        "      <td> negative regulation of extrinsic apoptotic sig...</td>\n",
        "    </tr>\n",
        "  </tbody>\n",
        "</table>\n",
        "<p>888 rows \u00d7 4 columns</p>\n",
        "</div>"
       ],
       "metadata": {},
       "output_type": "pyout",
       "prompt_number": 1077,
       "text": [
        "               n_genes  out_of  rank  \\\n",
        "go_identifier                          \n",
        "GO:0000014           1       0     0   \n",
        "GO:0000060           1       0     0   \n",
        "GO:0000075           2       0     0   \n",
        "GO:0000082           2       0     0   \n",
        "GO:0000087           2       0     0   \n",
        "GO:0000088           1       0     0   \n",
        "GO:0000090           1       0     0   \n",
        "GO:0000096           1       0     0   \n",
        "GO:0000098           1       0     0   \n",
        "GO:0000122           2       0     0   \n",
        "GO:0000151           1       0     0   \n",
        "GO:0000166          14       0     0   \n",
        "GO:0000185           1       0     0   \n",
        "GO:0000255           1       0     0   \n",
        "GO:0000278           2       0     0   \n",
        "GO:0000390           1       0     0   \n",
        "GO:0000398           2       0     0   \n",
        "GO:0000422           1       0     0   \n",
        "GO:0000724           1       0     0   \n",
        "GO:0000737           1       0     0   \n",
        "GO:0000910           2       0     0   \n",
        "GO:0000922           1       0     0   \n",
        "GO:0000979           1       0     0   \n",
        "GO:0001103           1       0     0   \n",
        "GO:0001104           1       0     0   \n",
        "GO:0001106           1       0     0   \n",
        "GO:0001523           1       0     0   \n",
        "GO:0001525           2       0     0   \n",
        "GO:0001541           2       0     0   \n",
        "GO:0001553           1       0     0   \n",
        "...                ...     ...   ...   \n",
        "GO:0097105           1       0     0   \n",
        "GO:0097109           1       0     0   \n",
        "GO:0097112           1       0     0   \n",
        "GO:0097114           1       0     0   \n",
        "GO:0097116           1       0     0   \n",
        "GO:0097117           1       0     0   \n",
        "GO:0097118           1       0     0   \n",
        "GO:0097119           1       0     0   \n",
        "GO:0097120           1       0     0   \n",
        "GO:0097286           1       0     0   \n",
        "GO:0097449           1       0     0   \n",
        "GO:0098655           1       0     0   \n",
        "GO:1900025           1       0     0   \n",
        "GO:1902186           1       0     0   \n",
        "GO:1902305           1       0     0   \n",
        "GO:1902560           1       0     0   \n",
        "GO:1902749           1       0     0   \n",
        "GO:1902911           1       0     0   \n",
        "GO:1990090           1       0     0   \n",
        "GO:2000114           1       0     0   \n",
        "GO:2000134           1       0     0   \n",
        "GO:2000145           1       0     0   \n",
        "GO:2000304           1       0     0   \n",
        "GO:2000310           1       0     0   \n",
        "GO:2000311           1       0     0   \n",
        "GO:2000463           1       0     0   \n",
        "GO:2000599           1       0     0   \n",
        "GO:2001014           1       0     0   \n",
        "GO:2001234           1       0     0   \n",
        "GO:2001237           1       0     0   \n",
        "\n",
        "                                                         go_name  \n",
        "go_identifier                                                     \n",
        "GO:0000014     single-stranded DNA endodeoxyribonuclease acti...  \n",
        "GO:0000060            protein import into nucleus, translocation  \n",
        "GO:0000075                                 cell cycle checkpoint  \n",
        "GO:0000082                 G1/S transition of mitotic cell cycle  \n",
        "GO:0000087                                       mitotic M phase  \n",
        "GO:0000088                                      mitotic prophase  \n",
        "GO:0000090                                      mitotic anaphase  \n",
        "GO:0000096                   sulfur amino acid metabolic process  \n",
        "GO:0000098                   sulfur amino acid catabolic process  \n",
        "GO:0000122     negative regulation of transcription from RNA ...  \n",
        "GO:0000151                              ubiquitin ligase complex  \n",
        "GO:0000166                                    nucleotide binding  \n",
        "GO:0000185                         activation of MAPKKK activity  \n",
        "GO:0000255                           allantoin metabolic process  \n",
        "GO:0000278                                    mitotic cell cycle  \n",
        "GO:0000390                      spliceosomal complex disassembly  \n",
        "GO:0000398                        mRNA splicing, via spliceosome  \n",
        "GO:0000422                             mitochondrion degradation  \n",
        "GO:0000724     double-strand break repair via homologous reco...  \n",
        "GO:0000737                DNA catabolic process, endonucleolytic  \n",
        "GO:0000910                                           cytokinesis  \n",
        "GO:0000922                                          spindle pole  \n",
        "GO:0000979     RNA polymerase II core promoter sequence-speci...  \n",
        "GO:0001103     RNA polymerase II repressing transcription fac...  \n",
        "GO:0001104     RNA polymerase II transcription cofactor activity  \n",
        "GO:0001106     RNA polymerase II transcription corepressor ac...  \n",
        "GO:0001523                            retinoid metabolic process  \n",
        "GO:0001525                                          angiogenesis  \n",
        "GO:0001541                          ovarian follicle development  \n",
        "GO:0001553                                         luteinization  \n",
        "...                                                          ...  \n",
        "GO:0097105                         presynaptic membrane assembly  \n",
        "GO:0097109                     neuroligin family protein binding  \n",
        "GO:0097112           gamma-aminobutyric acid receptor clustering  \n",
        "GO:0097114              N-methyl-D-aspartate receptor clustering  \n",
        "GO:0097116                                   gephyrin clustering  \n",
        "GO:0097117        guanylate kinase-associated protein clustering  \n",
        "GO:0097118                                 neuroligin clustering  \n",
        "GO:0097119            postsynaptic density protein 95 clustering  \n",
        "GO:0097120                      receptor localization to synapse  \n",
        "GO:0097286                                       iron ion import  \n",
        "GO:0097449                                  astrocyte projection  \n",
        "GO:0098655                        cation transmembrane transport  \n",
        "GO:1900025     negative regulation of substrate adhesion-depe...  \n",
        "GO:1902186            regulation of viral release from host cell  \n",
        "GO:1902305      regulation of sodium ion transmembrane transport  \n",
        "GO:1902560                                 GMP reductase complex  \n",
        "GO:1902749        regulation of cell cycle G2/M phase transition  \n",
        "GO:1902911                                protein kinase complex  \n",
        "GO:1990090     cellular response to nerve growth factor stimulus  \n",
        "GO:2000114          regulation of establishment of cell polarity  \n",
        "GO:2000134     negative regulation of G1/S transition of mito...  \n",
        "GO:2000145                           regulation of cell motility  \n",
        "GO:2000304     positive regulation of ceramide biosynthetic p...  \n",
        "GO:2000310     regulation of N-methyl-D-aspartate selective g...  \n",
        "GO:2000311     regulation of alpha-amino-3-hydroxy-5-methyl-4...  \n",
        "GO:2000463     positive regulation of excitatory postsynaptic...  \n",
        "GO:2000599       negative regulation of cyclin catabolic process  \n",
        "GO:2001014     regulation of skeletal muscle cell differentia...  \n",
        "GO:2001234     negative regulation of apoptotic signaling pat...  \n",
        "GO:2001237     negative regulation of extrinsic apoptotic sig...  \n",
        "\n",
        "[888 rows x 4 columns]"
       ]
      }
     ],
     "prompt_number": 1077
    },
    {
     "cell_type": "code",
     "collapsed": false,
     "input": [
      "def permut_assoc(rod_s, rnd, gene_df, gene_to_go, top_n, max_dist):\n",
      "    \"\"\"\n",
      "    This is the main function.\n",
      "    Use with rnd = 0 to get the real assoc.\n",
      "    \"\"\"\n",
      "    s = shift_rod(rod_s, rnd)\n",
      "    s.sort(ascending=False, inplace=True)\n",
      "    top_s = s.iloc[:top_n]\n",
      "    cand_genes = get_genes(top_s, gene_df, max_dist=max_dist)\n",
      "    assoc = get_go_assoc(cand_genes, gene_to_go)\n",
      "    return assoc\n",
      "\n",
      "def get_go_assoc(gene_ls, gene_to_go):\n",
      "    \"\"\"\n",
      "    Get series with number of genes associated with each\n",
      "    category in gene_to_go\n",
      "    \"\"\"\n",
      "    s = gene_to_go.set_index(\"gene_symbol\").ix[gene_ls].groupby(\"go_identifier\").apply(len)\n",
      "    return s\n",
      "\n",
      "def multiple_permut_assoc(rod_s, gene_df, gene_to_go, top_n, max_dist, n_runs, rnd_seed=None):\n",
      "    if rnd_seed is not None:\n",
      "        np.random.seed(rnd_seed)\n",
      "    assoc_table = pd.concat([permut_assoc(rod_s, rnd, gene_df, gene_to_go, top_n, max_dist) for rnd in np.random.rand(n_runs)],axis=1)\n",
      "    assoc_table = assoc_table.fillna(0)\n",
      "    return assoc_table\n",
      "\n",
      "def save_permut_assoc_table(assoc_table,fn):\n",
      "    assoc_table.to_csv(fn)\n",
      "\n",
      "def get_initial_rank_table(real_assoc):\n",
      "    return pd.DataFrame({\"n_genes\":real_assoc.values,\"rank\":0,\"out_of\":0},index=real_assoc.index)\n",
      "\n",
      "def get_p_val(rank_table):\n",
      "    \"\"\"\n",
      "    Input:\n",
      "    \n",
      "    \"\"\"\n",
      "    r =  1-rank_table[\"rank\"]*1./(rank_table[\"out_of\"]+1)\n",
      "    r.sort()\n",
      "    return r\n",
      "\n",
      "def update_rank(rank_table,permut_assoc_fh):\n",
      "    assoc_table = pd.read_csv(permut_assoc_fh,index_col=0)\n",
      "    r = assoc_table.apply(lambda row: get_rank(row,rank_table),axis=1)\n",
      "    return r\n",
      "    \n",
      "def total_rank(rank_table, permut_fns):\n",
      "    rt = rank_table.copy()\n",
      "    for f in permut_fns:\n",
      "        rt = update_rank(rt,permut_assoc_fh)\n",
      "    return rt\n",
      "    \n",
      "#def save_p_value_total(rank_table, permut_fns):\n",
      "    \n",
      "    \n",
      "def empirical_rank(value,dist):\n",
      "    \"\"\"\n",
      "    get empirical p value of\n",
      "    value with respect to list of \n",
      "    values in dist\n",
      "    \"\"\"\n",
      "    array = np.append(value,dist)\n",
      "    temp = array.argsort()\n",
      "    ranks = np.empty(len(array), int)\n",
      "    ranks[temp] = np.arange(len(array))\n",
      "    return ranks[0]\n",
      "\n",
      "def get_rank(series,rank_df):\n",
      "    go = series.name\n",
      "    try:\n",
      "        go_s = rank_df.ix[go]\n",
      "    except KeyError:\n",
      "        go_s = pd.Series({\"n_genes\":0,\"rank\":0,\"out_of\":0})\n",
      "    real_val = go_s[\"n_genes\"]\n",
      "    old_rank = go_s[\"rank\"]\n",
      "    old_out_of = go_s[\"out_of\"]\n",
      "    rank = empirical_rank(real_val,series.values)\n",
      "    new_rank = old_rank + rank\n",
      "    new_out_of = old_out_of + len(series)\n",
      "    return pd.Series({\"n_genes\":real_val,\"rank\":new_rank,\"out_of\":new_out_of})\n",
      "\n",
      "def get_genes_per_go(gene_ls,gene_to_go):\n",
      "    s = gene_to_go.set_index(\"gene_symbol\").ix[gene_ls].groupby(\"go_identifier\").apply(lambda x: list(x.index))\n",
      "    s.name = \"genes\"\n",
      "    return sqst"
     ],
     "language": "python",
     "metadata": {},
     "outputs": [],
     "prompt_number": 1199
    },
    {
     "cell_type": "code",
     "collapsed": false,
     "input": [
      "r = get_go_assoc(cand_genes,gene_to_go)\n",
      "rank_table = get_initial_rank_table(r)"
     ],
     "language": "python",
     "metadata": {},
     "outputs": [],
     "prompt_number": 1195
    },
    {
     "cell_type": "code",
     "collapsed": false,
     "input": [
      "genes_per_go = pd.Series(r0.index.values,index=r0.index)"
     ],
     "language": "python",
     "metadata": {},
     "outputs": [],
     "prompt_number": 1178
    },
    {
     "cell_type": "code",
     "collapsed": false,
     "input": [],
     "language": "python",
     "metadata": {},
     "outputs": [],
     "prompt_number": 1190
    },
    {
     "cell_type": "code",
     "collapsed": false,
     "input": [
      "pd.concat([rank_table,gg], axis =1)"
     ],
     "language": "python",
     "metadata": {},
     "outputs": [
      {
       "html": [
        "<div style=\"max-height:1000px;max-width:1500px;overflow:auto;\">\n",
        "<table border=\"1\" class=\"dataframe\">\n",
        "  <thead>\n",
        "    <tr style=\"text-align: right;\">\n",
        "      <th></th>\n",
        "      <th>n_genes</th>\n",
        "      <th>out_of</th>\n",
        "      <th>rank</th>\n",
        "      <th>genes</th>\n",
        "    </tr>\n",
        "    <tr>\n",
        "      <th>go_identifier</th>\n",
        "      <th></th>\n",
        "      <th></th>\n",
        "      <th></th>\n",
        "      <th></th>\n",
        "    </tr>\n",
        "  </thead>\n",
        "  <tbody>\n",
        "    <tr>\n",
        "      <th>GO:0000075</th>\n",
        "      <td>  1</td>\n",
        "      <td> 0</td>\n",
        "      <td> 0</td>\n",
        "      <td>                                          [ANAPC1]</td>\n",
        "    </tr>\n",
        "    <tr>\n",
        "      <th>GO:0000079</th>\n",
        "      <td>  1</td>\n",
        "      <td> 0</td>\n",
        "      <td> 0</td>\n",
        "      <td>                                           [HERC5]</td>\n",
        "    </tr>\n",
        "    <tr>\n",
        "      <th>GO:0000087</th>\n",
        "      <td>  1</td>\n",
        "      <td> 0</td>\n",
        "      <td> 0</td>\n",
        "      <td>                                          [ANAPC1]</td>\n",
        "    </tr>\n",
        "    <tr>\n",
        "      <th>GO:0000090</th>\n",
        "      <td>  1</td>\n",
        "      <td> 0</td>\n",
        "      <td> 0</td>\n",
        "      <td>                                          [ANAPC1]</td>\n",
        "    </tr>\n",
        "    <tr>\n",
        "      <th>GO:0000122</th>\n",
        "      <td>  2</td>\n",
        "      <td> 0</td>\n",
        "      <td> 0</td>\n",
        "      <td>                                     [TP63, TENM2]</td>\n",
        "    </tr>\n",
        "    <tr>\n",
        "      <th>GO:0000139</th>\n",
        "      <td>  2</td>\n",
        "      <td> 0</td>\n",
        "      <td> 0</td>\n",
        "      <td>                                  [HS3ST4, SCARA3]</td>\n",
        "    </tr>\n",
        "    <tr>\n",
        "      <th>GO:0000166</th>\n",
        "      <td> 24</td>\n",
        "      <td> 0</td>\n",
        "      <td> 0</td>\n",
        "      <td> [SCN8A, RAPGEF3, SYK, RFK, EPHB1, RYK, ATP13A4...</td>\n",
        "    </tr>\n",
        "    <tr>\n",
        "      <th>GO:0000185</th>\n",
        "      <td>  1</td>\n",
        "      <td> 0</td>\n",
        "      <td> 0</td>\n",
        "      <td>                                            [GRM5]</td>\n",
        "    </tr>\n",
        "    <tr>\n",
        "      <th>GO:0000187</th>\n",
        "      <td>  2</td>\n",
        "      <td> 0</td>\n",
        "      <td> 0</td>\n",
        "      <td>                                      [SYK, NTRK3]</td>\n",
        "    </tr>\n",
        "    <tr>\n",
        "      <th>GO:0000226</th>\n",
        "      <td>  1</td>\n",
        "      <td> 0</td>\n",
        "      <td> 0</td>\n",
        "      <td>                                           [PARD3]</td>\n",
        "    </tr>\n",
        "    <tr>\n",
        "      <th>GO:0000278</th>\n",
        "      <td>  1</td>\n",
        "      <td> 0</td>\n",
        "      <td> 0</td>\n",
        "      <td>                                          [ANAPC1]</td>\n",
        "    </tr>\n",
        "    <tr>\n",
        "      <th>GO:0000287</th>\n",
        "      <td>  1</td>\n",
        "      <td> 0</td>\n",
        "      <td> 0</td>\n",
        "      <td>                                          [ATP8A2]</td>\n",
        "    </tr>\n",
        "    <tr>\n",
        "      <th>GO:0000506</th>\n",
        "      <td>  1</td>\n",
        "      <td> 0</td>\n",
        "      <td> 0</td>\n",
        "      <td>                                            [PIGY]</td>\n",
        "    </tr>\n",
        "    <tr>\n",
        "      <th>GO:0000723</th>\n",
        "      <td>  1</td>\n",
        "      <td> 0</td>\n",
        "      <td> 0</td>\n",
        "      <td>                                           [XRCC5]</td>\n",
        "    </tr>\n",
        "    <tr>\n",
        "      <th>GO:0000783</th>\n",
        "      <td>  1</td>\n",
        "      <td> 0</td>\n",
        "      <td> 0</td>\n",
        "      <td>                                           [XRCC5]</td>\n",
        "    </tr>\n",
        "    <tr>\n",
        "      <th>GO:0000784</th>\n",
        "      <td>  1</td>\n",
        "      <td> 0</td>\n",
        "      <td> 0</td>\n",
        "      <td>                                           [XRCC5]</td>\n",
        "    </tr>\n",
        "    <tr>\n",
        "      <th>GO:0000785</th>\n",
        "      <td>  1</td>\n",
        "      <td> 0</td>\n",
        "      <td> 0</td>\n",
        "      <td>                                            [TP63]</td>\n",
        "    </tr>\n",
        "    <tr>\n",
        "      <th>GO:0000790</th>\n",
        "      <td>  1</td>\n",
        "      <td> 0</td>\n",
        "      <td> 0</td>\n",
        "      <td>                                            [TP63]</td>\n",
        "    </tr>\n",
        "    <tr>\n",
        "      <th>GO:0000978</th>\n",
        "      <td>  1</td>\n",
        "      <td> 0</td>\n",
        "      <td> 0</td>\n",
        "      <td>                                           [NR3C1]</td>\n",
        "    </tr>\n",
        "    <tr>\n",
        "      <th>GO:0000989</th>\n",
        "      <td>  1</td>\n",
        "      <td> 0</td>\n",
        "      <td> 0</td>\n",
        "      <td>                                            [TP63]</td>\n",
        "    </tr>\n",
        "    <tr>\n",
        "      <th>GO:0001077</th>\n",
        "      <td>  3</td>\n",
        "      <td> 0</td>\n",
        "      <td> 0</td>\n",
        "      <td>                              [TP63, NR3C1, CREB5]</td>\n",
        "    </tr>\n",
        "    <tr>\n",
        "      <th>GO:0001401</th>\n",
        "      <td>  1</td>\n",
        "      <td> 0</td>\n",
        "      <td> 0</td>\n",
        "      <td>                                          [SAMM50]</td>\n",
        "    </tr>\n",
        "    <tr>\n",
        "      <th>GO:0001501</th>\n",
        "      <td>  3</td>\n",
        "      <td> 0</td>\n",
        "      <td> 0</td>\n",
        "      <td>                                [TP63, ANKH, ARSE]</td>\n",
        "    </tr>\n",
        "    <tr>\n",
        "      <th>GO:0001508</th>\n",
        "      <td>  1</td>\n",
        "      <td> 0</td>\n",
        "      <td> 0</td>\n",
        "      <td>                                           [AKAP6]</td>\n",
        "    </tr>\n",
        "    <tr>\n",
        "      <th>GO:0001518</th>\n",
        "      <td>  1</td>\n",
        "      <td> 0</td>\n",
        "      <td> 0</td>\n",
        "      <td>                                           [SCN8A]</td>\n",
        "    </tr>\n",
        "    <tr>\n",
        "      <th>GO:0001523</th>\n",
        "      <td>  1</td>\n",
        "      <td> 0</td>\n",
        "      <td> 0</td>\n",
        "      <td>                                            [GPC5]</td>\n",
        "    </tr>\n",
        "    <tr>\n",
        "      <th>GO:0001525</th>\n",
        "      <td>  4</td>\n",
        "      <td> 0</td>\n",
        "      <td> 0</td>\n",
        "      <td>                     [RAPGEF3, SYK, EPHB1, THSD7A]</td>\n",
        "    </tr>\n",
        "    <tr>\n",
        "      <th>GO:0001539</th>\n",
        "      <td>  1</td>\n",
        "      <td> 0</td>\n",
        "      <td> 0</td>\n",
        "      <td>                                          [DNAH17]</td>\n",
        "    </tr>\n",
        "    <tr>\n",
        "      <th>GO:0001541</th>\n",
        "      <td>  1</td>\n",
        "      <td> 0</td>\n",
        "      <td> 0</td>\n",
        "      <td>                                            [FSHB]</td>\n",
        "    </tr>\n",
        "    <tr>\n",
        "      <th>GO:0001558</th>\n",
        "      <td>  1</td>\n",
        "      <td> 0</td>\n",
        "      <td> 0</td>\n",
        "      <td>                                            [NRG3]</td>\n",
        "    </tr>\n",
        "    <tr>\n",
        "      <th>...</th>\n",
        "      <td>...</td>\n",
        "      <td>...</td>\n",
        "      <td>...</td>\n",
        "      <td>...</td>\n",
        "    </tr>\n",
        "    <tr>\n",
        "      <th>GO:1900087</th>\n",
        "      <td>  1</td>\n",
        "      <td> 0</td>\n",
        "      <td> 0</td>\n",
        "      <td>                                         [PPP1R1C]</td>\n",
        "    </tr>\n",
        "    <tr>\n",
        "      <th>GO:1900170</th>\n",
        "      <td>  1</td>\n",
        "      <td> 0</td>\n",
        "      <td> 0</td>\n",
        "      <td>                                           [NR3C1]</td>\n",
        "    </tr>\n",
        "    <tr>\n",
        "      <th>GO:1900180</th>\n",
        "      <td>  1</td>\n",
        "      <td> 0</td>\n",
        "      <td> 0</td>\n",
        "      <td>                                          [RASSF5]</td>\n",
        "    </tr>\n",
        "    <tr>\n",
        "      <th>GO:1901018</th>\n",
        "      <td>  1</td>\n",
        "      <td> 0</td>\n",
        "      <td> 0</td>\n",
        "      <td>                                            [ANK2]</td>\n",
        "    </tr>\n",
        "    <tr>\n",
        "      <th>GO:1901019</th>\n",
        "      <td>  1</td>\n",
        "      <td> 0</td>\n",
        "      <td> 0</td>\n",
        "      <td>                                            [ANK2]</td>\n",
        "    </tr>\n",
        "    <tr>\n",
        "      <th>GO:1901021</th>\n",
        "      <td>  1</td>\n",
        "      <td> 0</td>\n",
        "      <td> 0</td>\n",
        "      <td>                                            [ANK2]</td>\n",
        "    </tr>\n",
        "    <tr>\n",
        "      <th>GO:1901374</th>\n",
        "      <td>  1</td>\n",
        "      <td> 0</td>\n",
        "      <td> 0</td>\n",
        "      <td>                                         [SLC22A1]</td>\n",
        "    </tr>\n",
        "    <tr>\n",
        "      <th>GO:1901379</th>\n",
        "      <td>  1</td>\n",
        "      <td> 0</td>\n",
        "      <td> 0</td>\n",
        "      <td>                                            [DPP6]</td>\n",
        "    </tr>\n",
        "    <tr>\n",
        "      <th>GO:1901381</th>\n",
        "      <td>  2</td>\n",
        "      <td> 0</td>\n",
        "      <td> 0</td>\n",
        "      <td>                                     [DPP6, AKAP6]</td>\n",
        "    </tr>\n",
        "    <tr>\n",
        "      <th>GO:1902187</th>\n",
        "      <td>  1</td>\n",
        "      <td> 0</td>\n",
        "      <td> 0</td>\n",
        "      <td>                                          [TRIM35]</td>\n",
        "    </tr>\n",
        "    <tr>\n",
        "      <th>GO:1902261</th>\n",
        "      <td>  1</td>\n",
        "      <td> 0</td>\n",
        "      <td> 0</td>\n",
        "      <td>                                           [AKAP6]</td>\n",
        "    </tr>\n",
        "    <tr>\n",
        "      <th>GO:1902287</th>\n",
        "      <td>  1</td>\n",
        "      <td> 0</td>\n",
        "      <td> 0</td>\n",
        "      <td>                                          [PLXNA4]</td>\n",
        "    </tr>\n",
        "    <tr>\n",
        "      <th>GO:1902808</th>\n",
        "      <td>  1</td>\n",
        "      <td> 0</td>\n",
        "      <td> 0</td>\n",
        "      <td>                                            [TP63]</td>\n",
        "    </tr>\n",
        "    <tr>\n",
        "      <th>GO:1990254</th>\n",
        "      <td>  1</td>\n",
        "      <td> 0</td>\n",
        "      <td> 0</td>\n",
        "      <td>                                           [KRT74]</td>\n",
        "    </tr>\n",
        "    <tr>\n",
        "      <th>GO:2000010</th>\n",
        "      <td>  1</td>\n",
        "      <td> 0</td>\n",
        "      <td> 0</td>\n",
        "      <td>                                           [GPD1L]</td>\n",
        "    </tr>\n",
        "    <tr>\n",
        "      <th>GO:2000096</th>\n",
        "      <td>  1</td>\n",
        "      <td> 0</td>\n",
        "      <td> 0</td>\n",
        "      <td>                                           [MLLT3]</td>\n",
        "    </tr>\n",
        "    <tr>\n",
        "      <th>GO:2000249</th>\n",
        "      <td>  1</td>\n",
        "      <td> 0</td>\n",
        "      <td> 0</td>\n",
        "      <td>                                         [RAPGEF3]</td>\n",
        "    </tr>\n",
        "    <tr>\n",
        "      <th>GO:2000251</th>\n",
        "      <td>  1</td>\n",
        "      <td> 0</td>\n",
        "      <td> 0</td>\n",
        "      <td>                                           [NTRK3]</td>\n",
        "    </tr>\n",
        "    <tr>\n",
        "      <th>GO:2000271</th>\n",
        "      <td>  1</td>\n",
        "      <td> 0</td>\n",
        "      <td> 0</td>\n",
        "      <td>                                            [TP63]</td>\n",
        "    </tr>\n",
        "    <tr>\n",
        "      <th>GO:2000366</th>\n",
        "      <td>  1</td>\n",
        "      <td> 0</td>\n",
        "      <td> 0</td>\n",
        "      <td>                                           [ERBB4]</td>\n",
        "    </tr>\n",
        "    <tr>\n",
        "      <th>GO:2000381</th>\n",
        "      <td>  1</td>\n",
        "      <td> 0</td>\n",
        "      <td> 0</td>\n",
        "      <td>                                            [TP63]</td>\n",
        "    </tr>\n",
        "    <tr>\n",
        "      <th>GO:2000599</th>\n",
        "      <td>  1</td>\n",
        "      <td> 0</td>\n",
        "      <td> 0</td>\n",
        "      <td>                                           [NELL1]</td>\n",
        "    </tr>\n",
        "    <tr>\n",
        "      <th>GO:2000642</th>\n",
        "      <td>  1</td>\n",
        "      <td> 0</td>\n",
        "      <td> 0</td>\n",
        "      <td>                                            [SNX3]</td>\n",
        "    </tr>\n",
        "    <tr>\n",
        "      <th>GO:2000649</th>\n",
        "      <td>  1</td>\n",
        "      <td> 0</td>\n",
        "      <td> 0</td>\n",
        "      <td>                                           [GPD1L]</td>\n",
        "    </tr>\n",
        "    <tr>\n",
        "      <th>GO:2000773</th>\n",
        "      <td>  1</td>\n",
        "      <td> 0</td>\n",
        "      <td> 0</td>\n",
        "      <td>                                            [TP63]</td>\n",
        "    </tr>\n",
        "    <tr>\n",
        "      <th>GO:2001235</th>\n",
        "      <td>  1</td>\n",
        "      <td> 0</td>\n",
        "      <td> 0</td>\n",
        "      <td>                                            [TP63]</td>\n",
        "    </tr>\n",
        "    <tr>\n",
        "      <th>GO:2001237</th>\n",
        "      <td>  1</td>\n",
        "      <td> 0</td>\n",
        "      <td> 0</td>\n",
        "      <td>                                           [ITGA6]</td>\n",
        "    </tr>\n",
        "    <tr>\n",
        "      <th>GO:2001244</th>\n",
        "      <td>  1</td>\n",
        "      <td> 0</td>\n",
        "      <td> 0</td>\n",
        "      <td>                                          [IL20RA]</td>\n",
        "    </tr>\n",
        "    <tr>\n",
        "      <th>GO:2001257</th>\n",
        "      <td>  1</td>\n",
        "      <td> 0</td>\n",
        "      <td> 0</td>\n",
        "      <td>                                            [ANK2]</td>\n",
        "    </tr>\n",
        "    <tr>\n",
        "      <th>GO:2001259</th>\n",
        "      <td>  1</td>\n",
        "      <td> 0</td>\n",
        "      <td> 0</td>\n",
        "      <td>                                            [ANK2]</td>\n",
        "    </tr>\n",
        "  </tbody>\n",
        "</table>\n",
        "<p>1265 rows \u00d7 4 columns</p>\n",
        "</div>"
       ],
       "metadata": {},
       "output_type": "pyout",
       "prompt_number": 1203,
       "text": [
        "               n_genes  out_of  rank  \\\n",
        "go_identifier                          \n",
        "GO:0000075           1       0     0   \n",
        "GO:0000079           1       0     0   \n",
        "GO:0000087           1       0     0   \n",
        "GO:0000090           1       0     0   \n",
        "GO:0000122           2       0     0   \n",
        "GO:0000139           2       0     0   \n",
        "GO:0000166          24       0     0   \n",
        "GO:0000185           1       0     0   \n",
        "GO:0000187           2       0     0   \n",
        "GO:0000226           1       0     0   \n",
        "GO:0000278           1       0     0   \n",
        "GO:0000287           1       0     0   \n",
        "GO:0000506           1       0     0   \n",
        "GO:0000723           1       0     0   \n",
        "GO:0000783           1       0     0   \n",
        "GO:0000784           1       0     0   \n",
        "GO:0000785           1       0     0   \n",
        "GO:0000790           1       0     0   \n",
        "GO:0000978           1       0     0   \n",
        "GO:0000989           1       0     0   \n",
        "GO:0001077           3       0     0   \n",
        "GO:0001401           1       0     0   \n",
        "GO:0001501           3       0     0   \n",
        "GO:0001508           1       0     0   \n",
        "GO:0001518           1       0     0   \n",
        "GO:0001523           1       0     0   \n",
        "GO:0001525           4       0     0   \n",
        "GO:0001539           1       0     0   \n",
        "GO:0001541           1       0     0   \n",
        "GO:0001558           1       0     0   \n",
        "...                ...     ...   ...   \n",
        "GO:1900087           1       0     0   \n",
        "GO:1900170           1       0     0   \n",
        "GO:1900180           1       0     0   \n",
        "GO:1901018           1       0     0   \n",
        "GO:1901019           1       0     0   \n",
        "GO:1901021           1       0     0   \n",
        "GO:1901374           1       0     0   \n",
        "GO:1901379           1       0     0   \n",
        "GO:1901381           2       0     0   \n",
        "GO:1902187           1       0     0   \n",
        "GO:1902261           1       0     0   \n",
        "GO:1902287           1       0     0   \n",
        "GO:1902808           1       0     0   \n",
        "GO:1990254           1       0     0   \n",
        "GO:2000010           1       0     0   \n",
        "GO:2000096           1       0     0   \n",
        "GO:2000249           1       0     0   \n",
        "GO:2000251           1       0     0   \n",
        "GO:2000271           1       0     0   \n",
        "GO:2000366           1       0     0   \n",
        "GO:2000381           1       0     0   \n",
        "GO:2000599           1       0     0   \n",
        "GO:2000642           1       0     0   \n",
        "GO:2000649           1       0     0   \n",
        "GO:2000773           1       0     0   \n",
        "GO:2001235           1       0     0   \n",
        "GO:2001237           1       0     0   \n",
        "GO:2001244           1       0     0   \n",
        "GO:2001257           1       0     0   \n",
        "GO:2001259           1       0     0   \n",
        "\n",
        "                                                           genes  \n",
        "go_identifier                                                     \n",
        "GO:0000075                                              [ANAPC1]  \n",
        "GO:0000079                                               [HERC5]  \n",
        "GO:0000087                                              [ANAPC1]  \n",
        "GO:0000090                                              [ANAPC1]  \n",
        "GO:0000122                                         [TP63, TENM2]  \n",
        "GO:0000139                                      [HS3ST4, SCARA3]  \n",
        "GO:0000166     [SCN8A, RAPGEF3, SYK, RFK, EPHB1, RYK, ATP13A4...  \n",
        "GO:0000185                                                [GRM5]  \n",
        "GO:0000187                                          [SYK, NTRK3]  \n",
        "GO:0000226                                               [PARD3]  \n",
        "GO:0000278                                              [ANAPC1]  \n",
        "GO:0000287                                              [ATP8A2]  \n",
        "GO:0000506                                                [PIGY]  \n",
        "GO:0000723                                               [XRCC5]  \n",
        "GO:0000783                                               [XRCC5]  \n",
        "GO:0000784                                               [XRCC5]  \n",
        "GO:0000785                                                [TP63]  \n",
        "GO:0000790                                                [TP63]  \n",
        "GO:0000978                                               [NR3C1]  \n",
        "GO:0000989                                                [TP63]  \n",
        "GO:0001077                                  [TP63, NR3C1, CREB5]  \n",
        "GO:0001401                                              [SAMM50]  \n",
        "GO:0001501                                    [TP63, ANKH, ARSE]  \n",
        "GO:0001508                                               [AKAP6]  \n",
        "GO:0001518                                               [SCN8A]  \n",
        "GO:0001523                                                [GPC5]  \n",
        "GO:0001525                         [RAPGEF3, SYK, EPHB1, THSD7A]  \n",
        "GO:0001539                                              [DNAH17]  \n",
        "GO:0001541                                                [FSHB]  \n",
        "GO:0001558                                                [NRG3]  \n",
        "...                                                          ...  \n",
        "GO:1900087                                             [PPP1R1C]  \n",
        "GO:1900170                                               [NR3C1]  \n",
        "GO:1900180                                              [RASSF5]  \n",
        "GO:1901018                                                [ANK2]  \n",
        "GO:1901019                                                [ANK2]  \n",
        "GO:1901021                                                [ANK2]  \n",
        "GO:1901374                                             [SLC22A1]  \n",
        "GO:1901379                                                [DPP6]  \n",
        "GO:1901381                                         [DPP6, AKAP6]  \n",
        "GO:1902187                                              [TRIM35]  \n",
        "GO:1902261                                               [AKAP6]  \n",
        "GO:1902287                                              [PLXNA4]  \n",
        "GO:1902808                                                [TP63]  \n",
        "GO:1990254                                               [KRT74]  \n",
        "GO:2000010                                               [GPD1L]  \n",
        "GO:2000096                                               [MLLT3]  \n",
        "GO:2000249                                             [RAPGEF3]  \n",
        "GO:2000251                                               [NTRK3]  \n",
        "GO:2000271                                                [TP63]  \n",
        "GO:2000366                                               [ERBB4]  \n",
        "GO:2000381                                                [TP63]  \n",
        "GO:2000599                                               [NELL1]  \n",
        "GO:2000642                                                [SNX3]  \n",
        "GO:2000649                                               [GPD1L]  \n",
        "GO:2000773                                                [TP63]  \n",
        "GO:2001235                                                [TP63]  \n",
        "GO:2001237                                               [ITGA6]  \n",
        "GO:2001244                                              [IL20RA]  \n",
        "GO:2001257                                                [ANK2]  \n",
        "GO:2001259                                                [ANK2]  \n",
        "\n",
        "[1265 rows x 4 columns]"
       ]
      }
     ],
     "prompt_number": 1203
    },
    {
     "cell_type": "code",
     "collapsed": false,
     "input": [
      "gg = get_genes_per_go(cand_genes,gene_to_go)"
     ],
     "language": "python",
     "metadata": {},
     "outputs": [],
     "prompt_number": 1202
    },
    {
     "cell_type": "code",
     "collapsed": false,
     "input": [
      "r = permut_assoc(tot_sweed_df[\"Likelihood\"], 0.1, gene_df, gene_to_go, top_n = 500, max_dist=10000)"
     ],
     "language": "python",
     "metadata": {},
     "outputs": [],
     "prompt_number": 927
    },
    {
     "cell_type": "code",
     "collapsed": false,
     "input": [
      "real_assoc = permut_assoc(tot_sweed_df[\"Likelihood\"], 0, gene_df, gene_to_go, top_n=1000, max_dist=10000)"
     ],
     "language": "python",
     "metadata": {},
     "outputs": [],
     "prompt_number": 934
    },
    {
     "cell_type": "code",
     "collapsed": false,
     "input": [
      "assoc_table = multiple_permut_assoc(tot_sweed_df[\"Likelihood\"], gene_df, gene_to_go, top_n=1000, max_dist = 10000, n_runs=100, rnd_seed=None)"
     ],
     "language": "python",
     "metadata": {},
     "outputs": [],
     "prompt_number": 933
    },
    {
     "cell_type": "code",
     "collapsed": false,
     "input": [
      "def f(r):\n",
      "    print r\n",
      "    raise"
     ],
     "language": "python",
     "metadata": {},
     "outputs": [],
     "prompt_number": 952
    },
    {
     "cell_type": "code",
     "collapsed": false,
     "input": [
      "p_vals = assoc_table.apply(lambda r: get_p_val(r,real_assoc),axis=1)"
     ],
     "language": "python",
     "metadata": {},
     "outputs": [],
     "prompt_number": 959
    },
    {
     "cell_type": "code",
     "collapsed": false,
     "input": [
      "p_vals.sort()"
     ],
     "language": "python",
     "metadata": {},
     "outputs": [],
     "prompt_number": 960
    },
    {
     "cell_type": "code",
     "collapsed": false,
     "input": [
      "sab_enrich = pd.read_csv(eu(\"~/vervet_project/analyses/20140611_163_subpop_sweepfinder/_data/enrichment_n2000_d10000_permut50000_sab.tsv\"),\n",
      "                            index_col=[0],sep=\"\\t\")"
     ],
     "language": "python",
     "metadata": {},
     "outputs": [],
     "prompt_number": 1206
    },
    {
     "cell_type": "code",
     "collapsed": false,
     "input": [
      "len(gene_to_go.set_index(\"go_identifier\").index.diff(sab_enrich.index))"
     ],
     "language": "python",
     "metadata": {},
     "outputs": [
      {
       "metadata": {},
       "output_type": "pyout",
       "prompt_number": 1218,
       "text": [
        "432"
       ]
      }
     ],
     "prompt_number": 1218
    },
    {
     "cell_type": "code",
     "collapsed": false,
     "input": [
      "len(sab_enrich)"
     ],
     "language": "python",
     "metadata": {},
     "outputs": [
      {
       "metadata": {},
       "output_type": "pyout",
       "prompt_number": 1220,
       "text": [
        "14566"
       ]
      }
     ],
     "prompt_number": 1220
    },
    {
     "cell_type": "code",
     "collapsed": false,
     "input": [
      "len(gene_to_go.set_index(\"go_identifier\").index.unique())"
     ],
     "language": "python",
     "metadata": {},
     "outputs": [
      {
       "metadata": {},
       "output_type": "pyout",
       "prompt_number": 1222,
       "text": [
        "14998"
       ]
      }
     ],
     "prompt_number": 1222
    },
    {
     "cell_type": "code",
     "collapsed": false,
     "input": [],
     "language": "python",
     "metadata": {},
     "outputs": [],
     "prompt_number": 1223
    },
    {
     "cell_type": "heading",
     "level": 2,
     "metadata": {},
     "source": [
      "Functions as used in test_enrichment.py\n"
     ]
    },
    {
     "cell_type": "markdown",
     "metadata": {},
     "source": [
      "(in case of conflict the most recent version should be there)"
     ]
    },
    {
     "cell_type": "code",
     "collapsed": false,
     "input": [
      "from hs_vervet.scripts import test_enrichment as te"
     ],
     "language": "python",
     "metadata": {},
     "outputs": [],
     "prompt_number": 3
    },
    {
     "cell_type": "code",
     "collapsed": false,
     "input": [
      "s1 = pd.Series([1,2,3,4,8],index=[1,2,3,4,8])\n",
      "s2 = pd.Series([1,2,7,14],index=[1,2,7,14])"
     ],
     "language": "python",
     "metadata": {},
     "outputs": []
    },
    {
     "cell_type": "code",
     "collapsed": false,
     "input": [
      "df1 = pd.DataFrame(17, columns=[\"rank\",\"out_of\",\"n_genes\"], index=[\"G0:1\",\"G0:2\",\"G0:3\",\"G0:4\",\"G0:5\",\"G0:6\",\"G0:7\"])\n",
      "df2 = pd.DataFrame(19, columns=[\"rank\",\"out_of\",\"n_genes\"], index=[\"G0:1\",\"G0:3\",\"G0:4\",\"G0:5\",\"G0:6\",\"G0:7\"])"
     ],
     "language": "python",
     "metadata": {},
     "outputs": [],
     "prompt_number": 7
    },
    {
     "cell_type": "code",
     "collapsed": false,
     "input": [
      "df1[\"rank\"].add(df2[\"rank\"],fill_value=0)"
     ],
     "language": "python",
     "metadata": {},
     "outputs": [
      {
       "metadata": {},
       "output_type": "pyout",
       "prompt_number": 12,
       "text": [
        "G0:1    36\n",
        "G0:2    17\n",
        "G0:3    36\n",
        "G0:4    36\n",
        "G0:5    36\n",
        "G0:6    36\n",
        "G0:7    36\n",
        "Name: rank, dtype: float64"
       ]
      }
     ],
     "prompt_number": 12
    },
    {
     "cell_type": "code",
     "collapsed": false,
     "input": [
      "fn0 = open(eu(\"~/vervet_project/analyses/20140611_163_subpop_sweepfinder/_data/sweed_go_enrichment_q0.001_d10000_0_pys.temp.tsv\"))\n",
      "fn1 = open(eu(\"~/vervet_project/analyses/20140611_163_subpop_sweepfinder/_data/sweed_go_enrichment_q0.001_d10000_1_pys.temp.tsv\"))\n",
      "tot_rank = te.read_table(fn0,index_col=0)\n",
      "rank_table = te.read_table(fn1,index_col=0)"
     ],
     "language": "python",
     "metadata": {},
     "outputs": [],
     "prompt_number": 17
    },
    {
     "cell_type": "code",
     "collapsed": false,
     "input": [
      "tot_rank.index.diff(rank_table.index)"
     ],
     "language": "python",
     "metadata": {},
     "outputs": [
      {
       "metadata": {},
       "output_type": "pyout",
       "prompt_number": 20,
       "text": [
        "Index([u'000'], dtype='object')"
       ]
      }
     ],
     "prompt_number": 20
    },
    {
     "cell_type": "code",
     "collapsed": false,
     "input": [
      "rank_table.index.diff(tot_rank.index)"
     ],
     "language": "python",
     "metadata": {},
     "outputs": [
      {
       "metadata": {},
       "output_type": "pyout",
       "prompt_number": 21,
       "text": [
        "Index([u'00'], dtype='object')"
       ]
      }
     ],
     "prompt_number": 21
    },
    {
     "cell_type": "code",
     "collapsed": false,
     "input": [
      "l = [1,2,3,4,8]"
     ],
     "language": "python",
     "metadata": {},
     "outputs": [],
     "prompt_number": 22
    },
    {
     "cell_type": "code",
     "collapsed": false,
     "input": [
      "l[:-1]+l[-1:-3:-1]"
     ],
     "language": "python",
     "metadata": {},
     "outputs": [
      {
       "metadata": {},
       "output_type": "pyout",
       "prompt_number": 28,
       "text": [
        "[1, 2, 3, 4, 8, 4]"
       ]
      }
     ],
     "prompt_number": 28
    },
    {
     "cell_type": "code",
     "collapsed": false,
     "input": [],
     "language": "python",
     "metadata": {},
     "outputs": []
    }
   ],
   "metadata": {}
  }
 ]
}