{
 "metadata": {
  "name": "parse_vcf"
 },
 "nbformat": 3,
 "nbformat_minor": 0,
 "worksheets": [
  {
   "cells": [
    {
     "cell_type": "code",
     "collapsed": false,
     "input": [
      "import os, sys\n",
      "sys.path.insert(0, os.path.expanduser('~/lib/python'))\n",
      "sys.path.insert(0, os.path.join(os.path.expanduser('~/script')))"
     ],
     "language": "python",
     "metadata": {},
     "outputs": [],
     "prompt_number": 1
    },
    {
     "cell_type": "code",
     "collapsed": false,
     "input": [
      "import vcf\n",
      "import pandas as pd\n",
      "import hs_vervet.tools.hs_vervet_basics as hvb"
     ],
     "language": "python",
     "metadata": {},
     "outputs": [
      {
       "output_type": "stream",
       "stream": "stderr",
       "text": [
        "/home/GMI/hannes.svardal/.local/lib/python2.7/site-packages/pytz/__init__.py:35: UserWarning: Module vcf was already imported from /home/GMI/hannes.svardal/lib/python/vcf/__init__.pyc, but /home/GMI/hannes.svardal/.local/lib/python2.7/site-packages/PyVCF-0.6.4-py2.7.egg is being added to sys.path\n",
        "  from pkg_resources import resource_stream\n"
       ]
      }
     ],
     "prompt_number": 2
    },
    {
     "cell_type": "code",
     "collapsed": false,
     "input": [
      "vcf_fname=os.path.expanduser('~/vervet_data/data/cortex/African_sabaeus.vcf')"
     ],
     "language": "python",
     "metadata": {},
     "outputs": [],
     "prompt_number": 4
    },
    {
     "cell_type": "code",
     "collapsed": false,
     "input": [
      "vcf_fname=os.path.expanduser('~/vervet_project/analyses/20131121_UnifiedGenotyper_African_Sabaeus/_data/GATK_UnifiedGenotyper_Gambian_Sabaeus_CAEY_filter.vcf')"
     ],
     "language": "python",
     "metadata": {},
     "outputs": [],
     "prompt_number": 6
    },
    {
     "cell_type": "code",
     "collapsed": false,
     "input": [
      "vcf_reader = vcf.Reader(open(vcf_fname, 'r'))"
     ],
     "language": "python",
     "metadata": {},
     "outputs": [],
     "prompt_number": 7
    },
    {
     "cell_type": "code",
     "collapsed": false,
     "input": [
      "vcf_reader.__dict__"
     ],
     "language": "python",
     "metadata": {},
     "outputs": [
      {
       "output_type": "pyout",
       "prompt_number": 8,
       "text": [
        "{'_column_headers': ['CHROM',\n",
        "  'POS',\n",
        "  'ID',\n",
        "  'REF',\n",
        "  'ALT',\n",
        "  'QUAL',\n",
        "  'FILTER',\n",
        "  'INFO',\n",
        "  'FORMAT'],\n",
        " '_format_cache': {},\n",
        " '_header_lines': ['##fileformat=VCFv4.1',\n",
        "  '##FILTER=<ID=ExcessCov,Description=\"Overlaps a user-input mask\">',\n",
        "  '##FILTER=<ID=LowQual,Description=\"Low quality\">',\n",
        "  '##FILTER=<ID=PoorMapping,Description=\"Overlaps a user-input mask\">',\n",
        "  '##FORMAT=<ID=AD,Number=.,Type=Integer,Description=\"Allelic depths for the ref and alt alleles in the order listed\">',\n",
        "  '##FORMAT=<ID=DP,Number=1,Type=Integer,Description=\"Approximate read depth (reads with MQ=255 or with bad mates are filtered)\">',\n",
        "  '##FORMAT=<ID=GQ,Number=1,Type=Integer,Description=\"Genotype Quality\">',\n",
        "  '##FORMAT=<ID=GT,Number=1,Type=String,Description=\"Genotype\">',\n",
        "  '##FORMAT=<ID=PL,Number=G,Type=Integer,Description=\"Normalized, Phred-scaled likelihoods for genotypes as defined in the VCF specification\">',\n",
        "  '##INFO=<ID=AC,Number=A,Type=Integer,Description=\"Allele count in genotypes, for each ALT allele, in the same order as listed\">',\n",
        "  '##INFO=<ID=AF,Number=A,Type=Float,Description=\"Allele Frequency, for each ALT allele, in the same order as listed\">',\n",
        "  '##INFO=<ID=AN,Number=1,Type=Integer,Description=\"Total number of alleles in called genotypes\">',\n",
        "  '##INFO=<ID=BaseQRankSum,Number=1,Type=Float,Description=\"Z-score from Wilcoxon rank sum test of Alt Vs. Ref base qualities\">',\n",
        "  '##INFO=<ID=DP,Number=1,Type=Integer,Description=\"Approximate read depth; some reads may have been filtered\">',\n",
        "  '##INFO=<ID=DS,Number=0,Type=Flag,Description=\"Were any of the samples downsampled?\">',\n",
        "  '##INFO=<ID=Dels,Number=1,Type=Float,Description=\"Fraction of Reads Containing Spanning Deletions\">',\n",
        "  '##INFO=<ID=FS,Number=1,Type=Float,Description=\"Phred-scaled p-value using Fisher\\'s exact test to detect strand bias\">',\n",
        "  '##INFO=<ID=HaplotypeScore,Number=1,Type=Float,Description=\"Consistency of the site with at most two segregating haplotypes\">',\n",
        "  '##INFO=<ID=InbreedingCoeff,Number=1,Type=Float,Description=\"Inbreeding coefficient as estimated from the genotype likelihoods per-sample when compared against the Hardy-Weinberg expectation\">',\n",
        "  '##INFO=<ID=MLEAC,Number=A,Type=Integer,Description=\"Maximum likelihood expectation (MLE) for the allele counts (not necessarily the same as the AC), for each ALT allele, in the same order as listed\">',\n",
        "  '##INFO=<ID=MLEAF,Number=A,Type=Float,Description=\"Maximum likelihood expectation (MLE) for the allele frequency (not necessarily the same as the AF), for each ALT allele, in the same order as listed\">',\n",
        "  '##INFO=<ID=MQ,Number=1,Type=Float,Description=\"RMS Mapping Quality\">',\n",
        "  '##INFO=<ID=MQ0,Number=1,Type=Integer,Description=\"Total Mapping Quality Zero Reads\">',\n",
        "  '##INFO=<ID=MQRankSum,Number=1,Type=Float,Description=\"Z-score From Wilcoxon rank sum test of Alt vs. Ref read mapping qualities\">',\n",
        "  '##INFO=<ID=QD,Number=1,Type=Float,Description=\"Variant Confidence/Quality by Depth\">',\n",
        "  '##INFO=<ID=RPA,Number=.,Type=Integer,Description=\"Number of times tandem repeat unit is repeated, for each allele (including reference)\">',\n",
        "  '##INFO=<ID=RU,Number=1,Type=String,Description=\"Tandem repeat unit (bases)\">',\n",
        "  '##INFO=<ID=ReadPosRankSum,Number=1,Type=Float,Description=\"Z-score from Wilcoxon rank sum test of Alt vs. Ref read position bias\">',\n",
        "  '##INFO=<ID=STR,Number=0,Type=Flag,Description=\"Variant is a short tandem repeat\">',\n",
        "  '##UnifiedGenotyper=\"analysis_type=UnifiedGenotyper input_file=[data/db/individual_alignment/6154_3432_VGA00002_GA_vs_3488_by_method6_realigned1_reduced0_p4383_m227.bam, data/db/individual_alignment/6126_1436_VGA00006_GA_vs_3488_by_method6_realigned1_reduced0_p4355_m227.bam, data/db/individual_alignment/6164_3442_VGA00010_GA_vs_3488_by_method6_realigned1_reduced0_p4393_m227.bam, data/db/individual_alignment/6167_3445_VGA00019_GA_vs_3488_by_method6_realigned1_reduced0_p4396_m227.bam, data/db/individual_alignment/6129_1444_VGA00021_GA_vs_3488_by_method6_realigned1_reduced0_p4358_m227.bam, data/db/individual_alignment/6177_3497_VGA00025_GA_vs_3488_by_method6_realigned1_reduced0_p4406_m227.bam, data/db/individual_alignment/6137_1484_VGA00085_GA_vs_3488_by_method6_realigned1_reduced0_p4366_m227.bam, data/db/individual_alignment/6135_1469_VGA00090_GA_vs_3488_by_method6_realigned1_reduced0_p4364_m227.bam, data/db/individual_alignment/6128_1443_VGA00092_GA_vs_3488_by_method6_realigned1_reduced0_p4357_m227.bam, data/db/individual_alignment/6132_1458_VGA00098_GA_vs_3488_by_method6_realigned1_reduced0_p4361_m227.bam, data/db/individual_alignment/6146_3424_VGA00100_GA_vs_3488_by_method6_realigned1_reduced0_p4375_m227.bam, data/db/individual_alignment/6140_1506_VGA00101_GA_vs_3488_by_method6_realigned1_reduced0_p4369_m227.bam, data/db/individual_alignment/6147_3425_VGA00107_GA_vs_3488_by_method6_realigned1_reduced0_p4376_m227.bam, data/db/individual_alignment/6127_1441_VGA00114_GA_vs_3488_by_method6_realigned1_reduced0_p4356_m227.bam, data/db/individual_alignment/6159_3437_VGA00138_GA_vs_3488_by_method6_realigned1_reduced0_p4388_m227.bam, data/db/individual_alignment/6134_1462_VGA00142_GA_vs_3488_by_method6_realigned1_reduced0_p4363_m227.bam, data/db/individual_alignment/6131_1457_VGA00146_GA_vs_3488_by_method6_realigned1_reduced0_p4360_m227.bam, data/db/individual_alignment/6158_3436_VGA00151_GA_vs_3488_by_method6_realigned1_reduced0_p4387_m227.bam, data/db/individual_alignment/6143_1523_VGA00152_GA_vs_3488_by_method6_realigned1_reduced0_p4372_m227.bam, data/db/individual_alignment/6144_1524_VGA00153_GA_vs_3488_by_method6_realigned1_reduced0_p4373_m227.bam, data/db/individual_alignment/6157_3435_VGA00155_GA_vs_3488_by_method6_realigned1_reduced0_p4386_m227.bam, data/db/individual_alignment/6133_1459_VGA00156_GA_vs_3488_by_method6_realigned1_reduced0_p4362_m227.bam] read_buffer_size=null phone_home=STANDARD gatk_key=null tag=NA read_filter=[] intervals=[CAEY] excludeIntervals=null interval_set_rule=UNION interval_merging=ALL interval_padding=0 reference_sequence=data/reference/reference_3488/3488_indID1_codeVRC_ref_sequencer3_seqType1_filtered0_version3.fasta nonDeterministicRandomSeed=false disableDithering=false maxRuntime=-1 maxRuntimeUnits=MINUTES downsampling_type=BY_SAMPLE downsample_to_fraction=null downsample_to_coverage=250 baq=OFF baqGapOpenPenalty=40.0 fix_misencoded_quality_scores=false allow_potentially_misencoded_quality_scores=false performanceLog=null useOriginalQualities=false BQSR=null quantize_quals=0 disable_indel_quals=false emit_original_quals=false preserve_qscores_less_than=6 globalQScorePrior=-1.0 allow_bqsr_on_reduced_bams_despite_repeated_warnings=false defaultBaseQualities=-1 validation_strictness=SILENT remove_program_records=false keep_program_records=false unsafe=null disable_auto_index_creation_and_locking_when_reading_rods=false num_threads=4 num_cpu_threads_per_data_thread=1 num_io_threads=0 monitorThreadEfficiency=false num_bam_file_handles=null read_group_black_list=null pedigree=[] pedigreeString=[] pedigreeValidationType=STRICT allow_intervals_with_unindexed_bam=false generateShadowBCF=false logging_level=INFO log_to_file=null help=false version=false genotype_likelihoods_model=SNP pcr_error_rate=1.0E-4 computeSLOD=false annotateNDA=false pair_hmm_implementation=ORIGINAL min_base_quality_score=17 max_deletion_fraction=0.05 min_indel_count_for_genotyping=5 min_indel_fraction_per_sample=0.25 indelGapContinuationPenalty=10 indelGapOpenPenalty=45 indelHaplotypeSize=80 indelDebug=false ignoreSNPAlleles=false allReadsSP=false ignoreLaneInfo=false reference_sample_calls=(RodBinding name= source=UNBOUND) reference_sample_name=null sample_ploidy=2 min_quality_score=1 max_quality_score=40 site_quality_prior=20 min_power_threshold_for_calling=0.95 min_reference_depth=100 exclude_filtered_reference_sites=false heterozygosity=0.001 indel_heterozygosity=1.25E-4 genotyping_mode=DISCOVERY output_mode=EMIT_VARIANTS_ONLY standard_min_confidence_threshold_for_calling=50.0 standard_min_confidence_threshold_for_emitting=10.0 alleles=(RodBinding name= source=UNBOUND) max_alternate_alleles=6 input_prior=[] contamination_fraction_to_filter=0.05 contamination_fraction_per_sample_file=null p_nonref_model=EXACT_INDEPENDENT exactcallslog=null dbsnp=(RodBinding name= source=UNBOUND) comp=[] out=org.broadinstitute.sting.gatk.io.stubs.VariantContextWriterStub no_cmdline_in_header=org.broadinstitute.sting.gatk.io.stubs.VariantContextWriterStub sites_only=org.broadinstitute.sting.gatk.io.stubs.VariantContextWriterStub bcf=org.broadinstitute.sting.gatk.io.stubs.VariantContextWriterStub debug_file=null metrics_file=null annotation=[] excludeAnnotation=[] filter_mismatching_base_and_quals=false filter_bases_not_stored=false\"',\n",
        "  '##VariantFiltration=\"analysis_type=VariantFiltration input_file=[] read_buffer_size=null phone_home=STANDARD gatk_key=null tag=NA read_filter=[] intervals=null excludeIntervals=null interval_set_rule=UNION interval_merging=ALL interval_padding=0 reference_sequence=data/reference/reference_3488/3488_indID1_codeVRC_ref_sequencer3_seqType1_filtered0_version3.fasta nonDeterministicRandomSeed=false disableDithering=false maxRuntime=-1 maxRuntimeUnits=MINUTES downsampling_type=BY_SAMPLE downsample_to_fraction=null downsample_to_coverage=1000 baq=OFF baqGapOpenPenalty=40.0 fix_misencoded_quality_scores=false allow_potentially_misencoded_quality_scores=false performanceLog=null useOriginalQualities=false BQSR=null quantize_quals=0 disable_indel_quals=false emit_original_quals=false preserve_qscores_less_than=6 globalQScorePrior=-1.0 allow_bqsr_on_reduced_bams_despite_repeated_warnings=false defaultBaseQualities=-1 validation_strictness=SILENT remove_program_records=false keep_program_records=false unsafe=null disable_auto_index_creation_and_locking_when_reading_rods=false num_threads=1 num_cpu_threads_per_data_thread=1 num_io_threads=0 monitorThreadEfficiency=false num_bam_file_handles=null read_group_black_list=null pedigree=[] pedigreeString=[] pedigreeValidationType=STRICT allow_intervals_with_unindexed_bam=false generateShadowBCF=false logging_level=INFO log_to_file=null help=false version=false variant=(RodBinding name=variant source=analyses/20131121_UnifiedGenotyper_African_Sabaeus/_data/GATK_UnifiedGenotyper_Gambian_Sabaeus_CAEY.vcf) mask=(RodBinding name=mask source=analyses/20131129_filtering_masks/_data/excessive_coverage_regions_CAEY.bed) out=org.broadinstitute.sting.gatk.io.stubs.VariantContextWriterStub no_cmdline_in_header=org.broadinstitute.sting.gatk.io.stubs.VariantContextWriterStub sites_only=org.broadinstitute.sting.gatk.io.stubs.VariantContextWriterStub bcf=org.broadinstitute.sting.gatk.io.stubs.VariantContextWriterStub filterExpression=[] filterName=[] genotypeFilterExpression=[] genotypeFilterName=[] clusterSize=3 clusterWindowSize=0 maskExtension=0 maskName=ExcessCov filterNotInMask=false missingValuesInExpressionsShouldEvaluateAsFailing=false invalidatePreviousFilters=false filter_mismatching_base_and_quals=false filter_bases_not_stored=false\"',\n",
        "  '##contig=<ID=CAE1,length=126035930>',\n",
        "  '##contig=<ID=CAE2,length=90373283>',\n",
        "  '##contig=<ID=CAE3,length=92142175>',\n",
        "  '##contig=<ID=CAE4,length=91010382>',\n",
        "  '##contig=<ID=CAE5,length=75399963>',\n",
        "  '##contig=<ID=CAE6,length=50890351>',\n",
        "  '##contig=<ID=CAE7,length=135778131>',\n",
        "  '##contig=<ID=CAE8,length=139301422>',\n",
        "  '##contig=<ID=CAE9,length=125710982>',\n",
        "  '##contig=<ID=CAE10,length=128595539>',\n",
        "  '##contig=<ID=CAE11,length=128539186>',\n",
        "  '##contig=<ID=CAE12,length=108555830>',\n",
        "  '##contig=<ID=CAE13,length=98384682>',\n",
        "  '##contig=<ID=CAE14,length=107702431>',\n",
        "  '##contig=<ID=CAE15,length=91754291>',\n",
        "  '##contig=<ID=CAE16,length=75148670>',\n",
        "  '##contig=<ID=CAE17,length=71996105>',\n",
        "  '##contig=<ID=CAE18,length=72318688>',\n",
        "  '##contig=<ID=CAE19,length=33263144>',\n",
        "  '##contig=<ID=CAE20,length=130588469>',\n",
        "  '##contig=<ID=CAE21,length=127223203>',\n",
        "  '##contig=<ID=CAE22,length=101219884>',\n",
        "  '##contig=<ID=CAE23,length=82825804>',\n",
        "  '##contig=<ID=CAE24,length=84932903>',\n",
        "  '##contig=<ID=CAE25,length=85787240>',\n",
        "  '##contig=<ID=CAE26,length=58131712>',\n",
        "  '##contig=<ID=CAE27,length=48547382>',\n",
        "  '##contig=<ID=CAE28,length=21531802>',\n",
        "  '##contig=<ID=CAE29,length=24206276>',\n",
        "  '##contig=<ID=CAEX,length=130038232>',\n",
        "  '##contig=<ID=CAEY,length=6181219>',\n",
        "  '##reference=file:///lustre/scratch/projects/vervetmonkey/data/reference/reference_3488/3488_indID1_codeVRC_ref_sequencer3_seqType1_filtered0_version3.fasta'],\n",
        " '_prepend_chr': False,\n",
        " '_reader': <open file '/home/GMI/hannes.svardal/vervet_project/analyses/20131121_UnifiedGenotyper_African_Sabaeus/_data/GATK_UnifiedGenotyper_Gambian_Sabaeus_CAEY_filter.vcf', mode 'r' at 0x2f5c6f0>,\n",
        " '_sample_indexes': {'6126_1436_VGA00006_GA_vs_3488': 0,\n",
        "  '6127_1441_VGA00114_GA_vs_3488': 1,\n",
        "  '6128_1443_VGA00092_GA_vs_3488': 2,\n",
        "  '6129_1444_VGA00021_GA_vs_3488': 3,\n",
        "  '6131_1457_VGA00146_GA_vs_3488': 4,\n",
        "  '6132_1458_VGA00098_GA_vs_3488': 5,\n",
        "  '6133_1459_VGA00156_GA_vs_3488': 6,\n",
        "  '6134_1462_VGA00142_GA_vs_3488': 7,\n",
        "  '6135_1469_VGA00090_GA_vs_3488': 8,\n",
        "  '6137_1484_VGA00085_GA_vs_3488': 9,\n",
        "  '6140_1506_VGA00101_GA_vs_3488': 10,\n",
        "  '6143_1523_VGA00152_GA_vs_3488': 11,\n",
        "  '6144_1524_VGA00153_GA_vs_3488': 12,\n",
        "  '6146_3424_VGA00100_GA_vs_3488': 13,\n",
        "  '6147_3425_VGA00107_GA_vs_3488': 14,\n",
        "  '6154_3432_VGA00002_GA_vs_3488': 15,\n",
        "  '6157_3435_VGA00155_GA_vs_3488': 16,\n",
        "  '6158_3436_VGA00151_GA_vs_3488': 17,\n",
        "  '6159_3437_VGA00138_GA_vs_3488': 18,\n",
        "  '6164_3442_VGA00010_GA_vs_3488': 19,\n",
        "  '6167_3445_VGA00019_GA_vs_3488': 20,\n",
        "  '6177_3497_VGA00025_GA_vs_3488': 21},\n",
        " '_separator': '\\t| +',\n",
        " '_tabix': None,\n",
        " 'alts': OrderedDict(),\n",
        " 'contigs': OrderedDict([('CAE1', Contig(id='CAE1', length=126035930)), ('CAE2', Contig(id='CAE2', length=90373283)), ('CAE3', Contig(id='CAE3', length=92142175)), ('CAE4', Contig(id='CAE4', length=91010382)), ('CAE5', Contig(id='CAE5', length=75399963)), ('CAE6', Contig(id='CAE6', length=50890351)), ('CAE7', Contig(id='CAE7', length=135778131)), ('CAE8', Contig(id='CAE8', length=139301422)), ('CAE9', Contig(id='CAE9', length=125710982)), ('CAE10', Contig(id='CAE10', length=128595539)), ('CAE11', Contig(id='CAE11', length=128539186)), ('CAE12', Contig(id='CAE12', length=108555830)), ('CAE13', Contig(id='CAE13', length=98384682)), ('CAE14', Contig(id='CAE14', length=107702431)), ('CAE15', Contig(id='CAE15', length=91754291)), ('CAE16', Contig(id='CAE16', length=75148670)), ('CAE17', Contig(id='CAE17', length=71996105)), ('CAE18', Contig(id='CAE18', length=72318688)), ('CAE19', Contig(id='CAE19', length=33263144)), ('CAE20', Contig(id='CAE20', length=130588469)), ('CAE21', Contig(id='CAE21', length=127223203)), ('CAE22', Contig(id='CAE22', length=101219884)), ('CAE23', Contig(id='CAE23', length=82825804)), ('CAE24', Contig(id='CAE24', length=84932903)), ('CAE25', Contig(id='CAE25', length=85787240)), ('CAE26', Contig(id='CAE26', length=58131712)), ('CAE27', Contig(id='CAE27', length=48547382)), ('CAE28', Contig(id='CAE28', length=21531802)), ('CAE29', Contig(id='CAE29', length=24206276)), ('CAEX', Contig(id='CAEX', length=130038232)), ('CAEY', Contig(id='CAEY', length=6181219))]),\n",
        " 'filename': '/home/GMI/hannes.svardal/vervet_project/analyses/20131121_UnifiedGenotyper_African_Sabaeus/_data/GATK_UnifiedGenotyper_Gambian_Sabaeus_CAEY_filter.vcf',\n",
        " 'filters': OrderedDict([('ExcessCov', Filter(id='ExcessCov', desc='Overlaps a user-input mask')), ('LowQual', Filter(id='LowQual', desc='Low quality')), ('PoorMapping', Filter(id='PoorMapping', desc='Overlaps a user-input mask'))]),\n",
        " 'formats': OrderedDict([('AD', Format(id='AD', num=None, type='Integer', desc='Allelic depths for the ref and alt alleles in the order listed')), ('DP', Format(id='DP', num=1, type='Integer', desc='Approximate read depth (reads with MQ=255 or with bad mates are filtered)')), ('GQ', Format(id='GQ', num=1, type='Integer', desc='Genotype Quality')), ('GT', Format(id='GT', num=1, type='String', desc='Genotype')), ('PL', Format(id='PL', num=-2, type='Integer', desc='Normalized, Phred-scaled likelihoods for genotypes as defined in the VCF specification'))]),\n",
        " 'infos': OrderedDict([('AC', Info(id='AC', num=-1, type='Integer', desc='Allele count in genotypes, for each ALT allele, in the same order as listed')), ('AF', Info(id='AF', num=-1, type='Float', desc='Allele Frequency, for each ALT allele, in the same order as listed')), ('AN', Info(id='AN', num=1, type='Integer', desc='Total number of alleles in called genotypes')), ('BaseQRankSum', Info(id='BaseQRankSum', num=1, type='Float', desc='Z-score from Wilcoxon rank sum test of Alt Vs. Ref base qualities')), ('DP', Info(id='DP', num=1, type='Integer', desc='Approximate read depth; some reads may have been filtered')), ('DS', Info(id='DS', num=0, type='Flag', desc='Were any of the samples downsampled?')), ('Dels', Info(id='Dels', num=1, type='Float', desc='Fraction of Reads Containing Spanning Deletions')), ('FS', Info(id='FS', num=1, type='Float', desc=\"Phred-scaled p-value using Fisher's exact test to detect strand bias\")), ('HaplotypeScore', Info(id='HaplotypeScore', num=1, type='Float', desc='Consistency of the site with at most two segregating haplotypes')), ('InbreedingCoeff', Info(id='InbreedingCoeff', num=1, type='Float', desc='Inbreeding coefficient as estimated from the genotype likelihoods per-sample when compared against the Hardy-Weinberg expectation')), ('MLEAC', Info(id='MLEAC', num=-1, type='Integer', desc='Maximum likelihood expectation (MLE) for the allele counts (not necessarily the same as the AC), for each ALT allele, in the same order as listed')), ('MLEAF', Info(id='MLEAF', num=-1, type='Float', desc='Maximum likelihood expectation (MLE) for the allele frequency (not necessarily the same as the AF), for each ALT allele, in the same order as listed')), ('MQ', Info(id='MQ', num=1, type='Float', desc='RMS Mapping Quality')), ('MQ0', Info(id='MQ0', num=1, type='Integer', desc='Total Mapping Quality Zero Reads')), ('MQRankSum', Info(id='MQRankSum', num=1, type='Float', desc='Z-score From Wilcoxon rank sum test of Alt vs. Ref read mapping qualities')), ('QD', Info(id='QD', num=1, type='Float', desc='Variant Confidence/Quality by Depth')), ('RPA', Info(id='RPA', num=None, type='Integer', desc='Number of times tandem repeat unit is repeated, for each allele (including reference)')), ('RU', Info(id='RU', num=1, type='String', desc='Tandem repeat unit (bases)')), ('ReadPosRankSum', Info(id='ReadPosRankSum', num=1, type='Float', desc='Z-score from Wilcoxon rank sum test of Alt vs. Ref read position bias')), ('STR', Info(id='STR', num=0, type='Flag', desc='Variant is a short tandem repeat'))]),\n",
        " 'metadata': OrderedDict([('fileformat', 'VCFv4.1'), ('UnifiedGenotyper', ['\"analysis_type=UnifiedGenotyper input_file=[data/db/individual_alignment/6154_3432_VGA00002_GA_vs_3488_by_method6_realigned1_reduced0_p4383_m227.bam, data/db/individual_alignment/6126_1436_VGA00006_GA_vs_3488_by_method6_realigned1_reduced0_p4355_m227.bam, data/db/individual_alignment/6164_3442_VGA00010_GA_vs_3488_by_method6_realigned1_reduced0_p4393_m227.bam, data/db/individual_alignment/6167_3445_VGA00019_GA_vs_3488_by_method6_realigned1_reduced0_p4396_m227.bam, data/db/individual_alignment/6129_1444_VGA00021_GA_vs_3488_by_method6_realigned1_reduced0_p4358_m227.bam, data/db/individual_alignment/6177_3497_VGA00025_GA_vs_3488_by_method6_realigned1_reduced0_p4406_m227.bam, data/db/individual_alignment/6137_1484_VGA00085_GA_vs_3488_by_method6_realigned1_reduced0_p4366_m227.bam, data/db/individual_alignment/6135_1469_VGA00090_GA_vs_3488_by_method6_realigned1_reduced0_p4364_m227.bam, data/db/individual_alignment/6128_1443_VGA00092_GA_vs_3488_by_method6_realigned1_reduced0_p4357_m227.bam, data/db/individual_alignment/6132_1458_VGA00098_GA_vs_3488_by_method6_realigned1_reduced0_p4361_m227.bam, data/db/individual_alignment/6146_3424_VGA00100_GA_vs_3488_by_method6_realigned1_reduced0_p4375_m227.bam, data/db/individual_alignment/6140_1506_VGA00101_GA_vs_3488_by_method6_realigned1_reduced0_p4369_m227.bam, data/db/individual_alignment/6147_3425_VGA00107_GA_vs_3488_by_method6_realigned1_reduced0_p4376_m227.bam, data/db/individual_alignment/6127_1441_VGA00114_GA_vs_3488_by_method6_realigned1_reduced0_p4356_m227.bam, data/db/individual_alignment/6159_3437_VGA00138_GA_vs_3488_by_method6_realigned1_reduced0_p4388_m227.bam, data/db/individual_alignment/6134_1462_VGA00142_GA_vs_3488_by_method6_realigned1_reduced0_p4363_m227.bam, data/db/individual_alignment/6131_1457_VGA00146_GA_vs_3488_by_method6_realigned1_reduced0_p4360_m227.bam, data/db/individual_alignment/6158_3436_VGA00151_GA_vs_3488_by_method6_realigned1_reduced0_p4387_m227.bam, data/db/individual_alignment/6143_1523_VGA00152_GA_vs_3488_by_method6_realigned1_reduced0_p4372_m227.bam, data/db/individual_alignment/6144_1524_VGA00153_GA_vs_3488_by_method6_realigned1_reduced0_p4373_m227.bam, data/db/individual_alignment/6157_3435_VGA00155_GA_vs_3488_by_method6_realigned1_reduced0_p4386_m227.bam, data/db/individual_alignment/6133_1459_VGA00156_GA_vs_3488_by_method6_realigned1_reduced0_p4362_m227.bam] read_buffer_size=null phone_home=STANDARD gatk_key=null tag=NA read_filter=[] intervals=[CAEY] excludeIntervals=null interval_set_rule=UNION interval_merging=ALL interval_padding=0 reference_sequence=data/reference/reference_3488/3488_indID1_codeVRC_ref_sequencer3_seqType1_filtered0_version3.fasta nonDeterministicRandomSeed=false disableDithering=false maxRuntime=-1 maxRuntimeUnits=MINUTES downsampling_type=BY_SAMPLE downsample_to_fraction=null downsample_to_coverage=250 baq=OFF baqGapOpenPenalty=40.0 fix_misencoded_quality_scores=false allow_potentially_misencoded_quality_scores=false performanceLog=null useOriginalQualities=false BQSR=null quantize_quals=0 disable_indel_quals=false emit_original_quals=false preserve_qscores_less_than=6 globalQScorePrior=-1.0 allow_bqsr_on_reduced_bams_despite_repeated_warnings=false defaultBaseQualities=-1 validation_strictness=SILENT remove_program_records=false keep_program_records=false unsafe=null disable_auto_index_creation_and_locking_when_reading_rods=false num_threads=4 num_cpu_threads_per_data_thread=1 num_io_threads=0 monitorThreadEfficiency=false num_bam_file_handles=null read_group_black_list=null pedigree=[] pedigreeString=[] pedigreeValidationType=STRICT allow_intervals_with_unindexed_bam=false generateShadowBCF=false logging_level=INFO log_to_file=null help=false version=false genotype_likelihoods_model=SNP pcr_error_rate=1.0E-4 computeSLOD=false annotateNDA=false pair_hmm_implementation=ORIGINAL min_base_quality_score=17 max_deletion_fraction=0.05 min_indel_count_for_genotyping=5 min_indel_fraction_per_sample=0.25 indelGapContinuationPenalty=10 indelGapOpenPenalty=45 indelHaplotypeSize=80 indelDebug=false ignoreSNPAlleles=false allReadsSP=false ignoreLaneInfo=false reference_sample_calls=(RodBinding name= source=UNBOUND) reference_sample_name=null sample_ploidy=2 min_quality_score=1 max_quality_score=40 site_quality_prior=20 min_power_threshold_for_calling=0.95 min_reference_depth=100 exclude_filtered_reference_sites=false heterozygosity=0.001 indel_heterozygosity=1.25E-4 genotyping_mode=DISCOVERY output_mode=EMIT_VARIANTS_ONLY standard_min_confidence_threshold_for_calling=50.0 standard_min_confidence_threshold_for_emitting=10.0 alleles=(RodBinding name= source=UNBOUND) max_alternate_alleles=6 input_prior=[] contamination_fraction_to_filter=0.05 contamination_fraction_per_sample_file=null p_nonref_model=EXACT_INDEPENDENT exactcallslog=null dbsnp=(RodBinding name= source=UNBOUND) comp=[] out=org.broadinstitute.sting.gatk.io.stubs.VariantContextWriterStub no_cmdline_in_header=org.broadinstitute.sting.gatk.io.stubs.VariantContextWriterStub sites_only=org.broadinstitute.sting.gatk.io.stubs.VariantContextWriterStub bcf=org.broadinstitute.sting.gatk.io.stubs.VariantContextWriterStub debug_file=null metrics_file=null annotation=[] excludeAnnotation=[] filter_mismatching_base_and_quals=false filter_bases_not_stored=false\"']), ('VariantFiltration', ['\"analysis_type=VariantFiltration input_file=[] read_buffer_size=null phone_home=STANDARD gatk_key=null tag=NA read_filter=[] intervals=null excludeIntervals=null interval_set_rule=UNION interval_merging=ALL interval_padding=0 reference_sequence=data/reference/reference_3488/3488_indID1_codeVRC_ref_sequencer3_seqType1_filtered0_version3.fasta nonDeterministicRandomSeed=false disableDithering=false maxRuntime=-1 maxRuntimeUnits=MINUTES downsampling_type=BY_SAMPLE downsample_to_fraction=null downsample_to_coverage=1000 baq=OFF baqGapOpenPenalty=40.0 fix_misencoded_quality_scores=false allow_potentially_misencoded_quality_scores=false performanceLog=null useOriginalQualities=false BQSR=null quantize_quals=0 disable_indel_quals=false emit_original_quals=false preserve_qscores_less_than=6 globalQScorePrior=-1.0 allow_bqsr_on_reduced_bams_despite_repeated_warnings=false defaultBaseQualities=-1 validation_strictness=SILENT remove_program_records=false keep_program_records=false unsafe=null disable_auto_index_creation_and_locking_when_reading_rods=false num_threads=1 num_cpu_threads_per_data_thread=1 num_io_threads=0 monitorThreadEfficiency=false num_bam_file_handles=null read_group_black_list=null pedigree=[] pedigreeString=[] pedigreeValidationType=STRICT allow_intervals_with_unindexed_bam=false generateShadowBCF=false logging_level=INFO log_to_file=null help=false version=false variant=(RodBinding name=variant source=analyses/20131121_UnifiedGenotyper_African_Sabaeus/_data/GATK_UnifiedGenotyper_Gambian_Sabaeus_CAEY.vcf) mask=(RodBinding name=mask source=analyses/20131129_filtering_masks/_data/excessive_coverage_regions_CAEY.bed) out=org.broadinstitute.sting.gatk.io.stubs.VariantContextWriterStub no_cmdline_in_header=org.broadinstitute.sting.gatk.io.stubs.VariantContextWriterStub sites_only=org.broadinstitute.sting.gatk.io.stubs.VariantContextWriterStub bcf=org.broadinstitute.sting.gatk.io.stubs.VariantContextWriterStub filterExpression=[] filterName=[] genotypeFilterExpression=[] genotypeFilterName=[] clusterSize=3 clusterWindowSize=0 maskExtension=0 maskName=ExcessCov filterNotInMask=false missingValuesInExpressionsShouldEvaluateAsFailing=false invalidatePreviousFilters=false filter_mismatching_base_and_quals=false filter_bases_not_stored=false\"']), ('reference', 'file:///lustre/scratch/projects/vervetmonkey/data/reference/reference_3488/3488_indID1_codeVRC_ref_sequencer3_seqType1_filtered0_version3.fasta')]),\n",
        " 'reader': <generator object <genexpr> at 0x2fe7af0>,\n",
        " 'samples': ['6126_1436_VGA00006_GA_vs_3488',\n",
        "  '6127_1441_VGA00114_GA_vs_3488',\n",
        "  '6128_1443_VGA00092_GA_vs_3488',\n",
        "  '6129_1444_VGA00021_GA_vs_3488',\n",
        "  '6131_1457_VGA00146_GA_vs_3488',\n",
        "  '6132_1458_VGA00098_GA_vs_3488',\n",
        "  '6133_1459_VGA00156_GA_vs_3488',\n",
        "  '6134_1462_VGA00142_GA_vs_3488',\n",
        "  '6135_1469_VGA00090_GA_vs_3488',\n",
        "  '6137_1484_VGA00085_GA_vs_3488',\n",
        "  '6140_1506_VGA00101_GA_vs_3488',\n",
        "  '6143_1523_VGA00152_GA_vs_3488',\n",
        "  '6144_1524_VGA00153_GA_vs_3488',\n",
        "  '6146_3424_VGA00100_GA_vs_3488',\n",
        "  '6147_3425_VGA00107_GA_vs_3488',\n",
        "  '6154_3432_VGA00002_GA_vs_3488',\n",
        "  '6157_3435_VGA00155_GA_vs_3488',\n",
        "  '6158_3436_VGA00151_GA_vs_3488',\n",
        "  '6159_3437_VGA00138_GA_vs_3488',\n",
        "  '6164_3442_VGA00010_GA_vs_3488',\n",
        "  '6167_3445_VGA00019_GA_vs_3488',\n",
        "  '6177_3497_VGA00025_GA_vs_3488']}"
       ]
      }
     ],
     "prompt_number": 8
    },
    {
     "cell_type": "code",
     "collapsed": false,
     "input": [
      "genoytype_dict={None:[None,None],'0/0':[0,0],'0/1':[0,1],'1/0':[1,0],'1/1':[1,1]}"
     ],
     "language": "python",
     "metadata": {},
     "outputs": [],
     "prompt_number": 9
    },
    {
     "cell_type": "code",
     "collapsed": false,
     "input": [
      "def vcf_to_tsv(vcf_fname,tsv_basefname):\n",
      "    \"\"\"\n",
      "    parse vcf and write one tsv per chromosome\n",
      "    tsv_basefname ... /location/for/output/name\n",
      "    ttsv_basefname is complemented by _chrom.tsv\n",
      "    \"\"\"\n",
      "    tsv_basename=os.path.expanduser(tsv_basefname)\n",
      "    hvb.try_make_dirs(os.path.dirname(tsv_basename))\n",
      "    \n",
      "    genoytype_dict={None:[None,None],'0/0':[0,0],'0/1':[0,1],'1/0':[1,0],'1/1':[1,1],'0/2':[0,2],'2/2':[2,2],'1/2':[1,2]}\n",
      "    \n",
      "    def make_df_and_save(genotypes,snp_metadata):\n",
      "        data_df=pd.DataFrame(genotypes,columns=cols,index=np.array(snp_metadata)[:,1])\n",
      "        meta_df=pd.DataFrame(snp_metadata,columns=['CHROM','POS','ID','REF','ALT'],index=np.array(snp_metadata)[:,1])\n",
      "        hvb.save_genotype(data_df,tsv_basefname+\"_\"+meta_df['CHROM'].iloc[0]+'.tsv')\n",
      "        meta_df.to_csv(tsv_basefname+\"_\"+meta_df['CHROM'].iloc[0]+'_metadata'+'.tsv',sep='\\t')\n",
      "        return data_df, meta_df             \n",
      "    \n",
      "    with open(vcf_fname, 'r') as f:\n",
      "        vcf_reader = vcf.Reader(f)\n",
      "        chrom_last=None\n",
      "        \n",
      "        ucla_ids=[i for k in vcf_reader.samples for i in [k]*2]\n",
      "        strands=[0,1]*len(vcf_reader.samples)\n",
      "        tuples=list(zip(ucla_ids,strands))\n",
      "        cols=pd.MultiIndex.from_tuples(tuples)\n",
      "        \n",
      "        \n",
      "        genotypes=[]\n",
      "        snp_metadata=[]\n",
      "        \n",
      "        for i,record in enumerate(vcf_reader):\n",
      "            if not record.FILTER: # only those that passed the filters\n",
      "                chrom_current = record.CHROM\n",
      "                if chrom_last is not None and chrom_current!=chrom_last:\n",
      "                    #return genotypes\n",
      "                    #save the stuff \n",
      "                    make_df_and_save(genotypes,snp_metadata)\n",
      "                    #and re-initialize\n",
      "                    genotypes=[]\n",
      "                    snp_metadata=[]\n",
      "                genotypes_row=[k for call in record.samples for k in genoytype_dict[call['GT']]]\n",
      "                genotypes.append(genotypes_row)\n",
      "                snp_metadata_row=[record.CHROM,record.POS,record.ID,record.REF,record.ALT]\n",
      "                snp_metadata.append(snp_metadata_row)\n",
      "                chrom_last=chrom_current\n",
      "            #if i>1000: break\n",
      "        data_df,meta_df=make_df_and_save(genotypes,snp_metadata)\n",
      "    return data_df,meta_df      \n",
      "    "
     ],
     "language": "python",
     "metadata": {},
     "outputs": [],
     "prompt_number": 29
    },
    {
     "cell_type": "code",
     "collapsed": false,
     "input": [
      "vcf_fname"
     ],
     "language": "python",
     "metadata": {},
     "outputs": [
      {
       "output_type": "pyout",
       "prompt_number": 30,
       "text": [
        "'/home/GMI/hannes.svardal/vervet_project/analyses/20131121_UnifiedGenotyper_African_Sabaeus/_data/GATK_UnifiedGenotyper_Gambian_Sabaeus_CAEY_filter.vcf'"
       ]
      }
     ],
     "prompt_number": 30
    },
    {
     "cell_type": "code",
     "collapsed": false,
     "input": [
      "data_df, meta_df = vcf_to_tsv(vcf_fname,\"test/test1\")"
     ],
     "language": "python",
     "metadata": {},
     "outputs": [],
     "prompt_number": 31
    },
    {
     "cell_type": "code",
     "collapsed": false,
     "input": [
      "pd.__version__"
     ],
     "language": "python",
     "metadata": {},
     "outputs": [
      {
       "output_type": "pyout",
       "prompt_number": 35,
       "text": [
        "'0.12.0'"
       ]
      }
     ],
     "prompt_number": 35
    },
    {
     "cell_type": "code",
     "collapsed": false,
     "input": [
      "data_df.iloc[:10,:10]"
     ],
     "language": "python",
     "metadata": {},
     "outputs": [
      {
       "html": [
        "<div style=\"max-height:1000px;max-width:1500px;overflow:auto;\">\n",
        "<table border=\"1\" class=\"dataframe\">\n",
        "  <thead>\n",
        "    <tr>\n",
        "      <th></th>\n",
        "      <th colspan=\"2\" halign=\"left\">6126_1436_VGA00006_GA_vs_3488</th>\n",
        "      <th colspan=\"2\" halign=\"left\">6127_1441_VGA00114_GA_vs_3488</th>\n",
        "      <th colspan=\"2\" halign=\"left\">6128_1443_VGA00092_GA_vs_3488</th>\n",
        "      <th colspan=\"2\" halign=\"left\">6129_1444_VGA00021_GA_vs_3488</th>\n",
        "      <th colspan=\"2\" halign=\"left\">6131_1457_VGA00146_GA_vs_3488</th>\n",
        "    </tr>\n",
        "    <tr>\n",
        "      <th></th>\n",
        "      <th>0</th>\n",
        "      <th>1</th>\n",
        "      <th>0</th>\n",
        "      <th>1</th>\n",
        "      <th>0</th>\n",
        "      <th>1</th>\n",
        "      <th>0</th>\n",
        "      <th>1</th>\n",
        "      <th>0</th>\n",
        "      <th>1</th>\n",
        "    </tr>\n",
        "  </thead>\n",
        "  <tbody>\n",
        "    <tr>\n",
        "      <th>14905</th>\n",
        "      <td>NaN</td>\n",
        "      <td>NaN</td>\n",
        "      <td>NaN</td>\n",
        "      <td>NaN</td>\n",
        "      <td>  0</td>\n",
        "      <td>  0</td>\n",
        "      <td>NaN</td>\n",
        "      <td>NaN</td>\n",
        "      <td>NaN</td>\n",
        "      <td>NaN</td>\n",
        "    </tr>\n",
        "    <tr>\n",
        "      <th>22567</th>\n",
        "      <td>NaN</td>\n",
        "      <td>NaN</td>\n",
        "      <td>NaN</td>\n",
        "      <td>NaN</td>\n",
        "      <td>  0</td>\n",
        "      <td>  0</td>\n",
        "      <td>  1</td>\n",
        "      <td>  1</td>\n",
        "      <td>NaN</td>\n",
        "      <td>NaN</td>\n",
        "    </tr>\n",
        "    <tr>\n",
        "      <th>30131</th>\n",
        "      <td>NaN</td>\n",
        "      <td>NaN</td>\n",
        "      <td>NaN</td>\n",
        "      <td>NaN</td>\n",
        "      <td>NaN</td>\n",
        "      <td>NaN</td>\n",
        "      <td>NaN</td>\n",
        "      <td>NaN</td>\n",
        "      <td>NaN</td>\n",
        "      <td>NaN</td>\n",
        "    </tr>\n",
        "    <tr>\n",
        "      <th>32625</th>\n",
        "      <td>NaN</td>\n",
        "      <td>NaN</td>\n",
        "      <td>NaN</td>\n",
        "      <td>NaN</td>\n",
        "      <td>  1</td>\n",
        "      <td>  1</td>\n",
        "      <td>  1</td>\n",
        "      <td>  1</td>\n",
        "      <td>NaN</td>\n",
        "      <td>NaN</td>\n",
        "    </tr>\n",
        "    <tr>\n",
        "      <th>32635</th>\n",
        "      <td>NaN</td>\n",
        "      <td>NaN</td>\n",
        "      <td>NaN</td>\n",
        "      <td>NaN</td>\n",
        "      <td>  1</td>\n",
        "      <td>  1</td>\n",
        "      <td>  1</td>\n",
        "      <td>  1</td>\n",
        "      <td>NaN</td>\n",
        "      <td>NaN</td>\n",
        "    </tr>\n",
        "    <tr>\n",
        "      <th>32641</th>\n",
        "      <td>NaN</td>\n",
        "      <td>NaN</td>\n",
        "      <td>NaN</td>\n",
        "      <td>NaN</td>\n",
        "      <td>  1</td>\n",
        "      <td>  1</td>\n",
        "      <td>  1</td>\n",
        "      <td>  1</td>\n",
        "      <td>NaN</td>\n",
        "      <td>NaN</td>\n",
        "    </tr>\n",
        "    <tr>\n",
        "      <th>32644</th>\n",
        "      <td>NaN</td>\n",
        "      <td>NaN</td>\n",
        "      <td>NaN</td>\n",
        "      <td>NaN</td>\n",
        "      <td>  1</td>\n",
        "      <td>  1</td>\n",
        "      <td>  1</td>\n",
        "      <td>  1</td>\n",
        "      <td>NaN</td>\n",
        "      <td>NaN</td>\n",
        "    </tr>\n",
        "    <tr>\n",
        "      <th>32657</th>\n",
        "      <td>NaN</td>\n",
        "      <td>NaN</td>\n",
        "      <td>NaN</td>\n",
        "      <td>NaN</td>\n",
        "      <td>  1</td>\n",
        "      <td>  1</td>\n",
        "      <td>  1</td>\n",
        "      <td>  1</td>\n",
        "      <td>NaN</td>\n",
        "      <td>NaN</td>\n",
        "    </tr>\n",
        "    <tr>\n",
        "      <th>32870</th>\n",
        "      <td>NaN</td>\n",
        "      <td>NaN</td>\n",
        "      <td>NaN</td>\n",
        "      <td>NaN</td>\n",
        "      <td>  1</td>\n",
        "      <td>  1</td>\n",
        "      <td>NaN</td>\n",
        "      <td>NaN</td>\n",
        "      <td>NaN</td>\n",
        "      <td>NaN</td>\n",
        "    </tr>\n",
        "    <tr>\n",
        "      <th>32872</th>\n",
        "      <td>NaN</td>\n",
        "      <td>NaN</td>\n",
        "      <td>NaN</td>\n",
        "      <td>NaN</td>\n",
        "      <td>NaN</td>\n",
        "      <td>NaN</td>\n",
        "      <td>NaN</td>\n",
        "      <td>NaN</td>\n",
        "      <td>NaN</td>\n",
        "      <td>NaN</td>\n",
        "    </tr>\n",
        "  </tbody>\n",
        "</table>\n",
        "</div>"
       ],
       "output_type": "pyout",
       "prompt_number": 32,
       "text": [
        "       6126_1436_VGA00006_GA_vs_3488      6127_1441_VGA00114_GA_vs_3488      \\\n",
        "                                   0   1                              0   1   \n",
        "14905                            NaN NaN                            NaN NaN   \n",
        "22567                            NaN NaN                            NaN NaN   \n",
        "30131                            NaN NaN                            NaN NaN   \n",
        "32625                            NaN NaN                            NaN NaN   \n",
        "32635                            NaN NaN                            NaN NaN   \n",
        "32641                            NaN NaN                            NaN NaN   \n",
        "32644                            NaN NaN                            NaN NaN   \n",
        "32657                            NaN NaN                            NaN NaN   \n",
        "32870                            NaN NaN                            NaN NaN   \n",
        "32872                            NaN NaN                            NaN NaN   \n",
        "\n",
        "       6128_1443_VGA00092_GA_vs_3488      6129_1444_VGA00021_GA_vs_3488      \\\n",
        "                                   0   1                              0   1   \n",
        "14905                              0   0                            NaN NaN   \n",
        "22567                              0   0                              1   1   \n",
        "30131                            NaN NaN                            NaN NaN   \n",
        "32625                              1   1                              1   1   \n",
        "32635                              1   1                              1   1   \n",
        "32641                              1   1                              1   1   \n",
        "32644                              1   1                              1   1   \n",
        "32657                              1   1                              1   1   \n",
        "32870                              1   1                            NaN NaN   \n",
        "32872                            NaN NaN                            NaN NaN   \n",
        "\n",
        "       6131_1457_VGA00146_GA_vs_3488      \n",
        "                                   0   1  \n",
        "14905                            NaN NaN  \n",
        "22567                            NaN NaN  \n",
        "30131                            NaN NaN  \n",
        "32625                            NaN NaN  \n",
        "32635                            NaN NaN  \n",
        "32641                            NaN NaN  \n",
        "32644                            NaN NaN  \n",
        "32657                            NaN NaN  \n",
        "32870                            NaN NaN  \n",
        "32872                            NaN NaN  "
       ]
      }
     ],
     "prompt_number": 32
    },
    {
     "cell_type": "code",
     "collapsed": false,
     "input": [
      "data_df.iloc[:10,:10].columns"
     ],
     "language": "python",
     "metadata": {},
     "outputs": [
      {
       "output_type": "pyout",
       "prompt_number": 34,
       "text": [
        "MultiIndex\n",
        "[(u'6126_1436_VGA00006_GA_vs_3488', 0), (u'6126_1436_VGA00006_GA_vs_3488', 1), (u'6127_1441_VGA00114_GA_vs_3488', 0), (u'6127_1441_VGA00114_GA_vs_3488', 1), (u'6128_1443_VGA00092_GA_vs_3488', 0), (u'6128_1443_VGA00092_GA_vs_3488', 1), (u'6129_1444_VGA00021_GA_vs_3488', 0), (u'6129_1444_VGA00021_GA_vs_3488', 1), (u'6131_1457_VGA00146_GA_vs_3488', 0), (u'6131_1457_VGA00146_GA_vs_3488', 1)]"
       ]
      }
     ],
     "prompt_number": 34
    },
    {
     "cell_type": "code",
     "collapsed": false,
     "input": [
      "data_df.to_csv('test1.tsv',sep='\\t',tupleize_cols=False)"
     ],
     "language": "python",
     "metadata": {},
     "outputs": [],
     "prompt_number": 36
    },
    {
     "cell_type": "code",
     "collapsed": false,
     "input": [
      "data_df2=pd.read_csv('test1.tsv',sep='\\t',index_col=0,header=[0,1],tupleize_cols=False)"
     ],
     "language": "python",
     "metadata": {},
     "outputs": [],
     "prompt_number": 44
    },
    {
     "cell_type": "code",
     "collapsed": false,
     "input": [
      "data_df2.iloc[:10,:10]"
     ],
     "language": "python",
     "metadata": {},
     "outputs": [
      {
       "html": [
        "<div style=\"max-height:1000px;max-width:1500px;overflow:auto;\">\n",
        "<table border=\"1\" class=\"dataframe\">\n",
        "  <thead>\n",
        "    <tr>\n",
        "      <th></th>\n",
        "      <th colspan=\"2\" halign=\"left\">6126_1436_VGA00006_GA_vs_3488</th>\n",
        "      <th colspan=\"2\" halign=\"left\">6127_1441_VGA00114_GA_vs_3488</th>\n",
        "      <th colspan=\"2\" halign=\"left\">6128_1443_VGA00092_GA_vs_3488</th>\n",
        "      <th colspan=\"2\" halign=\"left\">6129_1444_VGA00021_GA_vs_3488</th>\n",
        "      <th colspan=\"2\" halign=\"left\">6131_1457_VGA00146_GA_vs_3488</th>\n",
        "    </tr>\n",
        "    <tr>\n",
        "      <th></th>\n",
        "      <th>0</th>\n",
        "      <th>1</th>\n",
        "      <th>0</th>\n",
        "      <th>1</th>\n",
        "      <th>0</th>\n",
        "      <th>1</th>\n",
        "      <th>0</th>\n",
        "      <th>1</th>\n",
        "      <th>0</th>\n",
        "      <th>1</th>\n",
        "    </tr>\n",
        "  </thead>\n",
        "  <tbody>\n",
        "    <tr>\n",
        "      <th>14905</th>\n",
        "      <td>NaN</td>\n",
        "      <td>NaN</td>\n",
        "      <td>NaN</td>\n",
        "      <td>NaN</td>\n",
        "      <td>  0</td>\n",
        "      <td>  0</td>\n",
        "      <td>NaN</td>\n",
        "      <td>NaN</td>\n",
        "      <td>NaN</td>\n",
        "      <td>NaN</td>\n",
        "    </tr>\n",
        "    <tr>\n",
        "      <th>22567</th>\n",
        "      <td>NaN</td>\n",
        "      <td>NaN</td>\n",
        "      <td>NaN</td>\n",
        "      <td>NaN</td>\n",
        "      <td>  0</td>\n",
        "      <td>  0</td>\n",
        "      <td>  1</td>\n",
        "      <td>  1</td>\n",
        "      <td>NaN</td>\n",
        "      <td>NaN</td>\n",
        "    </tr>\n",
        "    <tr>\n",
        "      <th>30131</th>\n",
        "      <td>NaN</td>\n",
        "      <td>NaN</td>\n",
        "      <td>NaN</td>\n",
        "      <td>NaN</td>\n",
        "      <td>NaN</td>\n",
        "      <td>NaN</td>\n",
        "      <td>NaN</td>\n",
        "      <td>NaN</td>\n",
        "      <td>NaN</td>\n",
        "      <td>NaN</td>\n",
        "    </tr>\n",
        "    <tr>\n",
        "      <th>32625</th>\n",
        "      <td>NaN</td>\n",
        "      <td>NaN</td>\n",
        "      <td>NaN</td>\n",
        "      <td>NaN</td>\n",
        "      <td>  1</td>\n",
        "      <td>  1</td>\n",
        "      <td>  1</td>\n",
        "      <td>  1</td>\n",
        "      <td>NaN</td>\n",
        "      <td>NaN</td>\n",
        "    </tr>\n",
        "    <tr>\n",
        "      <th>32635</th>\n",
        "      <td>NaN</td>\n",
        "      <td>NaN</td>\n",
        "      <td>NaN</td>\n",
        "      <td>NaN</td>\n",
        "      <td>  1</td>\n",
        "      <td>  1</td>\n",
        "      <td>  1</td>\n",
        "      <td>  1</td>\n",
        "      <td>NaN</td>\n",
        "      <td>NaN</td>\n",
        "    </tr>\n",
        "    <tr>\n",
        "      <th>32641</th>\n",
        "      <td>NaN</td>\n",
        "      <td>NaN</td>\n",
        "      <td>NaN</td>\n",
        "      <td>NaN</td>\n",
        "      <td>  1</td>\n",
        "      <td>  1</td>\n",
        "      <td>  1</td>\n",
        "      <td>  1</td>\n",
        "      <td>NaN</td>\n",
        "      <td>NaN</td>\n",
        "    </tr>\n",
        "    <tr>\n",
        "      <th>32644</th>\n",
        "      <td>NaN</td>\n",
        "      <td>NaN</td>\n",
        "      <td>NaN</td>\n",
        "      <td>NaN</td>\n",
        "      <td>  1</td>\n",
        "      <td>  1</td>\n",
        "      <td>  1</td>\n",
        "      <td>  1</td>\n",
        "      <td>NaN</td>\n",
        "      <td>NaN</td>\n",
        "    </tr>\n",
        "    <tr>\n",
        "      <th>32657</th>\n",
        "      <td>NaN</td>\n",
        "      <td>NaN</td>\n",
        "      <td>NaN</td>\n",
        "      <td>NaN</td>\n",
        "      <td>  1</td>\n",
        "      <td>  1</td>\n",
        "      <td>  1</td>\n",
        "      <td>  1</td>\n",
        "      <td>NaN</td>\n",
        "      <td>NaN</td>\n",
        "    </tr>\n",
        "    <tr>\n",
        "      <th>32870</th>\n",
        "      <td>NaN</td>\n",
        "      <td>NaN</td>\n",
        "      <td>NaN</td>\n",
        "      <td>NaN</td>\n",
        "      <td>  1</td>\n",
        "      <td>  1</td>\n",
        "      <td>NaN</td>\n",
        "      <td>NaN</td>\n",
        "      <td>NaN</td>\n",
        "      <td>NaN</td>\n",
        "    </tr>\n",
        "    <tr>\n",
        "      <th>32872</th>\n",
        "      <td>NaN</td>\n",
        "      <td>NaN</td>\n",
        "      <td>NaN</td>\n",
        "      <td>NaN</td>\n",
        "      <td>NaN</td>\n",
        "      <td>NaN</td>\n",
        "      <td>NaN</td>\n",
        "      <td>NaN</td>\n",
        "      <td>NaN</td>\n",
        "      <td>NaN</td>\n",
        "    </tr>\n",
        "  </tbody>\n",
        "</table>\n",
        "</div>"
       ],
       "output_type": "pyout",
       "prompt_number": 45,
       "text": [
        "       6126_1436_VGA00006_GA_vs_3488      6127_1441_VGA00114_GA_vs_3488      \\\n",
        "                                   0   1                              0   1   \n",
        "14905                            NaN NaN                            NaN NaN   \n",
        "22567                            NaN NaN                            NaN NaN   \n",
        "30131                            NaN NaN                            NaN NaN   \n",
        "32625                            NaN NaN                            NaN NaN   \n",
        "32635                            NaN NaN                            NaN NaN   \n",
        "32641                            NaN NaN                            NaN NaN   \n",
        "32644                            NaN NaN                            NaN NaN   \n",
        "32657                            NaN NaN                            NaN NaN   \n",
        "32870                            NaN NaN                            NaN NaN   \n",
        "32872                            NaN NaN                            NaN NaN   \n",
        "\n",
        "       6128_1443_VGA00092_GA_vs_3488      6129_1444_VGA00021_GA_vs_3488      \\\n",
        "                                   0   1                              0   1   \n",
        "14905                              0   0                            NaN NaN   \n",
        "22567                              0   0                              1   1   \n",
        "30131                            NaN NaN                            NaN NaN   \n",
        "32625                              1   1                              1   1   \n",
        "32635                              1   1                              1   1   \n",
        "32641                              1   1                              1   1   \n",
        "32644                              1   1                              1   1   \n",
        "32657                              1   1                              1   1   \n",
        "32870                              1   1                            NaN NaN   \n",
        "32872                            NaN NaN                            NaN NaN   \n",
        "\n",
        "       6131_1457_VGA00146_GA_vs_3488      \n",
        "                                   0   1  \n",
        "14905                            NaN NaN  \n",
        "22567                            NaN NaN  \n",
        "30131                            NaN NaN  \n",
        "32625                            NaN NaN  \n",
        "32635                            NaN NaN  \n",
        "32641                            NaN NaN  \n",
        "32644                            NaN NaN  \n",
        "32657                            NaN NaN  \n",
        "32870                            NaN NaN  \n",
        "32872                            NaN NaN  "
       ]
      }
     ],
     "prompt_number": 45
    },
    {
     "cell_type": "code",
     "collapsed": false,
     "input": [
      "|print genotypes[0],genotypes[1],genotypes[1000]"
     ],
     "language": "python",
     "metadata": {},
     "outputs": [
      {
       "output_type": "stream",
       "stream": "stdout",
       "text": [
        "[None, None, None, None, 0, 0, None, None, 0, 0, 0, 0, None, None, 0, 0, None, None, None, None, 0, 0, 0, 0, 1, 0, None, None, None, None, None, None, None, None, 0, 0, 0, 0, 0, 0, None, None, 0, 0] [None, None, None, None, 0, 0, None, None, 0, 0, 0, 0, None, None, 0, 0, None, None, None, None, 0, 0, 0, 0, 1, 0, None, None, None, None, None, None, None, None, 0, 0, 0, 0, 0, 0, None, None, 0, 0] [None, None, None, None, 0, 0, None, None, 0, 0, 0, 0, None, None, 0, 0, None, None, None, None, 0, 0, 0, 0, 1, 0, None, None, None, None, None, None, None, None, 0, 0, 0, 0, 0, 0, None, None, 0, 0]\n"
       ]
      }
     ],
     "prompt_number": 134
    },
    {
     "cell_type": "code",
     "collapsed": false,
     "input": [
      "hvb.try_make_dirs(\"/home/GMI/hannes.svardal/vervet_data/data/hs_format/cortex/test\")"
     ],
     "language": "python",
     "metadata": {},
     "outputs": [],
     "prompt_number": 111
    },
    {
     "cell_type": "code",
     "collapsed": false,
     "input": [
      "os.path.dirname(\"/home/GMI/hannes.svardal/vervet_data/data/hs_format/cortex\")"
     ],
     "language": "python",
     "metadata": {},
     "outputs": [
      {
       "output_type": "pyout",
       "prompt_number": 109,
       "text": [
        "'/home/GMI/hannes.svardal/vervet_data/data/hs_format'"
       ]
      }
     ],
     "prompt_number": 109
    },
    {
     "cell_type": "code",
     "collapsed": false,
     "input": [
      "dat.iloc[:10,:10]"
     ],
     "language": "python",
     "metadata": {},
     "outputs": [
      {
       "html": [
        "<div style=\"max-height:1000px;max-width:1500px;overflow:auto;\">\n",
        "<table border=\"1\" class=\"dataframe\">\n",
        "  <thead>\n",
        "    <tr>\n",
        "      <th></th>\n",
        "      <th colspan=\"2\" halign=\"left\">VGA00002</th>\n",
        "      <th colspan=\"2\" halign=\"left\">VGA00006</th>\n",
        "      <th colspan=\"2\" halign=\"left\">VGA00010</th>\n",
        "      <th colspan=\"2\" halign=\"left\">VGA00019</th>\n",
        "      <th colspan=\"2\" halign=\"left\">VGA00021</th>\n",
        "    </tr>\n",
        "    <tr>\n",
        "      <th></th>\n",
        "      <th>0</th>\n",
        "      <th>1</th>\n",
        "      <th>0</th>\n",
        "      <th>1</th>\n",
        "      <th>0</th>\n",
        "      <th>1</th>\n",
        "      <th>0</th>\n",
        "      <th>1</th>\n",
        "      <th>0</th>\n",
        "      <th>1</th>\n",
        "    </tr>\n",
        "  </thead>\n",
        "  <tbody>\n",
        "    <tr>\n",
        "      <th>1879</th>\n",
        "      <td> None</td>\n",
        "      <td> None</td>\n",
        "      <td> None</td>\n",
        "      <td> None</td>\n",
        "      <td> 0</td>\n",
        "      <td> 0</td>\n",
        "      <td> None</td>\n",
        "      <td> None</td>\n",
        "      <td> 0</td>\n",
        "      <td> 0</td>\n",
        "    </tr>\n",
        "    <tr>\n",
        "      <th>1915</th>\n",
        "      <td> None</td>\n",
        "      <td> None</td>\n",
        "      <td> None</td>\n",
        "      <td> None</td>\n",
        "      <td> 0</td>\n",
        "      <td> 0</td>\n",
        "      <td> None</td>\n",
        "      <td> None</td>\n",
        "      <td> 0</td>\n",
        "      <td> 0</td>\n",
        "    </tr>\n",
        "    <tr>\n",
        "      <th>3692</th>\n",
        "      <td> None</td>\n",
        "      <td> None</td>\n",
        "      <td> None</td>\n",
        "      <td> None</td>\n",
        "      <td> 0</td>\n",
        "      <td> 0</td>\n",
        "      <td> None</td>\n",
        "      <td> None</td>\n",
        "      <td> 0</td>\n",
        "      <td> 0</td>\n",
        "    </tr>\n",
        "    <tr>\n",
        "      <th>4404</th>\n",
        "      <td> None</td>\n",
        "      <td> None</td>\n",
        "      <td> None</td>\n",
        "      <td> None</td>\n",
        "      <td> 0</td>\n",
        "      <td> 0</td>\n",
        "      <td> None</td>\n",
        "      <td> None</td>\n",
        "      <td> 0</td>\n",
        "      <td> 0</td>\n",
        "    </tr>\n",
        "    <tr>\n",
        "      <th>4531</th>\n",
        "      <td> None</td>\n",
        "      <td> None</td>\n",
        "      <td> None</td>\n",
        "      <td> None</td>\n",
        "      <td> 0</td>\n",
        "      <td> 0</td>\n",
        "      <td> None</td>\n",
        "      <td> None</td>\n",
        "      <td> 0</td>\n",
        "      <td> 0</td>\n",
        "    </tr>\n",
        "    <tr>\n",
        "      <th>4623</th>\n",
        "      <td> None</td>\n",
        "      <td> None</td>\n",
        "      <td> None</td>\n",
        "      <td> None</td>\n",
        "      <td> 0</td>\n",
        "      <td> 0</td>\n",
        "      <td> None</td>\n",
        "      <td> None</td>\n",
        "      <td> 0</td>\n",
        "      <td> 0</td>\n",
        "    </tr>\n",
        "    <tr>\n",
        "      <th>5167</th>\n",
        "      <td> None</td>\n",
        "      <td> None</td>\n",
        "      <td> None</td>\n",
        "      <td> None</td>\n",
        "      <td> 0</td>\n",
        "      <td> 0</td>\n",
        "      <td> None</td>\n",
        "      <td> None</td>\n",
        "      <td> 0</td>\n",
        "      <td> 0</td>\n",
        "    </tr>\n",
        "    <tr>\n",
        "      <th>5560</th>\n",
        "      <td> None</td>\n",
        "      <td> None</td>\n",
        "      <td> None</td>\n",
        "      <td> None</td>\n",
        "      <td> 0</td>\n",
        "      <td> 0</td>\n",
        "      <td> None</td>\n",
        "      <td> None</td>\n",
        "      <td> 0</td>\n",
        "      <td> 0</td>\n",
        "    </tr>\n",
        "    <tr>\n",
        "      <th>5676</th>\n",
        "      <td> None</td>\n",
        "      <td> None</td>\n",
        "      <td> None</td>\n",
        "      <td> None</td>\n",
        "      <td> 0</td>\n",
        "      <td> 0</td>\n",
        "      <td> None</td>\n",
        "      <td> None</td>\n",
        "      <td> 0</td>\n",
        "      <td> 0</td>\n",
        "    </tr>\n",
        "    <tr>\n",
        "      <th>5728</th>\n",
        "      <td> None</td>\n",
        "      <td> None</td>\n",
        "      <td> None</td>\n",
        "      <td> None</td>\n",
        "      <td> 0</td>\n",
        "      <td> 0</td>\n",
        "      <td> None</td>\n",
        "      <td> None</td>\n",
        "      <td> 0</td>\n",
        "      <td> 0</td>\n",
        "    </tr>\n",
        "  </tbody>\n",
        "</table>\n",
        "</div>"
       ],
       "output_type": "pyout",
       "prompt_number": 94,
       "text": [
        "     VGA00002       VGA00006        VGA00010    VGA00019        VGA00021   \n",
        "            0     1        0     1         0  1        0     1         0  1\n",
        "1879     None  None     None  None         0  0     None  None         0  0\n",
        "1915     None  None     None  None         0  0     None  None         0  0\n",
        "3692     None  None     None  None         0  0     None  None         0  0\n",
        "4404     None  None     None  None         0  0     None  None         0  0\n",
        "4531     None  None     None  None         0  0     None  None         0  0\n",
        "4623     None  None     None  None         0  0     None  None         0  0\n",
        "5167     None  None     None  None         0  0     None  None         0  0\n",
        "5560     None  None     None  None         0  0     None  None         0  0\n",
        "5676     None  None     None  None         0  0     None  None         0  0\n",
        "5728     None  None     None  None         0  0     None  None         0  0"
       ]
      }
     ],
     "prompt_number": 94
    },
    {
     "cell_type": "code",
     "collapsed": false,
     "input": [
      "met.iloc[:,:]"
     ],
     "language": "python",
     "metadata": {},
     "outputs": [
      {
       "html": [
        "<div style=\"max-height:1000px;max-width:1500px;overflow:auto;\">\n",
        "<table border=\"1\" class=\"dataframe\">\n",
        "  <thead>\n",
        "    <tr style=\"text-align: right;\">\n",
        "      <th></th>\n",
        "      <th>CHROM</th>\n",
        "      <th>POS</th>\n",
        "      <th>ID</th>\n",
        "      <th>REF</th>\n",
        "      <th>ALT</th>\n",
        "    </tr>\n",
        "  </thead>\n",
        "  <tbody>\n",
        "    <tr>\n",
        "      <th>1879</th>\n",
        "      <td> CAE1</td>\n",
        "      <td> 1879</td>\n",
        "      <td> var_1940192</td>\n",
        "      <td> A</td>\n",
        "      <td> [G]</td>\n",
        "    </tr>\n",
        "    <tr>\n",
        "      <th>1915</th>\n",
        "      <td> CAE1</td>\n",
        "      <td> 1915</td>\n",
        "      <td> var_3019303</td>\n",
        "      <td> T</td>\n",
        "      <td> [C]</td>\n",
        "    </tr>\n",
        "    <tr>\n",
        "      <th>3692</th>\n",
        "      <td> CAE1</td>\n",
        "      <td> 3692</td>\n",
        "      <td> var_4761202</td>\n",
        "      <td> A</td>\n",
        "      <td> [G]</td>\n",
        "    </tr>\n",
        "    <tr>\n",
        "      <th>4404</th>\n",
        "      <td> CAE1</td>\n",
        "      <td> 4404</td>\n",
        "      <td> var_7643431</td>\n",
        "      <td> C</td>\n",
        "      <td> [T]</td>\n",
        "    </tr>\n",
        "    <tr>\n",
        "      <th>4531</th>\n",
        "      <td> CAE1</td>\n",
        "      <td> 4531</td>\n",
        "      <td> var_1814618</td>\n",
        "      <td> A</td>\n",
        "      <td> [G]</td>\n",
        "    </tr>\n",
        "    <tr>\n",
        "      <th>4623</th>\n",
        "      <td> CAE1</td>\n",
        "      <td> 4623</td>\n",
        "      <td> var_5844884</td>\n",
        "      <td> C</td>\n",
        "      <td> [T]</td>\n",
        "    </tr>\n",
        "    <tr>\n",
        "      <th>5167</th>\n",
        "      <td> CAE1</td>\n",
        "      <td> 5167</td>\n",
        "      <td> var_2995012</td>\n",
        "      <td> C</td>\n",
        "      <td> [T]</td>\n",
        "    </tr>\n",
        "    <tr>\n",
        "      <th>5560</th>\n",
        "      <td> CAE1</td>\n",
        "      <td> 5560</td>\n",
        "      <td>  var_142740</td>\n",
        "      <td> G</td>\n",
        "      <td> [A]</td>\n",
        "    </tr>\n",
        "    <tr>\n",
        "      <th>5676</th>\n",
        "      <td> CAE1</td>\n",
        "      <td> 5676</td>\n",
        "      <td> var_5163837</td>\n",
        "      <td> G</td>\n",
        "      <td> [A]</td>\n",
        "    </tr>\n",
        "    <tr>\n",
        "      <th>5728</th>\n",
        "      <td> CAE1</td>\n",
        "      <td> 5728</td>\n",
        "      <td> var_8422988</td>\n",
        "      <td> C</td>\n",
        "      <td> [T]</td>\n",
        "    </tr>\n",
        "    <tr>\n",
        "      <th>5857</th>\n",
        "      <td> CAE1</td>\n",
        "      <td> 5857</td>\n",
        "      <td>  var_464894</td>\n",
        "      <td> T</td>\n",
        "      <td> [C]</td>\n",
        "    </tr>\n",
        "    <tr>\n",
        "      <th>6564</th>\n",
        "      <td> CAE1</td>\n",
        "      <td> 6564</td>\n",
        "      <td> var_4432088</td>\n",
        "      <td> G</td>\n",
        "      <td> [A]</td>\n",
        "    </tr>\n",
        "    <tr>\n",
        "      <th>6762</th>\n",
        "      <td> CAE1</td>\n",
        "      <td> 6762</td>\n",
        "      <td> var_5503060</td>\n",
        "      <td> T</td>\n",
        "      <td> [C]</td>\n",
        "    </tr>\n",
        "    <tr>\n",
        "      <th>7478</th>\n",
        "      <td> CAE1</td>\n",
        "      <td> 7478</td>\n",
        "      <td> var_6877828</td>\n",
        "      <td> G</td>\n",
        "      <td> [A]</td>\n",
        "    </tr>\n",
        "    <tr>\n",
        "      <th>8944</th>\n",
        "      <td> CAE1</td>\n",
        "      <td> 8944</td>\n",
        "      <td> var_8436988</td>\n",
        "      <td> T</td>\n",
        "      <td> [C]</td>\n",
        "    </tr>\n",
        "    <tr>\n",
        "      <th>9055</th>\n",
        "      <td> CAE1</td>\n",
        "      <td> 9055</td>\n",
        "      <td>  var_676815</td>\n",
        "      <td> G</td>\n",
        "      <td> [A]</td>\n",
        "    </tr>\n",
        "    <tr>\n",
        "      <th>9671</th>\n",
        "      <td> CAE1</td>\n",
        "      <td> 9671</td>\n",
        "      <td> var_2042948</td>\n",
        "      <td> G</td>\n",
        "      <td> [T]</td>\n",
        "    </tr>\n",
        "    <tr>\n",
        "      <th>9894</th>\n",
        "      <td> CAE1</td>\n",
        "      <td> 9894</td>\n",
        "      <td> var_7368180</td>\n",
        "      <td> G</td>\n",
        "      <td> [A]</td>\n",
        "    </tr>\n",
        "  </tbody>\n",
        "</table>\n",
        "</div>"
       ],
       "output_type": "pyout",
       "prompt_number": 99,
       "text": [
        "     CHROM   POS           ID REF  ALT\n",
        "1879  CAE1  1879  var_1940192   A  [G]\n",
        "1915  CAE1  1915  var_3019303   T  [C]\n",
        "3692  CAE1  3692  var_4761202   A  [G]\n",
        "4404  CAE1  4404  var_7643431   C  [T]\n",
        "4531  CAE1  4531  var_1814618   A  [G]\n",
        "4623  CAE1  4623  var_5844884   C  [T]\n",
        "5167  CAE1  5167  var_2995012   C  [T]\n",
        "5560  CAE1  5560   var_142740   G  [A]\n",
        "5676  CAE1  5676  var_5163837   G  [A]\n",
        "5728  CAE1  5728  var_8422988   C  [T]\n",
        "5857  CAE1  5857   var_464894   T  [C]\n",
        "6564  CAE1  6564  var_4432088   G  [A]\n",
        "6762  CAE1  6762  var_5503060   T  [C]\n",
        "7478  CAE1  7478  var_6877828   G  [A]\n",
        "8944  CAE1  8944  var_8436988   T  [C]\n",
        "9055  CAE1  9055   var_676815   G  [A]\n",
        "9671  CAE1  9671  var_2042948   G  [T]\n",
        "9894  CAE1  9894  var_7368180   G  [A]"
       ]
      }
     ],
     "prompt_number": 99
    },
    {
     "cell_type": "code",
     "collapsed": false,
     "input": [],
     "language": "python",
     "metadata": {},
     "outputs": [],
     "prompt_number": 100
    },
    {
     "cell_type": "code",
     "collapsed": false,
     "input": [
      "hvb.save_genotype(dat,tsv_basename'test_12.tsv')"
     ],
     "language": "python",
     "metadata": {},
     "outputs": [],
     "prompt_number": 101
    },
    {
     "cell_type": "code",
     "collapsed": false,
     "input": [
      "dat2=hvb.read_genotype('test_12.tsv')"
     ],
     "language": "python",
     "metadata": {},
     "outputs": [
      {
       "ename": "IOError",
       "evalue": "File test_12.tsv does not exist",
       "output_type": "pyerr",
       "traceback": [
        "\u001b[1;31m---------------------------------------------------------------------------\u001b[0m\n\u001b[1;31mIOError\u001b[0m                                   Traceback (most recent call last)",
        "\u001b[1;32m<ipython-input-3-5281873b0540>\u001b[0m in \u001b[0;36m<module>\u001b[1;34m()\u001b[0m\n\u001b[1;32m----> 1\u001b[1;33m \u001b[0mdat2\u001b[0m\u001b[1;33m=\u001b[0m\u001b[0mhvb\u001b[0m\u001b[1;33m.\u001b[0m\u001b[0mread_genotype\u001b[0m\u001b[1;33m(\u001b[0m\u001b[1;34m'test_12.tsv'\u001b[0m\u001b[1;33m)\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n\u001b[0m",
        "\u001b[1;32m/home/GMI/hannes.svardal/script/hs_vervet/tools/hs_vervet_basics.pyc\u001b[0m in \u001b[0;36mread_genotype\u001b[1;34m(fname, *args, **kwargs)\u001b[0m\n\u001b[0;32m     47\u001b[0m \u001b[1;32mdef\u001b[0m \u001b[0mread_genotype\u001b[0m\u001b[1;33m(\u001b[0m\u001b[0mfname\u001b[0m\u001b[1;33m,\u001b[0m\u001b[1;33m*\u001b[0m\u001b[0margs\u001b[0m\u001b[1;33m,\u001b[0m\u001b[1;33m**\u001b[0m\u001b[0mkwargs\u001b[0m\u001b[1;33m)\u001b[0m\u001b[1;33m:\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n\u001b[0;32m     48\u001b[0m     \u001b[1;32mimport\u001b[0m \u001b[0mpandas\u001b[0m \u001b[1;32mas\u001b[0m \u001b[0mpd\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n\u001b[1;32m---> 49\u001b[1;33m     \u001b[0mdf1\u001b[0m\u001b[1;33m=\u001b[0m\u001b[0mpd\u001b[0m\u001b[1;33m.\u001b[0m\u001b[0mread_csv\u001b[0m\u001b[1;33m(\u001b[0m\u001b[0mfname\u001b[0m\u001b[1;33m,\u001b[0m\u001b[0msep\u001b[0m\u001b[1;33m=\u001b[0m\u001b[1;34m'\\t'\u001b[0m\u001b[1;33m,\u001b[0m\u001b[0mindex_col\u001b[0m\u001b[1;33m=\u001b[0m\u001b[1;36m0\u001b[0m\u001b[1;33m,\u001b[0m\u001b[0mna_values\u001b[0m\u001b[1;33m=\u001b[0m\u001b[1;34m'N'\u001b[0m\u001b[1;33m,\u001b[0m\u001b[1;33m*\u001b[0m\u001b[0margs\u001b[0m\u001b[1;33m,\u001b[0m\u001b[1;33m**\u001b[0m\u001b[0mkwargs\u001b[0m\u001b[1;33m)\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n\u001b[0m\u001b[0;32m     50\u001b[0m     df1.columns=pd.MultiIndex.from_tuples(df1.columns.map(literal_eval), \n\u001b[0;32m     51\u001b[0m                                               names=['ucla_id','haplotype'])\n",
        "\u001b[1;32m/net/gmi.oeaw.ac.at/nordborg/user/hannes.svardal/Downloads/pandas-0.12.0/pandas/io/parsers.pyc\u001b[0m in \u001b[0;36mparser_f\u001b[1;34m(filepath_or_buffer, sep, dialect, compression, doublequote, escapechar, quotechar, quoting, skipinitialspace, lineterminator, header, index_col, names, prefix, skiprows, skipfooter, skip_footer, na_values, na_fvalues, true_values, false_values, delimiter, converters, dtype, usecols, engine, delim_whitespace, as_recarray, na_filter, compact_ints, use_unsigned, low_memory, buffer_lines, warn_bad_lines, error_bad_lines, keep_default_na, thousands, comment, decimal, parse_dates, keep_date_col, dayfirst, date_parser, memory_map, nrows, iterator, chunksize, verbose, encoding, squeeze, mangle_dupe_cols, tupleize_cols)\u001b[0m\n\u001b[0;32m    398\u001b[0m             )\n\u001b[0;32m    399\u001b[0m \u001b[1;33m\u001b[0m\u001b[0m\n\u001b[1;32m--> 400\u001b[1;33m         \u001b[1;32mreturn\u001b[0m \u001b[0m_read\u001b[0m\u001b[1;33m(\u001b[0m\u001b[0mfilepath_or_buffer\u001b[0m\u001b[1;33m,\u001b[0m \u001b[0mkwds\u001b[0m\u001b[1;33m)\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n\u001b[0m\u001b[0;32m    401\u001b[0m \u001b[1;33m\u001b[0m\u001b[0m\n\u001b[0;32m    402\u001b[0m     \u001b[0mparser_f\u001b[0m\u001b[1;33m.\u001b[0m\u001b[0m__name__\u001b[0m \u001b[1;33m=\u001b[0m \u001b[0mname\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n",
        "\u001b[1;32m/net/gmi.oeaw.ac.at/nordborg/user/hannes.svardal/Downloads/pandas-0.12.0/pandas/io/parsers.pyc\u001b[0m in \u001b[0;36m_read\u001b[1;34m(filepath_or_buffer, kwds)\u001b[0m\n\u001b[0;32m    196\u001b[0m \u001b[1;33m\u001b[0m\u001b[0m\n\u001b[0;32m    197\u001b[0m     \u001b[1;31m# Create the parser.\u001b[0m\u001b[1;33m\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n\u001b[1;32m--> 198\u001b[1;33m     \u001b[0mparser\u001b[0m \u001b[1;33m=\u001b[0m \u001b[0mTextFileReader\u001b[0m\u001b[1;33m(\u001b[0m\u001b[0mfilepath_or_buffer\u001b[0m\u001b[1;33m,\u001b[0m \u001b[1;33m**\u001b[0m\u001b[0mkwds\u001b[0m\u001b[1;33m)\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n\u001b[0m\u001b[0;32m    199\u001b[0m \u001b[1;33m\u001b[0m\u001b[0m\n\u001b[0;32m    200\u001b[0m     \u001b[1;32mif\u001b[0m \u001b[0mnrows\u001b[0m \u001b[1;32mis\u001b[0m \u001b[1;32mnot\u001b[0m \u001b[0mNone\u001b[0m\u001b[1;33m:\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n",
        "\u001b[1;32m/net/gmi.oeaw.ac.at/nordborg/user/hannes.svardal/Downloads/pandas-0.12.0/pandas/io/parsers.pyc\u001b[0m in \u001b[0;36m__init__\u001b[1;34m(self, f, engine, **kwds)\u001b[0m\n\u001b[0;32m    477\u001b[0m             \u001b[0mself\u001b[0m\u001b[1;33m.\u001b[0m\u001b[0moptions\u001b[0m\u001b[1;33m[\u001b[0m\u001b[1;34m'has_index_names'\u001b[0m\u001b[1;33m]\u001b[0m \u001b[1;33m=\u001b[0m \u001b[0mkwds\u001b[0m\u001b[1;33m[\u001b[0m\u001b[1;34m'has_index_names'\u001b[0m\u001b[1;33m]\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n\u001b[0;32m    478\u001b[0m \u001b[1;33m\u001b[0m\u001b[0m\n\u001b[1;32m--> 479\u001b[1;33m         \u001b[0mself\u001b[0m\u001b[1;33m.\u001b[0m\u001b[0m_make_engine\u001b[0m\u001b[1;33m(\u001b[0m\u001b[0mself\u001b[0m\u001b[1;33m.\u001b[0m\u001b[0mengine\u001b[0m\u001b[1;33m)\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n\u001b[0m\u001b[0;32m    480\u001b[0m \u001b[1;33m\u001b[0m\u001b[0m\n\u001b[0;32m    481\u001b[0m     \u001b[1;32mdef\u001b[0m \u001b[0m_get_options_with_defaults\u001b[0m\u001b[1;33m(\u001b[0m\u001b[0mself\u001b[0m\u001b[1;33m,\u001b[0m \u001b[0mengine\u001b[0m\u001b[1;33m)\u001b[0m\u001b[1;33m:\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n",
        "\u001b[1;32m/net/gmi.oeaw.ac.at/nordborg/user/hannes.svardal/Downloads/pandas-0.12.0/pandas/io/parsers.pyc\u001b[0m in \u001b[0;36m_make_engine\u001b[1;34m(self, engine)\u001b[0m\n\u001b[0;32m    584\u001b[0m     \u001b[1;32mdef\u001b[0m \u001b[0m_make_engine\u001b[0m\u001b[1;33m(\u001b[0m\u001b[0mself\u001b[0m\u001b[1;33m,\u001b[0m \u001b[0mengine\u001b[0m\u001b[1;33m=\u001b[0m\u001b[1;34m'c'\u001b[0m\u001b[1;33m)\u001b[0m\u001b[1;33m:\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n\u001b[0;32m    585\u001b[0m         \u001b[1;32mif\u001b[0m \u001b[0mengine\u001b[0m \u001b[1;33m==\u001b[0m \u001b[1;34m'c'\u001b[0m\u001b[1;33m:\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n\u001b[1;32m--> 586\u001b[1;33m             \u001b[0mself\u001b[0m\u001b[1;33m.\u001b[0m\u001b[0m_engine\u001b[0m \u001b[1;33m=\u001b[0m \u001b[0mCParserWrapper\u001b[0m\u001b[1;33m(\u001b[0m\u001b[0mself\u001b[0m\u001b[1;33m.\u001b[0m\u001b[0mf\u001b[0m\u001b[1;33m,\u001b[0m \u001b[1;33m**\u001b[0m\u001b[0mself\u001b[0m\u001b[1;33m.\u001b[0m\u001b[0moptions\u001b[0m\u001b[1;33m)\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n\u001b[0m\u001b[0;32m    587\u001b[0m         \u001b[1;32melse\u001b[0m\u001b[1;33m:\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n\u001b[0;32m    588\u001b[0m             \u001b[1;32mif\u001b[0m \u001b[0mengine\u001b[0m \u001b[1;33m==\u001b[0m \u001b[1;34m'python'\u001b[0m\u001b[1;33m:\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n",
        "\u001b[1;32m/net/gmi.oeaw.ac.at/nordborg/user/hannes.svardal/Downloads/pandas-0.12.0/pandas/io/parsers.pyc\u001b[0m in \u001b[0;36m__init__\u001b[1;34m(self, src, **kwds)\u001b[0m\n\u001b[0;32m    955\u001b[0m         \u001b[0mkwds\u001b[0m\u001b[1;33m[\u001b[0m\u001b[1;34m'allow_leading_cols'\u001b[0m\u001b[1;33m]\u001b[0m \u001b[1;33m=\u001b[0m \u001b[0mself\u001b[0m\u001b[1;33m.\u001b[0m\u001b[0mindex_col\u001b[0m \u001b[1;32mis\u001b[0m \u001b[1;32mnot\u001b[0m \u001b[0mFalse\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n\u001b[0;32m    956\u001b[0m \u001b[1;33m\u001b[0m\u001b[0m\n\u001b[1;32m--> 957\u001b[1;33m         \u001b[0mself\u001b[0m\u001b[1;33m.\u001b[0m\u001b[0m_reader\u001b[0m \u001b[1;33m=\u001b[0m \u001b[0m_parser\u001b[0m\u001b[1;33m.\u001b[0m\u001b[0mTextReader\u001b[0m\u001b[1;33m(\u001b[0m\u001b[0msrc\u001b[0m\u001b[1;33m,\u001b[0m \u001b[1;33m**\u001b[0m\u001b[0mkwds\u001b[0m\u001b[1;33m)\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n\u001b[0m\u001b[0;32m    958\u001b[0m \u001b[1;33m\u001b[0m\u001b[0m\n\u001b[0;32m    959\u001b[0m         \u001b[1;31m# XXX\u001b[0m\u001b[1;33m\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n",
        "\u001b[1;32m/net/gmi.oeaw.ac.at/nordborg/user/hannes.svardal/Downloads/pandas-0.12.0/pandas/parser.so\u001b[0m in \u001b[0;36mpandas.parser.TextReader.__cinit__ (pandas/parser.c:2987)\u001b[1;34m()\u001b[0m\n",
        "\u001b[1;32m/net/gmi.oeaw.ac.at/nordborg/user/hannes.svardal/Downloads/pandas-0.12.0/pandas/parser.so\u001b[0m in \u001b[0;36mpandas.parser.TextReader._setup_parser_source (pandas/parser.c:5345)\u001b[1;34m()\u001b[0m\n",
        "\u001b[1;31mIOError\u001b[0m: File test_12.tsv does not exist"
       ]
      }
     ],
     "prompt_number": 3
    },
    {
     "cell_type": "code",
     "collapsed": false,
     "input": [
      "dat2.iloc[:10,:10]"
     ],
     "language": "python",
     "metadata": {},
     "outputs": [
      {
       "html": [
        "<div style=\"max-height:1000px;max-width:1500px;overflow:auto;\">\n",
        "<table border=\"1\" class=\"dataframe\">\n",
        "  <thead>\n",
        "    <tr>\n",
        "      <th>ucla_id</th>\n",
        "      <th colspan=\"2\" halign=\"left\">VGA00002</th>\n",
        "      <th colspan=\"2\" halign=\"left\">VGA00006</th>\n",
        "      <th colspan=\"2\" halign=\"left\">VGA00010</th>\n",
        "      <th colspan=\"2\" halign=\"left\">VGA00019</th>\n",
        "      <th colspan=\"2\" halign=\"left\">VGA00021</th>\n",
        "    </tr>\n",
        "    <tr>\n",
        "      <th>haplotype</th>\n",
        "      <th>0</th>\n",
        "      <th>1</th>\n",
        "      <th>0</th>\n",
        "      <th>1</th>\n",
        "      <th>0</th>\n",
        "      <th>1</th>\n",
        "      <th>0</th>\n",
        "      <th>1</th>\n",
        "      <th>0</th>\n",
        "      <th>1</th>\n",
        "    </tr>\n",
        "    <tr>\n",
        "      <th>snp_pos</th>\n",
        "      <th></th>\n",
        "      <th></th>\n",
        "      <th></th>\n",
        "      <th></th>\n",
        "      <th></th>\n",
        "      <th></th>\n",
        "      <th></th>\n",
        "      <th></th>\n",
        "      <th></th>\n",
        "      <th></th>\n",
        "    </tr>\n",
        "  </thead>\n",
        "  <tbody>\n",
        "    <tr>\n",
        "      <th>1879</th>\n",
        "      <td>NaN</td>\n",
        "      <td>NaN</td>\n",
        "      <td>NaN</td>\n",
        "      <td>NaN</td>\n",
        "      <td> 0</td>\n",
        "      <td> 0</td>\n",
        "      <td>NaN</td>\n",
        "      <td>NaN</td>\n",
        "      <td> 0</td>\n",
        "      <td> 0</td>\n",
        "    </tr>\n",
        "    <tr>\n",
        "      <th>1915</th>\n",
        "      <td>NaN</td>\n",
        "      <td>NaN</td>\n",
        "      <td>NaN</td>\n",
        "      <td>NaN</td>\n",
        "      <td> 0</td>\n",
        "      <td> 0</td>\n",
        "      <td>NaN</td>\n",
        "      <td>NaN</td>\n",
        "      <td> 0</td>\n",
        "      <td> 0</td>\n",
        "    </tr>\n",
        "    <tr>\n",
        "      <th>3692</th>\n",
        "      <td>NaN</td>\n",
        "      <td>NaN</td>\n",
        "      <td>NaN</td>\n",
        "      <td>NaN</td>\n",
        "      <td> 0</td>\n",
        "      <td> 0</td>\n",
        "      <td>NaN</td>\n",
        "      <td>NaN</td>\n",
        "      <td> 0</td>\n",
        "      <td> 0</td>\n",
        "    </tr>\n",
        "    <tr>\n",
        "      <th>4404</th>\n",
        "      <td>NaN</td>\n",
        "      <td>NaN</td>\n",
        "      <td>NaN</td>\n",
        "      <td>NaN</td>\n",
        "      <td> 0</td>\n",
        "      <td> 0</td>\n",
        "      <td>NaN</td>\n",
        "      <td>NaN</td>\n",
        "      <td> 0</td>\n",
        "      <td> 0</td>\n",
        "    </tr>\n",
        "    <tr>\n",
        "      <th>4531</th>\n",
        "      <td>NaN</td>\n",
        "      <td>NaN</td>\n",
        "      <td>NaN</td>\n",
        "      <td>NaN</td>\n",
        "      <td> 0</td>\n",
        "      <td> 0</td>\n",
        "      <td>NaN</td>\n",
        "      <td>NaN</td>\n",
        "      <td> 0</td>\n",
        "      <td> 0</td>\n",
        "    </tr>\n",
        "    <tr>\n",
        "      <th>4623</th>\n",
        "      <td>NaN</td>\n",
        "      <td>NaN</td>\n",
        "      <td>NaN</td>\n",
        "      <td>NaN</td>\n",
        "      <td> 0</td>\n",
        "      <td> 0</td>\n",
        "      <td>NaN</td>\n",
        "      <td>NaN</td>\n",
        "      <td> 0</td>\n",
        "      <td> 0</td>\n",
        "    </tr>\n",
        "    <tr>\n",
        "      <th>5167</th>\n",
        "      <td>NaN</td>\n",
        "      <td>NaN</td>\n",
        "      <td>NaN</td>\n",
        "      <td>NaN</td>\n",
        "      <td> 0</td>\n",
        "      <td> 0</td>\n",
        "      <td>NaN</td>\n",
        "      <td>NaN</td>\n",
        "      <td> 0</td>\n",
        "      <td> 0</td>\n",
        "    </tr>\n",
        "    <tr>\n",
        "      <th>5560</th>\n",
        "      <td>NaN</td>\n",
        "      <td>NaN</td>\n",
        "      <td>NaN</td>\n",
        "      <td>NaN</td>\n",
        "      <td> 0</td>\n",
        "      <td> 0</td>\n",
        "      <td>NaN</td>\n",
        "      <td>NaN</td>\n",
        "      <td> 0</td>\n",
        "      <td> 0</td>\n",
        "    </tr>\n",
        "    <tr>\n",
        "      <th>5676</th>\n",
        "      <td>NaN</td>\n",
        "      <td>NaN</td>\n",
        "      <td>NaN</td>\n",
        "      <td>NaN</td>\n",
        "      <td> 0</td>\n",
        "      <td> 0</td>\n",
        "      <td>NaN</td>\n",
        "      <td>NaN</td>\n",
        "      <td> 0</td>\n",
        "      <td> 0</td>\n",
        "    </tr>\n",
        "    <tr>\n",
        "      <th>5728</th>\n",
        "      <td>NaN</td>\n",
        "      <td>NaN</td>\n",
        "      <td>NaN</td>\n",
        "      <td>NaN</td>\n",
        "      <td> 0</td>\n",
        "      <td> 0</td>\n",
        "      <td>NaN</td>\n",
        "      <td>NaN</td>\n",
        "      <td> 0</td>\n",
        "      <td> 0</td>\n",
        "    </tr>\n",
        "  </tbody>\n",
        "</table>\n",
        "</div>"
       ],
       "output_type": "pyout",
       "prompt_number": 103,
       "text": [
        "ucla_id    VGA00002      VGA00006      VGA00010     VGA00019      VGA00021   \n",
        "haplotype         0   1         0   1         0  1         0   1         0  1\n",
        "snp_pos                                                                      \n",
        "1879            NaN NaN       NaN NaN         0  0       NaN NaN         0  0\n",
        "1915            NaN NaN       NaN NaN         0  0       NaN NaN         0  0\n",
        "3692            NaN NaN       NaN NaN         0  0       NaN NaN         0  0\n",
        "4404            NaN NaN       NaN NaN         0  0       NaN NaN         0  0\n",
        "4531            NaN NaN       NaN NaN         0  0       NaN NaN         0  0\n",
        "4623            NaN NaN       NaN NaN         0  0       NaN NaN         0  0\n",
        "5167            NaN NaN       NaN NaN         0  0       NaN NaN         0  0\n",
        "5560            NaN NaN       NaN NaN         0  0       NaN NaN         0  0\n",
        "5676            NaN NaN       NaN NaN         0  0       NaN NaN         0  0\n",
        "5728            NaN NaN       NaN NaN         0  0       NaN NaN         0  0"
       ]
      }
     ],
     "prompt_number": 103
    },
    {
     "cell_type": "code",
     "collapsed": false,
     "input": [],
     "language": "python",
     "metadata": {},
     "outputs": []
    }
   ],
   "metadata": {}
  }
 ]
}